// lodash/core.js
/**
 * @license
 * Lodash (Custom Build) <https://lodash.com/>
 * Build: `lodash core -o ./dist/lodash.core.js`
 * Copyright JS Foundation and other contributors <https://js.foundation/>
 * Released under MIT license <https://lodash.com/license>
 * Based on Underscore.js 1.8.3 <http://underscorejs.org/LICENSE>
 * Copyright Jeremy Ashkenas, DocumentCloud and Investigative Reporters & Editors
 */
;(function() ***REMOVED***

  /** Used as a safe reference for `undefined` in pre-ES5 environments. */
  var undefined;

  /** Used as the semantic version number. */
  var VERSION = '4.17.4';

  /** Error message constants. */
  var FUNC_ERROR_TEXT = 'Expected a function';

  /** Used to compose bitmasks for value comparisons. */
  var COMPARE_PARTIAL_FLAG = 1,
      COMPARE_UNORDERED_FLAG = 2;

  /** Used to compose bitmasks for function metadata. */
  var WRAP_BIND_FLAG = 1,
      WRAP_PARTIAL_FLAG = 32;

  /** Used as references for various `Number` constants. */
  var INFINITY = 1 / 0,
      MAX_SAFE_INTEGER = 9007199254740991;

  /** `Object#toString` result references. */
  var argsTag = '[object Arguments]',
      arrayTag = '[object Array]',
      asyncTag = '[object AsyncFunction]',
      boolTag = '[object Boolean]',
      dateTag = '[object Date]',
      errorTag = '[object Error]',
      funcTag = '[object Function]',
      genTag = '[object GeneratorFunction]',
      numberTag = '[object Number]',
      objectTag = '[object Object]',
      proxyTag = '[object Proxy]',
      regexpTag = '[object RegExp]',
      stringTag = '[object String]';

  /** Used to match HTML entities and HTML characters. */
  var reUnescapedHtml = /[&<>"']/g,
      reHasUnescapedHtml = RegExp(reUnescapedHtml.source);

  /** Used to map characters to HTML entities. */
  var htmlEscapes = ***REMOVED***
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#39;'
  ***REMOVED***;

  /** Detect free variable `global` from Node.js. */
  var freeGlobal = typeof global == 'object' && global && global.Object === Object && global;

  /** Detect free variable `self`. */
  var freeSelf = typeof self == 'object' && self && self.Object === Object && self;

  /** Used as a reference to the global object. */
  var root = freeGlobal || freeSelf || Function('return this')();

  /** Detect free variable `exports`. */
  var freeExports = typeof exports == 'object' && exports && !exports.nodeType && exports;

  /** Detect free variable `module`. */
  var freeModule = freeExports && typeof module == 'object' && module && !module.nodeType && module;

  /*--------------------------------------------------------------------------*/

  /**
   * Appends the elements of `values` to `array`.
   *
   * @private
   * @param ***REMOVED***Array***REMOVED*** array The array to modify.
   * @param ***REMOVED***Array***REMOVED*** values The values to append.
   * @returns ***REMOVED***Array***REMOVED*** Returns `array`.
   */
  function arrayPush(array, values) ***REMOVED***
    array.push.apply(array, values);
    return array;
  ***REMOVED***

  /**
   * The base implementation of `_.findIndex` and `_.findLastIndex` without
   * support for iteratee shorthands.
   *
   * @private
   * @param ***REMOVED***Array***REMOVED*** array The array to inspect.
   * @param ***REMOVED***Function***REMOVED*** predicate The function invoked per iteration.
   * @param ***REMOVED***number***REMOVED*** fromIndex The index to search from.
   * @param ***REMOVED***boolean***REMOVED*** [fromRight] Specify iterating from right to left.
   * @returns ***REMOVED***number***REMOVED*** Returns the index of the matched value, else `-1`.
   */
  function baseFindIndex(array, predicate, fromIndex, fromRight) ***REMOVED***
    var length = array.length,
        index = fromIndex + (fromRight ? 1 : -1);

    while ((fromRight ? index-- : ++index < length)) ***REMOVED***
      if (predicate(array[index], index, array)) ***REMOVED***
        return index;
      ***REMOVED***
    ***REMOVED***
    return -1;
  ***REMOVED***

  /**
   * The base implementation of `_.property` without support for deep paths.
   *
   * @private
   * @param ***REMOVED***string***REMOVED*** key The key of the property to get.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new accessor function.
   */
  function baseProperty(key) ***REMOVED***
    return function(object) ***REMOVED***
      return object == null ? undefined : object[key];
    ***REMOVED***;
  ***REMOVED***

  /**
   * The base implementation of `_.propertyOf` without support for deep paths.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to query.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new accessor function.
   */
  function basePropertyOf(object) ***REMOVED***
    return function(key) ***REMOVED***
      return object == null ? undefined : object[key];
    ***REMOVED***;
  ***REMOVED***

  /**
   * The base implementation of `_.reduce` and `_.reduceRight`, without support
   * for iteratee shorthands, which iterates over `collection` using `eachFunc`.
   *
   * @private
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** iteratee The function invoked per iteration.
   * @param ***REMOVED*******REMOVED*** accumulator The initial value.
   * @param ***REMOVED***boolean***REMOVED*** initAccum Specify using the first or last element of
   *  `collection` as the initial value.
   * @param ***REMOVED***Function***REMOVED*** eachFunc The function to iterate over `collection`.
   * @returns ***REMOVED*******REMOVED*** Returns the accumulated value.
   */
  function baseReduce(collection, iteratee, accumulator, initAccum, eachFunc) ***REMOVED***
    eachFunc(collection, function(value, index, collection) ***REMOVED***
      accumulator = initAccum
        ? (initAccum = false, value)
        : iteratee(accumulator, value, index, collection);
    ***REMOVED***);
    return accumulator;
  ***REMOVED***

  /**
   * The base implementation of `_.values` and `_.valuesIn` which creates an
   * array of `object` property values corresponding to the property names
   * of `props`.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to query.
   * @param ***REMOVED***Array***REMOVED*** props The property names to get values for.
   * @returns ***REMOVED***Object***REMOVED*** Returns the array of property values.
   */
  function baseValues(object, props) ***REMOVED***
    return baseMap(props, function(key) ***REMOVED***
      return object[key];
    ***REMOVED***);
  ***REMOVED***

  /**
   * Used by `_.escape` to convert characters to HTML entities.
   *
   * @private
   * @param ***REMOVED***string***REMOVED*** chr The matched character to escape.
   * @returns ***REMOVED***string***REMOVED*** Returns the escaped character.
   */
  var escapeHtmlChar = basePropertyOf(htmlEscapes);

  /**
   * Creates a unary function that invokes `func` with its argument transformed.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** func The function to wrap.
   * @param ***REMOVED***Function***REMOVED*** transform The argument transform.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new function.
   */
  function overArg(func, transform) ***REMOVED***
    return function(arg) ***REMOVED***
      return func(transform(arg));
    ***REMOVED***;
  ***REMOVED***

  /*--------------------------------------------------------------------------*/

  /** Used for built-in method references. */
  var arrayProto = Array.prototype,
      objectProto = Object.prototype;

  /** Used to check objects for own properties. */
  var hasOwnProperty = objectProto.hasOwnProperty;

  /** Used to generate unique IDs. */
  var idCounter = 0;

  /**
   * Used to resolve the
   * [`toStringTag`](http://ecma-international.org/ecma-262/7.0/#sec-object.prototype.tostring)
   * of values.
   */
  var nativeObjectToString = objectProto.toString;

  /** Used to restore the original `_` reference in `_.noConflict`. */
  var oldDash = root._;

  /** Built-in value references. */
  var objectCreate = Object.create,
      propertyIsEnumerable = objectProto.propertyIsEnumerable;

  /* Built-in method references for those with the same name as other `lodash` methods. */
  var nativeIsFinite = root.isFinite,
      nativeKeys = overArg(Object.keys, Object),
      nativeMax = Math.max;

  /*------------------------------------------------------------------------*/

  /**
   * Creates a `lodash` object which wraps `value` to enable implicit method
   * chain sequences. Methods that operate on and return arrays, collections,
   * and functions can be chained together. Methods that retrieve a single value
   * or may return a primitive value will automatically end the chain sequence
   * and return the unwrapped value. Otherwise, the value must be unwrapped
   * with `_#value`.
   *
   * Explicit chain sequences, which must be unwrapped with `_#value`, may be
   * enabled using `_.chain`.
   *
   * The execution of chained methods is lazy, that is, it's deferred until
   * `_#value` is implicitly or explicitly called.
   *
   * Lazy evaluation allows several methods to support shortcut fusion.
   * Shortcut fusion is an optimization to merge iteratee calls; this avoids
   * the creation of intermediate arrays and can greatly reduce the number of
   * iteratee executions. Sections of a chain sequence qualify for shortcut
   * fusion if the section is applied to an array and iteratees accept only
   * one argument. The heuristic for whether a section qualifies for shortcut
   * fusion is subject to change.
   *
   * Chaining is supported in custom builds as long as the `_#value` method is
   * directly or indirectly included in the build.
   *
   * In addition to lodash methods, wrappers have `Array` and `String` methods.
   *
   * The wrapper `Array` methods are:
   * `concat`, `join`, `pop`, `push`, `shift`, `sort`, `splice`, and `unshift`
   *
   * The wrapper `String` methods are:
   * `replace` and `split`
   *
   * The wrapper methods that support shortcut fusion are:
   * `at`, `compact`, `drop`, `dropRight`, `dropWhile`, `filter`, `find`,
   * `findLast`, `head`, `initial`, `last`, `map`, `reject`, `reverse`, `slice`,
   * `tail`, `take`, `takeRight`, `takeRightWhile`, `takeWhile`, and `toArray`
   *
   * The chainable wrapper methods are:
   * `after`, `ary`, `assign`, `assignIn`, `assignInWith`, `assignWith`, `at`,
   * `before`, `bind`, `bindAll`, `bindKey`, `castArray`, `chain`, `chunk`,
   * `commit`, `compact`, `concat`, `conforms`, `constant`, `countBy`, `create`,
   * `curry`, `debounce`, `defaults`, `defaultsDeep`, `defer`, `delay`,
   * `difference`, `differenceBy`, `differenceWith`, `drop`, `dropRight`,
   * `dropRightWhile`, `dropWhile`, `extend`, `extendWith`, `fill`, `filter`,
   * `flatMap`, `flatMapDeep`, `flatMapDepth`, `flatten`, `flattenDeep`,
   * `flattenDepth`, `flip`, `flow`, `flowRight`, `fromPairs`, `functions`,
   * `functionsIn`, `groupBy`, `initial`, `intersection`, `intersectionBy`,
   * `intersectionWith`, `invert`, `invertBy`, `invokeMap`, `iteratee`, `keyBy`,
   * `keys`, `keysIn`, `map`, `mapKeys`, `mapValues`, `matches`, `matchesProperty`,
   * `memoize`, `merge`, `mergeWith`, `method`, `methodOf`, `mixin`, `negate`,
   * `nthArg`, `omit`, `omitBy`, `once`, `orderBy`, `over`, `overArgs`,
   * `overEvery`, `overSome`, `partial`, `partialRight`, `partition`, `pick`,
   * `pickBy`, `plant`, `property`, `propertyOf`, `pull`, `pullAll`, `pullAllBy`,
   * `pullAllWith`, `pullAt`, `push`, `range`, `rangeRight`, `rearg`, `reject`,
   * `remove`, `rest`, `reverse`, `sampleSize`, `set`, `setWith`, `shuffle`,
   * `slice`, `sort`, `sortBy`, `splice`, `spread`, `tail`, `take`, `takeRight`,
   * `takeRightWhile`, `takeWhile`, `tap`, `throttle`, `thru`, `toArray`,
   * `toPairs`, `toPairsIn`, `toPath`, `toPlainObject`, `transform`, `unary`,
   * `union`, `unionBy`, `unionWith`, `uniq`, `uniqBy`, `uniqWith`, `unset`,
   * `unshift`, `unzip`, `unzipWith`, `update`, `updateWith`, `values`,
   * `valuesIn`, `without`, `wrap`, `xor`, `xorBy`, `xorWith`, `zip`,
   * `zipObject`, `zipObjectDeep`, and `zipWith`
   *
   * The wrapper methods that are **not** chainable by default are:
   * `add`, `attempt`, `camelCase`, `capitalize`, `ceil`, `clamp`, `clone`,
   * `cloneDeep`, `cloneDeepWith`, `cloneWith`, `conformsTo`, `deburr`,
   * `defaultTo`, `divide`, `each`, `eachRight`, `endsWith`, `eq`, `escape`,
   * `escapeRegExp`, `every`, `find`, `findIndex`, `findKey`, `findLast`,
   * `findLastIndex`, `findLastKey`, `first`, `floor`, `forEach`, `forEachRight`,
   * `forIn`, `forInRight`, `forOwn`, `forOwnRight`, `get`, `gt`, `gte`, `has`,
   * `hasIn`, `head`, `identity`, `includes`, `indexOf`, `inRange`, `invoke`,
   * `isArguments`, `isArray`, `isArrayBuffer`, `isArrayLike`, `isArrayLikeObject`,
   * `isBoolean`, `isBuffer`, `isDate`, `isElement`, `isEmpty`, `isEqual`,
   * `isEqualWith`, `isError`, `isFinite`, `isFunction`, `isInteger`, `isLength`,
   * `isMap`, `isMatch`, `isMatchWith`, `isNaN`, `isNative`, `isNil`, `isNull`,
   * `isNumber`, `isObject`, `isObjectLike`, `isPlainObject`, `isRegExp`,
   * `isSafeInteger`, `isSet`, `isString`, `isUndefined`, `isTypedArray`,
   * `isWeakMap`, `isWeakSet`, `join`, `kebabCase`, `last`, `lastIndexOf`,
   * `lowerCase`, `lowerFirst`, `lt`, `lte`, `max`, `maxBy`, `mean`, `meanBy`,
   * `min`, `minBy`, `multiply`, `noConflict`, `noop`, `now`, `nth`, `pad`,
   * `padEnd`, `padStart`, `parseInt`, `pop`, `random`, `reduce`, `reduceRight`,
   * `repeat`, `result`, `round`, `runInContext`, `sample`, `shift`, `size`,
   * `snakeCase`, `some`, `sortedIndex`, `sortedIndexBy`, `sortedLastIndex`,
   * `sortedLastIndexBy`, `startCase`, `startsWith`, `stubArray`, `stubFalse`,
   * `stubObject`, `stubString`, `stubTrue`, `subtract`, `sum`, `sumBy`,
   * `template`, `times`, `toFinite`, `toInteger`, `toJSON`, `toLength`,
   * `toLower`, `toNumber`, `toSafeInteger`, `toString`, `toUpper`, `trim`,
   * `trimEnd`, `trimStart`, `truncate`, `unescape`, `uniqueId`, `upperCase`,
   * `upperFirst`, `value`, and `words`
   *
   * @name _
   * @constructor
   * @category Seq
   * @param ***REMOVED*******REMOVED*** value The value to wrap in a `lodash` instance.
   * @returns ***REMOVED***Object***REMOVED*** Returns the new `lodash` wrapper instance.
   * @example
   *
   * function square(n) ***REMOVED***
   *   return n * n;
   * ***REMOVED***
   *
   * var wrapped = _([1, 2, 3]);
   *
   * // Returns an unwrapped value.
   * wrapped.reduce(_.add);
   * // => 6
   *
   * // Returns a wrapped value.
   * var squares = wrapped.map(square);
   *
   * _.isArray(squares);
   * // => false
   *
   * _.isArray(squares.value());
   * // => true
   */
  function lodash(value) ***REMOVED***
    return value instanceof LodashWrapper
      ? value
      : new LodashWrapper(value);
  ***REMOVED***

  /**
   * The base implementation of `_.create` without support for assigning
   * properties to the created object.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** proto The object to inherit from.
   * @returns ***REMOVED***Object***REMOVED*** Returns the new object.
   */
  var baseCreate = (function() ***REMOVED***
    function object() ***REMOVED******REMOVED***
    return function(proto) ***REMOVED***
      if (!isObject(proto)) ***REMOVED***
        return ***REMOVED******REMOVED***;
      ***REMOVED***
      if (objectCreate) ***REMOVED***
        return objectCreate(proto);
      ***REMOVED***
      object.prototype = proto;
      var result = new object;
      object.prototype = undefined;
      return result;
    ***REMOVED***;
  ***REMOVED***());

  /**
   * The base constructor for creating `lodash` wrapper objects.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to wrap.
   * @param ***REMOVED***boolean***REMOVED*** [chainAll] Enable explicit method chain sequences.
   */
  function LodashWrapper(value, chainAll) ***REMOVED***
    this.__wrapped__ = value;
    this.__actions__ = [];
    this.__chain__ = !!chainAll;
  ***REMOVED***

  LodashWrapper.prototype = baseCreate(lodash.prototype);
  LodashWrapper.prototype.constructor = LodashWrapper;

  /*------------------------------------------------------------------------*/

  /**
   * Assigns `value` to `key` of `object` if the existing value is not equivalent
   * using [`SameValueZero`](http://ecma-international.org/ecma-262/7.0/#sec-samevaluezero)
   * for equality comparisons.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to modify.
   * @param ***REMOVED***string***REMOVED*** key The key of the property to assign.
   * @param ***REMOVED*******REMOVED*** value The value to assign.
   */
  function assignValue(object, key, value) ***REMOVED***
    var objValue = object[key];
    if (!(hasOwnProperty.call(object, key) && eq(objValue, value)) ||
        (value === undefined && !(key in object))) ***REMOVED***
      baseAssignValue(object, key, value);
    ***REMOVED***
  ***REMOVED***

  /**
   * The base implementation of `assignValue` and `assignMergeValue` without
   * value checks.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to modify.
   * @param ***REMOVED***string***REMOVED*** key The key of the property to assign.
   * @param ***REMOVED*******REMOVED*** value The value to assign.
   */
  function baseAssignValue(object, key, value) ***REMOVED***
    object[key] = value;
  ***REMOVED***

  /**
   * The base implementation of `_.delay` and `_.defer` which accepts `args`
   * to provide to `func`.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** func The function to delay.
   * @param ***REMOVED***number***REMOVED*** wait The number of milliseconds to delay invocation.
   * @param ***REMOVED***Array***REMOVED*** args The arguments to provide to `func`.
   * @returns ***REMOVED***number|Object***REMOVED*** Returns the timer id or timeout object.
   */
  function baseDelay(func, wait, args) ***REMOVED***
    if (typeof func != 'function') ***REMOVED***
      throw new TypeError(FUNC_ERROR_TEXT);
    ***REMOVED***
    return setTimeout(function() ***REMOVED*** func.apply(undefined, args); ***REMOVED***, wait);
  ***REMOVED***

  /**
   * The base implementation of `_.forEach` without support for iteratee shorthands.
   *
   * @private
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** iteratee The function invoked per iteration.
   * @returns ***REMOVED***Array|Object***REMOVED*** Returns `collection`.
   */
  var baseEach = createBaseEach(baseForOwn);

  /**
   * The base implementation of `_.every` without support for iteratee shorthands.
   *
   * @private
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** predicate The function invoked per iteration.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if all elements pass the predicate check,
   *  else `false`
   */
  function baseEvery(collection, predicate) ***REMOVED***
    var result = true;
    baseEach(collection, function(value, index, collection) ***REMOVED***
      result = !!predicate(value, index, collection);
      return result;
    ***REMOVED***);
    return result;
  ***REMOVED***

  /**
   * The base implementation of methods like `_.max` and `_.min` which accepts a
   * `comparator` to determine the extremum value.
   *
   * @private
   * @param ***REMOVED***Array***REMOVED*** array The array to iterate over.
   * @param ***REMOVED***Function***REMOVED*** iteratee The iteratee invoked per iteration.
   * @param ***REMOVED***Function***REMOVED*** comparator The comparator used to compare values.
   * @returns ***REMOVED*******REMOVED*** Returns the extremum value.
   */
  function baseExtremum(array, iteratee, comparator) ***REMOVED***
    var index = -1,
        length = array.length;

    while (++index < length) ***REMOVED***
      var value = array[index],
          current = iteratee(value);

      if (current != null && (computed === undefined
            ? (current === current && !false)
            : comparator(current, computed)
          )) ***REMOVED***
        var computed = current,
            result = value;
      ***REMOVED***
    ***REMOVED***
    return result;
  ***REMOVED***

  /**
   * The base implementation of `_.filter` without support for iteratee shorthands.
   *
   * @private
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** predicate The function invoked per iteration.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new filtered array.
   */
  function baseFilter(collection, predicate) ***REMOVED***
    var result = [];
    baseEach(collection, function(value, index, collection) ***REMOVED***
      if (predicate(value, index, collection)) ***REMOVED***
        result.push(value);
      ***REMOVED***
    ***REMOVED***);
    return result;
  ***REMOVED***

  /**
   * The base implementation of `_.flatten` with support for restricting flattening.
   *
   * @private
   * @param ***REMOVED***Array***REMOVED*** array The array to flatten.
   * @param ***REMOVED***number***REMOVED*** depth The maximum recursion depth.
   * @param ***REMOVED***boolean***REMOVED*** [predicate=isFlattenable] The function invoked per iteration.
   * @param ***REMOVED***boolean***REMOVED*** [isStrict] Restrict to values that pass `predicate` checks.
   * @param ***REMOVED***Array***REMOVED*** [result=[]] The initial result value.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new flattened array.
   */
  function baseFlatten(array, depth, predicate, isStrict, result) ***REMOVED***
    var index = -1,
        length = array.length;

    predicate || (predicate = isFlattenable);
    result || (result = []);

    while (++index < length) ***REMOVED***
      var value = array[index];
      if (depth > 0 && predicate(value)) ***REMOVED***
        if (depth > 1) ***REMOVED***
          // Recursively flatten arrays (susceptible to call stack limits).
          baseFlatten(value, depth - 1, predicate, isStrict, result);
        ***REMOVED*** else ***REMOVED***
          arrayPush(result, value);
        ***REMOVED***
      ***REMOVED*** else if (!isStrict) ***REMOVED***
        result[result.length] = value;
      ***REMOVED***
    ***REMOVED***
    return result;
  ***REMOVED***

  /**
   * The base implementation of `baseForOwn` which iterates over `object`
   * properties returned by `keysFunc` and invokes `iteratee` for each property.
   * Iteratee functions may exit iteration early by explicitly returning `false`.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to iterate over.
   * @param ***REMOVED***Function***REMOVED*** iteratee The function invoked per iteration.
   * @param ***REMOVED***Function***REMOVED*** keysFunc The function to get the keys of `object`.
   * @returns ***REMOVED***Object***REMOVED*** Returns `object`.
   */
  var baseFor = createBaseFor();

  /**
   * The base implementation of `_.forOwn` without support for iteratee shorthands.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to iterate over.
   * @param ***REMOVED***Function***REMOVED*** iteratee The function invoked per iteration.
   * @returns ***REMOVED***Object***REMOVED*** Returns `object`.
   */
  function baseForOwn(object, iteratee) ***REMOVED***
    return object && baseFor(object, iteratee, keys);
  ***REMOVED***

  /**
   * The base implementation of `_.functions` which creates an array of
   * `object` function property names filtered from `props`.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to inspect.
   * @param ***REMOVED***Array***REMOVED*** props The property names to filter.
   * @returns ***REMOVED***Array***REMOVED*** Returns the function names.
   */
  function baseFunctions(object, props) ***REMOVED***
    return baseFilter(props, function(key) ***REMOVED***
      return isFunction(object[key]);
    ***REMOVED***);
  ***REMOVED***

  /**
   * The base implementation of `getTag` without fallbacks for buggy environments.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to query.
   * @returns ***REMOVED***string***REMOVED*** Returns the `toStringTag`.
   */
  function baseGetTag(value) ***REMOVED***
    return objectToString(value);
  ***REMOVED***

  /**
   * The base implementation of `_.gt` which doesn't coerce arguments.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to compare.
   * @param ***REMOVED*******REMOVED*** other The other value to compare.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is greater than `other`,
   *  else `false`.
   */
  function baseGt(value, other) ***REMOVED***
    return value > other;
  ***REMOVED***

  /**
   * The base implementation of `_.isArguments`.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is an `arguments` object,
   */
  var baseIsArguments = noop;

  /**
   * The base implementation of `_.isDate` without Node.js optimizations.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a date object, else `false`.
   */
  function baseIsDate(value) ***REMOVED***
    return isObjectLike(value) && baseGetTag(value) == dateTag;
  ***REMOVED***

  /**
   * The base implementation of `_.isEqual` which supports partial comparisons
   * and tracks traversed objects.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to compare.
   * @param ***REMOVED*******REMOVED*** other The other value to compare.
   * @param ***REMOVED***boolean***REMOVED*** bitmask The bitmask flags.
   *  1 - Unordered comparison
   *  2 - Partial comparison
   * @param ***REMOVED***Function***REMOVED*** [customizer] The function to customize comparisons.
   * @param ***REMOVED***Object***REMOVED*** [stack] Tracks traversed `value` and `other` objects.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the values are equivalent, else `false`.
   */
  function baseIsEqual(value, other, bitmask, customizer, stack) ***REMOVED***
    if (value === other) ***REMOVED***
      return true;
    ***REMOVED***
    if (value == null || other == null || (!isObjectLike(value) && !isObjectLike(other))) ***REMOVED***
      return value !== value && other !== other;
    ***REMOVED***
    return baseIsEqualDeep(value, other, bitmask, customizer, baseIsEqual, stack);
  ***REMOVED***

  /**
   * A specialized version of `baseIsEqual` for arrays and objects which performs
   * deep comparisons and tracks traversed objects enabling objects with circular
   * references to be compared.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to compare.
   * @param ***REMOVED***Object***REMOVED*** other The other object to compare.
   * @param ***REMOVED***number***REMOVED*** bitmask The bitmask flags. See `baseIsEqual` for more details.
   * @param ***REMOVED***Function***REMOVED*** customizer The function to customize comparisons.
   * @param ***REMOVED***Function***REMOVED*** equalFunc The function to determine equivalents of values.
   * @param ***REMOVED***Object***REMOVED*** [stack] Tracks traversed `object` and `other` objects.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the objects are equivalent, else `false`.
   */
  function baseIsEqualDeep(object, other, bitmask, customizer, equalFunc, stack) ***REMOVED***
    var objIsArr = isArray(object),
        othIsArr = isArray(other),
        objTag = objIsArr ? arrayTag : baseGetTag(object),
        othTag = othIsArr ? arrayTag : baseGetTag(other);

    objTag = objTag == argsTag ? objectTag : objTag;
    othTag = othTag == argsTag ? objectTag : othTag;

    var objIsObj = objTag == objectTag,
        othIsObj = othTag == objectTag,
        isSameTag = objTag == othTag;

    stack || (stack = []);
    var objStack = find(stack, function(entry) ***REMOVED***
      return entry[0] == object;
    ***REMOVED***);
    var othStack = find(stack, function(entry) ***REMOVED***
      return entry[0] == other;
    ***REMOVED***);
    if (objStack && othStack) ***REMOVED***
      return objStack[1] == other;
    ***REMOVED***
    stack.push([object, other]);
    stack.push([other, object]);
    if (isSameTag && !objIsObj) ***REMOVED***
      var result = (objIsArr)
        ? equalArrays(object, other, bitmask, customizer, equalFunc, stack)
        : equalByTag(object, other, objTag, bitmask, customizer, equalFunc, stack);
      stack.pop();
      return result;
    ***REMOVED***
    if (!(bitmask & COMPARE_PARTIAL_FLAG)) ***REMOVED***
      var objIsWrapped = objIsObj && hasOwnProperty.call(object, '__wrapped__'),
          othIsWrapped = othIsObj && hasOwnProperty.call(other, '__wrapped__');

      if (objIsWrapped || othIsWrapped) ***REMOVED***
        var objUnwrapped = objIsWrapped ? object.value() : object,
            othUnwrapped = othIsWrapped ? other.value() : other;

        var result = equalFunc(objUnwrapped, othUnwrapped, bitmask, customizer, stack);
        stack.pop();
        return result;
      ***REMOVED***
    ***REMOVED***
    if (!isSameTag) ***REMOVED***
      return false;
    ***REMOVED***
    var result = equalObjects(object, other, bitmask, customizer, equalFunc, stack);
    stack.pop();
    return result;
  ***REMOVED***

  /**
   * The base implementation of `_.isRegExp` without Node.js optimizations.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a regexp, else `false`.
   */
  function baseIsRegExp(value) ***REMOVED***
    return isObjectLike(value) && baseGetTag(value) == regexpTag;
  ***REMOVED***

  /**
   * The base implementation of `_.iteratee`.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** [value=_.identity] The value to convert to an iteratee.
   * @returns ***REMOVED***Function***REMOVED*** Returns the iteratee.
   */
  function baseIteratee(func) ***REMOVED***
    if (typeof func == 'function') ***REMOVED***
      return func;
    ***REMOVED***
    if (func == null) ***REMOVED***
      return identity;
    ***REMOVED***
    return (typeof func == 'object' ? baseMatches : baseProperty)(func);
  ***REMOVED***

  /**
   * The base implementation of `_.lt` which doesn't coerce arguments.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to compare.
   * @param ***REMOVED*******REMOVED*** other The other value to compare.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is less than `other`,
   *  else `false`.
   */
  function baseLt(value, other) ***REMOVED***
    return value < other;
  ***REMOVED***

  /**
   * The base implementation of `_.map` without support for iteratee shorthands.
   *
   * @private
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** iteratee The function invoked per iteration.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new mapped array.
   */
  function baseMap(collection, iteratee) ***REMOVED***
    var index = -1,
        result = isArrayLike(collection) ? Array(collection.length) : [];

    baseEach(collection, function(value, key, collection) ***REMOVED***
      result[++index] = iteratee(value, key, collection);
    ***REMOVED***);
    return result;
  ***REMOVED***

  /**
   * The base implementation of `_.matches` which doesn't clone `source`.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** source The object of property values to match.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new spec function.
   */
  function baseMatches(source) ***REMOVED***
    var props = nativeKeys(source);
    return function(object) ***REMOVED***
      var length = props.length;
      if (object == null) ***REMOVED***
        return !length;
      ***REMOVED***
      object = Object(object);
      while (length--) ***REMOVED***
        var key = props[length];
        if (!(key in object &&
              baseIsEqual(source[key], object[key], COMPARE_PARTIAL_FLAG | COMPARE_UNORDERED_FLAG)
            )) ***REMOVED***
          return false;
        ***REMOVED***
      ***REMOVED***
      return true;
    ***REMOVED***;
  ***REMOVED***

  /**
   * The base implementation of `_.pick` without support for individual
   * property identifiers.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The source object.
   * @param ***REMOVED***string[]***REMOVED*** paths The property paths to pick.
   * @returns ***REMOVED***Object***REMOVED*** Returns the new object.
   */
  function basePick(object, props) ***REMOVED***
    object = Object(object);
    return reduce(props, function(result, key) ***REMOVED***
      if (key in object) ***REMOVED***
        result[key] = object[key];
      ***REMOVED***
      return result;
    ***REMOVED***, ***REMOVED******REMOVED***);
  ***REMOVED***

  /**
   * The base implementation of `_.rest` which doesn't validate or coerce arguments.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** func The function to apply a rest parameter to.
   * @param ***REMOVED***number***REMOVED*** [start=func.length-1] The start position of the rest parameter.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new function.
   */
  function baseRest(func, start) ***REMOVED***
    return setToString(overRest(func, start, identity), func + '');
  ***REMOVED***

  /**
   * The base implementation of `_.slice` without an iteratee call guard.
   *
   * @private
   * @param ***REMOVED***Array***REMOVED*** array The array to slice.
   * @param ***REMOVED***number***REMOVED*** [start=0] The start position.
   * @param ***REMOVED***number***REMOVED*** [end=array.length] The end position.
   * @returns ***REMOVED***Array***REMOVED*** Returns the slice of `array`.
   */
  function baseSlice(array, start, end) ***REMOVED***
    var index = -1,
        length = array.length;

    if (start < 0) ***REMOVED***
      start = -start > length ? 0 : (length + start);
    ***REMOVED***
    end = end > length ? length : end;
    if (end < 0) ***REMOVED***
      end += length;
    ***REMOVED***
    length = start > end ? 0 : ((end - start) >>> 0);
    start >>>= 0;

    var result = Array(length);
    while (++index < length) ***REMOVED***
      result[index] = array[index + start];
    ***REMOVED***
    return result;
  ***REMOVED***

  /**
   * Copies the values of `source` to `array`.
   *
   * @private
   * @param ***REMOVED***Array***REMOVED*** source The array to copy values from.
   * @param ***REMOVED***Array***REMOVED*** [array=[]] The array to copy values to.
   * @returns ***REMOVED***Array***REMOVED*** Returns `array`.
   */
  function copyArray(source) ***REMOVED***
    return baseSlice(source, 0, source.length);
  ***REMOVED***

  /**
   * The base implementation of `_.some` without support for iteratee shorthands.
   *
   * @private
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** predicate The function invoked per iteration.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if any element passes the predicate check,
   *  else `false`.
   */
  function baseSome(collection, predicate) ***REMOVED***
    var result;

    baseEach(collection, function(value, index, collection) ***REMOVED***
      result = predicate(value, index, collection);
      return !result;
    ***REMOVED***);
    return !!result;
  ***REMOVED***

  /**
   * The base implementation of `wrapperValue` which returns the result of
   * performing a sequence of actions on the unwrapped `value`, where each
   * successive action is supplied the return value of the previous.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The unwrapped value.
   * @param ***REMOVED***Array***REMOVED*** actions Actions to perform to resolve the unwrapped value.
   * @returns ***REMOVED*******REMOVED*** Returns the resolved value.
   */
  function baseWrapperValue(value, actions) ***REMOVED***
    var result = value;
    return reduce(actions, function(result, action) ***REMOVED***
      return action.func.apply(action.thisArg, arrayPush([result], action.args));
    ***REMOVED***, result);
  ***REMOVED***

  /**
   * Compares values to sort them in ascending order.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to compare.
   * @param ***REMOVED*******REMOVED*** other The other value to compare.
   * @returns ***REMOVED***number***REMOVED*** Returns the sort order indicator for `value`.
   */
  function compareAscending(value, other) ***REMOVED***
    if (value !== other) ***REMOVED***
      var valIsDefined = value !== undefined,
          valIsNull = value === null,
          valIsReflexive = value === value,
          valIsSymbol = false;

      var othIsDefined = other !== undefined,
          othIsNull = other === null,
          othIsReflexive = other === other,
          othIsSymbol = false;

      if ((!othIsNull && !othIsSymbol && !valIsSymbol && value > other) ||
          (valIsSymbol && othIsDefined && othIsReflexive && !othIsNull && !othIsSymbol) ||
          (valIsNull && othIsDefined && othIsReflexive) ||
          (!valIsDefined && othIsReflexive) ||
          !valIsReflexive) ***REMOVED***
        return 1;
      ***REMOVED***
      if ((!valIsNull && !valIsSymbol && !othIsSymbol && value < other) ||
          (othIsSymbol && valIsDefined && valIsReflexive && !valIsNull && !valIsSymbol) ||
          (othIsNull && valIsDefined && valIsReflexive) ||
          (!othIsDefined && valIsReflexive) ||
          !othIsReflexive) ***REMOVED***
        return -1;
      ***REMOVED***
    ***REMOVED***
    return 0;
  ***REMOVED***

  /**
   * Copies properties of `source` to `object`.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** source The object to copy properties from.
   * @param ***REMOVED***Array***REMOVED*** props The property identifiers to copy.
   * @param ***REMOVED***Object***REMOVED*** [object=***REMOVED******REMOVED***] The object to copy properties to.
   * @param ***REMOVED***Function***REMOVED*** [customizer] The function to customize copied values.
   * @returns ***REMOVED***Object***REMOVED*** Returns `object`.
   */
  function copyObject(source, props, object, customizer) ***REMOVED***
    var isNew = !object;
    object || (object = ***REMOVED******REMOVED***);

    var index = -1,
        length = props.length;

    while (++index < length) ***REMOVED***
      var key = props[index];

      var newValue = customizer
        ? customizer(object[key], source[key], key, object, source)
        : undefined;

      if (newValue === undefined) ***REMOVED***
        newValue = source[key];
      ***REMOVED***
      if (isNew) ***REMOVED***
        baseAssignValue(object, key, newValue);
      ***REMOVED*** else ***REMOVED***
        assignValue(object, key, newValue);
      ***REMOVED***
    ***REMOVED***
    return object;
  ***REMOVED***

  /**
   * Creates a function like `_.assign`.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** assigner The function to assign values.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new assigner function.
   */
  function createAssigner(assigner) ***REMOVED***
    return baseRest(function(object, sources) ***REMOVED***
      var index = -1,
          length = sources.length,
          customizer = length > 1 ? sources[length - 1] : undefined;

      customizer = (assigner.length > 3 && typeof customizer == 'function')
        ? (length--, customizer)
        : undefined;

      object = Object(object);
      while (++index < length) ***REMOVED***
        var source = sources[index];
        if (source) ***REMOVED***
          assigner(object, source, index, customizer);
        ***REMOVED***
      ***REMOVED***
      return object;
    ***REMOVED***);
  ***REMOVED***

  /**
   * Creates a `baseEach` or `baseEachRight` function.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** eachFunc The function to iterate over a collection.
   * @param ***REMOVED***boolean***REMOVED*** [fromRight] Specify iterating from right to left.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new base function.
   */
  function createBaseEach(eachFunc, fromRight) ***REMOVED***
    return function(collection, iteratee) ***REMOVED***
      if (collection == null) ***REMOVED***
        return collection;
      ***REMOVED***
      if (!isArrayLike(collection)) ***REMOVED***
        return eachFunc(collection, iteratee);
      ***REMOVED***
      var length = collection.length,
          index = fromRight ? length : -1,
          iterable = Object(collection);

      while ((fromRight ? index-- : ++index < length)) ***REMOVED***
        if (iteratee(iterable[index], index, iterable) === false) ***REMOVED***
          break;
        ***REMOVED***
      ***REMOVED***
      return collection;
    ***REMOVED***;
  ***REMOVED***

  /**
   * Creates a base function for methods like `_.forIn` and `_.forOwn`.
   *
   * @private
   * @param ***REMOVED***boolean***REMOVED*** [fromRight] Specify iterating from right to left.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new base function.
   */
  function createBaseFor(fromRight) ***REMOVED***
    return function(object, iteratee, keysFunc) ***REMOVED***
      var index = -1,
          iterable = Object(object),
          props = keysFunc(object),
          length = props.length;

      while (length--) ***REMOVED***
        var key = props[fromRight ? length : ++index];
        if (iteratee(iterable[key], key, iterable) === false) ***REMOVED***
          break;
        ***REMOVED***
      ***REMOVED***
      return object;
    ***REMOVED***;
  ***REMOVED***

  /**
   * Creates a function that produces an instance of `Ctor` regardless of
   * whether it was invoked as part of a `new` expression or by `call` or `apply`.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** Ctor The constructor to wrap.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new wrapped function.
   */
  function createCtor(Ctor) ***REMOVED***
    return function() ***REMOVED***
      // Use a `switch` statement to work with class constructors. See
      // http://ecma-international.org/ecma-262/7.0/#sec-ecmascript-function-objects-call-thisargument-argumentslist
      // for more details.
      var args = arguments;
      var thisBinding = baseCreate(Ctor.prototype),
          result = Ctor.apply(thisBinding, args);

      // Mimic the constructor's `return` behavior.
      // See https://es5.github.io/#x13.2.2 for more details.
      return isObject(result) ? result : thisBinding;
    ***REMOVED***;
  ***REMOVED***

  /**
   * Creates a `_.find` or `_.findLast` function.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** findIndexFunc The function to find the collection index.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new find function.
   */
  function createFind(findIndexFunc) ***REMOVED***
    return function(collection, predicate, fromIndex) ***REMOVED***
      var iterable = Object(collection);
      if (!isArrayLike(collection)) ***REMOVED***
        var iteratee = baseIteratee(predicate, 3);
        collection = keys(collection);
        predicate = function(key) ***REMOVED*** return iteratee(iterable[key], key, iterable); ***REMOVED***;
      ***REMOVED***
      var index = findIndexFunc(collection, predicate, fromIndex);
      return index > -1 ? iterable[iteratee ? collection[index] : index] : undefined;
    ***REMOVED***;
  ***REMOVED***

  /**
   * Creates a function that wraps `func` to invoke it with the `this` binding
   * of `thisArg` and `partials` prepended to the arguments it receives.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** func The function to wrap.
   * @param ***REMOVED***number***REMOVED*** bitmask The bitmask flags. See `createWrap` for more details.
   * @param ***REMOVED*******REMOVED*** thisArg The `this` binding of `func`.
   * @param ***REMOVED***Array***REMOVED*** partials The arguments to prepend to those provided to
   *  the new function.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new wrapped function.
   */
  function createPartial(func, bitmask, thisArg, partials) ***REMOVED***
    if (typeof func != 'function') ***REMOVED***
      throw new TypeError(FUNC_ERROR_TEXT);
    ***REMOVED***
    var isBind = bitmask & WRAP_BIND_FLAG,
        Ctor = createCtor(func);

    function wrapper() ***REMOVED***
      var argsIndex = -1,
          argsLength = arguments.length,
          leftIndex = -1,
          leftLength = partials.length,
          args = Array(leftLength + argsLength),
          fn = (this && this !== root && this instanceof wrapper) ? Ctor : func;

      while (++leftIndex < leftLength) ***REMOVED***
        args[leftIndex] = partials[leftIndex];
      ***REMOVED***
      while (argsLength--) ***REMOVED***
        args[leftIndex++] = arguments[++argsIndex];
      ***REMOVED***
      return fn.apply(isBind ? thisArg : this, args);
    ***REMOVED***
    return wrapper;
  ***REMOVED***

  /**
   * Used by `_.defaults` to customize its `_.assignIn` use to assign properties
   * of source objects to the destination object for all destination properties
   * that resolve to `undefined`.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** objValue The destination value.
   * @param ***REMOVED*******REMOVED*** srcValue The source value.
   * @param ***REMOVED***string***REMOVED*** key The key of the property to assign.
   * @param ***REMOVED***Object***REMOVED*** object The parent object of `objValue`.
   * @returns ***REMOVED*******REMOVED*** Returns the value to assign.
   */
  function customDefaultsAssignIn(objValue, srcValue, key, object) ***REMOVED***
    if (objValue === undefined ||
        (eq(objValue, objectProto[key]) && !hasOwnProperty.call(object, key))) ***REMOVED***
      return srcValue;
    ***REMOVED***
    return objValue;
  ***REMOVED***

  /**
   * A specialized version of `baseIsEqualDeep` for arrays with support for
   * partial deep comparisons.
   *
   * @private
   * @param ***REMOVED***Array***REMOVED*** array The array to compare.
   * @param ***REMOVED***Array***REMOVED*** other The other array to compare.
   * @param ***REMOVED***number***REMOVED*** bitmask The bitmask flags. See `baseIsEqual` for more details.
   * @param ***REMOVED***Function***REMOVED*** customizer The function to customize comparisons.
   * @param ***REMOVED***Function***REMOVED*** equalFunc The function to determine equivalents of values.
   * @param ***REMOVED***Object***REMOVED*** stack Tracks traversed `array` and `other` objects.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the arrays are equivalent, else `false`.
   */
  function equalArrays(array, other, bitmask, customizer, equalFunc, stack) ***REMOVED***
    var isPartial = bitmask & COMPARE_PARTIAL_FLAG,
        arrLength = array.length,
        othLength = other.length;

    if (arrLength != othLength && !(isPartial && othLength > arrLength)) ***REMOVED***
      return false;
    ***REMOVED***
    var index = -1,
        result = true,
        seen = (bitmask & COMPARE_UNORDERED_FLAG) ? [] : undefined;

    // Ignore non-index properties.
    while (++index < arrLength) ***REMOVED***
      var arrValue = array[index],
          othValue = other[index];

      var compared;
      if (compared !== undefined) ***REMOVED***
        if (compared) ***REMOVED***
          continue;
        ***REMOVED***
        result = false;
        break;
      ***REMOVED***
      // Recursively compare arrays (susceptible to call stack limits).
      if (seen) ***REMOVED***
        if (!baseSome(other, function(othValue, othIndex) ***REMOVED***
              if (!indexOf(seen, othIndex) &&
                  (arrValue === othValue || equalFunc(arrValue, othValue, bitmask, customizer, stack))) ***REMOVED***
                return seen.push(othIndex);
              ***REMOVED***
            ***REMOVED***)) ***REMOVED***
          result = false;
          break;
        ***REMOVED***
      ***REMOVED*** else if (!(
            arrValue === othValue ||
              equalFunc(arrValue, othValue, bitmask, customizer, stack)
          )) ***REMOVED***
        result = false;
        break;
      ***REMOVED***
    ***REMOVED***
    return result;
  ***REMOVED***

  /**
   * A specialized version of `baseIsEqualDeep` for comparing objects of
   * the same `toStringTag`.
   *
   * **Note:** This function only supports comparing values with tags of
   * `Boolean`, `Date`, `Error`, `Number`, `RegExp`, or `String`.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to compare.
   * @param ***REMOVED***Object***REMOVED*** other The other object to compare.
   * @param ***REMOVED***string***REMOVED*** tag The `toStringTag` of the objects to compare.
   * @param ***REMOVED***number***REMOVED*** bitmask The bitmask flags. See `baseIsEqual` for more details.
   * @param ***REMOVED***Function***REMOVED*** customizer The function to customize comparisons.
   * @param ***REMOVED***Function***REMOVED*** equalFunc The function to determine equivalents of values.
   * @param ***REMOVED***Object***REMOVED*** stack Tracks traversed `object` and `other` objects.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the objects are equivalent, else `false`.
   */
  function equalByTag(object, other, tag, bitmask, customizer, equalFunc, stack) ***REMOVED***
    switch (tag) ***REMOVED***

      case boolTag:
      case dateTag:
      case numberTag:
        // Coerce booleans to `1` or `0` and dates to milliseconds.
        // Invalid dates are coerced to `NaN`.
        return eq(+object, +other);

      case errorTag:
        return object.name == other.name && object.message == other.message;

      case regexpTag:
      case stringTag:
        // Coerce regexes to strings and treat strings, primitives and objects,
        // as equal. See http://www.ecma-international.org/ecma-262/7.0/#sec-regexp.prototype.tostring
        // for more details.
        return object == (other + '');

    ***REMOVED***
    return false;
  ***REMOVED***

  /**
   * A specialized version of `baseIsEqualDeep` for objects with support for
   * partial deep comparisons.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to compare.
   * @param ***REMOVED***Object***REMOVED*** other The other object to compare.
   * @param ***REMOVED***number***REMOVED*** bitmask The bitmask flags. See `baseIsEqual` for more details.
   * @param ***REMOVED***Function***REMOVED*** customizer The function to customize comparisons.
   * @param ***REMOVED***Function***REMOVED*** equalFunc The function to determine equivalents of values.
   * @param ***REMOVED***Object***REMOVED*** stack Tracks traversed `object` and `other` objects.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the objects are equivalent, else `false`.
   */
  function equalObjects(object, other, bitmask, customizer, equalFunc, stack) ***REMOVED***
    var isPartial = bitmask & COMPARE_PARTIAL_FLAG,
        objProps = keys(object),
        objLength = objProps.length,
        othProps = keys(other),
        othLength = othProps.length;

    if (objLength != othLength && !isPartial) ***REMOVED***
      return false;
    ***REMOVED***
    var index = objLength;
    while (index--) ***REMOVED***
      var key = objProps[index];
      if (!(isPartial ? key in other : hasOwnProperty.call(other, key))) ***REMOVED***
        return false;
      ***REMOVED***
    ***REMOVED***
    var result = true;

    var skipCtor = isPartial;
    while (++index < objLength) ***REMOVED***
      key = objProps[index];
      var objValue = object[key],
          othValue = other[key];

      var compared;
      // Recursively compare objects (susceptible to call stack limits).
      if (!(compared === undefined
            ? (objValue === othValue || equalFunc(objValue, othValue, bitmask, customizer, stack))
            : compared
          )) ***REMOVED***
        result = false;
        break;
      ***REMOVED***
      skipCtor || (skipCtor = key == 'constructor');
    ***REMOVED***
    if (result && !skipCtor) ***REMOVED***
      var objCtor = object.constructor,
          othCtor = other.constructor;

      // Non `Object` object instances with different constructors are not equal.
      if (objCtor != othCtor &&
          ('constructor' in object && 'constructor' in other) &&
          !(typeof objCtor == 'function' && objCtor instanceof objCtor &&
            typeof othCtor == 'function' && othCtor instanceof othCtor)) ***REMOVED***
        result = false;
      ***REMOVED***
    ***REMOVED***
    return result;
  ***REMOVED***

  /**
   * A specialized version of `baseRest` which flattens the rest array.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** func The function to apply a rest parameter to.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new function.
   */
  function flatRest(func) ***REMOVED***
    return setToString(overRest(func, undefined, flatten), func + '');
  ***REMOVED***

  /**
   * Checks if `value` is a flattenable `arguments` object or array.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is flattenable, else `false`.
   */
  function isFlattenable(value) ***REMOVED***
    return isArray(value) || isArguments(value);
  ***REMOVED***

  /**
   * This function is like
   * [`Object.keys`](http://ecma-international.org/ecma-262/7.0/#sec-object.keys)
   * except that it includes inherited enumerable properties.
   *
   * @private
   * @param ***REMOVED***Object***REMOVED*** object The object to query.
   * @returns ***REMOVED***Array***REMOVED*** Returns the array of property names.
   */
  function nativeKeysIn(object) ***REMOVED***
    var result = [];
    if (object != null) ***REMOVED***
      for (var key in Object(object)) ***REMOVED***
        result.push(key);
      ***REMOVED***
    ***REMOVED***
    return result;
  ***REMOVED***

  /**
   * Converts `value` to a string using `Object.prototype.toString`.
   *
   * @private
   * @param ***REMOVED*******REMOVED*** value The value to convert.
   * @returns ***REMOVED***string***REMOVED*** Returns the converted string.
   */
  function objectToString(value) ***REMOVED***
    return nativeObjectToString.call(value);
  ***REMOVED***

  /**
   * A specialized version of `baseRest` which transforms the rest array.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** func The function to apply a rest parameter to.
   * @param ***REMOVED***number***REMOVED*** [start=func.length-1] The start position of the rest parameter.
   * @param ***REMOVED***Function***REMOVED*** transform The rest array transform.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new function.
   */
  function overRest(func, start, transform) ***REMOVED***
    start = nativeMax(start === undefined ? (func.length - 1) : start, 0);
    return function() ***REMOVED***
      var args = arguments,
          index = -1,
          length = nativeMax(args.length - start, 0),
          array = Array(length);

      while (++index < length) ***REMOVED***
        array[index] = args[start + index];
      ***REMOVED***
      index = -1;
      var otherArgs = Array(start + 1);
      while (++index < start) ***REMOVED***
        otherArgs[index] = args[index];
      ***REMOVED***
      otherArgs[start] = transform(array);
      return func.apply(this, otherArgs);
    ***REMOVED***;
  ***REMOVED***

  /**
   * Sets the `toString` method of `func` to return `string`.
   *
   * @private
   * @param ***REMOVED***Function***REMOVED*** func The function to modify.
   * @param ***REMOVED***Function***REMOVED*** string The `toString` result.
   * @returns ***REMOVED***Function***REMOVED*** Returns `func`.
   */
  var setToString = identity;

  /*------------------------------------------------------------------------*/

  /**
   * Creates an array with all falsey values removed. The values `false`, `null`,
   * `0`, `""`, `undefined`, and `NaN` are falsey.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Array
   * @param ***REMOVED***Array***REMOVED*** array The array to compact.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new array of filtered values.
   * @example
   *
   * _.compact([0, 1, false, 2, '', 3]);
   * // => [1, 2, 3]
   */
  function compact(array) ***REMOVED***
    return baseFilter(array, Boolean);
  ***REMOVED***

  /**
   * Creates a new array concatenating `array` with any additional arrays
   * and/or values.
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @category Array
   * @param ***REMOVED***Array***REMOVED*** array The array to concatenate.
   * @param ***REMOVED***...****REMOVED*** [values] The values to concatenate.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new concatenated array.
   * @example
   *
   * var array = [1];
   * var other = _.concat(array, 2, [3], [[4]]);
   *
   * console.log(other);
   * // => [1, 2, 3, [4]]
   *
   * console.log(array);
   * // => [1]
   */
  function concat() ***REMOVED***
    var length = arguments.length;
    if (!length) ***REMOVED***
      return [];
    ***REMOVED***
    var args = Array(length - 1),
        array = arguments[0],
        index = length;

    while (index--) ***REMOVED***
      args[index - 1] = arguments[index];
    ***REMOVED***
    return arrayPush(isArray(array) ? copyArray(array) : [array], baseFlatten(args, 1));
  ***REMOVED***

  /**
   * This method is like `_.find` except that it returns the index of the first
   * element `predicate` returns truthy for instead of the element itself.
   *
   * @static
   * @memberOf _
   * @since 1.1.0
   * @category Array
   * @param ***REMOVED***Array***REMOVED*** array The array to inspect.
   * @param ***REMOVED***Function***REMOVED*** [predicate=_.identity] The function invoked per iteration.
   * @param ***REMOVED***number***REMOVED*** [fromIndex=0] The index to search from.
   * @returns ***REMOVED***number***REMOVED*** Returns the index of the found element, else `-1`.
   * @example
   *
   * var users = [
   *   ***REMOVED*** 'user': 'barney',  'active': false ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred',    'active': false ***REMOVED***,
   *   ***REMOVED*** 'user': 'pebbles', 'active': true ***REMOVED***
   * ];
   *
   * _.findIndex(users, function(o) ***REMOVED*** return o.user == 'barney'; ***REMOVED***);
   * // => 0
   *
   * // The `_.matches` iteratee shorthand.
   * _.findIndex(users, ***REMOVED*** 'user': 'fred', 'active': false ***REMOVED***);
   * // => 1
   *
   * // The `_.matchesProperty` iteratee shorthand.
   * _.findIndex(users, ['active', false]);
   * // => 0
   *
   * // The `_.property` iteratee shorthand.
   * _.findIndex(users, 'active');
   * // => 2
   */
  function findIndex(array, predicate, fromIndex) ***REMOVED***
    var length = array == null ? 0 : array.length;
    if (!length) ***REMOVED***
      return -1;
    ***REMOVED***
    var index = fromIndex == null ? 0 : toInteger(fromIndex);
    if (index < 0) ***REMOVED***
      index = nativeMax(length + index, 0);
    ***REMOVED***
    return baseFindIndex(array, baseIteratee(predicate, 3), index);
  ***REMOVED***

  /**
   * Flattens `array` a single level deep.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Array
   * @param ***REMOVED***Array***REMOVED*** array The array to flatten.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new flattened array.
   * @example
   *
   * _.flatten([1, [2, [3, [4]], 5]]);
   * // => [1, 2, [3, [4]], 5]
   */
  function flatten(array) ***REMOVED***
    var length = array == null ? 0 : array.length;
    return length ? baseFlatten(array, 1) : [];
  ***REMOVED***

  /**
   * Recursively flattens `array`.
   *
   * @static
   * @memberOf _
   * @since 3.0.0
   * @category Array
   * @param ***REMOVED***Array***REMOVED*** array The array to flatten.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new flattened array.
   * @example
   *
   * _.flattenDeep([1, [2, [3, [4]], 5]]);
   * // => [1, 2, 3, 4, 5]
   */
  function flattenDeep(array) ***REMOVED***
    var length = array == null ? 0 : array.length;
    return length ? baseFlatten(array, INFINITY) : [];
  ***REMOVED***

  /**
   * Gets the first element of `array`.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @alias first
   * @category Array
   * @param ***REMOVED***Array***REMOVED*** array The array to query.
   * @returns ***REMOVED*******REMOVED*** Returns the first element of `array`.
   * @example
   *
   * _.head([1, 2, 3]);
   * // => 1
   *
   * _.head([]);
   * // => undefined
   */
  function head(array) ***REMOVED***
    return (array && array.length) ? array[0] : undefined;
  ***REMOVED***

  /**
   * Gets the index at which the first occurrence of `value` is found in `array`
   * using [`SameValueZero`](http://ecma-international.org/ecma-262/7.0/#sec-samevaluezero)
   * for equality comparisons. If `fromIndex` is negative, it's used as the
   * offset from the end of `array`.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Array
   * @param ***REMOVED***Array***REMOVED*** array The array to inspect.
   * @param ***REMOVED*******REMOVED*** value The value to search for.
   * @param ***REMOVED***number***REMOVED*** [fromIndex=0] The index to search from.
   * @returns ***REMOVED***number***REMOVED*** Returns the index of the matched value, else `-1`.
   * @example
   *
   * _.indexOf([1, 2, 1, 2], 2);
   * // => 1
   *
   * // Search from the `fromIndex`.
   * _.indexOf([1, 2, 1, 2], 2, 2);
   * // => 3
   */
  function indexOf(array, value, fromIndex) ***REMOVED***
    var length = array == null ? 0 : array.length;
    if (typeof fromIndex == 'number') ***REMOVED***
      fromIndex = fromIndex < 0 ? nativeMax(length + fromIndex, 0) : fromIndex;
    ***REMOVED*** else ***REMOVED***
      fromIndex = 0;
    ***REMOVED***
    var index = (fromIndex || 0) - 1,
        isReflexive = value === value;

    while (++index < length) ***REMOVED***
      var other = array[index];
      if ((isReflexive ? other === value : other !== other)) ***REMOVED***
        return index;
      ***REMOVED***
    ***REMOVED***
    return -1;
  ***REMOVED***

  /**
   * Gets the last element of `array`.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Array
   * @param ***REMOVED***Array***REMOVED*** array The array to query.
   * @returns ***REMOVED*******REMOVED*** Returns the last element of `array`.
   * @example
   *
   * _.last([1, 2, 3]);
   * // => 3
   */
  function last(array) ***REMOVED***
    var length = array == null ? 0 : array.length;
    return length ? array[length - 1] : undefined;
  ***REMOVED***

  /**
   * Creates a slice of `array` from `start` up to, but not including, `end`.
   *
   * **Note:** This method is used instead of
   * [`Array#slice`](https://mdn.io/Array/slice) to ensure dense arrays are
   * returned.
   *
   * @static
   * @memberOf _
   * @since 3.0.0
   * @category Array
   * @param ***REMOVED***Array***REMOVED*** array The array to slice.
   * @param ***REMOVED***number***REMOVED*** [start=0] The start position.
   * @param ***REMOVED***number***REMOVED*** [end=array.length] The end position.
   * @returns ***REMOVED***Array***REMOVED*** Returns the slice of `array`.
   */
  function slice(array, start, end) ***REMOVED***
    var length = array == null ? 0 : array.length;
    start = start == null ? 0 : +start;
    end = end === undefined ? length : +end;
    return length ? baseSlice(array, start, end) : [];
  ***REMOVED***

  /*------------------------------------------------------------------------*/

  /**
   * Creates a `lodash` wrapper instance that wraps `value` with explicit method
   * chain sequences enabled. The result of such sequences must be unwrapped
   * with `_#value`.
   *
   * @static
   * @memberOf _
   * @since 1.3.0
   * @category Seq
   * @param ***REMOVED*******REMOVED*** value The value to wrap.
   * @returns ***REMOVED***Object***REMOVED*** Returns the new `lodash` wrapper instance.
   * @example
   *
   * var users = [
   *   ***REMOVED*** 'user': 'barney',  'age': 36 ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred',    'age': 40 ***REMOVED***,
   *   ***REMOVED*** 'user': 'pebbles', 'age': 1 ***REMOVED***
   * ];
   *
   * var youngest = _
   *   .chain(users)
   *   .sortBy('age')
   *   .map(function(o) ***REMOVED***
   *     return o.user + ' is ' + o.age;
   *   ***REMOVED***)
   *   .head()
   *   .value();
   * // => 'pebbles is 1'
   */
  function chain(value) ***REMOVED***
    var result = lodash(value);
    result.__chain__ = true;
    return result;
  ***REMOVED***

  /**
   * This method invokes `interceptor` and returns `value`. The interceptor
   * is invoked with one argument; (value). The purpose of this method is to
   * "tap into" a method chain sequence in order to modify intermediate results.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Seq
   * @param ***REMOVED*******REMOVED*** value The value to provide to `interceptor`.
   * @param ***REMOVED***Function***REMOVED*** interceptor The function to invoke.
   * @returns ***REMOVED*******REMOVED*** Returns `value`.
   * @example
   *
   * _([1, 2, 3])
   *  .tap(function(array) ***REMOVED***
   *    // Mutate input array.
   *    array.pop();
   *  ***REMOVED***)
   *  .reverse()
   *  .value();
   * // => [2, 1]
   */
  function tap(value, interceptor) ***REMOVED***
    interceptor(value);
    return value;
  ***REMOVED***

  /**
   * This method is like `_.tap` except that it returns the result of `interceptor`.
   * The purpose of this method is to "pass thru" values replacing intermediate
   * results in a method chain sequence.
   *
   * @static
   * @memberOf _
   * @since 3.0.0
   * @category Seq
   * @param ***REMOVED*******REMOVED*** value The value to provide to `interceptor`.
   * @param ***REMOVED***Function***REMOVED*** interceptor The function to invoke.
   * @returns ***REMOVED*******REMOVED*** Returns the result of `interceptor`.
   * @example
   *
   * _('  abc  ')
   *  .chain()
   *  .trim()
   *  .thru(function(value) ***REMOVED***
   *    return [value];
   *  ***REMOVED***)
   *  .value();
   * // => ['abc']
   */
  function thru(value, interceptor) ***REMOVED***
    return interceptor(value);
  ***REMOVED***

  /**
   * Creates a `lodash` wrapper instance with explicit method chain sequences enabled.
   *
   * @name chain
   * @memberOf _
   * @since 0.1.0
   * @category Seq
   * @returns ***REMOVED***Object***REMOVED*** Returns the new `lodash` wrapper instance.
   * @example
   *
   * var users = [
   *   ***REMOVED*** 'user': 'barney', 'age': 36 ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred',   'age': 40 ***REMOVED***
   * ];
   *
   * // A sequence without explicit chaining.
   * _(users).head();
   * // => ***REMOVED*** 'user': 'barney', 'age': 36 ***REMOVED***
   *
   * // A sequence with explicit chaining.
   * _(users)
   *   .chain()
   *   .head()
   *   .pick('user')
   *   .value();
   * // => ***REMOVED*** 'user': 'barney' ***REMOVED***
   */
  function wrapperChain() ***REMOVED***
    return chain(this);
  ***REMOVED***

  /**
   * Executes the chain sequence to resolve the unwrapped value.
   *
   * @name value
   * @memberOf _
   * @since 0.1.0
   * @alias toJSON, valueOf
   * @category Seq
   * @returns ***REMOVED*******REMOVED*** Returns the resolved unwrapped value.
   * @example
   *
   * _([1, 2, 3]).value();
   * // => [1, 2, 3]
   */
  function wrapperValue() ***REMOVED***
    return baseWrapperValue(this.__wrapped__, this.__actions__);
  ***REMOVED***

  /*------------------------------------------------------------------------*/

  /**
   * Checks if `predicate` returns truthy for **all** elements of `collection`.
   * Iteration is stopped once `predicate` returns falsey. The predicate is
   * invoked with three arguments: (value, index|key, collection).
   *
   * **Note:** This method returns `true` for
   * [empty collections](https://en.wikipedia.org/wiki/Empty_set) because
   * [everything is true](https://en.wikipedia.org/wiki/Vacuous_truth) of
   * elements of empty collections.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Collection
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** [predicate=_.identity] The function invoked per iteration.
   * @param- ***REMOVED***Object***REMOVED*** [guard] Enables use as an iteratee for methods like `_.map`.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if all elements pass the predicate check,
   *  else `false`.
   * @example
   *
   * _.every([true, 1, null, 'yes'], Boolean);
   * // => false
   *
   * var users = [
   *   ***REMOVED*** 'user': 'barney', 'age': 36, 'active': false ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred',   'age': 40, 'active': false ***REMOVED***
   * ];
   *
   * // The `_.matches` iteratee shorthand.
   * _.every(users, ***REMOVED*** 'user': 'barney', 'active': false ***REMOVED***);
   * // => false
   *
   * // The `_.matchesProperty` iteratee shorthand.
   * _.every(users, ['active', false]);
   * // => true
   *
   * // The `_.property` iteratee shorthand.
   * _.every(users, 'active');
   * // => false
   */
  function every(collection, predicate, guard) ***REMOVED***
    predicate = guard ? undefined : predicate;
    return baseEvery(collection, baseIteratee(predicate));
  ***REMOVED***

  /**
   * Iterates over elements of `collection`, returning an array of all elements
   * `predicate` returns truthy for. The predicate is invoked with three
   * arguments: (value, index|key, collection).
   *
   * **Note:** Unlike `_.remove`, this method returns a new array.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Collection
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** [predicate=_.identity] The function invoked per iteration.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new filtered array.
   * @see _.reject
   * @example
   *
   * var users = [
   *   ***REMOVED*** 'user': 'barney', 'age': 36, 'active': true ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred',   'age': 40, 'active': false ***REMOVED***
   * ];
   *
   * _.filter(users, function(o) ***REMOVED*** return !o.active; ***REMOVED***);
   * // => objects for ['fred']
   *
   * // The `_.matches` iteratee shorthand.
   * _.filter(users, ***REMOVED*** 'age': 36, 'active': true ***REMOVED***);
   * // => objects for ['barney']
   *
   * // The `_.matchesProperty` iteratee shorthand.
   * _.filter(users, ['active', false]);
   * // => objects for ['fred']
   *
   * // The `_.property` iteratee shorthand.
   * _.filter(users, 'active');
   * // => objects for ['barney']
   */
  function filter(collection, predicate) ***REMOVED***
    return baseFilter(collection, baseIteratee(predicate));
  ***REMOVED***

  /**
   * Iterates over elements of `collection`, returning the first element
   * `predicate` returns truthy for. The predicate is invoked with three
   * arguments: (value, index|key, collection).
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Collection
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to inspect.
   * @param ***REMOVED***Function***REMOVED*** [predicate=_.identity] The function invoked per iteration.
   * @param ***REMOVED***number***REMOVED*** [fromIndex=0] The index to search from.
   * @returns ***REMOVED*******REMOVED*** Returns the matched element, else `undefined`.
   * @example
   *
   * var users = [
   *   ***REMOVED*** 'user': 'barney',  'age': 36, 'active': true ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred',    'age': 40, 'active': false ***REMOVED***,
   *   ***REMOVED*** 'user': 'pebbles', 'age': 1,  'active': true ***REMOVED***
   * ];
   *
   * _.find(users, function(o) ***REMOVED*** return o.age < 40; ***REMOVED***);
   * // => object for 'barney'
   *
   * // The `_.matches` iteratee shorthand.
   * _.find(users, ***REMOVED*** 'age': 1, 'active': true ***REMOVED***);
   * // => object for 'pebbles'
   *
   * // The `_.matchesProperty` iteratee shorthand.
   * _.find(users, ['active', false]);
   * // => object for 'fred'
   *
   * // The `_.property` iteratee shorthand.
   * _.find(users, 'active');
   * // => object for 'barney'
   */
  var find = createFind(findIndex);

  /**
   * Iterates over elements of `collection` and invokes `iteratee` for each element.
   * The iteratee is invoked with three arguments: (value, index|key, collection).
   * Iteratee functions may exit iteration early by explicitly returning `false`.
   *
   * **Note:** As with other "Collections" methods, objects with a "length"
   * property are iterated like arrays. To avoid this behavior use `_.forIn`
   * or `_.forOwn` for object iteration.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @alias each
   * @category Collection
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** [iteratee=_.identity] The function invoked per iteration.
   * @returns ***REMOVED***Array|Object***REMOVED*** Returns `collection`.
   * @see _.forEachRight
   * @example
   *
   * _.forEach([1, 2], function(value) ***REMOVED***
   *   console.log(value);
   * ***REMOVED***);
   * // => Logs `1` then `2`.
   *
   * _.forEach(***REMOVED*** 'a': 1, 'b': 2 ***REMOVED***, function(value, key) ***REMOVED***
   *   console.log(key);
   * ***REMOVED***);
   * // => Logs 'a' then 'b' (iteration order is not guaranteed).
   */
  function forEach(collection, iteratee) ***REMOVED***
    return baseEach(collection, baseIteratee(iteratee));
  ***REMOVED***

  /**
   * Creates an array of values by running each element in `collection` thru
   * `iteratee`. The iteratee is invoked with three arguments:
   * (value, index|key, collection).
   *
   * Many lodash methods are guarded to work as iteratees for methods like
   * `_.every`, `_.filter`, `_.map`, `_.mapValues`, `_.reject`, and `_.some`.
   *
   * The guarded methods are:
   * `ary`, `chunk`, `curry`, `curryRight`, `drop`, `dropRight`, `every`,
   * `fill`, `invert`, `parseInt`, `random`, `range`, `rangeRight`, `repeat`,
   * `sampleSize`, `slice`, `some`, `sortBy`, `split`, `take`, `takeRight`,
   * `template`, `trim`, `trimEnd`, `trimStart`, and `words`
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Collection
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** [iteratee=_.identity] The function invoked per iteration.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new mapped array.
   * @example
   *
   * function square(n) ***REMOVED***
   *   return n * n;
   * ***REMOVED***
   *
   * _.map([4, 8], square);
   * // => [16, 64]
   *
   * _.map(***REMOVED*** 'a': 4, 'b': 8 ***REMOVED***, square);
   * // => [16, 64] (iteration order is not guaranteed)
   *
   * var users = [
   *   ***REMOVED*** 'user': 'barney' ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred' ***REMOVED***
   * ];
   *
   * // The `_.property` iteratee shorthand.
   * _.map(users, 'user');
   * // => ['barney', 'fred']
   */
  function map(collection, iteratee) ***REMOVED***
    return baseMap(collection, baseIteratee(iteratee));
  ***REMOVED***

  /**
   * Reduces `collection` to a value which is the accumulated result of running
   * each element in `collection` thru `iteratee`, where each successive
   * invocation is supplied the return value of the previous. If `accumulator`
   * is not given, the first element of `collection` is used as the initial
   * value. The iteratee is invoked with four arguments:
   * (accumulator, value, index|key, collection).
   *
   * Many lodash methods are guarded to work as iteratees for methods like
   * `_.reduce`, `_.reduceRight`, and `_.transform`.
   *
   * The guarded methods are:
   * `assign`, `defaults`, `defaultsDeep`, `includes`, `merge`, `orderBy`,
   * and `sortBy`
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Collection
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** [iteratee=_.identity] The function invoked per iteration.
   * @param ***REMOVED*******REMOVED*** [accumulator] The initial value.
   * @returns ***REMOVED*******REMOVED*** Returns the accumulated value.
   * @see _.reduceRight
   * @example
   *
   * _.reduce([1, 2], function(sum, n) ***REMOVED***
   *   return sum + n;
   * ***REMOVED***, 0);
   * // => 3
   *
   * _.reduce(***REMOVED*** 'a': 1, 'b': 2, 'c': 1 ***REMOVED***, function(result, value, key) ***REMOVED***
   *   (result[value] || (result[value] = [])).push(key);
   *   return result;
   * ***REMOVED***, ***REMOVED******REMOVED***);
   * // => ***REMOVED*** '1': ['a', 'c'], '2': ['b'] ***REMOVED*** (iteration order is not guaranteed)
   */
  function reduce(collection, iteratee, accumulator) ***REMOVED***
    return baseReduce(collection, baseIteratee(iteratee), accumulator, arguments.length < 3, baseEach);
  ***REMOVED***

  /**
   * Gets the size of `collection` by returning its length for array-like
   * values or the number of own enumerable string keyed properties for objects.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Collection
   * @param ***REMOVED***Array|Object|string***REMOVED*** collection The collection to inspect.
   * @returns ***REMOVED***number***REMOVED*** Returns the collection size.
   * @example
   *
   * _.size([1, 2, 3]);
   * // => 3
   *
   * _.size(***REMOVED*** 'a': 1, 'b': 2 ***REMOVED***);
   * // => 2
   *
   * _.size('pebbles');
   * // => 7
   */
  function size(collection) ***REMOVED***
    if (collection == null) ***REMOVED***
      return 0;
    ***REMOVED***
    collection = isArrayLike(collection) ? collection : nativeKeys(collection);
    return collection.length;
  ***REMOVED***

  /**
   * Checks if `predicate` returns truthy for **any** element of `collection`.
   * Iteration is stopped once `predicate` returns truthy. The predicate is
   * invoked with three arguments: (value, index|key, collection).
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Collection
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***Function***REMOVED*** [predicate=_.identity] The function invoked per iteration.
   * @param- ***REMOVED***Object***REMOVED*** [guard] Enables use as an iteratee for methods like `_.map`.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if any element passes the predicate check,
   *  else `false`.
   * @example
   *
   * _.some([null, 0, 'yes', false], Boolean);
   * // => true
   *
   * var users = [
   *   ***REMOVED*** 'user': 'barney', 'active': true ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred',   'active': false ***REMOVED***
   * ];
   *
   * // The `_.matches` iteratee shorthand.
   * _.some(users, ***REMOVED*** 'user': 'barney', 'active': false ***REMOVED***);
   * // => false
   *
   * // The `_.matchesProperty` iteratee shorthand.
   * _.some(users, ['active', false]);
   * // => true
   *
   * // The `_.property` iteratee shorthand.
   * _.some(users, 'active');
   * // => true
   */
  function some(collection, predicate, guard) ***REMOVED***
    predicate = guard ? undefined : predicate;
    return baseSome(collection, baseIteratee(predicate));
  ***REMOVED***

  /**
   * Creates an array of elements, sorted in ascending order by the results of
   * running each element in a collection thru each iteratee. This method
   * performs a stable sort, that is, it preserves the original sort order of
   * equal elements. The iteratees are invoked with one argument: (value).
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Collection
   * @param ***REMOVED***Array|Object***REMOVED*** collection The collection to iterate over.
   * @param ***REMOVED***...(Function|Function[])***REMOVED*** [iteratees=[_.identity]]
   *  The iteratees to sort by.
   * @returns ***REMOVED***Array***REMOVED*** Returns the new sorted array.
   * @example
   *
   * var users = [
   *   ***REMOVED*** 'user': 'fred',   'age': 48 ***REMOVED***,
   *   ***REMOVED*** 'user': 'barney', 'age': 36 ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred',   'age': 40 ***REMOVED***,
   *   ***REMOVED*** 'user': 'barney', 'age': 34 ***REMOVED***
   * ];
   *
   * _.sortBy(users, [function(o) ***REMOVED*** return o.user; ***REMOVED***]);
   * // => objects for [['barney', 36], ['barney', 34], ['fred', 48], ['fred', 40]]
   *
   * _.sortBy(users, ['user', 'age']);
   * // => objects for [['barney', 34], ['barney', 36], ['fred', 40], ['fred', 48]]
   */
  function sortBy(collection, iteratee) ***REMOVED***
    var index = 0;
    iteratee = baseIteratee(iteratee);

    return baseMap(baseMap(collection, function(value, key, collection) ***REMOVED***
      return ***REMOVED*** 'value': value, 'index': index++, 'criteria': iteratee(value, key, collection) ***REMOVED***;
    ***REMOVED***).sort(function(object, other) ***REMOVED***
      return compareAscending(object.criteria, other.criteria) || (object.index - other.index);
    ***REMOVED***), baseProperty('value'));
  ***REMOVED***

  /*------------------------------------------------------------------------*/

  /**
   * Creates a function that invokes `func`, with the `this` binding and arguments
   * of the created function, while it's called less than `n` times. Subsequent
   * calls to the created function return the result of the last `func` invocation.
   *
   * @static
   * @memberOf _
   * @since 3.0.0
   * @category Function
   * @param ***REMOVED***number***REMOVED*** n The number of calls at which `func` is no longer invoked.
   * @param ***REMOVED***Function***REMOVED*** func The function to restrict.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new restricted function.
   * @example
   *
   * jQuery(element).on('click', _.before(5, addContactToList));
   * // => Allows adding up to 4 contacts to the list.
   */
  function before(n, func) ***REMOVED***
    var result;
    if (typeof func != 'function') ***REMOVED***
      throw new TypeError(FUNC_ERROR_TEXT);
    ***REMOVED***
    n = toInteger(n);
    return function() ***REMOVED***
      if (--n > 0) ***REMOVED***
        result = func.apply(this, arguments);
      ***REMOVED***
      if (n <= 1) ***REMOVED***
        func = undefined;
      ***REMOVED***
      return result;
    ***REMOVED***;
  ***REMOVED***

  /**
   * Creates a function that invokes `func` with the `this` binding of `thisArg`
   * and `partials` prepended to the arguments it receives.
   *
   * The `_.bind.placeholder` value, which defaults to `_` in monolithic builds,
   * may be used as a placeholder for partially applied arguments.
   *
   * **Note:** Unlike native `Function#bind`, this method doesn't set the "length"
   * property of bound functions.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Function
   * @param ***REMOVED***Function***REMOVED*** func The function to bind.
   * @param ***REMOVED*******REMOVED*** thisArg The `this` binding of `func`.
   * @param ***REMOVED***...****REMOVED*** [partials] The arguments to be partially applied.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new bound function.
   * @example
   *
   * function greet(greeting, punctuation) ***REMOVED***
   *   return greeting + ' ' + this.user + punctuation;
   * ***REMOVED***
   *
   * var object = ***REMOVED*** 'user': 'fred' ***REMOVED***;
   *
   * var bound = _.bind(greet, object, 'hi');
   * bound('!');
   * // => 'hi fred!'
   *
   * // Bound with placeholders.
   * var bound = _.bind(greet, object, _, '!');
   * bound('hi');
   * // => 'hi fred!'
   */
  var bind = baseRest(function(func, thisArg, partials) ***REMOVED***
    return createPartial(func, WRAP_BIND_FLAG | WRAP_PARTIAL_FLAG, thisArg, partials);
  ***REMOVED***);

  /**
   * Defers invoking the `func` until the current call stack has cleared. Any
   * additional arguments are provided to `func` when it's invoked.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Function
   * @param ***REMOVED***Function***REMOVED*** func The function to defer.
   * @param ***REMOVED***...****REMOVED*** [args] The arguments to invoke `func` with.
   * @returns ***REMOVED***number***REMOVED*** Returns the timer id.
   * @example
   *
   * _.defer(function(text) ***REMOVED***
   *   console.log(text);
   * ***REMOVED***, 'deferred');
   * // => Logs 'deferred' after one millisecond.
   */
  var defer = baseRest(function(func, args) ***REMOVED***
    return baseDelay(func, 1, args);
  ***REMOVED***);

  /**
   * Invokes `func` after `wait` milliseconds. Any additional arguments are
   * provided to `func` when it's invoked.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Function
   * @param ***REMOVED***Function***REMOVED*** func The function to delay.
   * @param ***REMOVED***number***REMOVED*** wait The number of milliseconds to delay invocation.
   * @param ***REMOVED***...****REMOVED*** [args] The arguments to invoke `func` with.
   * @returns ***REMOVED***number***REMOVED*** Returns the timer id.
   * @example
   *
   * _.delay(function(text) ***REMOVED***
   *   console.log(text);
   * ***REMOVED***, 1000, 'later');
   * // => Logs 'later' after one second.
   */
  var delay = baseRest(function(func, wait, args) ***REMOVED***
    return baseDelay(func, toNumber(wait) || 0, args);
  ***REMOVED***);

  /**
   * Creates a function that negates the result of the predicate `func`. The
   * `func` predicate is invoked with the `this` binding and arguments of the
   * created function.
   *
   * @static
   * @memberOf _
   * @since 3.0.0
   * @category Function
   * @param ***REMOVED***Function***REMOVED*** predicate The predicate to negate.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new negated function.
   * @example
   *
   * function isEven(n) ***REMOVED***
   *   return n % 2 == 0;
   * ***REMOVED***
   *
   * _.filter([1, 2, 3, 4, 5, 6], _.negate(isEven));
   * // => [1, 3, 5]
   */
  function negate(predicate) ***REMOVED***
    if (typeof predicate != 'function') ***REMOVED***
      throw new TypeError(FUNC_ERROR_TEXT);
    ***REMOVED***
    return function() ***REMOVED***
      var args = arguments;
      return !predicate.apply(this, args);
    ***REMOVED***;
  ***REMOVED***

  /**
   * Creates a function that is restricted to invoking `func` once. Repeat calls
   * to the function return the value of the first invocation. The `func` is
   * invoked with the `this` binding and arguments of the created function.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Function
   * @param ***REMOVED***Function***REMOVED*** func The function to restrict.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new restricted function.
   * @example
   *
   * var initialize = _.once(createApplication);
   * initialize();
   * initialize();
   * // => `createApplication` is invoked once
   */
  function once(func) ***REMOVED***
    return before(2, func);
  ***REMOVED***

  /*------------------------------------------------------------------------*/

  /**
   * Creates a shallow clone of `value`.
   *
   * **Note:** This method is loosely based on the
   * [structured clone algorithm](https://mdn.io/Structured_clone_algorithm)
   * and supports cloning arrays, array buffers, booleans, date objects, maps,
   * numbers, `Object` objects, regexes, sets, strings, symbols, and typed
   * arrays. The own enumerable properties of `arguments` objects are cloned
   * as plain objects. An empty object is returned for uncloneable values such
   * as error objects, functions, DOM nodes, and WeakMaps.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to clone.
   * @returns ***REMOVED*******REMOVED*** Returns the cloned value.
   * @see _.cloneDeep
   * @example
   *
   * var objects = [***REMOVED*** 'a': 1 ***REMOVED***, ***REMOVED*** 'b': 2 ***REMOVED***];
   *
   * var shallow = _.clone(objects);
   * console.log(shallow[0] === objects[0]);
   * // => true
   */
  function clone(value) ***REMOVED***
    if (!isObject(value)) ***REMOVED***
      return value;
    ***REMOVED***
    return isArray(value) ? copyArray(value) : copyObject(value, nativeKeys(value));
  ***REMOVED***

  /**
   * Performs a
   * [`SameValueZero`](http://ecma-international.org/ecma-262/7.0/#sec-samevaluezero)
   * comparison between two values to determine if they are equivalent.
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to compare.
   * @param ***REMOVED*******REMOVED*** other The other value to compare.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the values are equivalent, else `false`.
   * @example
   *
   * var object = ***REMOVED*** 'a': 1 ***REMOVED***;
   * var other = ***REMOVED*** 'a': 1 ***REMOVED***;
   *
   * _.eq(object, object);
   * // => true
   *
   * _.eq(object, other);
   * // => false
   *
   * _.eq('a', 'a');
   * // => true
   *
   * _.eq('a', Object('a'));
   * // => false
   *
   * _.eq(NaN, NaN);
   * // => true
   */
  function eq(value, other) ***REMOVED***
    return value === other || (value !== value && other !== other);
  ***REMOVED***

  /**
   * Checks if `value` is likely an `arguments` object.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is an `arguments` object,
   *  else `false`.
   * @example
   *
   * _.isArguments(function() ***REMOVED*** return arguments; ***REMOVED***());
   * // => true
   *
   * _.isArguments([1, 2, 3]);
   * // => false
   */
  var isArguments = baseIsArguments(function() ***REMOVED*** return arguments; ***REMOVED***()) ? baseIsArguments : function(value) ***REMOVED***
    return isObjectLike(value) && hasOwnProperty.call(value, 'callee') &&
      !propertyIsEnumerable.call(value, 'callee');
  ***REMOVED***;

  /**
   * Checks if `value` is classified as an `Array` object.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is an array, else `false`.
   * @example
   *
   * _.isArray([1, 2, 3]);
   * // => true
   *
   * _.isArray(document.body.children);
   * // => false
   *
   * _.isArray('abc');
   * // => false
   *
   * _.isArray(_.noop);
   * // => false
   */
  var isArray = Array.isArray;

  /**
   * Checks if `value` is array-like. A value is considered array-like if it's
   * not a function and has a `value.length` that's an integer greater than or
   * equal to `0` and less than or equal to `Number.MAX_SAFE_INTEGER`.
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is array-like, else `false`.
   * @example
   *
   * _.isArrayLike([1, 2, 3]);
   * // => true
   *
   * _.isArrayLike(document.body.children);
   * // => true
   *
   * _.isArrayLike('abc');
   * // => true
   *
   * _.isArrayLike(_.noop);
   * // => false
   */
  function isArrayLike(value) ***REMOVED***
    return value != null && isLength(value.length) && !isFunction(value);
  ***REMOVED***

  /**
   * Checks if `value` is classified as a boolean primitive or object.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a boolean, else `false`.
   * @example
   *
   * _.isBoolean(false);
   * // => true
   *
   * _.isBoolean(null);
   * // => false
   */
  function isBoolean(value) ***REMOVED***
    return value === true || value === false ||
      (isObjectLike(value) && baseGetTag(value) == boolTag);
  ***REMOVED***

  /**
   * Checks if `value` is classified as a `Date` object.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a date object, else `false`.
   * @example
   *
   * _.isDate(new Date);
   * // => true
   *
   * _.isDate('Mon April 23 2012');
   * // => false
   */
  var isDate = baseIsDate;

  /**
   * Checks if `value` is an empty object, collection, map, or set.
   *
   * Objects are considered empty if they have no own enumerable string keyed
   * properties.
   *
   * Array-like values such as `arguments` objects, arrays, buffers, strings, or
   * jQuery-like collections are considered empty if they have a `length` of `0`.
   * Similarly, maps and sets are considered empty if they have a `size` of `0`.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is empty, else `false`.
   * @example
   *
   * _.isEmpty(null);
   * // => true
   *
   * _.isEmpty(true);
   * // => true
   *
   * _.isEmpty(1);
   * // => true
   *
   * _.isEmpty([1, 2, 3]);
   * // => false
   *
   * _.isEmpty(***REMOVED*** 'a': 1 ***REMOVED***);
   * // => false
   */
  function isEmpty(value) ***REMOVED***
    if (isArrayLike(value) &&
        (isArray(value) || isString(value) ||
          isFunction(value.splice) || isArguments(value))) ***REMOVED***
      return !value.length;
    ***REMOVED***
    return !nativeKeys(value).length;
  ***REMOVED***

  /**
   * Performs a deep comparison between two values to determine if they are
   * equivalent.
   *
   * **Note:** This method supports comparing arrays, array buffers, booleans,
   * date objects, error objects, maps, numbers, `Object` objects, regexes,
   * sets, strings, symbols, and typed arrays. `Object` objects are compared
   * by their own, not inherited, enumerable properties. Functions and DOM
   * nodes are compared by strict equality, i.e. `===`.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to compare.
   * @param ***REMOVED*******REMOVED*** other The other value to compare.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the values are equivalent, else `false`.
   * @example
   *
   * var object = ***REMOVED*** 'a': 1 ***REMOVED***;
   * var other = ***REMOVED*** 'a': 1 ***REMOVED***;
   *
   * _.isEqual(object, other);
   * // => true
   *
   * object === other;
   * // => false
   */
  function isEqual(value, other) ***REMOVED***
    return baseIsEqual(value, other);
  ***REMOVED***

  /**
   * Checks if `value` is a finite primitive number.
   *
   * **Note:** This method is based on
   * [`Number.isFinite`](https://mdn.io/Number/isFinite).
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a finite number, else `false`.
   * @example
   *
   * _.isFinite(3);
   * // => true
   *
   * _.isFinite(Number.MIN_VALUE);
   * // => true
   *
   * _.isFinite(Infinity);
   * // => false
   *
   * _.isFinite('3');
   * // => false
   */
  function isFinite(value) ***REMOVED***
    return typeof value == 'number' && nativeIsFinite(value);
  ***REMOVED***

  /**
   * Checks if `value` is classified as a `Function` object.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a function, else `false`.
   * @example
   *
   * _.isFunction(_);
   * // => true
   *
   * _.isFunction(/abc/);
   * // => false
   */
  function isFunction(value) ***REMOVED***
    if (!isObject(value)) ***REMOVED***
      return false;
    ***REMOVED***
    // The use of `Object#toString` avoids issues with the `typeof` operator
    // in Safari 9 which returns 'object' for typed arrays and other constructors.
    var tag = baseGetTag(value);
    return tag == funcTag || tag == genTag || tag == asyncTag || tag == proxyTag;
  ***REMOVED***

  /**
   * Checks if `value` is a valid array-like length.
   *
   * **Note:** This method is loosely based on
   * [`ToLength`](http://ecma-international.org/ecma-262/7.0/#sec-tolength).
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a valid length, else `false`.
   * @example
   *
   * _.isLength(3);
   * // => true
   *
   * _.isLength(Number.MIN_VALUE);
   * // => false
   *
   * _.isLength(Infinity);
   * // => false
   *
   * _.isLength('3');
   * // => false
   */
  function isLength(value) ***REMOVED***
    return typeof value == 'number' &&
      value > -1 && value % 1 == 0 && value <= MAX_SAFE_INTEGER;
  ***REMOVED***

  /**
   * Checks if `value` is the
   * [language type](http://www.ecma-international.org/ecma-262/7.0/#sec-ecmascript-language-types)
   * of `Object`. (e.g. arrays, functions, objects, regexes, `new Number(0)`, and `new String('')`)
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is an object, else `false`.
   * @example
   *
   * _.isObject(***REMOVED******REMOVED***);
   * // => true
   *
   * _.isObject([1, 2, 3]);
   * // => true
   *
   * _.isObject(_.noop);
   * // => true
   *
   * _.isObject(null);
   * // => false
   */
  function isObject(value) ***REMOVED***
    var type = typeof value;
    return value != null && (type == 'object' || type == 'function');
  ***REMOVED***

  /**
   * Checks if `value` is object-like. A value is object-like if it's not `null`
   * and has a `typeof` result of "object".
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is object-like, else `false`.
   * @example
   *
   * _.isObjectLike(***REMOVED******REMOVED***);
   * // => true
   *
   * _.isObjectLike([1, 2, 3]);
   * // => true
   *
   * _.isObjectLike(_.noop);
   * // => false
   *
   * _.isObjectLike(null);
   * // => false
   */
  function isObjectLike(value) ***REMOVED***
    return value != null && typeof value == 'object';
  ***REMOVED***

  /**
   * Checks if `value` is `NaN`.
   *
   * **Note:** This method is based on
   * [`Number.isNaN`](https://mdn.io/Number/isNaN) and is not the same as
   * global [`isNaN`](https://mdn.io/isNaN) which returns `true` for
   * `undefined` and other non-number values.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is `NaN`, else `false`.
   * @example
   *
   * _.isNaN(NaN);
   * // => true
   *
   * _.isNaN(new Number(NaN));
   * // => true
   *
   * isNaN(undefined);
   * // => true
   *
   * _.isNaN(undefined);
   * // => false
   */
  function isNaN(value) ***REMOVED***
    // An `NaN` primitive is the only value that is not equal to itself.
    // Perform the `toStringTag` check first to avoid errors with some
    // ActiveX objects in IE.
    return isNumber(value) && value != +value;
  ***REMOVED***

  /**
   * Checks if `value` is `null`.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is `null`, else `false`.
   * @example
   *
   * _.isNull(null);
   * // => true
   *
   * _.isNull(void 0);
   * // => false
   */
  function isNull(value) ***REMOVED***
    return value === null;
  ***REMOVED***

  /**
   * Checks if `value` is classified as a `Number` primitive or object.
   *
   * **Note:** To exclude `Infinity`, `-Infinity`, and `NaN`, which are
   * classified as numbers, use the `_.isFinite` method.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a number, else `false`.
   * @example
   *
   * _.isNumber(3);
   * // => true
   *
   * _.isNumber(Number.MIN_VALUE);
   * // => true
   *
   * _.isNumber(Infinity);
   * // => true
   *
   * _.isNumber('3');
   * // => false
   */
  function isNumber(value) ***REMOVED***
    return typeof value == 'number' ||
      (isObjectLike(value) && baseGetTag(value) == numberTag);
  ***REMOVED***

  /**
   * Checks if `value` is classified as a `RegExp` object.
   *
   * @static
   * @memberOf _
   * @since 0.1.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a regexp, else `false`.
   * @example
   *
   * _.isRegExp(/abc/);
   * // => true
   *
   * _.isRegExp('/abc/');
   * // => false
   */
  var isRegExp = baseIsRegExp;

  /**
   * Checks if `value` is classified as a `String` primitive or object.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a string, else `false`.
   * @example
   *
   * _.isString('abc');
   * // => true
   *
   * _.isString(1);
   * // => false
   */
  function isString(value) ***REMOVED***
    return typeof value == 'string' ||
      (!isArray(value) && isObjectLike(value) && baseGetTag(value) == stringTag);
  ***REMOVED***

  /**
   * Checks if `value` is `undefined`.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is `undefined`, else `false`.
   * @example
   *
   * _.isUndefined(void 0);
   * // => true
   *
   * _.isUndefined(null);
   * // => false
   */
  function isUndefined(value) ***REMOVED***
    return value === undefined;
  ***REMOVED***

  /**
   * Converts `value` to an array.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to convert.
   * @returns ***REMOVED***Array***REMOVED*** Returns the converted array.
   * @example
   *
   * _.toArray(***REMOVED*** 'a': 1, 'b': 2 ***REMOVED***);
   * // => [1, 2]
   *
   * _.toArray('abc');
   * // => ['a', 'b', 'c']
   *
   * _.toArray(1);
   * // => []
   *
   * _.toArray(null);
   * // => []
   */
  function toArray(value) ***REMOVED***
    if (!isArrayLike(value)) ***REMOVED***
      return values(value);
    ***REMOVED***
    return value.length ? copyArray(value) : [];
  ***REMOVED***

  /**
   * Converts `value` to an integer.
   *
   * **Note:** This method is loosely based on
   * [`ToInteger`](http://www.ecma-international.org/ecma-262/7.0/#sec-tointeger).
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to convert.
   * @returns ***REMOVED***number***REMOVED*** Returns the converted integer.
   * @example
   *
   * _.toInteger(3.2);
   * // => 3
   *
   * _.toInteger(Number.MIN_VALUE);
   * // => 0
   *
   * _.toInteger(Infinity);
   * // => 1.7976931348623157e+308
   *
   * _.toInteger('3.2');
   * // => 3
   */
  var toInteger = Number;

  /**
   * Converts `value` to a number.
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to process.
   * @returns ***REMOVED***number***REMOVED*** Returns the number.
   * @example
   *
   * _.toNumber(3.2);
   * // => 3.2
   *
   * _.toNumber(Number.MIN_VALUE);
   * // => 5e-324
   *
   * _.toNumber(Infinity);
   * // => Infinity
   *
   * _.toNumber('3.2');
   * // => 3.2
   */
  var toNumber = Number;

  /**
   * Converts `value` to a string. An empty string is returned for `null`
   * and `undefined` values. The sign of `-0` is preserved.
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @category Lang
   * @param ***REMOVED*******REMOVED*** value The value to convert.
   * @returns ***REMOVED***string***REMOVED*** Returns the converted string.
   * @example
   *
   * _.toString(null);
   * // => ''
   *
   * _.toString(-0);
   * // => '-0'
   *
   * _.toString([1, 2, 3]);
   * // => '1,2,3'
   */
  function toString(value) ***REMOVED***
    if (typeof value == 'string') ***REMOVED***
      return value;
    ***REMOVED***
    return value == null ? '' : (value + '');
  ***REMOVED***

  /*------------------------------------------------------------------------*/

  /**
   * Assigns own enumerable string keyed properties of source objects to the
   * destination object. Source objects are applied from left to right.
   * Subsequent sources overwrite property assignments of previous sources.
   *
   * **Note:** This method mutates `object` and is loosely based on
   * [`Object.assign`](https://mdn.io/Object/assign).
   *
   * @static
   * @memberOf _
   * @since 0.10.0
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The destination object.
   * @param ***REMOVED***...Object***REMOVED*** [sources] The source objects.
   * @returns ***REMOVED***Object***REMOVED*** Returns `object`.
   * @see _.assignIn
   * @example
   *
   * function Foo() ***REMOVED***
   *   this.a = 1;
   * ***REMOVED***
   *
   * function Bar() ***REMOVED***
   *   this.c = 3;
   * ***REMOVED***
   *
   * Foo.prototype.b = 2;
   * Bar.prototype.d = 4;
   *
   * _.assign(***REMOVED*** 'a': 0 ***REMOVED***, new Foo, new Bar);
   * // => ***REMOVED*** 'a': 1, 'c': 3 ***REMOVED***
   */
  var assign = createAssigner(function(object, source) ***REMOVED***
    copyObject(source, nativeKeys(source), object);
  ***REMOVED***);

  /**
   * This method is like `_.assign` except that it iterates over own and
   * inherited source properties.
   *
   * **Note:** This method mutates `object`.
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @alias extend
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The destination object.
   * @param ***REMOVED***...Object***REMOVED*** [sources] The source objects.
   * @returns ***REMOVED***Object***REMOVED*** Returns `object`.
   * @see _.assign
   * @example
   *
   * function Foo() ***REMOVED***
   *   this.a = 1;
   * ***REMOVED***
   *
   * function Bar() ***REMOVED***
   *   this.c = 3;
   * ***REMOVED***
   *
   * Foo.prototype.b = 2;
   * Bar.prototype.d = 4;
   *
   * _.assignIn(***REMOVED*** 'a': 0 ***REMOVED***, new Foo, new Bar);
   * // => ***REMOVED*** 'a': 1, 'b': 2, 'c': 3, 'd': 4 ***REMOVED***
   */
  var assignIn = createAssigner(function(object, source) ***REMOVED***
    copyObject(source, nativeKeysIn(source), object);
  ***REMOVED***);

  /**
   * This method is like `_.assignIn` except that it accepts `customizer`
   * which is invoked to produce the assigned values. If `customizer` returns
   * `undefined`, assignment is handled by the method instead. The `customizer`
   * is invoked with five arguments: (objValue, srcValue, key, object, source).
   *
   * **Note:** This method mutates `object`.
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @alias extendWith
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The destination object.
   * @param ***REMOVED***...Object***REMOVED*** sources The source objects.
   * @param ***REMOVED***Function***REMOVED*** [customizer] The function to customize assigned values.
   * @returns ***REMOVED***Object***REMOVED*** Returns `object`.
   * @see _.assignWith
   * @example
   *
   * function customizer(objValue, srcValue) ***REMOVED***
   *   return _.isUndefined(objValue) ? srcValue : objValue;
   * ***REMOVED***
   *
   * var defaults = _.partialRight(_.assignInWith, customizer);
   *
   * defaults(***REMOVED*** 'a': 1 ***REMOVED***, ***REMOVED*** 'b': 2 ***REMOVED***, ***REMOVED*** 'a': 3 ***REMOVED***);
   * // => ***REMOVED*** 'a': 1, 'b': 2 ***REMOVED***
   */
  var assignInWith = createAssigner(function(object, source, srcIndex, customizer) ***REMOVED***
    copyObject(source, keysIn(source), object, customizer);
  ***REMOVED***);

  /**
   * Creates an object that inherits from the `prototype` object. If a
   * `properties` object is given, its own enumerable string keyed properties
   * are assigned to the created object.
   *
   * @static
   * @memberOf _
   * @since 2.3.0
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** prototype The object to inherit from.
   * @param ***REMOVED***Object***REMOVED*** [properties] The properties to assign to the object.
   * @returns ***REMOVED***Object***REMOVED*** Returns the new object.
   * @example
   *
   * function Shape() ***REMOVED***
   *   this.x = 0;
   *   this.y = 0;
   * ***REMOVED***
   *
   * function Circle() ***REMOVED***
   *   Shape.call(this);
   * ***REMOVED***
   *
   * Circle.prototype = _.create(Shape.prototype, ***REMOVED***
   *   'constructor': Circle
   * ***REMOVED***);
   *
   * var circle = new Circle;
   * circle instanceof Circle;
   * // => true
   *
   * circle instanceof Shape;
   * // => true
   */
  function create(prototype, properties) ***REMOVED***
    var result = baseCreate(prototype);
    return properties == null ? result : assign(result, properties);
  ***REMOVED***

  /**
   * Assigns own and inherited enumerable string keyed properties of source
   * objects to the destination object for all destination properties that
   * resolve to `undefined`. Source objects are applied from left to right.
   * Once a property is set, additional values of the same property are ignored.
   *
   * **Note:** This method mutates `object`.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The destination object.
   * @param ***REMOVED***...Object***REMOVED*** [sources] The source objects.
   * @returns ***REMOVED***Object***REMOVED*** Returns `object`.
   * @see _.defaultsDeep
   * @example
   *
   * _.defaults(***REMOVED*** 'a': 1 ***REMOVED***, ***REMOVED*** 'b': 2 ***REMOVED***, ***REMOVED*** 'a': 3 ***REMOVED***);
   * // => ***REMOVED*** 'a': 1, 'b': 2 ***REMOVED***
   */
  var defaults = baseRest(function(args) ***REMOVED***
    args.push(undefined, customDefaultsAssignIn);
    return assignInWith.apply(undefined, args);
  ***REMOVED***);

  /**
   * Checks if `path` is a direct property of `object`.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The object to query.
   * @param ***REMOVED***Array|string***REMOVED*** path The path to check.
   * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `path` exists, else `false`.
   * @example
   *
   * var object = ***REMOVED*** 'a': ***REMOVED*** 'b': 2 ***REMOVED*** ***REMOVED***;
   * var other = _.create(***REMOVED*** 'a': _.create(***REMOVED*** 'b': 2 ***REMOVED***) ***REMOVED***);
   *
   * _.has(object, 'a');
   * // => true
   *
   * _.has(object, 'a.b');
   * // => true
   *
   * _.has(object, ['a', 'b']);
   * // => true
   *
   * _.has(other, 'a');
   * // => false
   */
  function has(object, path) ***REMOVED***
    return object != null && hasOwnProperty.call(object, path);
  ***REMOVED***

  /**
   * Creates an array of the own enumerable property names of `object`.
   *
   * **Note:** Non-object values are coerced to objects. See the
   * [ES spec](http://ecma-international.org/ecma-262/7.0/#sec-object.keys)
   * for more details.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The object to query.
   * @returns ***REMOVED***Array***REMOVED*** Returns the array of property names.
   * @example
   *
   * function Foo() ***REMOVED***
   *   this.a = 1;
   *   this.b = 2;
   * ***REMOVED***
   *
   * Foo.prototype.c = 3;
   *
   * _.keys(new Foo);
   * // => ['a', 'b'] (iteration order is not guaranteed)
   *
   * _.keys('hi');
   * // => ['0', '1']
   */
  var keys = nativeKeys;

  /**
   * Creates an array of the own and inherited enumerable property names of `object`.
   *
   * **Note:** Non-object values are coerced to objects.
   *
   * @static
   * @memberOf _
   * @since 3.0.0
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The object to query.
   * @returns ***REMOVED***Array***REMOVED*** Returns the array of property names.
   * @example
   *
   * function Foo() ***REMOVED***
   *   this.a = 1;
   *   this.b = 2;
   * ***REMOVED***
   *
   * Foo.prototype.c = 3;
   *
   * _.keysIn(new Foo);
   * // => ['a', 'b', 'c'] (iteration order is not guaranteed)
   */
  var keysIn = nativeKeysIn;

  /**
   * Creates an object composed of the picked `object` properties.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The source object.
   * @param ***REMOVED***...(string|string[])***REMOVED*** [paths] The property paths to pick.
   * @returns ***REMOVED***Object***REMOVED*** Returns the new object.
   * @example
   *
   * var object = ***REMOVED*** 'a': 1, 'b': '2', 'c': 3 ***REMOVED***;
   *
   * _.pick(object, ['a', 'c']);
   * // => ***REMOVED*** 'a': 1, 'c': 3 ***REMOVED***
   */
  var pick = flatRest(function(object, paths) ***REMOVED***
    return object == null ? ***REMOVED******REMOVED*** : basePick(object, paths);
  ***REMOVED***);

  /**
   * This method is like `_.get` except that if the resolved value is a
   * function it's invoked with the `this` binding of its parent object and
   * its result is returned.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The object to query.
   * @param ***REMOVED***Array|string***REMOVED*** path The path of the property to resolve.
   * @param ***REMOVED*******REMOVED*** [defaultValue] The value returned for `undefined` resolved values.
   * @returns ***REMOVED*******REMOVED*** Returns the resolved value.
   * @example
   *
   * var object = ***REMOVED*** 'a': [***REMOVED*** 'b': ***REMOVED*** 'c1': 3, 'c2': _.constant(4) ***REMOVED*** ***REMOVED***] ***REMOVED***;
   *
   * _.result(object, 'a[0].b.c1');
   * // => 3
   *
   * _.result(object, 'a[0].b.c2');
   * // => 4
   *
   * _.result(object, 'a[0].b.c3', 'default');
   * // => 'default'
   *
   * _.result(object, 'a[0].b.c3', _.constant('default'));
   * // => 'default'
   */
  function result(object, path, defaultValue) ***REMOVED***
    var value = object == null ? undefined : object[path];
    if (value === undefined) ***REMOVED***
      value = defaultValue;
    ***REMOVED***
    return isFunction(value) ? value.call(object) : value;
  ***REMOVED***

  /**
   * Creates an array of the own enumerable string keyed property values of `object`.
   *
   * **Note:** Non-object values are coerced to objects.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Object
   * @param ***REMOVED***Object***REMOVED*** object The object to query.
   * @returns ***REMOVED***Array***REMOVED*** Returns the array of property values.
   * @example
   *
   * function Foo() ***REMOVED***
   *   this.a = 1;
   *   this.b = 2;
   * ***REMOVED***
   *
   * Foo.prototype.c = 3;
   *
   * _.values(new Foo);
   * // => [1, 2] (iteration order is not guaranteed)
   *
   * _.values('hi');
   * // => ['h', 'i']
   */
  function values(object) ***REMOVED***
    return object == null ? [] : baseValues(object, keys(object));
  ***REMOVED***

  /*------------------------------------------------------------------------*/

  /**
   * Converts the characters "&", "<", ">", '"', and "'" in `string` to their
   * corresponding HTML entities.
   *
   * **Note:** No other characters are escaped. To escape additional
   * characters use a third-party library like [_he_](https://mths.be/he).
   *
   * Though the ">" character is escaped for symmetry, characters like
   * ">" and "/" don't need escaping in HTML and have no special meaning
   * unless they're part of a tag or unquoted attribute value. See
   * [Mathias Bynens's article](https://mathiasbynens.be/notes/ambiguous-ampersands)
   * (under "semi-related fun fact") for more details.
   *
   * When working with HTML you should always
   * [quote attribute values](http://wonko.com/post/html-escaping) to reduce
   * XSS vectors.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category String
   * @param ***REMOVED***string***REMOVED*** [string=''] The string to escape.
   * @returns ***REMOVED***string***REMOVED*** Returns the escaped string.
   * @example
   *
   * _.escape('fred, barney, & pebbles');
   * // => 'fred, barney, &amp; pebbles'
   */
  function escape(string) ***REMOVED***
    string = toString(string);
    return (string && reHasUnescapedHtml.test(string))
      ? string.replace(reUnescapedHtml, escapeHtmlChar)
      : string;
  ***REMOVED***

  /*------------------------------------------------------------------------*/

  /**
   * This method returns the first argument it receives.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Util
   * @param ***REMOVED*******REMOVED*** value Any value.
   * @returns ***REMOVED*******REMOVED*** Returns `value`.
   * @example
   *
   * var object = ***REMOVED*** 'a': 1 ***REMOVED***;
   *
   * console.log(_.identity(object) === object);
   * // => true
   */
  function identity(value) ***REMOVED***
    return value;
  ***REMOVED***

  /**
   * Creates a function that invokes `func` with the arguments of the created
   * function. If `func` is a property name, the created function returns the
   * property value for a given element. If `func` is an array or object, the
   * created function returns `true` for elements that contain the equivalent
   * source properties, otherwise it returns `false`.
   *
   * @static
   * @since 4.0.0
   * @memberOf _
   * @category Util
   * @param ***REMOVED*******REMOVED*** [func=_.identity] The value to convert to a callback.
   * @returns ***REMOVED***Function***REMOVED*** Returns the callback.
   * @example
   *
   * var users = [
   *   ***REMOVED*** 'user': 'barney', 'age': 36, 'active': true ***REMOVED***,
   *   ***REMOVED*** 'user': 'fred',   'age': 40, 'active': false ***REMOVED***
   * ];
   *
   * // The `_.matches` iteratee shorthand.
   * _.filter(users, _.iteratee(***REMOVED*** 'user': 'barney', 'active': true ***REMOVED***));
   * // => [***REMOVED*** 'user': 'barney', 'age': 36, 'active': true ***REMOVED***]
   *
   * // The `_.matchesProperty` iteratee shorthand.
   * _.filter(users, _.iteratee(['user', 'fred']));
   * // => [***REMOVED*** 'user': 'fred', 'age': 40 ***REMOVED***]
   *
   * // The `_.property` iteratee shorthand.
   * _.map(users, _.iteratee('user'));
   * // => ['barney', 'fred']
   *
   * // Create custom iteratee shorthands.
   * _.iteratee = _.wrap(_.iteratee, function(iteratee, func) ***REMOVED***
   *   return !_.isRegExp(func) ? iteratee(func) : function(string) ***REMOVED***
   *     return func.test(string);
   *   ***REMOVED***;
   * ***REMOVED***);
   *
   * _.filter(['abc', 'def'], /ef/);
   * // => ['def']
   */
  var iteratee = baseIteratee;

  /**
   * Creates a function that performs a partial deep comparison between a given
   * object and `source`, returning `true` if the given object has equivalent
   * property values, else `false`.
   *
   * **Note:** The created function is equivalent to `_.isMatch` with `source`
   * partially applied.
   *
   * Partial comparisons will match empty array and empty object `source`
   * values against any array or object value, respectively. See `_.isEqual`
   * for a list of supported value comparisons.
   *
   * @static
   * @memberOf _
   * @since 3.0.0
   * @category Util
   * @param ***REMOVED***Object***REMOVED*** source The object of property values to match.
   * @returns ***REMOVED***Function***REMOVED*** Returns the new spec function.
   * @example
   *
   * var objects = [
   *   ***REMOVED*** 'a': 1, 'b': 2, 'c': 3 ***REMOVED***,
   *   ***REMOVED*** 'a': 4, 'b': 5, 'c': 6 ***REMOVED***
   * ];
   *
   * _.filter(objects, _.matches(***REMOVED*** 'a': 4, 'c': 6 ***REMOVED***));
   * // => [***REMOVED*** 'a': 4, 'b': 5, 'c': 6 ***REMOVED***]
   */
  function matches(source) ***REMOVED***
    return baseMatches(assign(***REMOVED******REMOVED***, source));
  ***REMOVED***

  /**
   * Adds all own enumerable string keyed function properties of a source
   * object to the destination object. If `object` is a function, then methods
   * are added to its prototype as well.
   *
   * **Note:** Use `_.runInContext` to create a pristine `lodash` function to
   * avoid conflicts caused by modifying the original.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Util
   * @param ***REMOVED***Function|Object***REMOVED*** [object=lodash] The destination object.
   * @param ***REMOVED***Object***REMOVED*** source The object of functions to add.
   * @param ***REMOVED***Object***REMOVED*** [options=***REMOVED******REMOVED***] The options object.
   * @param ***REMOVED***boolean***REMOVED*** [options.chain=true] Specify whether mixins are chainable.
   * @returns ***REMOVED***Function|Object***REMOVED*** Returns `object`.
   * @example
   *
   * function vowels(string) ***REMOVED***
   *   return _.filter(string, function(v) ***REMOVED***
   *     return /[aeiou]/i.test(v);
   *   ***REMOVED***);
   * ***REMOVED***
   *
   * _.mixin(***REMOVED*** 'vowels': vowels ***REMOVED***);
   * _.vowels('fred');
   * // => ['e']
   *
   * _('fred').vowels().value();
   * // => ['e']
   *
   * _.mixin(***REMOVED*** 'vowels': vowels ***REMOVED***, ***REMOVED*** 'chain': false ***REMOVED***);
   * _('fred').vowels();
   * // => ['e']
   */
  function mixin(object, source, options) ***REMOVED***
    var props = keys(source),
        methodNames = baseFunctions(source, props);

    if (options == null &&
        !(isObject(source) && (methodNames.length || !props.length))) ***REMOVED***
      options = source;
      source = object;
      object = this;
      methodNames = baseFunctions(source, keys(source));
    ***REMOVED***
    var chain = !(isObject(options) && 'chain' in options) || !!options.chain,
        isFunc = isFunction(object);

    baseEach(methodNames, function(methodName) ***REMOVED***
      var func = source[methodName];
      object[methodName] = func;
      if (isFunc) ***REMOVED***
        object.prototype[methodName] = function() ***REMOVED***
          var chainAll = this.__chain__;
          if (chain || chainAll) ***REMOVED***
            var result = object(this.__wrapped__),
                actions = result.__actions__ = copyArray(this.__actions__);

            actions.push(***REMOVED*** 'func': func, 'args': arguments, 'thisArg': object ***REMOVED***);
            result.__chain__ = chainAll;
            return result;
          ***REMOVED***
          return func.apply(object, arrayPush([this.value()], arguments));
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***);

    return object;
  ***REMOVED***

  /**
   * Reverts the `_` variable to its previous value and returns a reference to
   * the `lodash` function.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Util
   * @returns ***REMOVED***Function***REMOVED*** Returns the `lodash` function.
   * @example
   *
   * var lodash = _.noConflict();
   */
  function noConflict() ***REMOVED***
    if (root._ === this) ***REMOVED***
      root._ = oldDash;
    ***REMOVED***
    return this;
  ***REMOVED***

  /**
   * This method returns `undefined`.
   *
   * @static
   * @memberOf _
   * @since 2.3.0
   * @category Util
   * @example
   *
   * _.times(2, _.noop);
   * // => [undefined, undefined]
   */
  function noop() ***REMOVED***
    // No operation performed.
  ***REMOVED***

  /**
   * Generates a unique ID. If `prefix` is given, the ID is appended to it.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Util
   * @param ***REMOVED***string***REMOVED*** [prefix=''] The value to prefix the ID with.
   * @returns ***REMOVED***string***REMOVED*** Returns the unique ID.
   * @example
   *
   * _.uniqueId('contact_');
   * // => 'contact_104'
   *
   * _.uniqueId();
   * // => '105'
   */
  function uniqueId(prefix) ***REMOVED***
    var id = ++idCounter;
    return toString(prefix) + id;
  ***REMOVED***

  /*------------------------------------------------------------------------*/

  /**
   * Computes the maximum value of `array`. If `array` is empty or falsey,
   * `undefined` is returned.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Math
   * @param ***REMOVED***Array***REMOVED*** array The array to iterate over.
   * @returns ***REMOVED*******REMOVED*** Returns the maximum value.
   * @example
   *
   * _.max([4, 2, 8, 6]);
   * // => 8
   *
   * _.max([]);
   * // => undefined
   */
  function max(array) ***REMOVED***
    return (array && array.length)
      ? baseExtremum(array, identity, baseGt)
      : undefined;
  ***REMOVED***

  /**
   * Computes the minimum value of `array`. If `array` is empty or falsey,
   * `undefined` is returned.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Math
   * @param ***REMOVED***Array***REMOVED*** array The array to iterate over.
   * @returns ***REMOVED*******REMOVED*** Returns the minimum value.
   * @example
   *
   * _.min([4, 2, 8, 6]);
   * // => 2
   *
   * _.min([]);
   * // => undefined
   */
  function min(array) ***REMOVED***
    return (array && array.length)
      ? baseExtremum(array, identity, baseLt)
      : undefined;
  ***REMOVED***

  /*------------------------------------------------------------------------*/

  // Add methods that return wrapped values in chain sequences.
  lodash.assignIn = assignIn;
  lodash.before = before;
  lodash.bind = bind;
  lodash.chain = chain;
  lodash.compact = compact;
  lodash.concat = concat;
  lodash.create = create;
  lodash.defaults = defaults;
  lodash.defer = defer;
  lodash.delay = delay;
  lodash.filter = filter;
  lodash.flatten = flatten;
  lodash.flattenDeep = flattenDeep;
  lodash.iteratee = iteratee;
  lodash.keys = keys;
  lodash.map = map;
  lodash.matches = matches;
  lodash.mixin = mixin;
  lodash.negate = negate;
  lodash.once = once;
  lodash.pick = pick;
  lodash.slice = slice;
  lodash.sortBy = sortBy;
  lodash.tap = tap;
  lodash.thru = thru;
  lodash.toArray = toArray;
  lodash.values = values;

  // Add aliases.
  lodash.extend = assignIn;

  // Add methods to `lodash.prototype`.
  mixin(lodash, lodash);

  /*------------------------------------------------------------------------*/

  // Add methods that return unwrapped values in chain sequences.
  lodash.clone = clone;
  lodash.escape = escape;
  lodash.every = every;
  lodash.find = find;
  lodash.forEach = forEach;
  lodash.has = has;
  lodash.head = head;
  lodash.identity = identity;
  lodash.indexOf = indexOf;
  lodash.isArguments = isArguments;
  lodash.isArray = isArray;
  lodash.isBoolean = isBoolean;
  lodash.isDate = isDate;
  lodash.isEmpty = isEmpty;
  lodash.isEqual = isEqual;
  lodash.isFinite = isFinite;
  lodash.isFunction = isFunction;
  lodash.isNaN = isNaN;
  lodash.isNull = isNull;
  lodash.isNumber = isNumber;
  lodash.isObject = isObject;
  lodash.isRegExp = isRegExp;
  lodash.isString = isString;
  lodash.isUndefined = isUndefined;
  lodash.last = last;
  lodash.max = max;
  lodash.min = min;
  lodash.noConflict = noConflict;
  lodash.noop = noop;
  lodash.reduce = reduce;
  lodash.result = result;
  lodash.size = size;
  lodash.some = some;
  lodash.uniqueId = uniqueId;

  // Add aliases.
  lodash.each = forEach;
  lodash.first = head;

  mixin(lodash, (function() ***REMOVED***
    var source = ***REMOVED******REMOVED***;
    baseForOwn(lodash, function(func, methodName) ***REMOVED***
      if (!hasOwnProperty.call(lodash.prototype, methodName)) ***REMOVED***
        source[methodName] = func;
      ***REMOVED***
    ***REMOVED***);
    return source;
  ***REMOVED***()), ***REMOVED*** 'chain': false ***REMOVED***);

  /*------------------------------------------------------------------------*/

  /**
   * The semantic version number.
   *
   * @static
   * @memberOf _
   * @type ***REMOVED***string***REMOVED***
   */
  lodash.VERSION = VERSION;

  // Add `Array` methods to `lodash.prototype`.
  baseEach(['pop', 'join', 'replace', 'reverse', 'split', 'push', 'shift', 'sort', 'splice', 'unshift'], function(methodName) ***REMOVED***
    var func = (/^(?:replace|split)$/.test(methodName) ? String.prototype : arrayProto)[methodName],
        chainName = /^(?:push|sort|unshift)$/.test(methodName) ? 'tap' : 'thru',
        retUnwrapped = /^(?:pop|join|replace|shift)$/.test(methodName);

    lodash.prototype[methodName] = function() ***REMOVED***
      var args = arguments;
      if (retUnwrapped && !this.__chain__) ***REMOVED***
        var value = this.value();
        return func.apply(isArray(value) ? value : [], args);
      ***REMOVED***
      return this[chainName](function(value) ***REMOVED***
        return func.apply(isArray(value) ? value : [], args);
      ***REMOVED***);
    ***REMOVED***;
  ***REMOVED***);

  // Add chain sequence methods to the `lodash` wrapper.
  lodash.prototype.toJSON = lodash.prototype.valueOf = lodash.prototype.value = wrapperValue;

  /*--------------------------------------------------------------------------*/

  // Some AMD build optimizers, like r.js, check for condition patterns like:
  if (typeof define == 'function' && typeof define.amd == 'object' && define.amd) ***REMOVED***
    // Expose Lodash on the global object to prevent errors when Lodash is
    // loaded by a script tag in the presence of an AMD loader.
    // See http://requirejs.org/docs/errors.html#mismatch for more details.
    // Use `_.noConflict` to remove Lodash from the global object.
    root._ = lodash;

    // Define as an anonymous module so, through path mapping, it can be
    // referenced as the "underscore" module.
    define(function() ***REMOVED***
      return lodash;
    ***REMOVED***);
  ***REMOVED***
  // Check for `exports` after `define` in case a build optimizer adds it.
  else if (freeModule) ***REMOVED***
    // Export for Node.js.
    (freeModule.exports = lodash)._ = lodash;
    // Export for CommonJS support.
    freeExports._ = lodash;
  ***REMOVED***
  else ***REMOVED***
    // Export to the global object.
    root._ = lodash;
  ***REMOVED***
***REMOVED***.call(this));
