// lodash/debounce.js
var isObject = require('./isObject'),
    now = require('./now'),
    toNumber = require('./toNumber');

/** Error message constants. */
var FUNC_ERROR_TEXT = 'Expected a function';

/* Built-in method references for those with the same name as other `lodash` methods. */
var nativeMax = Math.max,
    nativeMin = Math.min;

/**
 * Creates a debounced function that delays invoking `func` until after `wait`
 * milliseconds have elapsed since the last time the debounced function was
 * invoked. The debounced function comes with a `cancel` method to cancel
 * delayed `func` invocations and a `flush` method to immediately invoke them.
 * Provide `options` to indicate whether `func` should be invoked on the
 * leading and/or trailing edge of the `wait` timeout. The `func` is invoked
 * with the last arguments provided to the debounced function. Subsequent
 * calls to the debounced function return the result of the last `func`
 * invocation.
 *
 * **Note:** If `leading` and `trailing` options are `true`, `func` is
 * invoked on the trailing edge of the timeout only if the debounced function
 * is invoked more than once during the `wait` timeout.
 *
 * If `wait` is `0` and `leading` is `false`, `func` invocation is deferred
 * until to the next tick, similar to `setTimeout` with a timeout of `0`.
 *
 * See [David Corbacho's article](https://css-tricks.com/debouncing-throttling-explained-examples/)
 * for details over the differences between `_.debounce` and `_.throttle`.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Function
 * @param ***REMOVED***Function***REMOVED*** func The function to debounce.
 * @param ***REMOVED***number***REMOVED*** [wait=0] The number of milliseconds to delay.
 * @param ***REMOVED***Object***REMOVED*** [options=***REMOVED******REMOVED***] The options object.
 * @param ***REMOVED***boolean***REMOVED*** [options.leading=false]
 *  Specify invoking on the leading edge of the timeout.
 * @param ***REMOVED***number***REMOVED*** [options.maxWait]
 *  The maximum time `func` is allowed to be delayed before it's invoked.
 * @param ***REMOVED***boolean***REMOVED*** [options.trailing=true]
 *  Specify invoking on the trailing edge of the timeout.
 * @returns ***REMOVED***Function***REMOVED*** Returns the new debounced function.
 * @example
 *
 * // Avoid costly calculations while the window size is in flux.
 * jQuery(window).on('resize', _.debounce(calculateLayout, 150));
 *
 * // Invoke `sendMail` when clicked, debouncing subsequent calls.
 * jQuery(element).on('click', _.debounce(sendMail, 300, ***REMOVED***
 *   'leading': true,
 *   'trailing': false
 * ***REMOVED***));
 *
 * // Ensure `batchLog` is invoked once after 1 second of debounced calls.
 * var debounced = _.debounce(batchLog, 250, ***REMOVED*** 'maxWait': 1000 ***REMOVED***);
 * var source = new EventSource('/stream');
 * jQuery(source).on('message', debounced);
 *
 * // Cancel the trailing debounced invocation.
 * jQuery(window).on('popstate', debounced.cancel);
 */
function debounce(func, wait, options) ***REMOVED***
  var lastArgs,
      lastThis,
      maxWait,
      result,
      timerId,
      lastCallTime,
      lastInvokeTime = 0,
      leading = false,
      maxing = false,
      trailing = true;

  if (typeof func != 'function') ***REMOVED***
    throw new TypeError(FUNC_ERROR_TEXT);
  ***REMOVED***
  wait = toNumber(wait) || 0;
  if (isObject(options)) ***REMOVED***
    leading = !!options.leading;
    maxing = 'maxWait' in options;
    maxWait = maxing ? nativeMax(toNumber(options.maxWait) || 0, wait) : maxWait;
    trailing = 'trailing' in options ? !!options.trailing : trailing;
  ***REMOVED***

  function invokeFunc(time) ***REMOVED***
    var args = lastArgs,
        thisArg = lastThis;

    lastArgs = lastThis = undefined;
    lastInvokeTime = time;
    result = func.apply(thisArg, args);
    return result;
  ***REMOVED***

  function leadingEdge(time) ***REMOVED***
    // Reset any `maxWait` timer.
    lastInvokeTime = time;
    // Start the timer for the trailing edge.
    timerId = setTimeout(timerExpired, wait);
    // Invoke the leading edge.
    return leading ? invokeFunc(time) : result;
  ***REMOVED***

  function remainingWait(time) ***REMOVED***
    var timeSinceLastCall = time - lastCallTime,
        timeSinceLastInvoke = time - lastInvokeTime,
        result = wait - timeSinceLastCall;

    return maxing ? nativeMin(result, maxWait - timeSinceLastInvoke) : result;
  ***REMOVED***

  function shouldInvoke(time) ***REMOVED***
    var timeSinceLastCall = time - lastCallTime,
        timeSinceLastInvoke = time - lastInvokeTime;

    // Either this is the first call, activity has stopped and we're at the
    // trailing edge, the system time has gone backwards and we're treating
    // it as the trailing edge, or we've hit the `maxWait` limit.
    return (lastCallTime === undefined || (timeSinceLastCall >= wait) ||
      (timeSinceLastCall < 0) || (maxing && timeSinceLastInvoke >= maxWait));
  ***REMOVED***

  function timerExpired() ***REMOVED***
    var time = now();
    if (shouldInvoke(time)) ***REMOVED***
      return trailingEdge(time);
    ***REMOVED***
    // Restart the timer.
    timerId = setTimeout(timerExpired, remainingWait(time));
  ***REMOVED***

  function trailingEdge(time) ***REMOVED***
    timerId = undefined;

    // Only invoke if we have `lastArgs` which means `func` has been
    // debounced at least once.
    if (trailing && lastArgs) ***REMOVED***
      return invokeFunc(time);
    ***REMOVED***
    lastArgs = lastThis = undefined;
    return result;
  ***REMOVED***

  function cancel() ***REMOVED***
    if (timerId !== undefined) ***REMOVED***
      clearTimeout(timerId);
    ***REMOVED***
    lastInvokeTime = 0;
    lastArgs = lastCallTime = lastThis = timerId = undefined;
  ***REMOVED***

  function flush() ***REMOVED***
    return timerId === undefined ? result : trailingEdge(now());
  ***REMOVED***

  function debounced() ***REMOVED***
    var time = now(),
        isInvoking = shouldInvoke(time);

    lastArgs = arguments;
    lastThis = this;
    lastCallTime = time;

    if (isInvoking) ***REMOVED***
      if (timerId === undefined) ***REMOVED***
        return leadingEdge(lastCallTime);
      ***REMOVED***
      if (maxing) ***REMOVED***
        // Handle invocations in a tight loop.
        timerId = setTimeout(timerExpired, wait);
        return invokeFunc(lastCallTime);
      ***REMOVED***
    ***REMOVED***
    if (timerId === undefined) ***REMOVED***
      timerId = setTimeout(timerExpired, wait);
    ***REMOVED***
    return result;
  ***REMOVED***
  debounced.cancel = cancel;
  debounced.flush = flush;
  return debounced;
***REMOVED***

module.exports = debounce;
