// express/lib/router/index.js
/*!
 * express
 * Copyright(c) 2009-2013 TJ Holowaychuk
 * Copyright(c) 2013 Roman Shtylman
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 * @private
 */

var Route = require('./route');
var Layer = require('./layer');
var methods = require('methods');
var mixin = require('utils-merge');
var debug = require('debug')('express:router');
var deprecate = require('depd')('express');
var flatten = require('array-flatten');
var parseUrl = require('parseurl');

/**
 * Module variables.
 * @private
 */

var objectRegExp = /^\[object (\S+)\]$/;
var slice = Array.prototype.slice;
var toString = Object.prototype.toString;

/**
 * Initialize a new `Router` with the given `options`.
 *
 * @param ***REMOVED***Object***REMOVED*** options
 * @return ***REMOVED***Router***REMOVED*** which is an callable function
 * @public
 */

var proto = module.exports = function(options) ***REMOVED***
  var opts = options || ***REMOVED******REMOVED***;

  function router(req, res, next) ***REMOVED***
    router.handle(req, res, next);
  ***REMOVED***

  // mixin Router class functions
  router.__proto__ = proto;

  router.params = ***REMOVED******REMOVED***;
  router._params = [];
  router.caseSensitive = opts.caseSensitive;
  router.mergeParams = opts.mergeParams;
  router.strict = opts.strict;
  router.stack = [];

  return router;
***REMOVED***;

/**
 * Map the given param placeholder `name`(s) to the given callback.
 *
 * Parameter mapping is used to provide pre-conditions to routes
 * which use normalized placeholders. For example a _:user_id_ parameter
 * could automatically load a user's information from the database without
 * any additional code,
 *
 * The callback uses the same signature as middleware, the only difference
 * being that the value of the placeholder is passed, in this case the _id_
 * of the user. Once the `next()` function is invoked, just like middleware
 * it will continue on to execute the route, or subsequent parameter functions.
 *
 * Just like in middleware, you must either respond to the request or call next
 * to avoid stalling the request.
 *
 *  app.param('user_id', function(req, res, next, id)***REMOVED***
 *    User.find(id, function(err, user)***REMOVED***
 *      if (err) ***REMOVED***
 *        return next(err);
 *      ***REMOVED*** else if (!user) ***REMOVED***
 *        return next(new Error('failed to load user'));
 *      ***REMOVED***
 *      req.user = user;
 *      next();
 *    ***REMOVED***);
 *  ***REMOVED***);
 *
 * @param ***REMOVED***String***REMOVED*** name
 * @param ***REMOVED***Function***REMOVED*** fn
 * @return ***REMOVED***app***REMOVED*** for chaining
 * @public
 */

proto.param = function param(name, fn) ***REMOVED***
  // param logic
  if (typeof name === 'function') ***REMOVED***
    deprecate('router.param(fn): Refactor to use path params');
    this._params.push(name);
    return;
  ***REMOVED***

  // apply param functions
  var params = this._params;
  var len = params.length;
  var ret;

  if (name[0] === ':') ***REMOVED***
    deprecate('router.param(' + JSON.stringify(name) + ', fn): Use router.param(' + JSON.stringify(name.substr(1)) + ', fn) instead');
    name = name.substr(1);
  ***REMOVED***

  for (var i = 0; i < len; ++i) ***REMOVED***
    if (ret = params[i](name, fn)) ***REMOVED***
      fn = ret;
    ***REMOVED***
  ***REMOVED***

  // ensure we end up with a
  // middleware function
  if ('function' !== typeof fn) ***REMOVED***
    throw new Error('invalid param() call for ' + name + ', got ' + fn);
  ***REMOVED***

  (this.params[name] = this.params[name] || []).push(fn);
  return this;
***REMOVED***;

/**
 * Dispatch a req, res into the router.
 * @private
 */

proto.handle = function handle(req, res, out) ***REMOVED***
  var self = this;

  debug('dispatching %s %s', req.method, req.url);

  var search = 1 + req.url.indexOf('?');
  var pathlength = search ? search - 1 : req.url.length;
  var fqdn = req.url[0] !== '/' && 1 + req.url.substr(0, pathlength).indexOf('://');
  var protohost = fqdn ? req.url.substr(0, req.url.indexOf('/', 2 + fqdn)) : '';
  var idx = 0;
  var removed = '';
  var slashAdded = false;
  var paramcalled = ***REMOVED******REMOVED***;

  // store options for OPTIONS request
  // only used if OPTIONS request
  var options = [];

  // middleware and routes
  var stack = self.stack;

  // manage inter-router variables
  var parentParams = req.params;
  var parentUrl = req.baseUrl || '';
  var done = restore(out, req, 'baseUrl', 'next', 'params');

  // setup next layer
  req.next = next;

  // for options requests, respond with a default if nothing else responds
  if (req.method === 'OPTIONS') ***REMOVED***
    done = wrap(done, function(old, err) ***REMOVED***
      if (err || options.length === 0) return old(err);
      sendOptionsResponse(res, options, old);
    ***REMOVED***);
  ***REMOVED***

  // setup basic req values
  req.baseUrl = parentUrl;
  req.originalUrl = req.originalUrl || req.url;

  next();

  function next(err) ***REMOVED***
    var layerError = err === 'route'
      ? null
      : err;

    // remove added slash
    if (slashAdded) ***REMOVED***
      req.url = req.url.substr(1);
      slashAdded = false;
    ***REMOVED***

    // restore altered req.url
    if (removed.length !== 0) ***REMOVED***
      req.baseUrl = parentUrl;
      req.url = protohost + removed + req.url.substr(protohost.length);
      removed = '';
    ***REMOVED***

    // no more matching layers
    if (idx >= stack.length) ***REMOVED***
      setImmediate(done, layerError);
      return;
    ***REMOVED***

    // get pathname of request
    var path = getPathname(req);

    if (path == null) ***REMOVED***
      return done(layerError);
    ***REMOVED***

    // find next matching layer
    var layer;
    var match;
    var route;

    while (match !== true && idx < stack.length) ***REMOVED***
      layer = stack[idx++];
      match = matchLayer(layer, path);
      route = layer.route;

      if (typeof match !== 'boolean') ***REMOVED***
        // hold on to layerError
        layerError = layerError || match;
      ***REMOVED***

      if (match !== true) ***REMOVED***
        continue;
      ***REMOVED***

      if (!route) ***REMOVED***
        // process non-route handlers normally
        continue;
      ***REMOVED***

      if (layerError) ***REMOVED***
        // routes do not match with a pending error
        match = false;
        continue;
      ***REMOVED***

      var method = req.method;
      var has_method = route._handles_method(method);

      // build up automatic options response
      if (!has_method && method === 'OPTIONS') ***REMOVED***
        appendMethods(options, route._options());
      ***REMOVED***

      // don't even bother matching route
      if (!has_method && method !== 'HEAD') ***REMOVED***
        match = false;
        continue;
      ***REMOVED***
    ***REMOVED***

    // no match
    if (match !== true) ***REMOVED***
      return done(layerError);
    ***REMOVED***

    // store route for dispatch on change
    if (route) ***REMOVED***
      req.route = route;
    ***REMOVED***

    // Capture one-time layer values
    req.params = self.mergeParams
      ? mergeParams(layer.params, parentParams)
      : layer.params;
    var layerPath = layer.path;

    // this should be done for the layer
    self.process_params(layer, paramcalled, req, res, function (err) ***REMOVED***
      if (err) ***REMOVED***
        return next(layerError || err);
      ***REMOVED***

      if (route) ***REMOVED***
        return layer.handle_request(req, res, next);
      ***REMOVED***

      trim_prefix(layer, layerError, layerPath, path);
    ***REMOVED***);
  ***REMOVED***

  function trim_prefix(layer, layerError, layerPath, path) ***REMOVED***
    var c = path[layerPath.length];
    if (c && '/' !== c && '.' !== c) return next(layerError);

     // Trim off the part of the url that matches the route
     // middleware (.use stuff) needs to have the path stripped
    if (layerPath.length !== 0) ***REMOVED***
      debug('trim prefix (%s) from url %s', layerPath, req.url);
      removed = layerPath;
      req.url = protohost + req.url.substr(protohost.length + removed.length);

      // Ensure leading slash
      if (!fqdn && req.url[0] !== '/') ***REMOVED***
        req.url = '/' + req.url;
        slashAdded = true;
      ***REMOVED***

      // Setup base URL (no trailing slash)
      req.baseUrl = parentUrl + (removed[removed.length - 1] === '/'
        ? removed.substring(0, removed.length - 1)
        : removed);
    ***REMOVED***

    debug('%s %s : %s', layer.name, layerPath, req.originalUrl);

    if (layerError) ***REMOVED***
      layer.handle_error(layerError, req, res, next);
    ***REMOVED*** else ***REMOVED***
      layer.handle_request(req, res, next);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Process any parameters for the layer.
 * @private
 */

proto.process_params = function process_params(layer, called, req, res, done) ***REMOVED***
  var params = this.params;

  // captured parameters from the layer, keys and values
  var keys = layer.keys;

  // fast track
  if (!keys || keys.length === 0) ***REMOVED***
    return done();
  ***REMOVED***

  var i = 0;
  var name;
  var paramIndex = 0;
  var key;
  var paramVal;
  var paramCallbacks;
  var paramCalled;

  // process params in order
  // param callbacks can be async
  function param(err) ***REMOVED***
    if (err) ***REMOVED***
      return done(err);
    ***REMOVED***

    if (i >= keys.length ) ***REMOVED***
      return done();
    ***REMOVED***

    paramIndex = 0;
    key = keys[i++];

    if (!key) ***REMOVED***
      return done();
    ***REMOVED***

    name = key.name;
    paramVal = req.params[name];
    paramCallbacks = params[name];
    paramCalled = called[name];

    if (paramVal === undefined || !paramCallbacks) ***REMOVED***
      return param();
    ***REMOVED***

    // param previously called with same value or error occurred
    if (paramCalled && (paramCalled.match === paramVal
      || (paramCalled.error && paramCalled.error !== 'route'))) ***REMOVED***
      // restore value
      req.params[name] = paramCalled.value;

      // next param
      return param(paramCalled.error);
    ***REMOVED***

    called[name] = paramCalled = ***REMOVED***
      error: null,
      match: paramVal,
      value: paramVal
    ***REMOVED***;

    paramCallback();
  ***REMOVED***

  // single param callbacks
  function paramCallback(err) ***REMOVED***
    var fn = paramCallbacks[paramIndex++];

    // store updated value
    paramCalled.value = req.params[key.name];

    if (err) ***REMOVED***
      // store error
      paramCalled.error = err;
      param(err);
      return;
    ***REMOVED***

    if (!fn) return param();

    try ***REMOVED***
      fn(req, res, paramCallback, paramVal, key.name);
    ***REMOVED*** catch (e) ***REMOVED***
      paramCallback(e);
    ***REMOVED***
  ***REMOVED***

  param();
***REMOVED***;

/**
 * Use the given middleware function, with optional path, defaulting to "/".
 *
 * Use (like `.all`) will run for any http METHOD, but it will not add
 * handlers for those methods so OPTIONS requests will not consider `.use`
 * functions even if they could respond.
 *
 * The other difference is that _route_ path is stripped and not visible
 * to the handler function. The main effect of this feature is that mounted
 * handlers can operate without any code changes regardless of the "prefix"
 * pathname.
 *
 * @public
 */

proto.use = function use(fn) ***REMOVED***
  var offset = 0;
  var path = '/';

  // default path to '/'
  // disambiguate router.use([fn])
  if (typeof fn !== 'function') ***REMOVED***
    var arg = fn;

    while (Array.isArray(arg) && arg.length !== 0) ***REMOVED***
      arg = arg[0];
    ***REMOVED***

    // first arg is the path
    if (typeof arg !== 'function') ***REMOVED***
      offset = 1;
      path = fn;
    ***REMOVED***
  ***REMOVED***

  var callbacks = flatten(slice.call(arguments, offset));

  if (callbacks.length === 0) ***REMOVED***
    throw new TypeError('Router.use() requires middleware functions');
  ***REMOVED***

  for (var i = 0; i < callbacks.length; i++) ***REMOVED***
    var fn = callbacks[i];

    if (typeof fn !== 'function') ***REMOVED***
      throw new TypeError('Router.use() requires middleware function but got a ' + gettype(fn));
    ***REMOVED***

    // add the middleware
    debug('use %s %s', path, fn.name || '<anonymous>');

    var layer = new Layer(path, ***REMOVED***
      sensitive: this.caseSensitive,
      strict: false,
      end: false
    ***REMOVED***, fn);

    layer.route = undefined;

    this.stack.push(layer);
  ***REMOVED***

  return this;
***REMOVED***;

/**
 * Create a new Route for the given path.
 *
 * Each route contains a separate middleware stack and VERB handlers.
 *
 * See the Route api documentation for details on adding handlers
 * and middleware to routes.
 *
 * @param ***REMOVED***String***REMOVED*** path
 * @return ***REMOVED***Route***REMOVED***
 * @public
 */

proto.route = function route(path) ***REMOVED***
  var route = new Route(path);

  var layer = new Layer(path, ***REMOVED***
    sensitive: this.caseSensitive,
    strict: this.strict,
    end: true
  ***REMOVED***, route.dispatch.bind(route));

  layer.route = route;

  this.stack.push(layer);
  return route;
***REMOVED***;

// create Router#VERB functions
methods.concat('all').forEach(function(method)***REMOVED***
  proto[method] = function(path)***REMOVED***
    var route = this.route(path)
    route[method].apply(route, slice.call(arguments, 1));
    return this;
  ***REMOVED***;
***REMOVED***);

// append methods to a list of methods
function appendMethods(list, addition) ***REMOVED***
  for (var i = 0; i < addition.length; i++) ***REMOVED***
    var method = addition[i];
    if (list.indexOf(method) === -1) ***REMOVED***
      list.push(method);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// get pathname of request
function getPathname(req) ***REMOVED***
  try ***REMOVED***
    return parseUrl(req).pathname;
  ***REMOVED*** catch (err) ***REMOVED***
    return undefined;
  ***REMOVED***
***REMOVED***

// get type for error message
function gettype(obj) ***REMOVED***
  var type = typeof obj;

  if (type !== 'object') ***REMOVED***
    return type;
  ***REMOVED***

  // inspect [[Class]] for objects
  return toString.call(obj)
    .replace(objectRegExp, '$1');
***REMOVED***

/**
 * Match path to a layer.
 *
 * @param ***REMOVED***Layer***REMOVED*** layer
 * @param ***REMOVED***string***REMOVED*** path
 * @private
 */

function matchLayer(layer, path) ***REMOVED***
  try ***REMOVED***
    return layer.match(path);
  ***REMOVED*** catch (err) ***REMOVED***
    return err;
  ***REMOVED***
***REMOVED***

// merge params with parent params
function mergeParams(params, parent) ***REMOVED***
  if (typeof parent !== 'object' || !parent) ***REMOVED***
    return params;
  ***REMOVED***

  // make copy of parent for base
  var obj = mixin(***REMOVED******REMOVED***, parent);

  // simple non-numeric merging
  if (!(0 in params) || !(0 in parent)) ***REMOVED***
    return mixin(obj, params);
  ***REMOVED***

  var i = 0;
  var o = 0;

  // determine numeric gaps
  while (i in params) ***REMOVED***
    i++;
  ***REMOVED***

  while (o in parent) ***REMOVED***
    o++;
  ***REMOVED***

  // offset numeric indices in params before merge
  for (i--; i >= 0; i--) ***REMOVED***
    params[i + o] = params[i];

    // create holes for the merge when necessary
    if (i < o) ***REMOVED***
      delete params[i];
    ***REMOVED***
  ***REMOVED***

  return mixin(obj, params);
***REMOVED***

// restore obj props after function
function restore(fn, obj) ***REMOVED***
  var props = new Array(arguments.length - 2);
  var vals = new Array(arguments.length - 2);

  for (var i = 0; i < props.length; i++) ***REMOVED***
    props[i] = arguments[i + 2];
    vals[i] = obj[props[i]];
  ***REMOVED***

  return function(err)***REMOVED***
    // restore vals
    for (var i = 0; i < props.length; i++) ***REMOVED***
      obj[props[i]] = vals[i];
    ***REMOVED***

    return fn.apply(this, arguments);
  ***REMOVED***;
***REMOVED***

// send an OPTIONS response
function sendOptionsResponse(res, options, next) ***REMOVED***
  try ***REMOVED***
    var body = options.join(',');
    res.set('Allow', body);
    res.send(body);
  ***REMOVED*** catch (err) ***REMOVED***
    next(err);
  ***REMOVED***
***REMOVED***

// wrap a function
function wrap(old, fn) ***REMOVED***
  return function proxy() ***REMOVED***
    var args = new Array(arguments.length + 1);

    args[0] = old;
    for (var i = 0, len = arguments.length; i < len; i++) ***REMOVED***
      args[i + 1] = arguments[i];
    ***REMOVED***

    fn.apply(this, args);
  ***REMOVED***;
***REMOVED***
