// mongoose/lib/drivers/node-mongodb-native/connection.js
/*!
 * Module dependencies.
 */

var MongooseConnection = require('../../connection');
var mongo = require('mongodb');
var Db = mongo.Db;
var Server = mongo.Server;
var Mongos = mongo.Mongos;
var STATES = require('../../connectionstate');
var ReplSetServers = mongo.ReplSet;
var DisconnectedError = require('../../error/disconnected');

/**
 * A [node-mongodb-native](https://github.com/mongodb/node-mongodb-native) connection implementation.
 *
 * @inherits Connection
 * @api private
 */

function NativeConnection() ***REMOVED***
  MongooseConnection.apply(this, arguments);
  this._listening = false;
***REMOVED***

/**
 * Expose the possible connection states.
 * @api public
 */

NativeConnection.STATES = STATES;

/*!
 * Inherits from Connection.
 */

NativeConnection.prototype.__proto__ = MongooseConnection.prototype;

/**
 * Opens the connection to MongoDB.
 *
 * @param ***REMOVED***Function***REMOVED*** fn
 * @return ***REMOVED***Connection***REMOVED*** this
 * @api private
 */

NativeConnection.prototype.doOpen = function(fn) ***REMOVED***
  var _this = this;
  var server = new Server(this.host, this.port, this.options.server);

  if (this.options && this.options.mongos) ***REMOVED***
    var mongos = new Mongos([server], this.options.mongos);
    this.db = new Db(this.name, mongos, this.options.db);
  ***REMOVED*** else ***REMOVED***
    this.db = new Db(this.name, server, this.options.db);
  ***REMOVED***

  this.db.open(function(err) ***REMOVED***
    listen(_this);

    if (!mongos) ***REMOVED***
      server.s.server.on('error', function(error) ***REMOVED***
        if (/after \d+ attempts/.test(error.message)) ***REMOVED***
          _this.emit('error', new DisconnectedError(server.s.server.name));
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***

    if (err) return fn(err);

    fn();
  ***REMOVED***);

  return this;
***REMOVED***;

/**
 * Switches to a different database using the same connection pool.
 *
 * Returns a new connection object, with the new db.
 *
 * @param ***REMOVED***String***REMOVED*** name The database name
 * @return ***REMOVED***Connection***REMOVED*** New Connection Object
 * @api public
 */

NativeConnection.prototype.useDb = function(name) ***REMOVED***
  // we have to manually copy all of the attributes...
  var newConn = new this.constructor();
  newConn.name = name;
  newConn.base = this.base;
  newConn.collections = ***REMOVED******REMOVED***;
  newConn.models = ***REMOVED******REMOVED***;
  newConn.replica = this.replica;
  newConn.hosts = this.hosts;
  newConn.host = this.host;
  newConn.port = this.port;
  newConn.user = this.user;
  newConn.pass = this.pass;
  newConn.options = this.options;
  newConn._readyState = this._readyState;
  newConn._closeCalled = this._closeCalled;
  newConn._hasOpened = this._hasOpened;
  newConn._listening = false;

  // First, when we create another db object, we are not guaranteed to have a
  // db object to work with. So, in the case where we have a db object and it
  // is connected, we can just proceed with setting everything up. However, if
  // we do not have a db or the state is not connected, then we need to wait on
  // the 'open' event of the connection before doing the rest of the setup
  // the 'connected' event is the first time we'll have access to the db object

  var _this = this;

  if (this.db && this._readyState === STATES.connected) ***REMOVED***
    wireup();
  ***REMOVED*** else ***REMOVED***
    this.once('connected', wireup);
  ***REMOVED***

  function wireup() ***REMOVED***
    newConn.db = _this.db.db(name);
    newConn.onOpen();
    // setup the events appropriately
    listen(newConn);
  ***REMOVED***

  newConn.name = name;

  // push onto the otherDbs stack, this is used when state changes
  this.otherDbs.push(newConn);
  newConn.otherDbs.push(this);

  return newConn;
***REMOVED***;

/*!
 * Register listeners for important events and bubble appropriately.
 */

function listen(conn) ***REMOVED***
  if (conn.db._listening) ***REMOVED***
    return;
  ***REMOVED***
  conn.db._listening = true;

  conn.db.on('close', function() ***REMOVED***
    if (conn._closeCalled) return;

    // the driver never emits an `open` event. auto_reconnect still
    // emits a `close` event but since we never get another
    // `open` we can't emit close
    if (conn.db.serverConfig.autoReconnect) ***REMOVED***
      conn.readyState = STATES.disconnected;
      conn.emit('close');
      return;
    ***REMOVED***
    conn.onClose();
  ***REMOVED***);
  conn.db.on('error', function(err) ***REMOVED***
    conn.emit('error', err);
  ***REMOVED***);
  conn.db.on('reconnect', function() ***REMOVED***
    conn.readyState = STATES.connected;
    conn.emit('reconnected');
    conn.onOpen();
  ***REMOVED***);
  conn.db.on('timeout', function(err) ***REMOVED***
    var error = new Error(err && err.err || 'connection timeout');
    conn.emit('error', error);
  ***REMOVED***);
  conn.db.on('open', function(err, db) ***REMOVED***
    if (STATES.disconnected === conn.readyState && db && db.databaseName) ***REMOVED***
      conn.readyState = STATES.connected;
      conn.emit('reconnected');
    ***REMOVED***
  ***REMOVED***);
  conn.db.on('parseError', function(err) ***REMOVED***
    conn.emit('parseError', err);
  ***REMOVED***);
***REMOVED***

/**
 * Opens a connection to a MongoDB ReplicaSet.
 *
 * See description of [doOpen](#NativeConnection-doOpen) for server options. In this case `options.replset` is also passed to ReplSetServers.
 *
 * @param ***REMOVED***Function***REMOVED*** fn
 * @api private
 * @return ***REMOVED***Connection***REMOVED*** this
 */

NativeConnection.prototype.doOpenSet = function(fn) ***REMOVED***
  var servers = [],
      _this = this;

  this.hosts.forEach(function(server) ***REMOVED***
    var host = server.host || server.ipc;
    var port = server.port || 27017;
    servers.push(new Server(host, port, _this.options.server));
  ***REMOVED***);

  var server = this.options.mongos
    ? new Mongos(servers, this.options.mongos)
    : new ReplSetServers(servers, this.options.replset || this.options.replSet);
  this.db = new Db(this.name, server, this.options.db);

  this.db.on('fullsetup', function() ***REMOVED***
    _this.emit('fullsetup');
  ***REMOVED***);

  this.db.on('all', function() ***REMOVED***
    _this.emit('all');
  ***REMOVED***);

  this.db.open(function(err) ***REMOVED***
    if (err) return fn(err);
    fn();
    listen(_this);
  ***REMOVED***);

  return this;
***REMOVED***;

/**
 * Closes the connection
 *
 * @param ***REMOVED***Function***REMOVED*** fn
 * @return ***REMOVED***Connection***REMOVED*** this
 * @api private
 */

NativeConnection.prototype.doClose = function(fn) ***REMOVED***
  this.db.close(fn);
  return this;
***REMOVED***;

/**
 * Prepares default connection options for the node-mongodb-native driver.
 *
 * _NOTE: `passed` options take precedence over connection string options._
 *
 * @param ***REMOVED***Object***REMOVED*** passed options that were passed directly during connection
 * @param ***REMOVED***Object***REMOVED*** [connStrOptions] options that were passed in the connection string
 * @api private
 */

NativeConnection.prototype.parseOptions = function(passed, connStrOpts) ***REMOVED***
  var o = passed || ***REMOVED******REMOVED***;
  o.db || (o.db = ***REMOVED******REMOVED***);
  o.auth || (o.auth = ***REMOVED******REMOVED***);
  o.server || (o.server = ***REMOVED******REMOVED***);
  o.replset || (o.replset = o.replSet) || (o.replset = ***REMOVED******REMOVED***);
  o.server.socketOptions || (o.server.socketOptions = ***REMOVED******REMOVED***);
  o.replset.socketOptions || (o.replset.socketOptions = ***REMOVED******REMOVED***);
  o.mongos || (o.mongos = (connStrOpts && connStrOpts.mongos));
  (o.mongos === true) && (o.mongos = ***REMOVED******REMOVED***);

  var opts = connStrOpts || ***REMOVED******REMOVED***;
  Object.keys(opts).forEach(function(name) ***REMOVED***
    switch (name) ***REMOVED***
      case 'ssl':
        o.server.ssl = opts.ssl;
        o.replset.ssl = opts.ssl;
        o.mongos && (o.mongos.ssl = opts.ssl);
        break;
      case 'poolSize':
        if (typeof o.server[name] === 'undefined') ***REMOVED***
          o.server[name] = o.replset[name] = opts[name];
        ***REMOVED***
        break;
      case 'slaveOk':
        if (typeof o.server.slave_ok === 'undefined') ***REMOVED***
          o.server.slave_ok = opts[name];
        ***REMOVED***
        break;
      case 'autoReconnect':
        if (typeof o.server.auto_reconnect === 'undefined') ***REMOVED***
          o.server.auto_reconnect = opts[name];
        ***REMOVED***
        break;
      case 'socketTimeoutMS':
      case 'connectTimeoutMS':
        if (typeof o.server.socketOptions[name] === 'undefined') ***REMOVED***
          o.server.socketOptions[name] = o.replset.socketOptions[name] = opts[name];
        ***REMOVED***
        break;
      case 'authdb':
        if (typeof o.auth.authdb === 'undefined') ***REMOVED***
          o.auth.authdb = opts[name];
        ***REMOVED***
        break;
      case 'authSource':
        if (typeof o.auth.authSource === 'undefined') ***REMOVED***
          o.auth.authSource = opts[name];
        ***REMOVED***
        break;
      case 'retries':
      case 'reconnectWait':
      case 'rs_name':
        if (typeof o.replset[name] === 'undefined') ***REMOVED***
          o.replset[name] = opts[name];
        ***REMOVED***
        break;
      case 'replicaSet':
        if (typeof o.replset.rs_name === 'undefined') ***REMOVED***
          o.replset.rs_name = opts[name];
        ***REMOVED***
        break;
      case 'readSecondary':
        if (typeof o.replset.read_secondary === 'undefined') ***REMOVED***
          o.replset.read_secondary = opts[name];
        ***REMOVED***
        break;
      case 'nativeParser':
        if (typeof o.db.native_parser === 'undefined') ***REMOVED***
          o.db.native_parser = opts[name];
        ***REMOVED***
        break;
      case 'w':
      case 'safe':
      case 'fsync':
      case 'journal':
      case 'wtimeoutMS':
        if (typeof o.db[name] === 'undefined') ***REMOVED***
          o.db[name] = opts[name];
        ***REMOVED***
        break;
      case 'readPreference':
        if (typeof o.db.readPreference === 'undefined') ***REMOVED***
          o.db.readPreference = opts[name];
        ***REMOVED***
        break;
      case 'readPreferenceTags':
        if (typeof o.db.read_preference_tags === 'undefined') ***REMOVED***
          o.db.read_preference_tags = opts[name];
        ***REMOVED***
        break;
      case 'sslValidate':
        o.server.sslValidate = opts.sslValidate;
        o.replset.sslValidate = opts.sslValidate;
        o.mongos && (o.mongos.sslValidate = opts.sslValidate);
    ***REMOVED***
  ***REMOVED***);

  if (!('auto_reconnect' in o.server)) ***REMOVED***
    o.server.auto_reconnect = true;
  ***REMOVED***

  // mongoose creates its own ObjectIds
  o.db.forceServerObjectId = false;

  // default safe using new nomenclature
  if (!('journal' in o.db || 'j' in o.db ||
        'fsync' in o.db || 'safe' in o.db || 'w' in o.db)) ***REMOVED***
    o.db.w = 1;
  ***REMOVED***

  if (o.promiseLibrary) ***REMOVED***
    o.db.promiseLibrary = o.promiseLibrary;
  ***REMOVED***

  validate(o);
  return o;
***REMOVED***;

/*!
 * Validates the driver db options.
 *
 * @param ***REMOVED***Object***REMOVED*** o
 */

function validate(o) ***REMOVED***
  if (o.db.w === -1 || o.db.w === 0) ***REMOVED***
    if (o.db.journal || o.db.fsync || o.db.safe) ***REMOVED***
      throw new Error(
          'Invalid writeConcern: '
        + 'w set to -1 or 0 cannot be combined with safe|fsync|journal');
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/*!
 * Module exports.
 */

module.exports = NativeConnection;
