// media-typer/index.js
/*!
 * media-typer
 * Copyright(c) 2014 Douglas Christopher Wilson
 * MIT Licensed
 */

/**
 * RegExp to match *( ";" parameter ) in RFC 2616 sec 3.7
 *
 * parameter     = token "=" ( token | quoted-string )
 * token         = 1*<any CHAR except CTLs or separators>
 * separators    = "(" | ")" | "<" | ">" | "@"
 *               | "," | ";" | ":" | "\" | <">
 *               | "/" | "[" | "]" | "?" | "="
 *               | "***REMOVED***" | "***REMOVED***" | SP | HT
 * quoted-string = ( <"> *(qdtext | quoted-pair ) <"> )
 * qdtext        = <any TEXT except <">>
 * quoted-pair   = "\" CHAR
 * CHAR          = <any US-ASCII character (octets 0 - 127)>
 * TEXT          = <any OCTET except CTLs, but including LWS>
 * LWS           = [CRLF] 1*( SP | HT )
 * CRLF          = CR LF
 * CR            = <US-ASCII CR, carriage return (13)>
 * LF            = <US-ASCII LF, linefeed (10)>
 * SP            = <US-ASCII SP, space (32)>
 * SHT           = <US-ASCII HT, horizontal-tab (9)>
 * CTL           = <any US-ASCII control character (octets 0 - 31) and DEL (127)>
 * OCTET         = <any 8-bit sequence of data>
 */
var paramRegExp = /; *([!#$%&'\*\+\-\.0-9A-Z\^_`a-z\|~]+) *= *("(?:[ !\u0023-\u005b\u005d-\u007e\u0080-\u00ff]|\\[\u0020-\u007e])*"|[!#$%&'\*\+\-\.0-9A-Z\^_`a-z\|~]+) */g;
var textRegExp = /^[\u0020-\u007e\u0080-\u00ff]+$/
var tokenRegExp = /^[!#$%&'\*\+\-\.0-9A-Z\^_`a-z\|~]+$/

/**
 * RegExp to match quoted-pair in RFC 2616
 *
 * quoted-pair = "\" CHAR
 * CHAR        = <any US-ASCII character (octets 0 - 127)>
 */
var qescRegExp = /\\([\u0000-\u007f])/g;

/**
 * RegExp to match chars that must be quoted-pair in RFC 2616
 */
var quoteRegExp = /([\\"])/g;

/**
 * RegExp to match type in RFC 6838
 *
 * type-name = restricted-name
 * subtype-name = restricted-name
 * restricted-name = restricted-name-first *126restricted-name-chars
 * restricted-name-first  = ALPHA / DIGIT
 * restricted-name-chars  = ALPHA / DIGIT / "!" / "#" /
 *                          "$" / "&" / "-" / "^" / "_"
 * restricted-name-chars =/ "." ; Characters before first dot always
 *                              ; specify a facet name
 * restricted-name-chars =/ "+" ; Characters after last plus always
 *                              ; specify a structured syntax suffix
 * ALPHA =  %x41-5A / %x61-7A   ; A-Z / a-z
 * DIGIT =  %x30-39             ; 0-9
 */
var subtypeNameRegExp = /^[A-Za-z0-9][A-Za-z0-9!#$&^_.-]***REMOVED***0,126***REMOVED***$/
var typeNameRegExp = /^[A-Za-z0-9][A-Za-z0-9!#$&^_-]***REMOVED***0,126***REMOVED***$/
var typeRegExp = /^ *([A-Za-z0-9][A-Za-z0-9!#$&^_-]***REMOVED***0,126***REMOVED***)\/([A-Za-z0-9][A-Za-z0-9!#$&^_.+-]***REMOVED***0,126***REMOVED***) *$/;

/**
 * Module exports.
 */

exports.format = format
exports.parse = parse

/**
 * Format object to media type.
 *
 * @param ***REMOVED***object***REMOVED*** obj
 * @return ***REMOVED***string***REMOVED***
 * @api public
 */

function format(obj) ***REMOVED***
  if (!obj || typeof obj !== 'object') ***REMOVED***
    throw new TypeError('argument obj is required')
  ***REMOVED***

  var parameters = obj.parameters
  var subtype = obj.subtype
  var suffix = obj.suffix
  var type = obj.type

  if (!type || !typeNameRegExp.test(type)) ***REMOVED***
    throw new TypeError('invalid type')
  ***REMOVED***

  if (!subtype || !subtypeNameRegExp.test(subtype)) ***REMOVED***
    throw new TypeError('invalid subtype')
  ***REMOVED***

  // format as type/subtype
  var string = type + '/' + subtype

  // append +suffix
  if (suffix) ***REMOVED***
    if (!typeNameRegExp.test(suffix)) ***REMOVED***
      throw new TypeError('invalid suffix')
    ***REMOVED***

    string += '+' + suffix
  ***REMOVED***

  // append parameters
  if (parameters && typeof parameters === 'object') ***REMOVED***
    var param
    var params = Object.keys(parameters).sort()

    for (var i = 0; i < params.length; i++) ***REMOVED***
      param = params[i]

      if (!tokenRegExp.test(param)) ***REMOVED***
        throw new TypeError('invalid parameter name')
      ***REMOVED***

      string += '; ' + param + '=' + qstring(parameters[param])
    ***REMOVED***
  ***REMOVED***

  return string
***REMOVED***

/**
 * Parse media type to object.
 *
 * @param ***REMOVED***string|object***REMOVED*** string
 * @return ***REMOVED***Object***REMOVED***
 * @api public
 */

function parse(string) ***REMOVED***
  if (!string) ***REMOVED***
    throw new TypeError('argument string is required')
  ***REMOVED***

  // support req/res-like objects as argument
  if (typeof string === 'object') ***REMOVED***
    string = getcontenttype(string)
  ***REMOVED***

  if (typeof string !== 'string') ***REMOVED***
    throw new TypeError('argument string is required to be a string')
  ***REMOVED***

  var index = string.indexOf(';')
  var type = index !== -1
    ? string.substr(0, index)
    : string

  var key
  var match
  var obj = splitType(type)
  var params = ***REMOVED******REMOVED***
  var value

  paramRegExp.lastIndex = index

  while (match = paramRegExp.exec(string)) ***REMOVED***
    if (match.index !== index) ***REMOVED***
      throw new TypeError('invalid parameter format')
    ***REMOVED***

    index += match[0].length
    key = match[1].toLowerCase()
    value = match[2]

    if (value[0] === '"') ***REMOVED***
      // remove quotes and escapes
      value = value
        .substr(1, value.length - 2)
        .replace(qescRegExp, '$1')
    ***REMOVED***

    params[key] = value
  ***REMOVED***

  if (index !== -1 && index !== string.length) ***REMOVED***
    throw new TypeError('invalid parameter format')
  ***REMOVED***

  obj.parameters = params

  return obj
***REMOVED***

/**
 * Get content-type from req/res objects.
 *
 * @param ***REMOVED***object***REMOVED***
 * @return ***REMOVED***Object***REMOVED***
 * @api private
 */

function getcontenttype(obj) ***REMOVED***
  if (typeof obj.getHeader === 'function') ***REMOVED***
    // res-like
    return obj.getHeader('content-type')
  ***REMOVED***

  if (typeof obj.headers === 'object') ***REMOVED***
    // req-like
    return obj.headers && obj.headers['content-type']
  ***REMOVED***
***REMOVED***

/**
 * Quote a string if necessary.
 *
 * @param ***REMOVED***string***REMOVED*** val
 * @return ***REMOVED***string***REMOVED***
 * @api private
 */

function qstring(val) ***REMOVED***
  var str = String(val)

  // no need to quote tokens
  if (tokenRegExp.test(str)) ***REMOVED***
    return str
  ***REMOVED***

  if (str.length > 0 && !textRegExp.test(str)) ***REMOVED***
    throw new TypeError('invalid parameter value')
  ***REMOVED***

  return '"' + str.replace(quoteRegExp, '\\$1') + '"'
***REMOVED***

/**
 * Simply "type/subtype+siffx" into parts.
 *
 * @param ***REMOVED***string***REMOVED*** string
 * @return ***REMOVED***Object***REMOVED***
 * @api private
 */

function splitType(string) ***REMOVED***
  var match = typeRegExp.exec(string.toLowerCase())

  if (!match) ***REMOVED***
    throw new TypeError('invalid media type')
  ***REMOVED***

  var type = match[1]
  var subtype = match[2]
  var suffix

  // suffix after last +
  var index = subtype.lastIndexOf('+')
  if (index !== -1) ***REMOVED***
    suffix = subtype.substr(index + 1)
    subtype = subtype.substr(0, index)
  ***REMOVED***

  var obj = ***REMOVED***
    type: type,
    subtype: subtype,
    suffix: suffix
  ***REMOVED***

  return obj
***REMOVED***
