// async/autoInject.js
'use strict';

Object.defineProperty(exports, "__esModule", ***REMOVED***
    value: true
***REMOVED***);
exports.default = autoInject;

var _auto = require('./auto');

var _auto2 = _interopRequireDefault(_auto);

var _baseForOwn = require('lodash/_baseForOwn');

var _baseForOwn2 = _interopRequireDefault(_baseForOwn);

var _arrayMap = require('lodash/_arrayMap');

var _arrayMap2 = _interopRequireDefault(_arrayMap);

var _isArray = require('lodash/isArray');

var _isArray2 = _interopRequireDefault(_isArray);

var _trim = require('lodash/trim');

var _trim2 = _interopRequireDefault(_trim);

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj ***REMOVED***; ***REMOVED***

var FN_ARGS = /^(function)?\s*[^\(]*\(\s*([^\)]*)\)/m;
var FN_ARG_SPLIT = /,/;
var FN_ARG = /(=.+)?(\s*)$/;
var STRIP_COMMENTS = /((\/\/.*$)|(\/\*[\s\S]*?\*\/))/mg;

function parseParams(func) ***REMOVED***
    func = func.toString().replace(STRIP_COMMENTS, '');
    func = func.match(FN_ARGS)[2].replace(' ', '');
    func = func ? func.split(FN_ARG_SPLIT) : [];
    func = func.map(function (arg) ***REMOVED***
        return (0, _trim2.default)(arg.replace(FN_ARG, ''));
    ***REMOVED***);
    return func;
***REMOVED***

/**
 * A dependency-injected version of the [async.auto]***REMOVED***@link module:ControlFlow.auto***REMOVED*** function. Dependent
 * tasks are specified as parameters to the function, after the usual callback
 * parameter, with the parameter names matching the names of the tasks it
 * depends on. This can provide even more readable task graphs which can be
 * easier to maintain.
 *
 * If a final callback is specified, the task results are similarly injected,
 * specified as named parameters after the initial error parameter.
 *
 * The autoInject function is purely syntactic sugar and its semantics are
 * otherwise equivalent to [async.auto]***REMOVED***@link module:ControlFlow.auto***REMOVED***.
 *
 * @name autoInject
 * @static
 * @memberOf module:ControlFlow
 * @method
 * @see [async.auto]***REMOVED***@link module:ControlFlow.auto***REMOVED***
 * @category Control Flow
 * @param ***REMOVED***Object***REMOVED*** tasks - An object, each of whose properties is a function of
 * the form 'func([dependencies...], callback). The object's key of a property
 * serves as the name of the task defined by that property, i.e. can be used
 * when specifying requirements for other tasks.
 * * The `callback` parameter is a `callback(err, result)` which must be called
 *   when finished, passing an `error` (which can be `null`) and the result of
 *   the function's execution. The remaining parameters name other tasks on
 *   which the task is dependent, and the results from those tasks are the
 *   arguments of those parameters.
 * @param ***REMOVED***Function***REMOVED*** [callback] - An optional callback which is called when all
 * the tasks have been completed. It receives the `err` argument if any `tasks`
 * pass an error to their callback, and a `results` object with any completed
 * task results, similar to `auto`.
 * @example
 *
 * //  The example from `auto` can be rewritten as follows:
 * async.autoInject(***REMOVED***
 *     get_data: function(callback) ***REMOVED***
 *         // async code to get some data
 *         callback(null, 'data', 'converted to array');
 *     ***REMOVED***,
 *     make_folder: function(callback) ***REMOVED***
 *         // async code to create a directory to store a file in
 *         // this is run at the same time as getting the data
 *         callback(null, 'folder');
 *     ***REMOVED***,
 *     write_file: function(get_data, make_folder, callback) ***REMOVED***
 *         // once there is some data and the directory exists,
 *         // write the data to a file in the directory
 *         callback(null, 'filename');
 *     ***REMOVED***,
 *     email_link: function(write_file, callback) ***REMOVED***
 *         // once the file is written let's email a link to it...
 *         // write_file contains the filename returned by write_file.
 *         callback(null, ***REMOVED***'file':write_file, 'email':'user@example.com'***REMOVED***);
 *     ***REMOVED***
 * ***REMOVED***, function(err, results) ***REMOVED***
 *     console.log('err = ', err);
 *     console.log('email_link = ', results.email_link);
 * ***REMOVED***);
 *
 * // If you are using a JS minifier that mangles parameter names, `autoInject`
 * // will not work with plain functions, since the parameter names will be
 * // collapsed to a single letter identifier.  To work around this, you can
 * // explicitly specify the names of the parameters your task function needs
 * // in an array, similar to Angular.js dependency injection.
 *
 * // This still has an advantage over plain `auto`, since the results a task
 * // depends on are still spread into arguments.
 * async.autoInject(***REMOVED***
 *     //...
 *     write_file: ['get_data', 'make_folder', function(get_data, make_folder, callback) ***REMOVED***
 *         callback(null, 'filename');
 *     ***REMOVED***],
 *     email_link: ['write_file', function(write_file, callback) ***REMOVED***
 *         callback(null, ***REMOVED***'file':write_file, 'email':'user@example.com'***REMOVED***);
 *     ***REMOVED***]
 *     //...
 * ***REMOVED***, function(err, results) ***REMOVED***
 *     console.log('err = ', err);
 *     console.log('email_link = ', results.email_link);
 * ***REMOVED***);
 */
function autoInject(tasks, callback) ***REMOVED***
    var newTasks = ***REMOVED******REMOVED***;

    (0, _baseForOwn2.default)(tasks, function (taskFn, key) ***REMOVED***
        var params;

        if ((0, _isArray2.default)(taskFn)) ***REMOVED***
            params = taskFn.slice(0, -1);
            taskFn = taskFn[taskFn.length - 1];

            newTasks[key] = params.concat(params.length > 0 ? newTask : taskFn);
        ***REMOVED*** else if (taskFn.length === 1) ***REMOVED***
            // no dependencies, use the function as-is
            newTasks[key] = taskFn;
        ***REMOVED*** else ***REMOVED***
            params = parseParams(taskFn);
            if (taskFn.length === 0 && params.length === 0) ***REMOVED***
                throw new Error("autoInject task functions require explicit parameters.");
            ***REMOVED***

            params.pop();

            newTasks[key] = params.concat(newTask);
        ***REMOVED***

        function newTask(results, taskCb) ***REMOVED***
            var newArgs = (0, _arrayMap2.default)(params, function (name) ***REMOVED***
                return results[name];
            ***REMOVED***);
            newArgs.push(taskCb);
            taskFn.apply(null, newArgs);
        ***REMOVED***
    ***REMOVED***);

    (0, _auto2.default)(newTasks, callback);
***REMOVED***
module.exports = exports['default'];
