// async/during.js
'use strict';

Object.defineProperty(exports, "__esModule", ***REMOVED***
    value: true
***REMOVED***);
exports.default = during;

var _noop = require('lodash/noop');

var _noop2 = _interopRequireDefault(_noop);

var _onlyOnce = require('./internal/onlyOnce');

var _onlyOnce2 = _interopRequireDefault(_onlyOnce);

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj ***REMOVED***; ***REMOVED***

/**
 * Like [`whilst`]***REMOVED***@link module:ControlFlow.whilst***REMOVED***, except the `test` is an asynchronous function that
 * is passed a callback in the form of `function (err, truth)`. If error is
 * passed to `test` or `fn`, the main callback is immediately called with the
 * value of the error.
 *
 * @name during
 * @static
 * @memberOf module:ControlFlow
 * @method
 * @see [async.whilst]***REMOVED***@link module:ControlFlow.whilst***REMOVED***
 * @category Control Flow
 * @param ***REMOVED***Function***REMOVED*** test - asynchronous truth test to perform before each
 * execution of `fn`. Invoked with (callback).
 * @param ***REMOVED***Function***REMOVED*** fn - A function which is called each time `test` passes.
 * The function is passed a `callback(err)`, which must be called once it has
 * completed with an optional `err` argument. Invoked with (callback).
 * @param ***REMOVED***Function***REMOVED*** [callback] - A callback which is called after the test
 * function has failed and repeated execution of `fn` has stopped. `callback`
 * will be passed an error, if one occured, otherwise `null`.
 * @example
 *
 * var count = 0;
 *
 * async.during(
 *     function (callback) ***REMOVED***
 *         return callback(null, count < 5);
 *     ***REMOVED***,
 *     function (callback) ***REMOVED***
 *         count++;
 *         setTimeout(callback, 1000);
 *     ***REMOVED***,
 *     function (err) ***REMOVED***
 *         // 5 seconds have passed
 *     ***REMOVED***
 * );
 */
function during(test, fn, callback) ***REMOVED***
    callback = (0, _onlyOnce2.default)(callback || _noop2.default);

    function next(err) ***REMOVED***
        if (err) return callback(err);
        test(check);
    ***REMOVED***

    function check(err, truth) ***REMOVED***
        if (err) return callback(err);
        if (!truth) return callback(null);
        fn(next);
    ***REMOVED***

    test(check);
***REMOVED***
module.exports = exports['default'];
