// async/ensureAsync.js
'use strict';

Object.defineProperty(exports, "__esModule", ***REMOVED***
    value: true
***REMOVED***);
exports.default = ensureAsync;

var _setImmediate = require('./internal/setImmediate');

var _setImmediate2 = _interopRequireDefault(_setImmediate);

var _initialParams = require('./internal/initialParams');

var _initialParams2 = _interopRequireDefault(_initialParams);

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj ***REMOVED***; ***REMOVED***

/**
 * Wrap an async function and ensure it calls its callback on a later tick of
 * the event loop.  If the function already calls its callback on a next tick,
 * no extra deferral is added. This is useful for preventing stack overflows
 * (`RangeError: Maximum call stack size exceeded`) and generally keeping
 * [Zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony)
 * contained.
 *
 * @name ensureAsync
 * @static
 * @memberOf module:Utils
 * @method
 * @category Util
 * @param ***REMOVED***Function***REMOVED*** fn - an async function, one that expects a node-style
 * callback as its last argument.
 * @returns ***REMOVED***Function***REMOVED*** Returns a wrapped function with the exact same call
 * signature as the function passed in.
 * @example
 *
 * function sometimesAsync(arg, callback) ***REMOVED***
 *     if (cache[arg]) ***REMOVED***
 *         return callback(null, cache[arg]); // this would be synchronous!!
 *     ***REMOVED*** else ***REMOVED***
 *         doSomeIO(arg, callback); // this IO would be asynchronous
 *     ***REMOVED***
 * ***REMOVED***
 *
 * // this has a risk of stack overflows if many results are cached in a row
 * async.mapSeries(args, sometimesAsync, done);
 *
 * // this will defer sometimesAsync's callback if necessary,
 * // preventing stack overflows
 * async.mapSeries(args, async.ensureAsync(sometimesAsync), done);
 */
function ensureAsync(fn) ***REMOVED***
    return (0, _initialParams2.default)(function (args, callback) ***REMOVED***
        var sync = true;
        args.push(function () ***REMOVED***
            var innerArgs = arguments;
            if (sync) ***REMOVED***
                (0, _setImmediate2.default)(function () ***REMOVED***
                    callback.apply(null, innerArgs);
                ***REMOVED***);
            ***REMOVED*** else ***REMOVED***
                callback.apply(null, innerArgs);
            ***REMOVED***
        ***REMOVED***);
        fn.apply(this, args);
        sync = false;
    ***REMOVED***);
***REMOVED***
module.exports = exports['default'];
