// async/retry.js
'use strict';

Object.defineProperty(exports, "__esModule", ***REMOVED***
    value: true
***REMOVED***);
exports.default = retry;

var _noop = require('lodash/noop');

var _noop2 = _interopRequireDefault(_noop);

var _constant = require('lodash/constant');

var _constant2 = _interopRequireDefault(_constant);

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj ***REMOVED***; ***REMOVED***

/**
 * Attempts to get a successful response from `task` no more than `times` times
 * before returning an error. If the task is successful, the `callback` will be
 * passed the result of the successful task. If all attempts fail, the callback
 * will be passed the error and result (if any) of the final attempt.
 *
 * @name retry
 * @static
 * @memberOf module:ControlFlow
 * @method
 * @category Control Flow
 * @param ***REMOVED***Object|number***REMOVED*** [opts = ***REMOVED***times: 5, interval: 0***REMOVED***| 5] - Can be either an
 * object with `times` and `interval` or a number.
 * * `times` - The number of attempts to make before giving up.  The default
 *   is `5`.
 * * `interval` - The time to wait between retries, in milliseconds.  The
 *   default is `0`. The interval may also be specified as a function of the
 *   retry count (see example).
 * * `errorFilter` - An optional synchronous function that is invoked on
 *   erroneous result. If it returns `true` the retry attempts will continue;
 *   if the function returns `false` the retry flow is aborted with the current
 *   attempt's error and result being returned to the final callback.
 *   Invoked with (err).
 * * If `opts` is a number, the number specifies the number of times to retry,
 *   with the default interval of `0`.
 * @param ***REMOVED***Function***REMOVED*** task - A function which receives two arguments: (1) a
 * `callback(err, result)` which must be called when finished, passing `err`
 * (which can be `null`) and the `result` of the function's execution, and (2)
 * a `results` object, containing the results of the previously executed
 * functions (if nested inside another control flow). Invoked with
 * (callback, results).
 * @param ***REMOVED***Function***REMOVED*** [callback] - An optional callback which is called when the
 * task has succeeded, or after the final failed attempt. It receives the `err`
 * and `result` arguments of the last attempt at completing the `task`. Invoked
 * with (err, results).
 * @example
 *
 * // The `retry` function can be used as a stand-alone control flow by passing
 * // a callback, as shown below:
 *
 * // try calling apiMethod 3 times
 * async.retry(3, apiMethod, function(err, result) ***REMOVED***
 *     // do something with the result
 * ***REMOVED***);
 *
 * // try calling apiMethod 3 times, waiting 200 ms between each retry
 * async.retry(***REMOVED***times: 3, interval: 200***REMOVED***, apiMethod, function(err, result) ***REMOVED***
 *     // do something with the result
 * ***REMOVED***);
 *
 * // try calling apiMethod 10 times with exponential backoff
 * // (i.e. intervals of 100, 200, 400, 800, 1600, ... milliseconds)
 * async.retry(***REMOVED***
 *   times: 10,
 *   interval: function(retryCount) ***REMOVED***
 *     return 50 * Math.pow(2, retryCount);
 *   ***REMOVED***
 * ***REMOVED***, apiMethod, function(err, result) ***REMOVED***
 *     // do something with the result
 * ***REMOVED***);
 *
 * // try calling apiMethod the default 5 times no delay between each retry
 * async.retry(apiMethod, function(err, result) ***REMOVED***
 *     // do something with the result
 * ***REMOVED***);
 *
 * // try calling apiMethod only when error condition satisfies, all other
 * // errors will abort the retry control flow and return to final callback
 * async.retry(***REMOVED***
 *   errorFilter: function(err) ***REMOVED***
 *     return err.message === 'Temporary error'; // only retry on a specific error
 *   ***REMOVED***
 * ***REMOVED***, apiMethod, function(err, result) ***REMOVED***
 *     // do something with the result
 * ***REMOVED***);
 *
 * // It can also be embedded within other control flow functions to retry
 * // individual methods that are not as reliable, like this:
 * async.auto(***REMOVED***
 *     users: api.getUsers.bind(api),
 *     payments: async.retry(3, api.getPayments.bind(api))
 * ***REMOVED***, function(err, results) ***REMOVED***
 *     // do something with the results
 * ***REMOVED***);
 *
 */
function retry(opts, task, callback) ***REMOVED***
    var DEFAULT_TIMES = 5;
    var DEFAULT_INTERVAL = 0;

    var options = ***REMOVED***
        times: DEFAULT_TIMES,
        intervalFunc: (0, _constant2.default)(DEFAULT_INTERVAL)
    ***REMOVED***;

    function parseTimes(acc, t) ***REMOVED***
        if (typeof t === 'object') ***REMOVED***
            acc.times = +t.times || DEFAULT_TIMES;

            acc.intervalFunc = typeof t.interval === 'function' ? t.interval : (0, _constant2.default)(+t.interval || DEFAULT_INTERVAL);

            acc.errorFilter = t.errorFilter;
        ***REMOVED*** else if (typeof t === 'number' || typeof t === 'string') ***REMOVED***
            acc.times = +t || DEFAULT_TIMES;
        ***REMOVED*** else ***REMOVED***
            throw new Error("Invalid arguments for async.retry");
        ***REMOVED***
    ***REMOVED***

    if (arguments.length < 3 && typeof opts === 'function') ***REMOVED***
        callback = task || _noop2.default;
        task = opts;
    ***REMOVED*** else ***REMOVED***
        parseTimes(options, opts);
        callback = callback || _noop2.default;
    ***REMOVED***

    if (typeof task !== 'function') ***REMOVED***
        throw new Error("Invalid arguments for async.retry");
    ***REMOVED***

    var attempt = 1;
    function retryAttempt() ***REMOVED***
        task(function (err) ***REMOVED***
            if (err && attempt++ < options.times && (typeof options.errorFilter != 'function' || options.errorFilter(err))) ***REMOVED***
                setTimeout(retryAttempt, options.intervalFunc(attempt));
            ***REMOVED*** else ***REMOVED***
                callback.apply(null, arguments);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***

    retryAttempt();
***REMOVED***
module.exports = exports['default'];
