// body-parser/lib/types/text.js
/*!
 * body-parser
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 */

var bytes = require('bytes')
var contentType = require('content-type')
var debug = require('debug')('body-parser:text')
var read = require('../read')
var typeis = require('type-is')

/**
 * Module exports.
 */

module.exports = text

/**
 * Create a middleware to parse text bodies.
 *
 * @param ***REMOVED***object***REMOVED*** [options]
 * @return ***REMOVED***function***REMOVED***
 * @api public
 */

function text (options) ***REMOVED***
  var opts = options || ***REMOVED******REMOVED***

  var defaultCharset = opts.defaultCharset || 'utf-8'
  var inflate = opts.inflate !== false
  var limit = typeof opts.limit !== 'number'
    ? bytes.parse(opts.limit || '100kb')
    : opts.limit
  var type = opts.type || 'text/plain'
  var verify = opts.verify || false

  if (verify !== false && typeof verify !== 'function') ***REMOVED***
    throw new TypeError('option verify must be function')
  ***REMOVED***

  // create the appropriate type checking function
  var shouldParse = typeof type !== 'function'
    ? typeChecker(type)
    : type

  function parse (buf) ***REMOVED***
    return buf
  ***REMOVED***

  return function textParser (req, res, next) ***REMOVED***
    if (req._body) ***REMOVED***
      debug('body already parsed')
      next()
      return
    ***REMOVED***

    req.body = req.body || ***REMOVED******REMOVED***

    // skip requests without bodies
    if (!typeis.hasBody(req)) ***REMOVED***
      debug('skip empty body')
      next()
      return
    ***REMOVED***

    debug('content-type %j', req.headers['content-type'])

    // determine if request should be parsed
    if (!shouldParse(req)) ***REMOVED***
      debug('skip parsing')
      next()
      return
    ***REMOVED***

    // get charset
    var charset = getCharset(req) || defaultCharset

    // read
    read(req, res, next, parse, debug, ***REMOVED***
      encoding: charset,
      inflate: inflate,
      limit: limit,
      verify: verify
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

/**
 * Get the charset of a request.
 *
 * @param ***REMOVED***object***REMOVED*** req
 * @api private
 */

function getCharset (req) ***REMOVED***
  try ***REMOVED***
    return contentType.parse(req).parameters.charset.toLowerCase()
  ***REMOVED*** catch (e) ***REMOVED***
    return undefined
  ***REMOVED***
***REMOVED***

/**
 * Get the simple type checker.
 *
 * @param ***REMOVED***string***REMOVED*** type
 * @return ***REMOVED***function***REMOVED***
 */

function typeChecker (type) ***REMOVED***
  return function checkType (req) ***REMOVED***
    return Boolean(typeis(req, type))
  ***REMOVED***
***REMOVED***
