// readable-stream/lib/_stream_duplex.js
// a duplex stream is just a stream that is both readable and writable.
// Since JS doesn't have multiple prototypal inheritance, this class
// prototypally inherits from Readable, and then parasitically from
// Writable.

'use strict';

/*<replacement>*/

var objectKeys = Object.keys || function (obj) ***REMOVED***
  var keys = [];
  for (var key in obj) ***REMOVED***
    keys.push(key);
  ***REMOVED***return keys;
***REMOVED***;
/*</replacement>*/

module.exports = Duplex;

/*<replacement>*/
var processNextTick = require('process-nextick-args');
/*</replacement>*/

/*<replacement>*/
var util = require('core-util-is');
util.inherits = require('inherits');
/*</replacement>*/

var Readable = require('./_stream_readable');
var Writable = require('./_stream_writable');

util.inherits(Duplex, Readable);

var keys = objectKeys(Writable.prototype);
for (var v = 0; v < keys.length; v++) ***REMOVED***
  var method = keys[v];
  if (!Duplex.prototype[method]) Duplex.prototype[method] = Writable.prototype[method];
***REMOVED***

function Duplex(options) ***REMOVED***
  if (!(this instanceof Duplex)) return new Duplex(options);

  Readable.call(this, options);
  Writable.call(this, options);

  if (options && options.readable === false) this.readable = false;

  if (options && options.writable === false) this.writable = false;

  this.allowHalfOpen = true;
  if (options && options.allowHalfOpen === false) this.allowHalfOpen = false;

  this.once('end', onend);
***REMOVED***

// the no-half-open enforcer
function onend() ***REMOVED***
  // if we allow half-open state, or if the writable side ended,
  // then we're ok.
  if (this.allowHalfOpen || this._writableState.ended) return;

  // no more data can be written.
  // But allow more writes to happen in this tick.
  processNextTick(onEndNT, this);
***REMOVED***

function onEndNT(self) ***REMOVED***
  self.end();
***REMOVED***

function forEach(xs, f) ***REMOVED***
  for (var i = 0, l = xs.length; i < l; i++) ***REMOVED***
    f(xs[i], i);
  ***REMOVED***
***REMOVED***
