// mongoose/lib/index.js
'use strict';

/*!
 * Module dependencies.
 */

var Schema = require('./schema'),
    SchemaType = require('./schematype'),
    VirtualType = require('./virtualtype'),
    STATES = require('./connectionstate'),
    Types = require('./types'),
    Query = require('./query'),
    Model = require('./model'),
    Document = require('./document'),
    utils = require('./utils'),
    format = utils.toCollectionName,
    pkg = require('../package.json');

var querystring = require('querystring');

var Aggregate = require('./aggregate');
var PromiseProvider = require('./promise_provider');

/**
 * Mongoose constructor.
 *
 * The exports object of the `mongoose` module is an instance of this class.
 * Most apps will only use this one instance.
 *
 * @api public
 */

function Mongoose() ***REMOVED***
  this.connections = [];
  this.plugins = [];
  this.models = ***REMOVED******REMOVED***;
  this.modelSchemas = ***REMOVED******REMOVED***;
  // default global options
  this.options = ***REMOVED***
    pluralization: true
  ***REMOVED***;
  var conn = this.createConnection(); // default connection
  conn.models = this.models;
***REMOVED***

/**
 * Expose connection states for user-land
 *
 */
Mongoose.prototype.STATES = STATES;

/**
 * Sets mongoose options
 *
 * ####Example:
 *
 *     mongoose.set('test', value) // sets the 'test' option to `value`
 *
 *     mongoose.set('debug', true) // enable logging collection methods + arguments to the console
 *
 *     mongoose.set('debug', function(collectionName, methodName, arg1, arg2...) ***REMOVED******REMOVED***); // use custom function to log collection methods + arguments
 *
 * @param ***REMOVED***String***REMOVED*** key
 * @param ***REMOVED***String|Function***REMOVED*** value
 * @api public
 */

Mongoose.prototype.set = function(key, value) ***REMOVED***
  if (arguments.length === 1) ***REMOVED***
    return this.options[key];
  ***REMOVED***

  this.options[key] = value;
  return this;
***REMOVED***;
Mongoose.prototype.set.$hasSideEffects = true;

/**
 * Gets mongoose options
 *
 * ####Example:
 *
 *     mongoose.get('test') // returns the 'test' value
 *
 * @param ***REMOVED***String***REMOVED*** key
 * @method get
 * @api public
 */

Mongoose.prototype.get = Mongoose.prototype.set;

/*!
 * ReplSet connection string check.
 */

var rgxReplSet = /^.+,.+$/;

/**
 * Checks if ?replicaSet query parameter is specified in URI
 *
 * ####Example:
 *
 *     checkReplicaSetInUri('localhost:27000?replicaSet=rs0'); // true
 *
 * @param ***REMOVED***String***REMOVED*** uri
 * @return ***REMOVED***boolean***REMOVED***
 * @api private
 */

var checkReplicaSetInUri = function(uri) ***REMOVED***
  if (!uri) ***REMOVED***
    return false;
  ***REMOVED***

  var queryStringStart = uri.indexOf('?');
  var isReplicaSet = false;
  if (queryStringStart !== -1) ***REMOVED***
    try ***REMOVED***
      var obj = querystring.parse(uri.substr(queryStringStart + 1));
      if (obj && obj.replicaSet) ***REMOVED***
        isReplicaSet = true;
      ***REMOVED***
    ***REMOVED*** catch (e) ***REMOVED***
      return false;
    ***REMOVED***
  ***REMOVED***

  return isReplicaSet;
***REMOVED***;

/**
 * Creates a Connection instance.
 *
 * Each `connection` instance maps to a single database. This method is helpful when mangaging multiple db connections.
 *
 * If arguments are passed, they are proxied to either [Connection#open](#connection_Connection-open) or [Connection#openSet](#connection_Connection-openSet) appropriately. This means we can pass `db`, `server`, and `replset` options to the driver. _Note that the `safe` option specified in your schema will overwrite the `safe` db option specified here unless you set your schemas `safe` option to `undefined`. See [this](/docs/guide.html#safe) for more information._
 *
 * _Options passed take precedence over options included in connection strings._
 *
 * ####Example:
 *
 *     // with mongodb:// URI
 *     db = mongoose.createConnection('mongodb://user:pass@localhost:port/database');
 *
 *     // and options
 *     var opts = ***REMOVED*** db: ***REMOVED*** native_parser: true ***REMOVED******REMOVED***
 *     db = mongoose.createConnection('mongodb://user:pass@localhost:port/database', opts);
 *
 *     // replica sets
 *     db = mongoose.createConnection('mongodb://user:pass@localhost:port,anotherhost:port,yetanother:port/database');
 *
 *     // and options
 *     var opts = ***REMOVED*** replset: ***REMOVED*** strategy: 'ping', rs_name: 'testSet' ***REMOVED******REMOVED***
 *     db = mongoose.createConnection('mongodb://user:pass@localhost:port,anotherhost:port,yetanother:port/database', opts);
 *
 *     // with [host, database_name[, port] signature
 *     db = mongoose.createConnection('localhost', 'database', port)
 *
 *     // and options
 *     var opts = ***REMOVED*** server: ***REMOVED*** auto_reconnect: false ***REMOVED***, user: 'username', pass: 'mypassword' ***REMOVED***
 *     db = mongoose.createConnection('localhost', 'database', port, opts)
 *
 *     // initialize now, connect later
 *     db = mongoose.createConnection();
 *     db.open('localhost', 'database', port, [opts]);
 *
 * @param ***REMOVED***String***REMOVED*** [uri] a mongodb:// URI
 * @param ***REMOVED***Object***REMOVED*** [options] options to pass to the driver
 * @param ***REMOVED***Object***REMOVED*** [options.config] mongoose-specific options
 * @param ***REMOVED***Boolean***REMOVED*** [options.config.autoIndex] set to false to disable automatic index creation for all models associated with this connection.
 * @see Connection#open #connection_Connection-open
 * @see Connection#openSet #connection_Connection-openSet
 * @return ***REMOVED***Connection***REMOVED*** the created Connection object
 * @api public
 */

Mongoose.prototype.createConnection = function(uri, options) ***REMOVED***
  var conn = new Connection(this);
  this.connections.push(conn);

  var rsOption = options && (options.replset || options.replSet);
  if (arguments.length) ***REMOVED***
    if (rgxReplSet.test(arguments[0]) || checkReplicaSetInUri(arguments[0])) ***REMOVED***
      conn.openSet.apply(conn, arguments).catch(function() ***REMOVED******REMOVED***);
    ***REMOVED*** else if (rsOption &&
        (rsOption.replicaSet || rsOption.rs_name)) ***REMOVED***
      conn.openSet.apply(conn, arguments).catch(function() ***REMOVED******REMOVED***);
    ***REMOVED*** else ***REMOVED***
      conn.open.apply(conn, arguments).catch(function() ***REMOVED******REMOVED***);
    ***REMOVED***
  ***REMOVED***

  return conn;
***REMOVED***;
Mongoose.prototype.createConnection.$hasSideEffects = true;

/**
 * Opens the default mongoose connection.
 *
 * If arguments are passed, they are proxied to either
 * [Connection#open](#connection_Connection-open) or
 * [Connection#openSet](#connection_Connection-openSet) appropriately.
 *
 * _Options passed take precedence over options included in connection strings._
 *
 * ####Example:
 *
 *     mongoose.connect('mongodb://user:pass@localhost:port/database');
 *
 *     // replica sets
 *     var uri = 'mongodb://user:pass@localhost:port,anotherhost:port,yetanother:port/mydatabase';
 *     mongoose.connect(uri);
 *
 *     // with options
 *     mongoose.connect(uri, options);
 *
 *     // connecting to multiple mongos
 *     var uri = 'mongodb://hostA:27501,hostB:27501';
 *     var opts = ***REMOVED*** mongos: true ***REMOVED***;
 *     mongoose.connect(uri, opts);
 *
 *     // optional callback that gets fired when initial connection completed
 *     var uri = 'mongodb://nonexistent.domain:27000';
 *     mongoose.connect(uri, function(error) ***REMOVED***
 *       // if error is truthy, the initial connection failed.
 *     ***REMOVED***)
 *
 * @param ***REMOVED***String***REMOVED*** uri(s)
 * @param ***REMOVED***Object***REMOVED*** [options]
 * @param ***REMOVED***Function***REMOVED*** [callback]
 * @see Mongoose#createConnection #index_Mongoose-createConnection
 * @api public
 * @return ***REMOVED***MongooseThenable***REMOVED*** pseudo-promise wrapper around this
 */

Mongoose.prototype.connect = function() ***REMOVED***
  var conn = this.connection;
  if (rgxReplSet.test(arguments[0]) || checkReplicaSetInUri(arguments[0])) ***REMOVED***
    return new MongooseThenable(this, conn.openSet.apply(conn, arguments));
  ***REMOVED***

  return new MongooseThenable(this, conn.open.apply(conn, arguments));
***REMOVED***;
Mongoose.prototype.connect.$hasSideEffects = true;

/**
 * Disconnects all connections.
 *
 * @param ***REMOVED***Function***REMOVED*** [fn] called after all connection close.
 * @return ***REMOVED***MongooseThenable***REMOVED*** pseudo-promise wrapper around this
 * @api public
 */

Mongoose.prototype.disconnect = function(fn) ***REMOVED***
  var error;
  this.connections.forEach(function(conn) ***REMOVED***
    conn.close(function(err) ***REMOVED***
      if (error) ***REMOVED***
        return;
      ***REMOVED***
      if (err) ***REMOVED***
        error = err;
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***);

  var Promise = PromiseProvider.get();
  return new MongooseThenable(this, new Promise.ES6(function(resolve, reject) ***REMOVED***
    fn && fn(error);
    if (error) ***REMOVED***
      reject(error);
      return;
    ***REMOVED***
    resolve();
  ***REMOVED***));
***REMOVED***;
Mongoose.prototype.disconnect.$hasSideEffects = true;

/**
 * Defines a model or retrieves it.
 *
 * Models defined on the `mongoose` instance are available to all connection created by the same `mongoose` instance.
 *
 * ####Example:
 *
 *     var mongoose = require('mongoose');
 *
 *     // define an Actor model with this mongoose instance
 *     mongoose.model('Actor', new Schema(***REMOVED*** name: String ***REMOVED***));
 *
 *     // create a new connection
 *     var conn = mongoose.createConnection(..);
 *
 *     // retrieve the Actor model
 *     var Actor = conn.model('Actor');
 *
 * _When no `collection` argument is passed, Mongoose produces a collection name by passing the model `name` to the [utils.toCollectionName](#utils_exports.toCollectionName) method. This method pluralizes the name. If you don't like this behavior, either pass a collection name or set your schemas collection name option._
 *
 * ####Example:
 *
 *     var schema = new Schema(***REMOVED*** name: String ***REMOVED***, ***REMOVED*** collection: 'actor' ***REMOVED***);
 *
 *     // or
 *
 *     schema.set('collection', 'actor');
 *
 *     // or
 *
 *     var collectionName = 'actor'
 *     var M = mongoose.model('Actor', schema, collectionName)
 *
 * @param ***REMOVED***String|Function***REMOVED*** name model name or class extending Model
 * @param ***REMOVED***Schema***REMOVED*** [schema]
 * @param ***REMOVED***String***REMOVED*** [collection] name (optional, inferred from model name)
 * @param ***REMOVED***Boolean***REMOVED*** [skipInit] whether to skip initialization (defaults to false)
 * @api public
 */

Mongoose.prototype.model = function(name, schema, collection, skipInit) ***REMOVED***
  var model;
  if (typeof name === 'function') ***REMOVED***
    model = name;
    name = model.name;
    if (!(model.prototype instanceof Model)) ***REMOVED***
      throw new mongoose.Error('The provided class ' + name + ' must extend Model');
    ***REMOVED***
  ***REMOVED***

  if (typeof schema === 'string') ***REMOVED***
    collection = schema;
    schema = false;
  ***REMOVED***

  if (utils.isObject(schema) && !(schema.instanceOfSchema)) ***REMOVED***
    schema = new Schema(schema);
  ***REMOVED***
  if (schema && !schema.instanceOfSchema) ***REMOVED***
    throw new Error('The 2nd parameter to `mongoose.model()` should be a ' +
      'schema or a POJO');
  ***REMOVED***

  if (typeof collection === 'boolean') ***REMOVED***
    skipInit = collection;
    collection = null;
  ***REMOVED***

  // handle internal options from connection.model()
  var options;
  if (skipInit && utils.isObject(skipInit)) ***REMOVED***
    options = skipInit;
    skipInit = true;
  ***REMOVED*** else ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  // look up schema for the collection.
  if (!this.modelSchemas[name]) ***REMOVED***
    if (schema) ***REMOVED***
      // cache it so we only apply plugins once
      this.modelSchemas[name] = schema;
    ***REMOVED*** else ***REMOVED***
      throw new mongoose.Error.MissingSchemaError(name);
    ***REMOVED***
  ***REMOVED***

  if (schema) ***REMOVED***
    this._applyPlugins(schema);
  ***REMOVED***

  var sub;

  // connection.model() may be passing a different schema for
  // an existing model name. in this case don't read from cache.
  if (this.models[name] && options.cache !== false) ***REMOVED***
    if (schema && schema.instanceOfSchema && schema !== this.models[name].schema) ***REMOVED***
      throw new mongoose.Error.OverwriteModelError(name);
    ***REMOVED***

    if (collection) ***REMOVED***
      // subclass current model with alternate collection
      model = this.models[name];
      schema = model.prototype.schema;
      sub = model.__subclass(this.connection, schema, collection);
      // do not cache the sub model
      return sub;
    ***REMOVED***

    return this.models[name];
  ***REMOVED***

  // ensure a schema exists
  if (!schema) ***REMOVED***
    schema = this.modelSchemas[name];
    if (!schema) ***REMOVED***
      throw new mongoose.Error.MissingSchemaError(name);
    ***REMOVED***
  ***REMOVED***

  // Apply relevant "global" options to the schema
  if (!('pluralization' in schema.options)) schema.options.pluralization = this.options.pluralization;


  if (!collection) ***REMOVED***
    collection = schema.get('collection') || format(name, schema.options);
  ***REMOVED***

  var connection = options.connection || this.connection;
  model = this.Model.compile(model || name, schema, collection, connection, this);

  if (!skipInit) ***REMOVED***
    model.init();
  ***REMOVED***

  if (options.cache === false) ***REMOVED***
    return model;
  ***REMOVED***

  this.models[name] = model;
  return this.models[name];
***REMOVED***;
Mongoose.prototype.model.$hasSideEffects = true;

/**
 * Returns an array of model names created on this instance of Mongoose.
 *
 * ####Note:
 *
 * _Does not include names of models created using `connection.model()`._
 *
 * @api public
 * @return ***REMOVED***Array***REMOVED***
 */

Mongoose.prototype.modelNames = function() ***REMOVED***
  var names = Object.keys(this.models);
  return names;
***REMOVED***;
Mongoose.prototype.modelNames.$hasSideEffects = true;

/**
 * Applies global plugins to `schema`.
 *
 * @param ***REMOVED***Schema***REMOVED*** schema
 * @api private
 */

Mongoose.prototype._applyPlugins = function(schema) ***REMOVED***
  if (schema.$globalPluginsApplied) ***REMOVED***
    return;
  ***REMOVED***
  var i;
  var len;
  for (i = 0, len = this.plugins.length; i < len; ++i) ***REMOVED***
    schema.plugin(this.plugins[i][0], this.plugins[i][1]);
  ***REMOVED***
  schema.$globalPluginsApplied = true;
  for (i = 0, len = schema.childSchemas.length; i < len; ++i) ***REMOVED***
    this._applyPlugins(schema.childSchemas[i]);
  ***REMOVED***
***REMOVED***;
Mongoose.prototype._applyPlugins.$hasSideEffects = true;

/**
 * Declares a global plugin executed on all Schemas.
 *
 * Equivalent to calling `.plugin(fn)` on each Schema you create.
 *
 * @param ***REMOVED***Function***REMOVED*** fn plugin callback
 * @param ***REMOVED***Object***REMOVED*** [opts] optional options
 * @return ***REMOVED***Mongoose***REMOVED*** this
 * @see plugins ./plugins.html
 * @api public
 */

Mongoose.prototype.plugin = function(fn, opts) ***REMOVED***
  this.plugins.push([fn, opts]);
  return this;
***REMOVED***;
Mongoose.prototype.plugin.$hasSideEffects = true;

/**
 * The default connection of the mongoose module.
 *
 * ####Example:
 *
 *     var mongoose = require('mongoose');
 *     mongoose.connect(...);
 *     mongoose.connection.on('error', cb);
 *
 * This is the connection used by default for every model created using [mongoose.model](#index_Mongoose-model).
 *
 * @property connection
 * @return ***REMOVED***Connection***REMOVED***
 * @api public
 */

Mongoose.prototype.__defineGetter__('connection', function() ***REMOVED***
  return this.connections[0];
***REMOVED***);

Mongoose.prototype.__defineSetter__('connection', function(v) ***REMOVED***
  this.connections[0] = v;
***REMOVED***);

/*!
 * Driver depentend APIs
 */

var driver = global.MONGOOSE_DRIVER_PATH || './drivers/node-mongodb-native';

/*!
 * Connection
 */

var Connection = require(driver + '/connection');

/*!
 * Collection
 */

var Collection = require(driver + '/collection');

/**
 * The Mongoose Aggregate constructor
 *
 * @method Aggregate
 * @api public
 */

Mongoose.prototype.Aggregate = Aggregate;

/**
 * The Mongoose Collection constructor
 *
 * @method Collection
 * @api public
 */

Mongoose.prototype.Collection = Collection;

/**
 * The Mongoose [Connection](#connection_Connection) constructor
 *
 * @method Connection
 * @api public
 */

Mongoose.prototype.Connection = Connection;

/**
 * The Mongoose version
 *
 * @property version
 * @api public
 */

Mongoose.prototype.version = pkg.version;

/**
 * The Mongoose constructor
 *
 * The exports of the mongoose module is an instance of this class.
 *
 * ####Example:
 *
 *     var mongoose = require('mongoose');
 *     var mongoose2 = new mongoose.Mongoose();
 *
 * @method Mongoose
 * @api public
 */

Mongoose.prototype.Mongoose = Mongoose;

/**
 * The Mongoose [Schema](#schema_Schema) constructor
 *
 * ####Example:
 *
 *     var mongoose = require('mongoose');
 *     var Schema = mongoose.Schema;
 *     var CatSchema = new Schema(..);
 *
 * @method Schema
 * @api public
 */

Mongoose.prototype.Schema = Schema;

/**
 * The Mongoose [SchemaType](#schematype_SchemaType) constructor
 *
 * @method SchemaType
 * @api public
 */

Mongoose.prototype.SchemaType = SchemaType;

/**
 * The various Mongoose SchemaTypes.
 *
 * ####Note:
 *
 * _Alias of mongoose.Schema.Types for backwards compatibility._
 *
 * @property SchemaTypes
 * @see Schema.SchemaTypes #schema_Schema.Types
 * @api public
 */

Mongoose.prototype.SchemaTypes = Schema.Types;

/**
 * The Mongoose [VirtualType](#virtualtype_VirtualType) constructor
 *
 * @method VirtualType
 * @api public
 */

Mongoose.prototype.VirtualType = VirtualType;

/**
 * The various Mongoose Types.
 *
 * ####Example:
 *
 *     var mongoose = require('mongoose');
 *     var array = mongoose.Types.Array;
 *
 * ####Types:
 *
 * - [ObjectId](#types-objectid-js)
 * - [Buffer](#types-buffer-js)
 * - [SubDocument](#types-embedded-js)
 * - [Array](#types-array-js)
 * - [DocumentArray](#types-documentarray-js)
 *
 * Using this exposed access to the `ObjectId` type, we can construct ids on demand.
 *
 *     var ObjectId = mongoose.Types.ObjectId;
 *     var id1 = new ObjectId;
 *
 * @property Types
 * @api public
 */

Mongoose.prototype.Types = Types;

/**
 * The Mongoose [Query](#query_Query) constructor.
 *
 * @method Query
 * @api public
 */

Mongoose.prototype.Query = Query;

/**
 * The Mongoose [Promise](#promise_Promise) constructor.
 *
 * @method Promise
 * @api public
 */

Object.defineProperty(Mongoose.prototype, 'Promise', ***REMOVED***
  get: function() ***REMOVED***
    return PromiseProvider.get();
  ***REMOVED***,
  set: function(lib) ***REMOVED***
    PromiseProvider.set(lib);
  ***REMOVED***
***REMOVED***);

/**
 * Storage layer for mongoose promises
 *
 * @method PromiseProvider
 * @api public
 */

Mongoose.prototype.PromiseProvider = PromiseProvider;

/**
 * The Mongoose [Model](#model_Model) constructor.
 *
 * @method Model
 * @api public
 */

Mongoose.prototype.Model = Model;

/**
 * The Mongoose [Document](#document-js) constructor.
 *
 * @method Document
 * @api public
 */

Mongoose.prototype.Document = Document;

/**
 * The Mongoose DocumentProvider constructor.
 *
 * @method DocumentProvider
 * @api public
 */

Mongoose.prototype.DocumentProvider = require('./document_provider');

/**
 * The [MongooseError](#error_MongooseError) constructor.
 *
 * @method Error
 * @api public
 */

Mongoose.prototype.Error = require('./error');

/**
 * The Mongoose CastError constructor
 *
 * @method CastError
 * @param ***REMOVED***String***REMOVED*** type The name of the type
 * @param ***REMOVED***Any***REMOVED*** value The value that failed to cast
 * @param ***REMOVED***String***REMOVED*** path The path `a.b.c` in the doc where this cast error occurred
 * @param ***REMOVED***Error***REMOVED*** [reason] The original error that was thrown
 * @api public
 */

Mongoose.prototype.CastError = require('./error/cast');

/**
 * The [node-mongodb-native](https://github.com/mongodb/node-mongodb-native) driver Mongoose uses.
 *
 * @property mongo
 * @api public
 */

Mongoose.prototype.mongo = require('mongodb');

/**
 * The [mquery](https://github.com/aheckmann/mquery) query builder Mongoose uses.
 *
 * @property mquery
 * @api public
 */

Mongoose.prototype.mquery = require('mquery');

/**
 * Wraps the given Mongoose instance into a thenable (pseudo-promise). This
 * is so `connect()` and `disconnect()` can return a thenable while maintaining
 * backwards compatibility.
 *
 * @api private
 */

function MongooseThenable(mongoose, promise) ***REMOVED***
  var _this = this;
  for (var key in mongoose) ***REMOVED***
    if (typeof mongoose[key] === 'function' && mongoose[key].$hasSideEffects) ***REMOVED***
      (function(key) ***REMOVED***
        _this[key] = function() ***REMOVED***
          return mongoose[key].apply(mongoose, arguments);
        ***REMOVED***;
      ***REMOVED***)(key);
    ***REMOVED*** else if (['connection', 'connections'].indexOf(key) !== -1) ***REMOVED***
      _this[key] = mongoose[key];
    ***REMOVED***
  ***REMOVED***
  this.$opPromise = promise;
***REMOVED***

MongooseThenable.prototype = new Mongoose;

/**
 * Ability to use mongoose object as a pseudo-promise so `.connect().then()`
 * and `.disconnect().then()` are viable.
 *
 * @param ***REMOVED***Function***REMOVED*** onFulfilled
 * @param ***REMOVED***Function***REMOVED*** onRejected
 * @return ***REMOVED***Promise***REMOVED***
 * @api private
 */

MongooseThenable.prototype.then = function(onFulfilled, onRejected) ***REMOVED***
  var Promise = PromiseProvider.get();
  if (!this.$opPromise) ***REMOVED***
    return new Promise.ES6(function(resolve, reject) ***REMOVED***
      reject(new Error('Can only call `.then()` if connect() or disconnect() ' +
        'has been called'));
    ***REMOVED***).then(onFulfilled, onRejected);
  ***REMOVED***
  this.$opPromise.$hasHandler = true;
  return this.$opPromise.then(onFulfilled, onRejected);
***REMOVED***;

/**
 * Ability to use mongoose object as a pseudo-promise so `.connect().then()`
 * and `.disconnect().then()` are viable.
 *
 * @param ***REMOVED***Function***REMOVED*** onFulfilled
 * @param ***REMOVED***Function***REMOVED*** onRejected
 * @return ***REMOVED***Promise***REMOVED***
 * @api private
 */

MongooseThenable.prototype.catch = function(onRejected) ***REMOVED***
  return this.then(null, onRejected);
***REMOVED***;

/*!
 * The exports object is an instance of Mongoose.
 *
 * @api public
 */

var mongoose = module.exports = exports = new Mongoose;
