// mongoose/lib/promise.js
/*!
 * Module dependencies
 */

var MPromise = require('mpromise');
var util = require('util');

/**
 * Promise constructor.
 *
 * Promises are returned from executed queries. Example:
 *
 *     var query = Candy.find(***REMOVED*** bar: true ***REMOVED***);
 *     var promise = query.exec();
 *
 * DEPRECATED. Mongoose 5.0 will use native promises by default (or bluebird,
 * if native promises are not present) but still
 * support plugging in your own ES6-compatible promises library. Mongoose 5.0
 * will **not** support mpromise.
 *
 * @param ***REMOVED***Function***REMOVED*** fn a function which will be called when the promise is resolved that accepts `fn(err, ...)***REMOVED******REMOVED***` as signature
 * @inherits mpromise https://github.com/aheckmann/mpromise
 * @inherits NodeJS EventEmitter http://nodejs.org/api/events.html#events_class_events_eventemitter
 * @event `err`: Emits when the promise is rejected
 * @event `complete`: Emits when the promise is fulfilled
 * @api public
 * @deprecated
 */

function Promise(fn) ***REMOVED***
  MPromise.call(this, fn);
***REMOVED***

/**
 * ES6-style promise constructor wrapper around mpromise.
 *
 * @param ***REMOVED***Function***REMOVED*** resolver
 * @return ***REMOVED***Promise***REMOVED*** new promise
 * @api public
 */
Promise.ES6 = function(resolver) ***REMOVED***
  var promise = new Promise();

  // No try/catch for backwards compatibility
  resolver(
    function() ***REMOVED***
      promise.complete.apply(promise, arguments);
    ***REMOVED***,
    function(e) ***REMOVED***
      promise.error(e);
    ***REMOVED***);

  return promise;
***REMOVED***;

/*!
 * Inherit from mpromise
 */

Promise.prototype = Object.create(MPromise.prototype, ***REMOVED***
  constructor: ***REMOVED***
    value: Promise,
    enumerable: false,
    writable: true,
    configurable: true
  ***REMOVED***
***REMOVED***);

/*!
 * ignore
 */

Promise.prototype.then = util.deprecate(Promise.prototype.then,
  'Mongoose: mpromise (mongoose\'s default promise library) is deprecated, ' +
  'plug in your own promise library instead: ' +
  'http://mongoosejs.com/docs/promises.html');

/**
 * ES6-style `.catch()` shorthand
 *
 * @method catch
 * @memberOf Promise
 * @param ***REMOVED***Function***REMOVED*** onReject
 * @return ***REMOVED***Promise***REMOVED***
 * @api public
 */

Promise.prototype.catch = function(onReject) ***REMOVED***
  return this.then(null, onReject);
***REMOVED***;

/*!
 * Override event names for backward compatibility.
 */

Promise.SUCCESS = 'complete';
Promise.FAILURE = 'err';

/**
 * Adds `listener` to the `event`.
 *
 * If `event` is either the success or failure event and the event has already been emitted, the`listener` is called immediately and passed the results of the original emitted event.
 *
 * @see mpromise#on https://github.com/aheckmann/mpromise#on
 * @method on
 * @memberOf Promise
 * @param ***REMOVED***String***REMOVED*** event
 * @param ***REMOVED***Function***REMOVED*** listener
 * @return ***REMOVED***Promise***REMOVED*** this
 * @api public
 */

/**
 * Rejects this promise with `reason`.
 *
 * If the promise has already been fulfilled or rejected, not action is taken.
 *
 * @see mpromise#reject https://github.com/aheckmann/mpromise#reject
 * @method reject
 * @memberOf Promise
 * @param ***REMOVED***Object|String|Error***REMOVED*** reason
 * @return ***REMOVED***Promise***REMOVED*** this
 * @api public
 */

/**
 * Rejects this promise with `err`.
 *
 * If the promise has already been fulfilled or rejected, not action is taken.
 *
 * Differs from [#reject](#promise_Promise-reject) by first casting `err` to an `Error` if it is not `instanceof Error`.
 *
 * @api public
 * @param ***REMOVED***Error|String***REMOVED*** err
 * @return ***REMOVED***Promise***REMOVED*** this
 */

Promise.prototype.error = function(err) ***REMOVED***
  if (!(err instanceof Error)) ***REMOVED***
    if (err instanceof Object) ***REMOVED***
      err = util.inspect(err);
    ***REMOVED***
    err = new Error(err);
  ***REMOVED***
  return this.reject(err);
***REMOVED***;

/**
 * Resolves this promise to a rejected state if `err` is passed or a fulfilled state if no `err` is passed.
 *
 * If the promise has already been fulfilled or rejected, not action is taken.
 *
 * `err` will be cast to an Error if not already instanceof Error.
 *
 * _NOTE: overrides [mpromise#resolve](https://github.com/aheckmann/mpromise#resolve) to provide error casting._
 *
 * @param ***REMOVED***Error***REMOVED*** [err] error or null
 * @param ***REMOVED***Object***REMOVED*** [val] value to fulfill the promise with
 * @api public
 * @deprecated
 */

Promise.prototype.resolve = function(err) ***REMOVED***
  if (err) return this.error(err);
  return this.fulfill.apply(this, Array.prototype.slice.call(arguments, 1));
***REMOVED***;

/**
 * Adds a single function as a listener to both err and complete.
 *
 * It will be executed with traditional node.js argument position when the promise is resolved.
 *
 *     promise.addBack(function (err, args...) ***REMOVED***
 *       if (err) return handleError(err);
 *       console.log('success');
 *     ***REMOVED***)
 *
 * Alias of [mpromise#onResolve](https://github.com/aheckmann/mpromise#onresolve).
 *
 * _Deprecated. Use `onResolve` instead._
 *
 * @method addBack
 * @param ***REMOVED***Function***REMOVED*** listener
 * @return ***REMOVED***Promise***REMOVED*** this
 * @deprecated
 */

Promise.prototype.addBack = Promise.prototype.onResolve;

/**
 * Fulfills this promise with passed arguments.
 *
 * @method fulfill
 * @receiver Promise
 * @see https://github.com/aheckmann/mpromise#fulfill
 * @param ***REMOVED***any***REMOVED*** args
 * @api public
 * @deprecated
 */

/**
 * Fulfills this promise with passed arguments.
 *
 * Alias of [mpromise#fulfill](https://github.com/aheckmann/mpromise#fulfill).
 *
 * _Deprecated. Use `fulfill` instead._
 *
 * @method complete
 * @receiver Promise
 * @param ***REMOVED***any***REMOVED*** args
 * @api public
 * @deprecated
 */

Promise.prototype.complete = MPromise.prototype.fulfill;

/**
 * Adds a listener to the `complete` (success) event.
 *
 * Alias of [mpromise#onFulfill](https://github.com/aheckmann/mpromise#onfulfill).
 *
 * _Deprecated. Use `onFulfill` instead._
 *
 * @method addCallback
 * @param ***REMOVED***Function***REMOVED*** listener
 * @return ***REMOVED***Promise***REMOVED*** this
 * @api public
 * @deprecated
 */

Promise.prototype.addCallback = Promise.prototype.onFulfill;

/**
 * Adds a listener to the `err` (rejected) event.
 *
 * Alias of [mpromise#onReject](https://github.com/aheckmann/mpromise#onreject).
 *
 * _Deprecated. Use `onReject` instead._
 *
 * @method addErrback
 * @param ***REMOVED***Function***REMOVED*** listener
 * @return ***REMOVED***Promise***REMOVED*** this
 * @api public
 * @deprecated
 */

Promise.prototype.addErrback = Promise.prototype.onReject;

/**
 * Creates a new promise and returns it. If `onFulfill` or `onReject` are passed, they are added as SUCCESS/ERROR callbacks to this promise after the nextTick.
 *
 * Conforms to [promises/A+](https://github.com/promises-aplus/promises-spec) specification.
 *
 * ####Example:
 *
 *     var promise = Meetups.find(***REMOVED*** tags: 'javascript' ***REMOVED***).select('_id').exec();
 *     promise.then(function (meetups) ***REMOVED***
 *       var ids = meetups.map(function (m) ***REMOVED***
 *         return m._id;
 *       ***REMOVED***);
 *       return People.find(***REMOVED*** meetups: ***REMOVED*** $in: ids ***REMOVED***).exec();
 *     ***REMOVED***).then(function (people) ***REMOVED***
 *       if (people.length < 10000) ***REMOVED***
 *         throw new Error('Too few people!!!');
 *       ***REMOVED*** else ***REMOVED***
 *         throw new Error('Still need more people!!!');
 *       ***REMOVED***
 *     ***REMOVED***).then(null, function (err) ***REMOVED***
 *       assert.ok(err instanceof Error);
 *     ***REMOVED***);
 *
 * @see promises-A+ https://github.com/promises-aplus/promises-spec
 * @see mpromise#then https://github.com/aheckmann/mpromise#then
 * @method then
 * @memberOf Promise
 * @param ***REMOVED***Function***REMOVED*** onFulFill
 * @param ***REMOVED***Function***REMOVED*** onReject
 * @return ***REMOVED***Promise***REMOVED*** newPromise
 * @deprecated
 */

/**
 * Signifies that this promise was the last in a chain of `then()s`: if a handler passed to the call to `then` which produced this promise throws, the exception will go uncaught.
 *
 * ####Example:
 *
 *     var p = new Promise;
 *     p.then(function()***REMOVED*** throw new Error('shucks') ***REMOVED***);
 *     setTimeout(function () ***REMOVED***
 *       p.fulfill();
 *       // error was caught and swallowed by the promise returned from
 *       // p.then(). we either have to always register handlers on
 *       // the returned promises or we can do the following...
 *     ***REMOVED***, 10);
 *
 *     // this time we use .end() which prevents catching thrown errors
 *     var p = new Promise;
 *     var p2 = p.then(function()***REMOVED*** throw new Error('shucks') ***REMOVED***).end(); // <--
 *     setTimeout(function () ***REMOVED***
 *       p.fulfill(); // throws "shucks"
 *     ***REMOVED***, 10);
 *
 * @api public
 * @see mpromise#end https://github.com/aheckmann/mpromise#end
 * @method end
 * @memberOf Promise
 * @deprecated
 */

/*!
 * expose
 */

module.exports = Promise;
