// mongoose/lib/querycursor.js
/*!
 * Module dependencies.
 */

var PromiseProvider = require('./promise_provider');
var Readable = require('stream').Readable;
var helpers = require('./queryhelpers');
var util = require('util');

/**
 * A QueryCursor is a concurrency primitive for processing query results
 * one document at a time. A QueryCursor fulfills the [Node.js streams3 API](https://strongloop.com/strongblog/whats-new-io-js-beta-streams3/),
 * in addition to several other mechanisms for loading documents from MongoDB
 * one at a time.
 *
 * Unless you're an advanced user, do **not** instantiate this class directly.
 * Use [`Query#cursor()`](/docs/api.html#query_Query-cursor) instead.
 *
 * @param ***REMOVED***Query***REMOVED*** query
 * @param ***REMOVED***Object***REMOVED*** options query options passed to `.find()`
 * @inherits Readable
 * @event `cursor`: Emitted when the cursor is created
 * @event `error`: Emitted when an error occurred
 * @event `data`: Emitted when the stream is flowing and the next doc is ready
 * @event `end`: Emitted when the stream is exhausted
 * @api public
 */

function QueryCursor(query, options) ***REMOVED***
  Readable.call(this, ***REMOVED*** objectMode: true ***REMOVED***);

  this.cursor = null;
  this.query = query;
  this._transforms = options.transform ? [options.transform] : [];
  var _this = this;
  var model = query.model;
  model.collection.find(query._conditions, options, function(err, cursor) ***REMOVED***
    if (_this._error) ***REMOVED***
      cursor.close(function() ***REMOVED******REMOVED***);
      _this.listeners('error').length > 0 && _this.emit('error', _this._error);
    ***REMOVED***
    if (err) ***REMOVED***
      return _this.emit('error', err);
    ***REMOVED***
    _this.cursor = cursor;
    _this.emit('cursor', cursor);
  ***REMOVED***);
***REMOVED***

util.inherits(QueryCursor, Readable);

/*!
 * Necessary to satisfy the Readable API
 */

QueryCursor.prototype._read = function() ***REMOVED***
  var _this = this;
  _next(this, function(error, doc) ***REMOVED***
    if (error) ***REMOVED***
      return _this.emit('error', error);
    ***REMOVED***
    if (!doc) ***REMOVED***
      _this.push(null);
      return _this.cursor.close(function(error) ***REMOVED***
        if (error) ***REMOVED***
          return _this.emit('error', error);
        ***REMOVED***
        _this.emit('close');
      ***REMOVED***);
    ***REMOVED***
    _this.push(doc);
  ***REMOVED***);
***REMOVED***;

/**
 * Registers a transform function which subsequently maps documents retrieved
 * via the streams interface or `.next()`
 *
 * ####Example
 *
 *     // Map documents returned by `data` events
 *     Thing.
 *       find(***REMOVED*** name: /^hello/ ***REMOVED***).
 *       cursor().
 *       map(function (doc) ***REMOVED***
 *        doc.foo = "bar";
 *        return doc;
 *       ***REMOVED***)
 *       on('data', function(doc) ***REMOVED*** console.log(doc.foo); ***REMOVED***);
 *
 *     // Or map documents returned by `.next()`
 *     var cursor = Thing.find(***REMOVED*** name: /^hello/ ***REMOVED***).
 *       cursor().
 *       map(function (doc) ***REMOVED***
 *         doc.foo = "bar";
 *         return doc;
 *       ***REMOVED***);
 *     cursor.next(function(error, doc) ***REMOVED***
 *       console.log(doc.foo);
 *     ***REMOVED***);
 *
 * @param ***REMOVED***Function***REMOVED*** fn
 * @return ***REMOVED***QueryCursor***REMOVED***
 * @api public
 * @method map
 */

QueryCursor.prototype.map = function(fn) ***REMOVED***
  this._transforms.push(fn);
  return this;
***REMOVED***;

/*!
 * Marks this cursor as errored
 */

QueryCursor.prototype._markError = function(error) ***REMOVED***
  this._error = error;
  return this;
***REMOVED***;

/**
 * Marks this cursor as closed. Will stop streaming and subsequent calls to
 * `next()` will error.
 *
 * @param ***REMOVED***Function***REMOVED*** callback
 * @return ***REMOVED***Promise***REMOVED***
 * @api public
 * @method close
 * @emits close
 * @see MongoDB driver cursor#close http://mongodb.github.io/node-mongodb-native/2.1/api/Cursor.html#close
 */

QueryCursor.prototype.close = function(callback) ***REMOVED***
  var Promise = PromiseProvider.get();
  var _this = this;
  return new Promise.ES6(function(resolve, reject) ***REMOVED***
    _this.cursor.close(function(error) ***REMOVED***
      if (error) ***REMOVED***
        callback && callback(error);
        reject(error);
        return _this.listeners('error').length > 0 &&
          _this.emit('error', error);
      ***REMOVED***
      _this.emit('close');
      resolve();
      callback && callback();
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

/**
 * Get the next document from this cursor. Will return `null` when there are
 * no documents left.
 *
 * @param ***REMOVED***Function***REMOVED*** callback
 * @return ***REMOVED***Promise***REMOVED***
 * @api public
 * @method next
 */

QueryCursor.prototype.next = function(callback) ***REMOVED***
  var Promise = PromiseProvider.get();
  var _this = this;
  return new Promise.ES6(function(resolve, reject) ***REMOVED***
    _next(_this, function(error, doc) ***REMOVED***
      if (error) ***REMOVED***
        callback && callback(error);
        return reject(error);
      ***REMOVED***
      callback && callback(null, doc);
      resolve(doc);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

/**
 * Execute `fn` for every document in the cursor. If `fn` returns a promise,
 * will wait for the promise to resolve before iterating on to the next one.
 * Returns a promise that resolves when done.
 *
 * @param ***REMOVED***Function***REMOVED*** fn
 * @param ***REMOVED***Function***REMOVED*** [callback] executed when all docs have been processed
 * @return ***REMOVED***Promise***REMOVED***
 * @api public
 * @method eachAsync
 */

QueryCursor.prototype.eachAsync = function(fn, callback) ***REMOVED***
  var Promise = PromiseProvider.get();
  var _this = this;

  var handleNextResult = function(doc, callback) ***REMOVED***
    var promise = fn(doc);
    if (promise && typeof promise.then === 'function') ***REMOVED***
      promise.then(
        function() ***REMOVED*** callback(null); ***REMOVED***,
        function(error) ***REMOVED*** callback(error); ***REMOVED***);
    ***REMOVED*** else ***REMOVED***
      callback(null);
    ***REMOVED***
  ***REMOVED***;

  var iterate = function(callback) ***REMOVED***
    return _next(_this, function(error, doc) ***REMOVED***
      if (error) ***REMOVED***
        return callback(error);
      ***REMOVED***
      if (!doc) ***REMOVED***
        return callback(null);
      ***REMOVED***
      handleNextResult(doc, function(error) ***REMOVED***
        if (error) ***REMOVED***
          return callback(error);
        ***REMOVED***
        // Make sure to clear the stack re: gh-4697
        setTimeout(function() ***REMOVED***
          iterate(callback);
        ***REMOVED***, 0);
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***;

  return new Promise.ES6(function(resolve, reject) ***REMOVED***
    iterate(function(error) ***REMOVED***
      if (error) ***REMOVED***
        callback && callback(error);
        return reject(error);
      ***REMOVED***
      callback && callback(null);
      return resolve();
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

/*!
 * Get the next doc from the underlying cursor and mongooseify it
 * (populate, etc.)
 */

function _next(ctx, cb) ***REMOVED***
  var callback = cb;
  if (ctx._transforms.length) ***REMOVED***
    callback = function(err, doc) ***REMOVED***
      if (err || doc === null) return cb(err, doc);
      cb(err, ctx._transforms.reduce(function(doc, fn) ***REMOVED***
        return fn(doc);
      ***REMOVED***, doc));
    ***REMOVED***;
  ***REMOVED***

  if (ctx._error) ***REMOVED***
    return process.nextTick(function() ***REMOVED***
      callback(ctx._error);
    ***REMOVED***);
  ***REMOVED***

  if (ctx.cursor) ***REMOVED***
    ctx.cursor.next(function(error, doc) ***REMOVED***
      if (error) ***REMOVED***
        return callback(error);
      ***REMOVED***
      if (!doc) ***REMOVED***
        return callback(null, null);
      ***REMOVED***

      var opts = ctx.query._mongooseOptions;
      if (!opts.populate) ***REMOVED***
        return opts.lean === true ?
                callback(null, doc) :
                _create(ctx, doc, null, callback);
      ***REMOVED***

      var pop = helpers.preparePopulationOptionsMQ(ctx.query,
        ctx.query._mongooseOptions);
      pop.forEach(function(option) ***REMOVED***
        delete option.model;
      ***REMOVED***);
      pop.__noPromise = true;
      ctx.query.model.populate(doc, pop, function(err, doc) ***REMOVED***
        if (err) ***REMOVED***
          return callback(err);
        ***REMOVED***
        return opts.lean === true ?
          callback(null, doc) :
          _create(ctx, doc, pop, callback);
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
    ctx.once('cursor', function() ***REMOVED***
      _next(ctx, callback);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

/*!
 * Convert a raw doc into a full mongoose doc.
 */

function _create(ctx, doc, populatedIds, cb) ***REMOVED***
  var instance = helpers.createModel(ctx.query.model, doc, ctx.query._fields);
  var opts = populatedIds ?
    ***REMOVED*** populated: populatedIds ***REMOVED*** :
    undefined;

  instance.init(doc, opts, function(err) ***REMOVED***
    if (err) ***REMOVED***
      return cb(err);
    ***REMOVED***
    cb(null, instance);
  ***REMOVED***);
***REMOVED***

module.exports = QueryCursor;
