// bluebird/js/main/es5.js
var isES5 = (function()***REMOVED***
    "use strict";
    return this === undefined;
***REMOVED***)();

if (isES5) ***REMOVED***
    module.exports = ***REMOVED***
        freeze: Object.freeze,
        defineProperty: Object.defineProperty,
        getDescriptor: Object.getOwnPropertyDescriptor,
        keys: Object.keys,
        names: Object.getOwnPropertyNames,
        getPrototypeOf: Object.getPrototypeOf,
        isArray: Array.isArray,
        isES5: isES5,
        propertyIsWritable: function(obj, prop) ***REMOVED***
            var descriptor = Object.getOwnPropertyDescriptor(obj, prop);
            return !!(!descriptor || descriptor.writable || descriptor.set);
        ***REMOVED***
    ***REMOVED***;
***REMOVED*** else ***REMOVED***
    var has = ***REMOVED******REMOVED***.hasOwnProperty;
    var str = ***REMOVED******REMOVED***.toString;
    var proto = ***REMOVED******REMOVED***.constructor.prototype;

    var ObjectKeys = function (o) ***REMOVED***
        var ret = [];
        for (var key in o) ***REMOVED***
            if (has.call(o, key)) ***REMOVED***
                ret.push(key);
            ***REMOVED***
        ***REMOVED***
        return ret;
    ***REMOVED***;

    var ObjectGetDescriptor = function(o, key) ***REMOVED***
        return ***REMOVED***value: o[key]***REMOVED***;
    ***REMOVED***;

    var ObjectDefineProperty = function (o, key, desc) ***REMOVED***
        o[key] = desc.value;
        return o;
    ***REMOVED***;

    var ObjectFreeze = function (obj) ***REMOVED***
        return obj;
    ***REMOVED***;

    var ObjectGetPrototypeOf = function (obj) ***REMOVED***
        try ***REMOVED***
            return Object(obj).constructor.prototype;
        ***REMOVED***
        catch (e) ***REMOVED***
            return proto;
        ***REMOVED***
    ***REMOVED***;

    var ArrayIsArray = function (obj) ***REMOVED***
        try ***REMOVED***
            return str.call(obj) === "[object Array]";
        ***REMOVED***
        catch(e) ***REMOVED***
            return false;
        ***REMOVED***
    ***REMOVED***;

    module.exports = ***REMOVED***
        isArray: ArrayIsArray,
        keys: ObjectKeys,
        names: ObjectKeys,
        defineProperty: ObjectDefineProperty,
        getDescriptor: ObjectGetDescriptor,
        freeze: ObjectFreeze,
        getPrototypeOf: ObjectGetPrototypeOf,
        isES5: isES5,
        propertyIsWritable: function() ***REMOVED***
            return true;
        ***REMOVED***
    ***REMOVED***;
***REMOVED***
