// bluebird/js/main/promise_array.js
"use strict";
module.exports = function(Promise, INTERNAL, tryConvertToPromise,
    apiRejection) ***REMOVED***
var util = require("./util.js");
var isArray = util.isArray;

function toResolutionValue(val) ***REMOVED***
    switch(val) ***REMOVED***
    case -2: return [];
    case -3: return ***REMOVED******REMOVED***;
    ***REMOVED***
***REMOVED***

function PromiseArray(values) ***REMOVED***
    var promise = this._promise = new Promise(INTERNAL);
    var parent;
    if (values instanceof Promise) ***REMOVED***
        parent = values;
        promise._propagateFrom(parent, 1 | 4);
    ***REMOVED***
    this._values = values;
    this._length = 0;
    this._totalResolved = 0;
    this._init(undefined, -2);
***REMOVED***
PromiseArray.prototype.length = function () ***REMOVED***
    return this._length;
***REMOVED***;

PromiseArray.prototype.promise = function () ***REMOVED***
    return this._promise;
***REMOVED***;

PromiseArray.prototype._init = function init(_, resolveValueIfEmpty) ***REMOVED***
    var values = tryConvertToPromise(this._values, this._promise);
    if (values instanceof Promise) ***REMOVED***
        values = values._target();
        this._values = values;
        if (values._isFulfilled()) ***REMOVED***
            values = values._value();
            if (!isArray(values)) ***REMOVED***
                var err = new Promise.TypeError("expecting an array, a promise or a thenable\u000a\u000a    See http://goo.gl/s8MMhc\u000a");
                this.__hardReject__(err);
                return;
            ***REMOVED***
        ***REMOVED*** else if (values._isPending()) ***REMOVED***
            values._then(
                init,
                this._reject,
                undefined,
                this,
                resolveValueIfEmpty
           );
            return;
        ***REMOVED*** else ***REMOVED***
            this._reject(values._reason());
            return;
        ***REMOVED***
    ***REMOVED*** else if (!isArray(values)) ***REMOVED***
        this._promise._reject(apiRejection("expecting an array, a promise or a thenable\u000a\u000a    See http://goo.gl/s8MMhc\u000a")._reason());
        return;
    ***REMOVED***

    if (values.length === 0) ***REMOVED***
        if (resolveValueIfEmpty === -5) ***REMOVED***
            this._resolveEmptyArray();
        ***REMOVED***
        else ***REMOVED***
            this._resolve(toResolutionValue(resolveValueIfEmpty));
        ***REMOVED***
        return;
    ***REMOVED***
    var len = this.getActualLength(values.length);
    this._length = len;
    this._values = this.shouldCopyValues() ? new Array(len) : this._values;
    var promise = this._promise;
    for (var i = 0; i < len; ++i) ***REMOVED***
        var isResolved = this._isResolved();
        var maybePromise = tryConvertToPromise(values[i], promise);
        if (maybePromise instanceof Promise) ***REMOVED***
            maybePromise = maybePromise._target();
            if (isResolved) ***REMOVED***
                maybePromise._ignoreRejections();
            ***REMOVED*** else if (maybePromise._isPending()) ***REMOVED***
                maybePromise._proxyPromiseArray(this, i);
            ***REMOVED*** else if (maybePromise._isFulfilled()) ***REMOVED***
                this._promiseFulfilled(maybePromise._value(), i);
            ***REMOVED*** else ***REMOVED***
                this._promiseRejected(maybePromise._reason(), i);
            ***REMOVED***
        ***REMOVED*** else if (!isResolved) ***REMOVED***
            this._promiseFulfilled(maybePromise, i);
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

PromiseArray.prototype._isResolved = function () ***REMOVED***
    return this._values === null;
***REMOVED***;

PromiseArray.prototype._resolve = function (value) ***REMOVED***
    this._values = null;
    this._promise._fulfill(value);
***REMOVED***;

PromiseArray.prototype.__hardReject__ =
PromiseArray.prototype._reject = function (reason) ***REMOVED***
    this._values = null;
    this._promise._rejectCallback(reason, false, true);
***REMOVED***;

PromiseArray.prototype._promiseProgressed = function (progressValue, index) ***REMOVED***
    this._promise._progress(***REMOVED***
        index: index,
        value: progressValue
    ***REMOVED***);
***REMOVED***;


PromiseArray.prototype._promiseFulfilled = function (value, index) ***REMOVED***
    this._values[index] = value;
    var totalResolved = ++this._totalResolved;
    if (totalResolved >= this._length) ***REMOVED***
        this._resolve(this._values);
    ***REMOVED***
***REMOVED***;

PromiseArray.prototype._promiseRejected = function (reason, index) ***REMOVED***
    this._totalResolved++;
    this._reject(reason);
***REMOVED***;

PromiseArray.prototype.shouldCopyValues = function () ***REMOVED***
    return true;
***REMOVED***;

PromiseArray.prototype.getActualLength = function (len) ***REMOVED***
    return len;
***REMOVED***;

return PromiseArray;
***REMOVED***;
