// bluebird/js/main/promise.js
"use strict";
module.exports = function() ***REMOVED***
var makeSelfResolutionError = function () ***REMOVED***
    return new TypeError("circular promise resolution chain\u000a\u000a    See http://goo.gl/LhFpo0\u000a");
***REMOVED***;
var reflect = function() ***REMOVED***
    return new Promise.PromiseInspection(this._target());
***REMOVED***;
var apiRejection = function(msg) ***REMOVED***
    return Promise.reject(new TypeError(msg));
***REMOVED***;

var util = require("./util.js");

var getDomain;
if (util.isNode) ***REMOVED***
    getDomain = function() ***REMOVED***
        var ret = process.domain;
        if (ret === undefined) ret = null;
        return ret;
    ***REMOVED***;
***REMOVED*** else ***REMOVED***
    getDomain = function() ***REMOVED***
        return null;
    ***REMOVED***;
***REMOVED***
util.notEnumerableProp(Promise, "_getDomain", getDomain);

var UNDEFINED_BINDING = ***REMOVED******REMOVED***;
var async = require("./async.js");
var errors = require("./errors.js");
var TypeError = Promise.TypeError = errors.TypeError;
Promise.RangeError = errors.RangeError;
Promise.CancellationError = errors.CancellationError;
Promise.TimeoutError = errors.TimeoutError;
Promise.OperationalError = errors.OperationalError;
Promise.RejectionError = errors.OperationalError;
Promise.AggregateError = errors.AggregateError;
var INTERNAL = function()***REMOVED******REMOVED***;
var APPLY = ***REMOVED******REMOVED***;
var NEXT_FILTER = ***REMOVED***e: null***REMOVED***;
var tryConvertToPromise = require("./thenables.js")(Promise, INTERNAL);
var PromiseArray =
    require("./promise_array.js")(Promise, INTERNAL,
                                    tryConvertToPromise, apiRejection);
var CapturedTrace = require("./captured_trace.js")();
var isDebugging = require("./debuggability.js")(Promise, CapturedTrace);
 /*jshint unused:false*/
var createContext =
    require("./context.js")(Promise, CapturedTrace, isDebugging);
var CatchFilter = require("./catch_filter.js")(NEXT_FILTER);
var PromiseResolver = require("./promise_resolver.js");
var nodebackForPromise = PromiseResolver._nodebackForPromise;
var errorObj = util.errorObj;
var tryCatch = util.tryCatch;
function Promise(resolver) ***REMOVED***
    if (typeof resolver !== "function") ***REMOVED***
        throw new TypeError("the promise constructor requires a resolver function\u000a\u000a    See http://goo.gl/EC22Yn\u000a");
    ***REMOVED***
    if (this.constructor !== Promise) ***REMOVED***
        throw new TypeError("the promise constructor cannot be invoked directly\u000a\u000a    See http://goo.gl/KsIlge\u000a");
    ***REMOVED***
    this._bitField = 0;
    this._fulfillmentHandler0 = undefined;
    this._rejectionHandler0 = undefined;
    this._progressHandler0 = undefined;
    this._promise0 = undefined;
    this._receiver0 = undefined;
    this._settledValue = undefined;
    if (resolver !== INTERNAL) this._resolveFromResolver(resolver);
***REMOVED***

Promise.prototype.toString = function () ***REMOVED***
    return "[object Promise]";
***REMOVED***;

Promise.prototype.caught = Promise.prototype["catch"] = function (fn) ***REMOVED***
    var len = arguments.length;
    if (len > 1) ***REMOVED***
        var catchInstances = new Array(len - 1),
            j = 0, i;
        for (i = 0; i < len - 1; ++i) ***REMOVED***
            var item = arguments[i];
            if (typeof item === "function") ***REMOVED***
                catchInstances[j++] = item;
            ***REMOVED*** else ***REMOVED***
                return Promise.reject(
                    new TypeError("Catch filter must inherit from Error or be a simple predicate function\u000a\u000a    See http://goo.gl/o84o68\u000a"));
            ***REMOVED***
        ***REMOVED***
        catchInstances.length = j;
        fn = arguments[i];
        var catchFilter = new CatchFilter(catchInstances, fn, this);
        return this._then(undefined, catchFilter.doFilter, undefined,
            catchFilter, undefined);
    ***REMOVED***
    return this._then(undefined, fn, undefined, undefined, undefined);
***REMOVED***;

Promise.prototype.reflect = function () ***REMOVED***
    return this._then(reflect, reflect, undefined, this, undefined);
***REMOVED***;

Promise.prototype.then = function (didFulfill, didReject, didProgress) ***REMOVED***
    if (isDebugging() && arguments.length > 0 &&
        typeof didFulfill !== "function" &&
        typeof didReject !== "function") ***REMOVED***
        var msg = ".then() only accepts functions but was passed: " +
                util.classString(didFulfill);
        if (arguments.length > 1) ***REMOVED***
            msg += ", " + util.classString(didReject);
        ***REMOVED***
        this._warn(msg);
    ***REMOVED***
    return this._then(didFulfill, didReject, didProgress,
        undefined, undefined);
***REMOVED***;

Promise.prototype.done = function (didFulfill, didReject, didProgress) ***REMOVED***
    var promise = this._then(didFulfill, didReject, didProgress,
        undefined, undefined);
    promise._setIsFinal();
***REMOVED***;

Promise.prototype.spread = function (didFulfill, didReject) ***REMOVED***
    return this.all()._then(didFulfill, didReject, undefined, APPLY, undefined);
***REMOVED***;

Promise.prototype.isCancellable = function () ***REMOVED***
    return !this.isResolved() &&
        this._cancellable();
***REMOVED***;

Promise.prototype.toJSON = function () ***REMOVED***
    var ret = ***REMOVED***
        isFulfilled: false,
        isRejected: false,
        fulfillmentValue: undefined,
        rejectionReason: undefined
    ***REMOVED***;
    if (this.isFulfilled()) ***REMOVED***
        ret.fulfillmentValue = this.value();
        ret.isFulfilled = true;
    ***REMOVED*** else if (this.isRejected()) ***REMOVED***
        ret.rejectionReason = this.reason();
        ret.isRejected = true;
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.prototype.all = function () ***REMOVED***
    return new PromiseArray(this).promise();
***REMOVED***;

Promise.prototype.error = function (fn) ***REMOVED***
    return this.caught(util.originatesFromRejection, fn);
***REMOVED***;

Promise.is = function (val) ***REMOVED***
    return val instanceof Promise;
***REMOVED***;

Promise.fromNode = function(fn) ***REMOVED***
    var ret = new Promise(INTERNAL);
    var result = tryCatch(fn)(nodebackForPromise(ret));
    if (result === errorObj) ***REMOVED***
        ret._rejectCallback(result.e, true, true);
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.all = function (promises) ***REMOVED***
    return new PromiseArray(promises).promise();
***REMOVED***;

Promise.defer = Promise.pending = function () ***REMOVED***
    var promise = new Promise(INTERNAL);
    return new PromiseResolver(promise);
***REMOVED***;

Promise.cast = function (obj) ***REMOVED***
    var ret = tryConvertToPromise(obj);
    if (!(ret instanceof Promise)) ***REMOVED***
        var val = ret;
        ret = new Promise(INTERNAL);
        ret._fulfillUnchecked(val);
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.resolve = Promise.fulfilled = Promise.cast;

Promise.reject = Promise.rejected = function (reason) ***REMOVED***
    var ret = new Promise(INTERNAL);
    ret._captureStackTrace();
    ret._rejectCallback(reason, true);
    return ret;
***REMOVED***;

Promise.setScheduler = function(fn) ***REMOVED***
    if (typeof fn !== "function") throw new TypeError("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");
    var prev = async._schedule;
    async._schedule = fn;
    return prev;
***REMOVED***;

Promise.prototype._then = function (
    didFulfill,
    didReject,
    didProgress,
    receiver,
    internalData
) ***REMOVED***
    var haveInternalData = internalData !== undefined;
    var ret = haveInternalData ? internalData : new Promise(INTERNAL);

    if (!haveInternalData) ***REMOVED***
        ret._propagateFrom(this, 4 | 1);
        ret._captureStackTrace();
    ***REMOVED***

    var target = this._target();
    if (target !== this) ***REMOVED***
        if (receiver === undefined) receiver = this._boundTo;
        if (!haveInternalData) ret._setIsMigrated();
    ***REMOVED***

    var callbackIndex = target._addCallbacks(didFulfill,
                                             didReject,
                                             didProgress,
                                             ret,
                                             receiver,
                                             getDomain());

    if (target._isResolved() && !target._isSettlePromisesQueued()) ***REMOVED***
        async.invoke(
            target._settlePromiseAtPostResolution, target, callbackIndex);
    ***REMOVED***

    return ret;
***REMOVED***;

Promise.prototype._settlePromiseAtPostResolution = function (index) ***REMOVED***
    if (this._isRejectionUnhandled()) this._unsetRejectionIsUnhandled();
    this._settlePromiseAt(index);
***REMOVED***;

Promise.prototype._length = function () ***REMOVED***
    return this._bitField & 131071;
***REMOVED***;

Promise.prototype._isFollowingOrFulfilledOrRejected = function () ***REMOVED***
    return (this._bitField & 939524096) > 0;
***REMOVED***;

Promise.prototype._isFollowing = function () ***REMOVED***
    return (this._bitField & 536870912) === 536870912;
***REMOVED***;

Promise.prototype._setLength = function (len) ***REMOVED***
    this._bitField = (this._bitField & -131072) |
        (len & 131071);
***REMOVED***;

Promise.prototype._setFulfilled = function () ***REMOVED***
    this._bitField = this._bitField | 268435456;
***REMOVED***;

Promise.prototype._setRejected = function () ***REMOVED***
    this._bitField = this._bitField | 134217728;
***REMOVED***;

Promise.prototype._setFollowing = function () ***REMOVED***
    this._bitField = this._bitField | 536870912;
***REMOVED***;

Promise.prototype._setIsFinal = function () ***REMOVED***
    this._bitField = this._bitField | 33554432;
***REMOVED***;

Promise.prototype._isFinal = function () ***REMOVED***
    return (this._bitField & 33554432) > 0;
***REMOVED***;

Promise.prototype._cancellable = function () ***REMOVED***
    return (this._bitField & 67108864) > 0;
***REMOVED***;

Promise.prototype._setCancellable = function () ***REMOVED***
    this._bitField = this._bitField | 67108864;
***REMOVED***;

Promise.prototype._unsetCancellable = function () ***REMOVED***
    this._bitField = this._bitField & (~67108864);
***REMOVED***;

Promise.prototype._setIsMigrated = function () ***REMOVED***
    this._bitField = this._bitField | 4194304;
***REMOVED***;

Promise.prototype._unsetIsMigrated = function () ***REMOVED***
    this._bitField = this._bitField & (~4194304);
***REMOVED***;

Promise.prototype._isMigrated = function () ***REMOVED***
    return (this._bitField & 4194304) > 0;
***REMOVED***;

Promise.prototype._receiverAt = function (index) ***REMOVED***
    var ret = index === 0
        ? this._receiver0
        : this[
            index * 5 - 5 + 4];
    if (ret === UNDEFINED_BINDING) ***REMOVED***
        return undefined;
    ***REMOVED*** else if (ret === undefined && this._isBound()) ***REMOVED***
        return this._boundValue();
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.prototype._promiseAt = function (index) ***REMOVED***
    return index === 0
        ? this._promise0
        : this[index * 5 - 5 + 3];
***REMOVED***;

Promise.prototype._fulfillmentHandlerAt = function (index) ***REMOVED***
    return index === 0
        ? this._fulfillmentHandler0
        : this[index * 5 - 5 + 0];
***REMOVED***;

Promise.prototype._rejectionHandlerAt = function (index) ***REMOVED***
    return index === 0
        ? this._rejectionHandler0
        : this[index * 5 - 5 + 1];
***REMOVED***;

Promise.prototype._boundValue = function() ***REMOVED***
    var ret = this._boundTo;
    if (ret !== undefined) ***REMOVED***
        if (ret instanceof Promise) ***REMOVED***
            if (ret.isFulfilled()) ***REMOVED***
                return ret.value();
            ***REMOVED*** else ***REMOVED***
                return undefined;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.prototype._migrateCallbacks = function (follower, index) ***REMOVED***
    var fulfill = follower._fulfillmentHandlerAt(index);
    var reject = follower._rejectionHandlerAt(index);
    var progress = follower._progressHandlerAt(index);
    var promise = follower._promiseAt(index);
    var receiver = follower._receiverAt(index);
    if (promise instanceof Promise) promise._setIsMigrated();
    if (receiver === undefined) receiver = UNDEFINED_BINDING;
    this._addCallbacks(fulfill, reject, progress, promise, receiver, null);
***REMOVED***;

Promise.prototype._addCallbacks = function (
    fulfill,
    reject,
    progress,
    promise,
    receiver,
    domain
) ***REMOVED***
    var index = this._length();

    if (index >= 131071 - 5) ***REMOVED***
        index = 0;
        this._setLength(0);
    ***REMOVED***

    if (index === 0) ***REMOVED***
        this._promise0 = promise;
        if (receiver !== undefined) this._receiver0 = receiver;
        if (typeof fulfill === "function" && !this._isCarryingStackTrace()) ***REMOVED***
            this._fulfillmentHandler0 =
                domain === null ? fulfill : domain.bind(fulfill);
        ***REMOVED***
        if (typeof reject === "function") ***REMOVED***
            this._rejectionHandler0 =
                domain === null ? reject : domain.bind(reject);
        ***REMOVED***
        if (typeof progress === "function") ***REMOVED***
            this._progressHandler0 =
                domain === null ? progress : domain.bind(progress);
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        var base = index * 5 - 5;
        this[base + 3] = promise;
        this[base + 4] = receiver;
        if (typeof fulfill === "function") ***REMOVED***
            this[base + 0] =
                domain === null ? fulfill : domain.bind(fulfill);
        ***REMOVED***
        if (typeof reject === "function") ***REMOVED***
            this[base + 1] =
                domain === null ? reject : domain.bind(reject);
        ***REMOVED***
        if (typeof progress === "function") ***REMOVED***
            this[base + 2] =
                domain === null ? progress : domain.bind(progress);
        ***REMOVED***
    ***REMOVED***
    this._setLength(index + 1);
    return index;
***REMOVED***;

Promise.prototype._setProxyHandlers = function (receiver, promiseSlotValue) ***REMOVED***
    var index = this._length();

    if (index >= 131071 - 5) ***REMOVED***
        index = 0;
        this._setLength(0);
    ***REMOVED***
    if (index === 0) ***REMOVED***
        this._promise0 = promiseSlotValue;
        this._receiver0 = receiver;
    ***REMOVED*** else ***REMOVED***
        var base = index * 5 - 5;
        this[base + 3] = promiseSlotValue;
        this[base + 4] = receiver;
    ***REMOVED***
    this._setLength(index + 1);
***REMOVED***;

Promise.prototype._proxyPromiseArray = function (promiseArray, index) ***REMOVED***
    this._setProxyHandlers(promiseArray, index);
***REMOVED***;

Promise.prototype._resolveCallback = function(value, shouldBind) ***REMOVED***
    if (this._isFollowingOrFulfilledOrRejected()) return;
    if (value === this)
        return this._rejectCallback(makeSelfResolutionError(), false, true);
    var maybePromise = tryConvertToPromise(value, this);
    if (!(maybePromise instanceof Promise)) return this._fulfill(value);

    var propagationFlags = 1 | (shouldBind ? 4 : 0);
    this._propagateFrom(maybePromise, propagationFlags);
    var promise = maybePromise._target();
    if (promise._isPending()) ***REMOVED***
        var len = this._length();
        for (var i = 0; i < len; ++i) ***REMOVED***
            promise._migrateCallbacks(this, i);
        ***REMOVED***
        this._setFollowing();
        this._setLength(0);
        this._setFollowee(promise);
    ***REMOVED*** else if (promise._isFulfilled()) ***REMOVED***
        this._fulfillUnchecked(promise._value());
    ***REMOVED*** else ***REMOVED***
        this._rejectUnchecked(promise._reason(),
            promise._getCarriedStackTrace());
    ***REMOVED***
***REMOVED***;

Promise.prototype._rejectCallback =
function(reason, synchronous, shouldNotMarkOriginatingFromRejection) ***REMOVED***
    if (!shouldNotMarkOriginatingFromRejection) ***REMOVED***
        util.markAsOriginatingFromRejection(reason);
    ***REMOVED***
    var trace = util.ensureErrorObject(reason);
    var hasStack = trace === reason;
    this._attachExtraTrace(trace, synchronous ? hasStack : false);
    this._reject(reason, hasStack ? undefined : trace);
***REMOVED***;

Promise.prototype._resolveFromResolver = function (resolver) ***REMOVED***
    var promise = this;
    this._captureStackTrace();
    this._pushContext();
    var synchronous = true;
    var r = tryCatch(resolver)(function(value) ***REMOVED***
        if (promise === null) return;
        promise._resolveCallback(value);
        promise = null;
    ***REMOVED***, function (reason) ***REMOVED***
        if (promise === null) return;
        promise._rejectCallback(reason, synchronous);
        promise = null;
    ***REMOVED***);
    synchronous = false;
    this._popContext();

    if (r !== undefined && r === errorObj && promise !== null) ***REMOVED***
        promise._rejectCallback(r.e, true, true);
        promise = null;
    ***REMOVED***
***REMOVED***;

Promise.prototype._settlePromiseFromHandler = function (
    handler, receiver, value, promise
) ***REMOVED***
    if (promise._isRejected()) return;
    promise._pushContext();
    var x;
    if (receiver === APPLY && !this._isRejected()) ***REMOVED***
        x = tryCatch(handler).apply(this._boundValue(), value);
    ***REMOVED*** else ***REMOVED***
        x = tryCatch(handler).call(receiver, value);
    ***REMOVED***
    promise._popContext();

    if (x === errorObj || x === promise || x === NEXT_FILTER) ***REMOVED***
        var err = x === promise ? makeSelfResolutionError() : x.e;
        promise._rejectCallback(err, false, true);
    ***REMOVED*** else ***REMOVED***
        promise._resolveCallback(x);
    ***REMOVED***
***REMOVED***;

Promise.prototype._target = function() ***REMOVED***
    var ret = this;
    while (ret._isFollowing()) ret = ret._followee();
    return ret;
***REMOVED***;

Promise.prototype._followee = function() ***REMOVED***
    return this._rejectionHandler0;
***REMOVED***;

Promise.prototype._setFollowee = function(promise) ***REMOVED***
    this._rejectionHandler0 = promise;
***REMOVED***;

Promise.prototype._cleanValues = function () ***REMOVED***
    if (this._cancellable()) ***REMOVED***
        this._cancellationParent = undefined;
    ***REMOVED***
***REMOVED***;

Promise.prototype._propagateFrom = function (parent, flags) ***REMOVED***
    if ((flags & 1) > 0 && parent._cancellable()) ***REMOVED***
        this._setCancellable();
        this._cancellationParent = parent;
    ***REMOVED***
    if ((flags & 4) > 0 && parent._isBound()) ***REMOVED***
        this._setBoundTo(parent._boundTo);
    ***REMOVED***
***REMOVED***;

Promise.prototype._fulfill = function (value) ***REMOVED***
    if (this._isFollowingOrFulfilledOrRejected()) return;
    this._fulfillUnchecked(value);
***REMOVED***;

Promise.prototype._reject = function (reason, carriedStackTrace) ***REMOVED***
    if (this._isFollowingOrFulfilledOrRejected()) return;
    this._rejectUnchecked(reason, carriedStackTrace);
***REMOVED***;

Promise.prototype._settlePromiseAt = function (index) ***REMOVED***
    var promise = this._promiseAt(index);
    var isPromise = promise instanceof Promise;

    if (isPromise && promise._isMigrated()) ***REMOVED***
        promise._unsetIsMigrated();
        return async.invoke(this._settlePromiseAt, this, index);
    ***REMOVED***
    var handler = this._isFulfilled()
        ? this._fulfillmentHandlerAt(index)
        : this._rejectionHandlerAt(index);

    var carriedStackTrace =
        this._isCarryingStackTrace() ? this._getCarriedStackTrace() : undefined;
    var value = this._settledValue;
    var receiver = this._receiverAt(index);
    this._clearCallbackDataAtIndex(index);

    if (typeof handler === "function") ***REMOVED***
        if (!isPromise) ***REMOVED***
            handler.call(receiver, value, promise);
        ***REMOVED*** else ***REMOVED***
            this._settlePromiseFromHandler(handler, receiver, value, promise);
        ***REMOVED***
    ***REMOVED*** else if (receiver instanceof PromiseArray) ***REMOVED***
        if (!receiver._isResolved()) ***REMOVED***
            if (this._isFulfilled()) ***REMOVED***
                receiver._promiseFulfilled(value, promise);
            ***REMOVED***
            else ***REMOVED***
                receiver._promiseRejected(value, promise);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED*** else if (isPromise) ***REMOVED***
        if (this._isFulfilled()) ***REMOVED***
            promise._fulfill(value);
        ***REMOVED*** else ***REMOVED***
            promise._reject(value, carriedStackTrace);
        ***REMOVED***
    ***REMOVED***

    if (index >= 4 && (index & 31) === 4)
        async.invokeLater(this._setLength, this, 0);
***REMOVED***;

Promise.prototype._clearCallbackDataAtIndex = function(index) ***REMOVED***
    if (index === 0) ***REMOVED***
        if (!this._isCarryingStackTrace()) ***REMOVED***
            this._fulfillmentHandler0 = undefined;
        ***REMOVED***
        this._rejectionHandler0 =
        this._progressHandler0 =
        this._receiver0 =
        this._promise0 = undefined;
    ***REMOVED*** else ***REMOVED***
        var base = index * 5 - 5;
        this[base + 3] =
        this[base + 4] =
        this[base + 0] =
        this[base + 1] =
        this[base + 2] = undefined;
    ***REMOVED***
***REMOVED***;

Promise.prototype._isSettlePromisesQueued = function () ***REMOVED***
    return (this._bitField &
            -1073741824) === -1073741824;
***REMOVED***;

Promise.prototype._setSettlePromisesQueued = function () ***REMOVED***
    this._bitField = this._bitField | -1073741824;
***REMOVED***;

Promise.prototype._unsetSettlePromisesQueued = function () ***REMOVED***
    this._bitField = this._bitField & (~-1073741824);
***REMOVED***;

Promise.prototype._queueSettlePromises = function() ***REMOVED***
    async.settlePromises(this);
    this._setSettlePromisesQueued();
***REMOVED***;

Promise.prototype._fulfillUnchecked = function (value) ***REMOVED***
    if (value === this) ***REMOVED***
        var err = makeSelfResolutionError();
        this._attachExtraTrace(err);
        return this._rejectUnchecked(err, undefined);
    ***REMOVED***
    this._setFulfilled();
    this._settledValue = value;
    this._cleanValues();

    if (this._length() > 0) ***REMOVED***
        this._queueSettlePromises();
    ***REMOVED***
***REMOVED***;

Promise.prototype._rejectUncheckedCheckError = function (reason) ***REMOVED***
    var trace = util.ensureErrorObject(reason);
    this._rejectUnchecked(reason, trace === reason ? undefined : trace);
***REMOVED***;

Promise.prototype._rejectUnchecked = function (reason, trace) ***REMOVED***
    if (reason === this) ***REMOVED***
        var err = makeSelfResolutionError();
        this._attachExtraTrace(err);
        return this._rejectUnchecked(err);
    ***REMOVED***
    this._setRejected();
    this._settledValue = reason;
    this._cleanValues();

    if (this._isFinal()) ***REMOVED***
        async.throwLater(function(e) ***REMOVED***
            if ("stack" in e) ***REMOVED***
                async.invokeFirst(
                    CapturedTrace.unhandledRejection, undefined, e);
            ***REMOVED***
            throw e;
        ***REMOVED***, trace === undefined ? reason : trace);
        return;
    ***REMOVED***

    if (trace !== undefined && trace !== reason) ***REMOVED***
        this._setCarriedStackTrace(trace);
    ***REMOVED***

    if (this._length() > 0) ***REMOVED***
        this._queueSettlePromises();
    ***REMOVED*** else ***REMOVED***
        this._ensurePossibleRejectionHandled();
    ***REMOVED***
***REMOVED***;

Promise.prototype._settlePromises = function () ***REMOVED***
    this._unsetSettlePromisesQueued();
    var len = this._length();
    for (var i = 0; i < len; i++) ***REMOVED***
        this._settlePromiseAt(i);
    ***REMOVED***
***REMOVED***;

util.notEnumerableProp(Promise,
                       "_makeSelfResolutionError",
                       makeSelfResolutionError);

require("./progress.js")(Promise, PromiseArray);
require("./method.js")(Promise, INTERNAL, tryConvertToPromise, apiRejection);
require("./bind.js")(Promise, INTERNAL, tryConvertToPromise);
require("./finally.js")(Promise, NEXT_FILTER, tryConvertToPromise);
require("./direct_resolve.js")(Promise);
require("./synchronous_inspection.js")(Promise);
require("./join.js")(Promise, PromiseArray, tryConvertToPromise, INTERNAL);
Promise.Promise = Promise;
require('./map.js')(Promise, PromiseArray, apiRejection, tryConvertToPromise, INTERNAL);
require('./cancel.js')(Promise);
require('./using.js')(Promise, apiRejection, tryConvertToPromise, createContext);
require('./generators.js')(Promise, apiRejection, INTERNAL, tryConvertToPromise);
require('./nodeify.js')(Promise);
require('./call_get.js')(Promise);
require('./props.js')(Promise, PromiseArray, tryConvertToPromise, apiRejection);
require('./race.js')(Promise, INTERNAL, tryConvertToPromise, apiRejection);
require('./reduce.js')(Promise, PromiseArray, apiRejection, tryConvertToPromise, INTERNAL);
require('./settle.js')(Promise, PromiseArray);
require('./some.js')(Promise, PromiseArray, apiRejection);
require('./promisify.js')(Promise, INTERNAL);
require('./any.js')(Promise);
require('./each.js')(Promise, INTERNAL);
require('./timers.js')(Promise, INTERNAL);
require('./filter.js')(Promise, INTERNAL);
                                                         
    util.toFastProperties(Promise);                                          
    util.toFastProperties(Promise.prototype);                                
    function fillTypes(value) ***REMOVED***                                              
        var p = new Promise(INTERNAL);                                       
        p._fulfillmentHandler0 = value;                                      
        p._rejectionHandler0 = value;                                        
        p._progressHandler0 = value;                                         
        p._promise0 = value;                                                 
        p._receiver0 = value;                                                
        p._settledValue = value;                                             
    ***REMOVED***                                                                        
    // Complete slack tracking, opt out of field-type tracking and           
    // stabilize map                                                         
    fillTypes(***REMOVED***a: 1***REMOVED***);                                                       
    fillTypes(***REMOVED***b: 2***REMOVED***);                                                       
    fillTypes(***REMOVED***c: 3***REMOVED***);                                                       
    fillTypes(1);                                                            
    fillTypes(function()***REMOVED******REMOVED***);                                                 
    fillTypes(undefined);                                                    
    fillTypes(false);                                                        
    fillTypes(new Promise(INTERNAL));                                        
    CapturedTrace.setBounds(async.firstLineError, util.lastLineError);       
    return Promise;                                                          

***REMOVED***;
