// bluebird/js/main/promise_resolver.js
"use strict";
var util = require("./util.js");
var maybeWrapAsError = util.maybeWrapAsError;
var errors = require("./errors.js");
var TimeoutError = errors.TimeoutError;
var OperationalError = errors.OperationalError;
var haveGetters = util.haveGetters;
var es5 = require("./es5.js");

function isUntypedError(obj) ***REMOVED***
    return obj instanceof Error &&
        es5.getPrototypeOf(obj) === Error.prototype;
***REMOVED***

var rErrorKey = /^(?:name|message|stack|cause)$/;
function wrapAsOperationalError(obj) ***REMOVED***
    var ret;
    if (isUntypedError(obj)) ***REMOVED***
        ret = new OperationalError(obj);
        ret.name = obj.name;
        ret.message = obj.message;
        ret.stack = obj.stack;
        var keys = es5.keys(obj);
        for (var i = 0; i < keys.length; ++i) ***REMOVED***
            var key = keys[i];
            if (!rErrorKey.test(key)) ***REMOVED***
                ret[key] = obj[key];
            ***REMOVED***
        ***REMOVED***
        return ret;
    ***REMOVED***
    util.markAsOriginatingFromRejection(obj);
    return obj;
***REMOVED***

function nodebackForPromise(promise) ***REMOVED***
    return function(err, value) ***REMOVED***
        if (promise === null) return;

        if (err) ***REMOVED***
            var wrapped = wrapAsOperationalError(maybeWrapAsError(err));
            promise._attachExtraTrace(wrapped);
            promise._reject(wrapped);
        ***REMOVED*** else if (arguments.length > 2) ***REMOVED***
            var $_len = arguments.length;var args = new Array($_len - 1); for(var $_i = 1; $_i < $_len; ++$_i) ***REMOVED***args[$_i - 1] = arguments[$_i];***REMOVED***
            promise._fulfill(args);
        ***REMOVED*** else ***REMOVED***
            promise._fulfill(value);
        ***REMOVED***

        promise = null;
    ***REMOVED***;
***REMOVED***


var PromiseResolver;
if (!haveGetters) ***REMOVED***
    PromiseResolver = function (promise) ***REMOVED***
        this.promise = promise;
        this.asCallback = nodebackForPromise(promise);
        this.callback = this.asCallback;
    ***REMOVED***;
***REMOVED***
else ***REMOVED***
    PromiseResolver = function (promise) ***REMOVED***
        this.promise = promise;
    ***REMOVED***;
***REMOVED***
if (haveGetters) ***REMOVED***
    var prop = ***REMOVED***
        get: function() ***REMOVED***
            return nodebackForPromise(this.promise);
        ***REMOVED***
    ***REMOVED***;
    es5.defineProperty(PromiseResolver.prototype, "asCallback", prop);
    es5.defineProperty(PromiseResolver.prototype, "callback", prop);
***REMOVED***

PromiseResolver._nodebackForPromise = nodebackForPromise;

PromiseResolver.prototype.toString = function () ***REMOVED***
    return "[object PromiseResolver]";
***REMOVED***;

PromiseResolver.prototype.resolve =
PromiseResolver.prototype.fulfill = function (value) ***REMOVED***
    if (!(this instanceof PromiseResolver)) ***REMOVED***
        throw new TypeError("Illegal invocation, resolver resolve/reject must be called within a resolver context. Consider using the promise constructor instead.\u000a\u000a    See http://goo.gl/sdkXL9\u000a");
    ***REMOVED***
    this.promise._resolveCallback(value);
***REMOVED***;

PromiseResolver.prototype.reject = function (reason) ***REMOVED***
    if (!(this instanceof PromiseResolver)) ***REMOVED***
        throw new TypeError("Illegal invocation, resolver resolve/reject must be called within a resolver context. Consider using the promise constructor instead.\u000a\u000a    See http://goo.gl/sdkXL9\u000a");
    ***REMOVED***
    this.promise._rejectCallback(reason);
***REMOVED***;

PromiseResolver.prototype.progress = function (value) ***REMOVED***
    if (!(this instanceof PromiseResolver)) ***REMOVED***
        throw new TypeError("Illegal invocation, resolver resolve/reject must be called within a resolver context. Consider using the promise constructor instead.\u000a\u000a    See http://goo.gl/sdkXL9\u000a");
    ***REMOVED***
    this.promise._progress(value);
***REMOVED***;

PromiseResolver.prototype.cancel = function (err) ***REMOVED***
    this.promise.cancel(err);
***REMOVED***;

PromiseResolver.prototype.timeout = function () ***REMOVED***
    this.reject(new TimeoutError("timeout"));
***REMOVED***;

PromiseResolver.prototype.isResolved = function () ***REMOVED***
    return this.promise.isResolved();
***REMOVED***;

PromiseResolver.prototype.toJSON = function () ***REMOVED***
    return this.promise.toJSON();
***REMOVED***;

module.exports = PromiseResolver;
