// bluebird/js/main/promisify.js
"use strict";
module.exports = function(Promise, INTERNAL) ***REMOVED***
var THIS = ***REMOVED******REMOVED***;
var util = require("./util.js");
var nodebackForPromise = require("./promise_resolver.js")
    ._nodebackForPromise;
var withAppended = util.withAppended;
var maybeWrapAsError = util.maybeWrapAsError;
var canEvaluate = util.canEvaluate;
var TypeError = require("./errors").TypeError;
var defaultSuffix = "Async";
var defaultPromisified = ***REMOVED***__isPromisified__: true***REMOVED***;
var noCopyProps = [
    "arity",    "length",
    "name",
    "arguments",
    "caller",
    "callee",
    "prototype",
    "__isPromisified__"
];
var noCopyPropsPattern = new RegExp("^(?:" + noCopyProps.join("|") + ")$");

var defaultFilter = function(name) ***REMOVED***
    return util.isIdentifier(name) &&
        name.charAt(0) !== "_" &&
        name !== "constructor";
***REMOVED***;

function propsFilter(key) ***REMOVED***
    return !noCopyPropsPattern.test(key);
***REMOVED***

function isPromisified(fn) ***REMOVED***
    try ***REMOVED***
        return fn.__isPromisified__ === true;
    ***REMOVED***
    catch (e) ***REMOVED***
        return false;
    ***REMOVED***
***REMOVED***

function hasPromisified(obj, key, suffix) ***REMOVED***
    var val = util.getDataPropertyOrDefault(obj, key + suffix,
                                            defaultPromisified);
    return val ? isPromisified(val) : false;
***REMOVED***
function checkValid(ret, suffix, suffixRegexp) ***REMOVED***
    for (var i = 0; i < ret.length; i += 2) ***REMOVED***
        var key = ret[i];
        if (suffixRegexp.test(key)) ***REMOVED***
            var keyWithoutAsyncSuffix = key.replace(suffixRegexp, "");
            for (var j = 0; j < ret.length; j += 2) ***REMOVED***
                if (ret[j] === keyWithoutAsyncSuffix) ***REMOVED***
                    throw new TypeError("Cannot promisify an API that has normal methods with '%s'-suffix\u000a\u000a    See http://goo.gl/iWrZbw\u000a"
                        .replace("%s", suffix));
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***

function promisifiableMethods(obj, suffix, suffixRegexp, filter) ***REMOVED***
    var keys = util.inheritedDataKeys(obj);
    var ret = [];
    for (var i = 0; i < keys.length; ++i) ***REMOVED***
        var key = keys[i];
        var value = obj[key];
        var passesDefaultFilter = filter === defaultFilter
            ? true : defaultFilter(key, value, obj);
        if (typeof value === "function" &&
            !isPromisified(value) &&
            !hasPromisified(obj, key, suffix) &&
            filter(key, value, obj, passesDefaultFilter)) ***REMOVED***
            ret.push(key, value);
        ***REMOVED***
    ***REMOVED***
    checkValid(ret, suffix, suffixRegexp);
    return ret;
***REMOVED***

var escapeIdentRegex = function(str) ***REMOVED***
    return str.replace(/([$])/, "\\$");
***REMOVED***;

var makeNodePromisifiedEval;
if (!false) ***REMOVED***
var switchCaseArgumentOrder = function(likelyArgumentCount) ***REMOVED***
    var ret = [likelyArgumentCount];
    var min = Math.max(0, likelyArgumentCount - 1 - 3);
    for(var i = likelyArgumentCount - 1; i >= min; --i) ***REMOVED***
        ret.push(i);
    ***REMOVED***
    for(var i = likelyArgumentCount + 1; i <= 3; ++i) ***REMOVED***
        ret.push(i);
    ***REMOVED***
    return ret;
***REMOVED***;

var argumentSequence = function(argumentCount) ***REMOVED***
    return util.filledRange(argumentCount, "_arg", "");
***REMOVED***;

var parameterDeclaration = function(parameterCount) ***REMOVED***
    return util.filledRange(
        Math.max(parameterCount, 3), "_arg", "");
***REMOVED***;

var parameterCount = function(fn) ***REMOVED***
    if (typeof fn.length === "number") ***REMOVED***
        return Math.max(Math.min(fn.length, 1023 + 1), 0);
    ***REMOVED***
    return 0;
***REMOVED***;

makeNodePromisifiedEval =
function(callback, receiver, originalName, fn) ***REMOVED***
    var newParameterCount = Math.max(0, parameterCount(fn) - 1);
    var argumentOrder = switchCaseArgumentOrder(newParameterCount);
    var shouldProxyThis = typeof callback === "string" || receiver === THIS;

    function generateCallForArgumentCount(count) ***REMOVED***
        var args = argumentSequence(count).join(", ");
        var comma = count > 0 ? ", " : "";
        var ret;
        if (shouldProxyThis) ***REMOVED***
            ret = "ret = callback.call(this, ***REMOVED******REMOVED***args***REMOVED******REMOVED***, nodeback); break;\n";
        ***REMOVED*** else ***REMOVED***
            ret = receiver === undefined
                ? "ret = callback(***REMOVED******REMOVED***args***REMOVED******REMOVED***, nodeback); break;\n"
                : "ret = callback.call(receiver, ***REMOVED******REMOVED***args***REMOVED******REMOVED***, nodeback); break;\n";
        ***REMOVED***
        return ret.replace("***REMOVED******REMOVED***args***REMOVED******REMOVED***", args).replace(", ", comma);
    ***REMOVED***

    function generateArgumentSwitchCase() ***REMOVED***
        var ret = "";
        for (var i = 0; i < argumentOrder.length; ++i) ***REMOVED***
            ret += "case " + argumentOrder[i] +":" +
                generateCallForArgumentCount(argumentOrder[i]);
        ***REMOVED***

        ret += "                                                             \n\
        default:                                                             \n\
            var args = new Array(len + 1);                                   \n\
            var i = 0;                                                       \n\
            for (var i = 0; i < len; ++i) ***REMOVED***                                  \n\
               args[i] = arguments[i];                                       \n\
            ***REMOVED***                                                                \n\
            args[i] = nodeback;                                              \n\
            [CodeForCall]                                                    \n\
            break;                                                           \n\
        ".replace("[CodeForCall]", (shouldProxyThis
                                ? "ret = callback.apply(this, args);\n"
                                : "ret = callback.apply(receiver, args);\n"));
        return ret;
    ***REMOVED***

    var getFunctionCode = typeof callback === "string"
                                ? ("this != null ? this['"+callback+"'] : fn")
                                : "fn";

    return new Function("Promise",
                        "fn",
                        "receiver",
                        "withAppended",
                        "maybeWrapAsError",
                        "nodebackForPromise",
                        "tryCatch",
                        "errorObj",
                        "notEnumerableProp",
                        "INTERNAL","'use strict';                            \n\
        var ret = function (Parameters) ***REMOVED***                                    \n\
            'use strict';                                                    \n\
            var len = arguments.length;                                      \n\
            var promise = new Promise(INTERNAL);                             \n\
            promise._captureStackTrace();                                    \n\
            var nodeback = nodebackForPromise(promise);                      \n\
            var ret;                                                         \n\
            var callback = tryCatch([GetFunctionCode]);                      \n\
            switch(len) ***REMOVED***                                                    \n\
                [CodeForSwitchCase]                                          \n\
            ***REMOVED***                                                                \n\
            if (ret === errorObj) ***REMOVED***                                          \n\
                promise._rejectCallback(maybeWrapAsError(ret.e), true, true);\n\
            ***REMOVED***                                                                \n\
            return promise;                                                  \n\
        ***REMOVED***;                                                                   \n\
        notEnumerableProp(ret, '__isPromisified__', true);                   \n\
        return ret;                                                          \n\
        "
        .replace("Parameters", parameterDeclaration(newParameterCount))
        .replace("[CodeForSwitchCase]", generateArgumentSwitchCase())
        .replace("[GetFunctionCode]", getFunctionCode))(
            Promise,
            fn,
            receiver,
            withAppended,
            maybeWrapAsError,
            nodebackForPromise,
            util.tryCatch,
            util.errorObj,
            util.notEnumerableProp,
            INTERNAL
        );
***REMOVED***;
***REMOVED***

function makeNodePromisifiedClosure(callback, receiver, _, fn) ***REMOVED***
    var defaultThis = (function() ***REMOVED***return this;***REMOVED***)();
    var method = callback;
    if (typeof method === "string") ***REMOVED***
        callback = fn;
    ***REMOVED***
    function promisified() ***REMOVED***
        var _receiver = receiver;
        if (receiver === THIS) _receiver = this;
        var promise = new Promise(INTERNAL);
        promise._captureStackTrace();
        var cb = typeof method === "string" && this !== defaultThis
            ? this[method] : callback;
        var fn = nodebackForPromise(promise);
        try ***REMOVED***
            cb.apply(_receiver, withAppended(arguments, fn));
        ***REMOVED*** catch(e) ***REMOVED***
            promise._rejectCallback(maybeWrapAsError(e), true, true);
        ***REMOVED***
        return promise;
    ***REMOVED***
    util.notEnumerableProp(promisified, "__isPromisified__", true);
    return promisified;
***REMOVED***

var makeNodePromisified = canEvaluate
    ? makeNodePromisifiedEval
    : makeNodePromisifiedClosure;

function promisifyAll(obj, suffix, filter, promisifier) ***REMOVED***
    var suffixRegexp = new RegExp(escapeIdentRegex(suffix) + "$");
    var methods =
        promisifiableMethods(obj, suffix, suffixRegexp, filter);

    for (var i = 0, len = methods.length; i < len; i+= 2) ***REMOVED***
        var key = methods[i];
        var fn = methods[i+1];
        var promisifiedKey = key + suffix;
        if (promisifier === makeNodePromisified) ***REMOVED***
            obj[promisifiedKey] =
                makeNodePromisified(key, THIS, key, fn, suffix);
        ***REMOVED*** else ***REMOVED***
            var promisified = promisifier(fn, function() ***REMOVED***
                return makeNodePromisified(key, THIS, key, fn, suffix);
            ***REMOVED***);
            util.notEnumerableProp(promisified, "__isPromisified__", true);
            obj[promisifiedKey] = promisified;
        ***REMOVED***
    ***REMOVED***
    util.toFastProperties(obj);
    return obj;
***REMOVED***

function promisify(callback, receiver) ***REMOVED***
    return makeNodePromisified(callback, receiver, undefined, callback);
***REMOVED***

Promise.promisify = function (fn, receiver) ***REMOVED***
    if (typeof fn !== "function") ***REMOVED***
        throw new TypeError("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");
    ***REMOVED***
    if (isPromisified(fn)) ***REMOVED***
        return fn;
    ***REMOVED***
    var ret = promisify(fn, arguments.length < 2 ? THIS : receiver);
    util.copyDescriptors(fn, ret, propsFilter);
    return ret;
***REMOVED***;

Promise.promisifyAll = function (target, options) ***REMOVED***
    if (typeof target !== "function" && typeof target !== "object") ***REMOVED***
        throw new TypeError("the target of promisifyAll must be an object or a function\u000a\u000a    See http://goo.gl/9ITlV0\u000a");
    ***REMOVED***
    options = Object(options);
    var suffix = options.suffix;
    if (typeof suffix !== "string") suffix = defaultSuffix;
    var filter = options.filter;
    if (typeof filter !== "function") filter = defaultFilter;
    var promisifier = options.promisifier;
    if (typeof promisifier !== "function") promisifier = makeNodePromisified;

    if (!util.isIdentifier(suffix)) ***REMOVED***
        throw new RangeError("suffix must be a valid identifier\u000a\u000a    See http://goo.gl/8FZo5V\u000a");
    ***REMOVED***

    var keys = util.inheritedDataKeys(target);
    for (var i = 0; i < keys.length; ++i) ***REMOVED***
        var value = target[keys[i]];
        if (keys[i] !== "constructor" &&
            util.isClass(value)) ***REMOVED***
            promisifyAll(value.prototype, suffix, filter, promisifier);
            promisifyAll(value, suffix, filter, promisifier);
        ***REMOVED***
    ***REMOVED***

    return promisifyAll(target, suffix, filter, promisifier);
***REMOVED***;
***REMOVED***;
