// bluebird/js/main/queue.js
"use strict";
function arrayMove(src, srcIndex, dst, dstIndex, len) ***REMOVED***
    for (var j = 0; j < len; ++j) ***REMOVED***
        dst[j + dstIndex] = src[j + srcIndex];
        src[j + srcIndex] = void 0;
    ***REMOVED***
***REMOVED***

function Queue(capacity) ***REMOVED***
    this._capacity = capacity;
    this._length = 0;
    this._front = 0;
***REMOVED***

Queue.prototype._willBeOverCapacity = function (size) ***REMOVED***
    return this._capacity < size;
***REMOVED***;

Queue.prototype._pushOne = function (arg) ***REMOVED***
    var length = this.length();
    this._checkCapacity(length + 1);
    var i = (this._front + length) & (this._capacity - 1);
    this[i] = arg;
    this._length = length + 1;
***REMOVED***;

Queue.prototype._unshiftOne = function(value) ***REMOVED***
    var capacity = this._capacity;
    this._checkCapacity(this.length() + 1);
    var front = this._front;
    var i = (((( front - 1 ) &
                    ( capacity - 1) ) ^ capacity ) - capacity );
    this[i] = value;
    this._front = i;
    this._length = this.length() + 1;
***REMOVED***;

Queue.prototype.unshift = function(fn, receiver, arg) ***REMOVED***
    this._unshiftOne(arg);
    this._unshiftOne(receiver);
    this._unshiftOne(fn);
***REMOVED***;

Queue.prototype.push = function (fn, receiver, arg) ***REMOVED***
    var length = this.length() + 3;
    if (this._willBeOverCapacity(length)) ***REMOVED***
        this._pushOne(fn);
        this._pushOne(receiver);
        this._pushOne(arg);
        return;
    ***REMOVED***
    var j = this._front + length - 3;
    this._checkCapacity(length);
    var wrapMask = this._capacity - 1;
    this[(j + 0) & wrapMask] = fn;
    this[(j + 1) & wrapMask] = receiver;
    this[(j + 2) & wrapMask] = arg;
    this._length = length;
***REMOVED***;

Queue.prototype.shift = function () ***REMOVED***
    var front = this._front,
        ret = this[front];

    this[front] = undefined;
    this._front = (front + 1) & (this._capacity - 1);
    this._length--;
    return ret;
***REMOVED***;

Queue.prototype.length = function () ***REMOVED***
    return this._length;
***REMOVED***;

Queue.prototype._checkCapacity = function (size) ***REMOVED***
    if (this._capacity < size) ***REMOVED***
        this._resizeTo(this._capacity << 1);
    ***REMOVED***
***REMOVED***;

Queue.prototype._resizeTo = function (capacity) ***REMOVED***
    var oldCapacity = this._capacity;
    this._capacity = capacity;
    var front = this._front;
    var length = this._length;
    var moveItemsCount = (front + length) & (oldCapacity - 1);
    arrayMove(this, 0, this, oldCapacity, moveItemsCount);
***REMOVED***;

module.exports = Queue;
