// bluebird/js/main/synchronous_inspection.js
"use strict";
module.exports = function(Promise) ***REMOVED***
function PromiseInspection(promise) ***REMOVED***
    if (promise !== undefined) ***REMOVED***
        promise = promise._target();
        this._bitField = promise._bitField;
        this._settledValue = promise._settledValue;
    ***REMOVED***
    else ***REMOVED***
        this._bitField = 0;
        this._settledValue = undefined;
    ***REMOVED***
***REMOVED***

PromiseInspection.prototype.value = function () ***REMOVED***
    if (!this.isFulfilled()) ***REMOVED***
        throw new TypeError("cannot get fulfillment value of a non-fulfilled promise\u000a\u000a    See http://goo.gl/hc1DLj\u000a");
    ***REMOVED***
    return this._settledValue;
***REMOVED***;

PromiseInspection.prototype.error =
PromiseInspection.prototype.reason = function () ***REMOVED***
    if (!this.isRejected()) ***REMOVED***
        throw new TypeError("cannot get rejection reason of a non-rejected promise\u000a\u000a    See http://goo.gl/hPuiwB\u000a");
    ***REMOVED***
    return this._settledValue;
***REMOVED***;

PromiseInspection.prototype.isFulfilled =
Promise.prototype._isFulfilled = function () ***REMOVED***
    return (this._bitField & 268435456) > 0;
***REMOVED***;

PromiseInspection.prototype.isRejected =
Promise.prototype._isRejected = function () ***REMOVED***
    return (this._bitField & 134217728) > 0;
***REMOVED***;

PromiseInspection.prototype.isPending =
Promise.prototype._isPending = function () ***REMOVED***
    return (this._bitField & 402653184) === 0;
***REMOVED***;

PromiseInspection.prototype.isResolved =
Promise.prototype._isResolved = function () ***REMOVED***
    return (this._bitField & 402653184) > 0;
***REMOVED***;

Promise.prototype.isPending = function() ***REMOVED***
    return this._target()._isPending();
***REMOVED***;

Promise.prototype.isRejected = function() ***REMOVED***
    return this._target()._isRejected();
***REMOVED***;

Promise.prototype.isFulfilled = function() ***REMOVED***
    return this._target()._isFulfilled();
***REMOVED***;

Promise.prototype.isResolved = function() ***REMOVED***
    return this._target()._isResolved();
***REMOVED***;

Promise.prototype._value = function() ***REMOVED***
    return this._settledValue;
***REMOVED***;

Promise.prototype._reason = function() ***REMOVED***
    this._unsetRejectionIsUnhandled();
    return this._settledValue;
***REMOVED***;

Promise.prototype.value = function() ***REMOVED***
    var target = this._target();
    if (!target.isFulfilled()) ***REMOVED***
        throw new TypeError("cannot get fulfillment value of a non-fulfilled promise\u000a\u000a    See http://goo.gl/hc1DLj\u000a");
    ***REMOVED***
    return target._settledValue;
***REMOVED***;

Promise.prototype.reason = function() ***REMOVED***
    var target = this._target();
    if (!target.isRejected()) ***REMOVED***
        throw new TypeError("cannot get rejection reason of a non-rejected promise\u000a\u000a    See http://goo.gl/hPuiwB\u000a");
    ***REMOVED***
    target._unsetRejectionIsUnhandled();
    return target._settledValue;
***REMOVED***;


Promise.PromiseInspection = PromiseInspection;
***REMOVED***;
