// bluebird/js/main/thenables.js
"use strict";
module.exports = function(Promise, INTERNAL) ***REMOVED***
var util = require("./util.js");
var errorObj = util.errorObj;
var isObject = util.isObject;

function tryConvertToPromise(obj, context) ***REMOVED***
    if (isObject(obj)) ***REMOVED***
        if (obj instanceof Promise) ***REMOVED***
            return obj;
        ***REMOVED***
        else if (isAnyBluebirdPromise(obj)) ***REMOVED***
            var ret = new Promise(INTERNAL);
            obj._then(
                ret._fulfillUnchecked,
                ret._rejectUncheckedCheckError,
                ret._progressUnchecked,
                ret,
                null
            );
            return ret;
        ***REMOVED***
        var then = util.tryCatch(getThen)(obj);
        if (then === errorObj) ***REMOVED***
            if (context) context._pushContext();
            var ret = Promise.reject(then.e);
            if (context) context._popContext();
            return ret;
        ***REMOVED*** else if (typeof then === "function") ***REMOVED***
            return doThenable(obj, then, context);
        ***REMOVED***
    ***REMOVED***
    return obj;
***REMOVED***

function getThen(obj) ***REMOVED***
    return obj.then;
***REMOVED***

var hasProp = ***REMOVED******REMOVED***.hasOwnProperty;
function isAnyBluebirdPromise(obj) ***REMOVED***
    return hasProp.call(obj, "_promise0");
***REMOVED***

function doThenable(x, then, context) ***REMOVED***
    var promise = new Promise(INTERNAL);
    var ret = promise;
    if (context) context._pushContext();
    promise._captureStackTrace();
    if (context) context._popContext();
    var synchronous = true;
    var result = util.tryCatch(then).call(x,
                                        resolveFromThenable,
                                        rejectFromThenable,
                                        progressFromThenable);
    synchronous = false;
    if (promise && result === errorObj) ***REMOVED***
        promise._rejectCallback(result.e, true, true);
        promise = null;
    ***REMOVED***

    function resolveFromThenable(value) ***REMOVED***
        if (!promise) return;
        promise._resolveCallback(value);
        promise = null;
    ***REMOVED***

    function rejectFromThenable(reason) ***REMOVED***
        if (!promise) return;
        promise._rejectCallback(reason, synchronous, true);
        promise = null;
    ***REMOVED***

    function progressFromThenable(value) ***REMOVED***
        if (!promise) return;
        if (typeof promise._progress === "function") ***REMOVED***
            promise._progress(value);
        ***REMOVED***
    ***REMOVED***
    return ret;
***REMOVED***

return tryConvertToPromise;
***REMOVED***;
