// bluebird/js/main/util.js
"use strict";
var es5 = require("./es5.js");
var canEvaluate = typeof navigator == "undefined";
var haveGetters = (function()***REMOVED***
    try ***REMOVED***
        var o = ***REMOVED******REMOVED***;
        es5.defineProperty(o, "f", ***REMOVED***
            get: function () ***REMOVED***
                return 3;
            ***REMOVED***
        ***REMOVED***);
        return o.f === 3;
    ***REMOVED***
    catch (e) ***REMOVED***
        return false;
    ***REMOVED***

***REMOVED***)();

var errorObj = ***REMOVED***e: ***REMOVED******REMOVED******REMOVED***;
var tryCatchTarget;
function tryCatcher() ***REMOVED***
    try ***REMOVED***
        var target = tryCatchTarget;
        tryCatchTarget = null;
        return target.apply(this, arguments);
    ***REMOVED*** catch (e) ***REMOVED***
        errorObj.e = e;
        return errorObj;
    ***REMOVED***
***REMOVED***
function tryCatch(fn) ***REMOVED***
    tryCatchTarget = fn;
    return tryCatcher;
***REMOVED***

var inherits = function(Child, Parent) ***REMOVED***
    var hasProp = ***REMOVED******REMOVED***.hasOwnProperty;

    function T() ***REMOVED***
        this.constructor = Child;
        this.constructor$ = Parent;
        for (var propertyName in Parent.prototype) ***REMOVED***
            if (hasProp.call(Parent.prototype, propertyName) &&
                propertyName.charAt(propertyName.length-1) !== "$"
           ) ***REMOVED***
                this[propertyName + "$"] = Parent.prototype[propertyName];
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    T.prototype = Parent.prototype;
    Child.prototype = new T();
    return Child.prototype;
***REMOVED***;


function isPrimitive(val) ***REMOVED***
    return val == null || val === true || val === false ||
        typeof val === "string" || typeof val === "number";

***REMOVED***

function isObject(value) ***REMOVED***
    return !isPrimitive(value);
***REMOVED***

function maybeWrapAsError(maybeError) ***REMOVED***
    if (!isPrimitive(maybeError)) return maybeError;

    return new Error(safeToString(maybeError));
***REMOVED***

function withAppended(target, appendee) ***REMOVED***
    var len = target.length;
    var ret = new Array(len + 1);
    var i;
    for (i = 0; i < len; ++i) ***REMOVED***
        ret[i] = target[i];
    ***REMOVED***
    ret[i] = appendee;
    return ret;
***REMOVED***

function getDataPropertyOrDefault(obj, key, defaultValue) ***REMOVED***
    if (es5.isES5) ***REMOVED***
        var desc = Object.getOwnPropertyDescriptor(obj, key);

        if (desc != null) ***REMOVED***
            return desc.get == null && desc.set == null
                    ? desc.value
                    : defaultValue;
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        return ***REMOVED******REMOVED***.hasOwnProperty.call(obj, key) ? obj[key] : undefined;
    ***REMOVED***
***REMOVED***

function notEnumerableProp(obj, name, value) ***REMOVED***
    if (isPrimitive(obj)) return obj;
    var descriptor = ***REMOVED***
        value: value,
        configurable: true,
        enumerable: false,
        writable: true
    ***REMOVED***;
    es5.defineProperty(obj, name, descriptor);
    return obj;
***REMOVED***

function thrower(r) ***REMOVED***
    throw r;
***REMOVED***

var inheritedDataKeys = (function() ***REMOVED***
    var excludedPrototypes = [
        Array.prototype,
        Object.prototype,
        Function.prototype
    ];

    var isExcludedProto = function(val) ***REMOVED***
        for (var i = 0; i < excludedPrototypes.length; ++i) ***REMOVED***
            if (excludedPrototypes[i] === val) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED***;

    if (es5.isES5) ***REMOVED***
        var getKeys = Object.getOwnPropertyNames;
        return function(obj) ***REMOVED***
            var ret = [];
            var visitedKeys = Object.create(null);
            while (obj != null && !isExcludedProto(obj)) ***REMOVED***
                var keys;
                try ***REMOVED***
                    keys = getKeys(obj);
                ***REMOVED*** catch (e) ***REMOVED***
                    return ret;
                ***REMOVED***
                for (var i = 0; i < keys.length; ++i) ***REMOVED***
                    var key = keys[i];
                    if (visitedKeys[key]) continue;
                    visitedKeys[key] = true;
                    var desc = Object.getOwnPropertyDescriptor(obj, key);
                    if (desc != null && desc.get == null && desc.set == null) ***REMOVED***
                        ret.push(key);
                    ***REMOVED***
                ***REMOVED***
                obj = es5.getPrototypeOf(obj);
            ***REMOVED***
            return ret;
        ***REMOVED***;
    ***REMOVED*** else ***REMOVED***
        var hasProp = ***REMOVED******REMOVED***.hasOwnProperty;
        return function(obj) ***REMOVED***
            if (isExcludedProto(obj)) return [];
            var ret = [];

            /*jshint forin:false */
            enumeration: for (var key in obj) ***REMOVED***
                if (hasProp.call(obj, key)) ***REMOVED***
                    ret.push(key);
                ***REMOVED*** else ***REMOVED***
                    for (var i = 0; i < excludedPrototypes.length; ++i) ***REMOVED***
                        if (hasProp.call(excludedPrototypes[i], key)) ***REMOVED***
                            continue enumeration;
                        ***REMOVED***
                    ***REMOVED***
                    ret.push(key);
                ***REMOVED***
            ***REMOVED***
            return ret;
        ***REMOVED***;
    ***REMOVED***

***REMOVED***)();

var thisAssignmentPattern = /this\s*\.\s*\S+\s*=/;
function isClass(fn) ***REMOVED***
    try ***REMOVED***
        if (typeof fn === "function") ***REMOVED***
            var keys = es5.names(fn.prototype);

            var hasMethods = es5.isES5 && keys.length > 1;
            var hasMethodsOtherThanConstructor = keys.length > 0 &&
                !(keys.length === 1 && keys[0] === "constructor");
            var hasThisAssignmentAndStaticMethods =
                thisAssignmentPattern.test(fn + "") && es5.names(fn).length > 0;

            if (hasMethods || hasMethodsOtherThanConstructor ||
                hasThisAssignmentAndStaticMethods) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED*** catch (e) ***REMOVED***
        return false;
    ***REMOVED***
***REMOVED***

function toFastProperties(obj) ***REMOVED***
    /*jshint -W027,-W055,-W031*/
    function f() ***REMOVED******REMOVED***
    f.prototype = obj;
    var l = 8;
    while (l--) new f();
    return obj;
    eval(obj);
***REMOVED***

var rident = /^[a-z$_][a-z$_0-9]*$/i;
function isIdentifier(str) ***REMOVED***
    return rident.test(str);
***REMOVED***

function filledRange(count, prefix, suffix) ***REMOVED***
    var ret = new Array(count);
    for(var i = 0; i < count; ++i) ***REMOVED***
        ret[i] = prefix + i + suffix;
    ***REMOVED***
    return ret;
***REMOVED***

function safeToString(obj) ***REMOVED***
    try ***REMOVED***
        return obj + "";
    ***REMOVED*** catch (e) ***REMOVED***
        return "[no string representation]";
    ***REMOVED***
***REMOVED***

function markAsOriginatingFromRejection(e) ***REMOVED***
    try ***REMOVED***
        notEnumerableProp(e, "isOperational", true);
    ***REMOVED***
    catch(ignore) ***REMOVED******REMOVED***
***REMOVED***

function originatesFromRejection(e) ***REMOVED***
    if (e == null) return false;
    return ((e instanceof Error["__BluebirdErrorTypes__"].OperationalError) ||
        e["isOperational"] === true);
***REMOVED***

function canAttachTrace(obj) ***REMOVED***
    return obj instanceof Error && es5.propertyIsWritable(obj, "stack");
***REMOVED***

var ensureErrorObject = (function() ***REMOVED***
    if (!("stack" in new Error())) ***REMOVED***
        return function(value) ***REMOVED***
            if (canAttachTrace(value)) return value;
            try ***REMOVED***throw new Error(safeToString(value));***REMOVED***
            catch(err) ***REMOVED***return err;***REMOVED***
        ***REMOVED***;
    ***REMOVED*** else ***REMOVED***
        return function(value) ***REMOVED***
            if (canAttachTrace(value)) return value;
            return new Error(safeToString(value));
        ***REMOVED***;
    ***REMOVED***
***REMOVED***)();

function classString(obj) ***REMOVED***
    return ***REMOVED******REMOVED***.toString.call(obj);
***REMOVED***

function copyDescriptors(from, to, filter) ***REMOVED***
    var keys = es5.names(from);
    for (var i = 0; i < keys.length; ++i) ***REMOVED***
        var key = keys[i];
        if (filter(key)) ***REMOVED***
            try ***REMOVED***
                es5.defineProperty(to, key, es5.getDescriptor(from, key));
            ***REMOVED*** catch (ignore) ***REMOVED******REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***

var ret = ***REMOVED***
    isClass: isClass,
    isIdentifier: isIdentifier,
    inheritedDataKeys: inheritedDataKeys,
    getDataPropertyOrDefault: getDataPropertyOrDefault,
    thrower: thrower,
    isArray: es5.isArray,
    haveGetters: haveGetters,
    notEnumerableProp: notEnumerableProp,
    isPrimitive: isPrimitive,
    isObject: isObject,
    canEvaluate: canEvaluate,
    errorObj: errorObj,
    tryCatch: tryCatch,
    inherits: inherits,
    withAppended: withAppended,
    maybeWrapAsError: maybeWrapAsError,
    toFastProperties: toFastProperties,
    filledRange: filledRange,
    toString: safeToString,
    canAttachTrace: canAttachTrace,
    ensureErrorObject: ensureErrorObject,
    originatesFromRejection: originatesFromRejection,
    markAsOriginatingFromRejection: markAsOriginatingFromRejection,
    classString: classString,
    copyDescriptors: copyDescriptors,
    hasDevTools: typeof chrome !== "undefined" && chrome &&
                 typeof chrome.loadTimes === "function",
    isNode: typeof process !== "undefined" &&
        classString(process).toLowerCase() === "[object process]"
***REMOVED***;
ret.isRecentNode = ret.isNode && (function() ***REMOVED***
    var version = process.versions.node.split(".").map(Number);
    return (version[0] === 0 && version[1] > 10) || (version[0] > 0);
***REMOVED***)();

if (ret.isNode) ret.toFastProperties(process);

try ***REMOVED***throw new Error(); ***REMOVED*** catch (e) ***REMOVED***ret.lastLineError = e;***REMOVED***
module.exports = ret;
