// mime/mime.js
var path = require('path');
var fs = require('fs');

function Mime() ***REMOVED***
  // Map of extension -> mime type
  this.types = Object.create(null);

  // Map of mime type -> extension
  this.extensions = Object.create(null);
***REMOVED***

/**
 * Define mimetype -> extension mappings.  Each key is a mime-type that maps
 * to an array of extensions associated with the type.  The first extension is
 * used as the default extension for the type.
 *
 * e.g. mime.define(***REMOVED***'audio/ogg', ['oga', 'ogg', 'spx']***REMOVED***);
 *
 * @param map (Object) type definitions
 */
Mime.prototype.define = function (map) ***REMOVED***
  for (var type in map) ***REMOVED***
    var exts = map[type];
    for (var i = 0; i < exts.length; i++) ***REMOVED***
      if (process.env.DEBUG_MIME && this.types[exts]) ***REMOVED***
        console.warn(this._loading.replace(/.*\//, ''), 'changes "' + exts[i] + '" extension type from ' +
          this.types[exts] + ' to ' + type);
      ***REMOVED***

      this.types[exts[i]] = type;
    ***REMOVED***

    // Default extension is the first one we encounter
    if (!this.extensions[type]) ***REMOVED***
      this.extensions[type] = exts[0];
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Load an Apache2-style ".types" file
 *
 * This may be called multiple times (it's expected).  Where files declare
 * overlapping types/extensions, the last file wins.
 *
 * @param file (String) path of file to load.
 */
Mime.prototype.load = function(file) ***REMOVED***
  this._loading = file;
  // Read file and split into lines
  var map = ***REMOVED******REMOVED***,
      content = fs.readFileSync(file, 'ascii'),
      lines = content.split(/[\r\n]+/);

  lines.forEach(function(line) ***REMOVED***
    // Clean up whitespace/comments, and split into fields
    var fields = line.replace(/\s*#.*|^\s*|\s*$/g, '').split(/\s+/);
    map[fields.shift()] = fields;
  ***REMOVED***);

  this.define(map);

  this._loading = null;
***REMOVED***;

/**
 * Lookup a mime type based on extension
 */
Mime.prototype.lookup = function(path, fallback) ***REMOVED***
  var ext = path.replace(/.*[\.\/\\]/, '').toLowerCase();

  return this.types[ext] || fallback || this.default_type;
***REMOVED***;

/**
 * Return file extension associated with a mime type
 */
Mime.prototype.extension = function(mimeType) ***REMOVED***
  var type = mimeType.match(/^\s*([^;\s]*)(?:;|\s|$)/)[1].toLowerCase();
  return this.extensions[type];
***REMOVED***;

// Default instance
var mime = new Mime();

// Define built-in types
mime.define(require('./types.json'));

// Default type
mime.default_type = mime.lookup('bin');

//
// Additional API specific to the default instance
//

mime.Mime = Mime;

/**
 * Lookup a charset based on mime type.
 */
mime.charsets = ***REMOVED***
  lookup: function(mimeType, fallback) ***REMOVED***
    // Assume text types are utf8
    return (/^text\//).test(mimeType) ? 'UTF-8' : fallback;
  ***REMOVED***
***REMOVED***;

module.exports = mime;
