// bson/lib/bson/map.js
"use strict"

// We have an ES6 Map available, return the native instance
if(typeof global.Map !== 'undefined') ***REMOVED***
  module.exports = global.Map;
  module.exports.Map = global.Map;
***REMOVED*** else ***REMOVED***
  // We will return a polyfill
  var Map = function(array) ***REMOVED***
    this._keys = [];
    this._values = ***REMOVED******REMOVED***;

    for(var i = 0; i < array.length; i++) ***REMOVED***
      if(array[i] == null) continue;  // skip null and undefined
      var entry = array[i];
      var key = entry[0];
      var value = entry[1];
      // Add the key to the list of keys in order
      this._keys.push(key);
      // Add the key and value to the values dictionary with a point
      // to the location in the ordered keys list
      this._values[key] = ***REMOVED***v: value, i: this._keys.length - 1***REMOVED***;
    ***REMOVED***
  ***REMOVED***

  Map.prototype.clear = function() ***REMOVED***
    this._keys = [];
    this._values = ***REMOVED******REMOVED***;
  ***REMOVED***

  Map.prototype.delete = function(key) ***REMOVED***
    var value = this._values[key];
    if(value == null) return false;
    // Delete entry
    delete this._values[key];
    // Remove the key from the ordered keys list
    this._keys.splice(value.i, 1);
    return true;
  ***REMOVED***

  Map.prototype.entries = function() ***REMOVED***
    var self = this;
    var index = 0;

    return ***REMOVED***
      next: function() ***REMOVED***
        var key = self._keys[index++];
        return ***REMOVED***
          value: key !== undefined ? [key, self._values[key].v] : undefined,
          done: key !== undefined ? false : true
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***

  Map.prototype.forEach = function(callback, self) ***REMOVED***
    self = self || this;

    for(var i = 0; i < this._keys.length; i++) ***REMOVED***
      var key = this._keys[i];
      // Call the forEach callback
      callback.call(self, this._values[key].v, key, self);
    ***REMOVED***
  ***REMOVED***

  Map.prototype.get = function(key) ***REMOVED***
    return this._values[key] ? this._values[key].v : undefined;
  ***REMOVED***

  Map.prototype.has = function(key) ***REMOVED***
    return this._values[key] != null;
  ***REMOVED***

  Map.prototype.keys = function(key) ***REMOVED***
    var self = this;
    var index = 0;

    return ***REMOVED***
      next: function() ***REMOVED***
        var key = self._keys[index++];
        return ***REMOVED***
          value: key !== undefined ? key : undefined,
          done: key !== undefined ? false : true
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***

  Map.prototype.set = function(key, value) ***REMOVED***
    if(this._values[key]) ***REMOVED***
      this._values[key].v = value;
      return this;
    ***REMOVED***

    // Add the key to the list of keys in order
    this._keys.push(key);
    // Add the key and value to the values dictionary with a point
    // to the location in the ordered keys list
    this._values[key] = ***REMOVED***v: value, i: this._keys.length - 1***REMOVED***;
    return this;
  ***REMOVED***

  Map.prototype.values = function(key, value) ***REMOVED***
    var self = this;
    var index = 0;

    return ***REMOVED***
      next: function() ***REMOVED***
        var key = self._keys[index++];
        return ***REMOVED***
          value: key !== undefined ? self._values[key].v : undefined,
          done: key !== undefined ? false : true
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***

  // Last ismaster
  Object.defineProperty(Map.prototype, 'size', ***REMOVED***
    enumerable:true,
    get: function() ***REMOVED*** return this._keys.length; ***REMOVED***
  ***REMOVED***);

  module.exports = Map;
  module.exports.Map = Map;
***REMOVED***
