// mime-types/index.js
/*!
 * mime-types
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var db = require('mime-db')
var extname = require('path').extname

/**
 * Module variables.
 * @private
 */

var extractTypeRegExp = /^\s*([^;\s]*)(?:;|\s|$)/
var textTypeRegExp = /^text\//i

/**
 * Module exports.
 * @public
 */

exports.charset = charset
exports.charsets = ***REMOVED*** lookup: charset ***REMOVED***
exports.contentType = contentType
exports.extension = extension
exports.extensions = Object.create(null)
exports.lookup = lookup
exports.types = Object.create(null)

// Populate the extensions/types maps
populateMaps(exports.extensions, exports.types)

/**
 * Get the default charset for a MIME type.
 *
 * @param ***REMOVED***string***REMOVED*** type
 * @return ***REMOVED***boolean|string***REMOVED***
 */

function charset (type) ***REMOVED***
  if (!type || typeof type !== 'string') ***REMOVED***
    return false
  ***REMOVED***

  // TODO: use media-typer
  var match = extractTypeRegExp.exec(type)
  var mime = match && db[match[1].toLowerCase()]

  if (mime && mime.charset) ***REMOVED***
    return mime.charset
  ***REMOVED***

  // default text/* to utf-8
  if (match && textTypeRegExp.test(match[1])) ***REMOVED***
    return 'UTF-8'
  ***REMOVED***

  return false
***REMOVED***

/**
 * Create a full Content-Type header given a MIME type or extension.
 *
 * @param ***REMOVED***string***REMOVED*** str
 * @return ***REMOVED***boolean|string***REMOVED***
 */

function contentType (str) ***REMOVED***
  // TODO: should this even be in this module?
  if (!str || typeof str !== 'string') ***REMOVED***
    return false
  ***REMOVED***

  var mime = str.indexOf('/') === -1
    ? exports.lookup(str)
    : str

  if (!mime) ***REMOVED***
    return false
  ***REMOVED***

  // TODO: use content-type or other module
  if (mime.indexOf('charset') === -1) ***REMOVED***
    var charset = exports.charset(mime)
    if (charset) mime += '; charset=' + charset.toLowerCase()
  ***REMOVED***

  return mime
***REMOVED***

/**
 * Get the default extension for a MIME type.
 *
 * @param ***REMOVED***string***REMOVED*** type
 * @return ***REMOVED***boolean|string***REMOVED***
 */

function extension (type) ***REMOVED***
  if (!type || typeof type !== 'string') ***REMOVED***
    return false
  ***REMOVED***

  // TODO: use media-typer
  var match = extractTypeRegExp.exec(type)

  // get extensions
  var exts = match && exports.extensions[match[1].toLowerCase()]

  if (!exts || !exts.length) ***REMOVED***
    return false
  ***REMOVED***

  return exts[0]
***REMOVED***

/**
 * Lookup the MIME type for a file path/extension.
 *
 * @param ***REMOVED***string***REMOVED*** path
 * @return ***REMOVED***boolean|string***REMOVED***
 */

function lookup (path) ***REMOVED***
  if (!path || typeof path !== 'string') ***REMOVED***
    return false
  ***REMOVED***

  // get the extension ("ext" or ".ext" or full path)
  var extension = extname('x.' + path)
    .toLowerCase()
    .substr(1)

  if (!extension) ***REMOVED***
    return false
  ***REMOVED***

  return exports.types[extension] || false
***REMOVED***

/**
 * Populate the extensions and types maps.
 * @private
 */

function populateMaps (extensions, types) ***REMOVED***
  // source preference (least -> most)
  var preference = ['nginx', 'apache', undefined, 'iana']

  Object.keys(db).forEach(function forEachMimeType (type) ***REMOVED***
    var mime = db[type]
    var exts = mime.extensions

    if (!exts || !exts.length) ***REMOVED***
      return
    ***REMOVED***

    // mime -> extensions
    extensions[type] = exts

    // extension -> mime
    for (var i = 0; i < exts.length; i++) ***REMOVED***
      var extension = exts[i]

      if (types[extension]) ***REMOVED***
        var from = preference.indexOf(db[types[extension]].source)
        var to = preference.indexOf(mime.source)

        if (types[extension] !== 'application/octet-stream' &&
          from > to || (from === to && types[extension].substr(0, 12) === 'application/')) ***REMOVED***
          // skip the remapping
          continue
        ***REMOVED***
      ***REMOVED***

      // set the extension -> mime
      types[extension] = type
    ***REMOVED***
  ***REMOVED***)
***REMOVED***
