// qs/dist/qs.js
(function(f)***REMOVED***if(typeof exports==="object"&&typeof module!=="undefined")***REMOVED***module.exports=f()***REMOVED***else if(typeof define==="function"&&define.amd)***REMOVED***define([],f)***REMOVED***else***REMOVED***var g;if(typeof window!=="undefined")***REMOVED***g=window***REMOVED***else if(typeof global!=="undefined")***REMOVED***g=global***REMOVED***else if(typeof self!=="undefined")***REMOVED***g=self***REMOVED***else***REMOVED***g=this***REMOVED***g.Qs = f()***REMOVED******REMOVED***)(function()***REMOVED***var define,module,exports;return (function e(t,n,r)***REMOVED***function s(o,u)***REMOVED***if(!n[o])***REMOVED***if(!t[o])***REMOVED***var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f***REMOVED***var l=n[o]=***REMOVED***exports:***REMOVED******REMOVED******REMOVED***;t[o][0].call(l.exports,function(e)***REMOVED***var n=t[o][1][e];return s(n?n:e)***REMOVED***,l,l.exports,e,t,n,r)***REMOVED***return n[o].exports***REMOVED***var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s***REMOVED***)(***REMOVED***1:[function(require,module,exports)***REMOVED***
'use strict';

var Stringify = require('./stringify');
var Parse = require('./parse');

module.exports = ***REMOVED***
    stringify: Stringify,
    parse: Parse
***REMOVED***;

***REMOVED***,***REMOVED***"./parse":2,"./stringify":3***REMOVED***],2:[function(require,module,exports)***REMOVED***
'use strict';

var Utils = require('./utils');

var defaults = ***REMOVED***
    delimiter: '&',
    depth: 5,
    arrayLimit: 20,
    parameterLimit: 1000,
    strictNullHandling: false,
    plainObjects: false,
    allowPrototypes: false,
    allowDots: false,
    decoder: Utils.decode
***REMOVED***;

var parseValues = function parseValues(str, options) ***REMOVED***
    var obj = ***REMOVED******REMOVED***;
    var parts = str.split(options.delimiter, options.parameterLimit === Infinity ? undefined : options.parameterLimit);

    for (var i = 0; i < parts.length; ++i) ***REMOVED***
        var part = parts[i];
        var pos = part.indexOf(']=') === -1 ? part.indexOf('=') : part.indexOf(']=') + 1;

        if (pos === -1) ***REMOVED***
            obj[options.decoder(part)] = '';

            if (options.strictNullHandling) ***REMOVED***
                obj[options.decoder(part)] = null;
            ***REMOVED***
        ***REMOVED*** else ***REMOVED***
            var key = options.decoder(part.slice(0, pos));
            var val = options.decoder(part.slice(pos + 1));

            if (Object.prototype.hasOwnProperty.call(obj, key)) ***REMOVED***
                obj[key] = [].concat(obj[key]).concat(val);
            ***REMOVED*** else ***REMOVED***
                obj[key] = val;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    return obj;
***REMOVED***;

var parseObject = function parseObject(chain, val, options) ***REMOVED***
    if (!chain.length) ***REMOVED***
        return val;
    ***REMOVED***

    var root = chain.shift();

    var obj;
    if (root === '[]') ***REMOVED***
        obj = [];
        obj = obj.concat(parseObject(chain, val, options));
    ***REMOVED*** else ***REMOVED***
        obj = options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;
        var cleanRoot = root[0] === '[' && root[root.length - 1] === ']' ? root.slice(1, root.length - 1) : root;
        var index = parseInt(cleanRoot, 10);
        if (
            !isNaN(index) &&
            root !== cleanRoot &&
            String(index) === cleanRoot &&
            index >= 0 &&
            (options.parseArrays && index <= options.arrayLimit)
        ) ***REMOVED***
            obj = [];
            obj[index] = parseObject(chain, val, options);
        ***REMOVED*** else ***REMOVED***
            obj[cleanRoot] = parseObject(chain, val, options);
        ***REMOVED***
    ***REMOVED***

    return obj;
***REMOVED***;

var parseKeys = function parseKeys(givenKey, val, options) ***REMOVED***
    if (!givenKey) ***REMOVED***
        return;
    ***REMOVED***

    // Transform dot notation to bracket notation
    var key = options.allowDots ? givenKey.replace(/\.([^\.\[]+)/g, '[$1]') : givenKey;

    // The regex chunks

    var parent = /^([^\[\]]*)/;
    var child = /(\[[^\[\]]*\])/g;

    // Get the parent

    var segment = parent.exec(key);

    // Stash the parent if it exists

    var keys = [];
    if (segment[1]) ***REMOVED***
        // If we aren't using plain objects, optionally prefix keys
        // that would overwrite object prototype properties
        if (!options.plainObjects && Object.prototype.hasOwnProperty(segment[1])) ***REMOVED***
            if (!options.allowPrototypes) ***REMOVED***
                return;
            ***REMOVED***
        ***REMOVED***

        keys.push(segment[1]);
    ***REMOVED***

    // Loop through children appending to the array until we hit depth

    var i = 0;
    while ((segment = child.exec(key)) !== null && i < options.depth) ***REMOVED***
        i += 1;
        if (!options.plainObjects && Object.prototype.hasOwnProperty(segment[1].replace(/\[|\]/g, ''))) ***REMOVED***
            if (!options.allowPrototypes) ***REMOVED***
                continue;
            ***REMOVED***
        ***REMOVED***
        keys.push(segment[1]);
    ***REMOVED***

    // If there's a remainder, just add whatever is left

    if (segment) ***REMOVED***
        keys.push('[' + key.slice(segment.index) + ']');
    ***REMOVED***

    return parseObject(keys, val, options);
***REMOVED***;

module.exports = function (str, opts) ***REMOVED***
    var options = opts || ***REMOVED******REMOVED***;

    if (options.decoder !== null && options.decoder !== undefined && typeof options.decoder !== 'function') ***REMOVED***
        throw new TypeError('Decoder has to be a function.');
    ***REMOVED***

    options.delimiter = typeof options.delimiter === 'string' || Utils.isRegExp(options.delimiter) ? options.delimiter : defaults.delimiter;
    options.depth = typeof options.depth === 'number' ? options.depth : defaults.depth;
    options.arrayLimit = typeof options.arrayLimit === 'number' ? options.arrayLimit : defaults.arrayLimit;
    options.parseArrays = options.parseArrays !== false;
    options.decoder = typeof options.decoder === 'function' ? options.decoder : defaults.decoder;
    options.allowDots = typeof options.allowDots === 'boolean' ? options.allowDots : defaults.allowDots;
    options.plainObjects = typeof options.plainObjects === 'boolean' ? options.plainObjects : defaults.plainObjects;
    options.allowPrototypes = typeof options.allowPrototypes === 'boolean' ? options.allowPrototypes : defaults.allowPrototypes;
    options.parameterLimit = typeof options.parameterLimit === 'number' ? options.parameterLimit : defaults.parameterLimit;
    options.strictNullHandling = typeof options.strictNullHandling === 'boolean' ? options.strictNullHandling : defaults.strictNullHandling;

    if (str === '' || str === null || typeof str === 'undefined') ***REMOVED***
        return options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;
    ***REMOVED***

    var tempObj = typeof str === 'string' ? parseValues(str, options) : str;
    var obj = options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;

    // Iterate over the keys and setup the new object

    var keys = Object.keys(tempObj);
    for (var i = 0; i < keys.length; ++i) ***REMOVED***
        var key = keys[i];
        var newObj = parseKeys(key, tempObj[key], options);
        obj = Utils.merge(obj, newObj, options);
    ***REMOVED***

    return Utils.compact(obj);
***REMOVED***;

***REMOVED***,***REMOVED***"./utils":4***REMOVED***],3:[function(require,module,exports)***REMOVED***
'use strict';

var Utils = require('./utils');

var arrayPrefixGenerators = ***REMOVED***
    brackets: function brackets(prefix) ***REMOVED***
        return prefix + '[]';
    ***REMOVED***,
    indices: function indices(prefix, key) ***REMOVED***
        return prefix + '[' + key + ']';
    ***REMOVED***,
    repeat: function repeat(prefix) ***REMOVED***
        return prefix;
    ***REMOVED***
***REMOVED***;

var defaults = ***REMOVED***
    delimiter: '&',
    strictNullHandling: false,
    skipNulls: false,
    encode: true,
    encoder: Utils.encode
***REMOVED***;

var stringify = function stringify(object, prefix, generateArrayPrefix, strictNullHandling, skipNulls, encoder, filter, sort, allowDots) ***REMOVED***
    var obj = object;
    if (typeof filter === 'function') ***REMOVED***
        obj = filter(prefix, obj);
    ***REMOVED*** else if (obj instanceof Date) ***REMOVED***
        obj = obj.toISOString();
    ***REMOVED*** else if (obj === null) ***REMOVED***
        if (strictNullHandling) ***REMOVED***
            return encoder ? encoder(prefix) : prefix;
        ***REMOVED***

        obj = '';
    ***REMOVED***

    if (typeof obj === 'string' || typeof obj === 'number' || typeof obj === 'boolean' || Utils.isBuffer(obj)) ***REMOVED***
        if (encoder) ***REMOVED***
            return [encoder(prefix) + '=' + encoder(obj)];
        ***REMOVED***
        return [prefix + '=' + String(obj)];
    ***REMOVED***

    var values = [];

    if (typeof obj === 'undefined') ***REMOVED***
        return values;
    ***REMOVED***

    var objKeys;
    if (Array.isArray(filter)) ***REMOVED***
        objKeys = filter;
    ***REMOVED*** else ***REMOVED***
        var keys = Object.keys(obj);
        objKeys = sort ? keys.sort(sort) : keys;
    ***REMOVED***

    for (var i = 0; i < objKeys.length; ++i) ***REMOVED***
        var key = objKeys[i];

        if (skipNulls && obj[key] === null) ***REMOVED***
            continue;
        ***REMOVED***

        if (Array.isArray(obj)) ***REMOVED***
            values = values.concat(stringify(obj[key], generateArrayPrefix(prefix, key), generateArrayPrefix, strictNullHandling, skipNulls, encoder, filter, sort, allowDots));
        ***REMOVED*** else ***REMOVED***
            values = values.concat(stringify(obj[key], prefix + (allowDots ? '.' + key : '[' + key + ']'), generateArrayPrefix, strictNullHandling, skipNulls, encoder, filter, sort, allowDots));
        ***REMOVED***
    ***REMOVED***

    return values;
***REMOVED***;

module.exports = function (object, opts) ***REMOVED***
    var obj = object;
    var options = opts || ***REMOVED******REMOVED***;
    var delimiter = typeof options.delimiter === 'undefined' ? defaults.delimiter : options.delimiter;
    var strictNullHandling = typeof options.strictNullHandling === 'boolean' ? options.strictNullHandling : defaults.strictNullHandling;
    var skipNulls = typeof options.skipNulls === 'boolean' ? options.skipNulls : defaults.skipNulls;
    var encode = typeof options.encode === 'boolean' ? options.encode : defaults.encode;
    var encoder = encode ? (typeof options.encoder === 'function' ? options.encoder : defaults.encoder) : null;
    var sort = typeof options.sort === 'function' ? options.sort : null;
    var allowDots = typeof options.allowDots === 'undefined' ? false : options.allowDots;
    var objKeys;
    var filter;

    if (options.encoder !== null && options.encoder !== undefined && typeof options.encoder !== 'function') ***REMOVED***
        throw new TypeError('Encoder has to be a function.');
    ***REMOVED***

    if (typeof options.filter === 'function') ***REMOVED***
        filter = options.filter;
        obj = filter('', obj);
    ***REMOVED*** else if (Array.isArray(options.filter)) ***REMOVED***
        objKeys = filter = options.filter;
    ***REMOVED***

    var keys = [];

    if (typeof obj !== 'object' || obj === null) ***REMOVED***
        return '';
    ***REMOVED***

    var arrayFormat;
    if (options.arrayFormat in arrayPrefixGenerators) ***REMOVED***
        arrayFormat = options.arrayFormat;
    ***REMOVED*** else if ('indices' in options) ***REMOVED***
        arrayFormat = options.indices ? 'indices' : 'repeat';
    ***REMOVED*** else ***REMOVED***
        arrayFormat = 'indices';
    ***REMOVED***

    var generateArrayPrefix = arrayPrefixGenerators[arrayFormat];

    if (!objKeys) ***REMOVED***
        objKeys = Object.keys(obj);
    ***REMOVED***

    if (sort) ***REMOVED***
        objKeys.sort(sort);
    ***REMOVED***

    for (var i = 0; i < objKeys.length; ++i) ***REMOVED***
        var key = objKeys[i];

        if (skipNulls && obj[key] === null) ***REMOVED***
            continue;
        ***REMOVED***

        keys = keys.concat(stringify(obj[key], key, generateArrayPrefix, strictNullHandling, skipNulls, encoder, filter, sort, allowDots));
    ***REMOVED***

    return keys.join(delimiter);
***REMOVED***;

***REMOVED***,***REMOVED***"./utils":4***REMOVED***],4:[function(require,module,exports)***REMOVED***
'use strict';

var hexTable = (function () ***REMOVED***
    var array = new Array(256);
    for (var i = 0; i < 256; ++i) ***REMOVED***
        array[i] = '%' + ((i < 16 ? '0' : '') + i.toString(16)).toUpperCase();
    ***REMOVED***

    return array;
***REMOVED***());

exports.arrayToObject = function (source, options) ***REMOVED***
    var obj = options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;
    for (var i = 0; i < source.length; ++i) ***REMOVED***
        if (typeof source[i] !== 'undefined') ***REMOVED***
            obj[i] = source[i];
        ***REMOVED***
    ***REMOVED***

    return obj;
***REMOVED***;

exports.merge = function (target, source, options) ***REMOVED***
    if (!source) ***REMOVED***
        return target;
    ***REMOVED***

    if (typeof source !== 'object') ***REMOVED***
        if (Array.isArray(target)) ***REMOVED***
            target.push(source);
        ***REMOVED*** else if (typeof target === 'object') ***REMOVED***
            target[source] = true;
        ***REMOVED*** else ***REMOVED***
            return [target, source];
        ***REMOVED***

        return target;
    ***REMOVED***

    if (typeof target !== 'object') ***REMOVED***
        return [target].concat(source);
    ***REMOVED***

    var mergeTarget = target;
    if (Array.isArray(target) && !Array.isArray(source)) ***REMOVED***
        mergeTarget = exports.arrayToObject(target, options);
    ***REMOVED***

    return Object.keys(source).reduce(function (acc, key) ***REMOVED***
        var value = source[key];

        if (Object.prototype.hasOwnProperty.call(acc, key)) ***REMOVED***
            acc[key] = exports.merge(acc[key], value, options);
        ***REMOVED*** else ***REMOVED***
            acc[key] = value;
        ***REMOVED***
        return acc;
    ***REMOVED***, mergeTarget);
***REMOVED***;

exports.decode = function (str) ***REMOVED***
    try ***REMOVED***
        return decodeURIComponent(str.replace(/\+/g, ' '));
    ***REMOVED*** catch (e) ***REMOVED***
        return str;
    ***REMOVED***
***REMOVED***;

exports.encode = function (str) ***REMOVED***
    // This code was originally written by Brian White (mscdex) for the io.js core querystring library.
    // It has been adapted here for stricter adherence to RFC 3986
    if (str.length === 0) ***REMOVED***
        return str;
    ***REMOVED***

    var string = typeof str === 'string' ? str : String(str);

    var out = '';
    for (var i = 0; i < string.length; ++i) ***REMOVED***
        var c = string.charCodeAt(i);

        if (
            c === 0x2D || // -
            c === 0x2E || // .
            c === 0x5F || // _
            c === 0x7E || // ~
            (c >= 0x30 && c <= 0x39) || // 0-9
            (c >= 0x41 && c <= 0x5A) || // a-z
            (c >= 0x61 && c <= 0x7A) // A-Z
        ) ***REMOVED***
            out += string.charAt(i);
            continue;
        ***REMOVED***

        if (c < 0x80) ***REMOVED***
            out = out + hexTable[c];
            continue;
        ***REMOVED***

        if (c < 0x800) ***REMOVED***
            out = out + (hexTable[0xC0 | (c >> 6)] + hexTable[0x80 | (c & 0x3F)]);
            continue;
        ***REMOVED***

        if (c < 0xD800 || c >= 0xE000) ***REMOVED***
            out = out + (hexTable[0xE0 | (c >> 12)] + hexTable[0x80 | ((c >> 6) & 0x3F)] + hexTable[0x80 | (c & 0x3F)]);
            continue;
        ***REMOVED***

        i += 1;
        c = 0x10000 + (((c & 0x3FF) << 10) | (string.charCodeAt(i) & 0x3FF));
        out += hexTable[0xF0 | (c >> 18)] + hexTable[0x80 | ((c >> 12) & 0x3F)] + hexTable[0x80 | ((c >> 6) & 0x3F)] + hexTable[0x80 | (c & 0x3F)];
    ***REMOVED***

    return out;
***REMOVED***;

exports.compact = function (obj, references) ***REMOVED***
    if (typeof obj !== 'object' || obj === null) ***REMOVED***
        return obj;
    ***REMOVED***

    var refs = references || [];
    var lookup = refs.indexOf(obj);
    if (lookup !== -1) ***REMOVED***
        return refs[lookup];
    ***REMOVED***

    refs.push(obj);

    if (Array.isArray(obj)) ***REMOVED***
        var compacted = [];

        for (var i = 0; i < obj.length; ++i) ***REMOVED***
            if (obj[i] && typeof obj[i] === 'object') ***REMOVED***
                compacted.push(exports.compact(obj[i], refs));
            ***REMOVED*** else if (typeof obj[i] !== 'undefined') ***REMOVED***
                compacted.push(obj[i]);
            ***REMOVED***
        ***REMOVED***

        return compacted;
    ***REMOVED***

    var keys = Object.keys(obj);
    for (var j = 0; j < keys.length; ++j) ***REMOVED***
        var key = keys[j];
        obj[key] = exports.compact(obj[key], refs);
    ***REMOVED***

    return obj;
***REMOVED***;

exports.isRegExp = function (obj) ***REMOVED***
    return Object.prototype.toString.call(obj) === '[object RegExp]';
***REMOVED***;

exports.isBuffer = function (obj) ***REMOVED***
    if (obj === null || typeof obj === 'undefined') ***REMOVED***
        return false;
    ***REMOVED***

    return !!(obj.constructor && obj.constructor.isBuffer && obj.constructor.isBuffer(obj));
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***]***REMOVED***,***REMOVED******REMOVED***,[1])(1)
***REMOVED***);
