// winston/lib/winston/transports/console.js
/*
 * console.js: Transport for outputting to the console
 *
 * (C) 2010 Charlie Robbins
 * MIT LICENCE
 *
 */

var events = require('events'),
    os = require('os'),
    util = require('util'),
    common = require('../common'),
    Transport = require('./transport').Transport;

//
// ### function Console (options)
// #### @options ***REMOVED***Object***REMOVED*** Options for this instance.
// Constructor function for the Console transport object responsible
// for persisting log messages and metadata to a terminal or TTY.
//
var Console = exports.Console = function (options) ***REMOVED***
  Transport.call(this, options);
  options = options || ***REMOVED******REMOVED***;

  this.json         = options.json        || false;
  this.colorize     = options.colorize    || false;
  this.prettyPrint  = options.prettyPrint || false;
  this.timestamp    = typeof options.timestamp !== 'undefined' ? options.timestamp : false;
  this.showLevel    = options.showLevel === undefined ? true : options.showLevel;
  this.label        = options.label       || null;
  this.logstash     = options.logstash    || false;
  this.depth        = options.depth       || null;
  this.align        = options.align       || false;
  this.stderrLevels = setStderrLevels(options.stderrLevels, options.debugStdout);
  this.eol          = options.eol   || os.EOL;

  if (this.json) ***REMOVED***
    this.stringify = options.stringify || function (obj) ***REMOVED***
      return JSON.stringify(obj, null, 2);
    ***REMOVED***;
  ***REMOVED***

  //
  // Convert stderrLevels into an Object for faster key-lookup times than an Array.
  //
  // For backwards compatibility, stderrLevels defaults to ['error', 'debug']
  // or ['error'] depending on whether options.debugStdout is true.
  //
  function setStderrLevels (levels, debugStdout) ***REMOVED***
    var defaultMsg = 'Cannot have non-string elements in stderrLevels Array';
    if (debugStdout) ***REMOVED***
      if (levels) ***REMOVED***
        //
        // Don't allow setting both debugStdout and stderrLevels together,
        // since this could cause behaviour a programmer might not expect.
        //
        throw new Error('Cannot set debugStdout and stderrLevels together');
      ***REMOVED***

      return common.stringArrayToSet(['error'], defaultMsg);
    ***REMOVED***

    if (!levels) ***REMOVED***
      return common.stringArrayToSet(['error', 'debug'], defaultMsg);
    ***REMOVED*** else if (!(Array.isArray(levels))) ***REMOVED***
      throw new Error('Cannot set stderrLevels to type other than Array');
    ***REMOVED***

    return common.stringArrayToSet(levels, defaultMsg);
  ***REMOVED***;
***REMOVED***;

//
// Inherit from `winston.Transport`.
//
util.inherits(Console, Transport);

//
// Expose the name of this Transport on the prototype
//
Console.prototype.name = 'console';

//
// ### function log (level, msg, [meta], callback)
// #### @level ***REMOVED***string***REMOVED*** Level at which to log the message.
// #### @msg ***REMOVED***string***REMOVED*** Message to log
// #### @meta ***REMOVED***Object***REMOVED*** **Optional** Additional metadata to attach
// #### @callback ***REMOVED***function***REMOVED*** Continuation to respond to when complete.
// Core logging method exposed to Winston. Metadata is optional.
//
Console.prototype.log = function (level, msg, meta, callback) ***REMOVED***
  if (this.silent) ***REMOVED***
    return callback(null, true);
  ***REMOVED***

  var self = this,
      output;

  output = common.log(***REMOVED***
    colorize:    this.colorize,
    json:        this.json,
    level:       level,
    message:     msg,
    meta:        meta,
    stringify:   this.stringify,
    timestamp:   this.timestamp,
    showLevel:   this.showLevel,
    prettyPrint: this.prettyPrint,
    raw:         this.raw,
    label:       this.label,
    logstash:    this.logstash,
    depth:       this.depth,
    formatter:   this.formatter,
    align:       this.align,
    humanReadableUnhandledException: this.humanReadableUnhandledException
  ***REMOVED***);

  if (this.stderrLevels[level]) ***REMOVED***
    process.stderr.write(output + this.eol);
  ***REMOVED*** else ***REMOVED***
    process.stdout.write(output + this.eol);
  ***REMOVED***

  //
  // Emit the `logged` event immediately because the event loop
  // will not exit until `process.stdout` has drained anyway.
  //
  self.emit('logged');
  callback(null, true);
***REMOVED***;
