// winston/lib/winston/transports/memory.js
var events = require('events'),
    util = require('util'),
    common = require('../common'),
    Transport = require('./transport').Transport;

//
// ### function Memory (options)
// #### @options ***REMOVED***Object***REMOVED*** Options for this instance.
// Constructor function for the Memory transport object responsible
// for persisting log messages and metadata to a memory array of messages.
//
var Memory = exports.Memory = function (options) ***REMOVED***
  Transport.call(this, options);
  options = options || ***REMOVED******REMOVED***;

  this.errorOutput = [];
  this.writeOutput = [];

  this.json        = options.json        || false;
  this.colorize    = options.colorize    || false;
  this.prettyPrint = options.prettyPrint || false;
  this.timestamp   = typeof options.timestamp !== 'undefined' ? options.timestamp : false;
  this.showLevel   = options.showLevel === undefined ? true : options.showLevel;
  this.label       = options.label       || null;
  this.depth       = options.depth       || null;

  if (this.json) ***REMOVED***
    this.stringify = options.stringify || function (obj) ***REMOVED***
      return JSON.stringify(obj, null, 2);
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;

//
// Inherit from `winston.Transport`.
//
util.inherits(Memory, Transport);

//
// Expose the name of this Transport on the prototype
//
Memory.prototype.name = 'memory';

//
// ### function log (level, msg, [meta], callback)
// #### @level ***REMOVED***string***REMOVED*** Level at which to log the message.
// #### @msg ***REMOVED***string***REMOVED*** Message to log
// #### @meta ***REMOVED***Object***REMOVED*** **Optional** Additional metadata to attach
// #### @callback ***REMOVED***function***REMOVED*** Continuation to respond to when complete.
// Core logging method exposed to Winston. Metadata is optional.
//
Memory.prototype.log = function (level, msg, meta, callback) ***REMOVED***
  if (this.silent) ***REMOVED***
    return callback(null, true);
  ***REMOVED***

  var self = this,
      output;

  output = common.log(***REMOVED***
    colorize:    this.colorize,
    json:        this.json,
    level:       level,
    message:     msg,
    meta:        meta,
    stringify:   this.stringify,
    timestamp:   this.timestamp,
    prettyPrint: this.prettyPrint,
    raw:         this.raw,
    label:       this.label,
    depth:       this.depth,
    formatter:   this.formatter,
    humanReadableUnhandledException: this.humanReadableUnhandledException
  ***REMOVED***);

  if (level === 'error' || level === 'debug') ***REMOVED***
    this.errorOutput.push(output);
  ***REMOVED*** else ***REMOVED***
    this.writeOutput.push(output);
  ***REMOVED***

  self.emit('logged');
  callback(null, true);
***REMOVED***;

Memory.prototype.clearLogs = function () ***REMOVED***
  this.errorOutput = [];
  this.writeOutput = [];
***REMOVED***;
