// cookies/lib/cookies.js
var deprecate = require('depd')('cookies')
var Keygrip = require('keygrip')
var http = require('http')
var cache = ***REMOVED******REMOVED***

/**
 * RegExp to match field-content in RFC 7230 sec 3.2
 *
 * field-content = field-vchar [ 1*( SP / HTAB ) field-vchar ]
 * field-vchar   = VCHAR / obs-text
 * obs-text      = %x80-FF
 */

var fieldContentRegExp = /^[\u0009\u0020-\u007e\u0080-\u00ff]+$/;

function Cookies(request, response, options) ***REMOVED***
  if (!(this instanceof Cookies)) return new Cookies(request, response, options)

  this.secure = undefined
  this.request = request
  this.response = response

  if (options) ***REMOVED***
    if (Array.isArray(options)) ***REMOVED***
      // array of key strings
      deprecate('"keys" argument; provide using options ***REMOVED***"key": [...]***REMOVED***')
      this.keys = new Keygrip(options)
    ***REMOVED*** else if (options.constructor && options.constructor.name === 'Keygrip') ***REMOVED***
      // any keygrip constructor to allow different versions
      deprecate('"keys" argument; provide using options ***REMOVED***"key": keygrip***REMOVED***')
      this.keys = options
    ***REMOVED*** else ***REMOVED***
      this.keys = Array.isArray(options.keys) ? new Keygrip(options.keys) : options.keys
      this.secure = options.secure
    ***REMOVED***
  ***REMOVED***
***REMOVED***

Cookies.prototype.get = function(name, opts) ***REMOVED***
  var sigName = name + ".sig"
    , header, match, value, remote, data, index
    , signed = opts && opts.signed !== undefined ? opts.signed : !!this.keys

  header = this.request.headers["cookie"]
  if (!header) return

  match = header.match(getPattern(name))
  if (!match) return

  value = match[1]
  if (!opts || !signed) return value

  remote = this.get(sigName)
  if (!remote) return

  data = name + "=" + value
  if (!this.keys) throw new Error('.keys required for signed cookies');
  index = this.keys.index(data, remote)

  if (index < 0) ***REMOVED***
    this.set(sigName, null, ***REMOVED***path: "/", signed: false ***REMOVED***)
  ***REMOVED*** else ***REMOVED***
    index && this.set(sigName, this.keys.sign(data), ***REMOVED*** signed: false ***REMOVED***)
    return value
  ***REMOVED***
***REMOVED***;

Cookies.prototype.set = function(name, value, opts) ***REMOVED***
  var res = this.response
    , req = this.request
    , headers = res.getHeader("Set-Cookie") || []
    , secure = this.secure !== undefined ? !!this.secure : req.protocol === 'https' || req.connection.encrypted
    , cookie = new Cookie(name, value, opts)
    , signed = opts && opts.signed !== undefined ? opts.signed : !!this.keys

  if (typeof headers == "string") headers = [headers]

  if (!secure && opts && opts.secure) ***REMOVED***
    throw new Error('Cannot send secure cookie over unencrypted connection')
  ***REMOVED***

  cookie.secure = secure
  if (opts && "secure" in opts) cookie.secure = opts.secure

  if (opts && "secureProxy" in opts) ***REMOVED***
    deprecate('"secureProxy" option; use "secure" option, provide "secure" to constructor if needed')
    cookie.secure = opts.secureProxy
  ***REMOVED***

  headers = pushCookie(headers, cookie)

  if (opts && signed) ***REMOVED***
    if (!this.keys) throw new Error('.keys required for signed cookies');
    cookie.value = this.keys.sign(cookie.toString())
    cookie.name += ".sig"
    headers = pushCookie(headers, cookie)
  ***REMOVED***

  var setHeader = res.set ? http.OutgoingMessage.prototype.setHeader : res.setHeader
  setHeader.call(res, 'Set-Cookie', headers)
  return this
***REMOVED***;

function Cookie(name, value, attrs) ***REMOVED***
  if (!fieldContentRegExp.test(name)) ***REMOVED***
    throw new TypeError('argument name is invalid');
  ***REMOVED***

  if (value && !fieldContentRegExp.test(value)) ***REMOVED***
    throw new TypeError('argument value is invalid');
  ***REMOVED***

  value || (this.expires = new Date(0))

  this.name = name
  this.value = value || ""

  for (var name in attrs) ***REMOVED***
    this[name] = attrs[name]
  ***REMOVED***

  if (this.path && !fieldContentRegExp.test(this.path)) ***REMOVED***
    throw new TypeError('option path is invalid');
  ***REMOVED***

  if (this.domain && !fieldContentRegExp.test(this.domain)) ***REMOVED***
    throw new TypeError('option domain is invalid');
  ***REMOVED***
***REMOVED***

Cookie.prototype.path = "/";
Cookie.prototype.expires = undefined;
Cookie.prototype.domain = undefined;
Cookie.prototype.httpOnly = true;
Cookie.prototype.secure = false;
Cookie.prototype.overwrite = false;

Cookie.prototype.toString = function() ***REMOVED***
  return this.name + "=" + this.value
***REMOVED***;

Cookie.prototype.toHeader = function() ***REMOVED***
  var header = this.toString()

  if (this.maxAge) this.expires = new Date(Date.now() + this.maxAge);

  if (this.path     ) header += "; path=" + this.path
  if (this.expires  ) header += "; expires=" + this.expires.toUTCString()
  if (this.domain   ) header += "; domain=" + this.domain
  if (this.secure   ) header += "; secure"
  if (this.httpOnly ) header += "; httponly"

  return header
***REMOVED***;

// back-compat so maxage mirrors maxAge
Object.defineProperty(Cookie.prototype, 'maxage', ***REMOVED***
  configurable: true,
  enumerable: true,
  get: function () ***REMOVED*** return this.maxAge ***REMOVED***,
  set: function (val) ***REMOVED*** return this.maxAge = val ***REMOVED***
***REMOVED***);

function getPattern(name) ***REMOVED***
  if (cache[name]) return cache[name]

  return cache[name] = new RegExp(
    "(?:^|;) *" +
    name.replace(/[-[\]***REMOVED******REMOVED***()*+?.,\\^$|#\s]/g, "\\$&") +
    "=([^;]*)"
  )
***REMOVED***

function pushCookie(cookies, cookie) ***REMOVED***
  if (cookie.overwrite) ***REMOVED***
    cookies = cookies.filter(function(c) ***REMOVED*** return c.indexOf(cookie.name+'=') !== 0 ***REMOVED***)
  ***REMOVED***
  cookies.push(cookie.toHeader())
  return cookies
***REMOVED***

Cookies.connect = Cookies.express = function(keys) ***REMOVED***
  return function(req, res, next) ***REMOVED***
    req.cookies = res.cookies = new Cookies(req, res, ***REMOVED***
      keys: keys
    ***REMOVED***)

    next()
  ***REMOVED***
***REMOVED***

Cookies.Cookie = Cookie

module.exports = Cookies
