// moment/src/lib/units/hour.js
import ***REMOVED*** makeGetSet ***REMOVED*** from '../moment/get-set';
import ***REMOVED*** addFormatToken ***REMOVED*** from '../format/format';
import ***REMOVED*** addUnitAlias ***REMOVED*** from './aliases';
import ***REMOVED*** addUnitPriority ***REMOVED*** from './priorities';
import ***REMOVED*** addRegexToken, match1to2, match2, match3to4, match5to6 ***REMOVED*** from '../parse/regex';
import ***REMOVED*** addParseToken ***REMOVED*** from '../parse/token';
import ***REMOVED*** HOUR, MINUTE, SECOND ***REMOVED*** from './constants';
import toInt from '../utils/to-int';
import zeroFill from '../utils/zero-fill';
import getParsingFlags from '../create/parsing-flags';

// FORMATTING

function hFormat() ***REMOVED***
    return this.hours() % 12 || 12;
***REMOVED***

function kFormat() ***REMOVED***
    return this.hours() || 24;
***REMOVED***

addFormatToken('H', ['HH', 2], 0, 'hour');
addFormatToken('h', ['hh', 2], 0, hFormat);
addFormatToken('k', ['kk', 2], 0, kFormat);

addFormatToken('hmm', 0, 0, function () ***REMOVED***
    return '' + hFormat.apply(this) + zeroFill(this.minutes(), 2);
***REMOVED***);

addFormatToken('hmmss', 0, 0, function () ***REMOVED***
    return '' + hFormat.apply(this) + zeroFill(this.minutes(), 2) +
        zeroFill(this.seconds(), 2);
***REMOVED***);

addFormatToken('Hmm', 0, 0, function () ***REMOVED***
    return '' + this.hours() + zeroFill(this.minutes(), 2);
***REMOVED***);

addFormatToken('Hmmss', 0, 0, function () ***REMOVED***
    return '' + this.hours() + zeroFill(this.minutes(), 2) +
        zeroFill(this.seconds(), 2);
***REMOVED***);

function meridiem (token, lowercase) ***REMOVED***
    addFormatToken(token, 0, 0, function () ***REMOVED***
        return this.localeData().meridiem(this.hours(), this.minutes(), lowercase);
    ***REMOVED***);
***REMOVED***

meridiem('a', true);
meridiem('A', false);

// ALIASES

addUnitAlias('hour', 'h');

// PRIORITY
addUnitPriority('hour', 13);

// PARSING

function matchMeridiem (isStrict, locale) ***REMOVED***
    return locale._meridiemParse;
***REMOVED***

addRegexToken('a',  matchMeridiem);
addRegexToken('A',  matchMeridiem);
addRegexToken('H',  match1to2);
addRegexToken('h',  match1to2);
addRegexToken('HH', match1to2, match2);
addRegexToken('hh', match1to2, match2);

addRegexToken('hmm', match3to4);
addRegexToken('hmmss', match5to6);
addRegexToken('Hmm', match3to4);
addRegexToken('Hmmss', match5to6);

addParseToken(['H', 'HH'], HOUR);
addParseToken(['a', 'A'], function (input, array, config) ***REMOVED***
    config._isPm = config._locale.isPM(input);
    config._meridiem = input;
***REMOVED***);
addParseToken(['h', 'hh'], function (input, array, config) ***REMOVED***
    array[HOUR] = toInt(input);
    getParsingFlags(config).bigHour = true;
***REMOVED***);
addParseToken('hmm', function (input, array, config) ***REMOVED***
    var pos = input.length - 2;
    array[HOUR] = toInt(input.substr(0, pos));
    array[MINUTE] = toInt(input.substr(pos));
    getParsingFlags(config).bigHour = true;
***REMOVED***);
addParseToken('hmmss', function (input, array, config) ***REMOVED***
    var pos1 = input.length - 4;
    var pos2 = input.length - 2;
    array[HOUR] = toInt(input.substr(0, pos1));
    array[MINUTE] = toInt(input.substr(pos1, 2));
    array[SECOND] = toInt(input.substr(pos2));
    getParsingFlags(config).bigHour = true;
***REMOVED***);
addParseToken('Hmm', function (input, array, config) ***REMOVED***
    var pos = input.length - 2;
    array[HOUR] = toInt(input.substr(0, pos));
    array[MINUTE] = toInt(input.substr(pos));
***REMOVED***);
addParseToken('Hmmss', function (input, array, config) ***REMOVED***
    var pos1 = input.length - 4;
    var pos2 = input.length - 2;
    array[HOUR] = toInt(input.substr(0, pos1));
    array[MINUTE] = toInt(input.substr(pos1, 2));
    array[SECOND] = toInt(input.substr(pos2));
***REMOVED***);

// LOCALES

export function localeIsPM (input) ***REMOVED***
    // IE8 Quirks Mode & IE7 Standards Mode do not allow accessing strings like arrays
    // Using charAt should be more compatible.
    return ((input + '').toLowerCase().charAt(0) === 'p');
***REMOVED***

export var defaultLocaleMeridiemParse = /[ap]\.?m?\.?/i;
export function localeMeridiem (hours, minutes, isLower) ***REMOVED***
    if (hours > 11) ***REMOVED***
        return isLower ? 'pm' : 'PM';
    ***REMOVED*** else ***REMOVED***
        return isLower ? 'am' : 'AM';
    ***REMOVED***
***REMOVED***


// MOMENTS

// Setting the hour should keep the time, because the user explicitly
// specified which hour he wants. So trying to maintain the same hour (in
// a new timezone) makes sense. Adding/subtracting hours does not follow
// this rule.
export var getSetHour = makeGetSet('Hours', true);
