// moment/src/lib/units/week.js
import ***REMOVED*** addFormatToken ***REMOVED*** from '../format/format';
import ***REMOVED*** addUnitAlias ***REMOVED*** from './aliases';
import ***REMOVED*** addUnitPriority ***REMOVED*** from './priorities';
import ***REMOVED*** addRegexToken, match1to2, match2 ***REMOVED*** from '../parse/regex';
import ***REMOVED*** addWeekParseToken ***REMOVED*** from '../parse/token';
import toInt from '../utils/to-int';
import ***REMOVED*** createLocal ***REMOVED*** from '../create/local';
import ***REMOVED*** weekOfYear ***REMOVED*** from './week-calendar-utils';

// FORMATTING

addFormatToken('w', ['ww', 2], 'wo', 'week');
addFormatToken('W', ['WW', 2], 'Wo', 'isoWeek');

// ALIASES

addUnitAlias('week', 'w');
addUnitAlias('isoWeek', 'W');

// PRIORITIES

addUnitPriority('week', 5);
addUnitPriority('isoWeek', 5);

// PARSING

addRegexToken('w',  match1to2);
addRegexToken('ww', match1to2, match2);
addRegexToken('W',  match1to2);
addRegexToken('WW', match1to2, match2);

addWeekParseToken(['w', 'ww', 'W', 'WW'], function (input, week, config, token) ***REMOVED***
    week[token.substr(0, 1)] = toInt(input);
***REMOVED***);

// HELPERS

// LOCALES

export function localeWeek (mom) ***REMOVED***
    return weekOfYear(mom, this._week.dow, this._week.doy).week;
***REMOVED***

export var defaultLocaleWeek = ***REMOVED***
    dow : 0, // Sunday is the first day of the week.
    doy : 6  // The week that contains Jan 1st is the first week of the year.
***REMOVED***;

export function localeFirstDayOfWeek () ***REMOVED***
    return this._week.dow;
***REMOVED***

export function localeFirstDayOfYear () ***REMOVED***
    return this._week.doy;
***REMOVED***

// MOMENTS

export function getSetWeek (input) ***REMOVED***
    var week = this.localeData().week(this);
    return input == null ? week : this.add((input - week) * 7, 'd');
***REMOVED***

export function getSetISOWeek (input) ***REMOVED***
    var week = weekOfYear(this, 1, 4).week;
    return input == null ? week : this.add((input - week) * 7, 'd');
***REMOVED***
