// moment/src/lib/units/week-year.js
import ***REMOVED*** addFormatToken ***REMOVED*** from '../format/format';
import ***REMOVED*** addUnitAlias ***REMOVED*** from './aliases';
import ***REMOVED*** addUnitPriority ***REMOVED*** from './priorities';
import ***REMOVED*** addRegexToken, match1to2, match1to4, match1to6, match2, match4, match6, matchSigned ***REMOVED*** from '../parse/regex';
import ***REMOVED*** addWeekParseToken ***REMOVED*** from '../parse/token';
import ***REMOVED*** weekOfYear, weeksInYear, dayOfYearFromWeeks ***REMOVED*** from './week-calendar-utils';
import toInt from '../utils/to-int';
import ***REMOVED*** hooks ***REMOVED*** from '../utils/hooks';
import ***REMOVED*** createLocal ***REMOVED*** from '../create/local';
import ***REMOVED*** createUTCDate ***REMOVED*** from '../create/date-from-array';

// FORMATTING

addFormatToken(0, ['gg', 2], 0, function () ***REMOVED***
    return this.weekYear() % 100;
***REMOVED***);

addFormatToken(0, ['GG', 2], 0, function () ***REMOVED***
    return this.isoWeekYear() % 100;
***REMOVED***);

function addWeekYearFormatToken (token, getter) ***REMOVED***
    addFormatToken(0, [token, token.length], 0, getter);
***REMOVED***

addWeekYearFormatToken('gggg',     'weekYear');
addWeekYearFormatToken('ggggg',    'weekYear');
addWeekYearFormatToken('GGGG',  'isoWeekYear');
addWeekYearFormatToken('GGGGG', 'isoWeekYear');

// ALIASES

addUnitAlias('weekYear', 'gg');
addUnitAlias('isoWeekYear', 'GG');

// PRIORITY

addUnitPriority('weekYear', 1);
addUnitPriority('isoWeekYear', 1);


// PARSING

addRegexToken('G',      matchSigned);
addRegexToken('g',      matchSigned);
addRegexToken('GG',     match1to2, match2);
addRegexToken('gg',     match1to2, match2);
addRegexToken('GGGG',   match1to4, match4);
addRegexToken('gggg',   match1to4, match4);
addRegexToken('GGGGG',  match1to6, match6);
addRegexToken('ggggg',  match1to6, match6);

addWeekParseToken(['gggg', 'ggggg', 'GGGG', 'GGGGG'], function (input, week, config, token) ***REMOVED***
    week[token.substr(0, 2)] = toInt(input);
***REMOVED***);

addWeekParseToken(['gg', 'GG'], function (input, week, config, token) ***REMOVED***
    week[token] = hooks.parseTwoDigitYear(input);
***REMOVED***);

// MOMENTS

export function getSetWeekYear (input) ***REMOVED***
    return getSetWeekYearHelper.call(this,
            input,
            this.week(),
            this.weekday(),
            this.localeData()._week.dow,
            this.localeData()._week.doy);
***REMOVED***

export function getSetISOWeekYear (input) ***REMOVED***
    return getSetWeekYearHelper.call(this,
            input, this.isoWeek(), this.isoWeekday(), 1, 4);
***REMOVED***

export function getISOWeeksInYear () ***REMOVED***
    return weeksInYear(this.year(), 1, 4);
***REMOVED***

export function getWeeksInYear () ***REMOVED***
    var weekInfo = this.localeData()._week;
    return weeksInYear(this.year(), weekInfo.dow, weekInfo.doy);
***REMOVED***

function getSetWeekYearHelper(input, week, weekday, dow, doy) ***REMOVED***
    var weeksTarget;
    if (input == null) ***REMOVED***
        return weekOfYear(this, dow, doy).year;
    ***REMOVED*** else ***REMOVED***
        weeksTarget = weeksInYear(input, dow, doy);
        if (week > weeksTarget) ***REMOVED***
            week = weeksTarget;
        ***REMOVED***
        return setWeekAll.call(this, input, week, weekday, dow, doy);
    ***REMOVED***
***REMOVED***

function setWeekAll(weekYear, week, weekday, dow, doy) ***REMOVED***
    var dayOfYearData = dayOfYearFromWeeks(weekYear, week, weekday, dow, doy),
        date = createUTCDate(dayOfYearData.year, 0, dayOfYearData.dayOfYear);

    this.year(date.getUTCFullYear());
    this.month(date.getUTCMonth());
    this.date(date.getUTCDate());
    return this;
***REMOVED***
