// moment/src/lib/units/week-calendar-utils.js
import ***REMOVED*** daysInYear ***REMOVED*** from './year';
import ***REMOVED*** createLocal ***REMOVED*** from '../create/local';
import ***REMOVED*** createUTCDate ***REMOVED*** from '../create/date-from-array';

// start-of-first-week - start-of-year
function firstWeekOffset(year, dow, doy) ***REMOVED***
    var // first-week day -- which january is always in the first week (4 for iso, 1 for other)
        fwd = 7 + dow - doy,
        // first-week day local weekday -- which local weekday is fwd
        fwdlw = (7 + createUTCDate(year, 0, fwd).getUTCDay() - dow) % 7;

    return -fwdlw + fwd - 1;
***REMOVED***

//http://en.wikipedia.org/wiki/ISO_week_date#Calculating_a_date_given_the_year.2C_week_number_and_weekday
export function dayOfYearFromWeeks(year, week, weekday, dow, doy) ***REMOVED***
    var localWeekday = (7 + weekday - dow) % 7,
        weekOffset = firstWeekOffset(year, dow, doy),
        dayOfYear = 1 + 7 * (week - 1) + localWeekday + weekOffset,
        resYear, resDayOfYear;

    if (dayOfYear <= 0) ***REMOVED***
        resYear = year - 1;
        resDayOfYear = daysInYear(resYear) + dayOfYear;
    ***REMOVED*** else if (dayOfYear > daysInYear(year)) ***REMOVED***
        resYear = year + 1;
        resDayOfYear = dayOfYear - daysInYear(year);
    ***REMOVED*** else ***REMOVED***
        resYear = year;
        resDayOfYear = dayOfYear;
    ***REMOVED***

    return ***REMOVED***
        year: resYear,
        dayOfYear: resDayOfYear
    ***REMOVED***;
***REMOVED***

export function weekOfYear(mom, dow, doy) ***REMOVED***
    var weekOffset = firstWeekOffset(mom.year(), dow, doy),
        week = Math.floor((mom.dayOfYear() - weekOffset - 1) / 7) + 1,
        resWeek, resYear;

    if (week < 1) ***REMOVED***
        resYear = mom.year() - 1;
        resWeek = week + weeksInYear(resYear, dow, doy);
    ***REMOVED*** else if (week > weeksInYear(mom.year(), dow, doy)) ***REMOVED***
        resWeek = week - weeksInYear(mom.year(), dow, doy);
        resYear = mom.year() + 1;
    ***REMOVED*** else ***REMOVED***
        resYear = mom.year();
        resWeek = week;
    ***REMOVED***

    return ***REMOVED***
        week: resWeek,
        year: resYear
    ***REMOVED***;
***REMOVED***

export function weeksInYear(year, dow, doy) ***REMOVED***
    var weekOffset = firstWeekOffset(year, dow, doy),
        weekOffsetNext = firstWeekOffset(year + 1, dow, doy);
    return (daysInYear(year) - weekOffset + weekOffsetNext) / 7;
***REMOVED***
