// moment/src/lib/duration/as.js
import ***REMOVED*** daysToMonths, monthsToDays ***REMOVED*** from './bubble';
import ***REMOVED*** normalizeUnits ***REMOVED*** from '../units/aliases';
import toInt from '../utils/to-int';

export function as (units) ***REMOVED***
    var days;
    var months;
    var milliseconds = this._milliseconds;

    units = normalizeUnits(units);

    if (units === 'month' || units === 'year') ***REMOVED***
        days   = this._days   + milliseconds / 864e5;
        months = this._months + daysToMonths(days);
        return units === 'month' ? months : months / 12;
    ***REMOVED*** else ***REMOVED***
        // handle milliseconds separately because of floating point math errors (issue #1867)
        days = this._days + Math.round(monthsToDays(this._months));
        switch (units) ***REMOVED***
            case 'week'   : return days / 7     + milliseconds / 6048e5;
            case 'day'    : return days         + milliseconds / 864e5;
            case 'hour'   : return days * 24    + milliseconds / 36e5;
            case 'minute' : return days * 1440  + milliseconds / 6e4;
            case 'second' : return days * 86400 + milliseconds / 1000;
            // Math.floor prevents floating point math errors here
            case 'millisecond': return Math.floor(days * 864e5) + milliseconds;
            default: throw new Error('Unknown unit ' + units);
        ***REMOVED***
    ***REMOVED***
***REMOVED***

// TODO: Use this.as('ms')?
export function valueOf () ***REMOVED***
    return (
        this._milliseconds +
        this._days * 864e5 +
        (this._months % 12) * 2592e6 +
        toInt(this._months / 12) * 31536e6
    );
***REMOVED***

function makeAs (alias) ***REMOVED***
    return function () ***REMOVED***
        return this.as(alias);
    ***REMOVED***;
***REMOVED***

export var asMilliseconds = makeAs('ms');
export var asSeconds      = makeAs('s');
export var asMinutes      = makeAs('m');
export var asHours        = makeAs('h');
export var asDays         = makeAs('d');
export var asWeeks        = makeAs('w');
export var asMonths       = makeAs('M');
export var asYears        = makeAs('y');
