// mongoose/lib/error/cast.js
/*!
 * Module dependencies.
 */

var MongooseError = require('../error.js');
var util = require('util');

/**
 * Casting Error constructor.
 *
 * @param ***REMOVED***String***REMOVED*** type
 * @param ***REMOVED***String***REMOVED*** value
 * @inherits MongooseError
 * @api private
 */

function CastError(type, value, path, reason) ***REMOVED***
  var stringValue = util.inspect(value);
  stringValue = stringValue.replace(/^'/, '"').replace(/'$/, '"');
  if (stringValue.charAt(0) !== '"') ***REMOVED***
    stringValue = '"' + stringValue + '"';
  ***REMOVED***
  MongooseError.call(this, 'Cast to ' + type + ' failed for value ' +
    stringValue + ' at path "' + path + '"');
  if (Error.captureStackTrace) ***REMOVED***
    Error.captureStackTrace(this);
  ***REMOVED*** else ***REMOVED***
    this.stack = new Error().stack;
  ***REMOVED***
  this.stringValue = stringValue;
  this.name = 'CastError';
  this.kind = type;
  this.value = value;
  this.path = path;
  this.reason = reason;
***REMOVED***

/*!
 * Inherits from MongooseError.
 */

CastError.prototype = Object.create(MongooseError.prototype);
CastError.prototype.constructor = MongooseError;

/*!
 * ignore
 */

CastError.prototype.setModel = function(model) ***REMOVED***
  this.model = model;
  this.message = 'Cast to ' + this.kind + ' failed for value ' +
    this.stringValue + ' at path "' + this.path + '"' + ' for model "' +
    model.modelName + '"';
***REMOVED***;

/*!
 * exports
 */

module.exports = CastError;
