// mongoose/lib/error/validator.js
/*!
 * Module dependencies.
 */

var MongooseError = require('../error.js');

/**
 * Schema validator error
 *
 * @param ***REMOVED***Object***REMOVED*** properties
 * @inherits MongooseError
 * @api private
 */

function ValidatorError(properties) ***REMOVED***
  var msg = properties.message;
  if (!msg) ***REMOVED***
    msg = MongooseError.messages.general.default;
  ***REMOVED***

  var message = this.formatMessage(msg, properties);
  MongooseError.call(this, message);
  if (Error.captureStackTrace) ***REMOVED***
    Error.captureStackTrace(this);
  ***REMOVED*** else ***REMOVED***
    this.stack = new Error().stack;
  ***REMOVED***
  this.properties = properties;
  this.name = 'ValidatorError';
  this.kind = properties.type;
  this.path = properties.path;
  this.value = properties.value;
***REMOVED***

/*!
 * Inherits from MongooseError
 */

ValidatorError.prototype = Object.create(MongooseError.prototype);
ValidatorError.prototype.constructor = MongooseError;

/*!
 * The object used to define this validator. Not enumerable to hide
 * it from `require('util').inspect()` output re: gh-3925
 */

Object.defineProperty(ValidatorError.prototype, 'properties', ***REMOVED***
  enumerable: false,
  writable: true,
  value: null
***REMOVED***);

/*!
 * Formats error messages
 */

ValidatorError.prototype.formatMessage = function(msg, properties) ***REMOVED***
  var propertyNames = Object.keys(properties);
  for (var i = 0; i < propertyNames.length; ++i) ***REMOVED***
    var propertyName = propertyNames[i];
    if (propertyName === 'message') ***REMOVED***
      continue;
    ***REMOVED***
    msg = msg.replace('***REMOVED***' + propertyName.toUpperCase() + '***REMOVED***', properties[propertyName]);
  ***REMOVED***
  return msg;
***REMOVED***;

/*!
 * toString helper
 */

ValidatorError.prototype.toString = function() ***REMOVED***
  return this.message;
***REMOVED***;

/*!
 * exports
 */

module.exports = ValidatorError;
