// muri/lib/index.js
// muri

/**
 * MongoDB URI parser as described here:
 * http://www.mongodb.org/display/DOCS/Connections
 */

/**
 * Module dependencies
 */

var qs = require('querystring');

/**
 * Defaults
 */

var DEFAULT_PORT = 27017;
var DEFAULT_DB = 'test';
var ADMIN_DB = 'admin';

/**
 * Muri
 */

module.exports = exports = function muri (str) ***REMOVED***
  if (!/^mongodb:\/\//.test(str)) ***REMOVED***
    throw new Error('Invalid mongodb uri. Must begin with "mongodb://"'
                  + '\n  Received: ' + str);
  ***REMOVED***

  var ret = ***REMOVED***
      hosts: []
    , db: DEFAULT_DB
    , options: ***REMOVED******REMOVED***
  ***REMOVED***

  var match = /^mongodb:\/\/([^?]+)(\??.*)$/.exec(str);
  if (!match || '/' == match[1]) ***REMOVED***
    throw new Error('Invalid mongodb uri. Missing hostname');
  ***REMOVED***

  var uris = match[1];
  var path = match[2];
  var db;

  uris.split(',').forEach(function (uri) ***REMOVED***
    var o = parse(uri);

    if (o.host) ***REMOVED***
      ret.hosts.push(***REMOVED***
          host: o.host
        , port: parseInt(o.port, 10)
      ***REMOVED***)

      if (!db && o.db) ***REMOVED***
        db = o.db;
      ***REMOVED***
    ***REMOVED*** else if (o.ipc) ***REMOVED***
      ret.hosts.push(***REMOVED*** ipc: o.ipc ***REMOVED***);
    ***REMOVED***

    if (o.auth) ***REMOVED***
      ret.auth = ***REMOVED***
          user: o.auth.user
        , pass: o.auth.pass
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***)

  if (!ret.hosts.length) ***REMOVED***
    throw new Error('Invalid mongodb uri. Missing hostname');
  ***REMOVED***

  var parts = path.split('?');

  if (!db) ***REMOVED***
    if (parts[0]) ***REMOVED***
      db = parts[0].replace(/^\//, '');
    ***REMOVED*** else ***REMOVED***
      // deal with ipc formats
      db = /\/([^\.]+)$/.exec(match[1]);
      if (db && db[1]) ***REMOVED***
        db = db[1];
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (db) ***REMOVED***
    ret.db = db;
  ***REMOVED*** else if (ret.auth) ***REMOVED***
    ret.db = ADMIN_DB;
  ***REMOVED***

  if (parts[1]) ***REMOVED***
    ret.options = options(parts[1]);
  ***REMOVED***

  return ret;
***REMOVED***

/**
 * Parse str into key/val pairs casting values appropriately.
 */

function options (str) ***REMOVED***
  var sep = /;/.test(str)
    ? ';'
    : '&';

  var ret = qs.parse(str, sep);

  Object.keys(ret).forEach(function (key) ***REMOVED***
    var val = ret[key];
    if ('readPreferenceTags' == key) ***REMOVED***
      val = readPref(val);
      if (val) ***REMOVED***
        ret[key] = Array.isArray(val)
          ? val
          : [val];
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      ret[key] = format(val);
    ***REMOVED***
  ***REMOVED***);

  return ret;
***REMOVED***

function format (val) ***REMOVED***
  var num;

  if ('true' == val) ***REMOVED***
    return true;
  ***REMOVED*** else if ('false' == val) ***REMOVED***
    return false;
  ***REMOVED*** else ***REMOVED***
    num = parseInt(+val, 10);
    if (!isNaN(num)) ***REMOVED***
      return num;
    ***REMOVED***
  ***REMOVED***

  return val;
***REMOVED***

function readPref (val) ***REMOVED***
  var ret;

  if (Array.isArray(val)) ***REMOVED***
    ret = val.map(readPref).filter(Boolean);
    return ret.length
      ? ret
      : undefined
  ***REMOVED***

  var pair = val.split(',');
  var hasKeys;
  ret = ***REMOVED******REMOVED***;

  pair.forEach(function (kv) ***REMOVED***
    kv = (kv || '').trim();
    if (!kv) return;
    hasKeys = true;
    var split = kv.split(':');
    ret[split[0]] = format(split[1]);
  ***REMOVED***);

  return hasKeys && ret;
***REMOVED***

var ipcRgx = /\.sock/;

function parse (uriString) ***REMOVED***
  // do not use require('url').parse b/c it can't handle # in username or pwd
  // mongo uris are strange

  var uri = uriString;
  var ret = ***REMOVED******REMOVED***;
  var parts;
  var auth;
  var ipcs;

  // skip protocol
  uri = uri.replace(/^mongodb:\/\//, '');

  // auth
  if (/@/.test(uri)) ***REMOVED***
    parts = uri.split(/@/);
    auth = parts[0];
    uri = parts[1];

    parts = auth.split(':');
    ret.auth = ***REMOVED******REMOVED***;
    ret.auth.user = parts[0];
    ret.auth.pass = parts[1];
  ***REMOVED***

  // unix domain sockets
  if (ipcRgx.test(uri)) ***REMOVED***
    ipcs = uri.split(ipcRgx);
    ret.ipc = ipcs[0] + '.sock';

    // included a database?
    if (ipcs[1]) ***REMOVED***
      // strip leading / from database name
      ipcs[1] = ipcs[1].replace(/^\//, '');

      if (ipcs[1]) ***REMOVED***
        ret.db = ipcs[1];
      ***REMOVED***
    ***REMOVED***

    return ret;
  ***REMOVED***

  // database name
  parts = uri.split('/');
  if (parts[1]) ret.db = parts[1];

  // ipv6, [ip]:host
  if (/^\[[^\]]+\]:\d+/.test(parts[0])) ***REMOVED***
    ret.host = parts[0].substring(1, parts[0].indexOf(']:'));
    ret.port = parts[0].substring(parts[0].indexOf(']:') + ']:'.length);
  ***REMOVED*** else ***REMOVED***
    // host:port
    parts = parts[0].split(':');
    ret.host = parts[0];
    ret.port = parts[1] || DEFAULT_PORT;
  ***REMOVED***

  return ret;
***REMOVED***

/**
 * Version
 */

module.exports.version = JSON.parse(
  require('fs').readFileSync(__dirname + '/../package.json', 'utf8')
).version;
