// depd/lib/compat/index.js
/*!
 * depd
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var Buffer = require('buffer')
var EventEmitter = require('events').EventEmitter

/**
 * Module exports.
 * @public
 */

lazyProperty(module.exports, 'bufferConcat', function bufferConcat() ***REMOVED***
  return Buffer.concat || require('./buffer-concat')
***REMOVED***)

lazyProperty(module.exports, 'callSiteToString', function callSiteToString() ***REMOVED***
  var limit = Error.stackTraceLimit
  var obj = ***REMOVED******REMOVED***
  var prep = Error.prepareStackTrace

  function prepareObjectStackTrace(obj, stack) ***REMOVED***
    return stack
  ***REMOVED***

  Error.prepareStackTrace = prepareObjectStackTrace
  Error.stackTraceLimit = 2

  // capture the stack
  Error.captureStackTrace(obj)

  // slice the stack
  var stack = obj.stack.slice()

  Error.prepareStackTrace = prep
  Error.stackTraceLimit = limit

  return stack[0].toString ? toString : require('./callsite-tostring')
***REMOVED***)

lazyProperty(module.exports, 'eventListenerCount', function eventListenerCount() ***REMOVED***
  return EventEmitter.listenerCount || require('./event-listener-count')
***REMOVED***)

/**
 * Define a lazy property.
 */

function lazyProperty(obj, prop, getter) ***REMOVED***
  function get() ***REMOVED***
    var val = getter()

    Object.defineProperty(obj, prop, ***REMOVED***
      configurable: true,
      enumerable: true,
      value: val
    ***REMOVED***)

    return val
  ***REMOVED***

  Object.defineProperty(obj, prop, ***REMOVED***
    configurable: true,
    enumerable: true,
    get: get
  ***REMOVED***)
***REMOVED***

/**
 * Call toString() on the obj
 */

function toString(obj) ***REMOVED***
  return obj.toString()
***REMOVED***
