// serve-static/index.js
/*!
 * serve-static
 * Copyright(c) 2010 Sencha Inc.
 * Copyright(c) 2011 TJ Holowaychuk
 * Copyright(c) 2014-2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var encodeUrl = require('encodeurl')
var escapeHtml = require('escape-html')
var parseUrl = require('parseurl')
var resolve = require('path').resolve
var send = require('send')
var url = require('url')

/**
 * Module exports.
 * @public
 */

module.exports = serveStatic
module.exports.mime = send.mime

/**
 * @param ***REMOVED***string***REMOVED*** root
 * @param ***REMOVED***object***REMOVED*** [options]
 * @return ***REMOVED***function***REMOVED***
 * @public
 */

function serveStatic (root, options) ***REMOVED***
  if (!root) ***REMOVED***
    throw new TypeError('root path required')
  ***REMOVED***

  if (typeof root !== 'string') ***REMOVED***
    throw new TypeError('root path must be a string')
  ***REMOVED***

  // copy options object
  var opts = Object.create(options || null)

  // fall-though
  var fallthrough = opts.fallthrough !== false

  // default redirect
  var redirect = opts.redirect !== false

  // headers listener
  var setHeaders = opts.setHeaders

  if (setHeaders && typeof setHeaders !== 'function') ***REMOVED***
    throw new TypeError('option setHeaders must be function')
  ***REMOVED***

  // setup options for send
  opts.maxage = opts.maxage || opts.maxAge || 0
  opts.root = resolve(root)

  // construct directory listener
  var onDirectory = redirect
    ? createRedirectDirectoryListener()
    : createNotFoundDirectoryListener()

  return function serveStatic (req, res, next) ***REMOVED***
    if (req.method !== 'GET' && req.method !== 'HEAD') ***REMOVED***
      if (fallthrough) ***REMOVED***
        return next()
      ***REMOVED***

      // method not allowed
      res.statusCode = 405
      res.setHeader('Allow', 'GET, HEAD')
      res.setHeader('Content-Length', '0')
      res.end()
      return
    ***REMOVED***

    var forwardError = !fallthrough
    var originalUrl = parseUrl.original(req)
    var path = parseUrl(req).pathname

    // make sure redirect occurs at mount
    if (path === '/' && originalUrl.pathname.substr(-1) !== '/') ***REMOVED***
      path = ''
    ***REMOVED***

    // create send stream
    var stream = send(req, path, opts)

    // add directory handler
    stream.on('directory', onDirectory)

    // add headers listener
    if (setHeaders) ***REMOVED***
      stream.on('headers', setHeaders)
    ***REMOVED***

    // add file listener for fallthrough
    if (fallthrough) ***REMOVED***
      stream.on('file', function onFile () ***REMOVED***
        // once file is determined, always forward error
        forwardError = true
      ***REMOVED***)
    ***REMOVED***

    // forward errors
    stream.on('error', function error (err) ***REMOVED***
      if (forwardError || !(err.statusCode < 500)) ***REMOVED***
        next(err)
        return
      ***REMOVED***

      next()
    ***REMOVED***)

    // pipe
    stream.pipe(res)
  ***REMOVED***
***REMOVED***

/**
 * Collapse all leading slashes into a single slash
 * @private
 */
function collapseLeadingSlashes (str) ***REMOVED***
  for (var i = 0; i < str.length; i++) ***REMOVED***
    if (str[i] !== '/') ***REMOVED***
      break
    ***REMOVED***
  ***REMOVED***

  return i > 1
    ? '/' + str.substr(i)
    : str
***REMOVED***

/**
 * Create a directory listener that just 404s.
 * @private
 */

function createNotFoundDirectoryListener () ***REMOVED***
  return function notFound () ***REMOVED***
    this.error(404)
  ***REMOVED***
***REMOVED***

/**
 * Create a directory listener that performs a redirect.
 * @private
 */

function createRedirectDirectoryListener () ***REMOVED***
  return function redirect () ***REMOVED***
    if (this.hasTrailingSlash()) ***REMOVED***
      this.error(404)
      return
    ***REMOVED***

    // get original URL
    var originalUrl = parseUrl.original(this.req)

    // append trailing slash
    originalUrl.path = null
    originalUrl.pathname = collapseLeadingSlashes(originalUrl.pathname + '/')

    // reformat the URL
    var loc = encodeUrl(url.format(originalUrl))
    var msg = 'Redirecting to <a href="' + escapeHtml(loc) + '">' + escapeHtml(loc) + '</a>\n'
    var res = this.res

    // send redirect response
    res.statusCode = 301
    res.setHeader('Content-Type', 'text/html; charset=UTF-8')
    res.setHeader('Content-Length', Buffer.byteLength(msg))
    res.setHeader('X-Content-Type-Options', 'nosniff')
    res.setHeader('Location', loc)
    res.end(msg)
  ***REMOVED***
***REMOVED***
