// mongodb/lib/cursor.js
"use strict";

var inherits = require('util').inherits
  , f = require('util').format
  , formattedOrderClause = require('./utils').formattedOrderClause
  , handleCallback = require('./utils').handleCallback
  , ReadPreference = require('./read_preference')
  , MongoError = require('mongodb-core').MongoError
  , Readable = require('stream').Readable || require('readable-stream').Readable
  , Define = require('./metadata')
  , CoreCursor = require('mongodb-core').Cursor
  , Map = require('mongodb-core').BSON.Map
  , CoreReadPreference = require('mongodb-core').ReadPreference;

/**
 * @fileOverview The **Cursor** class is an internal class that embodies a cursor on MongoDB
 * allowing for iteration over the results returned from the underlying query. It supports
 * one by one document iteration, conversion to an array or can be iterated as a Node 0.10.X
 * or higher stream
 *
 * **CURSORS Cannot directly be instantiated**
 * @example
 * var MongoClient = require('mongodb').MongoClient,
 *   test = require('assert');
 * // Connection url
 * var url = 'mongodb://localhost:27017/test';
 * // Connect using MongoClient
 * MongoClient.connect(url, function(err, db) ***REMOVED***
 *   // Create a collection we want to drop later
 *   var col = db.collection('createIndexExample1');
 *   // Insert a bunch of documents
 *   col.insert([***REMOVED***a:1, b:1***REMOVED***
 *     , ***REMOVED***a:2, b:2***REMOVED***, ***REMOVED***a:3, b:3***REMOVED***
 *     , ***REMOVED***a:4, b:4***REMOVED***], ***REMOVED***w:1***REMOVED***, function(err, result) ***REMOVED***
 *     test.equal(null, err);
 *
 *     // Show that duplicate records got dropped
 *     col.find(***REMOVED******REMOVED***).toArray(function(err, items) ***REMOVED***
 *       test.equal(null, err);
 *       test.equal(4, items.length);
 *       db.close();
 *     ***REMOVED***);
 *   ***REMOVED***);
 * ***REMOVED***);
 */

/**
 * Namespace provided by the mongodb-core and node.js
 * @external CoreCursor
 * @external Readable
 */

// Flags allowed for cursor
var flags = ['tailable', 'oplogReplay', 'noCursorTimeout', 'awaitData', 'exhaust', 'partial'];
var fields = ['numberOfRetries', 'tailableRetryInterval'];
var push = Array.prototype.push;

/**
 * Creates a new Cursor instance (INTERNAL TYPE, do not instantiate directly)
 * @class Cursor
 * @extends external:CoreCursor
 * @extends external:Readable
 * @property ***REMOVED***string***REMOVED*** sortValue Cursor query sort setting.
 * @property ***REMOVED***boolean***REMOVED*** timeout Is Cursor able to time out.
 * @property ***REMOVED***ReadPreference***REMOVED*** readPreference Get cursor ReadPreference.
 * @fires Cursor#data
 * @fires Cursor#end
 * @fires Cursor#close
 * @fires Cursor#readable
 * @return ***REMOVED***Cursor***REMOVED*** a Cursor instance.
 * @example
 * Cursor cursor options.
 *
 * collection.find(***REMOVED******REMOVED***).project(***REMOVED***a:1***REMOVED***)                             // Create a projection of field a
 * collection.find(***REMOVED******REMOVED***).skip(1).limit(10)                          // Skip 1 and limit 10
 * collection.find(***REMOVED******REMOVED***).batchSize(5)                               // Set batchSize on cursor to 5
 * collection.find(***REMOVED******REMOVED***).filter(***REMOVED***a:1***REMOVED***)                              // Set query on the cursor
 * collection.find(***REMOVED******REMOVED***).comment('add a comment')                   // Add a comment to the query, allowing to correlate queries
 * collection.find(***REMOVED******REMOVED***).addCursorFlag('tailable', true)            // Set cursor as tailable
 * collection.find(***REMOVED******REMOVED***).addCursorFlag('oplogReplay', true)         // Set cursor as oplogReplay
 * collection.find(***REMOVED******REMOVED***).addCursorFlag('noCursorTimeout', true)     // Set cursor as noCursorTimeout
 * collection.find(***REMOVED******REMOVED***).addCursorFlag('awaitData', true)           // Set cursor as awaitData
 * collection.find(***REMOVED******REMOVED***).addCursorFlag('partial', true)             // Set cursor as partial
 * collection.find(***REMOVED******REMOVED***).addQueryModifier('$orderby', ***REMOVED***a:1***REMOVED***)        // Set $orderby ***REMOVED***a:1***REMOVED***
 * collection.find(***REMOVED******REMOVED***).max(10)                                    // Set the cursor maxScan
 * collection.find(***REMOVED******REMOVED***).maxScan(10)                                // Set the cursor maxScan
 * collection.find(***REMOVED******REMOVED***).maxTimeMS(1000)                            // Set the cursor maxTimeMS
 * collection.find(***REMOVED******REMOVED***).min(100)                                   // Set the cursor min
 * collection.find(***REMOVED******REMOVED***).returnKey(10)                              // Set the cursor returnKey
 * collection.find(***REMOVED******REMOVED***).setReadPreference(ReadPreference.PRIMARY)  // Set the cursor readPreference
 * collection.find(***REMOVED******REMOVED***).showRecordId(true)                         // Set the cursor showRecordId
 * collection.find(***REMOVED******REMOVED***).snapshot(true)                             // Set the cursor snapshot
 * collection.find(***REMOVED******REMOVED***).sort([['a', 1]])                           // Sets the sort order of the cursor query
 * collection.find(***REMOVED******REMOVED***).hint('a_1')                                // Set the cursor hint
 *
 * All options are chainable, so one can do the following.
 *
 * collection.find(***REMOVED******REMOVED***).maxTimeMS(1000).maxScan(100).skip(1).toArray(..)
 */
var Cursor = function(bson, ns, cmd, options, topology, topologyOptions) ***REMOVED***
  CoreCursor.apply(this, Array.prototype.slice.call(arguments, 0));
  var self = this;
  var state = Cursor.INIT;
  var streamOptions = ***REMOVED******REMOVED***;

  // Tailable cursor options
  var numberOfRetries = options.numberOfRetries || 5;
  var tailableRetryInterval = options.tailableRetryInterval || 500;
  var currentNumberOfRetries = numberOfRetries;

  // Get the promiseLibrary
  var promiseLibrary = options.promiseLibrary;

  // No promise library selected fall back
  if(!promiseLibrary) ***REMOVED***
    promiseLibrary = typeof global.Promise == 'function' ?
      global.Promise : require('es6-promise').Promise;
  ***REMOVED***

  // Set up
  Readable.call(this, ***REMOVED***objectMode: true***REMOVED***);

  // Internal cursor state
  this.s = ***REMOVED***
    // Tailable cursor options
      numberOfRetries: numberOfRetries
    , tailableRetryInterval: tailableRetryInterval
    , currentNumberOfRetries: currentNumberOfRetries
    // State
    , state: state
    // Stream options
    , streamOptions: streamOptions
    // BSON
    , bson: bson
    // Namespace
    , ns: ns
    // Command
    , cmd: cmd
    // Options
    , options: options
    // Topology
    , topology: topology
    // Topology options
    , topologyOptions: topologyOptions
    // Promise library
    , promiseLibrary: promiseLibrary
    // Current doc
    , currentDoc: null
  ***REMOVED***

  // Translate correctly
  if(self.s.options.noCursorTimeout == true) ***REMOVED***
    self.addCursorFlag('noCursorTimeout', true);
  ***REMOVED***

  // Set the sort value
  this.sortValue = self.s.cmd.sort;
***REMOVED***

/**
 * Cursor stream data event, fired for each document in the cursor.
 *
 * @event Cursor#data
 * @type ***REMOVED***object***REMOVED***
 */

/**
 * Cursor stream end event
 *
 * @event Cursor#end
 * @type ***REMOVED***null***REMOVED***
 */

/**
 * Cursor stream close event
 *
 * @event Cursor#close
 * @type ***REMOVED***null***REMOVED***
 */

/**
 * Cursor stream readable event
 *
 * @event Cursor#readable
 * @type ***REMOVED***null***REMOVED***
 */

// Inherit from Readable
inherits(Cursor, Readable);

// Map core cursor _next method so we can apply mapping
CoreCursor.prototype._next = CoreCursor.prototype.next;

for(var name in CoreCursor.prototype) ***REMOVED***
  Cursor.prototype[name] = CoreCursor.prototype[name];
***REMOVED***

var define = Cursor.define = new Define('Cursor', Cursor, true);

/**
 * Check if there is any document still available in the cursor
 * @method
 * @param ***REMOVED***Cursor~resultCallback***REMOVED*** [callback] The result callback.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.hasNext = function(callback) ***REMOVED***
  var self = this;

  // Execute using callback
  if(typeof callback == 'function') ***REMOVED***
    if(self.s.currentDoc)***REMOVED***
      return callback(null, true);
    ***REMOVED*** else ***REMOVED***
      return nextObject(self, function(err, doc) ***REMOVED***
        if(!doc) return callback(null, false);
        self.s.currentDoc = doc;
        callback(null, true);
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  // Return a Promise
  return new this.s.promiseLibrary(function(resolve, reject) ***REMOVED***
    if(self.s.currentDoc)***REMOVED***
      resolve(true);
    ***REMOVED*** else ***REMOVED***
      nextObject(self, function(err, doc) ***REMOVED***
        if(self.s.state == Cursor.CLOSED || self.isDead()) return resolve(false);
        if(err) return reject(err);
        if(!doc) return resolve(false);
        self.s.currentDoc = doc;
        resolve(true);
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

define.classMethod('hasNext', ***REMOVED***callback: true, promise:true***REMOVED***);

/**
 * Get the next available document from the cursor, returns null if no more documents are available.
 * @method
 * @param ***REMOVED***Cursor~resultCallback***REMOVED*** [callback] The result callback.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.next = function(callback) ***REMOVED***
  var self = this;

  // Execute using callback
  if(typeof callback == 'function') ***REMOVED***
    // Return the currentDoc if someone called hasNext first
    if(self.s.currentDoc) ***REMOVED***
      var doc = self.s.currentDoc;
      self.s.currentDoc = null;
      return callback(null, doc);
    ***REMOVED***

    // Return the next object
    return nextObject(self, callback)
  ***REMOVED***

  // Return a Promise
  return new this.s.promiseLibrary(function(resolve, reject) ***REMOVED***
    // Return the currentDoc if someone called hasNext first
    if(self.s.currentDoc) ***REMOVED***
      var doc = self.s.currentDoc;
      self.s.currentDoc = null;
      return resolve(doc);
    ***REMOVED***

    nextObject(self, function(err, r) ***REMOVED***
      if(err) return reject(err);
      resolve(r);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

define.classMethod('next', ***REMOVED***callback: true, promise:true***REMOVED***);

/**
 * Set the cursor query
 * @method
 * @param ***REMOVED***object***REMOVED*** filter The filter object used for the cursor.
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.filter = function(filter) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.query = filter;
  return this;
***REMOVED***

define.classMethod('filter', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the cursor maxScan
 * @method
 * @param ***REMOVED***object***REMOVED*** maxScan Constrains the query to only scan the specified number of documents when fulfilling the query
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.maxScan = function(maxScan) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.maxScan = maxScan;
  return this;
***REMOVED***

define.classMethod('maxScan', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the cursor hint
 * @method
 * @param ***REMOVED***object***REMOVED*** hint If specified, then the query system will only consider plans using the hinted index.
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.hint = function(hint) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.hint = hint;
  return this;
***REMOVED***

define.classMethod('hint', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the cursor min
 * @method
 * @param ***REMOVED***object***REMOVED*** min Specify a $min value to specify the inclusive lower bound for a specific index in order to constrain the results of find(). The $min specifies the lower bound for all keys of a specific index in order.
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.min = function(min) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.min = min;
  return this;
***REMOVED***

define.classMethod('min', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the cursor max
 * @method
 * @param ***REMOVED***object***REMOVED*** max Specify a $max value to specify the exclusive upper bound for a specific index in order to constrain the results of find(). The $max specifies the upper bound for all keys of a specific index in order.
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.max = function(max) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.max = max;
  return this;
***REMOVED***

define.classMethod('max', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the cursor returnKey
 * @method
 * @param ***REMOVED***object***REMOVED*** returnKey Only return the index field or fields for the results of the query. If $returnKey is set to true and the query does not use an index to perform the read operation, the returned documents will not contain any fields. Use one of the following forms:
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.returnKey = function(value) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.returnKey = value;
  return this;
***REMOVED***

define.classMethod('returnKey', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the cursor showRecordId
 * @method
 * @param ***REMOVED***object***REMOVED*** showRecordId The $showDiskLoc option has now been deprecated and replaced with the showRecordId field. $showDiskLoc will still be accepted for OP_QUERY stye find.
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.showRecordId = function(value) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.showDiskLoc = value;
  return this;
***REMOVED***

define.classMethod('showRecordId', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the cursor snapshot
 * @method
 * @param ***REMOVED***object***REMOVED*** snapshot The $snapshot operator prevents the cursor from returning a document more than once because an intervening write operation results in a move of the document.
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.snapshot = function(value) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.snapshot = value;
  return this;
***REMOVED***

define.classMethod('snapshot', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set a node.js specific cursor option
 * @method
 * @param ***REMOVED***string***REMOVED*** field The cursor option to set ['numberOfRetries', 'tailableRetryInterval'].
 * @param ***REMOVED***object***REMOVED*** value The field value.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.setCursorOption = function(field, value) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  if(fields.indexOf(field) == -1) throw MongoError.create(***REMOVED***message: f("option %s not a supported option %s", field, fields), driver:true ***REMOVED***);
  this.s[field] = value;
  if(field == 'numberOfRetries')
    this.s.currentNumberOfRetries = value;
  return this;
***REMOVED***

define.classMethod('setCursorOption', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Add a cursor flag to the cursor
 * @method
 * @param ***REMOVED***string***REMOVED*** flag The flag to set, must be one of following ['tailable', 'oplogReplay', 'noCursorTimeout', 'awaitData', 'partial'].
 * @param ***REMOVED***boolean***REMOVED*** value The flag boolean value.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.addCursorFlag = function(flag, value) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  if(flags.indexOf(flag) == -1) throw MongoError.create(***REMOVED***message: f("flag %s not a supported flag %s", flag, flags), driver:true ***REMOVED***);
  if(typeof value != 'boolean') throw MongoError.create(***REMOVED***message: f("flag %s must be a boolean value", flag), driver:true***REMOVED***);
  this.s.cmd[flag] = value;
  return this;
***REMOVED***

define.classMethod('addCursorFlag', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Add a query modifier to the cursor query
 * @method
 * @param ***REMOVED***string***REMOVED*** name The query modifier (must start with $, such as $orderby etc)
 * @param ***REMOVED***boolean***REMOVED*** value The flag boolean value.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.addQueryModifier = function(name, value) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  if(name[0] != '$') throw MongoError.create(***REMOVED***message: f("%s is not a valid query modifier"), driver:true***REMOVED***);
  // Strip of the $
  var field = name.substr(1);
  // Set on the command
  this.s.cmd[field] = value;
  // Deal with the special case for sort
  if(field == 'orderby') this.s.cmd.sort = this.s.cmd[field];
  return this;
***REMOVED***

define.classMethod('addQueryModifier', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Add a comment to the cursor query allowing for tracking the comment in the log.
 * @method
 * @param ***REMOVED***string***REMOVED*** value The comment attached to this query.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.comment = function(value) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.comment = value;
  return this;
***REMOVED***

define.classMethod('comment', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set a maxAwaitTimeMS on a tailing cursor query to allow to customize the timeout value for the option awaitData (Only supported on MongoDB 3.2 or higher, ignored otherwise)
 * @method
 * @param ***REMOVED***number***REMOVED*** value Number of milliseconds to wait before aborting the tailed query.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.maxAwaitTimeMS = function(value) ***REMOVED***
  if(typeof value != 'number') throw MongoError.create(***REMOVED***message: "maxAwaitTimeMS must be a number", driver:true***REMOVED***);
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.maxAwaitTimeMS = value;
  return this;
***REMOVED***

define.classMethod('maxAwaitTimeMS', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set a maxTimeMS on the cursor query, allowing for hard timeout limits on queries (Only supported on MongoDB 2.6 or higher)
 * @method
 * @param ***REMOVED***number***REMOVED*** value Number of milliseconds to wait before aborting the query.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.maxTimeMS = function(value) ***REMOVED***
  if(typeof value != 'number') throw MongoError.create(***REMOVED***message: "maxTimeMS must be a number", driver:true***REMOVED***);
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.maxTimeMS = value;
  return this;
***REMOVED***

define.classMethod('maxTimeMS', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

Cursor.prototype.maxTimeMs = Cursor.prototype.maxTimeMS;

define.classMethod('maxTimeMs', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Sets a field projection for the query.
 * @method
 * @param ***REMOVED***object***REMOVED*** value The field projection object.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.project = function(value) ***REMOVED***
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  this.s.cmd.fields = value;
  return this;
***REMOVED***

define.classMethod('project', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Sets the sort order of the cursor query.
 * @method
 * @param ***REMOVED***(string|array|object)***REMOVED*** keyOrList The key or keys set for the sort.
 * @param ***REMOVED***number***REMOVED*** [direction] The direction of the sorting (1 or -1).
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.sort = function(keyOrList, direction) ***REMOVED***
  if(this.s.options.tailable) throw MongoError.create(***REMOVED***message: "Tailable cursor doesn't support sorting", driver:true***REMOVED***);
  if(this.s.state == Cursor.CLOSED || this.s.state == Cursor.OPEN || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  var order = keyOrList;

  // We have an array of arrays, we need to preserve the order of the sort
  // so we will us a Map
  if(Array.isArray(order) && Array.isArray(order[0])) ***REMOVED***
    order = new Map(order.map(function(x) ***REMOVED***
      var value = [x[0], null];
      if(x[1] == 'asc') ***REMOVED***
        value[1] = 1;
      ***REMOVED*** else if(x[1] == 'desc') ***REMOVED***
        value[1] = -1;
      ***REMOVED*** else if(x[1] == 1 || x[1] == -1) ***REMOVED***
        value[1] = x[1];
      ***REMOVED*** else ***REMOVED***
        throw new MongoError("Illegal sort clause, must be of the form [['field1', '(ascending|descending)'], ['field2', '(ascending|descending)']]");
      ***REMOVED***

      return value;
    ***REMOVED***));
  ***REMOVED***

  if(direction != null) ***REMOVED***
    order = [[keyOrList, direction]];
  ***REMOVED***

  this.s.cmd.sort = order;
  this.sortValue = order;
  return this;
***REMOVED***

define.classMethod('sort', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the batch size for the cursor.
 * @method
 * @param ***REMOVED***number***REMOVED*** value The batchSize for the cursor.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.batchSize = function(value) ***REMOVED***
  if(this.s.options.tailable) throw MongoError.create(***REMOVED***message: "Tailable cursor doesn't support batchSize", driver:true***REMOVED***);
  if(this.s.state == Cursor.CLOSED || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  if(typeof value != 'number') throw MongoError.create(***REMOVED***message: "batchSize requires an integer", driver:true***REMOVED***);
  this.s.cmd.batchSize = value;
  this.setCursorBatchSize(value);
  return this;
***REMOVED***

define.classMethod('batchSize', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the collation options for the cursor.
 * @method
 * @param ***REMOVED***object***REMOVED*** value The cursor collation options (MongoDB 3.4 or higher) settings for update operation (see 3.4 documentation for available fields).
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.collation = function(value) ***REMOVED***
  this.s.cmd.collation = value;
  return this;
***REMOVED***

define.classMethod('collation', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the limit for the cursor.
 * @method
 * @param ***REMOVED***number***REMOVED*** value The limit for the cursor query.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.limit = function(value) ***REMOVED***
  if(this.s.options.tailable) throw MongoError.create(***REMOVED***message: "Tailable cursor doesn't support limit", driver:true***REMOVED***);
  if(this.s.state == Cursor.OPEN || this.s.state == Cursor.CLOSED || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  if(typeof value != 'number') throw MongoError.create(***REMOVED***message: "limit requires an integer", driver:true***REMOVED***);
  this.s.cmd.limit = value;
  // this.cursorLimit = value;
  this.setCursorLimit(value);
  return this;
***REMOVED***

define.classMethod('limit', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Set the skip for the cursor.
 * @method
 * @param ***REMOVED***number***REMOVED*** value The skip for the cursor query.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.skip = function(value) ***REMOVED***
  if(this.s.options.tailable) throw MongoError.create(***REMOVED***message: "Tailable cursor doesn't support skip", driver:true***REMOVED***);
  if(this.s.state == Cursor.OPEN || this.s.state == Cursor.CLOSED || this.isDead()) throw MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***);
  if(typeof value != 'number') throw MongoError.create(***REMOVED***message: "skip requires an integer", driver:true***REMOVED***);
  this.s.cmd.skip = value;
  this.setCursorSkip(value);
  return this;
***REMOVED***

define.classMethod('skip', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * The callback format for results
 * @callback Cursor~resultCallback
 * @param ***REMOVED***MongoError***REMOVED*** error An error instance representing the error during the execution.
 * @param ***REMOVED***(object|null|boolean)***REMOVED*** result The result object if the command was executed successfully.
 */

/**
 * Clone the cursor
 * @function external:CoreCursor#clone
 * @return ***REMOVED***Cursor***REMOVED***
 */

/**
 * Resets the cursor
 * @function external:CoreCursor#rewind
 * @return ***REMOVED***null***REMOVED***
 */

/**
 * Get the next available document from the cursor, returns null if no more documents are available.
 * @method
 * @param ***REMOVED***Cursor~resultCallback***REMOVED*** [callback] The result callback.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @deprecated
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.nextObject = Cursor.prototype.next;

var nextObject = function(self, callback) ***REMOVED***
  if(self.s.state == Cursor.CLOSED || self.isDead && self.isDead()) return handleCallback(callback, MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***));
  if(self.s.state == Cursor.INIT && self.s.cmd.sort) ***REMOVED***
    try ***REMOVED***
      self.s.cmd.sort = formattedOrderClause(self.s.cmd.sort);
    ***REMOVED*** catch(err) ***REMOVED***
      return handleCallback(callback, err);
    ***REMOVED***
  ***REMOVED***

  // Get the next object
  self._next(function(err, doc) ***REMOVED***
    self.s.state = Cursor.OPEN;
    if(err) return handleCallback(callback, err);
    handleCallback(callback, null, doc);
  ***REMOVED***);
***REMOVED***

define.classMethod('nextObject', ***REMOVED***callback: true, promise:true***REMOVED***);

// Trampoline emptying the number of retrieved items
// without incurring a nextTick operation
var loop = function(self, callback) ***REMOVED***
  // No more items we are done
  if(self.bufferedCount() == 0) return;
  // Get the next document
  self._next(callback);
  // Loop
  return loop;
***REMOVED***

Cursor.prototype.next = Cursor.prototype.nextObject;

define.classMethod('next', ***REMOVED***callback: true, promise:true***REMOVED***);

/**
 * Iterates over all the documents for this cursor. As with *****REMOVED***cursor.toArray***REMOVED*****,
 * not all of the elements will be iterated if this cursor had been previouly accessed.
 * In that case, *****REMOVED***cursor.rewind***REMOVED***** can be used to reset the cursor. However, unlike
 * *****REMOVED***cursor.toArray***REMOVED*****, the cursor will only hold a maximum of batch size elements
 * at any given time if batch size is specified. Otherwise, the caller is responsible
 * for making sure that the entire result can fit the memory.
 * @method
 * @deprecated
 * @param ***REMOVED***Cursor~resultCallback***REMOVED*** callback The result callback.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***null***REMOVED***
 */
Cursor.prototype.each = function(callback) ***REMOVED***
  // Rewind cursor state
  this.rewind();
  // Set current cursor to INIT
  this.s.state = Cursor.INIT;
  // Run the query
  _each(this, callback);
***REMOVED***;

define.classMethod('each', ***REMOVED***callback: true, promise:false***REMOVED***);

// Run the each loop
var _each = function(self, callback) ***REMOVED***
  if(!callback) throw MongoError.create(***REMOVED***message: 'callback is mandatory', driver:true***REMOVED***);
  if(self.isNotified()) return;
  if(self.s.state == Cursor.CLOSED || self.isDead()) ***REMOVED***
    return handleCallback(callback, MongoError.create(***REMOVED***message: "Cursor is closed", driver:true***REMOVED***));
  ***REMOVED***

  if(self.s.state == Cursor.INIT) self.s.state = Cursor.OPEN;

  // Define function to avoid global scope escape
  var fn = null;
  // Trampoline all the entries
  if(self.bufferedCount() > 0) ***REMOVED***
    while(fn = loop(self, callback)) fn(self, callback);
    _each(self, callback);
  ***REMOVED*** else ***REMOVED***
    self.next(function(err, item) ***REMOVED***
      if(err) return handleCallback(callback, err);
      if(item == null) ***REMOVED***
        self.s.state = Cursor.CLOSED;
        return handleCallback(callback, null, null);
      ***REMOVED***

      if(handleCallback(callback, null, item) == false) return;
      _each(self, callback);
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

/**
 * The callback format for the forEach iterator method
 * @callback Cursor~iteratorCallback
 * @param ***REMOVED***Object***REMOVED*** doc An emitted document for the iterator
 */

/**
 * The callback error format for the forEach iterator method
 * @callback Cursor~endCallback
 * @param ***REMOVED***MongoError***REMOVED*** error An error instance representing the error during the execution.
 */

/**
 * Iterates over all the documents for this cursor using the iterator, callback pattern.
 * @method
 * @param ***REMOVED***Cursor~iteratorCallback***REMOVED*** iterator The iteration callback.
 * @param ***REMOVED***Cursor~endCallback***REMOVED*** callback The end callback.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***null***REMOVED***
 */
Cursor.prototype.forEach = function(iterator, callback) ***REMOVED***
  this.each(function(err, doc)***REMOVED***
    if(err) ***REMOVED*** callback(err); return false; ***REMOVED***
    if(doc != null) ***REMOVED*** iterator(doc); return true; ***REMOVED***
    if(doc == null && callback) ***REMOVED***
      var internalCallback = callback;
      callback = null;
      internalCallback(null);
      return false;
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

define.classMethod('forEach', ***REMOVED***callback: true, promise:false***REMOVED***);

/**
 * Set the ReadPreference for the cursor.
 * @method
 * @param ***REMOVED***(string|ReadPreference)***REMOVED*** readPreference The new read preference for the cursor.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.setReadPreference = function(r) ***REMOVED***
  if(this.s.state != Cursor.INIT) throw MongoError.create(***REMOVED***message: 'cannot change cursor readPreference after cursor has been accessed', driver:true***REMOVED***);
  if(r instanceof ReadPreference) ***REMOVED***
    this.s.options.readPreference = new CoreReadPreference(r.mode, r.tags, ***REMOVED***maxStalenessSeconds: r.maxStalenessSeconds***REMOVED***);
  ***REMOVED*** else if(typeof r == 'string')***REMOVED***
    this.s.options.readPreference = new CoreReadPreference(r);
  ***REMOVED*** else if(r instanceof CoreReadPreference) ***REMOVED***
    this.s.options.readPreference = r;
  ***REMOVED***

  return this;
***REMOVED***

define.classMethod('setReadPreference', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * The callback format for results
 * @callback Cursor~toArrayResultCallback
 * @param ***REMOVED***MongoError***REMOVED*** error An error instance representing the error during the execution.
 * @param ***REMOVED***object[]***REMOVED*** documents All the documents the satisfy the cursor.
 */

/**
 * Returns an array of documents. The caller is responsible for making sure that there
 * is enough memory to store the results. Note that the array only contain partial
 * results when this cursor had been previouly accessed. In that case,
 * cursor.rewind() can be used to reset the cursor.
 * @method
 * @param ***REMOVED***Cursor~toArrayResultCallback***REMOVED*** [callback] The result callback.
 * @throws ***REMOVED***MongoError***REMOVED***
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.toArray = function(callback) ***REMOVED***
  var self = this;
  if(self.s.options.tailable) throw MongoError.create(***REMOVED***message: 'Tailable cursor cannot be converted to array', driver:true***REMOVED***);

  // Execute using callback
  if(typeof callback == 'function') return toArray(self, callback);

  // Return a Promise
  return new this.s.promiseLibrary(function(resolve, reject) ***REMOVED***
    toArray(self, function(err, r) ***REMOVED***
      if(err) return reject(err);
      resolve(r);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

var toArray = function(self, callback) ***REMOVED***
  var items = [];

  // Reset cursor
  self.rewind();
  self.s.state = Cursor.INIT;

  // Fetch all the documents
  var fetchDocs = function() ***REMOVED***
    self._next(function(err, doc) ***REMOVED***
      if(err) return handleCallback(callback, err);
      if(doc == null) ***REMOVED***
        self.s.state = Cursor.CLOSED;
        return handleCallback(callback, null, items);
      ***REMOVED***

      // Add doc to items
      items.push(doc)

      // Get all buffered objects
      if(self.bufferedCount() > 0) ***REMOVED***
        var docs = self.readBufferedDocuments(self.bufferedCount())

        // Transform the doc if transform method added
        if(self.s.transforms && typeof self.s.transforms.doc == 'function') ***REMOVED***
          docs = docs.map(self.s.transforms.doc);
        ***REMOVED***

        push.apply(items, docs);
      ***REMOVED***

      // Attempt a fetch
      fetchDocs();
    ***REMOVED***)
  ***REMOVED***

  fetchDocs();
***REMOVED***

define.classMethod('toArray', ***REMOVED***callback: true, promise:true***REMOVED***);

/**
 * The callback format for results
 * @callback Cursor~countResultCallback
 * @param ***REMOVED***MongoError***REMOVED*** error An error instance representing the error during the execution.
 * @param ***REMOVED***number***REMOVED*** count The count of documents.
 */

/**
 * Get the count of documents for this cursor
 * @method
 * @param ***REMOVED***boolean***REMOVED*** [applySkipLimit=true] Should the count command apply limit and skip settings on the cursor or in the passed in options.
 * @param ***REMOVED***object***REMOVED*** [options=null] Optional settings.
 * @param ***REMOVED***number***REMOVED*** [options.skip=null] The number of documents to skip.
 * @param ***REMOVED***number***REMOVED*** [options.limit=null] The maximum amounts to count before aborting.
 * @param ***REMOVED***number***REMOVED*** [options.maxTimeMS=null] Number of miliseconds to wait before aborting the query.
 * @param ***REMOVED***string***REMOVED*** [options.hint=null] An index name hint for the query.
 * @param ***REMOVED***(ReadPreference|string)***REMOVED*** [options.readPreference=null] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param ***REMOVED***Cursor~countResultCallback***REMOVED*** [callback] The result callback.
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.count = function(applySkipLimit, opts, callback) ***REMOVED***
  var self = this;
  if(self.s.cmd.query == null) throw MongoError.create(***REMOVED***message: "count can only be used with find command", driver:true***REMOVED***);
  if(typeof opts == 'function') callback = opts, opts = ***REMOVED******REMOVED***;
  opts = opts || ***REMOVED******REMOVED***;

  // Execute using callback
  if(typeof callback == 'function') return count(self, applySkipLimit, opts, callback);

  // Return a Promise
  return new this.s.promiseLibrary(function(resolve, reject) ***REMOVED***
    count(self, applySkipLimit, opts, function(err, r) ***REMOVED***
      if(err) return reject(err);
      resolve(r);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

var count = function(self, applySkipLimit, opts, callback) ***REMOVED***
  if(typeof applySkipLimit == 'function') ***REMOVED***
    callback = applySkipLimit;
    applySkipLimit = true;
  ***REMOVED***

  if(applySkipLimit) ***REMOVED***
    if(typeof self.cursorSkip() == 'number') opts.skip = self.cursorSkip();
    if(typeof self.cursorLimit() == 'number') opts.limit = self.cursorLimit();
  ***REMOVED***

  // Command
  var delimiter = self.s.ns.indexOf('.');

  var command = ***REMOVED***
    'count': self.s.ns.substr(delimiter+1), 'query': self.s.cmd.query
  ***REMOVED***

  if(typeof opts.maxTimeMS == 'number') ***REMOVED***
    command.maxTimeMS = opts.maxTimeMS;
  ***REMOVED*** else if(self.s.cmd && typeof self.s.cmd.maxTimeMS == 'number') ***REMOVED***
    command.maxTimeMS = self.s.cmd.maxTimeMS;
  ***REMOVED***

  // Merge in any options
  if(opts.skip) command.skip = opts.skip;
  if(opts.limit) command.limit = opts.limit;
  if(self.s.options.hint) command.hint = self.s.options.hint;

  // Execute the command
  self.topology.command(f("%s.$cmd", self.s.ns.substr(0, delimiter))
    , command, function(err, result) ***REMOVED***
      callback(err, result ? result.result.n : null)
    ***REMOVED***);
***REMOVED***

define.classMethod('count', ***REMOVED***callback: true, promise:true***REMOVED***);

/**
 * Close the cursor, sending a KillCursor command and emitting close.
 * @method
 * @param ***REMOVED***Cursor~resultCallback***REMOVED*** [callback] The result callback.
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.close = function(callback) ***REMOVED***
  this.s.state = Cursor.CLOSED;
  // Kill the cursor
  this.kill();
  // Emit the close event for the cursor
  this.emit('close');
  // Callback if provided
  if(typeof callback == 'function') return handleCallback(callback, null, this);
  // Return a Promise
  return new this.s.promiseLibrary(function(resolve) ***REMOVED***
    resolve();
  ***REMOVED***);
***REMOVED***

define.classMethod('close', ***REMOVED***callback: true, promise:true***REMOVED***);

/**
 * Map all documents using the provided function
 * @method
 * @param ***REMOVED***function***REMOVED*** [transform] The mapping transformation method.
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.map = function(transform) ***REMOVED***
  this.cursorState.transforms = ***REMOVED*** doc: transform ***REMOVED***;
  return this;
***REMOVED***

define.classMethod('map', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Is the cursor closed
 * @method
 * @return ***REMOVED***boolean***REMOVED***
 */
Cursor.prototype.isClosed = function() ***REMOVED***
  return this.isDead();
***REMOVED***

define.classMethod('isClosed', ***REMOVED***callback: false, promise:false, returns: [Boolean]***REMOVED***);

Cursor.prototype.destroy = function(err) ***REMOVED***
  if(err) this.emit('error', err);
  this.pause();
  this.close();
***REMOVED***

define.classMethod('destroy', ***REMOVED***callback: false, promise:false***REMOVED***);

/**
 * Return a modified Readable stream including a possible transform method.
 * @method
 * @param ***REMOVED***object***REMOVED*** [options=null] Optional settings.
 * @param ***REMOVED***function***REMOVED*** [options.transform=null] A transformation method applied to each document emitted by the stream.
 * @return ***REMOVED***Cursor***REMOVED***
 */
Cursor.prototype.stream = function(options) ***REMOVED***
  this.s.streamOptions = options || ***REMOVED******REMOVED***;
  return this;
***REMOVED***

define.classMethod('stream', ***REMOVED***callback: false, promise:false, returns: [Cursor]***REMOVED***);

/**
 * Execute the explain for the cursor
 * @method
 * @param ***REMOVED***Cursor~resultCallback***REMOVED*** [callback] The result callback.
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Cursor.prototype.explain = function(callback) ***REMOVED***
  var self = this;
  this.s.cmd.explain = true;

  // Do we have a readConcern
  if(this.s.cmd.readConcern) ***REMOVED***
    delete this.s.cmd['readConcern'];
  ***REMOVED***

  // Execute using callback
  if(typeof callback == 'function') return this._next(callback);

  // Return a Promise
  return new this.s.promiseLibrary(function(resolve, reject) ***REMOVED***
    self._next(function(err, r) ***REMOVED***
      if(err) return reject(err);
      resolve(r);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

define.classMethod('explain', ***REMOVED***callback: true, promise:true***REMOVED***);

Cursor.prototype._read = function() ***REMOVED***
  var self = this;
  if(self.s.state == Cursor.CLOSED || self.isDead()) ***REMOVED***
    return self.push(null);
  ***REMOVED***

  // Get the next item
  self.nextObject(function(err, result) ***REMOVED***
    if(err) ***REMOVED***
      if(self.listeners('error') && self.listeners('error').length > 0) ***REMOVED***
        self.emit('error', err);
      ***REMOVED***
      if(!self.isDead()) self.close();

      // Emit end event
      self.emit('end');
      return self.emit('finish');
    ***REMOVED***

    // If we provided a transformation method
    if(typeof self.s.streamOptions.transform == 'function' && result != null) ***REMOVED***
      return self.push(self.s.streamOptions.transform(result));
    ***REMOVED***

    // If we provided a map function
    if(self.cursorState.transforms && typeof self.cursorState.transforms.doc == 'function' && result != null) ***REMOVED***
      return self.push(self.cursorState.transforms.doc(result));
    ***REMOVED***

    // Return the result
    self.push(result);
  ***REMOVED***);
***REMOVED***

Object.defineProperty(Cursor.prototype, 'readPreference', ***REMOVED***
  enumerable:true,
  get: function() ***REMOVED***
    if (!this || !this.s) ***REMOVED***
      return null;
    ***REMOVED***

    return this.s.options.readPreference;
  ***REMOVED***
***REMOVED***);

Object.defineProperty(Cursor.prototype, 'namespace', ***REMOVED***
  enumerable: true,
  get: function() ***REMOVED***
    if (!this || !this.s) ***REMOVED***
      return null;
    ***REMOVED***

    // TODO: refactor this logic into core
    var ns = this.s.ns || '';
    var firstDot = ns.indexOf('.');
    if (firstDot < 0) ***REMOVED***
      return ***REMOVED***
        database: this.s.ns,
        collection: ''
      ***REMOVED***;
    ***REMOVED***
    return ***REMOVED***
      database: ns.substr(0, firstDot),
      collection: ns.substr(firstDot + 1)
    ***REMOVED***;
  ***REMOVED***
***REMOVED***);

/**
 * The read() method pulls some data out of the internal buffer and returns it. If there is no data available, then it will return null.
 * @function external:Readable#read
 * @param ***REMOVED***number***REMOVED*** size Optional argument to specify how much data to read.
 * @return ***REMOVED***(String | Buffer | null)***REMOVED***
 */

/**
 * Call this function to cause the stream to return strings of the specified encoding instead of Buffer objects.
 * @function external:Readable#setEncoding
 * @param ***REMOVED***string***REMOVED*** encoding The encoding to use.
 * @return ***REMOVED***null***REMOVED***
 */

/**
 * This method will cause the readable stream to resume emitting data events.
 * @function external:Readable#resume
 * @return ***REMOVED***null***REMOVED***
 */

/**
 * This method will cause a stream in flowing-mode to stop emitting data events. Any data that becomes available will remain in the internal buffer.
 * @function external:Readable#pause
 * @return ***REMOVED***null***REMOVED***
 */

/**
 * This method pulls all the data out of a readable stream, and writes it to the supplied destination, automatically managing the flow so that the destination is not overwhelmed by a fast readable stream.
 * @function external:Readable#pipe
 * @param ***REMOVED***Writable***REMOVED*** destination The destination for writing data
 * @param ***REMOVED***object***REMOVED*** [options] Pipe options
 * @return ***REMOVED***null***REMOVED***
 */

/**
 * This method will remove the hooks set up for a previous pipe() call.
 * @function external:Readable#unpipe
 * @param ***REMOVED***Writable***REMOVED*** [destination] The destination for writing data
 * @return ***REMOVED***null***REMOVED***
 */

/**
 * This is useful in certain cases where a stream is being consumed by a parser, which needs to "un-consume" some data that it has optimistically pulled out of the source, so that the stream can be passed on to some other party.
 * @function external:Readable#unshift
 * @param ***REMOVED***(Buffer|string)***REMOVED*** chunk Chunk of data to unshift onto the read queue.
 * @return ***REMOVED***null***REMOVED***
 */

/**
 * Versions of Node prior to v0.10 had streams that did not implement the entire Streams API as it is today. (See "Compatibility" below for more information.)
 * @function external:Readable#wrap
 * @param ***REMOVED***Stream***REMOVED*** stream An "old style" readable stream.
 * @return ***REMOVED***null***REMOVED***
 */

Cursor.INIT = 0;
Cursor.OPEN = 1;
Cursor.CLOSED = 2;
Cursor.GET_MORE = 3;

module.exports = Cursor;
