// mongodb/lib/metadata.js
var f = require('util').format;

var Define = function(name, object, stream) ***REMOVED***
  this.name = name;
  this.object = object;
  this.stream = typeof stream == 'boolean' ? stream : false;
  this.instrumentations = ***REMOVED******REMOVED***;
***REMOVED***

Define.prototype.classMethod = function(name, options) ***REMOVED***
  var keys = Object.keys(options).sort();
  var key = generateKey(keys, options);

  // Add a list of instrumentations
  if(this.instrumentations[key] == null) ***REMOVED***
    this.instrumentations[key] = ***REMOVED***
      methods: [], options: options
    ***REMOVED***
  ***REMOVED***

  // Push to list of method for this instrumentation
  this.instrumentations[key].methods.push(name);
***REMOVED***

var generateKey = function(keys, options) ***REMOVED***
  var parts = [];
  for(var i = 0; i < keys.length; i++) ***REMOVED***
    parts.push(f('%s=%s', keys[i], options[keys[i]]));
  ***REMOVED***

  return parts.join();
***REMOVED***

Define.prototype.staticMethod = function(name, options) ***REMOVED***
  options.static = true;
  var keys = Object.keys(options).sort();
  var key = generateKey(keys, options);

  // Add a list of instrumentations
  if(this.instrumentations[key] == null) ***REMOVED***
    this.instrumentations[key] = ***REMOVED***
      methods: [], options: options
    ***REMOVED***
  ***REMOVED***

  // Push to list of method for this instrumentation
  this.instrumentations[key].methods.push(name);
***REMOVED***

Define.prototype.generate = function() ***REMOVED***
  // Generate the return object
  var object = ***REMOVED***
    name: this.name, obj: this.object, stream: this.stream,
    instrumentations: []
  ***REMOVED***

  for(var name in this.instrumentations) ***REMOVED***
    object.instrumentations.push(this.instrumentations[name]);
  ***REMOVED***

  return object;
***REMOVED***

module.exports = Define;
