// mongodb/lib/read_preference.js
"use strict";

/**
 * @fileOverview The **ReadPreference** class is a class that represents a MongoDB ReadPreference and is
 * used to construct connections.
 *
 * @example
 * var Db = require('mongodb').Db,
 *   ReplSet = require('mongodb').ReplSet,
 *   Server = require('mongodb').Server,
 *   ReadPreference = require('mongodb').ReadPreference,
 *   test = require('assert');
 * // Connect using ReplSet
 * var server = new Server('localhost', 27017);
 * var db = new Db('test', new ReplSet([server]));
 * db.open(function(err, db) ***REMOVED***
 *   test.equal(null, err);
 *   // Perform a read
 *   var cursor = db.collection('t').find(***REMOVED******REMOVED***);
 *   cursor.setReadPreference(ReadPreference.PRIMARY);
 *   cursor.toArray(function(err, docs) ***REMOVED***
 *     test.equal(null, err);
 *     db.close();
 *   ***REMOVED***);
 * ***REMOVED***);
 */

/**
 * Creates a new ReadPreference instance
 *
 * Read Preferences
 *  - **ReadPreference.PRIMARY**, Read from primary only. All operations produce an error (throw an exception where applicable) if primary is unavailable. Cannot be combined with tags (This is the default.).
 *  - **ReadPreference.PRIMARY_PREFERRED**, Read from primary if available, otherwise a secondary.
 *  - **ReadPreference.SECONDARY**, Read from secondary if available, otherwise error.
 *  - **ReadPreference.SECONDARY_PREFERRED**, Read from a secondary if available, otherwise read from the primary.
 *  - **ReadPreference.NEAREST**, All modes read from among the nearest candidates, but unlike other modes, NEAREST will include both the primary and all secondaries in the random selection.
 *
 * @class
 * @param ***REMOVED***string***REMOVED*** mode The ReadPreference mode as listed above.
 * @param ***REMOVED***array|object***REMOVED*** tags An object representing read preference tags.
 * @param ***REMOVED***object***REMOVED*** [options] Additional read preference options
 * @param ***REMOVED***number***REMOVED*** [options.maxStalenessSeconds] Max Secondary Read Stalleness in Seconds
 * @return ***REMOVED***ReadPreference***REMOVED*** a ReadPreference instance.
 */
var ReadPreference = function(mode, tags, options) ***REMOVED***
  if(!(this instanceof ReadPreference)) ***REMOVED***
    return new ReadPreference(mode, tags, options);
  ***REMOVED***

  this._type = 'ReadPreference';
  this.mode = mode;
  this.tags = tags;
  this.options =  options;

  // If no tags were passed in
  if(tags && typeof tags == 'object' && !Array.isArray(tags)) ***REMOVED***
    if(tags.maxStalenessSeconds) ***REMOVED***
      this.options = tags;
      this.tags = null;
    ***REMOVED***
  ***REMOVED***

  // Add the maxStalenessSeconds value to the read Preference
  if(this.options && this.options.maxStalenessSeconds) ***REMOVED***
    this.maxStalenessSeconds = this.options.maxStalenessSeconds;
  ***REMOVED***
***REMOVED***

/**
 * Validate if a mode is legal
 *
 * @method
 * @param ***REMOVED***string***REMOVED*** mode The string representing the read preference mode.
 * @return ***REMOVED***boolean***REMOVED***
 */
ReadPreference.isValid = function(_mode) ***REMOVED***
  return (_mode == ReadPreference.PRIMARY || _mode == ReadPreference.PRIMARY_PREFERRED
    || _mode == ReadPreference.SECONDARY || _mode == ReadPreference.SECONDARY_PREFERRED
    || _mode == ReadPreference.NEAREST
    || _mode == true || _mode == false || _mode == null);
***REMOVED***

/**
 * Validate if a mode is legal
 *
 * @method
 * @param ***REMOVED***string***REMOVED*** mode The string representing the read preference mode.
 * @return ***REMOVED***boolean***REMOVED***
 */
ReadPreference.prototype.isValid = function(mode) ***REMOVED***
  var _mode = typeof mode == 'string' ? mode : this.mode;
  return ReadPreference.isValid(_mode);
***REMOVED***

/**
 * @ignore
 */
ReadPreference.prototype.toObject = function() ***REMOVED***
  var object = ***REMOVED***mode:this.mode***REMOVED***;

  if(this.tags != null) ***REMOVED***
    object['tags'] = this.tags;
  ***REMOVED***

  if(this.maxStalenessSeconds) ***REMOVED***
    object['maxStalenessSeconds'] = this.maxStalenessSeconds;
  ***REMOVED***

  return object;
***REMOVED***

/**
 * @ignore
 */
ReadPreference.prototype.toJSON = function() ***REMOVED***
  return this.toObject();
***REMOVED***

/**
 * @ignore
 */
ReadPreference.PRIMARY = 'primary';
ReadPreference.PRIMARY_PREFERRED = 'primaryPreferred';
ReadPreference.SECONDARY = 'secondary';
ReadPreference.SECONDARY_PREFERRED = 'secondaryPreferred';
ReadPreference.NEAREST = 'nearest'

/**
 * @ignore
 */
module.exports = ReadPreference;
