// joi/lib/index.js
// Load modules

var Any = require('./any');
var Cast = require('./cast');
var Ref = require('./ref');


// Declare internals

var internals = ***REMOVED***
    alternatives: require('./alternatives'),
    array: require('./array'),
    boolean: require('./boolean'),
    binary: require('./binary'),
    date: require('./date'),
    number: require('./number'),
    object: require('./object'),
    string: require('./string')
***REMOVED***;


internals.root = function () ***REMOVED***

    var any = new Any();

    var root = any.clone();
    root.any = function () ***REMOVED***

        return any;
    ***REMOVED***;

    root.alternatives = root.alt = function () ***REMOVED***

        return arguments.length ? internals.alternatives.try.apply(internals.alternatives, arguments) : internals.alternatives;
    ***REMOVED***;

    root.array = function () ***REMOVED***

        return internals.array;
    ***REMOVED***;

    root.boolean = root.bool = function () ***REMOVED***

        return internals.boolean;
    ***REMOVED***;

    root.binary = function () ***REMOVED***

        return internals.binary;
    ***REMOVED***;

    root.date = function () ***REMOVED***

        return internals.date;
    ***REMOVED***;

    root.func = function () ***REMOVED***

        return internals.object._func();
    ***REMOVED***;

    root.number = function () ***REMOVED***

        return internals.number;
    ***REMOVED***;

    root.object = function () ***REMOVED***

        return arguments.length ? internals.object.keys.apply(internals.object, arguments) : internals.object;
    ***REMOVED***;

    root.string = function () ***REMOVED***

        return internals.string;
    ***REMOVED***;

    root.ref = function () ***REMOVED***

        return Ref.create.apply(null, arguments);
    ***REMOVED***;

    root.isRef = function (ref) ***REMOVED***

        return Ref.isRef(ref);
    ***REMOVED***;

    root.validate = function (value /*, [schema], [options], callback */) ***REMOVED***

        var last = arguments[arguments.length - 1];
        var callback = typeof last === 'function' ? last : null;

        var count = arguments.length - (callback ? 1 : 0);
        if (count === 1) ***REMOVED***
            return any.validate(value, callback);
        ***REMOVED***

        var options = count === 3 ? arguments[2] : ***REMOVED******REMOVED***;
        var schema = root.compile(arguments[1]);

        return schema._validateWithOptions(value, options, callback);
    ***REMOVED***;

    root.describe = function () ***REMOVED***

        var schema = arguments.length ? root.compile(arguments[0]) : any;
        return schema.describe();
    ***REMOVED***;

    root.compile = function (schema) ***REMOVED***

        try ***REMOVED***
            return Cast.schema(schema);
        ***REMOVED***
        catch (err) ***REMOVED***
            if (err.hasOwnProperty('path')) ***REMOVED***
                err.message += '(' + err.path + ')';
            ***REMOVED***
            throw err;
        ***REMOVED***
    ***REMOVED***;

    root.assert = function (value, schema, message) ***REMOVED***

        root.attempt(value, schema, message);
    ***REMOVED***;

    root.attempt = function (value, schema, message) ***REMOVED***

        var result = root.validate(value, schema);
        var error = result.error;
        if (error) ***REMOVED***
            if (!message) ***REMOVED***
                error.message = error.annotate();
                throw error;
            ***REMOVED***

            if (!(message instanceof Error)) ***REMOVED***
                error.message = message + ' ' + error.annotate();
                throw error;
            ***REMOVED***

            throw message;
        ***REMOVED***

        return result.value;
    ***REMOVED***;

    return root;
***REMOVED***;


module.exports = internals.root();
