// send/index.js
/*!
 * send
 * Copyright(c) 2012 TJ Holowaychuk
 * Copyright(c) 2014-2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var createError = require('http-errors')
var debug = require('debug')('send')
var deprecate = require('depd')('send')
var destroy = require('destroy')
var encodeUrl = require('encodeurl')
var escapeHtml = require('escape-html')
var etag = require('etag')
var EventEmitter = require('events').EventEmitter
var fresh = require('fresh')
var fs = require('fs')
var mime = require('mime')
var ms = require('ms')
var onFinished = require('on-finished')
var parseRange = require('range-parser')
var path = require('path')
var statuses = require('statuses')
var Stream = require('stream')
var util = require('util')

/**
 * Path function references.
 * @private
 */

var extname = path.extname
var join = path.join
var normalize = path.normalize
var resolve = path.resolve
var sep = path.sep

/**
 * Regular expression for identifying a bytes Range header.
 * @private
 */

var BYTES_RANGE_REGEXP = /^ *bytes=/

/**
 * Maximum value allowed for the max age.
 * @private
 */

var MAX_MAXAGE = 60 * 60 * 24 * 365 * 1000 // 1 year

/**
 * Regular expression to match a path with a directory up component.
 * @private
 */

var UP_PATH_REGEXP = /(?:^|[\\\/])\.\.(?:[\\\/]|$)/

/**
 * Module exports.
 * @public
 */

module.exports = send
module.exports.mime = mime

/**
 * Shim EventEmitter.listenerCount for node.js < 0.10
 */

/* istanbul ignore next */
var listenerCount = EventEmitter.listenerCount ||
  function (emitter, type) ***REMOVED*** return emitter.listeners(type).length ***REMOVED***

/**
 * Return a `SendStream` for `req` and `path`.
 *
 * @param ***REMOVED***object***REMOVED*** req
 * @param ***REMOVED***string***REMOVED*** path
 * @param ***REMOVED***object***REMOVED*** [options]
 * @return ***REMOVED***SendStream***REMOVED***
 * @public
 */

function send (req, path, options) ***REMOVED***
  return new SendStream(req, path, options)
***REMOVED***

/**
 * Initialize a `SendStream` with the given `path`.
 *
 * @param ***REMOVED***Request***REMOVED*** req
 * @param ***REMOVED***String***REMOVED*** path
 * @param ***REMOVED***object***REMOVED*** [options]
 * @private
 */

function SendStream (req, path, options) ***REMOVED***
  Stream.call(this)

  var opts = options || ***REMOVED******REMOVED***

  this.options = opts
  this.path = path
  this.req = req

  this._acceptRanges = opts.acceptRanges !== undefined
    ? Boolean(opts.acceptRanges)
    : true

  this._cacheControl = opts.cacheControl !== undefined
    ? Boolean(opts.cacheControl)
    : true

  this._etag = opts.etag !== undefined
    ? Boolean(opts.etag)
    : true

  this._dotfiles = opts.dotfiles !== undefined
    ? opts.dotfiles
    : 'ignore'

  if (this._dotfiles !== 'ignore' && this._dotfiles !== 'allow' && this._dotfiles !== 'deny') ***REMOVED***
    throw new TypeError('dotfiles option must be "allow", "deny", or "ignore"')
  ***REMOVED***

  this._hidden = Boolean(opts.hidden)

  if (opts.hidden !== undefined) ***REMOVED***
    deprecate('hidden: use dotfiles: \'' + (this._hidden ? 'allow' : 'ignore') + '\' instead')
  ***REMOVED***

  // legacy support
  if (opts.dotfiles === undefined) ***REMOVED***
    this._dotfiles = undefined
  ***REMOVED***

  this._extensions = opts.extensions !== undefined
    ? normalizeList(opts.extensions, 'extensions option')
    : []

  this._index = opts.index !== undefined
    ? normalizeList(opts.index, 'index option')
    : ['index.html']

  this._lastModified = opts.lastModified !== undefined
    ? Boolean(opts.lastModified)
    : true

  this._maxage = opts.maxAge || opts.maxage
  this._maxage = typeof this._maxage === 'string'
    ? ms(this._maxage)
    : Number(this._maxage)
  this._maxage = !isNaN(this._maxage)
    ? Math.min(Math.max(0, this._maxage), MAX_MAXAGE)
    : 0

  this._root = opts.root
    ? resolve(opts.root)
    : null

  if (!this._root && opts.from) ***REMOVED***
    this.from(opts.from)
  ***REMOVED***
***REMOVED***

/**
 * Inherits from `Stream`.
 */

util.inherits(SendStream, Stream)

/**
 * Enable or disable etag generation.
 *
 * @param ***REMOVED***Boolean***REMOVED*** val
 * @return ***REMOVED***SendStream***REMOVED***
 * @api public
 */

SendStream.prototype.etag = deprecate.function(function etag (val) ***REMOVED***
  this._etag = Boolean(val)
  debug('etag %s', this._etag)
  return this
***REMOVED***, 'send.etag: pass etag as option')

/**
 * Enable or disable "hidden" (dot) files.
 *
 * @param ***REMOVED***Boolean***REMOVED*** path
 * @return ***REMOVED***SendStream***REMOVED***
 * @api public
 */

SendStream.prototype.hidden = deprecate.function(function hidden (val) ***REMOVED***
  this._hidden = Boolean(val)
  this._dotfiles = undefined
  debug('hidden %s', this._hidden)
  return this
***REMOVED***, 'send.hidden: use dotfiles option')

/**
 * Set index `paths`, set to a falsy
 * value to disable index support.
 *
 * @param ***REMOVED***String|Boolean|Array***REMOVED*** paths
 * @return ***REMOVED***SendStream***REMOVED***
 * @api public
 */

SendStream.prototype.index = deprecate.function(function index (paths) ***REMOVED***
  var index = !paths ? [] : normalizeList(paths, 'paths argument')
  debug('index %o', paths)
  this._index = index
  return this
***REMOVED***, 'send.index: pass index as option')

/**
 * Set root `path`.
 *
 * @param ***REMOVED***String***REMOVED*** path
 * @return ***REMOVED***SendStream***REMOVED***
 * @api public
 */

SendStream.prototype.root = function root (path) ***REMOVED***
  this._root = resolve(String(path))
  debug('root %s', this._root)
  return this
***REMOVED***

SendStream.prototype.from = deprecate.function(SendStream.prototype.root,
  'send.from: pass root as option')

SendStream.prototype.root = deprecate.function(SendStream.prototype.root,
  'send.root: pass root as option')

/**
 * Set max-age to `maxAge`.
 *
 * @param ***REMOVED***Number***REMOVED*** maxAge
 * @return ***REMOVED***SendStream***REMOVED***
 * @api public
 */

SendStream.prototype.maxage = deprecate.function(function maxage (maxAge) ***REMOVED***
  this._maxage = typeof maxAge === 'string'
    ? ms(maxAge)
    : Number(maxAge)
  this._maxage = !isNaN(this._maxage)
    ? Math.min(Math.max(0, this._maxage), MAX_MAXAGE)
    : 0
  debug('max-age %d', this._maxage)
  return this
***REMOVED***, 'send.maxage: pass maxAge as option')

/**
 * Emit error with `status`.
 *
 * @param ***REMOVED***number***REMOVED*** status
 * @param ***REMOVED***Error***REMOVED*** [error]
 * @private
 */

SendStream.prototype.error = function error (status, error) ***REMOVED***
  // emit if listeners instead of responding
  if (listenerCount(this, 'error') !== 0) ***REMOVED***
    return this.emit('error', createError(error, status, ***REMOVED***
      expose: false
    ***REMOVED***))
  ***REMOVED***

  var res = this.res
  var msg = statuses[status]

  // clear existing headers
  clearHeaders(res)

  // add error headers
  if (error && error.headers) ***REMOVED***
    setHeaders(res, error.headers)
  ***REMOVED***

  // send basic response
  res.statusCode = status
  res.setHeader('Content-Type', 'text/plain; charset=UTF-8')
  res.setHeader('Content-Length', Buffer.byteLength(msg))
  res.setHeader('X-Content-Type-Options', 'nosniff')
  res.end(msg)
***REMOVED***

/**
 * Check if the pathname ends with "/".
 *
 * @return ***REMOVED***Boolean***REMOVED***
 * @api private
 */

SendStream.prototype.hasTrailingSlash = function hasTrailingSlash () ***REMOVED***
  return this.path[this.path.length - 1] === '/'
***REMOVED***

/**
 * Check if this is a conditional GET request.
 *
 * @return ***REMOVED***Boolean***REMOVED***
 * @api private
 */

SendStream.prototype.isConditionalGET = function isConditionalGET () ***REMOVED***
  return this.req.headers['if-none-match'] ||
    this.req.headers['if-modified-since']
***REMOVED***

/**
 * Strip content-* header fields.
 *
 * @private
 */

SendStream.prototype.removeContentHeaderFields = function removeContentHeaderFields () ***REMOVED***
  var res = this.res
  var headers = Object.keys(res._headers || ***REMOVED******REMOVED***)

  for (var i = 0; i < headers.length; i++) ***REMOVED***
    var header = headers[i]
    if (header.substr(0, 8) === 'content-' && header !== 'content-location') ***REMOVED***
      res.removeHeader(header)
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * Respond with 304 not modified.
 *
 * @api private
 */

SendStream.prototype.notModified = function notModified () ***REMOVED***
  var res = this.res
  debug('not modified')
  this.removeContentHeaderFields()
  res.statusCode = 304
  res.end()
***REMOVED***

/**
 * Raise error that headers already sent.
 *
 * @api private
 */

SendStream.prototype.headersAlreadySent = function headersAlreadySent () ***REMOVED***
  var err = new Error('Can\'t set headers after they are sent.')
  debug('headers already sent')
  this.error(500, err)
***REMOVED***

/**
 * Check if the request is cacheable, aka
 * responded with 2xx or 304 (see RFC 2616 section 14.2***REMOVED***5,6***REMOVED***).
 *
 * @return ***REMOVED***Boolean***REMOVED***
 * @api private
 */

SendStream.prototype.isCachable = function isCachable () ***REMOVED***
  var statusCode = this.res.statusCode
  return (statusCode >= 200 && statusCode < 300) ||
    statusCode === 304
***REMOVED***

/**
 * Handle stat() error.
 *
 * @param ***REMOVED***Error***REMOVED*** error
 * @private
 */

SendStream.prototype.onStatError = function onStatError (error) ***REMOVED***
  switch (error.code) ***REMOVED***
    case 'ENAMETOOLONG':
    case 'ENOENT':
    case 'ENOTDIR':
      this.error(404, error)
      break
    default:
      this.error(500, error)
      break
  ***REMOVED***
***REMOVED***

/**
 * Check if the cache is fresh.
 *
 * @return ***REMOVED***Boolean***REMOVED***
 * @api private
 */

SendStream.prototype.isFresh = function isFresh () ***REMOVED***
  return fresh(this.req.headers, this.res._headers)
***REMOVED***

/**
 * Check if the range is fresh.
 *
 * @return ***REMOVED***Boolean***REMOVED***
 * @api private
 */

SendStream.prototype.isRangeFresh = function isRangeFresh () ***REMOVED***
  var ifRange = this.req.headers['if-range']

  if (!ifRange) ***REMOVED***
    return true
  ***REMOVED***

  return ~ifRange.indexOf('"')
    ? ~ifRange.indexOf(this.res._headers['etag'])
    : Date.parse(this.res._headers['last-modified']) <= Date.parse(ifRange)
***REMOVED***

/**
 * Redirect to path.
 *
 * @param ***REMOVED***string***REMOVED*** path
 * @private
 */

SendStream.prototype.redirect = function redirect (path) ***REMOVED***
  if (listenerCount(this, 'directory') !== 0) ***REMOVED***
    this.emit('directory')
    return
  ***REMOVED***

  if (this.hasTrailingSlash()) ***REMOVED***
    this.error(403)
    return
  ***REMOVED***

  var loc = encodeUrl(collapseLeadingSlashes(path + '/'))
  var msg = 'Redirecting to <a href="' + escapeHtml(loc) + '">' + escapeHtml(loc) + '</a>\n'
  var res = this.res

  // redirect
  res.statusCode = 301
  res.setHeader('Content-Type', 'text/html; charset=UTF-8')
  res.setHeader('Content-Length', Buffer.byteLength(msg))
  res.setHeader('X-Content-Type-Options', 'nosniff')
  res.setHeader('Location', loc)
  res.end(msg)
***REMOVED***

/**
 * Pipe to `res.
 *
 * @param ***REMOVED***Stream***REMOVED*** res
 * @return ***REMOVED***Stream***REMOVED*** res
 * @api public
 */

SendStream.prototype.pipe = function pipe (res) ***REMOVED***
  // root path
  var root = this._root

  // references
  this.res = res

  // decode the path
  var path = decode(this.path)
  if (path === -1) ***REMOVED***
    this.error(400)
    return res
  ***REMOVED***

  // null byte(s)
  if (~path.indexOf('\0')) ***REMOVED***
    this.error(400)
    return res
  ***REMOVED***

  var parts
  if (root !== null) ***REMOVED***
    // malicious path
    if (UP_PATH_REGEXP.test(normalize('.' + sep + path))) ***REMOVED***
      debug('malicious path "%s"', path)
      this.error(403)
      return res
    ***REMOVED***

    // join / normalize from optional root dir
    path = normalize(join(root, path))
    root = normalize(root + sep)

    // explode path parts
    parts = path.substr(root.length).split(sep)
  ***REMOVED*** else ***REMOVED***
    // ".." is malicious without "root"
    if (UP_PATH_REGEXP.test(path)) ***REMOVED***
      debug('malicious path "%s"', path)
      this.error(403)
      return res
    ***REMOVED***

    // explode path parts
    parts = normalize(path).split(sep)

    // resolve the path
    path = resolve(path)
  ***REMOVED***

  // dotfile handling
  if (containsDotFile(parts)) ***REMOVED***
    var access = this._dotfiles

    // legacy support
    if (access === undefined) ***REMOVED***
      access = parts[parts.length - 1][0] === '.'
        ? (this._hidden ? 'allow' : 'ignore')
        : 'allow'
    ***REMOVED***

    debug('%s dotfile "%s"', access, path)
    switch (access) ***REMOVED***
      case 'allow':
        break
      case 'deny':
        this.error(403)
        return res
      case 'ignore':
      default:
        this.error(404)
        return res
    ***REMOVED***
  ***REMOVED***

  // index file support
  if (this._index.length && this.path[this.path.length - 1] === '/') ***REMOVED***
    this.sendIndex(path)
    return res
  ***REMOVED***

  this.sendFile(path)
  return res
***REMOVED***

/**
 * Transfer `path`.
 *
 * @param ***REMOVED***String***REMOVED*** path
 * @api public
 */

SendStream.prototype.send = function send (path, stat) ***REMOVED***
  var len = stat.size
  var options = this.options
  var opts = ***REMOVED******REMOVED***
  var res = this.res
  var req = this.req
  var ranges = req.headers.range
  var offset = options.start || 0

  if (res._header) ***REMOVED***
    // impossible to send now
    this.headersAlreadySent()
    return
  ***REMOVED***

  debug('pipe "%s"', path)

  // set header fields
  this.setHeader(path, stat)

  // set content-type
  this.type(path)

  // conditional GET support
  if (this.isConditionalGET() && this.isCachable() && this.isFresh()) ***REMOVED***
    this.notModified()
    return
  ***REMOVED***

  // adjust len to start/end options
  len = Math.max(0, len - offset)
  if (options.end !== undefined) ***REMOVED***
    var bytes = options.end - offset + 1
    if (len > bytes) len = bytes
  ***REMOVED***

  // Range support
  if (this._acceptRanges && BYTES_RANGE_REGEXP.test(ranges)) ***REMOVED***
    // parse
    ranges = parseRange(len, ranges, ***REMOVED***
      combine: true
    ***REMOVED***)

    // If-Range support
    if (!this.isRangeFresh()) ***REMOVED***
      debug('range stale')
      ranges = -2
    ***REMOVED***

    // unsatisfiable
    if (ranges === -1) ***REMOVED***
      debug('range unsatisfiable')

      // Content-Range
      res.setHeader('Content-Range', contentRange('bytes', len))

      // 416 Requested Range Not Satisfiable
      return this.error(416, ***REMOVED***
        headers: ***REMOVED***'Content-Range': res.getHeader('Content-Range')***REMOVED***
      ***REMOVED***)
    ***REMOVED***

    // valid (syntactically invalid/multiple ranges are treated as a regular response)
    if (ranges !== -2 && ranges.length === 1) ***REMOVED***
      debug('range %j', ranges)

      // Content-Range
      res.statusCode = 206
      res.setHeader('Content-Range', contentRange('bytes', len, ranges[0]))

      // adjust for requested range
      offset += ranges[0].start
      len = ranges[0].end - ranges[0].start + 1
    ***REMOVED***
  ***REMOVED***

  // clone options
  for (var prop in options) ***REMOVED***
    opts[prop] = options[prop]
  ***REMOVED***

  // set read options
  opts.start = offset
  opts.end = Math.max(offset, offset + len - 1)

  // content-length
  res.setHeader('Content-Length', len)

  // HEAD support
  if (req.method === 'HEAD') ***REMOVED***
    res.end()
    return
  ***REMOVED***

  this.stream(path, opts)
***REMOVED***

/**
 * Transfer file for `path`.
 *
 * @param ***REMOVED***String***REMOVED*** path
 * @api private
 */
SendStream.prototype.sendFile = function sendFile (path) ***REMOVED***
  var i = 0
  var self = this

  debug('stat "%s"', path)
  fs.stat(path, function onstat (err, stat) ***REMOVED***
    if (err && err.code === 'ENOENT' && !extname(path) && path[path.length - 1] !== sep) ***REMOVED***
      // not found, check extensions
      return next(err)
    ***REMOVED***
    if (err) return self.onStatError(err)
    if (stat.isDirectory()) return self.redirect(self.path)
    self.emit('file', path, stat)
    self.send(path, stat)
  ***REMOVED***)

  function next (err) ***REMOVED***
    if (self._extensions.length <= i) ***REMOVED***
      return err
        ? self.onStatError(err)
        : self.error(404)
    ***REMOVED***

    var p = path + '.' + self._extensions[i++]

    debug('stat "%s"', p)
    fs.stat(p, function (err, stat) ***REMOVED***
      if (err) return next(err)
      if (stat.isDirectory()) return next()
      self.emit('file', p, stat)
      self.send(p, stat)
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

/**
 * Transfer index for `path`.
 *
 * @param ***REMOVED***String***REMOVED*** path
 * @api private
 */
SendStream.prototype.sendIndex = function sendIndex (path) ***REMOVED***
  var i = -1
  var self = this

  function next (err) ***REMOVED***
    if (++i >= self._index.length) ***REMOVED***
      if (err) return self.onStatError(err)
      return self.error(404)
    ***REMOVED***

    var p = join(path, self._index[i])

    debug('stat "%s"', p)
    fs.stat(p, function (err, stat) ***REMOVED***
      if (err) return next(err)
      if (stat.isDirectory()) return next()
      self.emit('file', p, stat)
      self.send(p, stat)
    ***REMOVED***)
  ***REMOVED***

  next()
***REMOVED***

/**
 * Stream `path` to the response.
 *
 * @param ***REMOVED***String***REMOVED*** path
 * @param ***REMOVED***Object***REMOVED*** options
 * @api private
 */

SendStream.prototype.stream = function stream (path, options) ***REMOVED***
  // TODO: this is all lame, refactor meeee
  var finished = false
  var self = this
  var res = this.res

  // pipe
  var stream = fs.createReadStream(path, options)
  this.emit('stream', stream)
  stream.pipe(res)

  // response finished, done with the fd
  onFinished(res, function onfinished () ***REMOVED***
    finished = true
    destroy(stream)
  ***REMOVED***)

  // error handling code-smell
  stream.on('error', function onerror (err) ***REMOVED***
    // request already finished
    if (finished) return

    // clean up stream
    finished = true
    destroy(stream)

    // error
    self.onStatError(err)
  ***REMOVED***)

  // end
  stream.on('end', function onend () ***REMOVED***
    self.emit('end')
  ***REMOVED***)
***REMOVED***

/**
 * Set content-type based on `path`
 * if it hasn't been explicitly set.
 *
 * @param ***REMOVED***String***REMOVED*** path
 * @api private
 */

SendStream.prototype.type = function type (path) ***REMOVED***
  var res = this.res

  if (res.getHeader('Content-Type')) return

  var type = mime.lookup(path)

  if (!type) ***REMOVED***
    debug('no content-type')
    return
  ***REMOVED***

  var charset = mime.charsets.lookup(type)

  debug('content-type %s', type)
  res.setHeader('Content-Type', type + (charset ? '; charset=' + charset : ''))
***REMOVED***

/**
 * Set response header fields, most
 * fields may be pre-defined.
 *
 * @param ***REMOVED***String***REMOVED*** path
 * @param ***REMOVED***Object***REMOVED*** stat
 * @api private
 */

SendStream.prototype.setHeader = function setHeader (path, stat) ***REMOVED***
  var res = this.res

  this.emit('headers', res, path, stat)

  if (this._acceptRanges && !res.getHeader('Accept-Ranges')) ***REMOVED***
    debug('accept ranges')
    res.setHeader('Accept-Ranges', 'bytes')
  ***REMOVED***

  if (this._cacheControl && !res.getHeader('Cache-Control')) ***REMOVED***
    var cacheControl = 'public, max-age=' + Math.floor(this._maxage / 1000)
    debug('cache-control %s', cacheControl)
    res.setHeader('Cache-Control', cacheControl)
  ***REMOVED***

  if (this._lastModified && !res.getHeader('Last-Modified')) ***REMOVED***
    var modified = stat.mtime.toUTCString()
    debug('modified %s', modified)
    res.setHeader('Last-Modified', modified)
  ***REMOVED***

  if (this._etag && !res.getHeader('ETag')) ***REMOVED***
    var val = etag(stat)
    debug('etag %s', val)
    res.setHeader('ETag', val)
  ***REMOVED***
***REMOVED***

/**
 * Clear all headers from a response.
 *
 * @param ***REMOVED***object***REMOVED*** res
 * @private
 */

function clearHeaders (res) ***REMOVED***
  res._headers = ***REMOVED******REMOVED***
  res._headerNames = ***REMOVED******REMOVED***
***REMOVED***

/**
 * Collapse all leading slashes into a single slash
 *
 * @param ***REMOVED***string***REMOVED*** str
 * @private
 */
function collapseLeadingSlashes (str) ***REMOVED***
  for (var i = 0; i < str.length; i++) ***REMOVED***
    if (str[i] !== '/') ***REMOVED***
      break
    ***REMOVED***
  ***REMOVED***

  return i > 1
    ? '/' + str.substr(i)
    : str
***REMOVED***

/**
 * Determine if path parts contain a dotfile.
 *
 * @api private
 */

function containsDotFile (parts) ***REMOVED***
  for (var i = 0; i < parts.length; i++) ***REMOVED***
    if (parts[i][0] === '.') ***REMOVED***
      return true
    ***REMOVED***
  ***REMOVED***

  return false
***REMOVED***

/**
 * Create a Content-Range header.
 *
 * @param ***REMOVED***string***REMOVED*** type
 * @param ***REMOVED***number***REMOVED*** size
 * @param ***REMOVED***array***REMOVED*** [range]
 */

function contentRange (type, size, range) ***REMOVED***
  return type + ' ' + (range ? range.start + '-' + range.end : '*') + '/' + size
***REMOVED***

/**
 * decodeURIComponent.
 *
 * Allows V8 to only deoptimize this fn instead of all
 * of send().
 *
 * @param ***REMOVED***String***REMOVED*** path
 * @api private
 */

function decode (path) ***REMOVED***
  try ***REMOVED***
    return decodeURIComponent(path)
  ***REMOVED*** catch (err) ***REMOVED***
    return -1
  ***REMOVED***
***REMOVED***

/**
 * Normalize the index option into an array.
 *
 * @param ***REMOVED***boolean|string|array***REMOVED*** val
 * @param ***REMOVED***string***REMOVED*** name
 * @private
 */

function normalizeList (val, name) ***REMOVED***
  var list = [].concat(val || [])

  for (var i = 0; i < list.length; i++) ***REMOVED***
    if (typeof list[i] !== 'string') ***REMOVED***
      throw new TypeError(name + ' must be array of strings or false')
    ***REMOVED***
  ***REMOVED***

  return list
***REMOVED***

/**
 * Set an object of headers on a response.
 *
 * @param ***REMOVED***object***REMOVED*** res
 * @param ***REMOVED***object***REMOVED*** headers
 * @private
 */

function setHeaders (res, headers) ***REMOVED***
  var keys = Object.keys(headers)

  for (var i = 0; i < keys.length; i++) ***REMOVED***
    var key = keys[i]
    res.setHeader(key, headers[key])
  ***REMOVED***
***REMOVED***
