// raw-body/index.js
/*!
 * raw-body
 * Copyright(c) 2013-2014 Jonathan Ong
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var bytes = require('bytes')
var iconv = require('iconv-lite')
var unpipe = require('unpipe')

/**
 * Module exports.
 * @public
 */

module.exports = getRawBody

/**
 * Module variables.
 * @private
 */

var iconvEncodingMessageRegExp = /^Encoding not recognized: /

/**
 * Get the decoder for a given encoding.
 *
 * @param ***REMOVED***string***REMOVED*** encoding
 * @private
 */

function getDecoder (encoding) ***REMOVED***
  if (!encoding) return null

  try ***REMOVED***
    return iconv.getDecoder(encoding)
  ***REMOVED*** catch (e) ***REMOVED***
    // error getting decoder
    if (!iconvEncodingMessageRegExp.test(e.message)) throw e

    // the encoding was not found
    throw createError(415, 'specified encoding unsupported', 'encoding.unsupported', ***REMOVED***
      encoding: encoding
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

/**
 * Get the raw body of a stream (typically HTTP).
 *
 * @param ***REMOVED***object***REMOVED*** stream
 * @param ***REMOVED***object|string|function***REMOVED*** [options]
 * @param ***REMOVED***function***REMOVED*** [callback]
 * @public
 */

function getRawBody (stream, options, callback) ***REMOVED***
  var done = callback
  var opts = options || ***REMOVED******REMOVED***

  if (options === true || typeof options === 'string') ***REMOVED***
    // short cut for encoding
    opts = ***REMOVED***
      encoding: options
    ***REMOVED***
  ***REMOVED***

  if (typeof options === 'function') ***REMOVED***
    done = options
    opts = ***REMOVED******REMOVED***
  ***REMOVED***

  // validate callback is a function, if provided
  if (done !== undefined && typeof done !== 'function') ***REMOVED***
    throw new TypeError('argument callback must be a function')
  ***REMOVED***

  // require the callback without promises
  if (!done && !global.Promise) ***REMOVED***
    throw new TypeError('argument callback is required')
  ***REMOVED***

  // get encoding
  var encoding = opts.encoding !== true
    ? opts.encoding
    : 'utf-8'

  // convert the limit to an integer
  var limit = bytes.parse(opts.limit)

  // convert the expected length to an integer
  var length = opts.length != null && !isNaN(opts.length)
    ? parseInt(opts.length, 10)
    : null

  if (done) ***REMOVED***
    // classic callback style
    return readStream(stream, encoding, length, limit, done)
  ***REMOVED***

  return new Promise(function executor (resolve, reject) ***REMOVED***
    readStream(stream, encoding, length, limit, function onRead (err, buf) ***REMOVED***
      if (err) return reject(err)
      resolve(buf)
    ***REMOVED***)
  ***REMOVED***)
***REMOVED***

/**
 * Halt a stream.
 *
 * @param ***REMOVED***Object***REMOVED*** stream
 * @private
 */

function halt (stream) ***REMOVED***
  // unpipe everything from the stream
  unpipe(stream)

  // pause stream
  if (typeof stream.pause === 'function') ***REMOVED***
    stream.pause()
  ***REMOVED***
***REMOVED***

/**
 * Make a serializable error object.
 *
 * To create serializable errors you must re-set message so
 * that it is enumerable and you must re configure the type
 * property so that is writable and enumerable.
 *
 * @param ***REMOVED***number***REMOVED*** status
 * @param ***REMOVED***string***REMOVED*** message
 * @param ***REMOVED***string***REMOVED*** type
 * @param ***REMOVED***object***REMOVED*** props
 * @private
 */

function createError (status, message, type, props) ***REMOVED***
  var error = new Error()

  // capture stack trace
  Error.captureStackTrace(error, createError)

  // set free-form properties
  for (var prop in props) ***REMOVED***
    error[prop] = props[prop]
  ***REMOVED***

  // set message
  error.message = message

  // set status
  error.status = status
  error.statusCode = status

  // set type
  Object.defineProperty(error, 'type', ***REMOVED***
    value: type,
    enumerable: true,
    writable: true,
    configurable: true
  ***REMOVED***)

  return error
***REMOVED***

/**
 * Read the data from the stream.
 *
 * @param ***REMOVED***object***REMOVED*** stream
 * @param ***REMOVED***string***REMOVED*** encoding
 * @param ***REMOVED***number***REMOVED*** length
 * @param ***REMOVED***number***REMOVED*** limit
 * @param ***REMOVED***function***REMOVED*** callback
 * @public
 */

function readStream (stream, encoding, length, limit, callback) ***REMOVED***
  var complete = false
  var sync = true

  // check the length and limit options.
  // note: we intentionally leave the stream paused,
  // so users should handle the stream themselves.
  if (limit !== null && length !== null && length > limit) ***REMOVED***
    return done(createError(413, 'request entity too large', 'entity.too.large', ***REMOVED***
      expected: length,
      length: length,
      limit: limit
    ***REMOVED***))
  ***REMOVED***

  // streams1: assert request encoding is buffer.
  // streams2+: assert the stream encoding is buffer.
  //   stream._decoder: streams1
  //   state.encoding: streams2
  //   state.decoder: streams2, specifically < 0.10.6
  var state = stream._readableState
  if (stream._decoder || (state && (state.encoding || state.decoder))) ***REMOVED***
    // developer error
    return done(createError(500, 'stream encoding should not be set', 'stream.encoding.set'))
  ***REMOVED***

  var received = 0
  var decoder

  try ***REMOVED***
    decoder = getDecoder(encoding)
  ***REMOVED*** catch (err) ***REMOVED***
    return done(err)
  ***REMOVED***

  var buffer = decoder
    ? ''
    : []

  // attach listeners
  stream.on('aborted', onAborted)
  stream.on('close', cleanup)
  stream.on('data', onData)
  stream.on('end', onEnd)
  stream.on('error', onEnd)

  // mark sync section complete
  sync = false

  function done () ***REMOVED***
    var args = new Array(arguments.length)

    // copy arguments
    for (var i = 0; i < args.length; i++) ***REMOVED***
      args[i] = arguments[i]
    ***REMOVED***

    // mark complete
    complete = true

    if (sync) ***REMOVED***
      process.nextTick(invokeCallback)
    ***REMOVED*** else ***REMOVED***
      invokeCallback()
    ***REMOVED***

    function invokeCallback () ***REMOVED***
      cleanup()

      if (args[0]) ***REMOVED***
        // halt the stream on error
        halt(stream)
      ***REMOVED***

      callback.apply(null, args)
    ***REMOVED***
  ***REMOVED***

  function onAborted () ***REMOVED***
    if (complete) return

    done(createError(400, 'request aborted', 'request.aborted', ***REMOVED***
      code: 'ECONNABORTED',
      expected: length,
      length: length,
      received: received
    ***REMOVED***))
  ***REMOVED***

  function onData (chunk) ***REMOVED***
    if (complete) return

    received += chunk.length
    decoder
      ? buffer += decoder.write(chunk)
      : buffer.push(chunk)

    if (limit !== null && received > limit) ***REMOVED***
      done(createError(413, 'request entity too large', 'entity.too.large', ***REMOVED***
        limit: limit,
        received: received
      ***REMOVED***))
    ***REMOVED***
  ***REMOVED***

  function onEnd (err) ***REMOVED***
    if (complete) return
    if (err) return done(err)

    if (length !== null && received !== length) ***REMOVED***
      done(createError(400, 'request size did not match content length', 'request.size.invalid', ***REMOVED***
        expected: length,
        length: length,
        received: received
      ***REMOVED***))
    ***REMOVED*** else ***REMOVED***
      var string = decoder
        ? buffer + (decoder.end() || '')
        : Buffer.concat(buffer)
      done(null, string)
    ***REMOVED***
  ***REMOVED***

  function cleanup () ***REMOVED***
    buffer = null

    stream.removeListener('aborted', onAborted)
    stream.removeListener('data', onData)
    stream.removeListener('end', onEnd)
    stream.removeListener('error', onEnd)
    stream.removeListener('close', cleanup)
  ***REMOVED***
***REMOVED***
