// mongodb-core/lib/connection/logger.js
"use strict";

var f = require('util').format
  , MongoError = require('../error');

// Filters for classes
var classFilters = ***REMOVED******REMOVED***;
var filteredClasses = ***REMOVED******REMOVED***;
var level = null;
// Save the process id
var pid = process.pid;
// current logger
var currentLogger = null;

/**
 * Creates a new Logger instance
 * @class
 * @param ***REMOVED***string***REMOVED*** className The Class name associated with the logging instance
 * @param ***REMOVED***object***REMOVED*** [options=null] Optional settings.
 * @param ***REMOVED***Function***REMOVED*** [options.logger=null] Custom logger function;
 * @param ***REMOVED***string***REMOVED*** [options.loggerLevel=error] Override default global log level.
 * @return ***REMOVED***Logger***REMOVED*** a Logger instance.
 */
var Logger = function(className, options) ***REMOVED***
  if(!(this instanceof Logger)) return new Logger(className, options);
  options = options || ***REMOVED******REMOVED***;

  // Current reference
  this.className = className;

  // Current logger
  if(options.logger) ***REMOVED***
    currentLogger = options.logger;
  ***REMOVED*** else if(currentLogger == null) ***REMOVED***
    currentLogger = console.log;
  ***REMOVED***

  // Set level of logging, default is error
  if(options.loggerLevel) ***REMOVED***
    level = options.loggerLevel || 'error';
  ***REMOVED***

  // Add all class names
  if(filteredClasses[this.className] == null) classFilters[this.className] =  true;
***REMOVED***

/**
 * Log a message at the debug level
 * @method
 * @param ***REMOVED***string***REMOVED*** message The message to log
 * @param ***REMOVED***object***REMOVED*** object additional meta data to log
 * @return ***REMOVED***null***REMOVED***
 */
Logger.prototype.debug = function(message, object) ***REMOVED***
  if(this.isDebug()
    && ((Object.keys(filteredClasses).length > 0 && filteredClasses[this.className])
      || (Object.keys(filteredClasses).length == 0 && classFilters[this.className]))) ***REMOVED***
    var dateTime = new Date().getTime();
    var msg = f("[%s-%s:%s] %s %s", 'DEBUG', this.className, pid, dateTime, message);
    var state = ***REMOVED***
      type: 'debug', message: message, className: this.className, pid: pid, date: dateTime
    ***REMOVED***;
    if(object) state.meta = object;
    currentLogger(msg, state);
  ***REMOVED***
***REMOVED***

/**
 * Log a message at the warn level
 * @method
 * @param ***REMOVED***string***REMOVED*** message The message to log
 * @param ***REMOVED***object***REMOVED*** object additional meta data to log
 * @return ***REMOVED***null***REMOVED***
 */
Logger.prototype.warn = function(message, object) ***REMOVED***
  if(this.isWarn()
    && ((Object.keys(filteredClasses).length > 0 && filteredClasses[this.className])
      || (Object.keys(filteredClasses).length == 0 && classFilters[this.className]))) ***REMOVED***
    var dateTime = new Date().getTime();
    var msg = f("[%s-%s:%s] %s %s", 'WARN', this.className, pid, dateTime, message);
    var state = ***REMOVED***
      type: 'warn', message: message, className: this.className, pid: pid, date: dateTime
    ***REMOVED***;
    if(object) state.meta = object;
    currentLogger(msg, state);
  ***REMOVED***
***REMOVED***,

/**
 * Log a message at the info level
 * @method
 * @param ***REMOVED***string***REMOVED*** message The message to log
 * @param ***REMOVED***object***REMOVED*** object additional meta data to log
 * @return ***REMOVED***null***REMOVED***
 */
Logger.prototype.info = function(message, object) ***REMOVED***
  if(this.isInfo()
    && ((Object.keys(filteredClasses).length > 0 && filteredClasses[this.className])
      || (Object.keys(filteredClasses).length == 0 && classFilters[this.className]))) ***REMOVED***
    var dateTime = new Date().getTime();
    var msg = f("[%s-%s:%s] %s %s", 'INFO', this.className, pid, dateTime, message);
    var state = ***REMOVED***
      type: 'info', message: message, className: this.className, pid: pid, date: dateTime
    ***REMOVED***;
    if(object) state.meta = object;
    currentLogger(msg, state);
  ***REMOVED***
***REMOVED***,

/**
 * Log a message at the error level
 * @method
 * @param ***REMOVED***string***REMOVED*** message The message to log
 * @param ***REMOVED***object***REMOVED*** object additional meta data to log
 * @return ***REMOVED***null***REMOVED***
 */
Logger.prototype.error = function(message, object) ***REMOVED***
  if(this.isError()
    && ((Object.keys(filteredClasses).length > 0 && filteredClasses[this.className])
      || (Object.keys(filteredClasses).length == 0 && classFilters[this.className]))) ***REMOVED***
    var dateTime = new Date().getTime();
    var msg = f("[%s-%s:%s] %s %s", 'ERROR', this.className, pid, dateTime, message);
    var state = ***REMOVED***
      type: 'error', message: message, className: this.className, pid: pid, date: dateTime
    ***REMOVED***;
    if(object) state.meta = object;
    currentLogger(msg, state);
  ***REMOVED***
***REMOVED***,

/**
 * Is the logger set at info level
 * @method
 * @return ***REMOVED***boolean***REMOVED***
 */
Logger.prototype.isInfo = function() ***REMOVED***
  return level == 'info' || level == 'debug';
***REMOVED***,

/**
 * Is the logger set at error level
 * @method
 * @return ***REMOVED***boolean***REMOVED***
 */
Logger.prototype.isError = function() ***REMOVED***
  return level == 'error' || level == 'info' || level == 'debug';
***REMOVED***,

/**
 * Is the logger set at error level
 * @method
 * @return ***REMOVED***boolean***REMOVED***
 */
Logger.prototype.isWarn = function() ***REMOVED***
  return level == 'error' || level == 'warn' || level == 'info' || level == 'debug';
***REMOVED***,

/**
 * Is the logger set at debug level
 * @method
 * @return ***REMOVED***boolean***REMOVED***
 */
Logger.prototype.isDebug = function() ***REMOVED***
  return level == 'debug';
***REMOVED***

/**
 * Resets the logger to default settings, error and no filtered classes
 * @method
 * @return ***REMOVED***null***REMOVED***
 */
Logger.reset = function() ***REMOVED***
  level = 'error';
  filteredClasses = ***REMOVED******REMOVED***;
***REMOVED***

/**
 * Get the current logger function
 * @method
 * @return ***REMOVED***function***REMOVED***
 */
Logger.currentLogger = function() ***REMOVED***
  return currentLogger;
***REMOVED***

/**
 * Set the current logger function
 * @method
 * @param ***REMOVED***function***REMOVED*** logger Logger function.
 * @return ***REMOVED***null***REMOVED***
 */
Logger.setCurrentLogger = function(logger) ***REMOVED***
  if(typeof logger != 'function') throw new MongoError("current logger must be a function");
  currentLogger = logger;
***REMOVED***

/**
 * Set what classes to log.
 * @method
 * @param ***REMOVED***string***REMOVED*** type The type of filter (currently only class)
 * @param ***REMOVED***string[]***REMOVED*** values The filters to apply
 * @return ***REMOVED***null***REMOVED***
 */
Logger.filter = function(type, values) ***REMOVED***
  if(type == 'class' && Array.isArray(values)) ***REMOVED***
    filteredClasses = ***REMOVED******REMOVED***;

    values.forEach(function(x) ***REMOVED***
      filteredClasses[x] = true;
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

/**
 * Set the current log level
 * @method
 * @param ***REMOVED***string***REMOVED*** level Set current log level (debug, info, error)
 * @return ***REMOVED***null***REMOVED***
 */
Logger.setLevel = function(_level) ***REMOVED***
  if(_level != 'info' && _level != 'error' && _level != 'debug' && _level != 'warn') ***REMOVED***
    throw new Error(f("%s is an illegal logging level", _level));
  ***REMOVED***

  level = _level;
***REMOVED***

module.exports = Logger;
