// moment/src/lib/format/format.js
import zeroFill from '../utils/zero-fill';

export var formattingTokens = /(\[[^\[]*\])|(\\)?([Hh]mm(ss)?|Mo|MM?M?M?|Do|DDDo|DD?D?D?|ddd?d?|do?|w[o|w]?|W[o|W]?|Qo?|YYYYYY|YYYYY|YYYY|YY|gg(ggg?)?|GG(GGG?)?|e|E|a|A|hh?|HH?|kk?|mm?|ss?|S***REMOVED***1,9***REMOVED***|x|X|zz?|ZZ?|.)/g;

var localFormattingTokens = /(\[[^\[]*\])|(\\)?(LTS|LT|LL?L?L?|l***REMOVED***1,4***REMOVED***)/g;

var formatFunctions = ***REMOVED******REMOVED***;

export var formatTokenFunctions = ***REMOVED******REMOVED***;

// token:    'M'
// padded:   ['MM', 2]
// ordinal:  'Mo'
// callback: function () ***REMOVED*** this.month() + 1 ***REMOVED***
export function addFormatToken (token, padded, ordinal, callback) ***REMOVED***
    var func = callback;
    if (typeof callback === 'string') ***REMOVED***
        func = function () ***REMOVED***
            return this[callback]();
        ***REMOVED***;
    ***REMOVED***
    if (token) ***REMOVED***
        formatTokenFunctions[token] = func;
    ***REMOVED***
    if (padded) ***REMOVED***
        formatTokenFunctions[padded[0]] = function () ***REMOVED***
            return zeroFill(func.apply(this, arguments), padded[1], padded[2]);
        ***REMOVED***;
    ***REMOVED***
    if (ordinal) ***REMOVED***
        formatTokenFunctions[ordinal] = function () ***REMOVED***
            return this.localeData().ordinal(func.apply(this, arguments), token);
        ***REMOVED***;
    ***REMOVED***
***REMOVED***

function removeFormattingTokens(input) ***REMOVED***
    if (input.match(/\[[\s\S]/)) ***REMOVED***
        return input.replace(/^\[|\]$/g, '');
    ***REMOVED***
    return input.replace(/\\/g, '');
***REMOVED***

function makeFormatFunction(format) ***REMOVED***
    var array = format.match(formattingTokens), i, length;

    for (i = 0, length = array.length; i < length; i++) ***REMOVED***
        if (formatTokenFunctions[array[i]]) ***REMOVED***
            array[i] = formatTokenFunctions[array[i]];
        ***REMOVED*** else ***REMOVED***
            array[i] = removeFormattingTokens(array[i]);
        ***REMOVED***
    ***REMOVED***

    return function (mom) ***REMOVED***
        var output = '', i;
        for (i = 0; i < length; i++) ***REMOVED***
            output += array[i] instanceof Function ? array[i].call(mom, format) : array[i];
        ***REMOVED***
        return output;
    ***REMOVED***;
***REMOVED***

// format date using native date object
export function formatMoment(m, format) ***REMOVED***
    if (!m.isValid()) ***REMOVED***
        return m.localeData().invalidDate();
    ***REMOVED***

    format = expandFormat(format, m.localeData());
    formatFunctions[format] = formatFunctions[format] || makeFormatFunction(format);

    return formatFunctions[format](m);
***REMOVED***

export function expandFormat(format, locale) ***REMOVED***
    var i = 5;

    function replaceLongDateFormatTokens(input) ***REMOVED***
        return locale.longDateFormat(input) || input;
    ***REMOVED***

    localFormattingTokens.lastIndex = 0;
    while (i >= 0 && localFormattingTokens.test(format)) ***REMOVED***
        format = format.replace(localFormattingTokens, replaceLongDateFormatTokens);
        localFormattingTokens.lastIndex = 0;
        i -= 1;
    ***REMOVED***

    return format;
***REMOVED***
