// debug/debug.js

/**
 * This is the common logic for both the Node.js and web browser
 * implementations of `debug()`.
 *
 * Expose `debug()` as the module.
 */

exports = module.exports = debug;
exports.coerce = coerce;
exports.disable = disable;
exports.enable = enable;
exports.enabled = enabled;
exports.humanize = require('ms');

/**
 * The currently active debug mode names, and names to skip.
 */

exports.names = [];
exports.skips = [];

/**
 * Map of special "%n" handling functions, for the debug "format" argument.
 *
 * Valid key names are a single, lowercased letter, i.e. "n".
 */

exports.formatters = ***REMOVED******REMOVED***;

/**
 * Previously assigned color.
 */

var prevColor = 0;

/**
 * Previous log timestamp.
 */

var prevTime;

/**
 * Select a color.
 *
 * @return ***REMOVED***Number***REMOVED***
 * @api private
 */

function selectColor() ***REMOVED***
  return exports.colors[prevColor++ % exports.colors.length];
***REMOVED***

/**
 * Create a debugger with the given `namespace`.
 *
 * @param ***REMOVED***String***REMOVED*** namespace
 * @return ***REMOVED***Function***REMOVED***
 * @api public
 */

function debug(namespace) ***REMOVED***

  // define the `disabled` version
  function disabled() ***REMOVED***
  ***REMOVED***
  disabled.enabled = false;

  // define the `enabled` version
  function enabled() ***REMOVED***

    var self = enabled;

    // set `diff` timestamp
    var curr = +new Date();
    var ms = curr - (prevTime || curr);
    self.diff = ms;
    self.prev = prevTime;
    self.curr = curr;
    prevTime = curr;

    // add the `color` if not set
    if (null == self.useColors) self.useColors = exports.useColors();
    if (null == self.color && self.useColors) self.color = selectColor();

    var args = Array.prototype.slice.call(arguments);

    args[0] = exports.coerce(args[0]);

    if ('string' !== typeof args[0]) ***REMOVED***
      // anything else let's inspect with %o
      args = ['%o'].concat(args);
    ***REMOVED***

    // apply any `formatters` transformations
    var index = 0;
    args[0] = args[0].replace(/%([a-z%])/g, function(match, format) ***REMOVED***
      // if we encounter an escaped % then don't increase the array index
      if (match === '%%') return match;
      index++;
      var formatter = exports.formatters[format];
      if ('function' === typeof formatter) ***REMOVED***
        var val = args[index];
        match = formatter.call(self, val);

        // now we need to remove `args[index]` since it's inlined in the `format`
        args.splice(index, 1);
        index--;
      ***REMOVED***
      return match;
    ***REMOVED***);

    if ('function' === typeof exports.formatArgs) ***REMOVED***
      args = exports.formatArgs.apply(self, args);
    ***REMOVED***
    var logFn = enabled.log || exports.log || console.log.bind(console);
    logFn.apply(self, args);
  ***REMOVED***
  enabled.enabled = true;

  var fn = exports.enabled(namespace) ? enabled : disabled;

  fn.namespace = namespace;

  return fn;
***REMOVED***

/**
 * Enables a debug mode by namespaces. This can include modes
 * separated by a colon and wildcards.
 *
 * @param ***REMOVED***String***REMOVED*** namespaces
 * @api public
 */

function enable(namespaces) ***REMOVED***
  exports.save(namespaces);

  var split = (namespaces || '').split(/[\s,]+/);
  var len = split.length;

  for (var i = 0; i < len; i++) ***REMOVED***
    if (!split[i]) continue; // ignore empty strings
    namespaces = split[i].replace(/\*/g, '.*?');
    if (namespaces[0] === '-') ***REMOVED***
      exports.skips.push(new RegExp('^' + namespaces.substr(1) + '$'));
    ***REMOVED*** else ***REMOVED***
      exports.names.push(new RegExp('^' + namespaces + '$'));
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * Disable debug output.
 *
 * @api public
 */

function disable() ***REMOVED***
  exports.enable('');
***REMOVED***

/**
 * Returns true if the given mode name is enabled, false otherwise.
 *
 * @param ***REMOVED***String***REMOVED*** name
 * @return ***REMOVED***Boolean***REMOVED***
 * @api public
 */

function enabled(name) ***REMOVED***
  var i, len;
  for (i = 0, len = exports.skips.length; i < len; i++) ***REMOVED***
    if (exports.skips[i].test(name)) ***REMOVED***
      return false;
    ***REMOVED***
  ***REMOVED***
  for (i = 0, len = exports.names.length; i < len; i++) ***REMOVED***
    if (exports.names[i].test(name)) ***REMOVED***
      return true;
    ***REMOVED***
  ***REMOVED***
  return false;
***REMOVED***

/**
 * Coerce `val`.
 *
 * @param ***REMOVED***Mixed***REMOVED*** val
 * @return ***REMOVED***Mixed***REMOVED***
 * @api private
 */

function coerce(val) ***REMOVED***
  if (val instanceof Error) return val.stack || val.message;
  return val;
***REMOVED***
