// random-bytes/index.js
/*!
 * random-bytes
 * Copyright(c) 2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var crypto = require('crypto')

/**
 * Module variables.
 * @private
 */

var generateAttempts = crypto.randomBytes === crypto.pseudoRandomBytes ? 1 : 3

/**
 * Module exports.
 * @public
 */

module.exports = randomBytes
module.exports.sync = randomBytesSync

/**
 * Generates strong pseudo-random bytes.
 *
 * @param ***REMOVED***number***REMOVED*** size
 * @param ***REMOVED***function***REMOVED*** [callback]
 * @return ***REMOVED***Promise***REMOVED***
 * @public
 */

function randomBytes(size, callback) ***REMOVED***
  // validate callback is a function, if provided
  if (callback !== undefined && typeof callback !== 'function') ***REMOVED***
    throw new TypeError('argument callback must be a function')
  ***REMOVED***

  // require the callback without promises
  if (!callback && !global.Promise) ***REMOVED***
    throw new TypeError('argument callback is required')
  ***REMOVED***

  if (callback) ***REMOVED***
    // classic callback style
    return generateRandomBytes(size, generateAttempts, callback)
  ***REMOVED***

  return new Promise(function executor(resolve, reject) ***REMOVED***
    generateRandomBytes(size, generateAttempts, function onRandomBytes(err, str) ***REMOVED***
      if (err) return reject(err)
      resolve(str)
    ***REMOVED***)
  ***REMOVED***)
***REMOVED***

/**
 * Generates strong pseudo-random bytes sync.
 *
 * @param ***REMOVED***number***REMOVED*** size
 * @return ***REMOVED***Buffer***REMOVED***
 * @public
 */

function randomBytesSync(size) ***REMOVED***
  var err = null

  for (var i = 0; i < generateAttempts; i++) ***REMOVED***
    try ***REMOVED***
      return crypto.randomBytes(size)
    ***REMOVED*** catch (e) ***REMOVED***
      err = e
    ***REMOVED***
  ***REMOVED***

  throw err
***REMOVED***

/**
 * Generates strong pseudo-random bytes.
 *
 * @param ***REMOVED***number***REMOVED*** size
 * @param ***REMOVED***number***REMOVED*** attempts
 * @param ***REMOVED***function***REMOVED*** callback
 * @private
 */

function generateRandomBytes(size, attempts, callback) ***REMOVED***
  crypto.randomBytes(size, function onRandomBytes(err, buf) ***REMOVED***
    if (!err) return callback(null, buf)
    if (!--attempts) return callback(err)
    setTimeout(generateRandomBytes.bind(null, size, attempts, callback), 10)
  ***REMOVED***)
***REMOVED***
