// mongodb/lib/gridfs-stream/upload.js
var core = require('mongodb-core');
var crypto = require('crypto');
var stream = require('stream');
var util = require('util');

var ERROR_NAMESPACE_NOT_FOUND = 26;

module.exports = GridFSBucketWriteStream;

/**
 * A writable stream that enables you to write buffers to GridFS.
 *
 * Do not instantiate this class directly. Use `openUploadStream()` instead.
 *
 * @class
 * @param ***REMOVED***GridFSBucket***REMOVED*** bucket Handle for this stream's corresponding bucket
 * @param ***REMOVED***string***REMOVED*** filename The value of the 'filename' key in the files doc
 * @param ***REMOVED***object***REMOVED*** [options=null] Optional settings.
 * @param ***REMOVED***string|number|object***REMOVED*** [options.id=null] Custom file id for the GridFS file.
 * @param ***REMOVED***number***REMOVED*** [options.chunkSizeBytes=null] The chunk size to use, in bytes
 * @param ***REMOVED***number***REMOVED*** [options.w=null] The write concern
 * @param ***REMOVED***number***REMOVED*** [options.wtimeout=null] The write concern timeout
 * @param ***REMOVED***number***REMOVED*** [options.j=null] The journal write concern
 * @fires GridFSBucketWriteStream#error
 * @fires GridFSBucketWriteStream#finish
 * @return ***REMOVED***GridFSBucketWriteStream***REMOVED*** a GridFSBucketWriteStream instance.
 */

function GridFSBucketWriteStream(bucket, filename, options) ***REMOVED***
  options = options || ***REMOVED******REMOVED***;
  this.bucket = bucket;
  this.chunks = bucket.s._chunksCollection;
  this.filename = filename;
  this.files = bucket.s._filesCollection;
  this.options = options;

  this.id = options.id ? options.id : core.BSON.ObjectId();
  this.chunkSizeBytes = this.options.chunkSizeBytes;
  this.bufToStore = new Buffer(this.chunkSizeBytes);
  this.length = 0;
  this.md5 = crypto.createHash('md5');
  this.n = 0;
  this.pos = 0;
  this.state = ***REMOVED***
    streamEnd: false,
    outstandingRequests: 0,
    errored: false,
    aborted: false,
    promiseLibrary: this.bucket.s.promiseLibrary
  ***REMOVED***;

  if (!this.bucket.s.calledOpenUploadStream) ***REMOVED***
    this.bucket.s.calledOpenUploadStream = true;

    var _this = this;
    checkIndexes(this, function() ***REMOVED***
      _this.bucket.s.checkedIndexes = true;
      _this.bucket.emit('index');
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

util.inherits(GridFSBucketWriteStream, stream.Writable);

/**
 * An error occurred
 *
 * @event GridFSBucketWriteStream#error
 * @type ***REMOVED***Error***REMOVED***
 */

/**
 * `end()` was called and the write stream successfully wrote the file
 * metadata and all the chunks to MongoDB.
 *
 * @event GridFSBucketWriteStream#finish
 * @type ***REMOVED***object***REMOVED***
 */

/**
 * Write a buffer to the stream.
 *
 * @method
 * @param ***REMOVED***Buffer***REMOVED*** chunk Buffer to write
 * @param ***REMOVED***String***REMOVED*** encoding Optional encoding for the buffer
 * @param ***REMOVED***Function***REMOVED*** callback Function to call when the chunk was added to the buffer, or if the entire chunk was persisted to MongoDB if this chunk caused a flush.
 * @return ***REMOVED***Boolean***REMOVED*** False if this write required flushing a chunk to MongoDB. True otherwise.
 */

GridFSBucketWriteStream.prototype.write = function(chunk, encoding, callback) ***REMOVED***
  var _this = this;
  return waitForIndexes(this, function() ***REMOVED***
    return doWrite(_this, chunk, encoding, callback);
  ***REMOVED***);
***REMOVED***;

/**
 * Places this write stream into an aborted state (all future writes fail)
 * and deletes all chunks that have already been written.
 *
 * @method
 * @param ***REMOVED***GridFSBucket~errorCallback***REMOVED*** callback called when chunks are successfully removed or error occurred
 * @return ***REMOVED***Promise***REMOVED*** if no callback specified
 */

GridFSBucketWriteStream.prototype.abort = function(callback) ***REMOVED***
  if (this.state.streamEnd) ***REMOVED***
    var error = new Error('Cannot abort a stream that has already completed');
    if (typeof callback == 'function') ***REMOVED***
      return callback(error);
    ***REMOVED***
    return this.state.promiseLibrary.reject(error);
  ***REMOVED***
  if (this.state.aborted) ***REMOVED***
    error = new Error('Cannot call abort() on a stream twice');
    if (typeof callback == 'function') ***REMOVED***
      return callback(error);
    ***REMOVED***
    return this.state.promiseLibrary.reject(error);
  ***REMOVED***
  this.state.aborted = true;
  this.chunks.deleteMany(***REMOVED*** files_id: this.id ***REMOVED***, function(error) ***REMOVED***
    if(typeof callback == 'function') callback(error);
  ***REMOVED***);
***REMOVED***;

/**
 * Tells the stream that no more data will be coming in. The stream will
 * persist the remaining data to MongoDB, write the files document, and
 * then emit a 'finish' event.
 *
 * @method
 * @param ***REMOVED***Buffer***REMOVED*** chunk Buffer to write
 * @param ***REMOVED***String***REMOVED*** encoding Optional encoding for the buffer
 * @param ***REMOVED***Function***REMOVED*** callback Function to call when all files and chunks have been persisted to MongoDB
 */

GridFSBucketWriteStream.prototype.end = function(chunk, encoding, callback) ***REMOVED***
  var _this = this;
  if(typeof chunk == 'function') ***REMOVED***
    callback = chunk, chunk = null, encoding = null;
  ***REMOVED*** else if(typeof encoding == 'function') ***REMOVED***
    callback = encoding, encoding = null;
  ***REMOVED***

  if (checkAborted(this, callback)) ***REMOVED***
    return;
  ***REMOVED***
  this.state.streamEnd = true;

  if (callback) ***REMOVED***
    this.once('finish', function(result) ***REMOVED***
      callback(null, result);
    ***REMOVED***);
  ***REMOVED***

  if (!chunk) ***REMOVED***
    waitForIndexes(this, function() ***REMOVED***
      writeRemnant(_this);
    ***REMOVED***);
    return;
  ***REMOVED***

  this.write(chunk, encoding, function() ***REMOVED***
    writeRemnant(_this);
  ***REMOVED***);
***REMOVED***;

/**
 * @ignore
 */

function __handleError(_this, error, callback) ***REMOVED***
  if (_this.state.errored) ***REMOVED***
    return;
  ***REMOVED***
  _this.state.errored = true;
  if (callback) ***REMOVED***
    return callback(error);
  ***REMOVED***
  _this.emit('error', error);
***REMOVED***

/**
 * @ignore
 */

function createChunkDoc(filesId, n, data) ***REMOVED***
  return ***REMOVED***
    _id: core.BSON.ObjectId(),
    files_id: filesId,
    n: n,
    data: data
  ***REMOVED***;
***REMOVED***

/**
 * @ignore
 */

function checkChunksIndex(_this, callback) ***REMOVED***
  _this.chunks.listIndexes().toArray(function(error, indexes) ***REMOVED***
    if (error) ***REMOVED***
      // Collection doesn't exist so create index
      if (error.code === ERROR_NAMESPACE_NOT_FOUND) ***REMOVED***
        var index = ***REMOVED*** files_id: 1, n: 1 ***REMOVED***;
        _this.chunks.createIndex(index, ***REMOVED*** background: false, unique: true ***REMOVED***, function(error) ***REMOVED***
          if (error) ***REMOVED***
            return callback(error);
          ***REMOVED***

          callback();
        ***REMOVED***);
        return;
      ***REMOVED***
      return callback(error);
    ***REMOVED***

    var hasChunksIndex = false;
    indexes.forEach(function(index) ***REMOVED***
      if (index.key) ***REMOVED***
        var keys = Object.keys(index.key);
        if (keys.length === 2 && index.key.files_id === 1 &&
            index.key.n === 1) ***REMOVED***
          hasChunksIndex = true;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***);

    if (hasChunksIndex) ***REMOVED***
      callback();
    ***REMOVED*** else ***REMOVED***
      index = ***REMOVED*** files_id: 1, n: 1 ***REMOVED***;
      var indexOptions = getWriteOptions(_this);

      indexOptions.background = false;
      indexOptions.unique = true;

      _this.chunks.createIndex(index, indexOptions, function(error) ***REMOVED***
        if (error) ***REMOVED***
          return callback(error);
        ***REMOVED***

        callback();
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

/**
 * @ignore
 */

function checkDone(_this, callback) ***REMOVED***
  if (_this.state.streamEnd &&
      _this.state.outstandingRequests === 0 &&
      !_this.state.errored) ***REMOVED***
    var filesDoc = createFilesDoc(_this.id, _this.length, _this.chunkSizeBytes,
      _this.md5.digest('hex'), _this.filename, _this.options.contentType,
      _this.options.aliases, _this.options.metadata);

    if (checkAborted(_this, callback)) ***REMOVED***
      return false;
    ***REMOVED***

    _this.files.insert(filesDoc, getWriteOptions(_this), function(error) ***REMOVED***
      if (error) ***REMOVED***
        return __handleError(_this, error, callback);
      ***REMOVED***
      _this.emit('finish', filesDoc);
    ***REMOVED***);

    return true;
  ***REMOVED***

  return false;
***REMOVED***

/**
 * @ignore
 */

function checkIndexes(_this, callback) ***REMOVED***
  _this.files.findOne(***REMOVED******REMOVED***, ***REMOVED*** _id: 1 ***REMOVED***, function(error, doc) ***REMOVED***
    if (error) ***REMOVED***
      return callback(error);
    ***REMOVED***
    if (doc) ***REMOVED***
      return callback();
    ***REMOVED***

    _this.files.listIndexes().toArray(function(error, indexes) ***REMOVED***
      if (error) ***REMOVED***
        // Collection doesn't exist so create index
        if (error.code === ERROR_NAMESPACE_NOT_FOUND) ***REMOVED***
          var index = ***REMOVED*** filename: 1, uploadDate: 1 ***REMOVED***;
          _this.files.createIndex(index, ***REMOVED*** background: false ***REMOVED***, function(error) ***REMOVED***
            if (error) ***REMOVED***
              return callback(error);
            ***REMOVED***

            checkChunksIndex(_this, callback);
          ***REMOVED***);
          return;
        ***REMOVED***
        return callback(error);
      ***REMOVED***

      var hasFileIndex = false;
      indexes.forEach(function(index) ***REMOVED***
        var keys = Object.keys(index.key);
        if (keys.length === 2 && index.key.filename === 1 &&
            index.key.uploadDate === 1) ***REMOVED***
          hasFileIndex = true;
        ***REMOVED***
      ***REMOVED***);

      if (hasFileIndex) ***REMOVED***
        checkChunksIndex(_this, callback);
      ***REMOVED*** else ***REMOVED***
        index = ***REMOVED*** filename: 1, uploadDate: 1 ***REMOVED***;

        var indexOptions = getWriteOptions(_this);

        indexOptions.background = false;

        _this.files.createIndex(index, indexOptions, function(error) ***REMOVED***
          if (error) ***REMOVED***
            return callback(error);
          ***REMOVED***

          checkChunksIndex(_this, callback);
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

/**
 * @ignore
 */

function createFilesDoc(_id, length, chunkSize, md5, filename, contentType,
  aliases, metadata) ***REMOVED***
  var ret = ***REMOVED***
    _id: _id,
    length: length,
    chunkSize: chunkSize,
    uploadDate: new Date(),
    md5: md5,
    filename: filename
  ***REMOVED***;

  if (contentType) ***REMOVED***
    ret.contentType = contentType;
  ***REMOVED***

  if (aliases) ***REMOVED***
    ret.aliases = aliases;
  ***REMOVED***

  if (metadata) ***REMOVED***
    ret.metadata = metadata;
  ***REMOVED***

  return ret;
***REMOVED***

/**
 * @ignore
 */

function doWrite(_this, chunk, encoding, callback) ***REMOVED***
  if (checkAborted(_this, callback)) ***REMOVED***
    return false;
  ***REMOVED***

  var inputBuf = (Buffer.isBuffer(chunk)) ?
    chunk : new Buffer(chunk, encoding);

  _this.length += inputBuf.length;

  // Input is small enough to fit in our buffer
  if (_this.pos + inputBuf.length < _this.chunkSizeBytes) ***REMOVED***
    inputBuf.copy(_this.bufToStore, _this.pos);
    _this.pos += inputBuf.length;

    callback && callback();

    // Note that we reverse the typical semantics of write's return value
    // to be compatible with node's `.pipe()` function.
    // True means client can keep writing.
    return true;
  ***REMOVED***

  // Otherwise, buffer is too big for current chunk, so we need to flush
  // to MongoDB.
  var inputBufRemaining = inputBuf.length;
  var spaceRemaining = _this.chunkSizeBytes - _this.pos;
  var numToCopy = Math.min(spaceRemaining, inputBuf.length);
  var outstandingRequests = 0;
  while (inputBufRemaining > 0) ***REMOVED***
    var inputBufPos = inputBuf.length - inputBufRemaining;
    inputBuf.copy(_this.bufToStore, _this.pos,
      inputBufPos, inputBufPos + numToCopy);
    _this.pos += numToCopy;
    spaceRemaining -= numToCopy;
    if (spaceRemaining === 0) ***REMOVED***
      _this.md5.update(_this.bufToStore);
      var doc = createChunkDoc(_this.id, _this.n, _this.bufToStore);
      ++_this.state.outstandingRequests;
      ++outstandingRequests;

      if (checkAborted(_this, callback)) ***REMOVED***
        return false;
      ***REMOVED***

      _this.chunks.insert(doc, getWriteOptions(_this), function(error) ***REMOVED***
        if (error) ***REMOVED***
          return __handleError(_this, error);
        ***REMOVED***
        --_this.state.outstandingRequests;
        --outstandingRequests;
        if (!outstandingRequests) ***REMOVED***
          _this.emit('drain', doc);
          callback && callback();
          checkDone(_this);
        ***REMOVED***
      ***REMOVED***);

      spaceRemaining = _this.chunkSizeBytes;
      _this.pos = 0;
      ++_this.n;
    ***REMOVED***
    inputBufRemaining -= numToCopy;
    numToCopy = Math.min(spaceRemaining, inputBufRemaining);
  ***REMOVED***

  // Note that we reverse the typical semantics of write's return value
  // to be compatible with node's `.pipe()` function.
  // False means the client should wait for the 'drain' event.
  return false;
***REMOVED***

/**
 * @ignore
 */

function getWriteOptions(_this) ***REMOVED***
  var obj = ***REMOVED******REMOVED***;
  if (_this.options.writeConcern) ***REMOVED***
    obj.w = _this.options.writeConcern.w;
    obj.wtimeout = _this.options.writeConcern.wtimeout;
    obj.j = _this.options.writeConcern.j;
  ***REMOVED***
  return obj;
***REMOVED***

/**
 * @ignore
 */

function waitForIndexes(_this, callback) ***REMOVED***
  if (_this.bucket.s.checkedIndexes) ***REMOVED***
    return callback(false);
  ***REMOVED***

  _this.bucket.once('index', function() ***REMOVED***
    callback(true);
  ***REMOVED***);

  return true;
***REMOVED***

/**
 * @ignore
 */

function writeRemnant(_this, callback) ***REMOVED***
  // Buffer is empty, so don't bother to insert
  if (_this.pos === 0) ***REMOVED***
    return checkDone(_this, callback);
  ***REMOVED***

  ++_this.state.outstandingRequests;

  // Create a new buffer to make sure the buffer isn't bigger than it needs
  // to be.
  var remnant = new Buffer(_this.pos);
  _this.bufToStore.copy(remnant, 0, 0, _this.pos);
  _this.md5.update(remnant);
  var doc = createChunkDoc(_this.id, _this.n, remnant);

  // If the stream was aborted, do not write remnant
  if (checkAborted(_this, callback)) ***REMOVED***
    return false;
  ***REMOVED***

  _this.chunks.insert(doc, getWriteOptions(_this), function(error) ***REMOVED***
    if (error) ***REMOVED***
      return __handleError(_this, error);
    ***REMOVED***
    --_this.state.outstandingRequests;
    checkDone(_this);
  ***REMOVED***);
***REMOVED***

/**
 * @ignore
 */

function checkAborted(_this, callback) ***REMOVED***
  if (_this.state.aborted) ***REMOVED***
    if(typeof callback == 'function') ***REMOVED***
      callback(new Error('this stream has been aborted'));
    ***REMOVED***
    return true;
  ***REMOVED***
  return false;
***REMOVED***
