// jws/lib/sign-stream.js
/*global module*/
var base64url = require('base64url');
var DataStream = require('./data-stream');
var jwa = require('jwa');
var Stream = require('stream');
var toString = require('./tostring');
var util = require('util');

function jwsSecuredInput(header, payload, encoding) ***REMOVED***
  encoding = encoding || 'utf8';
  var encodedHeader = base64url(toString(header), 'binary');
  var encodedPayload = base64url(toString(payload), encoding);
  return util.format('%s.%s', encodedHeader, encodedPayload);
***REMOVED***

function jwsSign(opts) ***REMOVED***
  var header = opts.header;
  var payload = opts.payload;
  var secretOrKey = opts.secret || opts.privateKey;
  var encoding = opts.encoding;
  var algo = jwa(header.alg);
  var securedInput = jwsSecuredInput(header, payload, encoding);
  var signature = algo.sign(securedInput, secretOrKey);
  return util.format('%s.%s', securedInput, signature);
***REMOVED***

function SignStream(opts) ***REMOVED***
  var secret = opts.secret||opts.privateKey||opts.key;
  var secretStream = new DataStream(secret);
  this.readable = true;
  this.header = opts.header;
  this.encoding = opts.encoding;
  this.secret = this.privateKey = this.key = secretStream;
  this.payload = new DataStream(opts.payload);
  this.secret.once('close', function () ***REMOVED***
    if (!this.payload.writable && this.readable)
      this.sign();
  ***REMOVED***.bind(this));

  this.payload.once('close', function () ***REMOVED***
    if (!this.secret.writable && this.readable)
      this.sign();
  ***REMOVED***.bind(this));
***REMOVED***
util.inherits(SignStream, Stream);

SignStream.prototype.sign = function sign() ***REMOVED***
  try ***REMOVED***
    var signature = jwsSign(***REMOVED***
      header: this.header,
      payload: this.payload.buffer,
      secret: this.secret.buffer,
      encoding: this.encoding
    ***REMOVED***);
    this.emit('done', signature);
    this.emit('data', signature);
    this.emit('end');
    this.readable = false;
    return signature;
  ***REMOVED*** catch (e) ***REMOVED***
    this.readable = false;
    this.emit('error', e);
    this.emit('close');
  ***REMOVED***
***REMOVED***;

SignStream.sign = jwsSign;

module.exports = SignStream;
