// jws/lib/verify-stream.js
/*global module*/
var base64url = require('base64url');
var DataStream = require('./data-stream');
var jwa = require('jwa');
var Stream = require('stream');
var toString = require('./tostring');
var util = require('util');
var JWS_REGEX = /^[a-zA-Z0-9\-_]+?\.[a-zA-Z0-9\-_]+?\.([a-zA-Z0-9\-_]+)?$/;

function isObject(thing) ***REMOVED***
  return Object.prototype.toString.call(thing) === '[object Object]';
***REMOVED***

function safeJsonParse(thing) ***REMOVED***
  if (isObject(thing))
    return thing;
  try ***REMOVED*** return JSON.parse(thing); ***REMOVED***
  catch (e) ***REMOVED*** return undefined; ***REMOVED***
***REMOVED***

function headerFromJWS(jwsSig) ***REMOVED***
  var encodedHeader = jwsSig.split('.', 1)[0];
  return safeJsonParse(base64url.decode(encodedHeader, 'binary'));
***REMOVED***

function securedInputFromJWS(jwsSig) ***REMOVED***
  return jwsSig.split('.', 2).join('.');
***REMOVED***

function signatureFromJWS(jwsSig) ***REMOVED***
  return jwsSig.split('.')[2];
***REMOVED***

function payloadFromJWS(jwsSig, encoding) ***REMOVED***
  encoding = encoding || 'utf8';
  var payload = jwsSig.split('.')[1];
  return base64url.decode(payload, encoding);
***REMOVED***

function isValidJws(string) ***REMOVED***
  return JWS_REGEX.test(string) && !!headerFromJWS(string);
***REMOVED***

function jwsVerify(jwsSig, algorithm, secretOrKey) ***REMOVED***
  if (!algorithm) ***REMOVED***
    var err = new Error("Missing algorithm parameter for jws.verify");
    err.code = "MISSING_ALGORITHM";
    throw err;
  ***REMOVED***
  jwsSig = toString(jwsSig);
  var signature = signatureFromJWS(jwsSig);
  var securedInput = securedInputFromJWS(jwsSig);
  var algo = jwa(algorithm);
  return algo.verify(securedInput, signature, secretOrKey);
***REMOVED***

function jwsDecode(jwsSig, opts) ***REMOVED***
  opts = opts || ***REMOVED******REMOVED***;
  jwsSig = toString(jwsSig);

  if (!isValidJws(jwsSig))
    return null;

  var header = headerFromJWS(jwsSig);

  if (!header)
    return null;

  var payload = payloadFromJWS(jwsSig);
  if (header.typ === 'JWT' || opts.json)
    payload = JSON.parse(payload, opts.encoding);

  return ***REMOVED***
    header: header,
    payload: payload,
    signature: signatureFromJWS(jwsSig)
  ***REMOVED***;
***REMOVED***

function VerifyStream(opts) ***REMOVED***
  opts = opts || ***REMOVED******REMOVED***;
  var secretOrKey = opts.secret||opts.publicKey||opts.key;
  var secretStream = new DataStream(secretOrKey);
  this.readable = true;
  this.algorithm = opts.algorithm;
  this.encoding = opts.encoding;
  this.secret = this.publicKey = this.key = secretStream;
  this.signature = new DataStream(opts.signature);
  this.secret.once('close', function () ***REMOVED***
    if (!this.signature.writable && this.readable)
      this.verify();
  ***REMOVED***.bind(this));

  this.signature.once('close', function () ***REMOVED***
    if (!this.secret.writable && this.readable)
      this.verify();
  ***REMOVED***.bind(this));
***REMOVED***
util.inherits(VerifyStream, Stream);
VerifyStream.prototype.verify = function verify() ***REMOVED***
  try ***REMOVED***
    var valid = jwsVerify(this.signature.buffer, this.algorithm, this.key.buffer);
    var obj = jwsDecode(this.signature.buffer, this.encoding);
    this.emit('done', valid, obj);
    this.emit('data', valid);
    this.emit('end');
    this.readable = false;
    return valid;
  ***REMOVED*** catch (e) ***REMOVED***
    this.readable = false;
    this.emit('error', e);
    this.emit('close');
  ***REMOVED***
***REMOVED***;

VerifyStream.decode = jwsDecode;
VerifyStream.isValid = isValidJws;
VerifyStream.verify = jwsVerify;

module.exports = VerifyStream;
