// accepts/index.js
/*!
 * accepts
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var Negotiator = require('negotiator')
var mime = require('mime-types')

/**
 * Module exports.
 * @public
 */

module.exports = Accepts

/**
 * Create a new Accepts object for the given req.
 *
 * @param ***REMOVED***object***REMOVED*** req
 * @public
 */

function Accepts(req) ***REMOVED***
  if (!(this instanceof Accepts))
    return new Accepts(req)

  this.headers = req.headers
  this.negotiator = new Negotiator(req)
***REMOVED***

/**
 * Check if the given `type(s)` is acceptable, returning
 * the best match when true, otherwise `undefined`, in which
 * case you should respond with 406 "Not Acceptable".
 *
 * The `type` value may be a single mime type string
 * such as "application/json", the extension name
 * such as "json" or an array `["json", "html", "text/plain"]`. When a list
 * or array is given the _best_ match, if any is returned.
 *
 * Examples:
 *
 *     // Accept: text/html
 *     this.types('html');
 *     // => "html"
 *
 *     // Accept: text/*, application/json
 *     this.types('html');
 *     // => "html"
 *     this.types('text/html');
 *     // => "text/html"
 *     this.types('json', 'text');
 *     // => "json"
 *     this.types('application/json');
 *     // => "application/json"
 *
 *     // Accept: text/*, application/json
 *     this.types('image/png');
 *     this.types('png');
 *     // => undefined
 *
 *     // Accept: text/*;q=.5, application/json
 *     this.types(['html', 'json']);
 *     this.types('html', 'json');
 *     // => "json"
 *
 * @param ***REMOVED***String|Array***REMOVED*** types...
 * @return ***REMOVED***String|Array|Boolean***REMOVED***
 * @public
 */

Accepts.prototype.type =
Accepts.prototype.types = function (types_) ***REMOVED***
  var types = types_

  // support flattened arguments
  if (types && !Array.isArray(types)) ***REMOVED***
    types = new Array(arguments.length)
    for (var i = 0; i < types.length; i++) ***REMOVED***
      types[i] = arguments[i]
    ***REMOVED***
  ***REMOVED***

  // no types, return all requested types
  if (!types || types.length === 0) ***REMOVED***
    return this.negotiator.mediaTypes()
  ***REMOVED***

  if (!this.headers.accept) return types[0];
  var mimes = types.map(extToMime);
  var accepts = this.negotiator.mediaTypes(mimes.filter(validMime));
  var first = accepts[0];
  if (!first) return false;
  return types[mimes.indexOf(first)];
***REMOVED***

/**
 * Return accepted encodings or best fit based on `encodings`.
 *
 * Given `Accept-Encoding: gzip, deflate`
 * an array sorted by quality is returned:
 *
 *     ['gzip', 'deflate']
 *
 * @param ***REMOVED***String|Array***REMOVED*** encodings...
 * @return ***REMOVED***String|Array***REMOVED***
 * @public
 */

Accepts.prototype.encoding =
Accepts.prototype.encodings = function (encodings_) ***REMOVED***
  var encodings = encodings_

  // support flattened arguments
  if (encodings && !Array.isArray(encodings)) ***REMOVED***
    encodings = new Array(arguments.length)
    for (var i = 0; i < encodings.length; i++) ***REMOVED***
      encodings[i] = arguments[i]
    ***REMOVED***
  ***REMOVED***

  // no encodings, return all requested encodings
  if (!encodings || encodings.length === 0) ***REMOVED***
    return this.negotiator.encodings()
  ***REMOVED***

  return this.negotiator.encodings(encodings)[0] || false
***REMOVED***

/**
 * Return accepted charsets or best fit based on `charsets`.
 *
 * Given `Accept-Charset: utf-8, iso-8859-1;q=0.2, utf-7;q=0.5`
 * an array sorted by quality is returned:
 *
 *     ['utf-8', 'utf-7', 'iso-8859-1']
 *
 * @param ***REMOVED***String|Array***REMOVED*** charsets...
 * @return ***REMOVED***String|Array***REMOVED***
 * @public
 */

Accepts.prototype.charset =
Accepts.prototype.charsets = function (charsets_) ***REMOVED***
  var charsets = charsets_

  // support flattened arguments
  if (charsets && !Array.isArray(charsets)) ***REMOVED***
    charsets = new Array(arguments.length)
    for (var i = 0; i < charsets.length; i++) ***REMOVED***
      charsets[i] = arguments[i]
    ***REMOVED***
  ***REMOVED***

  // no charsets, return all requested charsets
  if (!charsets || charsets.length === 0) ***REMOVED***
    return this.negotiator.charsets()
  ***REMOVED***

  return this.negotiator.charsets(charsets)[0] || false
***REMOVED***

/**
 * Return accepted languages or best fit based on `langs`.
 *
 * Given `Accept-Language: en;q=0.8, es, pt`
 * an array sorted by quality is returned:
 *
 *     ['es', 'pt', 'en']
 *
 * @param ***REMOVED***String|Array***REMOVED*** langs...
 * @return ***REMOVED***Array|String***REMOVED***
 * @public
 */

Accepts.prototype.lang =
Accepts.prototype.langs =
Accepts.prototype.language =
Accepts.prototype.languages = function (languages_) ***REMOVED***
  var languages = languages_

  // support flattened arguments
  if (languages && !Array.isArray(languages)) ***REMOVED***
    languages = new Array(arguments.length)
    for (var i = 0; i < languages.length; i++) ***REMOVED***
      languages[i] = arguments[i]
    ***REMOVED***
  ***REMOVED***

  // no languages, return all requested languages
  if (!languages || languages.length === 0) ***REMOVED***
    return this.negotiator.languages()
  ***REMOVED***

  return this.negotiator.languages(languages)[0] || false
***REMOVED***

/**
 * Convert extnames to mime.
 *
 * @param ***REMOVED***String***REMOVED*** type
 * @return ***REMOVED***String***REMOVED***
 * @private
 */

function extToMime(type) ***REMOVED***
  return type.indexOf('/') === -1
    ? mime.lookup(type)
    : type
***REMOVED***

/**
 * Check if mime is valid.
 *
 * @param ***REMOVED***String***REMOVED*** type
 * @return ***REMOVED***String***REMOVED***
 * @private
 */

function validMime(type) ***REMOVED***
  return typeof type === 'string';
***REMOVED***
