// qs/lib/parse.js
'use strict';

var Utils = require('./utils');

var defaults = ***REMOVED***
    delimiter: '&',
    depth: 5,
    arrayLimit: 20,
    parameterLimit: 1000,
    strictNullHandling: false,
    plainObjects: false,
    allowPrototypes: false,
    allowDots: false,
    decoder: Utils.decode
***REMOVED***;

var parseValues = function parseValues(str, options) ***REMOVED***
    var obj = ***REMOVED******REMOVED***;
    var parts = str.split(options.delimiter, options.parameterLimit === Infinity ? undefined : options.parameterLimit);

    for (var i = 0; i < parts.length; ++i) ***REMOVED***
        var part = parts[i];
        var pos = part.indexOf(']=') === -1 ? part.indexOf('=') : part.indexOf(']=') + 1;

        if (pos === -1) ***REMOVED***
            obj[options.decoder(part)] = '';

            if (options.strictNullHandling) ***REMOVED***
                obj[options.decoder(part)] = null;
            ***REMOVED***
        ***REMOVED*** else ***REMOVED***
            var key = options.decoder(part.slice(0, pos));
            var val = options.decoder(part.slice(pos + 1));

            if (Object.prototype.hasOwnProperty.call(obj, key)) ***REMOVED***
                obj[key] = [].concat(obj[key]).concat(val);
            ***REMOVED*** else ***REMOVED***
                obj[key] = val;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    return obj;
***REMOVED***;

var parseObject = function parseObject(chain, val, options) ***REMOVED***
    if (!chain.length) ***REMOVED***
        return val;
    ***REMOVED***

    var root = chain.shift();

    var obj;
    if (root === '[]') ***REMOVED***
        obj = [];
        obj = obj.concat(parseObject(chain, val, options));
    ***REMOVED*** else ***REMOVED***
        obj = options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;
        var cleanRoot = root[0] === '[' && root[root.length - 1] === ']' ? root.slice(1, root.length - 1) : root;
        var index = parseInt(cleanRoot, 10);
        if (
            !isNaN(index) &&
            root !== cleanRoot &&
            String(index) === cleanRoot &&
            index >= 0 &&
            (options.parseArrays && index <= options.arrayLimit)
        ) ***REMOVED***
            obj = [];
            obj[index] = parseObject(chain, val, options);
        ***REMOVED*** else ***REMOVED***
            obj[cleanRoot] = parseObject(chain, val, options);
        ***REMOVED***
    ***REMOVED***

    return obj;
***REMOVED***;

var parseKeys = function parseKeys(givenKey, val, options) ***REMOVED***
    if (!givenKey) ***REMOVED***
        return;
    ***REMOVED***

    // Transform dot notation to bracket notation
    var key = options.allowDots ? givenKey.replace(/\.([^\.\[]+)/g, '[$1]') : givenKey;

    // The regex chunks

    var parent = /^([^\[\]]*)/;
    var child = /(\[[^\[\]]*\])/g;

    // Get the parent

    var segment = parent.exec(key);

    // Stash the parent if it exists

    var keys = [];
    if (segment[1]) ***REMOVED***
        // If we aren't using plain objects, optionally prefix keys
        // that would overwrite object prototype properties
        if (!options.plainObjects && Object.prototype.hasOwnProperty(segment[1])) ***REMOVED***
            if (!options.allowPrototypes) ***REMOVED***
                return;
            ***REMOVED***
        ***REMOVED***

        keys.push(segment[1]);
    ***REMOVED***

    // Loop through children appending to the array until we hit depth

    var i = 0;
    while ((segment = child.exec(key)) !== null && i < options.depth) ***REMOVED***
        i += 1;
        if (!options.plainObjects && Object.prototype.hasOwnProperty(segment[1].replace(/\[|\]/g, ''))) ***REMOVED***
            if (!options.allowPrototypes) ***REMOVED***
                continue;
            ***REMOVED***
        ***REMOVED***
        keys.push(segment[1]);
    ***REMOVED***

    // If there's a remainder, just add whatever is left

    if (segment) ***REMOVED***
        keys.push('[' + key.slice(segment.index) + ']');
    ***REMOVED***

    return parseObject(keys, val, options);
***REMOVED***;

module.exports = function (str, opts) ***REMOVED***
    var options = opts || ***REMOVED******REMOVED***;

    if (options.decoder !== null && options.decoder !== undefined && typeof options.decoder !== 'function') ***REMOVED***
        throw new TypeError('Decoder has to be a function.');
    ***REMOVED***

    options.delimiter = typeof options.delimiter === 'string' || Utils.isRegExp(options.delimiter) ? options.delimiter : defaults.delimiter;
    options.depth = typeof options.depth === 'number' ? options.depth : defaults.depth;
    options.arrayLimit = typeof options.arrayLimit === 'number' ? options.arrayLimit : defaults.arrayLimit;
    options.parseArrays = options.parseArrays !== false;
    options.decoder = typeof options.decoder === 'function' ? options.decoder : defaults.decoder;
    options.allowDots = typeof options.allowDots === 'boolean' ? options.allowDots : defaults.allowDots;
    options.plainObjects = typeof options.plainObjects === 'boolean' ? options.plainObjects : defaults.plainObjects;
    options.allowPrototypes = typeof options.allowPrototypes === 'boolean' ? options.allowPrototypes : defaults.allowPrototypes;
    options.parameterLimit = typeof options.parameterLimit === 'number' ? options.parameterLimit : defaults.parameterLimit;
    options.strictNullHandling = typeof options.strictNullHandling === 'boolean' ? options.strictNullHandling : defaults.strictNullHandling;

    if (str === '' || str === null || typeof str === 'undefined') ***REMOVED***
        return options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;
    ***REMOVED***

    var tempObj = typeof str === 'string' ? parseValues(str, options) : str;
    var obj = options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;

    // Iterate over the keys and setup the new object

    var keys = Object.keys(tempObj);
    for (var i = 0; i < keys.length; ++i) ***REMOVED***
        var key = keys[i];
        var newObj = parseKeys(key, tempObj[key], options);
        obj = Utils.merge(obj, newObj, options);
    ***REMOVED***

    return Utils.compact(obj);
***REMOVED***;
