// express/lib/view.js
/*!
 * express
 * Copyright(c) 2009-2013 TJ Holowaychuk
 * Copyright(c) 2013 Roman Shtylman
 * Copyright(c) 2014-2015 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 * @private
 */

var debug = require('debug')('express:view');
var path = require('path');
var fs = require('fs');
var utils = require('./utils');

/**
 * Module variables.
 * @private
 */

var dirname = path.dirname;
var basename = path.basename;
var extname = path.extname;
var join = path.join;
var resolve = path.resolve;

/**
 * Module exports.
 * @public
 */

module.exports = View;

/**
 * Initialize a new `View` with the given `name`.
 *
 * Options:
 *
 *   - `defaultEngine` the default template engine name
 *   - `engines` template engine require() cache
 *   - `root` root path for view lookup
 *
 * @param ***REMOVED***string***REMOVED*** name
 * @param ***REMOVED***object***REMOVED*** options
 * @public
 */

function View(name, options) ***REMOVED***
  var opts = options || ***REMOVED******REMOVED***;

  this.defaultEngine = opts.defaultEngine;
  this.ext = extname(name);
  this.name = name;
  this.root = opts.root;

  if (!this.ext && !this.defaultEngine) ***REMOVED***
    throw new Error('No default engine was specified and no extension was provided.');
  ***REMOVED***

  var fileName = name;

  if (!this.ext) ***REMOVED***
    // get extension from default engine name
    this.ext = this.defaultEngine[0] !== '.'
      ? '.' + this.defaultEngine
      : this.defaultEngine;

    fileName += this.ext;
  ***REMOVED***

  if (!opts.engines[this.ext]) ***REMOVED***
    // load engine
    opts.engines[this.ext] = require(this.ext.substr(1)).__express;
  ***REMOVED***

  // store loaded engine
  this.engine = opts.engines[this.ext];

  // lookup path
  this.path = this.lookup(fileName);
***REMOVED***

/**
 * Lookup view by the given `name`
 *
 * @param ***REMOVED***string***REMOVED*** name
 * @private
 */

View.prototype.lookup = function lookup(name) ***REMOVED***
  var path;
  var roots = [].concat(this.root);

  debug('lookup "%s"', name);

  for (var i = 0; i < roots.length && !path; i++) ***REMOVED***
    var root = roots[i];

    // resolve the path
    var loc = resolve(root, name);
    var dir = dirname(loc);
    var file = basename(loc);

    // resolve the file
    path = this.resolve(dir, file);
  ***REMOVED***

  return path;
***REMOVED***;

/**
 * Render with the given options.
 *
 * @param ***REMOVED***object***REMOVED*** options
 * @param ***REMOVED***function***REMOVED*** callback
 * @private
 */

View.prototype.render = function render(options, callback) ***REMOVED***
  debug('render "%s"', this.path);
  this.engine(this.path, options, callback);
***REMOVED***;

/**
 * Resolve the file within the given directory.
 *
 * @param ***REMOVED***string***REMOVED*** dir
 * @param ***REMOVED***string***REMOVED*** file
 * @private
 */

View.prototype.resolve = function resolve(dir, file) ***REMOVED***
  var ext = this.ext;

  // <path>.<ext>
  var path = join(dir, file);
  var stat = tryStat(path);

  if (stat && stat.isFile()) ***REMOVED***
    return path;
  ***REMOVED***

  // <path>/index.<ext>
  path = join(dir, basename(file, ext), 'index' + ext);
  stat = tryStat(path);

  if (stat && stat.isFile()) ***REMOVED***
    return path;
  ***REMOVED***
***REMOVED***;

/**
 * Return a stat, maybe.
 *
 * @param ***REMOVED***string***REMOVED*** path
 * @return ***REMOVED***fs.Stats***REMOVED***
 * @private
 */

function tryStat(path) ***REMOVED***
  debug('stat "%s"', path);

  try ***REMOVED***
    return fs.statSync(path);
  ***REMOVED*** catch (e) ***REMOVED***
    return undefined;
  ***REMOVED***
***REMOVED***
