// colors/lib/custom/zalgo.js
// please no
module['exports'] = function zalgo(text, options) ***REMOVED***
  text = text || "   he is here   ";
  var soul = ***REMOVED***
    "up" : [
      '̍', '̎', '̄', '̅',
      '̿', '̑', '̆', '̐',
      '͒', '͗', '͑', '̇',
      '̈', '̊', '͂', '̓',
      '̈', '͊', '͋', '͌',
      '̃', '̂', '̌', '͐',
      '̀', '́', '̋', '̏',
      '̒', '̓', '̔', '̽',
      '̉', 'ͣ', 'ͤ', 'ͥ',
      'ͦ', 'ͧ', 'ͨ', 'ͩ',
      'ͪ', 'ͫ', 'ͬ', 'ͭ',
      'ͮ', 'ͯ', '̾', '͛',
      '͆', '̚'
    ],
    "down" : [
      '̖', '̗', '̘', '̙',
      '̜', '̝', '̞', '̟',
      '̠', '̤', '̥', '̦',
      '̩', '̪', '̫', '̬',
      '̭', '̮', '̯', '̰',
      '̱', '̲', '̳', '̹',
      '̺', '̻', '̼', 'ͅ',
      '͇', '͈', '͉', '͍',
      '͎', '͓', '͔', '͕',
      '͖', '͙', '͚', '̣'
    ],
    "mid" : [
      '̕', '̛', '̀', '́',
      '͘', '̡', '̢', '̧',
      '̨', '̴', '̵', '̶',
      '͜', '͝', '͞',
      '͟', '͠', '͢', '̸',
      '̷', '͡', ' ҉'
    ]
  ***REMOVED***,
  all = [].concat(soul.up, soul.down, soul.mid),
  zalgo = ***REMOVED******REMOVED***;

  function randomNumber(range) ***REMOVED***
    var r = Math.floor(Math.random() * range);
    return r;
  ***REMOVED***

  function is_char(character) ***REMOVED***
    var bool = false;
    all.filter(function (i) ***REMOVED***
      bool = (i === character);
    ***REMOVED***);
    return bool;
  ***REMOVED***
  

  function heComes(text, options) ***REMOVED***
    var result = '', counts, l;
    options = options || ***REMOVED******REMOVED***;
    options["up"] = options["up"] || true;
    options["mid"] = options["mid"] || true;
    options["down"] = options["down"] || true;
    options["size"] = options["size"] || "maxi";
    text = text.split('');
    for (l in text) ***REMOVED***
      if (is_char(l)) ***REMOVED***
        continue;
      ***REMOVED***
      result = result + text[l];
      counts = ***REMOVED***"up" : 0, "down" : 0, "mid" : 0***REMOVED***;
      switch (options.size) ***REMOVED***
      case 'mini':
        counts.up = randomNumber(8);
        counts.min = randomNumber(2);
        counts.down = randomNumber(8);
        break;
      case 'maxi':
        counts.up = randomNumber(16) + 3;
        counts.min = randomNumber(4) + 1;
        counts.down = randomNumber(64) + 3;
        break;
      default:
        counts.up = randomNumber(8) + 1;
        counts.mid = randomNumber(6) / 2;
        counts.down = randomNumber(8) + 1;
        break;
      ***REMOVED***

      var arr = ["up", "mid", "down"];
      for (var d in arr) ***REMOVED***
        var index = arr[d];
        for (var i = 0 ; i <= counts[index]; i++) ***REMOVED***
          if (options[index]) ***REMOVED***
            result = result + soul[index][randomNumber(soul[index].length)];
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return result;
  ***REMOVED***
  // don't summon him
  return heComes(text);
***REMOVED***
