// hooks-fixed/test.js
var hooks = require('./hooks')
  , should = require('should')
  , assert = require('assert')
  , _ = require('underscore');

// TODO Add in test for making sure all pres get called if pre is defined directly on an instance.
// TODO Test for calling `done` twice or `next` twice in the same function counts only once
module.exports = ***REMOVED***
  'should be able to assign multiple hooks at once': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook(***REMOVED***
      hook1: function (a) ***REMOVED******REMOVED***,
      hook2: function (b) ***REMOVED******REMOVED***
    ***REMOVED***);
    var a = new A();
    assert.equal(typeof a.hook1, 'function');
    assert.equal(typeof a.hook2, 'function');
  ***REMOVED***,
  'should run without pres and posts when not present': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    var a = new A();
    a.save();
    a.value.should.equal(1);
  ***REMOVED***,
  'should run with pres when present': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      this.preValue = 2;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.value.should.equal(1);
    a.preValue.should.equal(2);
  ***REMOVED***,
  'should run with posts when present': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.post('save', function (next) ***REMOVED***
      this.value = 2;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.value.should.equal(2);
  ***REMOVED***,
  'should run pres and posts when present': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      this.preValue = 2;
      next();
    ***REMOVED***);
    A.post('save', function (next) ***REMOVED***
      this.value = 3;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.value.should.equal(3);
    a.preValue.should.equal(2);
  ***REMOVED***,
  'should run posts after pres': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      this.override = 100;
      next();
    ***REMOVED***);
    A.post('save', function (next) ***REMOVED***
      this.override = 200;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.value.should.equal(1);
    a.override.should.equal(200);
  ***REMOVED***,
  'should not run a hook if a pre fails': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    var counter = 0;
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***, function (err) ***REMOVED***
      counter++;
    ***REMOVED***);
    A.pre('save', true, function (next, done) ***REMOVED***
      next(new Error());
    ***REMOVED***);
    var a = new A();
    a.save();
    counter.should.equal(1);
    assert.equal(typeof a.value, 'undefined');
  ***REMOVED***,
  'should be able to run multiple pres': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      this.v1 = 1;
      next();
    ***REMOVED***).pre('save', function (next) ***REMOVED***
      this.v2 = 2;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.v1.should.equal(1);
    a.v2.should.equal(2);
  ***REMOVED***,
  'should run multiple pres until a pre fails and not call the hook': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***, function (err) ***REMOVED******REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      this.v1 = 1;
      next();
    ***REMOVED***).pre('save', function (next) ***REMOVED***
      next(new Error());
    ***REMOVED***).pre('save', function (next) ***REMOVED***
      this.v3 = 3;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.v1.should.equal(1);
    assert.equal(typeof a.v3, 'undefined');
    assert.equal(typeof a.value, 'undefined');
  ***REMOVED***,
  'should be able to run multiple posts': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.post('save', function (next) ***REMOVED***
      this.value = 2;
      next();
    ***REMOVED***).post('save', function (next) ***REMOVED***
      this.value = 3.14;
      next();
    ***REMOVED***).post('save', function (next) ***REMOVED***
      this.v3 = 3;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    assert.equal(a.value, 3.14);
    assert.equal(a.v3, 3);
  ***REMOVED***,
  'should run only posts up until an error': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***, function (err) ***REMOVED******REMOVED***);
    A.post('save', function (next) ***REMOVED***
      this.value = 2;
      next();
    ***REMOVED***).post('save', function (next) ***REMOVED***
      this.value = 3;
      next(new Error());
    ***REMOVED***).post('save', function (next) ***REMOVED***
      this.value = 4;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.value.should.equal(3);
  ***REMOVED***,
  "should fall back first to the hook method's last argument as the error handler if it is a function of arity 1 or 2": function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    var counter = 0;
    A.hook('save', function (callback) ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.pre('save', true, function (next, done) ***REMOVED***
      next(new Error());
    ***REMOVED***);
    var a = new A();
    a.save( function (err) ***REMOVED***
      if (err instanceof Error) counter++;
    ***REMOVED***);
    counter.should.equal(1);
    should.deepEqual(undefined, a.value);
  ***REMOVED***,
  'should fall back second to the default error handler if specified': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    var counter = 0;
    A.hook('save', function (callback) ***REMOVED***
      this.value = 1;
    ***REMOVED***, function (err) ***REMOVED***
      if (err instanceof Error) counter++;
    ***REMOVED***);
    A.pre('save', true, function (next, done) ***REMOVED***
      next(new Error());
    ***REMOVED***);
    var a = new A();
    a.save();
    counter.should.equal(1);
    should.deepEqual(undefined, a.value);
  ***REMOVED***,
  'fallback default error handler should scope to the object': function () ***REMOVED***
    var A = function () ***REMOVED***
      this.counter = 0;
    ***REMOVED***;
    _.extend(A, hooks);
    var counter = 0;
    A.hook('save', function (callback) ***REMOVED***
      this.value = 1;
    ***REMOVED***, function (err) ***REMOVED***
      if (err instanceof Error) this.counter++;
    ***REMOVED***);
    A.pre('save', true, function (next, done) ***REMOVED***
      next(new Error());
    ***REMOVED***);
    var a = new A();
    a.save();
    a.counter.should.equal(1);
    should.deepEqual(undefined, a.value);
  ***REMOVED***,
  'should fall back last to throwing the error': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    var counter = 0;
    A.hook('save', function (err) ***REMOVED***
      if (err instanceof Error) return counter++;
      this.value = 1;
    ***REMOVED***);
    A.pre('save', true, function (next, done) ***REMOVED***
      next(new Error());
    ***REMOVED***);
    var a = new A();
    var didCatch = false;
    try ***REMOVED***
      a.save();
    ***REMOVED*** catch (e) ***REMOVED***
      didCatch = true;
      e.should.be.an.instanceof(Error);
      counter.should.equal(0);
      assert.equal(typeof a.value, 'undefined');
    ***REMOVED***
    didCatch.should.be.true;
  ***REMOVED***,
  "should proceed without mutating arguments if `next(null|undefined)` is called in a serial pre, and the last argument of the target method is a callback with node-like signature function (err, obj) ***REMOVED***...***REMOVED*** or function (err) ***REMOVED***...***REMOVED***": function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    var counter = 0;
    A.prototype.save = function (callback) ***REMOVED***
      this.value = 1;
      callback();
    ***REMOVED***;
    A.pre('save', function (next) ***REMOVED***
      next(null);
    ***REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      next(undefined);
    ***REMOVED***);
    var a = new A();
    a.save( function (err) ***REMOVED***
      if (err instanceof Error) counter++;
      else counter--;
    ***REMOVED***);
    counter.should.equal(-1);
    a.value.should.eql(1);
  ***REMOVED***,
  "should proceed with mutating arguments if `next(null|undefined)` is callback in a serial pre, and the last argument of the target method is not a function": function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.prototype.set = function (v) ***REMOVED***
      this.value = v;
    ***REMOVED***;
    A.pre('set', function (next) ***REMOVED***
      next(undefined);
    ***REMOVED***);
    A.pre('set', function (next) ***REMOVED***
      next(null);
    ***REMOVED***);
    var a = new A();
    a.set(1);
    should.strictEqual(null, a.value);
  ***REMOVED***,
  'should not run any posts if a pre fails': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 2;
    ***REMOVED***, function (err) ***REMOVED******REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      this.value = 1;
      next(new Error());
    ***REMOVED***).post('save', function (next) ***REMOVED***
      this.value = 3;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.value.should.equal(1);
  ***REMOVED***,

  "can pass the hook's arguments verbatim to pres": function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('set', function (path, val) ***REMOVED***
      this[path] = val;
    ***REMOVED***);
    A.pre('set', function (next, path, val) ***REMOVED***
      path.should.equal('hello');
      val.should.equal('world');
      next();
    ***REMOVED***);
    var a = new A();
    a.set('hello', 'world');
    a.hello.should.equal('world');
  ***REMOVED***,
//  "can pass the hook's arguments as an array to pres": function () ***REMOVED***
//    // Great for dynamic arity - e.g., slice(...)
//    var A = function () ***REMOVED******REMOVED***;
//    _.extend(A, hooks);
//    A.hook('set', function (path, val) ***REMOVED***
//      this[path] = val;
//    ***REMOVED***);
//    A.pre('set', function (next, hello, world) ***REMOVED***
//      hello.should.equal('hello');
//      world.should.equal('world');
//      next();
//    ***REMOVED***);
//    var a = new A();
//    a.set('hello', 'world');
//    assert.equal(a.hello, 'world');
//  ***REMOVED***,
  "can pass the hook's arguments verbatim to posts": function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('set', function (path, val) ***REMOVED***
      this[path] = val;
    ***REMOVED***);
    A.post('set', function (next, path, val) ***REMOVED***
      path.should.equal('hello');
      val.should.equal('world');
      next();
    ***REMOVED***);
    var a = new A();
    a.set('hello', 'world');
    assert.equal(a.hello, 'world');
  ***REMOVED***,
//  "can pass the hook's arguments as an array to posts": function () ***REMOVED***
//    var A = function () ***REMOVED******REMOVED***;
//    _.extend(A, hooks);
//    A.hook('set', function (path, val) ***REMOVED***
//      this[path] = val;
//    ***REMOVED***);
//    A.post('set', function (next, halt, args) ***REMOVED***
//      assert.equal(args[0], 'hello');
//      assert.equal(args[1], 'world');
//      next();
//    ***REMOVED***);
//    var a = new A();
//    a.set('hello', 'world');
//    assert.equal(a.hello, 'world');
//  ***REMOVED***,
  "pres should be able to modify and pass on a modified version of the hook's arguments": function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('set', function (path, val) ***REMOVED***
      this[path] = val;
      assert.equal(arguments[2], 'optional');
    ***REMOVED***);
    A.pre('set', function (next, path, val) ***REMOVED***
      next('foo', 'bar');
    ***REMOVED***);
    A.pre('set', function (next, path, val) ***REMOVED***
      assert.equal(path, 'foo');
      assert.equal(val, 'bar');
      next('rock', 'says', 'optional');
    ***REMOVED***);
    A.pre('set', function (next, path, val, opt) ***REMOVED***
      assert.equal(path, 'rock');
      assert.equal(val, 'says');
      assert.equal(opt, 'optional');
      next();
    ***REMOVED***);
    var a = new A();
    a.set('hello', 'world');
    assert.equal(typeof a.hello, 'undefined');
    a.rock.should.equal('says');
  ***REMOVED***,
  'posts should see the modified version of arguments if the pres modified them': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('set', function (path, val) ***REMOVED***
      this[path] = val;
    ***REMOVED***);
    A.pre('set', function (next, path, val) ***REMOVED***
      next('foo', 'bar');
    ***REMOVED***);
    A.post('set', function (next, path, val) ***REMOVED***
      path.should.equal('foo');
      val.should.equal('bar');
    ***REMOVED***);
    var a = new A();
    a.set('hello', 'world');
    assert.equal(typeof a.hello, 'undefined');
    a.foo.should.equal('bar');
  ***REMOVED***,
  'should pad missing arguments (relative to expected arguments of the hook) with null': function () ***REMOVED***
    // Otherwise, with hookFn = function (a, b, next, ),
    // if we use hookFn(a), then because the pre functions are of the form
    // preFn = function (a, b, next, ), then it actually gets executed with
    // preFn(a, next, ), so when we call next() from within preFn, we are actually
    // calling ()

    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('set', function (path, val, opts) ***REMOVED***
      this[path] = val;
    ***REMOVED***);
    A.pre('set', function (next, path, val, opts) ***REMOVED***
      next('foo', 'bar');
      assert.equal(typeof opts, 'undefined');
    ***REMOVED***);
    var a = new A();
    a.set('hello', 'world');
  ***REMOVED***,

  'should not invoke the target method until all asynchronous middleware have invoked dones': function () ***REMOVED***
    var counter = 0;
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('set', function (path, val) ***REMOVED***
      counter++;
      this[path] = val;
      counter.should.equal(7);
    ***REMOVED***);
    A.pre('set', function (next, path, val) ***REMOVED***
      counter++;
      next();
    ***REMOVED***);
    A.pre('set', true, function (next, done, path, val) ***REMOVED***
      counter++;
      setTimeout(function () ***REMOVED***
        counter++;
        done();
      ***REMOVED***, 1000);
      next();
    ***REMOVED***);
    A.pre('set', function (next, path, val) ***REMOVED***
      counter++;
      next();
    ***REMOVED***);
    A.pre('set', true, function (next, done, path, val) ***REMOVED***
      counter++;
      setTimeout(function () ***REMOVED***
        counter++;
        done();
      ***REMOVED***, 500);
      next();
    ***REMOVED***);
    var a = new A();
    a.set('hello', 'world');
  ***REMOVED***,

  'invoking a method twice should run its async middleware twice': function () ***REMOVED***
    var counter = 0;
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('set', function (path, val) ***REMOVED***
      this[path] = val;
      if (path === 'hello') counter.should.equal(1);
      if (path === 'foo') counter.should.equal(2);
    ***REMOVED***);
    A.pre('set', true, function (next, done, path, val) ***REMOVED***
      setTimeout(function () ***REMOVED***
        counter++;
        done();
      ***REMOVED***, 1000);
      next();
    ***REMOVED***);
    var a = new A();
    a.set('hello', 'world');
    a.set('foo', 'bar');
  ***REMOVED***,

  'calling the same done multiple times should have the effect of only calling it once': function () ***REMOVED***
    var A = function () ***REMOVED***
      this.acked = false;
    ***REMOVED***;
    _.extend(A, hooks);
    A.hook('ack', function () ***REMOVED***
      console.log("UH OH, YOU SHOULD NOT BE SEEING THIS");
      this.acked = true;
    ***REMOVED***);
    A.pre('ack', true, function (next, done) ***REMOVED***
      next();
      done();
      done();
    ***REMOVED***);
    A.pre('ack', true, function (next, done) ***REMOVED***
      next();
      // Notice that done() is not invoked here
    ***REMOVED***);
    var a = new A();
    a.ack();
    setTimeout( function () ***REMOVED***
      a.acked.should.be.false;
    ***REMOVED***, 1000);
  ***REMOVED***,

  'calling the same next multiple times should have the effect of only calling it once': function (beforeExit) ***REMOVED***
    var A = function () ***REMOVED***
      this.acked = false;
    ***REMOVED***;
    _.extend(A, hooks);
    A.hook('ack', function () ***REMOVED***
      console.log("UH OH, YOU SHOULD NOT BE SEEING THIS");
      this.acked = true;
    ***REMOVED***);
    A.pre('ack', function (next) ***REMOVED***
      // force a throw to re-exec next()
      try ***REMOVED***
        next(new Error('bam'));
      ***REMOVED*** catch (err) ***REMOVED***
        next();
      ***REMOVED***
    ***REMOVED***);
    A.pre('ack', function (next) ***REMOVED***
      next();
    ***REMOVED***);
    var a = new A();
    a.ack();
    beforeExit( function () ***REMOVED***
      a.acked.should.be.false;
    ***REMOVED***);
  ***REMOVED***,

  'asynchronous middleware should be able to pass an error via `done`, stopping the middleware chain': function () ***REMOVED***
    var counter = 0;
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('set', function (path, val, fn) ***REMOVED***
      counter++;
      this[path] = val;
      fn(null);
    ***REMOVED***);
    A.pre('set', true, function (next, done, path, val, fn) ***REMOVED***
      setTimeout(function () ***REMOVED***
        counter++;
        done(new Error);
      ***REMOVED***, 1000);
      next();
    ***REMOVED***);
    var a = new A();
    a.set('hello', 'world', function (err) ***REMOVED***
      err.should.be.an.instanceof(Error);
      should.strictEqual(undefined, a['hello']);
      counter.should.eql(1);
    ***REMOVED***);
  ***REMOVED***,

  'should be able to remove a particular pre': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***
      , preTwo;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      this.preValueOne = 2;
      next();
    ***REMOVED***);
    A.pre('save', preTwo = function (next) ***REMOVED***
      this.preValueTwo = 4;
      next();
    ***REMOVED***);
    A.removePre('save', preTwo);
    var a = new A();
    a.save();
    a.value.should.equal(1);
    a.preValueOne.should.equal(2);
    should.strictEqual(undefined, a.preValueTwo);
  ***REMOVED***,

  'should be able to remove all pres associated with a hook': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      this.preValueOne = 2;
      next();
    ***REMOVED***);
    A.pre('save', function (next) ***REMOVED***
      this.preValueTwo = 4;
      next();
    ***REMOVED***);
    A.removePre('save');
    var a = new A();
    a.save();
    a.value.should.equal(1);
    should.strictEqual(undefined, a.preValueOne);
    should.strictEqual(undefined, a.preValueTwo);
  ***REMOVED***,

  'should be able to remove a particular post': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***
      , postTwo;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.post('save', function (next) ***REMOVED***
      this.postValueOne = 2;
      next();
    ***REMOVED***);
    A.post('save', postTwo = function (next) ***REMOVED***
      this.postValueTwo = 4;
      next();
    ***REMOVED***);
    A.removePost('save', postTwo);
    var a = new A();
    a.save();
    a.value.should.equal(1);
    a.postValueOne.should.equal(2);
    should.strictEqual(undefined, a.postValueTwo);
  ***REMOVED***,

  'should be able to remove all posts associated with a hook': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function () ***REMOVED***
      this.value = 1;
    ***REMOVED***);
    A.post('save', function (next) ***REMOVED***
      this.postValueOne = 2;
      next();
    ***REMOVED***);
    A.post('save', function (next) ***REMOVED***
      this.postValueTwo = 4;
      next();
    ***REMOVED***);
    A.removePost('save');
    var a = new A();
    a.save();
    a.value.should.equal(1);
    should.strictEqual(undefined, a.postValueOne);
    should.strictEqual(undefined, a.postValueTwo);
  ***REMOVED***,

  '#pre should lazily make a method hookable': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.prototype.save = function () ***REMOVED***
      this.value = 1;
    ***REMOVED***;
    A.pre('save', function (next) ***REMOVED***
      this.preValue = 2;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.value.should.equal(1);
    a.preValue.should.equal(2);
  ***REMOVED***,

  '#pre lazily making a method hookable should be able to provide a default errorHandler as the last argument': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    var preValue = "";
    _.extend(A, hooks);
    A.prototype.save = function () ***REMOVED***
      this.value = 1;
    ***REMOVED***;
    A.pre('save', function (next) ***REMOVED***
      next(new Error);
    ***REMOVED***, function (err) ***REMOVED***
      preValue = 'ERROR';
    ***REMOVED***);
    var a = new A();
    a.save();
    should.strictEqual(undefined, a.value);
    preValue.should.equal('ERROR');
  ***REMOVED***,

  '#post should lazily make a method hookable': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.prototype.save = function () ***REMOVED***
      this.value = 1;
    ***REMOVED***;
    A.post('save', function (next) ***REMOVED***
      this.value = 2;
      next();
    ***REMOVED***);
    var a = new A();
    a.save();
    a.value.should.equal(2);
  ***REMOVED***,

  "a lazy hooks setup should handle errors via a method's last argument, if it's a callback": function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.prototype.save = function (fn) ***REMOVED******REMOVED***;
    A.pre('save', function (next) ***REMOVED***
      next(new Error("hi there"));
    ***REMOVED***);
    var a = new A();
    a.save( function (err) ***REMOVED***
      err.should.be.an.instanceof(Error);
    ***REMOVED***);
  ***REMOVED***,

  'should intercept method callbacks for post handlers': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function (val, callback) ***REMOVED***
      this.value = val;
      callback();
    ***REMOVED***);
    A.post('save', function (next) ***REMOVED***
      assert.equal(a.value, 2);
      this.value += 2;
      setTimeout(next, 10);
    ***REMOVED***).post('save', function (next) ***REMOVED***
      assert.equal(a.value, 4);
      this.value += 3;
      setTimeout(next, 10);
    ***REMOVED***).post('save', function (next) ***REMOVED***
      assert.equal(a.value, 7);
      this.value2 = 3;
      setTimeout(next, 10);
    ***REMOVED***);
    var a = new A();
    a.save(2, function()***REMOVED***
      assert.equal(a.value, 7);
      assert.equal(a.value2, 3);
    ***REMOVED***);
  ***REMOVED***,

  'should handle parallel followed by serial': function () ***REMOVED***
    var A = function () ***REMOVED******REMOVED***;
    _.extend(A, hooks);
    A.hook('save', function (val, callback) ***REMOVED***
      this.value = val;
      callback();
    ***REMOVED***);
    A.pre('save', true, function(next, done) ***REMOVED***
      process.nextTick(function() ***REMOVED***
        done();
      ***REMOVED***);
      next();
    ***REMOVED***).pre('save', function(done) ***REMOVED***
      process.nextTick(function() ***REMOVED***
        done();
      ***REMOVED***);
    ***REMOVED***);
    var a = new A();
    a.save(2, function()***REMOVED***
      assert.ok(true);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;
