// colors/lib/colors.js
/*

The MIT License (MIT)

Original Library 
  - Copyright (c) Marak Squires

Additional functionality
 - Copyright (c) Sindre Sorhus <sindresorhus@gmail.com> (sindresorhus.com)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

*/

var colors = ***REMOVED******REMOVED***;
module['exports'] = colors;

colors.themes = ***REMOVED******REMOVED***;

var ansiStyles = colors.styles = require('./styles');
var defineProps = Object.defineProperties;

colors.supportsColor = require('./system/supports-colors');

if (typeof colors.enabled === "undefined") ***REMOVED***
  colors.enabled = colors.supportsColor;
***REMOVED***

colors.stripColors = colors.strip = function(str)***REMOVED***
  return ("" + str).replace(/\x1B\[\d+m/g, '');
***REMOVED***;


var stylize = colors.stylize = function stylize (str, style) ***REMOVED***
  return ansiStyles[style].open + str + ansiStyles[style].close;
***REMOVED***

var matchOperatorsRe = /[|\\***REMOVED******REMOVED***()[\]^$+*?.]/g;
var escapeStringRegexp = function (str) ***REMOVED***
  if (typeof str !== 'string') ***REMOVED***
    throw new TypeError('Expected a string');
  ***REMOVED***
  return str.replace(matchOperatorsRe,  '\\$&');
***REMOVED***

function build(_styles) ***REMOVED***
  var builder = function builder() ***REMOVED***
    return applyStyle.apply(builder, arguments);
  ***REMOVED***;
  builder._styles = _styles;
  // __proto__ is used because we must return a function, but there is
  // no way to create a function with a different prototype.
  builder.__proto__ = proto;
  return builder;
***REMOVED***

var styles = (function () ***REMOVED***
  var ret = ***REMOVED******REMOVED***;
  ansiStyles.grey = ansiStyles.gray;
  Object.keys(ansiStyles).forEach(function (key) ***REMOVED***
    ansiStyles[key].closeRe = new RegExp(escapeStringRegexp(ansiStyles[key].close), 'g');
    ret[key] = ***REMOVED***
      get: function () ***REMOVED***
        return build(this._styles.concat(key));
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***);
  return ret;
***REMOVED***)();

var proto = defineProps(function colors() ***REMOVED******REMOVED***, styles);

function applyStyle() ***REMOVED***
  var args = arguments;
  var argsLen = args.length;
  var str = argsLen !== 0 && String(arguments[0]);
  if (argsLen > 1) ***REMOVED***
    for (var a = 1; a < argsLen; a++) ***REMOVED***
      str += ' ' + args[a];
    ***REMOVED***
  ***REMOVED***

  if (!colors.enabled || !str) ***REMOVED***
    return str;
  ***REMOVED***

  var nestedStyles = this._styles;

  var i = nestedStyles.length;
  while (i--) ***REMOVED***
    var code = ansiStyles[nestedStyles[i]];
    str = code.open + str.replace(code.closeRe, code.open) + code.close;
  ***REMOVED***

  return str;
***REMOVED***

function applyTheme (theme) ***REMOVED***
  for (var style in theme) ***REMOVED***
    (function(style)***REMOVED***
      colors[style] = function(str)***REMOVED***
        return colors[theme[style]](str);
      ***REMOVED***;
    ***REMOVED***)(style)
  ***REMOVED***
***REMOVED***

colors.setTheme = function (theme) ***REMOVED***
  if (typeof theme === 'string') ***REMOVED***
    try ***REMOVED***
      colors.themes[theme] = require(theme);
      applyTheme(colors.themes[theme]);
      return colors.themes[theme];
    ***REMOVED*** catch (err) ***REMOVED***
      console.log(err);
      return err;
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    applyTheme(theme);
  ***REMOVED***
***REMOVED***;

function init() ***REMOVED***
  var ret = ***REMOVED******REMOVED***;
  Object.keys(styles).forEach(function (name) ***REMOVED***
    ret[name] = ***REMOVED***
      get: function () ***REMOVED***
        return build([name]);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***);
  return ret;
***REMOVED***

var sequencer = function sequencer (map, str) ***REMOVED***
  var exploded = str.split(""), i = 0;
  exploded = exploded.map(map);
  return exploded.join("");
***REMOVED***;

// custom formatter methods
colors.trap = require('./custom/trap');
colors.zalgo = require('./custom/zalgo');

// maps
colors.maps = ***REMOVED******REMOVED***;
colors.maps.america = require('./maps/america');
colors.maps.zebra = require('./maps/zebra');
colors.maps.rainbow = require('./maps/rainbow');
colors.maps.random = require('./maps/random')

for (var map in colors.maps) ***REMOVED***
  (function(map)***REMOVED***
    colors[map] = function (str) ***REMOVED***
      return sequencer(colors.maps[map], str);
    ***REMOVED***
  ***REMOVED***)(map)
***REMOVED***

defineProps(colors, init());
