// async/internal/DoublyLinkedList.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
    value: true
***REMOVED***);
exports.default = DLL;
// Simple doubly linked list (https://en.wikipedia.org/wiki/Doubly_linked_list) implementation
// used for queues. This implementation assumes that the node provided by the user can be modified
// to adjust the next and last properties. We implement only the minimal functionality
// for queue support.
function DLL() ***REMOVED***
    this.head = this.tail = null;
    this.length = 0;
***REMOVED***

function setInitial(dll, node) ***REMOVED***
    dll.length = 1;
    dll.head = dll.tail = node;
***REMOVED***

DLL.prototype.removeLink = function (node) ***REMOVED***
    if (node.prev) node.prev.next = node.next;else this.head = node.next;
    if (node.next) node.next.prev = node.prev;else this.tail = node.prev;

    node.prev = node.next = null;
    this.length -= 1;
    return node;
***REMOVED***;

DLL.prototype.empty = DLL;

DLL.prototype.insertAfter = function (node, newNode) ***REMOVED***
    newNode.prev = node;
    newNode.next = node.next;
    if (node.next) node.next.prev = newNode;else this.tail = newNode;
    node.next = newNode;
    this.length += 1;
***REMOVED***;

DLL.prototype.insertBefore = function (node, newNode) ***REMOVED***
    newNode.prev = node.prev;
    newNode.next = node;
    if (node.prev) node.prev.next = newNode;else this.head = newNode;
    node.prev = newNode;
    this.length += 1;
***REMOVED***;

DLL.prototype.unshift = function (node) ***REMOVED***
    if (this.head) this.insertBefore(this.head, node);else setInitial(this, node);
***REMOVED***;

DLL.prototype.push = function (node) ***REMOVED***
    if (this.tail) this.insertAfter(this.tail, node);else setInitial(this, node);
***REMOVED***;

DLL.prototype.shift = function () ***REMOVED***
    return this.head && this.removeLink(this.head);
***REMOVED***;

DLL.prototype.pop = function () ***REMOVED***
    return this.tail && this.removeLink(this.tail);
***REMOVED***;
module.exports = exports["default"];
