// joi/test/any.js
// Load modules

var Lab = require('lab');
var Code = require('code');
var Joi = require('../lib');
var Helper = require('./helper');


// Declare internals

var internals = ***REMOVED******REMOVED***;


// Test shortcuts

var lab = exports.lab = Lab.script();
var describe = lab.describe;
var it = lab.it;
var expect = Code.expect;


describe('any', function () ***REMOVED***

    describe('#equal', function () ***REMOVED***

        it('validates valid values', function (done) ***REMOVED***

            Helper.validate(Joi.equal(4), [
                [4, true],
                [5, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    describe('#not', function () ***REMOVED***

        it('validates invalid values', function (done) ***REMOVED***

            Helper.validate(Joi.not(5), [
                [4, true],
                [5, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    describe('#exist', function () ***REMOVED***

        it('validates required values', function (done) ***REMOVED***

            Helper.validate(Joi.exist(), [
                [4, true],
                [undefined, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    describe('#strict', function () ***REMOVED***

        it('validates without converting', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                array: Joi.array().items(Joi.string().min(5), Joi.number().min(3))
            ***REMOVED***).strict();

            Helper.validate(schema, [
                [***REMOVED*** array: ['12345'] ***REMOVED***, true],
                [***REMOVED*** array: ['1'] ***REMOVED***, false],
                [***REMOVED*** array: [3] ***REMOVED***, true],
                [***REMOVED*** array: ['12345', 3] ***REMOVED***, true],
                [***REMOVED*** array: ['3'] ***REMOVED***, false],
                [***REMOVED*** array: [1] ***REMOVED***, false]
            ], done);
        ***REMOVED***);

        it('can be disabled', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                array: Joi.array().items(Joi.string().min(5), Joi.number().min(3))
            ***REMOVED***).strict().strict(false);

            Helper.validate(schema, [
                [***REMOVED*** array: ['12345'] ***REMOVED***, true],
                [***REMOVED*** array: ['1'] ***REMOVED***, false],
                [***REMOVED*** array: [3] ***REMOVED***, true],
                [***REMOVED*** array: ['12345', 3] ***REMOVED***, true],
                [***REMOVED*** array: ['3'] ***REMOVED***, true],
                [***REMOVED*** array: [1] ***REMOVED***, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    describe('#options', function () ***REMOVED***

        it('adds to existing options', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED*** b: Joi.number().strict().options(***REMOVED*** convert: true ***REMOVED***) ***REMOVED***);
            var input = ***REMOVED*** b: '2' ***REMOVED***;
            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.b).to.equal(2);
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('throws with an invalid option', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.any().options(***REMOVED*** foo: 'bar' ***REMOVED***);
            ***REMOVED***).to.throw('unknown key foo');
            done();
        ***REMOVED***);

        it('throws with an invalid option type', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.any().options(***REMOVED*** convert: 'yes' ***REMOVED***);
            ***REMOVED***).to.throw('convert should be of type boolean');
            done();
        ***REMOVED***);

        it('throws with an invalid option value', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.any().options(***REMOVED*** presence: 'yes' ***REMOVED***);
            ***REMOVED***).to.throw('presence should be one of required, optional, forbidden, ignore');
            done();
        ***REMOVED***);

        it('does not throw with multiple options including presence key', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.any().options(***REMOVED*** presence: 'optional', raw: true ***REMOVED***);
            ***REMOVED***).to.not.throw();
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#label', function () ***REMOVED***

        it('adds to existing options', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED*** b: Joi.string().email().label('Custom label') ***REMOVED***);
            var input = ***REMOVED*** b: 'not_a_valid_email' ***REMOVED***;
            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.exist();
                expect(err.details[0].message).to.equal('"Custom label" must be a valid email');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('throws when label is missing', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.label();
            ***REMOVED***).to.throw('Label name must be a non-empty string');
            done();
        ***REMOVED***);

        it('can describe a label', function (done) ***REMOVED***

            var schema = Joi.object().label('lbl').describe();
            expect(schema).to.deep.equal(***REMOVED*** type: 'object', label: 'lbl' ***REMOVED***);
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#strict', function () ***REMOVED***

        it('adds to existing options', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED*** b: Joi.number().options(***REMOVED*** convert: true ***REMOVED***).strict() ***REMOVED***);
            var input = ***REMOVED*** b: '2' ***REMOVED***;
            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.exist();
                expect(value.b).to.equal('2');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('#raw', function () ***REMOVED***

        it('gives the raw input', function (done) ***REMOVED***

            var tests = [
                [Joi.array(), '[1,2,3]'],
                [Joi.binary(), 'abc'],
                [Joi.boolean(), 'false'],
                [Joi.date().format('YYYYMMDD'), '19700101'],
                [Joi.number(), '12'],
                [Joi.object(), '***REMOVED*** "a": 1 ***REMOVED***'],
                [Joi.any().strict(), 'abc']
            ];

            tests.forEach(function (test) ***REMOVED***

                var baseSchema = test[0];
                var input = test[1];
                var schemas = [
                    baseSchema.raw(),
                    baseSchema.raw(true),
                    baseSchema.options(***REMOVED*** raw: true ***REMOVED***)
                ];

                schemas.forEach(function (schema) ***REMOVED***

                    schema.raw().validate(input, function (err, value) ***REMOVED***

                        expect(err).to.not.exist();
                        expect(value).to.equal(input);
                    ***REMOVED***);
                ***REMOVED***);
            ***REMOVED***);

            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#default', function () ***REMOVED***

        it('sets the value', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED*** foo: Joi.string().default('test') ***REMOVED***);
            var input = ***REMOVED******REMOVED***;

            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.foo).to.equal('test');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('throws when value is a method and no description is provided', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.object(***REMOVED***
                    foo: Joi.string().default(function () ***REMOVED***

                        return 'test';

                    ***REMOVED***)
                ***REMOVED***);
            ***REMOVED***).to.throw();

            done();
        ***REMOVED***);

        it('allows passing description as a property of a default method', function (done) ***REMOVED***

            var defaultFn = function () ***REMOVED***

                return 'test';
            ***REMOVED***;
            defaultFn.description = 'test';

            expect(function () ***REMOVED***

                Joi.object(***REMOVED*** foo: Joi.string().default(defaultFn) ***REMOVED***);
            ***REMOVED***).to.not.throw();

            done();
        ***REMOVED***);

        it('sets the value when passing a method', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                foo: Joi.string().default(function () ***REMOVED***

                    return 'test';
                ***REMOVED***, 'testing')
            ***REMOVED***);
            var input = ***REMOVED******REMOVED***;

            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.foo).to.equal('test');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('executes the default method each time validate is called', function (done) ***REMOVED***

            var count = 0;
            var schema = Joi.object(***REMOVED***
                foo: Joi.number().default(function () ***REMOVED***

                    return ++count;
                ***REMOVED***, 'incrementer')
            ***REMOVED***);
            var input = ***REMOVED******REMOVED***;

            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.foo).to.equal(1);
            ***REMOVED***);

            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.foo).to.equal(2);
            ***REMOVED***);

            done();
        ***REMOVED***);

        it('passes a clone of the context if the default method accepts an argument', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                foo: Joi.string().default(function (context) ***REMOVED***

                    return context.bar + 'ing';
                ***REMOVED***, 'testing'),
                bar: Joi.string()
            ***REMOVED***);
            var input = ***REMOVED*** bar: 'test' ***REMOVED***;

            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.foo).to.equal('testing');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('does not modify the original object when modifying the clone in a default method', function (done) ***REMOVED***

            var defaultFn = function (context) ***REMOVED***

                context.bar = 'broken';
                return 'test';
            ***REMOVED***;
            defaultFn.description = 'testing';

            var schema = Joi.object(***REMOVED***
                foo: Joi.string().default(defaultFn),
                bar: Joi.string()
            ***REMOVED***);
            var input = ***REMOVED*** bar: 'test' ***REMOVED***;

            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.bar).to.equal('test');
                expect(value.foo).to.equal('test');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('passes undefined as the context if the default method has no parent', function (done) ***REMOVED***

            var c;
            var methodCalled = false;
            var schema = Joi.string().default(function (context) ***REMOVED***

                methodCalled = true;
                c = context;
                return 'test';
            ***REMOVED***, 'testing');

            schema.validate(undefined, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(methodCalled).to.equal(true);
                expect(c).to.equal(undefined);
                expect(value).to.equal('test');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('allows passing a method with no description to default when the object being validated is a function', function (done) ***REMOVED***

            var defaultFn = function () ***REMOVED***

                return 'just a function';
            ***REMOVED***;

            var schema;
            expect(function () ***REMOVED***

                schema = Joi.func().default(defaultFn);
            ***REMOVED***).to.not.throw();

            schema.validate(undefined, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value).to.deep.equal(defaultFn);
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('allows passing a method that generates a default method when validating a function', function (done) ***REMOVED***

            var defaultFn = function () ***REMOVED***

                return 'just a function';
            ***REMOVED***;

            var defaultGeneratorFn = function () ***REMOVED***

                return defaultFn;
            ***REMOVED***;
            defaultGeneratorFn.description = 'generate a default fn';

            var schema;
            expect(function () ***REMOVED***

                schema = Joi.func().default(defaultGeneratorFn);
            ***REMOVED***).to.not.throw();

            schema.validate(undefined, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value).to.deep.equal(defaultFn);
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('allows passing a ref as a default without a description', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.string(),
                b: Joi.string().default(Joi.ref('a'))
            ***REMOVED***);

            schema.validate(***REMOVED*** a: 'test' ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.a).to.equal('test');
                expect(value.b).to.equal('test');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('ignores description when passing a ref as a default', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.string(),
                b: Joi.string().default(Joi.ref('a'), 'this is a ref')
            ***REMOVED***);

            schema.validate(***REMOVED*** a: 'test' ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.a).to.equal('test');
                expect(value.b).to.equal('test');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('catches errors in default methods', function (done) ***REMOVED***

            var defaultFn = function () ***REMOVED***

                throw new Error('boom');
            ***REMOVED***;
            defaultFn.description = 'broken method';

            var schema = Joi.string().default(defaultFn);

            schema.validate(undefined, function (err, value) ***REMOVED***

                expect(err).to.exist();
                expect(err.details).to.have.length(1);
                expect(err.details[0].message).to.contain('threw an error when running default method');
                expect(err.details[0].type).to.equal('any.default');
                expect(err.details[0].context).to.be.an.instanceof(Error);
                expect(err.details[0].context.message).to.equal('boom');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('should not overide a value when value is given', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED*** foo: Joi.string().default('bar') ***REMOVED***);
            var input = ***REMOVED*** foo: 'test' ***REMOVED***;

            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.foo).to.equal('test');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('sets value based on condition (outer)', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.boolean(),
                b: Joi.boolean().default(false).when('a', ***REMOVED*** is: true, then: Joi.required(), otherwise: Joi.forbidden() ***REMOVED***)
            ***REMOVED***);

            schema.validate(***REMOVED*** a: false ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.b).to.equal(false);
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('sets value based on condition (inner)', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.boolean(),
                b: Joi.boolean().when('a', ***REMOVED*** is: true, then: Joi.default(false), otherwise: Joi.forbidden() ***REMOVED***)
            ***REMOVED***);

            schema.validate(***REMOVED*** a: true ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.b).to.equal(false);
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('creates deep defaults', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.number().default(42),
                b: Joi.object(***REMOVED***
                    c: Joi.boolean().default(true),
                    d: Joi.string()
                ***REMOVED***).default()
            ***REMOVED***).default();

            Helper.validate(schema, [
                [undefined, true, null, ***REMOVED*** a: 42, b: ***REMOVED*** c: true ***REMOVED*** ***REMOVED***],
                [***REMOVED*** a: 24 ***REMOVED***, true, null, ***REMOVED*** a: 24, b: ***REMOVED*** c: true ***REMOVED*** ***REMOVED***]
            ], done);
        ***REMOVED***);

        it('should not affect objects other than object when called without an argument', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.number().default()
            ***REMOVED***).default();

            Helper.validate(schema, [
                [undefined, true, null, ***REMOVED******REMOVED***],
                [***REMOVED*** a: 24 ***REMOVED***, true, null, ***REMOVED*** a: 24 ***REMOVED***]
            ], done);
        ***REMOVED***);

        it('should set default value as a clone', function (done) ***REMOVED***

            var defaultValue = ***REMOVED*** bar: 'val' ***REMOVED***;
            var schema = Joi.object(***REMOVED*** foo: Joi.object().default(defaultValue) ***REMOVED***);
            var input = ***REMOVED******REMOVED***;

            schema.validate(input, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.foo).to.not.equal(defaultValue);
                expect(value.foo).to.only.deep.include(***REMOVED*** bar: 'val' ***REMOVED***);

                value.foo.bar = 'mutated';

                schema.validate(input, function (err2, value2) ***REMOVED***

                    expect(err2).to.not.exist();
                    expect(value2.foo).to.not.equal(defaultValue);
                    expect(value2.foo).to.only.deep.include(***REMOVED*** bar: 'val' ***REMOVED***);

                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should not apply default values if the noDefaults option is enquire', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.string().default('foo'),
                b: Joi.number()
            ***REMOVED***);

            var input = ***REMOVED*** b: 42 ***REMOVED***;

            Joi.validate(input, schema, ***REMOVED*** noDefaults: true ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.a).to.not.exist();
                expect(value.b).to.be.equal(42);

                done();
            ***REMOVED***);
        ***REMOVED***);

        it('should not apply default values from functions if the noDefaults option is enquire', function (done) ***REMOVED***

            var func = function (context) ***REMOVED***

                return 'foo';
            ***REMOVED***;

            func.description = 'test parameter';

            var schema = Joi.object(***REMOVED***
                a: Joi.string().default(func),
                b: Joi.number()
            ***REMOVED***);

            var input = ***REMOVED*** b: 42 ***REMOVED***;

            Joi.validate(input, schema, ***REMOVED*** noDefaults: true ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.a).to.not.exist();
                expect(value.b).to.be.equal(42);

                done();
            ***REMOVED***);
        ***REMOVED***);

        it('should not apply default values from references if the noDefaults option is enquire', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.string().default(Joi.ref('b')),
                b: Joi.number()
            ***REMOVED***);

            var input = ***REMOVED*** b: 42 ***REMOVED***;

            Joi.validate(input, schema, ***REMOVED*** noDefaults: true ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value.a).to.not.exist();
                expect(value.b).to.be.equal(42);

                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('#optional', function () ***REMOVED***

        it('validates optional with default required', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.any(),
                b: Joi.any(),
                c: ***REMOVED***
                    d: Joi.any()
                ***REMOVED***
            ***REMOVED***).options(***REMOVED*** presence: 'required' ***REMOVED***);

            Helper.validate(schema, [
                [***REMOVED*** a: 5 ***REMOVED***, false],
                [***REMOVED*** a: 5, b: 6 ***REMOVED***, false],
                [***REMOVED*** a: 5, b: 6, c: ***REMOVED******REMOVED*** ***REMOVED***, false],
                [***REMOVED*** a: 5, b: 6, c: ***REMOVED*** d: 7 ***REMOVED*** ***REMOVED***, true],
                [***REMOVED******REMOVED***, false],
                [***REMOVED*** b: 5 ***REMOVED***, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    describe('#forbidden', function () ***REMOVED***

        it('validates forbidden', function (done) ***REMOVED***

            var schema = ***REMOVED***
                a: Joi.number(),
                b: Joi.forbidden()
            ***REMOVED***;

            Helper.validate(schema, [
                [***REMOVED*** a: 5 ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 6 ***REMOVED***, false],
                [***REMOVED*** a: 'a' ***REMOVED***, false],
                [***REMOVED******REMOVED***, true],
                [***REMOVED*** b: undefined ***REMOVED***, true],
                [***REMOVED*** b: null ***REMOVED***, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    describe('#strip', function () ***REMOVED***

        it('validates and returns undefined', function (done) ***REMOVED***

            var schema = Joi.string().strip();

            schema.validate('test', function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value).to.not.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('validates and returns an error', function (done) ***REMOVED***

            var schema = Joi.string().strip();

            schema.validate(1, function (err, value) ***REMOVED***

                expect(err).to.exist();
                expect(err.message).to.equal('"value" must be a string');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('#description', function () ***REMOVED***

        it('sets the description', function (done) ***REMOVED***

            var b = Joi.description('my description');
            expect(b._description).to.equal('my description');
            done();
        ***REMOVED***);

        it('throws when description is missing', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.description();
            ***REMOVED***).to.throw('Description must be a non-empty string');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#notes', function () ***REMOVED***

        it('sets the notes', function (done) ***REMOVED***

            var b = Joi.notes(['a']).notes('my notes');
            expect(b._notes).to.deep.equal(['a', 'my notes']);
            done();
        ***REMOVED***);

        it('throws when notes are missing', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.notes();
            ***REMOVED***).to.throw('Notes must be a non-empty string or array');
            done();
        ***REMOVED***);

        it('throws when notes are invalid', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.notes(5);
            ***REMOVED***).to.throw('Notes must be a non-empty string or array');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#tags', function () ***REMOVED***

        it('sets the tags', function (done) ***REMOVED***

            var b = Joi.tags(['tag1', 'tag2']).tags('tag3');
            expect(b._tags).to.include('tag1');
            expect(b._tags).to.include('tag2');
            expect(b._tags).to.include('tag3');
            done();
        ***REMOVED***);

        it('throws when tags are missing', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.tags();
            ***REMOVED***).to.throw('Tags must be a non-empty string or array');
            done();
        ***REMOVED***);

        it('throws when tags are invalid', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.tags(5);
            ***REMOVED***).to.throw('Tags must be a non-empty string or array');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#meta', function () ***REMOVED***

        it('sets the meta', function (done) ***REMOVED***

            var meta = ***REMOVED*** prop: 'val', prop2: 3 ***REMOVED***;
            var b = Joi.meta(meta);
            expect(b.describe().meta).to.deep.equal([meta]);

            b = b.meta(***REMOVED*** other: true ***REMOVED***);
            expect(b.describe().meta).to.deep.equal([meta, ***REMOVED***
                other: true
            ***REMOVED***]);

            done();
        ***REMOVED***);

        it('throws when meta is missing', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.meta();
            ***REMOVED***).to.throw('Meta cannot be undefined');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#example', function () ***REMOVED***

        it('sets an example', function (done) ***REMOVED***

            var schema = Joi.valid(5, 6, 7).example(5);
            expect(schema._examples).to.include(5);
            expect(schema.describe().examples).to.deep.equal([5]);
            done();
        ***REMOVED***);

        it('throws when tags are missing', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.example();
            ***REMOVED***).to.throw('Missing example');
            done();
        ***REMOVED***);

        it('throws when example fails own rules', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.valid(5, 6, 7).example(4);
            ***REMOVED***).to.throw('Bad example: "value" must be one of [5, 6, 7]');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#unit', function () ***REMOVED***

        it('sets the unit', function (done) ***REMOVED***

            var b = Joi.unit('milliseconds');
            expect(b._unit).to.equal('milliseconds');
            expect(b.describe().unit).to.equal('milliseconds');
            done();
        ***REMOVED***);

        it('throws when unit is missing', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.unit();
            ***REMOVED***).to.throw('Unit name must be a non-empty string');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#_validate', function () ***REMOVED***

        it('checks value after conversion', function (done) ***REMOVED***

            var schema = Joi.number().invalid(2);
            Joi.validate('2', schema, ***REMOVED*** abortEarly: false ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('#concat', function () ***REMOVED***

        it('throws when schema is not any', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.string().concat(Joi.number());
            ***REMOVED***).to.throw('Cannot merge type string with another type: number');
            done();
        ***REMOVED***);

        it('throws when schema is missing', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.string().concat();
            ***REMOVED***).to.throw('Invalid schema object');
            done();
        ***REMOVED***);

        it('throws when schema is invalid', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.string().concat(1);
            ***REMOVED***).to.throw('Invalid schema object');
            done();
        ***REMOVED***);

        it('merges two schemas (settings)', function (done) ***REMOVED***

            var a = Joi.number().options(***REMOVED*** convert: true ***REMOVED***);
            var b = Joi.options(***REMOVED*** convert: false ***REMOVED***);

            Helper.validate(a, [
                [1, true], ['1', true]
            ]);

            Helper.validate(a.concat(b), [
                [1, true], ['1', false]
            ], done);
        ***REMOVED***);

        it('merges two schemas (valid)', function (done) ***REMOVED***

            var a = Joi.string().valid('a');
            var b = Joi.string().valid('b');

            Helper.validate(a, [
                ['a', true],
                ['b', false]
            ]);

            Helper.validate(b, [
                ['b', true],
                ['a', false]
            ]);

            Helper.validate(a.concat(b), [
                ['a', true],
                ['b', true]
            ], done);
        ***REMOVED***);

        it('merges two schemas (invalid)', function (done) ***REMOVED***

            var a = Joi.string().invalid('a');
            var b = Joi.invalid('b');

            Helper.validate(a, [
                ['b', true], ['a', false]
            ]);

            Helper.validate(b, [
                ['a', true], ['b', false]
            ]);

            Helper.validate(a.concat(b), [
                ['a', false], ['b', false]
            ], done);
        ***REMOVED***);

        it('merges two schemas (valid/invalid)', function (done) ***REMOVED***

            var a = Joi.string().valid('a').invalid('b');
            var b = Joi.string().valid('b').invalid('a');

            Helper.validate(a, [
                ['a', true],
                ['b', false]
            ]);

            Helper.validate(b, [
                ['b', true],
                ['a', false]
            ]);

            Helper.validate(a.concat(b), [
                ['a', false],
                ['b', true]
            ], done);
        ***REMOVED***);

        it('merges two schemas (tests)', function (done) ***REMOVED***

            var a = Joi.number().min(5);
            var b = Joi.number().max(10);

            Helper.validate(a, [
                [4, false], [11, true]
            ]);

            Helper.validate(b, [
                [6, true], [11, false]
            ]);

            Helper.validate(a.concat(b), [
                [4, false], [6, true], [11, false]
            ], done);
        ***REMOVED***);

        it('merges two schemas (flags)', function (done) ***REMOVED***

            var a = Joi.string().valid('a');
            var b = Joi.string().insensitive();

            Helper.validate(a, [
                ['a', true], ['A', false], ['b', false]
            ]);

            Helper.validate(a.concat(b), [
                ['a', true], ['A', true], ['b', false]
            ], done);
        ***REMOVED***);

        it('overrides and append information', function (done) ***REMOVED***

            var a = Joi.description('a').unit('a').tags('a').example('a');
            var b = Joi.description('b').unit('b').tags('b').example('b');

            var desc = a.concat(b).describe();
            expect(desc).to.deep.equal(***REMOVED***
                type: 'any',
                description: 'b',
                tags: ['a', 'b'],
                examples: ['a', 'b'],
                unit: 'b'
            ***REMOVED***);
            done();
        ***REMOVED***);

        it('merges two objects (any key + specific key)', function (done) ***REMOVED***

            var a = Joi.object();
            var b = Joi.object(***REMOVED*** b: 1 ***REMOVED***);

            Helper.validate(a, [
                [***REMOVED*** b: 1 ***REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, true]
            ]);

            Helper.validate(b, [
                [***REMOVED*** b: 1 ***REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, false]
            ]);

            Helper.validate(a.concat(b), [
                [***REMOVED*** b: 1 ***REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, false]
            ]);

            Helper.validate(b.concat(a), [
                [***REMOVED*** b: 1 ***REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, false]
            ], done);
        ***REMOVED***);

        it('merges two objects (no key + any key)', function (done) ***REMOVED***

            var a = Joi.object(***REMOVED******REMOVED***);
            var b = Joi.object();

            Helper.validate(a, [
                [***REMOVED******REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, false]
            ]);

            Helper.validate(b, [
                [***REMOVED******REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, true]
            ]);

            Helper.validate(a.concat(b), [
                [***REMOVED******REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, false]
            ]);

            Helper.validate(b.concat(a), [
                [***REMOVED******REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, false]
            ], done);
        ***REMOVED***);

        it('merges two objects (key + key)', function (done) ***REMOVED***

            var a = Joi.object(***REMOVED*** a: 1 ***REMOVED***);
            var b = Joi.object(***REMOVED*** b: 2 ***REMOVED***);

            Helper.validate(a, [
                [***REMOVED*** a: 1 ***REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, false]
            ]);

            Helper.validate(b, [
                [***REMOVED*** a: 1 ***REMOVED***, false], [***REMOVED*** b: 2 ***REMOVED***, true]
            ]);

            Helper.validate(a.concat(b), [
                [***REMOVED*** a: 1 ***REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, true]
            ]);

            Helper.validate(b.concat(a), [
                [***REMOVED*** a: 1 ***REMOVED***, true], [***REMOVED*** b: 2 ***REMOVED***, true]
            ], done);
        ***REMOVED***);

        it('merges two objects (renames)', function (done) ***REMOVED***

            var a = Joi.object(***REMOVED*** a: 1 ***REMOVED***).rename('c', 'a');
            var b = Joi.object(***REMOVED*** b: 2 ***REMOVED***).rename('d', 'b');

            a.concat(b).validate(***REMOVED*** c: 1, d: 2 ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                expect(value).to.deep.equal(***REMOVED*** a: 1, b: 2 ***REMOVED***);
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('merges two objects (deps)', function (done) ***REMOVED***

            var a = Joi.object(***REMOVED*** a: 1 ***REMOVED***);
            var b = Joi.object(***REMOVED*** b: 2 ***REMOVED***).and('b', 'a');

            a.concat(b).validate(***REMOVED*** a: 1, b: 2 ***REMOVED***, function (err, value) ***REMOVED***

                expect(err).to.not.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('merges two objects (same key)', function (done) ***REMOVED***

            var a = Joi.object(***REMOVED*** a: 1, b: 2, c: 3 ***REMOVED***);
            var b = Joi.object(***REMOVED*** b: 1, c: 2, a: 3 ***REMOVED***);

            var ab = a.concat(b);

            Helper.validate(a, [
                [***REMOVED*** a: 1, b: 2, c: 3 ***REMOVED***, true],
                [***REMOVED*** a: 3, b: 1, c: 2 ***REMOVED***, false]
            ]);

            Helper.validate(b, [
                [***REMOVED*** a: 1, b: 2, c: 3 ***REMOVED***, false],
                [***REMOVED*** a: 3, b: 1, c: 2 ***REMOVED***, true]
            ]);

            Helper.validate(ab, [
                [***REMOVED*** a: 1, b: 2, c: 3 ***REMOVED***, true],
                [***REMOVED*** a: 3, b: 1, c: 2 ***REMOVED***, true],
                [***REMOVED*** a: 1, b: 2, c: 2 ***REMOVED***, true],
                [***REMOVED*** a: 1, b: 2, c: 4 ***REMOVED***, false]
            ], done);
        ***REMOVED***);

        it('throws when schema key types do not match', function (done) ***REMOVED***

            var a = Joi.object(***REMOVED*** a: Joi.number() ***REMOVED***);
            var b = Joi.object(***REMOVED*** a: Joi.string() ***REMOVED***);

            expect(function () ***REMOVED***

                a.concat(b);
            ***REMOVED***).to.throw('Cannot merge type number with another type: string');
            done();
        ***REMOVED***);

        it('merges two alternatives with references', function (done) ***REMOVED***

            var schema = ***REMOVED***
                a: ***REMOVED*** c: Joi.number() ***REMOVED***,
                b: Joi.alternatives(Joi.ref('a.c')).concat(Joi.alternatives(Joi.ref('c'))),
                c: Joi.number()
            ***REMOVED***;

            Helper.validate(schema, [
                [***REMOVED*** a: ***REMOVED******REMOVED*** ***REMOVED***, true],
                [***REMOVED*** a: ***REMOVED*** c: '5' ***REMOVED***, b: 5 ***REMOVED***, true],
                [***REMOVED*** a: ***REMOVED*** c: '5' ***REMOVED***, b: 6, c: '6' ***REMOVED***, true],
                [***REMOVED*** a: ***REMOVED*** c: '5' ***REMOVED***, b: 7, c: '6' ***REMOVED***, false]
            ], done);
        ***REMOVED***);

        it('merges meta properly', function (done) ***REMOVED***

            var metaA = ***REMOVED*** a: 1 ***REMOVED***;
            var metaB = ***REMOVED*** b: 1 ***REMOVED***;
            var a = Joi.any().meta(metaA);
            var b = Joi.any().meta(metaB);
            var c = Joi.any();
            var d = Joi.any();

            expect(a.concat(b)._meta).to.deep.equal([***REMOVED*** a: 1 ***REMOVED***, ***REMOVED*** b: 1 ***REMOVED***]);
            expect(a.concat(c)._meta).to.deep.equal([metaA]);
            expect(b.concat(c)._meta).to.deep.equal([metaB]);
            expect(c.concat(d)._meta).to.deep.equal([]);

            done();
        ***REMOVED***);

        it('merges into an any', function (done) ***REMOVED***

            var a = Joi.any().required();
            var b = Joi.number().only(0);

            expect(function () ***REMOVED***

                a.concat(b);
            ***REMOVED***).to.not.throw();

            var schema = a.concat(b);
            expect(schema.validate().error.message).to.equal('"value" is required');
            expect(schema.validate(1).error.message).to.equal('"value" must be one of [0]');

            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#when', function () ***REMOVED***

        it('throws when options are invalid', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.when('a');
            ***REMOVED***).to.throw('Invalid options');
            done();
        ***REMOVED***);

        it('forks type into alternatives', function (done) ***REMOVED***

            var schema = ***REMOVED***
                a: Joi.any(),
                b: Joi.string().valid('x').when('a', ***REMOVED*** is: 5, then: Joi.valid('y'), otherwise: Joi.valid('z') ***REMOVED***)
            ***REMOVED***;

            Helper.validate(schema, [
                [***REMOVED*** a: 5, b: 'x' ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 'y' ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 'z' ***REMOVED***, false],
                [***REMOVED*** a: 1, b: 'x' ***REMOVED***, true],
                [***REMOVED*** a: 1, b: 'y' ***REMOVED***, false],
                [***REMOVED*** a: 1, b: 'z' ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 'a' ***REMOVED***, false],
                [***REMOVED*** b: 'a' ***REMOVED***, false]
            ], done);
        ***REMOVED***);

        it('forks type into alternatives (only then)', function (done) ***REMOVED***

            var schema = ***REMOVED***
                a: Joi.any(),
                b: Joi.string().valid('x').when('a', ***REMOVED*** is: 5, then: Joi.valid('y') ***REMOVED***)
            ***REMOVED***;

            Helper.validate(schema, [
                [***REMOVED*** a: 5, b: 'x' ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 'y' ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 'z' ***REMOVED***, false],
                [***REMOVED*** a: 1, b: 'x' ***REMOVED***, true],
                [***REMOVED*** a: 1, b: 'y' ***REMOVED***, false],
                [***REMOVED*** a: 1, b: 'z' ***REMOVED***, false],
                [***REMOVED*** a: 5, b: 'a' ***REMOVED***, false],
                [***REMOVED*** b: 'a' ***REMOVED***, false]
            ], done);
        ***REMOVED***);

        it('forks type into alternatives (only otherwise)', function (done) ***REMOVED***

            var schema = ***REMOVED***
                a: Joi.any(),
                b: Joi.string().valid('x').when('a', ***REMOVED*** is: 5, otherwise: Joi.valid('z') ***REMOVED***)
            ***REMOVED***;

            Helper.validate(schema, [
                [***REMOVED*** a: 5, b: 'x' ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 'y' ***REMOVED***, false],
                [***REMOVED*** a: 5, b: 'z' ***REMOVED***, false],
                [***REMOVED*** a: 1, b: 'x' ***REMOVED***, true],
                [***REMOVED*** a: 1, b: 'y' ***REMOVED***, false],
                [***REMOVED*** a: 1, b: 'z' ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 'a' ***REMOVED***, false],
                [***REMOVED*** b: 'a' ***REMOVED***, false]
            ], done);
        ***REMOVED***);

        it('forks type into alternatives (with a schema)', function (done) ***REMOVED***

            var schema = ***REMOVED***
                a: Joi.any(),
                b: Joi.string().valid('x').when('a', ***REMOVED*** is: Joi.number().only(5).required(), then: Joi.valid('y') ***REMOVED***)
            ***REMOVED***;

            Helper.validate(schema, [
                [***REMOVED*** a: 5, b: 'x' ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 'y' ***REMOVED***, true],
                [***REMOVED*** a: 5, b: 'z' ***REMOVED***, false],
                [***REMOVED*** a: 1, b: 'x' ***REMOVED***, true],
                [***REMOVED*** a: 1, b: 'y' ***REMOVED***, false],
                [***REMOVED*** a: 1, b: 'z' ***REMOVED***, false],
                [***REMOVED*** a: 5, b: 'a' ***REMOVED***, false],
                [***REMOVED*** b: 'a' ***REMOVED***, false]
            ], done);
        ***REMOVED***);

        it('makes peer required', function (done) ***REMOVED***

            var schema = ***REMOVED***
                a: Joi.when('b', ***REMOVED*** is: 5, then: Joi.required() ***REMOVED***),
                b: Joi.any()
            ***REMOVED***;

            Helper.validate(schema, [
                [***REMOVED*** b: 5 ***REMOVED***, false],
                [***REMOVED*** b: 6 ***REMOVED***, true],
                [***REMOVED*** a: 'b' ***REMOVED***, true],
                [***REMOVED*** b: 5, a: 'x' ***REMOVED***, true]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    describe('#requiredKeys', function () ***REMOVED***

        it('should set keys as required', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED*** a: 0, b: 0, c: ***REMOVED*** d: 0, e: ***REMOVED*** f: 0 ***REMOVED*** ***REMOVED***, g: ***REMOVED*** h: 0 ***REMOVED*** ***REMOVED***)
                .requiredKeys('a', 'b', 'c.d', 'c.e.f', 'g');
            Helper.validate(schema, [
                [***REMOVED******REMOVED***, false],
                [***REMOVED*** a: 0 ***REMOVED***, false],
                [***REMOVED*** a: 0, b: 0 ***REMOVED***, false],
                [***REMOVED*** a: 0, b: 0, g: ***REMOVED******REMOVED*** ***REMOVED***, true],
                [***REMOVED*** a: 0, b: 0, c: ***REMOVED******REMOVED***, g: ***REMOVED******REMOVED*** ***REMOVED***, false],
                [***REMOVED*** a: 0, b: 0, c: ***REMOVED*** d: 0 ***REMOVED***, g: ***REMOVED******REMOVED*** ***REMOVED***, true],
                [***REMOVED*** a: 0, b: 0, c: ***REMOVED*** d: 0, e: ***REMOVED******REMOVED*** ***REMOVED***, g: ***REMOVED******REMOVED*** ***REMOVED***, false],
                [***REMOVED*** a: 0, b: 0, c: ***REMOVED*** d: 0, e: ***REMOVED*** f: 0 ***REMOVED*** ***REMOVED***, g: ***REMOVED******REMOVED*** ***REMOVED***, true]
            ], done);
        ***REMOVED***);

        it('should work on types other than objects', function (done) ***REMOVED***

            var schemas = [Joi.array(), Joi.binary(), Joi.boolean(), Joi.date(), Joi.func(), Joi.number(), Joi.string()];
            schemas.forEach(function (schema) ***REMOVED***

                expect(function () ***REMOVED***

                    schema.applyFunctionToChildren([''], 'required');
                ***REMOVED***).to.not.throw();

                expect(function () ***REMOVED***

                    schema.applyFunctionToChildren(['', 'a'], 'required');
                ***REMOVED***).to.throw();

                expect(function () ***REMOVED***

                    schema.applyFunctionToChildren(['a'], 'required');
                ***REMOVED***).to.throw();
            ***REMOVED***);

            done();
        ***REMOVED***);

        it('should throw on unknown key', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.object(***REMOVED*** a: 0, b: 0 ***REMOVED***).requiredKeys('a', 'c', 'b', 'd', 'd.e.f');
            ***REMOVED***).to.throw(Error, 'unknown key(s) c, d');

            expect(function () ***REMOVED***

                Joi.object(***REMOVED*** a: 0, b: 0 ***REMOVED***).requiredKeys('a', 'b', 'a.c.d');
            ***REMOVED***).to.throw(Error, 'unknown key(s) a.c.d');

            done();
        ***REMOVED***);

        it('should throw on empty object', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.object().requiredKeys('a', 'c', 'b', 'd');
            ***REMOVED***).to.throw(Error, 'unknown key(s) a, b, c, d');
            done();
        ***REMOVED***);

        it('should not modify original object', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED*** a: 0 ***REMOVED***);
            var requiredSchema = schema.requiredKeys('a');
            schema.validate(***REMOVED******REMOVED***, function (err) ***REMOVED***

                expect(err).to.not.exist();

                requiredSchema.validate(***REMOVED******REMOVED***, function (err) ***REMOVED***

                    expect(err).to.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('#optionalKeys', function () ***REMOVED***

        it('should set keys as optional', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED*** a: Joi.number().required(), b: Joi.number().required() ***REMOVED***).optionalKeys('a', 'b');
            Helper.validate(schema, [
                [***REMOVED******REMOVED***, true],
                [***REMOVED*** a: 0 ***REMOVED***, true],
                [***REMOVED*** a: 0, b: 0 ***REMOVED***, true]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    describe('#empty', function () ***REMOVED***

        it('should void values when considered empty', function (done) ***REMOVED***

            var schema = Joi.string().empty('');
            Helper.validate(schema, [
                [undefined, true, null, undefined],
                ['abc', true, null, 'abc'],
                ['', true, null, undefined]
            ], done);
        ***REMOVED***);

        it('should override any previous empty', function (done) ***REMOVED***

            var schema = Joi.string().empty('').empty('abc');
            Helper.validate(schema, [
                [undefined, true, null, undefined],
                ['abc', true, null, undefined],
                ['', false, null, '"value" is not allowed to be empty'],
                ['def', true, null, 'def']
            ], done);
        ***REMOVED***);

        it('should be possible to reset the empty value', function (done) ***REMOVED***

            var schema = Joi.string().empty('').empty();
            Helper.validate(schema, [
                [undefined, true, null, undefined],
                ['abc', true, null, 'abc'],
                ['', false, null, '"value" is not allowed to be empty']
            ], done);
        ***REMOVED***);

        it('should have no effect if only reset is used', function (done) ***REMOVED***

            var schema = Joi.string().empty();
            Helper.validate(schema, [
                [undefined, true, null, undefined],
                ['abc', true, null, 'abc'],
                ['', false, null, '"value" is not allowed to be empty']
            ], done);
        ***REMOVED***);

        it('should work with dependencies', function (done) ***REMOVED***

            var schema = Joi.object(***REMOVED***
                a: Joi.string().empty(''),
                b: Joi.string().empty('')
            ***REMOVED***).or('a', 'b');

            Helper.validate(schema, [
                [***REMOVED******REMOVED***, false, null, '"value" must contain at least one of [a, b]'],
                [***REMOVED*** a: '' ***REMOVED***, false, null, '"value" must contain at least one of [a, b]'],
                [***REMOVED*** a: 'a' ***REMOVED***, true, null, ***REMOVED*** a: 'a' ***REMOVED***],
                [***REMOVED*** a: '', b: 'b' ***REMOVED***, true, null, ***REMOVED*** b: 'b' ***REMOVED***]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    describe('Set', function () ***REMOVED***

        describe('#add', function () ***REMOVED***

            it('throws when adding a non ref function', function (done) ***REMOVED***

                expect(function () ***REMOVED***

                    Joi.valid(function () ***REMOVED*** ***REMOVED***);
                ***REMOVED***).to.throw('Value cannot be an object or function');
                done();
            ***REMOVED***);

            it('throws when adding an object function', function (done) ***REMOVED***

                expect(function () ***REMOVED***

                    Joi.valid(***REMOVED******REMOVED***);
                ***REMOVED***).to.throw('Value cannot be an object or function');
                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('#has', function () ***REMOVED***

            it('compares date to null', function (done) ***REMOVED***

                var any = Joi.any().clone();
                any._valids.add(null);
                expect(any._valids.has(new Date())).to.equal(false);
                done();
            ***REMOVED***);

            it('compares buffer to null', function (done) ***REMOVED***

                var any = Joi.any().clone();
                any._valids.add(null);
                expect(any._valids.has(new Buffer(''))).to.equal(false);
                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('#values', function () ***REMOVED***

            it('returns array', function (done) ***REMOVED***

                var a = Joi.any().valid('x').invalid('y');
                var b = a.invalid('x');
                expect(a._valids.values().length).to.equal(1);
                expect(b._valids.values().length).to.equal(0);
                expect(a._invalids.values().length).to.equal(1);
                expect(b._invalids.values().length).to.equal(2);
                done();
            ***REMOVED***);

            it('strips undefined', function (done) ***REMOVED***

                var any = Joi.any().clone();
                any._valids.add(undefined);
                expect(any._valids.values(***REMOVED*** stripUndefined: true ***REMOVED***)).to.not.include(undefined);
                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('#allow', function () ***REMOVED***

            it('allows valid values to be set', function (done) ***REMOVED***

                expect(function () ***REMOVED***

                    Joi.any().allow(true, 1, 'hello', new Date());
                ***REMOVED***).not.to.throw();
                done();
            ***REMOVED***);

            it('throws when passed undefined', function (done) ***REMOVED***

                expect(function () ***REMOVED***

                    Joi.any().allow(undefined);
                ***REMOVED***).to.throw(Error, 'Cannot call allow/valid/invalid with undefined');
                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('#valid', function () ***REMOVED***

            it('allows valid values to be set', function (done) ***REMOVED***

                expect(function () ***REMOVED***

                    Joi.any().valid(true, 1, 'hello', new Date());
                ***REMOVED***).not.to.throw();
                done();
            ***REMOVED***);

            it('throws when passed undefined', function (done) ***REMOVED***

                expect(function () ***REMOVED***

                    Joi.any().valid(undefined);
                ***REMOVED***).to.throw(Error, 'Cannot call allow/valid/invalid with undefined');
                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('#invalid', function () ***REMOVED***

            it('allows invalid values to be set', function (done) ***REMOVED***

                expect(function () ***REMOVED***

                    Joi.any().valid(true, 1, 'hello', new Date());
                ***REMOVED***).not.to.throw();
                done();
            ***REMOVED***);

            it('throws when passed undefined', function (done) ***REMOVED***

                expect(function () ***REMOVED***

                    Joi.any().invalid(undefined);
                ***REMOVED***).to.throw('Cannot call allow/valid/invalid with undefined');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***);
