// joi/test/index.js
// Load modules

var Lab = require('lab');
var Code = require('code');
var Joi = require('../lib');
var Helper = require('./helper');


// Declare internals

var internals = ***REMOVED******REMOVED***;


// Test shortcuts

var lab = exports.lab = Lab.script();
var describe = lab.describe;
var it = lab.it;
var expect = Code.expect;


describe('Joi', function () ***REMOVED***

    it('validates object', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            a: Joi.number().min(0).max(3),
            b: Joi.string().valid('a', 'b', 'c'),
            c: Joi.string().email().optional()
        ***REMOVED***).without('a', 'none');

        var obj = ***REMOVED***
            a: 1,
            b: 'a',
            c: 'joe@example.com'
        ***REMOVED***;

        schema.validate(obj, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('keeps schema immutable', function (done) ***REMOVED***

        var a = Joi.string();
        var b = a.valid('b');

        Helper.validate(a, [
            ['a', true],
            ['b', true],
            [5, false]
        ], function () ***REMOVED***

            Helper.validate(b, [
                ['a', false],
                ['b', true],
                [5, false]
            ], done);
        ***REMOVED***);

    ***REMOVED***);

    it('validates null', function (done) ***REMOVED***

        Joi.string().validate(null, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(err.annotate()).to.equal('***REMOVED***\n  \u001b[41m\"value\"\u001b[0m\u001b[31m [1]: -- missing --\u001b[0m\n***REMOVED***\n\u001b[31m\n[1] "value" must be a string\u001b[0m');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates null schema', function (done) ***REMOVED***

        Helper.validate(null, [
            ['a', false],
            [null, true]
        ], done);
    ***REMOVED***);

    it('validates number literal', function (done) ***REMOVED***

        Helper.validate(5, [
            [6, false],
            [5, true]
        ], done);
    ***REMOVED***);

    it('validates string literal', function (done) ***REMOVED***

        Helper.validate('5', [
            ['6', false],
            ['5', true]
        ], done);
    ***REMOVED***);

    it('validates boolean literal', function (done) ***REMOVED***

        Helper.validate(true, [
            [false, false],
            [true, true]
        ], done);
    ***REMOVED***);

    it('validates date literal', function (done) ***REMOVED***

        var now = Date.now();
        Helper.validate(new Date(now), [
            [new Date(now), true],
            [now, true],
            [now * 2, false]
        ], done);
    ***REMOVED***);

    it('validates complex literal', function (done) ***REMOVED***

        var schema = ['key', 5, ***REMOVED*** a: true, b: [/^a/, 'boom'] ***REMOVED***];
        Helper.validate(schema, [
            ['key', true],
            [5, true],
            ['other', false],
            [6, false],
            [***REMOVED*** c: 5 ***REMOVED***, false],
            [***REMOVED******REMOVED***, true],
            [***REMOVED*** b: 'abc' ***REMOVED***, true],
            [***REMOVED*** a: true, b: 'boom' ***REMOVED***, true],
            [***REMOVED*** a: 5, b: 'a' ***REMOVED***, false]
        ], done);
    ***REMOVED***);

    it('validates a compiled complex literal', function (done) ***REMOVED***

        var schema = Joi.compile(['key', 5, ***REMOVED*** a: true, b: [/^a/, 'boom'] ***REMOVED***]);
        Helper.validate(schema, [
            ['key', true],
            [5, true],
            ['other', false],
            [6, false],
            [***REMOVED*** c: 5 ***REMOVED***, false],
            [***REMOVED******REMOVED***, true],
            [***REMOVED*** b: 'abc' ***REMOVED***, true],
            [***REMOVED*** a: true, b: 'boom' ***REMOVED***, true],
            [***REMOVED*** a: 5, b: 'a' ***REMOVED***, false]
        ], done);
    ***REMOVED***);

    it('validates regex directly', function (done) ***REMOVED***

        Joi.compile(/^5$/).validate('5', function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            Joi.compile(/.***REMOVED***2***REMOVED***/).validate('6', function (err2, value2) ***REMOVED***

                expect(err2).to.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('validated with', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            txt: Joi.string(),
            upc: Joi.string()
        ***REMOVED***).with('txt', 'upc');

        Joi.validate(***REMOVED*** txt: 'a' ***REMOVED***, schema, ***REMOVED*** abortEarly: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err.message).to.equal('"txt" missing required peer "upc"');

            Helper.validate(schema, [
                [***REMOVED*** upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: undefined ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: 'test' ***REMOVED***, true]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    it('validated without', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            txt: Joi.string(),
            upc: Joi.string()
        ***REMOVED***).without('txt', 'upc');

        Joi.validate(***REMOVED*** txt: 'a', upc: 'b' ***REMOVED***, schema, ***REMOVED*** abortEarly: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err.message).to.equal('"txt" conflict with forbidden peer "upc"');

            Helper.validate(schema, [
                [***REMOVED*** upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: undefined ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: 'test' ***REMOVED***, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    it('validates xor', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            txt: Joi.string(),
            upc: Joi.string()
        ***REMOVED***).xor('txt', 'upc');

        Joi.validate(***REMOVED******REMOVED***, schema, ***REMOVED*** abortEarly: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err.message).to.equal('"value" must contain at least one of [txt, upc]');

            Helper.validate(schema, [
                [***REMOVED*** upc: null ***REMOVED***, false],
                [***REMOVED*** upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: null ***REMOVED***, false],
                [***REMOVED*** txt: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: null, upc: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: undefined, upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: undefined ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: undefined ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: '' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: 'test' ***REMOVED***, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    it('validates multiple peers xor', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            txt: Joi.string(),
            upc: Joi.string(),
            code: Joi.string()
        ***REMOVED***).xor('txt', 'upc', 'code');

        Helper.validate(schema, [
            [***REMOVED*** upc: 'test' ***REMOVED***, true],
            [***REMOVED*** txt: 'test' ***REMOVED***, true],
            [***REMOVED******REMOVED***, false]
        ], done);
    ***REMOVED***);

    it('validates xor with number types', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            code: Joi.number(),
            upc: Joi.number()
        ***REMOVED***).xor('code', 'upc');

        Helper.validate(schema, [
            [***REMOVED*** upc: 123 ***REMOVED***, true],
            [***REMOVED*** code: 456 ***REMOVED***, true],
            [***REMOVED*** code: 456, upc: 123 ***REMOVED***, false],
            [***REMOVED******REMOVED***, false]
        ], done);
    ***REMOVED***);

    it('validates xor when empty value of peer allowed', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            code: Joi.string(),
            upc: Joi.string().allow('')
        ***REMOVED***).xor('code', 'upc');

        Helper.validate(schema, [
            [***REMOVED*** upc: '' ***REMOVED***, true],
            [***REMOVED*** upc: '123' ***REMOVED***, true],
            [***REMOVED*** code: '456' ***REMOVED***, true],
            [***REMOVED*** code: '456', upc: '' ***REMOVED***, false],
            [***REMOVED******REMOVED***, false]
        ], done);
    ***REMOVED***);

    it('validates or', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            txt: Joi.string(),
            upc: Joi.string().allow(null, ''),
            code: Joi.number()
        ***REMOVED***).or('txt', 'upc', 'code');

        Joi.validate(***REMOVED******REMOVED***, schema, ***REMOVED*** abortEarly: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err.message).to.equal('"value" must contain at least one of [txt, upc, code]');

            Helper.validate(schema, [
                [***REMOVED*** upc: null ***REMOVED***, true],
                [***REMOVED*** upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: null ***REMOVED***, false],
                [***REMOVED*** txt: 'test' ***REMOVED***, true],
                [***REMOVED*** code: null ***REMOVED***, false],
                [***REMOVED*** code: 123 ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, true],
                [***REMOVED*** txt: '', upc: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: null, upc: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: undefined, upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: undefined ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, true],
                [***REMOVED*** txt: '', upc: undefined ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: undefined, code: 999 ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: undefined, code: undefined ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: '' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: 'test', code: 322 ***REMOVED***, true]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    it('validates and', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            txt: Joi.string(),
            upc: Joi.string().allow(null, ''),
            code: Joi.number()
        ***REMOVED***).and('txt', 'upc', 'code');

        Joi.validate(***REMOVED*** txt: 'x' ***REMOVED***, schema, ***REMOVED*** abortEarly: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err.message).to.equal('"value" contains [txt] without its required peers [upc, code]');

            Helper.validate(schema, [
                [***REMOVED******REMOVED***, true],
                [***REMOVED*** upc: null ***REMOVED***, false],
                [***REMOVED*** upc: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: null ***REMOVED***, false],
                [***REMOVED*** txt: 'test' ***REMOVED***, false],
                [***REMOVED*** code: null ***REMOVED***, false],
                [***REMOVED*** code: 123 ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: null, upc: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: undefined, upc: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: undefined ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: undefined ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: undefined, code: 999 ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: undefined, code: undefined ***REMOVED***, false],
                [***REMOVED*** txt: '', upc: '' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: 'test' ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: 'test', code: 322 ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: null, code: 322 ***REMOVED***, true]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    it('validates nand()', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            txt: Joi.string(),
            upc: Joi.string().allow(null, ''),
            code: Joi.number()
        ***REMOVED***).nand('txt', 'upc', 'code');

        Joi.validate(***REMOVED*** txt: 'x', upc: 'y', code: 123 ***REMOVED***, schema, ***REMOVED*** abortEarly: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err.message).to.equal('"txt" must not exist simultaneously with [upc, code]');

            Helper.validate(schema, [
                [***REMOVED******REMOVED***, true],
                [***REMOVED*** upc: null ***REMOVED***, true],
                [***REMOVED*** upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test' ***REMOVED***, true],
                [***REMOVED*** code: 123 ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, true],
                [***REMOVED*** txt: undefined, upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: undefined ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: '' ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: null ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: undefined, code: 999 ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: 'test' ***REMOVED***, true],
                [***REMOVED*** txt: 'test', upc: 'test', code: 322 ***REMOVED***, false],
                [***REMOVED*** txt: 'test', upc: null, code: 322 ***REMOVED***, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    it('validates an array of valid types', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            auth: [
                Joi.object(***REMOVED***
                    mode: Joi.string().valid('required', 'optional', 'try').allow(null)
                ***REMOVED***).allow(null),
                Joi.string(),
                Joi.boolean()
            ]
        ***REMOVED***);

        schema.validate(***REMOVED*** auth: ***REMOVED*** mode: 'none' ***REMOVED*** ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(err.message).to.equal('child "auth" fails because [child "mode" fails because ["mode" must be one of [required, optional, try, null]], "auth" must be a string, "auth" must be a boolean]');

            Helper.validate(schema, [
                [***REMOVED*** auth: ***REMOVED*** mode: 'try' ***REMOVED*** ***REMOVED***, true],
                [***REMOVED*** something: undefined ***REMOVED***, false],
                [***REMOVED*** auth: ***REMOVED*** something: undefined ***REMOVED*** ***REMOVED***, false],
                [***REMOVED*** auth: null ***REMOVED***, true],
                [***REMOVED*** auth: undefined ***REMOVED***, true],
                [***REMOVED******REMOVED***, true],
                [***REMOVED*** auth: true ***REMOVED***, true],
                [***REMOVED*** auth: 123 ***REMOVED***, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    it('validates alternatives', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            auth: Joi.alternatives(
                Joi.object(***REMOVED***
                    mode: Joi.string().valid('required', 'optional', 'try').allow(null)
                ***REMOVED***).allow(null),
                Joi.string(),
                Joi.boolean()
            )
        ***REMOVED***);

        schema.validate(***REMOVED*** auth: ***REMOVED*** mode: 'none' ***REMOVED*** ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(err.message).to.equal('child "auth" fails because [child "mode" fails because ["mode" must be one of [required, optional, try, null]], "auth" must be a string, "auth" must be a boolean]');

            Helper.validate(schema, [
                [***REMOVED*** auth: ***REMOVED*** mode: 'try' ***REMOVED*** ***REMOVED***, true],
                [***REMOVED*** something: undefined ***REMOVED***, false],
                [***REMOVED*** auth: ***REMOVED*** something: undefined ***REMOVED*** ***REMOVED***, false],
                [***REMOVED*** auth: null ***REMOVED***, true],
                [***REMOVED*** auth: undefined ***REMOVED***, true],
                [***REMOVED******REMOVED***, true],
                [***REMOVED*** auth: true ***REMOVED***, true],
                [***REMOVED*** auth: 123 ***REMOVED***, false]
            ], done);
        ***REMOVED***);
    ***REMOVED***);

    it('validates required alternatives', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: Joi.alternatives(
                Joi.string().required(),
                Joi.boolean().required()
            )
        ***REMOVED***;

        Helper.validate(schema, [
            [***REMOVED*** a: null ***REMOVED***, false],
            [***REMOVED*** a: undefined ***REMOVED***, true],
            [***REMOVED******REMOVED***, true],
            [***REMOVED*** a: true ***REMOVED***, true],
            [***REMOVED*** a: 'true' ***REMOVED***, true],
            [***REMOVED*** a: 123 ***REMOVED***, false],
            [***REMOVED*** a: ***REMOVED*** c: 1 ***REMOVED*** ***REMOVED***, false],
            [***REMOVED*** b: undefined ***REMOVED***, false]
        ], done);
    ***REMOVED***);

    it('validates required [] alternatives', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: [
                Joi.string().required(),
                Joi.boolean().required()
            ]
        ***REMOVED***;

        Helper.validate(schema, [
            [***REMOVED*** a: null ***REMOVED***, false],
            [***REMOVED*** a: undefined ***REMOVED***, true],
            [***REMOVED******REMOVED***, true],
            [***REMOVED*** a: true ***REMOVED***, true],
            [***REMOVED*** a: 'true' ***REMOVED***, true],
            [***REMOVED*** a: 123 ***REMOVED***, false],
            [***REMOVED*** a: ***REMOVED*** c: 1 ***REMOVED*** ***REMOVED***, false],
            [***REMOVED*** b: undefined ***REMOVED***, false]
        ], done);
    ***REMOVED***);

    it('validates an array of string with valid', function (done) ***REMOVED***

        var schema = ***REMOVED***
            brand: Joi.array().items(Joi.string().valid('amex', 'visa'))
        ***REMOVED***;

        Helper.validate(schema, [
            [***REMOVED*** brand: ['amex'] ***REMOVED***, true],
            [***REMOVED*** brand: ['visa', 'mc'] ***REMOVED***, false]
        ], done);
    ***REMOVED***);

    it('validates pre and post convert value', function (done) ***REMOVED***

        var schema = Joi.number().valid(5);

        Helper.validate(schema, [
            [5, true],
            ['5', true]
        ], done);
    ***REMOVED***);

    it('does not change object when validation fails', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            a: Joi.number().valid(2)
        ***REMOVED***);

        var obj = ***REMOVED***
            a: '5'
        ***REMOVED***;

        schema.validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(value.a).to.equal('5');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('does not set optional keys when missing', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            a: Joi.number()
        ***REMOVED***);

        var obj = ***REMOVED******REMOVED***;

        schema.validate(obj, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            expect(value.hasOwnProperty('a')).to.equal(false);
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('invalidates pre and post convert value', function (done) ***REMOVED***

        var schema = Joi.number().invalid(5);

        Helper.validate(schema, [
            [5, false],
            ['5', false]
        ], done);
    ***REMOVED***);

    it('invalidates missing peers', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            username: Joi.string(),
            password: Joi.string()
        ***REMOVED***).with('username', 'password').without('password', 'access_token');

        schema.validate(***REMOVED*** username: 'bob' ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates config where the root item is a joi type', function (done) ***REMOVED***

        Joi.boolean().allow(null).validate(true, function (err, value) ***REMOVED***

            expect(err).to.be.null();
            Joi.object().validate(***REMOVED*** auth: ***REMOVED*** mode: 'try' ***REMOVED*** ***REMOVED***, function (err2, value2) ***REMOVED***

                expect(err2).to.be.null();

                Joi.object().validate(true, function (err3, value3) ***REMOVED***

                    expect(err3.message).to.contain('"value" must be an object');

                    Joi.string().validate(true, function (err4, value4) ***REMOVED***

                        expect(err4.message).to.contain('"value" must be a string');

                        Joi.string().email().validate('test@test.com', function (err5, value5) ***REMOVED***

                            expect(err5).to.be.null();
                            Joi.object(***REMOVED*** param: Joi.string().required() ***REMOVED***).validate(***REMOVED*** param: 'item' ***REMOVED***, function (err6, value6) ***REMOVED***

                                expect(err6).to.be.null();
                                done();
                            ***REMOVED***);
                        ***REMOVED***);
                    ***REMOVED***);
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('converts string to number', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            a: Joi.number()
        ***REMOVED***);

        var input = ***REMOVED*** a: '5' ***REMOVED***;
        schema.validate(input, function (err, value) ***REMOVED***

            expect(err).to.be.null();
            expect(value.a).to.equal(5);
            expect(input.a).to.equal('5');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('allows unknown keys in objects if no schema was given', function (done) ***REMOVED***

        Joi.object().validate(***REMOVED*** foo: 'bar' ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails on unknown keys in objects if a schema was given', function (done) ***REMOVED***

        Joi.object(***REMOVED******REMOVED***).validate(***REMOVED*** foo: 'bar' ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(err.message).to.equal('"foo" is not allowed');

            Joi.compile(***REMOVED******REMOVED***).validate(***REMOVED*** foo: 'bar' ***REMOVED***, function (err2, value2) ***REMOVED***

                expect(err2).to.exist();
                expect(err2.message).to.equal('"foo" is not allowed');

                Joi.compile(***REMOVED*** other: Joi.number() ***REMOVED***).validate(***REMOVED*** foo: 'bar' ***REMOVED***, function (err3, value3) ***REMOVED***

                    expect(err3).to.exist();
                    expect(err3.message).to.equal('"foo" is not allowed');

                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('validates an unknown option', function (done) ***REMOVED***

        var config = ***REMOVED***
            auth: Joi.object(***REMOVED***
                mode: Joi.string().valid('required', 'optional', 'try').allow(null)
            ***REMOVED***).allow(null)
        ***REMOVED***;

        Joi.compile(config).validate(***REMOVED*** auth: ***REMOVED*** unknown: true ***REMOVED*** ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.not.be.null();
            expect(err.message).to.contain('"unknown" is not allowed');

            Joi.compile(config).validate(***REMOVED*** something: false ***REMOVED***, function (err2, value2) ***REMOVED***

                expect(err2).to.not.be.null();
                expect(err2.message).to.contain('"something" is not allowed');

                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('validates required key with multiple options', function (done) ***REMOVED***

        var config = ***REMOVED***
            module: Joi.alternatives([
                Joi.object(***REMOVED***
                    compile: Joi.func().required(),
                    execute: Joi.func()
                ***REMOVED***),
                Joi.string()
            ]).required()
        ***REMOVED***;

        Joi.compile(config).validate(***REMOVED******REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(err.message).to.contain('"module" is required');

            Joi.compile(config).validate(***REMOVED*** module: 'test' ***REMOVED***, function (err2, value2) ***REMOVED***

                expect(err2).to.be.null();

                Joi.compile(config).validate(***REMOVED*** module: ***REMOVED******REMOVED*** ***REMOVED***, function (err3, value3) ***REMOVED***

                    expect(err3).to.not.be.null();
                    expect(err3.message).to.contain('"compile" is required');
                    expect(err3.message).to.contain('"module" must be a string');

                    Joi.compile(config).validate(***REMOVED*** module: ***REMOVED*** compile: function () ***REMOVED*** ***REMOVED*** ***REMOVED*** ***REMOVED***, function (err4, value4) ***REMOVED***

                        expect(err4).to.be.null();
                        done();
                    ***REMOVED***);
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('validates key with required alternatives', function (done) ***REMOVED***

        var config = ***REMOVED***
            module: Joi.alt().try(
                Joi.object(***REMOVED***
                    compile: Joi.func().required(),
                    execute: Joi.func()
                ***REMOVED***).required(),
                Joi.string().required()
            )
        ***REMOVED***;

        Joi.compile(config).validate(***REMOVED******REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates required key with alternatives', function (done) ***REMOVED***

        var config = ***REMOVED***
            module: Joi.alt().try(
                Joi.object(***REMOVED***
                    compile: Joi.func().required(),
                    execute: Joi.func()
                ***REMOVED***),
                Joi.string()
            ).required()
        ***REMOVED***;

        Joi.compile(config).validate(***REMOVED******REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(err.message).to.contain('"module" is required');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('does not require optional numbers', function (done) ***REMOVED***

        var config = ***REMOVED***
            position: Joi.number(),
            suggestion: Joi.string()
        ***REMOVED***;

        Joi.compile(config).validate(***REMOVED*** suggestion: 'something' ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.be.null();

            Joi.compile(config).validate(***REMOVED*** position: 1 ***REMOVED***, function (err2, value2) ***REMOVED***

                expect(err2).to.be.null();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('does not require optional objects', function (done) ***REMOVED***

        var config = ***REMOVED***
            position: Joi.number(),
            suggestion: Joi.object()
        ***REMOVED***;

        Joi.compile(config).validate(***REMOVED*** suggestion: ***REMOVED******REMOVED*** ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.be.null();

            Joi.compile(config).validate(***REMOVED*** position: 1 ***REMOVED***, function (err2, value2) ***REMOVED***

                expect(err2).to.be.null();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('validates object successfully when config has an array of types', function (done) ***REMOVED***

        var schema = ***REMOVED***
            f: [Joi.number(), Joi.boolean()],
            g: [Joi.string(), Joi.object()]
        ***REMOVED***;

        var obj = ***REMOVED***
            f: true,
            g: 'test'
        ***REMOVED***;

        Joi.compile(schema).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates object successfully when config allows for optional key and key is missing', function (done) ***REMOVED***

        var schema = ***REMOVED***
            h: Joi.number(),
            i: Joi.string(),
            j: Joi.object()
        ***REMOVED***;

        var obj = ***REMOVED***
            h: 12,
            i: 'test'
        ***REMOVED***;

        Joi.compile(schema).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: Joi.number().min(0).max(3),
            b: Joi.string().valid('a', 'b', 'c'),
            c: Joi.string().email().optional()
        ***REMOVED***;

        var obj = ***REMOVED***
            a: 10,
            b: 'a',
            c: 'joe@example.com'
        ***REMOVED***;

        Joi.compile(schema).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation when the wrong types are supplied', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: Joi.number().min(0).max(3),
            b: Joi.string().valid('a', 'b', 'c'),
            c: Joi.string().email().optional()
        ***REMOVED***;

        var obj = ***REMOVED***
            a: 'a',
            b: 'a',
            c: 'joe@example.com'
        ***REMOVED***;

        Joi.compile(schema).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation when missing a required parameter', function (done) ***REMOVED***

        var obj = ***REMOVED***
            c: 10
        ***REMOVED***;

        Joi.compile(***REMOVED*** a: Joi.string().required() ***REMOVED***).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation when missing a required parameter within an object config', function (done) ***REMOVED***

        var obj = ***REMOVED***
            a: ***REMOVED******REMOVED***
        ***REMOVED***;

        Joi.compile(***REMOVED*** a: Joi.object(***REMOVED*** b: Joi.string().required() ***REMOVED***) ***REMOVED***).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation when parameter is required to be an object but is given as string', function (done) ***REMOVED***

        var obj = ***REMOVED***
            a: 'a string'
        ***REMOVED***;

        Joi.compile(***REMOVED*** a: Joi.object(***REMOVED*** b: Joi.string().required() ***REMOVED***) ***REMOVED***).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates when parameter is required to be an object and is given correctly as a json string', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: Joi.object(***REMOVED***
                b: Joi.string().required()
            ***REMOVED***)
        ***REMOVED***;

        var input = ***REMOVED***
            a: '***REMOVED***"b":"string"***REMOVED***'
        ***REMOVED***;

        Joi.validate(input, schema, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            expect(input.a).to.equal('***REMOVED***"b":"string"***REMOVED***');
            expect(value.a.b).to.equal('string');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation when parameter is required to be an object but is given as a json string that is incorrect (number instead of string)', function (done) ***REMOVED***

        var obj = ***REMOVED***
            a: '***REMOVED***"b":2***REMOVED***'
        ***REMOVED***;

        Joi.object(***REMOVED*** a: Joi.object(***REMOVED*** b: Joi.string().required() ***REMOVED***) ***REMOVED***).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation when parameter is required to be an Array but is given as string', function (done) ***REMOVED***

        var obj = ***REMOVED***
            a: 'an array'
        ***REMOVED***;

        Joi.object(***REMOVED*** a: Joi.array() ***REMOVED***).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates when parameter is required to be an Array and is given correctly as a json string', function (done) ***REMOVED***

        var obj = ***REMOVED***
            a: '[1,2]'
        ***REMOVED***;

        Joi.object(***REMOVED*** a: Joi.array() ***REMOVED***).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.be.null();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation when parameter is required to be an Array but is given as a json that is incorrect (object instead of array)', function (done) ***REMOVED***

        var obj = ***REMOVED***
            a: '***REMOVED***"b":2***REMOVED***'
        ***REMOVED***;

        Joi.object(***REMOVED*** a: Joi.object(***REMOVED*** b: Joi.string().required() ***REMOVED***) ***REMOVED***).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation when config is an array and fails', function (done) ***REMOVED***

        var schema = ***REMOVED***
            d: [Joi.string(), Joi.boolean()],
            e: [Joi.number(), Joi.object()]
        ***REMOVED***;

        var obj = ***REMOVED***
            d: 10,
            e: 'a'
        ***REMOVED***;

        Joi.compile(schema).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation when config is an array and fails with extra keys', function (done) ***REMOVED***

        var schema = ***REMOVED***
            d: [Joi.string(), Joi.boolean()],
            e: [Joi.number(), Joi.object()]
        ***REMOVED***;

        var obj = ***REMOVED***
            a: 10,
            b: 'a'
        ***REMOVED***;

        Joi.compile(schema).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails validation with extra keys', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: Joi.number()
        ***REMOVED***;

        var obj = ***REMOVED***
            a: 1,
            b: 'a'
        ***REMOVED***;

        Joi.compile(schema).validate(obj, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates missing optional key with string condition', function (done) ***REMOVED***

        var schema = ***REMOVED***
            key: Joi.string().alphanum(false).min(8)
        ***REMOVED***;

        Joi.compile(schema).validate(***REMOVED******REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates with extra keys and remove them when stripUnknown is set', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: Joi.number().min(0).max(3),
            b: Joi.string().valid('a', 'b', 'c'),
            c: Joi.string().email().optional()
        ***REMOVED***;

        var obj = ***REMOVED***
            a: 1,
            b: 'a',
            d: 'c'
        ***REMOVED***;

        Joi.validate(obj, schema, ***REMOVED*** stripUnknown: true, allowUnknown: true ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.be.null();
            expect(value).to.deep.equal(***REMOVED*** a: 1, b: 'a' ***REMOVED***);
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates dependencies when stripUnknown is set', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            a: Joi.number(),
            b: Joi.string()
        ***REMOVED***).and('a', 'b');

        var obj = ***REMOVED***
            a: 1,
            foo: 'bar'
        ***REMOVED***;

        Joi.validate(obj, schema, ***REMOVED*** stripUnknown: true ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(err.message).to.equal('"value" contains [a] without its required peers [b]');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('fails to validate with incorrect property when asked to strip unkown keys without aborting early', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: Joi.number().min(0).max(3),
            b: Joi.string().valid('a', 'b', 'c'),
            c: Joi.string().email().optional()
        ***REMOVED***;

        var obj = ***REMOVED***
            a: 1,
            b: 'f',
            d: 'c'
        ***REMOVED***;

        Joi.validate(obj, schema, ***REMOVED*** stripUnknown: true, abortEarly: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should pass validation with extra keys when allowUnknown is set', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: Joi.number().min(0).max(3),
            b: Joi.string().valid('a', 'b', 'c'),
            c: Joi.string().email().optional()
        ***REMOVED***;

        var obj = ***REMOVED***
            a: 1,
            b: 'a',
            d: 'c'
        ***REMOVED***;

        Joi.validate(obj, schema, ***REMOVED*** allowUnknown: true ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.be.null();
            expect(value).to.deep.equal(***REMOVED*** a: 1, b: 'a', d: 'c' ***REMOVED***);
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should pass validation with extra keys set', function (done) ***REMOVED***

        var localConfig = Joi.object(***REMOVED***
            a: Joi.number().min(0).max(3),
            b: Joi.string().valid('a', 'b', 'c')
        ***REMOVED***).options(***REMOVED*** allowUnknown: true ***REMOVED***);

        var obj = ***REMOVED***
            a: 1,
            b: 'a',
            d: 'c'
        ***REMOVED***;

        localConfig.validate(obj, function (err, value) ***REMOVED***

            expect(err).to.be.null();
            expect(value).to.deep.equal(***REMOVED*** a: 1, b: 'a', d: 'c' ***REMOVED***);

            localConfig.validate(value, function (err2, value2) ***REMOVED***

                expect(err2).to.be.null();
                expect(value2).to.deep.equal(***REMOVED*** a: 1, b: 'a', d: 'c' ***REMOVED***);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);


    it('should pass validation with extra keys and remove them when skipExtraKeys is set locally', function (done) ***REMOVED***

        var localConfig = Joi.object(***REMOVED***
            a: Joi.number().min(0).max(3),
            b: Joi.string().valid('a', 'b', 'c')
        ***REMOVED***).options(***REMOVED*** stripUnknown: true, allowUnknown: true ***REMOVED***);

        var obj = ***REMOVED***
            a: 1,
            b: 'a',
            d: 'c'
        ***REMOVED***;

        localConfig.validate(obj, function (err, value) ***REMOVED***

            expect(err).to.be.null();
            expect(value).to.deep.equal(***REMOVED*** a: 1, b: 'a' ***REMOVED***);

            localConfig.validate(value, function (err2, value2) ***REMOVED***

                expect(err2).to.be.null();
                expect(value2).to.deep.equal(***REMOVED*** a: 1, b: 'a' ***REMOVED***);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('should work when the skipFunctions setting is enabled', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED*** username: Joi.string() ***REMOVED***).options(***REMOVED*** skipFunctions: true ***REMOVED***);
        var input = ***REMOVED*** username: 'test', func: function () ***REMOVED*** ***REMOVED*** ***REMOVED***;
        Joi.validate(input, schema, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should work when the skipFunctions setting is disabled', function (done) ***REMOVED***

        var schema = ***REMOVED*** username: Joi.string() ***REMOVED***;
        var input = ***REMOVED*** username: 'test', func: function () ***REMOVED*** ***REMOVED*** ***REMOVED***;

        Joi.validate(input, schema, ***REMOVED*** skipFunctions: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(err.message).to.contain('"func" is not allowed');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should not convert values when convert is false', function (done) ***REMOVED***

        var schema = ***REMOVED***
            arr: Joi.array().items(Joi.string())
        ***REMOVED***;

        var input = ***REMOVED*** arr: 'foo' ***REMOVED***;
        Joi.validate(input, schema, ***REMOVED*** convert: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('full errors when abortEarly is false', function (done) ***REMOVED***

        var schema = ***REMOVED***
            a: Joi.string(),
            b: Joi.string()
        ***REMOVED***;

        var input = ***REMOVED*** a: 1, b: 2 ***REMOVED***;

        Joi.validate(input, schema, function (errOne, valueOne) ***REMOVED***

            Joi.validate(input, schema, ***REMOVED*** abortEarly: false ***REMOVED***, function (errFull, valueFull) ***REMOVED***

                expect(errOne).to.exist();
                expect(errFull).to.exist();
                expect(errFull.details.length).to.be.greaterThan(errOne.details.length);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('errors multiple times when abortEarly is false in a complex object', function (done) ***REMOVED***

        var schema = Joi.object(***REMOVED***
            test: Joi.array().items(Joi.object().keys(***REMOVED***
                foo: Joi.string().required().max(3),
                bar: Joi.string().max(5)
            ***REMOVED***)),
            test2: Joi.object(***REMOVED***
                test3: Joi.array().items(Joi.object().keys(***REMOVED***
                    foo: Joi.string().required().max(3),
                    bar: Joi.string().max(5),
                    baz: Joi.object(***REMOVED***
                        test4: Joi.array().items(Joi.object().keys(***REMOVED***
                            foo: Joi.string().required().max(3),
                            bar: Joi.string().max(5)
                        ***REMOVED***))
                    ***REMOVED***)
                ***REMOVED***))
            ***REMOVED***)
        ***REMOVED***);

        var input = ***REMOVED***
            test: [***REMOVED***
                foo: 'test1',
                bar: 'testfailed'
            ***REMOVED***],
            test2: ***REMOVED***
                test3: [***REMOVED***
                    foo: '123'
                ***REMOVED***, ***REMOVED***
                    foo: 'test1',
                    bar: 'testfailed'
                ***REMOVED***, ***REMOVED***
                    foo: '123',
                    baz: ***REMOVED***
                        test4: [***REMOVED***
                            foo: 'test1',
                            baz: '123'
                        ***REMOVED***]
                    ***REMOVED***
                ***REMOVED***]
            ***REMOVED***
        ***REMOVED***;

        Joi.validate(input, schema, ***REMOVED*** abortEarly: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            expect(err.details).to.have.length(6);
            expect(err.details).to.deep.equal([***REMOVED***
                message: '"foo" length must be less than or equal to 3 characters long',
                path: 'test.0.foo',
                type: 'string.max',
                context: ***REMOVED*** limit: 3, value: 'test1', key: 'foo', encoding: undefined ***REMOVED***
            ***REMOVED***, ***REMOVED***
                message: '"bar" length must be less than or equal to 5 characters long',
                path: 'test.0.bar',
                type: 'string.max',
                context: ***REMOVED*** limit: 5, value: 'testfailed', key: 'bar', encoding: undefined ***REMOVED***
            ***REMOVED***, ***REMOVED***
                message: '"foo" length must be less than or equal to 3 characters long',
                path: 'test2.test3.1.foo',
                type: 'string.max',
                context: ***REMOVED*** limit: 3, value: 'test1', key: 'foo', encoding: undefined ***REMOVED***
            ***REMOVED***, ***REMOVED***
                message: '"bar" length must be less than or equal to 5 characters long',
                path: 'test2.test3.1.bar',
                type: 'string.max',
                context: ***REMOVED*** limit: 5, value: 'testfailed', key: 'bar', encoding: undefined ***REMOVED***
            ***REMOVED***, ***REMOVED***
                message: '"foo" length must be less than or equal to 3 characters long',
                path: 'test2.test3.2.baz.test4.0.foo',
                type: 'string.max',
                context: ***REMOVED*** limit: 3, value: 'test1', key: 'foo', encoding: undefined ***REMOVED***
            ***REMOVED***, ***REMOVED***
                message: '"baz" is not allowed',
                path: 'test2.test3.2.baz.test4.0.baz',
                type: 'object.allowUnknown',
                context: ***REMOVED*** key: 'baz' ***REMOVED***
            ***REMOVED***]);
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates using the root any object', function (done) ***REMOVED***

        var any = Joi;
        any.validate('abc', function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('validates using the root any object (no callback)', function (done) ***REMOVED***

        var any = Joi;
        var result = any.validate('abc');
        expect(result.error).to.not.exist();
        expect(result.value).to.equal('abc');
        done();
    ***REMOVED***);

    it('accepts no options', function (done) ***REMOVED***

        Joi.validate('test', Joi.string(), function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('accepts no options (no callback)', function (done) ***REMOVED***

        var result = Joi.validate('test', Joi.string());
        expect(result.error).to.not.exist();
        expect(result.value).to.equal('test');
        done();
    ***REMOVED***);

    it('accepts options', function (done) ***REMOVED***

        Joi.validate('5', Joi.number(), ***REMOVED*** convert: false ***REMOVED***, function (err, value) ***REMOVED***

            expect(err).to.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('accepts options (no callback)', function (done) ***REMOVED***

        var result = Joi.validate('5', Joi.number(), ***REMOVED*** convert: false ***REMOVED***);
        expect(result.error).to.exist();
        done();
    ***REMOVED***);

    it('accepts null options', function (done) ***REMOVED***

        Joi.validate('test', Joi.string(), null, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('accepts undefined options', function (done) ***REMOVED***

        Joi.validate('test', Joi.string(), undefined, function (err, value) ***REMOVED***

            expect(err).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#describe', function () ***REMOVED***

        var defaultFn = function () ***REMOVED***

            return 'test';
        ***REMOVED***;
        defaultFn.description = 'testing';

        var defaultDescribedFn = function () ***REMOVED***

            return 'test';
        ***REMOVED***;

        var defaultRef = Joi.ref('xor');

        var schema = Joi.object(***REMOVED***
            sub: ***REMOVED***
                email: Joi.string().email(),
                date: Joi.date(),
                child: Joi.object(***REMOVED***
                    alphanum: Joi.string().alphanum()
                ***REMOVED***)
            ***REMOVED***,
            min: [Joi.number(), Joi.string().min(3)],
            max: Joi.string().max(3),
            required: Joi.string().required(),
            xor: Joi.string(),
            renamed: Joi.string().valid('456'),
            notEmpty: Joi.string().required().description('a').notes('b').tags('c'),
            empty: Joi.string().empty('').strip(),
            defaultRef: Joi.string().default(defaultRef, 'not here'),
            defaultFn: Joi.string().default(defaultFn, 'not here'),
            defaultDescribedFn: Joi.string().default(defaultDescribedFn, 'described test')
        ***REMOVED***).rename('renamed', 'required').without('required', 'xor').without('xor', 'required');

        var result = ***REMOVED***
            type: 'object',
            children: ***REMOVED***
                sub: ***REMOVED***
                    type: 'object',
                    children: ***REMOVED***
                        email: ***REMOVED***
                            type: 'string',
                            invalids: [''],
                            rules: [***REMOVED*** name: 'email' ***REMOVED***]
                        ***REMOVED***,
                        date: ***REMOVED***
                            type: 'date'
                        ***REMOVED***,
                        child: ***REMOVED***
                            type: 'object',
                            children: ***REMOVED***
                                alphanum: ***REMOVED***
                                    type: 'string',
                                    invalids: [''],
                                    rules: [***REMOVED*** name: 'alphanum' ***REMOVED***]
                                ***REMOVED***
                            ***REMOVED***
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***,
                min: ***REMOVED***
                    type: 'alternatives',
                    alternatives: [
                        ***REMOVED***
                            type: 'number',
                            invalids: [Infinity, -Infinity]
                        ***REMOVED***,
                        ***REMOVED***
                            type: 'string',
                            invalids: [''],
                            rules: [***REMOVED*** name: 'min', arg: 3 ***REMOVED***]
                        ***REMOVED***
                    ]
                ***REMOVED***,
                max: ***REMOVED***
                    type: 'string',
                    invalids: [''],
                    rules: [***REMOVED*** name: 'max', arg: 3 ***REMOVED***]
                ***REMOVED***,
                required: ***REMOVED***
                    type: 'string',
                    flags: ***REMOVED***
                        presence: 'required'
                    ***REMOVED***,
                    invalids: ['']
                ***REMOVED***,
                xor: ***REMOVED***
                    type: 'string',
                    invalids: ['']
                ***REMOVED***,
                renamed: ***REMOVED***
                    type: 'string',
                    flags: ***REMOVED***
                        allowOnly: true
                    ***REMOVED***,
                    valids: ['456'],
                    invalids: ['']
                ***REMOVED***,
                notEmpty: ***REMOVED***
                    type: 'string',
                    flags: ***REMOVED***
                        presence: 'required'
                    ***REMOVED***,
                    description: 'a',
                    notes: ['b'],
                    tags: ['c'],
                    invalids: ['']
                ***REMOVED***,
                empty: ***REMOVED***
                    type: 'string',
                    flags: ***REMOVED***
                        empty: ***REMOVED***
                            type: 'string',
                            flags: ***REMOVED***
                                allowOnly: true
                            ***REMOVED***,
                            valids: ['']
                        ***REMOVED***,
                        strip: true
                    ***REMOVED***,
                    invalids: ['']
                ***REMOVED***,
                defaultRef: ***REMOVED***
                    type: 'string',
                    flags: ***REMOVED***
                        default: defaultRef
                    ***REMOVED***,
                    invalids: ['']
                ***REMOVED***,
                defaultFn: ***REMOVED***
                    type: 'string',
                    flags: ***REMOVED***
                        default: defaultFn
                    ***REMOVED***,
                    invalids: ['']
                ***REMOVED***,
                defaultDescribedFn: ***REMOVED***
                    type: 'string',
                    flags: ***REMOVED***
                        default: defaultDescribedFn
                    ***REMOVED***,
                    invalids: ['']
                ***REMOVED***
            ***REMOVED***,
            dependencies: [
                ***REMOVED***
                    type: 'without',
                    key: 'required',
                    peers: ['xor']
                ***REMOVED***,
                ***REMOVED***
                    type: 'without',
                    key: 'xor',
                    peers: ['required']
                ***REMOVED***
            ]
        ***REMOVED***;

        it('describes schema (direct)', function (done) ***REMOVED***

            var description = schema.describe();
            expect(description).to.deep.equal(result);
            expect(description.children.defaultRef.flags.default.description).to.not.exist();
            expect(description.children.defaultFn.flags.default.description).to.equal('testing');
            expect(description.children.defaultDescribedFn.flags.default.description).to.equal('described test');
            done();
        ***REMOVED***);

        it('describes schema (root)', function (done) ***REMOVED***

            var description = Joi.describe(schema);
            expect(description).to.deep.equal(result);
            done();
        ***REMOVED***);

        it('describes schema (any)', function (done) ***REMOVED***

            var any = Joi;
            var description = any.describe();
            expect(description).to.deep.equal(***REMOVED***
                type: 'any'
            ***REMOVED***);
            done();
        ***REMOVED***);

        it('describes schema without invalids', function (done) ***REMOVED***

            var description = Joi.allow(null).describe();
            expect(description.invalids).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#assert', function () ***REMOVED***

        it('does not have a return value', function (done) ***REMOVED***

            var result;
            expect(function () ***REMOVED***

                result = Joi.assert('4', Joi.number());
            ***REMOVED***).to.not.throw();
            expect(result).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#attempt', function () ***REMOVED***

        it('throws on invalid value', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.attempt('x', Joi.number());
            ***REMOVED***).to.throw('"x"\n\u001b[31m\n[1] "value" must be a number\u001b[0m');
            done();
        ***REMOVED***);

        it('does not throw on valid value', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.attempt('4', Joi.number());
            ***REMOVED***).to.not.throw();
            done();
        ***REMOVED***);

        it('returns validated structure', function (done) ***REMOVED***

            var valid;
            expect(function () ***REMOVED***

                valid = Joi.attempt('4', Joi.number());
            ***REMOVED***).to.not.throw();
            expect(valid).to.equal(4);
            done();
        ***REMOVED***);

        it('throws on invalid value with message', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.attempt('x', Joi.number(), 'the reason is');
            ***REMOVED***).to.throw('the reason is "x"\n\u001b[31m\n[1] "value" must be a number\u001b[0m');
            done();
        ***REMOVED***);

        it('throws on invalid value with message as error', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.attempt('x', Joi.number(), new Error('invalid value'));
            ***REMOVED***).to.throw('invalid value');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('#compile', function () ***REMOVED***

        it('throws an error on invalid value', function (done) ***REMOVED***

            expect(function () ***REMOVED***

                Joi.compile(undefined);
            ***REMOVED***).to.throw(Error, 'Invalid schema content: ');
            done();
        ***REMOVED***);

        it('shows path to errors in object', function (done) ***REMOVED***

            var schema = ***REMOVED***
                a: ***REMOVED***
                    b: ***REMOVED***
                        c: ***REMOVED***
                            d: undefined
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***;

            expect(function () ***REMOVED***

                Joi.compile(schema);
            ***REMOVED***).to.throw(Error, 'Invalid schema content: (a.b.c.d)');
            done();
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***);
