// bluebird/js/browser/bluebird.js
/* @preserve
 * The MIT License (MIT)
 * 
 * Copyright (c) 2013-2015 Petka Antonov
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * 
 */
/**
 * bluebird build version 2.10.2
 * Features enabled: core, race, call_get, generators, map, nodeify, promisify, props, reduce, settle, some, cancel, using, filter, any, each, timers
*/
!function(e)***REMOVED***if("object"==typeof exports&&"undefined"!=typeof module)module.exports=e();else if("function"==typeof define&&define.amd)define([],e);else***REMOVED***var f;"undefined"!=typeof window?f=window:"undefined"!=typeof global?f=global:"undefined"!=typeof self&&(f=self),f.Promise=e()***REMOVED******REMOVED***(function()***REMOVED***var define,module,exports;return (function e(t,n,r)***REMOVED***function s(o,u)***REMOVED***if(!n[o])***REMOVED***if(!t[o])***REMOVED***var a=typeof _dereq_=="function"&&_dereq_;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f***REMOVED***var l=n[o]=***REMOVED***exports:***REMOVED******REMOVED******REMOVED***;t[o][0].call(l.exports,function(e)***REMOVED***var n=t[o][1][e];return s(n?n:e)***REMOVED***,l,l.exports,e,t,n,r)***REMOVED***return n[o].exports***REMOVED***var i=typeof _dereq_=="function"&&_dereq_;for(var o=0;o<r.length;o++)s(r[o]);return s***REMOVED***)(***REMOVED***1:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise) ***REMOVED***
var SomePromiseArray = Promise._SomePromiseArray;
function any(promises) ***REMOVED***
    var ret = new SomePromiseArray(promises);
    var promise = ret.promise();
    ret.setHowMany(1);
    ret.setUnwrap();
    ret.init();
    return promise;
***REMOVED***

Promise.any = function (promises) ***REMOVED***
    return any(promises);
***REMOVED***;

Promise.prototype.any = function () ***REMOVED***
    return any(this);
***REMOVED***;

***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],2:[function(_dereq_,module,exports)***REMOVED***
"use strict";
var firstLineError;
try ***REMOVED***throw new Error(); ***REMOVED*** catch (e) ***REMOVED***firstLineError = e;***REMOVED***
var schedule = _dereq_("./schedule.js");
var Queue = _dereq_("./queue.js");
var util = _dereq_("./util.js");

function Async() ***REMOVED***
    this._isTickUsed = false;
    this._lateQueue = new Queue(16);
    this._normalQueue = new Queue(16);
    this._trampolineEnabled = true;
    var self = this;
    this.drainQueues = function () ***REMOVED***
        self._drainQueues();
    ***REMOVED***;
    this._schedule =
        schedule.isStatic ? schedule(this.drainQueues) : schedule;
***REMOVED***

Async.prototype.disableTrampolineIfNecessary = function() ***REMOVED***
    if (util.hasDevTools) ***REMOVED***
        this._trampolineEnabled = false;
    ***REMOVED***
***REMOVED***;

Async.prototype.enableTrampoline = function() ***REMOVED***
    if (!this._trampolineEnabled) ***REMOVED***
        this._trampolineEnabled = true;
        this._schedule = function(fn) ***REMOVED***
            setTimeout(fn, 0);
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;

Async.prototype.haveItemsQueued = function () ***REMOVED***
    return this._normalQueue.length() > 0;
***REMOVED***;

Async.prototype.throwLater = function(fn, arg) ***REMOVED***
    if (arguments.length === 1) ***REMOVED***
        arg = fn;
        fn = function () ***REMOVED*** throw arg; ***REMOVED***;
    ***REMOVED***
    if (typeof setTimeout !== "undefined") ***REMOVED***
        setTimeout(function() ***REMOVED***
            fn(arg);
        ***REMOVED***, 0);
    ***REMOVED*** else try ***REMOVED***
        this._schedule(function() ***REMOVED***
            fn(arg);
        ***REMOVED***);
    ***REMOVED*** catch (e) ***REMOVED***
        throw new Error("No async scheduler available\u000a\u000a    See http://goo.gl/m3OTXk\u000a");
    ***REMOVED***
***REMOVED***;

function AsyncInvokeLater(fn, receiver, arg) ***REMOVED***
    this._lateQueue.push(fn, receiver, arg);
    this._queueTick();
***REMOVED***

function AsyncInvoke(fn, receiver, arg) ***REMOVED***
    this._normalQueue.push(fn, receiver, arg);
    this._queueTick();
***REMOVED***

function AsyncSettlePromises(promise) ***REMOVED***
    this._normalQueue._pushOne(promise);
    this._queueTick();
***REMOVED***

if (!util.hasDevTools) ***REMOVED***
    Async.prototype.invokeLater = AsyncInvokeLater;
    Async.prototype.invoke = AsyncInvoke;
    Async.prototype.settlePromises = AsyncSettlePromises;
***REMOVED*** else ***REMOVED***
    if (schedule.isStatic) ***REMOVED***
        schedule = function(fn) ***REMOVED*** setTimeout(fn, 0); ***REMOVED***;
    ***REMOVED***
    Async.prototype.invokeLater = function (fn, receiver, arg) ***REMOVED***
        if (this._trampolineEnabled) ***REMOVED***
            AsyncInvokeLater.call(this, fn, receiver, arg);
        ***REMOVED*** else ***REMOVED***
            this._schedule(function() ***REMOVED***
                setTimeout(function() ***REMOVED***
                    fn.call(receiver, arg);
                ***REMOVED***, 100);
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;

    Async.prototype.invoke = function (fn, receiver, arg) ***REMOVED***
        if (this._trampolineEnabled) ***REMOVED***
            AsyncInvoke.call(this, fn, receiver, arg);
        ***REMOVED*** else ***REMOVED***
            this._schedule(function() ***REMOVED***
                fn.call(receiver, arg);
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;

    Async.prototype.settlePromises = function(promise) ***REMOVED***
        if (this._trampolineEnabled) ***REMOVED***
            AsyncSettlePromises.call(this, promise);
        ***REMOVED*** else ***REMOVED***
            this._schedule(function() ***REMOVED***
                promise._settlePromises();
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
***REMOVED***

Async.prototype.invokeFirst = function (fn, receiver, arg) ***REMOVED***
    this._normalQueue.unshift(fn, receiver, arg);
    this._queueTick();
***REMOVED***;

Async.prototype._drainQueue = function(queue) ***REMOVED***
    while (queue.length() > 0) ***REMOVED***
        var fn = queue.shift();
        if (typeof fn !== "function") ***REMOVED***
            fn._settlePromises();
            continue;
        ***REMOVED***
        var receiver = queue.shift();
        var arg = queue.shift();
        fn.call(receiver, arg);
    ***REMOVED***
***REMOVED***;

Async.prototype._drainQueues = function () ***REMOVED***
    this._drainQueue(this._normalQueue);
    this._reset();
    this._drainQueue(this._lateQueue);
***REMOVED***;

Async.prototype._queueTick = function () ***REMOVED***
    if (!this._isTickUsed) ***REMOVED***
        this._isTickUsed = true;
        this._schedule(this.drainQueues);
    ***REMOVED***
***REMOVED***;

Async.prototype._reset = function () ***REMOVED***
    this._isTickUsed = false;
***REMOVED***;

module.exports = new Async();
module.exports.firstLineError = firstLineError;

***REMOVED***,***REMOVED***"./queue.js":28,"./schedule.js":31,"./util.js":38***REMOVED***],3:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, INTERNAL, tryConvertToPromise) ***REMOVED***
var rejectThis = function(_, e) ***REMOVED***
    this._reject(e);
***REMOVED***;

var targetRejected = function(e, context) ***REMOVED***
    context.promiseRejectionQueued = true;
    context.bindingPromise._then(rejectThis, rejectThis, null, this, e);
***REMOVED***;

var bindingResolved = function(thisArg, context) ***REMOVED***
    if (this._isPending()) ***REMOVED***
        this._resolveCallback(context.target);
    ***REMOVED***
***REMOVED***;

var bindingRejected = function(e, context) ***REMOVED***
    if (!context.promiseRejectionQueued) this._reject(e);
***REMOVED***;

Promise.prototype.bind = function (thisArg) ***REMOVED***
    var maybePromise = tryConvertToPromise(thisArg);
    var ret = new Promise(INTERNAL);
    ret._propagateFrom(this, 1);
    var target = this._target();

    ret._setBoundTo(maybePromise);
    if (maybePromise instanceof Promise) ***REMOVED***
        var context = ***REMOVED***
            promiseRejectionQueued: false,
            promise: ret,
            target: target,
            bindingPromise: maybePromise
        ***REMOVED***;
        target._then(INTERNAL, targetRejected, ret._progress, ret, context);
        maybePromise._then(
            bindingResolved, bindingRejected, ret._progress, ret, context);
    ***REMOVED*** else ***REMOVED***
        ret._resolveCallback(target);
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.prototype._setBoundTo = function (obj) ***REMOVED***
    if (obj !== undefined) ***REMOVED***
        this._bitField = this._bitField | 131072;
        this._boundTo = obj;
    ***REMOVED*** else ***REMOVED***
        this._bitField = this._bitField & (~131072);
    ***REMOVED***
***REMOVED***;

Promise.prototype._isBound = function () ***REMOVED***
    return (this._bitField & 131072) === 131072;
***REMOVED***;

Promise.bind = function (thisArg, value) ***REMOVED***
    var maybePromise = tryConvertToPromise(thisArg);
    var ret = new Promise(INTERNAL);

    ret._setBoundTo(maybePromise);
    if (maybePromise instanceof Promise) ***REMOVED***
        maybePromise._then(function() ***REMOVED***
            ret._resolveCallback(value);
        ***REMOVED***, ret._reject, ret._progress, ret, null);
    ***REMOVED*** else ***REMOVED***
        ret._resolveCallback(value);
    ***REMOVED***
    return ret;
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],4:[function(_dereq_,module,exports)***REMOVED***
"use strict";
var old;
if (typeof Promise !== "undefined") old = Promise;
function noConflict() ***REMOVED***
    try ***REMOVED*** if (Promise === bluebird) Promise = old; ***REMOVED***
    catch (e) ***REMOVED******REMOVED***
    return bluebird;
***REMOVED***
var bluebird = _dereq_("./promise.js")();
bluebird.noConflict = noConflict;
module.exports = bluebird;

***REMOVED***,***REMOVED***"./promise.js":23***REMOVED***],5:[function(_dereq_,module,exports)***REMOVED***
"use strict";
var cr = Object.create;
if (cr) ***REMOVED***
    var callerCache = cr(null);
    var getterCache = cr(null);
    callerCache[" size"] = getterCache[" size"] = 0;
***REMOVED***

module.exports = function(Promise) ***REMOVED***
var util = _dereq_("./util.js");
var canEvaluate = util.canEvaluate;
var isIdentifier = util.isIdentifier;

var getMethodCaller;
var getGetter;
if (!true) ***REMOVED***
var makeMethodCaller = function (methodName) ***REMOVED***
    return new Function("ensureMethod", "                                    \n\
        return function(obj) ***REMOVED***                                               \n\
            'use strict'                                                     \n\
            var len = this.length;                                           \n\
            ensureMethod(obj, 'methodName');                                 \n\
            switch(len) ***REMOVED***                                                    \n\
                case 1: return obj.methodName(this[0]);                      \n\
                case 2: return obj.methodName(this[0], this[1]);             \n\
                case 3: return obj.methodName(this[0], this[1], this[2]);    \n\
                case 0: return obj.methodName();                             \n\
                default:                                                     \n\
                    return obj.methodName.apply(obj, this);                  \n\
            ***REMOVED***                                                                \n\
        ***REMOVED***;                                                                   \n\
        ".replace(/methodName/g, methodName))(ensureMethod);
***REMOVED***;

var makeGetter = function (propertyName) ***REMOVED***
    return new Function("obj", "                                             \n\
        'use strict';                                                        \n\
        return obj.propertyName;                                             \n\
        ".replace("propertyName", propertyName));
***REMOVED***;

var getCompiled = function(name, compiler, cache) ***REMOVED***
    var ret = cache[name];
    if (typeof ret !== "function") ***REMOVED***
        if (!isIdentifier(name)) ***REMOVED***
            return null;
        ***REMOVED***
        ret = compiler(name);
        cache[name] = ret;
        cache[" size"]++;
        if (cache[" size"] > 512) ***REMOVED***
            var keys = Object.keys(cache);
            for (var i = 0; i < 256; ++i) delete cache[keys[i]];
            cache[" size"] = keys.length - 256;
        ***REMOVED***
    ***REMOVED***
    return ret;
***REMOVED***;

getMethodCaller = function(name) ***REMOVED***
    return getCompiled(name, makeMethodCaller, callerCache);
***REMOVED***;

getGetter = function(name) ***REMOVED***
    return getCompiled(name, makeGetter, getterCache);
***REMOVED***;
***REMOVED***

function ensureMethod(obj, methodName) ***REMOVED***
    var fn;
    if (obj != null) fn = obj[methodName];
    if (typeof fn !== "function") ***REMOVED***
        var message = "Object " + util.classString(obj) + " has no method '" +
            util.toString(methodName) + "'";
        throw new Promise.TypeError(message);
    ***REMOVED***
    return fn;
***REMOVED***

function caller(obj) ***REMOVED***
    var methodName = this.pop();
    var fn = ensureMethod(obj, methodName);
    return fn.apply(obj, this);
***REMOVED***
Promise.prototype.call = function (methodName) ***REMOVED***
    var $_len = arguments.length;var args = new Array($_len - 1); for(var $_i = 1; $_i < $_len; ++$_i) ***REMOVED***args[$_i - 1] = arguments[$_i];***REMOVED***
    if (!true) ***REMOVED***
        if (canEvaluate) ***REMOVED***
            var maybeCaller = getMethodCaller(methodName);
            if (maybeCaller !== null) ***REMOVED***
                return this._then(
                    maybeCaller, undefined, undefined, args, undefined);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    args.push(methodName);
    return this._then(caller, undefined, undefined, args, undefined);
***REMOVED***;

function namedGetter(obj) ***REMOVED***
    return obj[this];
***REMOVED***
function indexedGetter(obj) ***REMOVED***
    var index = +this;
    if (index < 0) index = Math.max(0, index + obj.length);
    return obj[index];
***REMOVED***
Promise.prototype.get = function (propertyName) ***REMOVED***
    var isIndex = (typeof propertyName === "number");
    var getter;
    if (!isIndex) ***REMOVED***
        if (canEvaluate) ***REMOVED***
            var maybeGetter = getGetter(propertyName);
            getter = maybeGetter !== null ? maybeGetter : namedGetter;
        ***REMOVED*** else ***REMOVED***
            getter = namedGetter;
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        getter = indexedGetter;
    ***REMOVED***
    return this._then(getter, undefined, undefined, propertyName, undefined);
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],6:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise) ***REMOVED***
var errors = _dereq_("./errors.js");
var async = _dereq_("./async.js");
var CancellationError = errors.CancellationError;

Promise.prototype._cancel = function (reason) ***REMOVED***
    if (!this.isCancellable()) return this;
    var parent;
    var promiseToReject = this;
    while ((parent = promiseToReject._cancellationParent) !== undefined &&
        parent.isCancellable()) ***REMOVED***
        promiseToReject = parent;
    ***REMOVED***
    this._unsetCancellable();
    promiseToReject._target()._rejectCallback(reason, false, true);
***REMOVED***;

Promise.prototype.cancel = function (reason) ***REMOVED***
    if (!this.isCancellable()) return this;
    if (reason === undefined) reason = new CancellationError();
    async.invokeLater(this._cancel, this, reason);
    return this;
***REMOVED***;

Promise.prototype.cancellable = function () ***REMOVED***
    if (this._cancellable()) return this;
    async.enableTrampoline();
    this._setCancellable();
    this._cancellationParent = undefined;
    return this;
***REMOVED***;

Promise.prototype.uncancellable = function () ***REMOVED***
    var ret = this.then();
    ret._unsetCancellable();
    return ret;
***REMOVED***;

Promise.prototype.fork = function (didFulfill, didReject, didProgress) ***REMOVED***
    var ret = this._then(didFulfill, didReject, didProgress,
                         undefined, undefined);

    ret._setCancellable();
    ret._cancellationParent = undefined;
    return ret;
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./async.js":2,"./errors.js":13***REMOVED***],7:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function() ***REMOVED***
var async = _dereq_("./async.js");
var util = _dereq_("./util.js");
var bluebirdFramePattern =
    /[\\\/]bluebird[\\\/]js[\\\/](main|debug|zalgo|instrumented)/;
var stackFramePattern = null;
var formatStack = null;
var indentStackFrames = false;
var warn;

function CapturedTrace(parent) ***REMOVED***
    this._parent = parent;
    var length = this._length = 1 + (parent === undefined ? 0 : parent._length);
    captureStackTrace(this, CapturedTrace);
    if (length > 32) this.uncycle();
***REMOVED***
util.inherits(CapturedTrace, Error);

CapturedTrace.prototype.uncycle = function() ***REMOVED***
    var length = this._length;
    if (length < 2) return;
    var nodes = [];
    var stackToIndex = ***REMOVED******REMOVED***;

    for (var i = 0, node = this; node !== undefined; ++i) ***REMOVED***
        nodes.push(node);
        node = node._parent;
    ***REMOVED***
    length = this._length = i;
    for (var i = length - 1; i >= 0; --i) ***REMOVED***
        var stack = nodes[i].stack;
        if (stackToIndex[stack] === undefined) ***REMOVED***
            stackToIndex[stack] = i;
        ***REMOVED***
    ***REMOVED***
    for (var i = 0; i < length; ++i) ***REMOVED***
        var currentStack = nodes[i].stack;
        var index = stackToIndex[currentStack];
        if (index !== undefined && index !== i) ***REMOVED***
            if (index > 0) ***REMOVED***
                nodes[index - 1]._parent = undefined;
                nodes[index - 1]._length = 1;
            ***REMOVED***
            nodes[i]._parent = undefined;
            nodes[i]._length = 1;
            var cycleEdgeNode = i > 0 ? nodes[i - 1] : this;

            if (index < length - 1) ***REMOVED***
                cycleEdgeNode._parent = nodes[index + 1];
                cycleEdgeNode._parent.uncycle();
                cycleEdgeNode._length =
                    cycleEdgeNode._parent._length + 1;
            ***REMOVED*** else ***REMOVED***
                cycleEdgeNode._parent = undefined;
                cycleEdgeNode._length = 1;
            ***REMOVED***
            var currentChildLength = cycleEdgeNode._length + 1;
            for (var j = i - 2; j >= 0; --j) ***REMOVED***
                nodes[j]._length = currentChildLength;
                currentChildLength++;
            ***REMOVED***
            return;
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

CapturedTrace.prototype.parent = function() ***REMOVED***
    return this._parent;
***REMOVED***;

CapturedTrace.prototype.hasParent = function() ***REMOVED***
    return this._parent !== undefined;
***REMOVED***;

CapturedTrace.prototype.attachExtraTrace = function(error) ***REMOVED***
    if (error.__stackCleaned__) return;
    this.uncycle();
    var parsed = CapturedTrace.parseStackAndMessage(error);
    var message = parsed.message;
    var stacks = [parsed.stack];

    var trace = this;
    while (trace !== undefined) ***REMOVED***
        stacks.push(cleanStack(trace.stack.split("\n")));
        trace = trace._parent;
    ***REMOVED***
    removeCommonRoots(stacks);
    removeDuplicateOrEmptyJumps(stacks);
    util.notEnumerableProp(error, "stack", reconstructStack(message, stacks));
    util.notEnumerableProp(error, "__stackCleaned__", true);
***REMOVED***;

function reconstructStack(message, stacks) ***REMOVED***
    for (var i = 0; i < stacks.length - 1; ++i) ***REMOVED***
        stacks[i].push("From previous event:");
        stacks[i] = stacks[i].join("\n");
    ***REMOVED***
    if (i < stacks.length) ***REMOVED***
        stacks[i] = stacks[i].join("\n");
    ***REMOVED***
    return message + "\n" + stacks.join("\n");
***REMOVED***

function removeDuplicateOrEmptyJumps(stacks) ***REMOVED***
    for (var i = 0; i < stacks.length; ++i) ***REMOVED***
        if (stacks[i].length === 0 ||
            ((i + 1 < stacks.length) && stacks[i][0] === stacks[i+1][0])) ***REMOVED***
            stacks.splice(i, 1);
            i--;
        ***REMOVED***
    ***REMOVED***
***REMOVED***

function removeCommonRoots(stacks) ***REMOVED***
    var current = stacks[0];
    for (var i = 1; i < stacks.length; ++i) ***REMOVED***
        var prev = stacks[i];
        var currentLastIndex = current.length - 1;
        var currentLastLine = current[currentLastIndex];
        var commonRootMeetPoint = -1;

        for (var j = prev.length - 1; j >= 0; --j) ***REMOVED***
            if (prev[j] === currentLastLine) ***REMOVED***
                commonRootMeetPoint = j;
                break;
            ***REMOVED***
        ***REMOVED***

        for (var j = commonRootMeetPoint; j >= 0; --j) ***REMOVED***
            var line = prev[j];
            if (current[currentLastIndex] === line) ***REMOVED***
                current.pop();
                currentLastIndex--;
            ***REMOVED*** else ***REMOVED***
                break;
            ***REMOVED***
        ***REMOVED***
        current = prev;
    ***REMOVED***
***REMOVED***

function cleanStack(stack) ***REMOVED***
    var ret = [];
    for (var i = 0; i < stack.length; ++i) ***REMOVED***
        var line = stack[i];
        var isTraceLine = stackFramePattern.test(line) ||
            "    (No stack trace)" === line;
        var isInternalFrame = isTraceLine && shouldIgnore(line);
        if (isTraceLine && !isInternalFrame) ***REMOVED***
            if (indentStackFrames && line.charAt(0) !== " ") ***REMOVED***
                line = "    " + line;
            ***REMOVED***
            ret.push(line);
        ***REMOVED***
    ***REMOVED***
    return ret;
***REMOVED***

function stackFramesAsArray(error) ***REMOVED***
    var stack = error.stack.replace(/\s+$/g, "").split("\n");
    for (var i = 0; i < stack.length; ++i) ***REMOVED***
        var line = stack[i];
        if ("    (No stack trace)" === line || stackFramePattern.test(line)) ***REMOVED***
            break;
        ***REMOVED***
    ***REMOVED***
    if (i > 0) ***REMOVED***
        stack = stack.slice(i);
    ***REMOVED***
    return stack;
***REMOVED***

CapturedTrace.parseStackAndMessage = function(error) ***REMOVED***
    var stack = error.stack;
    var message = error.toString();
    stack = typeof stack === "string" && stack.length > 0
                ? stackFramesAsArray(error) : ["    (No stack trace)"];
    return ***REMOVED***
        message: message,
        stack: cleanStack(stack)
    ***REMOVED***;
***REMOVED***;

CapturedTrace.formatAndLogError = function(error, title) ***REMOVED***
    if (typeof console !== "undefined") ***REMOVED***
        var message;
        if (typeof error === "object" || typeof error === "function") ***REMOVED***
            var stack = error.stack;
            message = title + formatStack(stack, error);
        ***REMOVED*** else ***REMOVED***
            message = title + String(error);
        ***REMOVED***
        if (typeof warn === "function") ***REMOVED***
            warn(message);
        ***REMOVED*** else if (typeof console.log === "function" ||
            typeof console.log === "object") ***REMOVED***
            console.log(message);
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

CapturedTrace.unhandledRejection = function (reason) ***REMOVED***
    CapturedTrace.formatAndLogError(reason, "^--- With additional stack trace: ");
***REMOVED***;

CapturedTrace.isSupported = function () ***REMOVED***
    return typeof captureStackTrace === "function";
***REMOVED***;

CapturedTrace.fireRejectionEvent =
function(name, localHandler, reason, promise) ***REMOVED***
    var localEventFired = false;
    try ***REMOVED***
        if (typeof localHandler === "function") ***REMOVED***
            localEventFired = true;
            if (name === "rejectionHandled") ***REMOVED***
                localHandler(promise);
            ***REMOVED*** else ***REMOVED***
                localHandler(reason, promise);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED*** catch (e) ***REMOVED***
        async.throwLater(e);
    ***REMOVED***

    var globalEventFired = false;
    try ***REMOVED***
        globalEventFired = fireGlobalEvent(name, reason, promise);
    ***REMOVED*** catch (e) ***REMOVED***
        globalEventFired = true;
        async.throwLater(e);
    ***REMOVED***

    var domEventFired = false;
    if (fireDomEvent) ***REMOVED***
        try ***REMOVED***
            domEventFired = fireDomEvent(name.toLowerCase(), ***REMOVED***
                reason: reason,
                promise: promise
            ***REMOVED***);
        ***REMOVED*** catch (e) ***REMOVED***
            domEventFired = true;
            async.throwLater(e);
        ***REMOVED***
    ***REMOVED***

    if (!globalEventFired && !localEventFired && !domEventFired &&
        name === "unhandledRejection") ***REMOVED***
        CapturedTrace.formatAndLogError(reason, "Unhandled rejection ");
    ***REMOVED***
***REMOVED***;

function formatNonError(obj) ***REMOVED***
    var str;
    if (typeof obj === "function") ***REMOVED***
        str = "[function " +
            (obj.name || "anonymous") +
            "]";
    ***REMOVED*** else ***REMOVED***
        str = obj.toString();
        var ruselessToString = /\[object [a-zA-Z0-9$_]+\]/;
        if (ruselessToString.test(str)) ***REMOVED***
            try ***REMOVED***
                var newStr = JSON.stringify(obj);
                str = newStr;
            ***REMOVED***
            catch(e) ***REMOVED***

            ***REMOVED***
        ***REMOVED***
        if (str.length === 0) ***REMOVED***
            str = "(empty array)";
        ***REMOVED***
    ***REMOVED***
    return ("(<" + snip(str) + ">, no stack trace)");
***REMOVED***

function snip(str) ***REMOVED***
    var maxChars = 41;
    if (str.length < maxChars) ***REMOVED***
        return str;
    ***REMOVED***
    return str.substr(0, maxChars - 3) + "...";
***REMOVED***

var shouldIgnore = function() ***REMOVED*** return false; ***REMOVED***;
var parseLineInfoRegex = /[\/<\(]([^:\/]+):(\d+):(?:\d+)\)?\s*$/;
function parseLineInfo(line) ***REMOVED***
    var matches = line.match(parseLineInfoRegex);
    if (matches) ***REMOVED***
        return ***REMOVED***
            fileName: matches[1],
            line: parseInt(matches[2], 10)
        ***REMOVED***;
    ***REMOVED***
***REMOVED***
CapturedTrace.setBounds = function(firstLineError, lastLineError) ***REMOVED***
    if (!CapturedTrace.isSupported()) return;
    var firstStackLines = firstLineError.stack.split("\n");
    var lastStackLines = lastLineError.stack.split("\n");
    var firstIndex = -1;
    var lastIndex = -1;
    var firstFileName;
    var lastFileName;
    for (var i = 0; i < firstStackLines.length; ++i) ***REMOVED***
        var result = parseLineInfo(firstStackLines[i]);
        if (result) ***REMOVED***
            firstFileName = result.fileName;
            firstIndex = result.line;
            break;
        ***REMOVED***
    ***REMOVED***
    for (var i = 0; i < lastStackLines.length; ++i) ***REMOVED***
        var result = parseLineInfo(lastStackLines[i]);
        if (result) ***REMOVED***
            lastFileName = result.fileName;
            lastIndex = result.line;
            break;
        ***REMOVED***
    ***REMOVED***
    if (firstIndex < 0 || lastIndex < 0 || !firstFileName || !lastFileName ||
        firstFileName !== lastFileName || firstIndex >= lastIndex) ***REMOVED***
        return;
    ***REMOVED***

    shouldIgnore = function(line) ***REMOVED***
        if (bluebirdFramePattern.test(line)) return true;
        var info = parseLineInfo(line);
        if (info) ***REMOVED***
            if (info.fileName === firstFileName &&
                (firstIndex <= info.line && info.line <= lastIndex)) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED***;
***REMOVED***;

var captureStackTrace = (function stackDetection() ***REMOVED***
    var v8stackFramePattern = /^\s*at\s*/;
    var v8stackFormatter = function(stack, error) ***REMOVED***
        if (typeof stack === "string") return stack;

        if (error.name !== undefined &&
            error.message !== undefined) ***REMOVED***
            return error.toString();
        ***REMOVED***
        return formatNonError(error);
    ***REMOVED***;

    if (typeof Error.stackTraceLimit === "number" &&
        typeof Error.captureStackTrace === "function") ***REMOVED***
        Error.stackTraceLimit = Error.stackTraceLimit + 6;
        stackFramePattern = v8stackFramePattern;
        formatStack = v8stackFormatter;
        var captureStackTrace = Error.captureStackTrace;

        shouldIgnore = function(line) ***REMOVED***
            return bluebirdFramePattern.test(line);
        ***REMOVED***;
        return function(receiver, ignoreUntil) ***REMOVED***
            Error.stackTraceLimit = Error.stackTraceLimit + 6;
            captureStackTrace(receiver, ignoreUntil);
            Error.stackTraceLimit = Error.stackTraceLimit - 6;
        ***REMOVED***;
    ***REMOVED***
    var err = new Error();

    if (typeof err.stack === "string" &&
        err.stack.split("\n")[0].indexOf("stackDetection@") >= 0) ***REMOVED***
        stackFramePattern = /@/;
        formatStack = v8stackFormatter;
        indentStackFrames = true;
        return function captureStackTrace(o) ***REMOVED***
            o.stack = new Error().stack;
        ***REMOVED***;
    ***REMOVED***

    var hasStackAfterThrow;
    try ***REMOVED*** throw new Error(); ***REMOVED***
    catch(e) ***REMOVED***
        hasStackAfterThrow = ("stack" in e);
    ***REMOVED***
    if (!("stack" in err) && hasStackAfterThrow &&
        typeof Error.stackTraceLimit === "number") ***REMOVED***
        stackFramePattern = v8stackFramePattern;
        formatStack = v8stackFormatter;
        return function captureStackTrace(o) ***REMOVED***
            Error.stackTraceLimit = Error.stackTraceLimit + 6;
            try ***REMOVED*** throw new Error(); ***REMOVED***
            catch(e) ***REMOVED*** o.stack = e.stack; ***REMOVED***
            Error.stackTraceLimit = Error.stackTraceLimit - 6;
        ***REMOVED***;
    ***REMOVED***

    formatStack = function(stack, error) ***REMOVED***
        if (typeof stack === "string") return stack;

        if ((typeof error === "object" ||
            typeof error === "function") &&
            error.name !== undefined &&
            error.message !== undefined) ***REMOVED***
            return error.toString();
        ***REMOVED***
        return formatNonError(error);
    ***REMOVED***;

    return null;

***REMOVED***)([]);

var fireDomEvent;
var fireGlobalEvent = (function() ***REMOVED***
    if (util.isNode) ***REMOVED***
        return function(name, reason, promise) ***REMOVED***
            if (name === "rejectionHandled") ***REMOVED***
                return process.emit(name, promise);
            ***REMOVED*** else ***REMOVED***
                return process.emit(name, reason, promise);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED*** else ***REMOVED***
        var customEventWorks = false;
        var anyEventWorks = true;
        try ***REMOVED***
            var ev = new self.CustomEvent("test");
            customEventWorks = ev instanceof CustomEvent;
        ***REMOVED*** catch (e) ***REMOVED******REMOVED***
        if (!customEventWorks) ***REMOVED***
            try ***REMOVED***
                var event = document.createEvent("CustomEvent");
                event.initCustomEvent("testingtheevent", false, true, ***REMOVED******REMOVED***);
                self.dispatchEvent(event);
            ***REMOVED*** catch (e) ***REMOVED***
                anyEventWorks = false;
            ***REMOVED***
        ***REMOVED***
        if (anyEventWorks) ***REMOVED***
            fireDomEvent = function(type, detail) ***REMOVED***
                var event;
                if (customEventWorks) ***REMOVED***
                    event = new self.CustomEvent(type, ***REMOVED***
                        detail: detail,
                        bubbles: false,
                        cancelable: true
                    ***REMOVED***);
                ***REMOVED*** else if (self.dispatchEvent) ***REMOVED***
                    event = document.createEvent("CustomEvent");
                    event.initCustomEvent(type, false, true, detail);
                ***REMOVED***

                return event ? !self.dispatchEvent(event) : false;
            ***REMOVED***;
        ***REMOVED***

        var toWindowMethodNameMap = ***REMOVED******REMOVED***;
        toWindowMethodNameMap["unhandledRejection"] = ("on" +
            "unhandledRejection").toLowerCase();
        toWindowMethodNameMap["rejectionHandled"] = ("on" +
            "rejectionHandled").toLowerCase();

        return function(name, reason, promise) ***REMOVED***
            var methodName = toWindowMethodNameMap[name];
            var method = self[methodName];
            if (!method) return false;
            if (name === "rejectionHandled") ***REMOVED***
                method.call(self, promise);
            ***REMOVED*** else ***REMOVED***
                method.call(self, reason, promise);
            ***REMOVED***
            return true;
        ***REMOVED***;
    ***REMOVED***
***REMOVED***)();

if (typeof console !== "undefined" && typeof console.warn !== "undefined") ***REMOVED***
    warn = function (message) ***REMOVED***
        console.warn(message);
    ***REMOVED***;
    if (util.isNode && process.stderr.isTTY) ***REMOVED***
        warn = function(message) ***REMOVED***
            process.stderr.write("\u001b[31m" + message + "\u001b[39m\n");
        ***REMOVED***;
    ***REMOVED*** else if (!util.isNode && typeof (new Error().stack) === "string") ***REMOVED***
        warn = function(message) ***REMOVED***
            console.warn("%c" + message, "color: red");
        ***REMOVED***;
    ***REMOVED***
***REMOVED***

return CapturedTrace;
***REMOVED***;

***REMOVED***,***REMOVED***"./async.js":2,"./util.js":38***REMOVED***],8:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(NEXT_FILTER) ***REMOVED***
var util = _dereq_("./util.js");
var errors = _dereq_("./errors.js");
var tryCatch = util.tryCatch;
var errorObj = util.errorObj;
var keys = _dereq_("./es5.js").keys;
var TypeError = errors.TypeError;

function CatchFilter(instances, callback, promise) ***REMOVED***
    this._instances = instances;
    this._callback = callback;
    this._promise = promise;
***REMOVED***

function safePredicate(predicate, e) ***REMOVED***
    var safeObject = ***REMOVED******REMOVED***;
    var retfilter = tryCatch(predicate).call(safeObject, e);

    if (retfilter === errorObj) return retfilter;

    var safeKeys = keys(safeObject);
    if (safeKeys.length) ***REMOVED***
        errorObj.e = new TypeError("Catch filter must inherit from Error or be a simple predicate function\u000a\u000a    See http://goo.gl/o84o68\u000a");
        return errorObj;
    ***REMOVED***
    return retfilter;
***REMOVED***

CatchFilter.prototype.doFilter = function (e) ***REMOVED***
    var cb = this._callback;
    var promise = this._promise;
    var boundTo = promise._boundValue();
    for (var i = 0, len = this._instances.length; i < len; ++i) ***REMOVED***
        var item = this._instances[i];
        var itemIsErrorType = item === Error ||
            (item != null && item.prototype instanceof Error);

        if (itemIsErrorType && e instanceof item) ***REMOVED***
            var ret = tryCatch(cb).call(boundTo, e);
            if (ret === errorObj) ***REMOVED***
                NEXT_FILTER.e = ret.e;
                return NEXT_FILTER;
            ***REMOVED***
            return ret;
        ***REMOVED*** else if (typeof item === "function" && !itemIsErrorType) ***REMOVED***
            var shouldHandle = safePredicate(item, e);
            if (shouldHandle === errorObj) ***REMOVED***
                e = errorObj.e;
                break;
            ***REMOVED*** else if (shouldHandle) ***REMOVED***
                var ret = tryCatch(cb).call(boundTo, e);
                if (ret === errorObj) ***REMOVED***
                    NEXT_FILTER.e = ret.e;
                    return NEXT_FILTER;
                ***REMOVED***
                return ret;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    NEXT_FILTER.e = e;
    return NEXT_FILTER;
***REMOVED***;

return CatchFilter;
***REMOVED***;

***REMOVED***,***REMOVED***"./errors.js":13,"./es5.js":14,"./util.js":38***REMOVED***],9:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, CapturedTrace, isDebugging) ***REMOVED***
var contextStack = [];
function Context() ***REMOVED***
    this._trace = new CapturedTrace(peekContext());
***REMOVED***
Context.prototype._pushContext = function () ***REMOVED***
    if (!isDebugging()) return;
    if (this._trace !== undefined) ***REMOVED***
        contextStack.push(this._trace);
    ***REMOVED***
***REMOVED***;

Context.prototype._popContext = function () ***REMOVED***
    if (!isDebugging()) return;
    if (this._trace !== undefined) ***REMOVED***
        contextStack.pop();
    ***REMOVED***
***REMOVED***;

function createContext() ***REMOVED***
    if (isDebugging()) return new Context();
***REMOVED***

function peekContext() ***REMOVED***
    var lastIndex = contextStack.length - 1;
    if (lastIndex >= 0) ***REMOVED***
        return contextStack[lastIndex];
    ***REMOVED***
    return undefined;
***REMOVED***

Promise.prototype._peekContext = peekContext;
Promise.prototype._pushContext = Context.prototype._pushContext;
Promise.prototype._popContext = Context.prototype._popContext;

return createContext;
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],10:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, CapturedTrace) ***REMOVED***
var getDomain = Promise._getDomain;
var async = _dereq_("./async.js");
var Warning = _dereq_("./errors.js").Warning;
var util = _dereq_("./util.js");
var canAttachTrace = util.canAttachTrace;
var unhandledRejectionHandled;
var possiblyUnhandledRejection;
var debugging = false || (util.isNode &&
                    (!!process.env["BLUEBIRD_DEBUG"] ||
                     process.env["NODE_ENV"] === "development"));

if (util.isNode && process.env["BLUEBIRD_DEBUG"] == 0) debugging = false;

if (debugging) ***REMOVED***
    async.disableTrampolineIfNecessary();
***REMOVED***

Promise.prototype._ignoreRejections = function() ***REMOVED***
    this._unsetRejectionIsUnhandled();
    this._bitField = this._bitField | 16777216;
***REMOVED***;

Promise.prototype._ensurePossibleRejectionHandled = function () ***REMOVED***
    if ((this._bitField & 16777216) !== 0) return;
    this._setRejectionIsUnhandled();
    async.invokeLater(this._notifyUnhandledRejection, this, undefined);
***REMOVED***;

Promise.prototype._notifyUnhandledRejectionIsHandled = function () ***REMOVED***
    CapturedTrace.fireRejectionEvent("rejectionHandled",
                                  unhandledRejectionHandled, undefined, this);
***REMOVED***;

Promise.prototype._notifyUnhandledRejection = function () ***REMOVED***
    if (this._isRejectionUnhandled()) ***REMOVED***
        var reason = this._getCarriedStackTrace() || this._settledValue;
        this._setUnhandledRejectionIsNotified();
        CapturedTrace.fireRejectionEvent("unhandledRejection",
                                      possiblyUnhandledRejection, reason, this);
    ***REMOVED***
***REMOVED***;

Promise.prototype._setUnhandledRejectionIsNotified = function () ***REMOVED***
    this._bitField = this._bitField | 524288;
***REMOVED***;

Promise.prototype._unsetUnhandledRejectionIsNotified = function () ***REMOVED***
    this._bitField = this._bitField & (~524288);
***REMOVED***;

Promise.prototype._isUnhandledRejectionNotified = function () ***REMOVED***
    return (this._bitField & 524288) > 0;
***REMOVED***;

Promise.prototype._setRejectionIsUnhandled = function () ***REMOVED***
    this._bitField = this._bitField | 2097152;
***REMOVED***;

Promise.prototype._unsetRejectionIsUnhandled = function () ***REMOVED***
    this._bitField = this._bitField & (~2097152);
    if (this._isUnhandledRejectionNotified()) ***REMOVED***
        this._unsetUnhandledRejectionIsNotified();
        this._notifyUnhandledRejectionIsHandled();
    ***REMOVED***
***REMOVED***;

Promise.prototype._isRejectionUnhandled = function () ***REMOVED***
    return (this._bitField & 2097152) > 0;
***REMOVED***;

Promise.prototype._setCarriedStackTrace = function (capturedTrace) ***REMOVED***
    this._bitField = this._bitField | 1048576;
    this._fulfillmentHandler0 = capturedTrace;
***REMOVED***;

Promise.prototype._isCarryingStackTrace = function () ***REMOVED***
    return (this._bitField & 1048576) > 0;
***REMOVED***;

Promise.prototype._getCarriedStackTrace = function () ***REMOVED***
    return this._isCarryingStackTrace()
        ? this._fulfillmentHandler0
        : undefined;
***REMOVED***;

Promise.prototype._captureStackTrace = function () ***REMOVED***
    if (debugging) ***REMOVED***
        this._trace = new CapturedTrace(this._peekContext());
    ***REMOVED***
    return this;
***REMOVED***;

Promise.prototype._attachExtraTrace = function (error, ignoreSelf) ***REMOVED***
    if (debugging && canAttachTrace(error)) ***REMOVED***
        var trace = this._trace;
        if (trace !== undefined) ***REMOVED***
            if (ignoreSelf) trace = trace._parent;
        ***REMOVED***
        if (trace !== undefined) ***REMOVED***
            trace.attachExtraTrace(error);
        ***REMOVED*** else if (!error.__stackCleaned__) ***REMOVED***
            var parsed = CapturedTrace.parseStackAndMessage(error);
            util.notEnumerableProp(error, "stack",
                parsed.message + "\n" + parsed.stack.join("\n"));
            util.notEnumerableProp(error, "__stackCleaned__", true);
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

Promise.prototype._warn = function(message) ***REMOVED***
    var warning = new Warning(message);
    var ctx = this._peekContext();
    if (ctx) ***REMOVED***
        ctx.attachExtraTrace(warning);
    ***REMOVED*** else ***REMOVED***
        var parsed = CapturedTrace.parseStackAndMessage(warning);
        warning.stack = parsed.message + "\n" + parsed.stack.join("\n");
    ***REMOVED***
    CapturedTrace.formatAndLogError(warning, "");
***REMOVED***;

Promise.onPossiblyUnhandledRejection = function (fn) ***REMOVED***
    var domain = getDomain();
    possiblyUnhandledRejection =
        typeof fn === "function" ? (domain === null ? fn : domain.bind(fn))
                                 : undefined;
***REMOVED***;

Promise.onUnhandledRejectionHandled = function (fn) ***REMOVED***
    var domain = getDomain();
    unhandledRejectionHandled =
        typeof fn === "function" ? (domain === null ? fn : domain.bind(fn))
                                 : undefined;
***REMOVED***;

Promise.longStackTraces = function () ***REMOVED***
    if (async.haveItemsQueued() &&
        debugging === false
   ) ***REMOVED***
        throw new Error("cannot enable long stack traces after promises have been created\u000a\u000a    See http://goo.gl/DT1qyG\u000a");
    ***REMOVED***
    debugging = CapturedTrace.isSupported();
    if (debugging) ***REMOVED***
        async.disableTrampolineIfNecessary();
    ***REMOVED***
***REMOVED***;

Promise.hasLongStackTraces = function () ***REMOVED***
    return debugging && CapturedTrace.isSupported();
***REMOVED***;

if (!CapturedTrace.isSupported()) ***REMOVED***
    Promise.longStackTraces = function()***REMOVED******REMOVED***;
    debugging = false;
***REMOVED***

return function() ***REMOVED***
    return debugging;
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./async.js":2,"./errors.js":13,"./util.js":38***REMOVED***],11:[function(_dereq_,module,exports)***REMOVED***
"use strict";
var util = _dereq_("./util.js");
var isPrimitive = util.isPrimitive;

module.exports = function(Promise) ***REMOVED***
var returner = function () ***REMOVED***
    return this;
***REMOVED***;
var thrower = function () ***REMOVED***
    throw this;
***REMOVED***;
var returnUndefined = function() ***REMOVED******REMOVED***;
var throwUndefined = function() ***REMOVED***
    throw undefined;
***REMOVED***;

var wrapper = function (value, action) ***REMOVED***
    if (action === 1) ***REMOVED***
        return function () ***REMOVED***
            throw value;
        ***REMOVED***;
    ***REMOVED*** else if (action === 2) ***REMOVED***
        return function () ***REMOVED***
            return value;
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;


Promise.prototype["return"] =
Promise.prototype.thenReturn = function (value) ***REMOVED***
    if (value === undefined) return this.then(returnUndefined);

    if (isPrimitive(value)) ***REMOVED***
        return this._then(
            wrapper(value, 2),
            undefined,
            undefined,
            undefined,
            undefined
       );
    ***REMOVED*** else if (value instanceof Promise) ***REMOVED***
        value._ignoreRejections();
    ***REMOVED***
    return this._then(returner, undefined, undefined, value, undefined);
***REMOVED***;

Promise.prototype["throw"] =
Promise.prototype.thenThrow = function (reason) ***REMOVED***
    if (reason === undefined) return this.then(throwUndefined);

    if (isPrimitive(reason)) ***REMOVED***
        return this._then(
            wrapper(reason, 1),
            undefined,
            undefined,
            undefined,
            undefined
       );
    ***REMOVED***
    return this._then(thrower, undefined, undefined, reason, undefined);
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],12:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, INTERNAL) ***REMOVED***
var PromiseReduce = Promise.reduce;

Promise.prototype.each = function (fn) ***REMOVED***
    return PromiseReduce(this, fn, null, INTERNAL);
***REMOVED***;

Promise.each = function (promises, fn) ***REMOVED***
    return PromiseReduce(promises, fn, null, INTERNAL);
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],13:[function(_dereq_,module,exports)***REMOVED***
"use strict";
var es5 = _dereq_("./es5.js");
var Objectfreeze = es5.freeze;
var util = _dereq_("./util.js");
var inherits = util.inherits;
var notEnumerableProp = util.notEnumerableProp;

function subError(nameProperty, defaultMessage) ***REMOVED***
    function SubError(message) ***REMOVED***
        if (!(this instanceof SubError)) return new SubError(message);
        notEnumerableProp(this, "message",
            typeof message === "string" ? message : defaultMessage);
        notEnumerableProp(this, "name", nameProperty);
        if (Error.captureStackTrace) ***REMOVED***
            Error.captureStackTrace(this, this.constructor);
        ***REMOVED*** else ***REMOVED***
            Error.call(this);
        ***REMOVED***
    ***REMOVED***
    inherits(SubError, Error);
    return SubError;
***REMOVED***

var _TypeError, _RangeError;
var Warning = subError("Warning", "warning");
var CancellationError = subError("CancellationError", "cancellation error");
var TimeoutError = subError("TimeoutError", "timeout error");
var AggregateError = subError("AggregateError", "aggregate error");
try ***REMOVED***
    _TypeError = TypeError;
    _RangeError = RangeError;
***REMOVED*** catch(e) ***REMOVED***
    _TypeError = subError("TypeError", "type error");
    _RangeError = subError("RangeError", "range error");
***REMOVED***

var methods = ("join pop push shift unshift slice filter forEach some " +
    "every map indexOf lastIndexOf reduce reduceRight sort reverse").split(" ");

for (var i = 0; i < methods.length; ++i) ***REMOVED***
    if (typeof Array.prototype[methods[i]] === "function") ***REMOVED***
        AggregateError.prototype[methods[i]] = Array.prototype[methods[i]];
    ***REMOVED***
***REMOVED***

es5.defineProperty(AggregateError.prototype, "length", ***REMOVED***
    value: 0,
    configurable: false,
    writable: true,
    enumerable: true
***REMOVED***);
AggregateError.prototype["isOperational"] = true;
var level = 0;
AggregateError.prototype.toString = function() ***REMOVED***
    var indent = Array(level * 4 + 1).join(" ");
    var ret = "\n" + indent + "AggregateError of:" + "\n";
    level++;
    indent = Array(level * 4 + 1).join(" ");
    for (var i = 0; i < this.length; ++i) ***REMOVED***
        var str = this[i] === this ? "[Circular AggregateError]" : this[i] + "";
        var lines = str.split("\n");
        for (var j = 0; j < lines.length; ++j) ***REMOVED***
            lines[j] = indent + lines[j];
        ***REMOVED***
        str = lines.join("\n");
        ret += str + "\n";
    ***REMOVED***
    level--;
    return ret;
***REMOVED***;

function OperationalError(message) ***REMOVED***
    if (!(this instanceof OperationalError))
        return new OperationalError(message);
    notEnumerableProp(this, "name", "OperationalError");
    notEnumerableProp(this, "message", message);
    this.cause = message;
    this["isOperational"] = true;

    if (message instanceof Error) ***REMOVED***
        notEnumerableProp(this, "message", message.message);
        notEnumerableProp(this, "stack", message.stack);
    ***REMOVED*** else if (Error.captureStackTrace) ***REMOVED***
        Error.captureStackTrace(this, this.constructor);
    ***REMOVED***

***REMOVED***
inherits(OperationalError, Error);

var errorTypes = Error["__BluebirdErrorTypes__"];
if (!errorTypes) ***REMOVED***
    errorTypes = Objectfreeze(***REMOVED***
        CancellationError: CancellationError,
        TimeoutError: TimeoutError,
        OperationalError: OperationalError,
        RejectionError: OperationalError,
        AggregateError: AggregateError
    ***REMOVED***);
    notEnumerableProp(Error, "__BluebirdErrorTypes__", errorTypes);
***REMOVED***

module.exports = ***REMOVED***
    Error: Error,
    TypeError: _TypeError,
    RangeError: _RangeError,
    CancellationError: errorTypes.CancellationError,
    OperationalError: errorTypes.OperationalError,
    TimeoutError: errorTypes.TimeoutError,
    AggregateError: errorTypes.AggregateError,
    Warning: Warning
***REMOVED***;

***REMOVED***,***REMOVED***"./es5.js":14,"./util.js":38***REMOVED***],14:[function(_dereq_,module,exports)***REMOVED***
var isES5 = (function()***REMOVED***
    "use strict";
    return this === undefined;
***REMOVED***)();

if (isES5) ***REMOVED***
    module.exports = ***REMOVED***
        freeze: Object.freeze,
        defineProperty: Object.defineProperty,
        getDescriptor: Object.getOwnPropertyDescriptor,
        keys: Object.keys,
        names: Object.getOwnPropertyNames,
        getPrototypeOf: Object.getPrototypeOf,
        isArray: Array.isArray,
        isES5: isES5,
        propertyIsWritable: function(obj, prop) ***REMOVED***
            var descriptor = Object.getOwnPropertyDescriptor(obj, prop);
            return !!(!descriptor || descriptor.writable || descriptor.set);
        ***REMOVED***
    ***REMOVED***;
***REMOVED*** else ***REMOVED***
    var has = ***REMOVED******REMOVED***.hasOwnProperty;
    var str = ***REMOVED******REMOVED***.toString;
    var proto = ***REMOVED******REMOVED***.constructor.prototype;

    var ObjectKeys = function (o) ***REMOVED***
        var ret = [];
        for (var key in o) ***REMOVED***
            if (has.call(o, key)) ***REMOVED***
                ret.push(key);
            ***REMOVED***
        ***REMOVED***
        return ret;
    ***REMOVED***;

    var ObjectGetDescriptor = function(o, key) ***REMOVED***
        return ***REMOVED***value: o[key]***REMOVED***;
    ***REMOVED***;

    var ObjectDefineProperty = function (o, key, desc) ***REMOVED***
        o[key] = desc.value;
        return o;
    ***REMOVED***;

    var ObjectFreeze = function (obj) ***REMOVED***
        return obj;
    ***REMOVED***;

    var ObjectGetPrototypeOf = function (obj) ***REMOVED***
        try ***REMOVED***
            return Object(obj).constructor.prototype;
        ***REMOVED***
        catch (e) ***REMOVED***
            return proto;
        ***REMOVED***
    ***REMOVED***;

    var ArrayIsArray = function (obj) ***REMOVED***
        try ***REMOVED***
            return str.call(obj) === "[object Array]";
        ***REMOVED***
        catch(e) ***REMOVED***
            return false;
        ***REMOVED***
    ***REMOVED***;

    module.exports = ***REMOVED***
        isArray: ArrayIsArray,
        keys: ObjectKeys,
        names: ObjectKeys,
        defineProperty: ObjectDefineProperty,
        getDescriptor: ObjectGetDescriptor,
        freeze: ObjectFreeze,
        getPrototypeOf: ObjectGetPrototypeOf,
        isES5: isES5,
        propertyIsWritable: function() ***REMOVED***
            return true;
        ***REMOVED***
    ***REMOVED***;
***REMOVED***

***REMOVED***,***REMOVED******REMOVED***],15:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, INTERNAL) ***REMOVED***
var PromiseMap = Promise.map;

Promise.prototype.filter = function (fn, options) ***REMOVED***
    return PromiseMap(this, fn, options, INTERNAL);
***REMOVED***;

Promise.filter = function (promises, fn, options) ***REMOVED***
    return PromiseMap(promises, fn, options, INTERNAL);
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],16:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, NEXT_FILTER, tryConvertToPromise) ***REMOVED***
var util = _dereq_("./util.js");
var isPrimitive = util.isPrimitive;
var thrower = util.thrower;

function returnThis() ***REMOVED***
    return this;
***REMOVED***
function throwThis() ***REMOVED***
    throw this;
***REMOVED***
function return$(r) ***REMOVED***
    return function() ***REMOVED***
        return r;
    ***REMOVED***;
***REMOVED***
function throw$(r) ***REMOVED***
    return function() ***REMOVED***
        throw r;
    ***REMOVED***;
***REMOVED***
function promisedFinally(ret, reasonOrValue, isFulfilled) ***REMOVED***
    var then;
    if (isPrimitive(reasonOrValue)) ***REMOVED***
        then = isFulfilled ? return$(reasonOrValue) : throw$(reasonOrValue);
    ***REMOVED*** else ***REMOVED***
        then = isFulfilled ? returnThis : throwThis;
    ***REMOVED***
    return ret._then(then, thrower, undefined, reasonOrValue, undefined);
***REMOVED***

function finallyHandler(reasonOrValue) ***REMOVED***
    var promise = this.promise;
    var handler = this.handler;

    var ret = promise._isBound()
                    ? handler.call(promise._boundValue())
                    : handler();

    if (ret !== undefined) ***REMOVED***
        var maybePromise = tryConvertToPromise(ret, promise);
        if (maybePromise instanceof Promise) ***REMOVED***
            maybePromise = maybePromise._target();
            return promisedFinally(maybePromise, reasonOrValue,
                                    promise.isFulfilled());
        ***REMOVED***
    ***REMOVED***

    if (promise.isRejected()) ***REMOVED***
        NEXT_FILTER.e = reasonOrValue;
        return NEXT_FILTER;
    ***REMOVED*** else ***REMOVED***
        return reasonOrValue;
    ***REMOVED***
***REMOVED***

function tapHandler(value) ***REMOVED***
    var promise = this.promise;
    var handler = this.handler;

    var ret = promise._isBound()
                    ? handler.call(promise._boundValue(), value)
                    : handler(value);

    if (ret !== undefined) ***REMOVED***
        var maybePromise = tryConvertToPromise(ret, promise);
        if (maybePromise instanceof Promise) ***REMOVED***
            maybePromise = maybePromise._target();
            return promisedFinally(maybePromise, value, true);
        ***REMOVED***
    ***REMOVED***
    return value;
***REMOVED***

Promise.prototype._passThroughHandler = function (handler, isFinally) ***REMOVED***
    if (typeof handler !== "function") return this.then();

    var promiseAndHandler = ***REMOVED***
        promise: this,
        handler: handler
    ***REMOVED***;

    return this._then(
            isFinally ? finallyHandler : tapHandler,
            isFinally ? finallyHandler : undefined, undefined,
            promiseAndHandler, undefined);
***REMOVED***;

Promise.prototype.lastly =
Promise.prototype["finally"] = function (handler) ***REMOVED***
    return this._passThroughHandler(handler, true);
***REMOVED***;

Promise.prototype.tap = function (handler) ***REMOVED***
    return this._passThroughHandler(handler, false);
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],17:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise,
                          apiRejection,
                          INTERNAL,
                          tryConvertToPromise) ***REMOVED***
var errors = _dereq_("./errors.js");
var TypeError = errors.TypeError;
var util = _dereq_("./util.js");
var errorObj = util.errorObj;
var tryCatch = util.tryCatch;
var yieldHandlers = [];

function promiseFromYieldHandler(value, yieldHandlers, traceParent) ***REMOVED***
    for (var i = 0; i < yieldHandlers.length; ++i) ***REMOVED***
        traceParent._pushContext();
        var result = tryCatch(yieldHandlers[i])(value);
        traceParent._popContext();
        if (result === errorObj) ***REMOVED***
            traceParent._pushContext();
            var ret = Promise.reject(errorObj.e);
            traceParent._popContext();
            return ret;
        ***REMOVED***
        var maybePromise = tryConvertToPromise(result, traceParent);
        if (maybePromise instanceof Promise) return maybePromise;
    ***REMOVED***
    return null;
***REMOVED***

function PromiseSpawn(generatorFunction, receiver, yieldHandler, stack) ***REMOVED***
    var promise = this._promise = new Promise(INTERNAL);
    promise._captureStackTrace();
    this._stack = stack;
    this._generatorFunction = generatorFunction;
    this._receiver = receiver;
    this._generator = undefined;
    this._yieldHandlers = typeof yieldHandler === "function"
        ? [yieldHandler].concat(yieldHandlers)
        : yieldHandlers;
***REMOVED***

PromiseSpawn.prototype.promise = function () ***REMOVED***
    return this._promise;
***REMOVED***;

PromiseSpawn.prototype._run = function () ***REMOVED***
    this._generator = this._generatorFunction.call(this._receiver);
    this._receiver =
        this._generatorFunction = undefined;
    this._next(undefined);
***REMOVED***;

PromiseSpawn.prototype._continue = function (result) ***REMOVED***
    if (result === errorObj) ***REMOVED***
        return this._promise._rejectCallback(result.e, false, true);
    ***REMOVED***

    var value = result.value;
    if (result.done === true) ***REMOVED***
        this._promise._resolveCallback(value);
    ***REMOVED*** else ***REMOVED***
        var maybePromise = tryConvertToPromise(value, this._promise);
        if (!(maybePromise instanceof Promise)) ***REMOVED***
            maybePromise =
                promiseFromYieldHandler(maybePromise,
                                        this._yieldHandlers,
                                        this._promise);
            if (maybePromise === null) ***REMOVED***
                this._throw(
                    new TypeError(
                        "A value %s was yielded that could not be treated as a promise\u000a\u000a    See http://goo.gl/4Y4pDk\u000a\u000a".replace("%s", value) +
                        "From coroutine:\u000a" +
                        this._stack.split("\n").slice(1, -7).join("\n")
                    )
                );
                return;
            ***REMOVED***
        ***REMOVED***
        maybePromise._then(
            this._next,
            this._throw,
            undefined,
            this,
            null
       );
    ***REMOVED***
***REMOVED***;

PromiseSpawn.prototype._throw = function (reason) ***REMOVED***
    this._promise._attachExtraTrace(reason);
    this._promise._pushContext();
    var result = tryCatch(this._generator["throw"])
        .call(this._generator, reason);
    this._promise._popContext();
    this._continue(result);
***REMOVED***;

PromiseSpawn.prototype._next = function (value) ***REMOVED***
    this._promise._pushContext();
    var result = tryCatch(this._generator.next).call(this._generator, value);
    this._promise._popContext();
    this._continue(result);
***REMOVED***;

Promise.coroutine = function (generatorFunction, options) ***REMOVED***
    if (typeof generatorFunction !== "function") ***REMOVED***
        throw new TypeError("generatorFunction must be a function\u000a\u000a    See http://goo.gl/6Vqhm0\u000a");
    ***REMOVED***
    var yieldHandler = Object(options).yieldHandler;
    var PromiseSpawn$ = PromiseSpawn;
    var stack = new Error().stack;
    return function () ***REMOVED***
        var generator = generatorFunction.apply(this, arguments);
        var spawn = new PromiseSpawn$(undefined, undefined, yieldHandler,
                                      stack);
        spawn._generator = generator;
        spawn._next(undefined);
        return spawn.promise();
    ***REMOVED***;
***REMOVED***;

Promise.coroutine.addYieldHandler = function(fn) ***REMOVED***
    if (typeof fn !== "function") throw new TypeError("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");
    yieldHandlers.push(fn);
***REMOVED***;

Promise.spawn = function (generatorFunction) ***REMOVED***
    if (typeof generatorFunction !== "function") ***REMOVED***
        return apiRejection("generatorFunction must be a function\u000a\u000a    See http://goo.gl/6Vqhm0\u000a");
    ***REMOVED***
    var spawn = new PromiseSpawn(generatorFunction, this);
    var ret = spawn.promise();
    spawn._run(Promise.spawn);
    return ret;
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./errors.js":13,"./util.js":38***REMOVED***],18:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports =
function(Promise, PromiseArray, tryConvertToPromise, INTERNAL) ***REMOVED***
var util = _dereq_("./util.js");
var canEvaluate = util.canEvaluate;
var tryCatch = util.tryCatch;
var errorObj = util.errorObj;
var reject;

if (!true) ***REMOVED***
if (canEvaluate) ***REMOVED***
    var thenCallback = function(i) ***REMOVED***
        return new Function("value", "holder", "                             \n\
            'use strict';                                                    \n\
            holder.pIndex = value;                                           \n\
            holder.checkFulfillment(this);                                   \n\
            ".replace(/Index/g, i));
    ***REMOVED***;

    var caller = function(count) ***REMOVED***
        var values = [];
        for (var i = 1; i <= count; ++i) values.push("holder.p" + i);
        return new Function("holder", "                                      \n\
            'use strict';                                                    \n\
            var callback = holder.fn;                                        \n\
            return callback(values);                                         \n\
            ".replace(/values/g, values.join(", ")));
    ***REMOVED***;
    var thenCallbacks = [];
    var callers = [undefined];
    for (var i = 1; i <= 5; ++i) ***REMOVED***
        thenCallbacks.push(thenCallback(i));
        callers.push(caller(i));
    ***REMOVED***

    var Holder = function(total, fn) ***REMOVED***
        this.p1 = this.p2 = this.p3 = this.p4 = this.p5 = null;
        this.fn = fn;
        this.total = total;
        this.now = 0;
    ***REMOVED***;

    Holder.prototype.callers = callers;
    Holder.prototype.checkFulfillment = function(promise) ***REMOVED***
        var now = this.now;
        now++;
        var total = this.total;
        if (now >= total) ***REMOVED***
            var handler = this.callers[total];
            promise._pushContext();
            var ret = tryCatch(handler)(this);
            promise._popContext();
            if (ret === errorObj) ***REMOVED***
                promise._rejectCallback(ret.e, false, true);
            ***REMOVED*** else ***REMOVED***
                promise._resolveCallback(ret);
            ***REMOVED***
        ***REMOVED*** else ***REMOVED***
            this.now = now;
        ***REMOVED***
    ***REMOVED***;

    var reject = function (reason) ***REMOVED***
        this._reject(reason);
    ***REMOVED***;
***REMOVED***
***REMOVED***

Promise.join = function () ***REMOVED***
    var last = arguments.length - 1;
    var fn;
    if (last > 0 && typeof arguments[last] === "function") ***REMOVED***
        fn = arguments[last];
        if (!true) ***REMOVED***
            if (last < 6 && canEvaluate) ***REMOVED***
                var ret = new Promise(INTERNAL);
                ret._captureStackTrace();
                var holder = new Holder(last, fn);
                var callbacks = thenCallbacks;
                for (var i = 0; i < last; ++i) ***REMOVED***
                    var maybePromise = tryConvertToPromise(arguments[i], ret);
                    if (maybePromise instanceof Promise) ***REMOVED***
                        maybePromise = maybePromise._target();
                        if (maybePromise._isPending()) ***REMOVED***
                            maybePromise._then(callbacks[i], reject,
                                               undefined, ret, holder);
                        ***REMOVED*** else if (maybePromise._isFulfilled()) ***REMOVED***
                            callbacks[i].call(ret,
                                              maybePromise._value(), holder);
                        ***REMOVED*** else ***REMOVED***
                            ret._reject(maybePromise._reason());
                        ***REMOVED***
                    ***REMOVED*** else ***REMOVED***
                        callbacks[i].call(ret, maybePromise, holder);
                    ***REMOVED***
                ***REMOVED***
                return ret;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    var $_len = arguments.length;var args = new Array($_len); for(var $_i = 0; $_i < $_len; ++$_i) ***REMOVED***args[$_i] = arguments[$_i];***REMOVED***
    if (fn) args.pop();
    var ret = new PromiseArray(args).promise();
    return fn !== undefined ? ret.spread(fn) : ret;
***REMOVED***;

***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],19:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise,
                          PromiseArray,
                          apiRejection,
                          tryConvertToPromise,
                          INTERNAL) ***REMOVED***
var getDomain = Promise._getDomain;
var async = _dereq_("./async.js");
var util = _dereq_("./util.js");
var tryCatch = util.tryCatch;
var errorObj = util.errorObj;
var PENDING = ***REMOVED******REMOVED***;
var EMPTY_ARRAY = [];

function MappingPromiseArray(promises, fn, limit, _filter) ***REMOVED***
    this.constructor$(promises);
    this._promise._captureStackTrace();
    var domain = getDomain();
    this._callback = domain === null ? fn : domain.bind(fn);
    this._preservedValues = _filter === INTERNAL
        ? new Array(this.length())
        : null;
    this._limit = limit;
    this._inFlight = 0;
    this._queue = limit >= 1 ? [] : EMPTY_ARRAY;
    async.invoke(init, this, undefined);
***REMOVED***
util.inherits(MappingPromiseArray, PromiseArray);
function init() ***REMOVED***this._init$(undefined, -2);***REMOVED***

MappingPromiseArray.prototype._init = function () ***REMOVED******REMOVED***;

MappingPromiseArray.prototype._promiseFulfilled = function (value, index) ***REMOVED***
    var values = this._values;
    var length = this.length();
    var preservedValues = this._preservedValues;
    var limit = this._limit;
    if (values[index] === PENDING) ***REMOVED***
        values[index] = value;
        if (limit >= 1) ***REMOVED***
            this._inFlight--;
            this._drainQueue();
            if (this._isResolved()) return;
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        if (limit >= 1 && this._inFlight >= limit) ***REMOVED***
            values[index] = value;
            this._queue.push(index);
            return;
        ***REMOVED***
        if (preservedValues !== null) preservedValues[index] = value;

        var callback = this._callback;
        var receiver = this._promise._boundValue();
        this._promise._pushContext();
        var ret = tryCatch(callback).call(receiver, value, index, length);
        this._promise._popContext();
        if (ret === errorObj) return this._reject(ret.e);

        var maybePromise = tryConvertToPromise(ret, this._promise);
        if (maybePromise instanceof Promise) ***REMOVED***
            maybePromise = maybePromise._target();
            if (maybePromise._isPending()) ***REMOVED***
                if (limit >= 1) this._inFlight++;
                values[index] = PENDING;
                return maybePromise._proxyPromiseArray(this, index);
            ***REMOVED*** else if (maybePromise._isFulfilled()) ***REMOVED***
                ret = maybePromise._value();
            ***REMOVED*** else ***REMOVED***
                return this._reject(maybePromise._reason());
            ***REMOVED***
        ***REMOVED***
        values[index] = ret;
    ***REMOVED***
    var totalResolved = ++this._totalResolved;
    if (totalResolved >= length) ***REMOVED***
        if (preservedValues !== null) ***REMOVED***
            this._filter(values, preservedValues);
        ***REMOVED*** else ***REMOVED***
            this._resolve(values);
        ***REMOVED***

    ***REMOVED***
***REMOVED***;

MappingPromiseArray.prototype._drainQueue = function () ***REMOVED***
    var queue = this._queue;
    var limit = this._limit;
    var values = this._values;
    while (queue.length > 0 && this._inFlight < limit) ***REMOVED***
        if (this._isResolved()) return;
        var index = queue.pop();
        this._promiseFulfilled(values[index], index);
    ***REMOVED***
***REMOVED***;

MappingPromiseArray.prototype._filter = function (booleans, values) ***REMOVED***
    var len = values.length;
    var ret = new Array(len);
    var j = 0;
    for (var i = 0; i < len; ++i) ***REMOVED***
        if (booleans[i]) ret[j++] = values[i];
    ***REMOVED***
    ret.length = j;
    this._resolve(ret);
***REMOVED***;

MappingPromiseArray.prototype.preservedValues = function () ***REMOVED***
    return this._preservedValues;
***REMOVED***;

function map(promises, fn, options, _filter) ***REMOVED***
    var limit = typeof options === "object" && options !== null
        ? options.concurrency
        : 0;
    limit = typeof limit === "number" &&
        isFinite(limit) && limit >= 1 ? limit : 0;
    return new MappingPromiseArray(promises, fn, limit, _filter);
***REMOVED***

Promise.prototype.map = function (fn, options) ***REMOVED***
    if (typeof fn !== "function") return apiRejection("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");

    return map(this, fn, options, null).promise();
***REMOVED***;

Promise.map = function (promises, fn, options, _filter) ***REMOVED***
    if (typeof fn !== "function") return apiRejection("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");
    return map(promises, fn, options, _filter).promise();
***REMOVED***;


***REMOVED***;

***REMOVED***,***REMOVED***"./async.js":2,"./util.js":38***REMOVED***],20:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports =
function(Promise, INTERNAL, tryConvertToPromise, apiRejection) ***REMOVED***
var util = _dereq_("./util.js");
var tryCatch = util.tryCatch;

Promise.method = function (fn) ***REMOVED***
    if (typeof fn !== "function") ***REMOVED***
        throw new Promise.TypeError("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");
    ***REMOVED***
    return function () ***REMOVED***
        var ret = new Promise(INTERNAL);
        ret._captureStackTrace();
        ret._pushContext();
        var value = tryCatch(fn).apply(this, arguments);
        ret._popContext();
        ret._resolveFromSyncValue(value);
        return ret;
    ***REMOVED***;
***REMOVED***;

Promise.attempt = Promise["try"] = function (fn, args, ctx) ***REMOVED***
    if (typeof fn !== "function") ***REMOVED***
        return apiRejection("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");
    ***REMOVED***
    var ret = new Promise(INTERNAL);
    ret._captureStackTrace();
    ret._pushContext();
    var value = util.isArray(args)
        ? tryCatch(fn).apply(ctx, args)
        : tryCatch(fn).call(ctx, args);
    ret._popContext();
    ret._resolveFromSyncValue(value);
    return ret;
***REMOVED***;

Promise.prototype._resolveFromSyncValue = function (value) ***REMOVED***
    if (value === util.errorObj) ***REMOVED***
        this._rejectCallback(value.e, false, true);
    ***REMOVED*** else ***REMOVED***
        this._resolveCallback(value, true);
    ***REMOVED***
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],21:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise) ***REMOVED***
var util = _dereq_("./util.js");
var async = _dereq_("./async.js");
var tryCatch = util.tryCatch;
var errorObj = util.errorObj;

function spreadAdapter(val, nodeback) ***REMOVED***
    var promise = this;
    if (!util.isArray(val)) return successAdapter.call(promise, val, nodeback);
    var ret =
        tryCatch(nodeback).apply(promise._boundValue(), [null].concat(val));
    if (ret === errorObj) ***REMOVED***
        async.throwLater(ret.e);
    ***REMOVED***
***REMOVED***

function successAdapter(val, nodeback) ***REMOVED***
    var promise = this;
    var receiver = promise._boundValue();
    var ret = val === undefined
        ? tryCatch(nodeback).call(receiver, null)
        : tryCatch(nodeback).call(receiver, null, val);
    if (ret === errorObj) ***REMOVED***
        async.throwLater(ret.e);
    ***REMOVED***
***REMOVED***
function errorAdapter(reason, nodeback) ***REMOVED***
    var promise = this;
    if (!reason) ***REMOVED***
        var target = promise._target();
        var newReason = target._getCarriedStackTrace();
        newReason.cause = reason;
        reason = newReason;
    ***REMOVED***
    var ret = tryCatch(nodeback).call(promise._boundValue(), reason);
    if (ret === errorObj) ***REMOVED***
        async.throwLater(ret.e);
    ***REMOVED***
***REMOVED***

Promise.prototype.asCallback =
Promise.prototype.nodeify = function (nodeback, options) ***REMOVED***
    if (typeof nodeback == "function") ***REMOVED***
        var adapter = successAdapter;
        if (options !== undefined && Object(options).spread) ***REMOVED***
            adapter = spreadAdapter;
        ***REMOVED***
        this._then(
            adapter,
            errorAdapter,
            undefined,
            this,
            nodeback
        );
    ***REMOVED***
    return this;
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./async.js":2,"./util.js":38***REMOVED***],22:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, PromiseArray) ***REMOVED***
var util = _dereq_("./util.js");
var async = _dereq_("./async.js");
var tryCatch = util.tryCatch;
var errorObj = util.errorObj;

Promise.prototype.progressed = function (handler) ***REMOVED***
    return this._then(undefined, undefined, handler, undefined, undefined);
***REMOVED***;

Promise.prototype._progress = function (progressValue) ***REMOVED***
    if (this._isFollowingOrFulfilledOrRejected()) return;
    this._target()._progressUnchecked(progressValue);

***REMOVED***;

Promise.prototype._progressHandlerAt = function (index) ***REMOVED***
    return index === 0
        ? this._progressHandler0
        : this[(index << 2) + index - 5 + 2];
***REMOVED***;

Promise.prototype._doProgressWith = function (progression) ***REMOVED***
    var progressValue = progression.value;
    var handler = progression.handler;
    var promise = progression.promise;
    var receiver = progression.receiver;

    var ret = tryCatch(handler).call(receiver, progressValue);
    if (ret === errorObj) ***REMOVED***
        if (ret.e != null &&
            ret.e.name !== "StopProgressPropagation") ***REMOVED***
            var trace = util.canAttachTrace(ret.e)
                ? ret.e : new Error(util.toString(ret.e));
            promise._attachExtraTrace(trace);
            promise._progress(ret.e);
        ***REMOVED***
    ***REMOVED*** else if (ret instanceof Promise) ***REMOVED***
        ret._then(promise._progress, null, null, promise, undefined);
    ***REMOVED*** else ***REMOVED***
        promise._progress(ret);
    ***REMOVED***
***REMOVED***;


Promise.prototype._progressUnchecked = function (progressValue) ***REMOVED***
    var len = this._length();
    var progress = this._progress;
    for (var i = 0; i < len; i++) ***REMOVED***
        var handler = this._progressHandlerAt(i);
        var promise = this._promiseAt(i);
        if (!(promise instanceof Promise)) ***REMOVED***
            var receiver = this._receiverAt(i);
            if (typeof handler === "function") ***REMOVED***
                handler.call(receiver, progressValue, promise);
            ***REMOVED*** else if (receiver instanceof PromiseArray &&
                       !receiver._isResolved()) ***REMOVED***
                receiver._promiseProgressed(progressValue, promise);
            ***REMOVED***
            continue;
        ***REMOVED***

        if (typeof handler === "function") ***REMOVED***
            async.invoke(this._doProgressWith, this, ***REMOVED***
                handler: handler,
                promise: promise,
                receiver: this._receiverAt(i),
                value: progressValue
            ***REMOVED***);
        ***REMOVED*** else ***REMOVED***
            async.invoke(progress, promise, progressValue);
        ***REMOVED***
    ***REMOVED***
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./async.js":2,"./util.js":38***REMOVED***],23:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function() ***REMOVED***
var makeSelfResolutionError = function () ***REMOVED***
    return new TypeError("circular promise resolution chain\u000a\u000a    See http://goo.gl/LhFpo0\u000a");
***REMOVED***;
var reflect = function() ***REMOVED***
    return new Promise.PromiseInspection(this._target());
***REMOVED***;
var apiRejection = function(msg) ***REMOVED***
    return Promise.reject(new TypeError(msg));
***REMOVED***;

var util = _dereq_("./util.js");

var getDomain;
if (util.isNode) ***REMOVED***
    getDomain = function() ***REMOVED***
        var ret = process.domain;
        if (ret === undefined) ret = null;
        return ret;
    ***REMOVED***;
***REMOVED*** else ***REMOVED***
    getDomain = function() ***REMOVED***
        return null;
    ***REMOVED***;
***REMOVED***
util.notEnumerableProp(Promise, "_getDomain", getDomain);

var UNDEFINED_BINDING = ***REMOVED******REMOVED***;
var async = _dereq_("./async.js");
var errors = _dereq_("./errors.js");
var TypeError = Promise.TypeError = errors.TypeError;
Promise.RangeError = errors.RangeError;
Promise.CancellationError = errors.CancellationError;
Promise.TimeoutError = errors.TimeoutError;
Promise.OperationalError = errors.OperationalError;
Promise.RejectionError = errors.OperationalError;
Promise.AggregateError = errors.AggregateError;
var INTERNAL = function()***REMOVED******REMOVED***;
var APPLY = ***REMOVED******REMOVED***;
var NEXT_FILTER = ***REMOVED***e: null***REMOVED***;
var tryConvertToPromise = _dereq_("./thenables.js")(Promise, INTERNAL);
var PromiseArray =
    _dereq_("./promise_array.js")(Promise, INTERNAL,
                                    tryConvertToPromise, apiRejection);
var CapturedTrace = _dereq_("./captured_trace.js")();
var isDebugging = _dereq_("./debuggability.js")(Promise, CapturedTrace);
 /*jshint unused:false*/
var createContext =
    _dereq_("./context.js")(Promise, CapturedTrace, isDebugging);
var CatchFilter = _dereq_("./catch_filter.js")(NEXT_FILTER);
var PromiseResolver = _dereq_("./promise_resolver.js");
var nodebackForPromise = PromiseResolver._nodebackForPromise;
var errorObj = util.errorObj;
var tryCatch = util.tryCatch;
function Promise(resolver) ***REMOVED***
    if (typeof resolver !== "function") ***REMOVED***
        throw new TypeError("the promise constructor requires a resolver function\u000a\u000a    See http://goo.gl/EC22Yn\u000a");
    ***REMOVED***
    if (this.constructor !== Promise) ***REMOVED***
        throw new TypeError("the promise constructor cannot be invoked directly\u000a\u000a    See http://goo.gl/KsIlge\u000a");
    ***REMOVED***
    this._bitField = 0;
    this._fulfillmentHandler0 = undefined;
    this._rejectionHandler0 = undefined;
    this._progressHandler0 = undefined;
    this._promise0 = undefined;
    this._receiver0 = undefined;
    this._settledValue = undefined;
    if (resolver !== INTERNAL) this._resolveFromResolver(resolver);
***REMOVED***

Promise.prototype.toString = function () ***REMOVED***
    return "[object Promise]";
***REMOVED***;

Promise.prototype.caught = Promise.prototype["catch"] = function (fn) ***REMOVED***
    var len = arguments.length;
    if (len > 1) ***REMOVED***
        var catchInstances = new Array(len - 1),
            j = 0, i;
        for (i = 0; i < len - 1; ++i) ***REMOVED***
            var item = arguments[i];
            if (typeof item === "function") ***REMOVED***
                catchInstances[j++] = item;
            ***REMOVED*** else ***REMOVED***
                return Promise.reject(
                    new TypeError("Catch filter must inherit from Error or be a simple predicate function\u000a\u000a    See http://goo.gl/o84o68\u000a"));
            ***REMOVED***
        ***REMOVED***
        catchInstances.length = j;
        fn = arguments[i];
        var catchFilter = new CatchFilter(catchInstances, fn, this);
        return this._then(undefined, catchFilter.doFilter, undefined,
            catchFilter, undefined);
    ***REMOVED***
    return this._then(undefined, fn, undefined, undefined, undefined);
***REMOVED***;

Promise.prototype.reflect = function () ***REMOVED***
    return this._then(reflect, reflect, undefined, this, undefined);
***REMOVED***;

Promise.prototype.then = function (didFulfill, didReject, didProgress) ***REMOVED***
    if (isDebugging() && arguments.length > 0 &&
        typeof didFulfill !== "function" &&
        typeof didReject !== "function") ***REMOVED***
        var msg = ".then() only accepts functions but was passed: " +
                util.classString(didFulfill);
        if (arguments.length > 1) ***REMOVED***
            msg += ", " + util.classString(didReject);
        ***REMOVED***
        this._warn(msg);
    ***REMOVED***
    return this._then(didFulfill, didReject, didProgress,
        undefined, undefined);
***REMOVED***;

Promise.prototype.done = function (didFulfill, didReject, didProgress) ***REMOVED***
    var promise = this._then(didFulfill, didReject, didProgress,
        undefined, undefined);
    promise._setIsFinal();
***REMOVED***;

Promise.prototype.spread = function (didFulfill, didReject) ***REMOVED***
    return this.all()._then(didFulfill, didReject, undefined, APPLY, undefined);
***REMOVED***;

Promise.prototype.isCancellable = function () ***REMOVED***
    return !this.isResolved() &&
        this._cancellable();
***REMOVED***;

Promise.prototype.toJSON = function () ***REMOVED***
    var ret = ***REMOVED***
        isFulfilled: false,
        isRejected: false,
        fulfillmentValue: undefined,
        rejectionReason: undefined
    ***REMOVED***;
    if (this.isFulfilled()) ***REMOVED***
        ret.fulfillmentValue = this.value();
        ret.isFulfilled = true;
    ***REMOVED*** else if (this.isRejected()) ***REMOVED***
        ret.rejectionReason = this.reason();
        ret.isRejected = true;
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.prototype.all = function () ***REMOVED***
    return new PromiseArray(this).promise();
***REMOVED***;

Promise.prototype.error = function (fn) ***REMOVED***
    return this.caught(util.originatesFromRejection, fn);
***REMOVED***;

Promise.is = function (val) ***REMOVED***
    return val instanceof Promise;
***REMOVED***;

Promise.fromNode = function(fn) ***REMOVED***
    var ret = new Promise(INTERNAL);
    var result = tryCatch(fn)(nodebackForPromise(ret));
    if (result === errorObj) ***REMOVED***
        ret._rejectCallback(result.e, true, true);
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.all = function (promises) ***REMOVED***
    return new PromiseArray(promises).promise();
***REMOVED***;

Promise.defer = Promise.pending = function () ***REMOVED***
    var promise = new Promise(INTERNAL);
    return new PromiseResolver(promise);
***REMOVED***;

Promise.cast = function (obj) ***REMOVED***
    var ret = tryConvertToPromise(obj);
    if (!(ret instanceof Promise)) ***REMOVED***
        var val = ret;
        ret = new Promise(INTERNAL);
        ret._fulfillUnchecked(val);
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.resolve = Promise.fulfilled = Promise.cast;

Promise.reject = Promise.rejected = function (reason) ***REMOVED***
    var ret = new Promise(INTERNAL);
    ret._captureStackTrace();
    ret._rejectCallback(reason, true);
    return ret;
***REMOVED***;

Promise.setScheduler = function(fn) ***REMOVED***
    if (typeof fn !== "function") throw new TypeError("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");
    var prev = async._schedule;
    async._schedule = fn;
    return prev;
***REMOVED***;

Promise.prototype._then = function (
    didFulfill,
    didReject,
    didProgress,
    receiver,
    internalData
) ***REMOVED***
    var haveInternalData = internalData !== undefined;
    var ret = haveInternalData ? internalData : new Promise(INTERNAL);

    if (!haveInternalData) ***REMOVED***
        ret._propagateFrom(this, 4 | 1);
        ret._captureStackTrace();
    ***REMOVED***

    var target = this._target();
    if (target !== this) ***REMOVED***
        if (receiver === undefined) receiver = this._boundTo;
        if (!haveInternalData) ret._setIsMigrated();
    ***REMOVED***

    var callbackIndex = target._addCallbacks(didFulfill,
                                             didReject,
                                             didProgress,
                                             ret,
                                             receiver,
                                             getDomain());

    if (target._isResolved() && !target._isSettlePromisesQueued()) ***REMOVED***
        async.invoke(
            target._settlePromiseAtPostResolution, target, callbackIndex);
    ***REMOVED***

    return ret;
***REMOVED***;

Promise.prototype._settlePromiseAtPostResolution = function (index) ***REMOVED***
    if (this._isRejectionUnhandled()) this._unsetRejectionIsUnhandled();
    this._settlePromiseAt(index);
***REMOVED***;

Promise.prototype._length = function () ***REMOVED***
    return this._bitField & 131071;
***REMOVED***;

Promise.prototype._isFollowingOrFulfilledOrRejected = function () ***REMOVED***
    return (this._bitField & 939524096) > 0;
***REMOVED***;

Promise.prototype._isFollowing = function () ***REMOVED***
    return (this._bitField & 536870912) === 536870912;
***REMOVED***;

Promise.prototype._setLength = function (len) ***REMOVED***
    this._bitField = (this._bitField & -131072) |
        (len & 131071);
***REMOVED***;

Promise.prototype._setFulfilled = function () ***REMOVED***
    this._bitField = this._bitField | 268435456;
***REMOVED***;

Promise.prototype._setRejected = function () ***REMOVED***
    this._bitField = this._bitField | 134217728;
***REMOVED***;

Promise.prototype._setFollowing = function () ***REMOVED***
    this._bitField = this._bitField | 536870912;
***REMOVED***;

Promise.prototype._setIsFinal = function () ***REMOVED***
    this._bitField = this._bitField | 33554432;
***REMOVED***;

Promise.prototype._isFinal = function () ***REMOVED***
    return (this._bitField & 33554432) > 0;
***REMOVED***;

Promise.prototype._cancellable = function () ***REMOVED***
    return (this._bitField & 67108864) > 0;
***REMOVED***;

Promise.prototype._setCancellable = function () ***REMOVED***
    this._bitField = this._bitField | 67108864;
***REMOVED***;

Promise.prototype._unsetCancellable = function () ***REMOVED***
    this._bitField = this._bitField & (~67108864);
***REMOVED***;

Promise.prototype._setIsMigrated = function () ***REMOVED***
    this._bitField = this._bitField | 4194304;
***REMOVED***;

Promise.prototype._unsetIsMigrated = function () ***REMOVED***
    this._bitField = this._bitField & (~4194304);
***REMOVED***;

Promise.prototype._isMigrated = function () ***REMOVED***
    return (this._bitField & 4194304) > 0;
***REMOVED***;

Promise.prototype._receiverAt = function (index) ***REMOVED***
    var ret = index === 0
        ? this._receiver0
        : this[
            index * 5 - 5 + 4];
    if (ret === UNDEFINED_BINDING) ***REMOVED***
        return undefined;
    ***REMOVED*** else if (ret === undefined && this._isBound()) ***REMOVED***
        return this._boundValue();
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.prototype._promiseAt = function (index) ***REMOVED***
    return index === 0
        ? this._promise0
        : this[index * 5 - 5 + 3];
***REMOVED***;

Promise.prototype._fulfillmentHandlerAt = function (index) ***REMOVED***
    return index === 0
        ? this._fulfillmentHandler0
        : this[index * 5 - 5 + 0];
***REMOVED***;

Promise.prototype._rejectionHandlerAt = function (index) ***REMOVED***
    return index === 0
        ? this._rejectionHandler0
        : this[index * 5 - 5 + 1];
***REMOVED***;

Promise.prototype._boundValue = function() ***REMOVED***
    var ret = this._boundTo;
    if (ret !== undefined) ***REMOVED***
        if (ret instanceof Promise) ***REMOVED***
            if (ret.isFulfilled()) ***REMOVED***
                return ret.value();
            ***REMOVED*** else ***REMOVED***
                return undefined;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.prototype._migrateCallbacks = function (follower, index) ***REMOVED***
    var fulfill = follower._fulfillmentHandlerAt(index);
    var reject = follower._rejectionHandlerAt(index);
    var progress = follower._progressHandlerAt(index);
    var promise = follower._promiseAt(index);
    var receiver = follower._receiverAt(index);
    if (promise instanceof Promise) promise._setIsMigrated();
    if (receiver === undefined) receiver = UNDEFINED_BINDING;
    this._addCallbacks(fulfill, reject, progress, promise, receiver, null);
***REMOVED***;

Promise.prototype._addCallbacks = function (
    fulfill,
    reject,
    progress,
    promise,
    receiver,
    domain
) ***REMOVED***
    var index = this._length();

    if (index >= 131071 - 5) ***REMOVED***
        index = 0;
        this._setLength(0);
    ***REMOVED***

    if (index === 0) ***REMOVED***
        this._promise0 = promise;
        if (receiver !== undefined) this._receiver0 = receiver;
        if (typeof fulfill === "function" && !this._isCarryingStackTrace()) ***REMOVED***
            this._fulfillmentHandler0 =
                domain === null ? fulfill : domain.bind(fulfill);
        ***REMOVED***
        if (typeof reject === "function") ***REMOVED***
            this._rejectionHandler0 =
                domain === null ? reject : domain.bind(reject);
        ***REMOVED***
        if (typeof progress === "function") ***REMOVED***
            this._progressHandler0 =
                domain === null ? progress : domain.bind(progress);
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        var base = index * 5 - 5;
        this[base + 3] = promise;
        this[base + 4] = receiver;
        if (typeof fulfill === "function") ***REMOVED***
            this[base + 0] =
                domain === null ? fulfill : domain.bind(fulfill);
        ***REMOVED***
        if (typeof reject === "function") ***REMOVED***
            this[base + 1] =
                domain === null ? reject : domain.bind(reject);
        ***REMOVED***
        if (typeof progress === "function") ***REMOVED***
            this[base + 2] =
                domain === null ? progress : domain.bind(progress);
        ***REMOVED***
    ***REMOVED***
    this._setLength(index + 1);
    return index;
***REMOVED***;

Promise.prototype._setProxyHandlers = function (receiver, promiseSlotValue) ***REMOVED***
    var index = this._length();

    if (index >= 131071 - 5) ***REMOVED***
        index = 0;
        this._setLength(0);
    ***REMOVED***
    if (index === 0) ***REMOVED***
        this._promise0 = promiseSlotValue;
        this._receiver0 = receiver;
    ***REMOVED*** else ***REMOVED***
        var base = index * 5 - 5;
        this[base + 3] = promiseSlotValue;
        this[base + 4] = receiver;
    ***REMOVED***
    this._setLength(index + 1);
***REMOVED***;

Promise.prototype._proxyPromiseArray = function (promiseArray, index) ***REMOVED***
    this._setProxyHandlers(promiseArray, index);
***REMOVED***;

Promise.prototype._resolveCallback = function(value, shouldBind) ***REMOVED***
    if (this._isFollowingOrFulfilledOrRejected()) return;
    if (value === this)
        return this._rejectCallback(makeSelfResolutionError(), false, true);
    var maybePromise = tryConvertToPromise(value, this);
    if (!(maybePromise instanceof Promise)) return this._fulfill(value);

    var propagationFlags = 1 | (shouldBind ? 4 : 0);
    this._propagateFrom(maybePromise, propagationFlags);
    var promise = maybePromise._target();
    if (promise._isPending()) ***REMOVED***
        var len = this._length();
        for (var i = 0; i < len; ++i) ***REMOVED***
            promise._migrateCallbacks(this, i);
        ***REMOVED***
        this._setFollowing();
        this._setLength(0);
        this._setFollowee(promise);
    ***REMOVED*** else if (promise._isFulfilled()) ***REMOVED***
        this._fulfillUnchecked(promise._value());
    ***REMOVED*** else ***REMOVED***
        this._rejectUnchecked(promise._reason(),
            promise._getCarriedStackTrace());
    ***REMOVED***
***REMOVED***;

Promise.prototype._rejectCallback =
function(reason, synchronous, shouldNotMarkOriginatingFromRejection) ***REMOVED***
    if (!shouldNotMarkOriginatingFromRejection) ***REMOVED***
        util.markAsOriginatingFromRejection(reason);
    ***REMOVED***
    var trace = util.ensureErrorObject(reason);
    var hasStack = trace === reason;
    this._attachExtraTrace(trace, synchronous ? hasStack : false);
    this._reject(reason, hasStack ? undefined : trace);
***REMOVED***;

Promise.prototype._resolveFromResolver = function (resolver) ***REMOVED***
    var promise = this;
    this._captureStackTrace();
    this._pushContext();
    var synchronous = true;
    var r = tryCatch(resolver)(function(value) ***REMOVED***
        if (promise === null) return;
        promise._resolveCallback(value);
        promise = null;
    ***REMOVED***, function (reason) ***REMOVED***
        if (promise === null) return;
        promise._rejectCallback(reason, synchronous);
        promise = null;
    ***REMOVED***);
    synchronous = false;
    this._popContext();

    if (r !== undefined && r === errorObj && promise !== null) ***REMOVED***
        promise._rejectCallback(r.e, true, true);
        promise = null;
    ***REMOVED***
***REMOVED***;

Promise.prototype._settlePromiseFromHandler = function (
    handler, receiver, value, promise
) ***REMOVED***
    if (promise._isRejected()) return;
    promise._pushContext();
    var x;
    if (receiver === APPLY && !this._isRejected()) ***REMOVED***
        x = tryCatch(handler).apply(this._boundValue(), value);
    ***REMOVED*** else ***REMOVED***
        x = tryCatch(handler).call(receiver, value);
    ***REMOVED***
    promise._popContext();

    if (x === errorObj || x === promise || x === NEXT_FILTER) ***REMOVED***
        var err = x === promise ? makeSelfResolutionError() : x.e;
        promise._rejectCallback(err, false, true);
    ***REMOVED*** else ***REMOVED***
        promise._resolveCallback(x);
    ***REMOVED***
***REMOVED***;

Promise.prototype._target = function() ***REMOVED***
    var ret = this;
    while (ret._isFollowing()) ret = ret._followee();
    return ret;
***REMOVED***;

Promise.prototype._followee = function() ***REMOVED***
    return this._rejectionHandler0;
***REMOVED***;

Promise.prototype._setFollowee = function(promise) ***REMOVED***
    this._rejectionHandler0 = promise;
***REMOVED***;

Promise.prototype._cleanValues = function () ***REMOVED***
    if (this._cancellable()) ***REMOVED***
        this._cancellationParent = undefined;
    ***REMOVED***
***REMOVED***;

Promise.prototype._propagateFrom = function (parent, flags) ***REMOVED***
    if ((flags & 1) > 0 && parent._cancellable()) ***REMOVED***
        this._setCancellable();
        this._cancellationParent = parent;
    ***REMOVED***
    if ((flags & 4) > 0 && parent._isBound()) ***REMOVED***
        this._setBoundTo(parent._boundTo);
    ***REMOVED***
***REMOVED***;

Promise.prototype._fulfill = function (value) ***REMOVED***
    if (this._isFollowingOrFulfilledOrRejected()) return;
    this._fulfillUnchecked(value);
***REMOVED***;

Promise.prototype._reject = function (reason, carriedStackTrace) ***REMOVED***
    if (this._isFollowingOrFulfilledOrRejected()) return;
    this._rejectUnchecked(reason, carriedStackTrace);
***REMOVED***;

Promise.prototype._settlePromiseAt = function (index) ***REMOVED***
    var promise = this._promiseAt(index);
    var isPromise = promise instanceof Promise;

    if (isPromise && promise._isMigrated()) ***REMOVED***
        promise._unsetIsMigrated();
        return async.invoke(this._settlePromiseAt, this, index);
    ***REMOVED***
    var handler = this._isFulfilled()
        ? this._fulfillmentHandlerAt(index)
        : this._rejectionHandlerAt(index);

    var carriedStackTrace =
        this._isCarryingStackTrace() ? this._getCarriedStackTrace() : undefined;
    var value = this._settledValue;
    var receiver = this._receiverAt(index);
    this._clearCallbackDataAtIndex(index);

    if (typeof handler === "function") ***REMOVED***
        if (!isPromise) ***REMOVED***
            handler.call(receiver, value, promise);
        ***REMOVED*** else ***REMOVED***
            this._settlePromiseFromHandler(handler, receiver, value, promise);
        ***REMOVED***
    ***REMOVED*** else if (receiver instanceof PromiseArray) ***REMOVED***
        if (!receiver._isResolved()) ***REMOVED***
            if (this._isFulfilled()) ***REMOVED***
                receiver._promiseFulfilled(value, promise);
            ***REMOVED***
            else ***REMOVED***
                receiver._promiseRejected(value, promise);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED*** else if (isPromise) ***REMOVED***
        if (this._isFulfilled()) ***REMOVED***
            promise._fulfill(value);
        ***REMOVED*** else ***REMOVED***
            promise._reject(value, carriedStackTrace);
        ***REMOVED***
    ***REMOVED***

    if (index >= 4 && (index & 31) === 4)
        async.invokeLater(this._setLength, this, 0);
***REMOVED***;

Promise.prototype._clearCallbackDataAtIndex = function(index) ***REMOVED***
    if (index === 0) ***REMOVED***
        if (!this._isCarryingStackTrace()) ***REMOVED***
            this._fulfillmentHandler0 = undefined;
        ***REMOVED***
        this._rejectionHandler0 =
        this._progressHandler0 =
        this._receiver0 =
        this._promise0 = undefined;
    ***REMOVED*** else ***REMOVED***
        var base = index * 5 - 5;
        this[base + 3] =
        this[base + 4] =
        this[base + 0] =
        this[base + 1] =
        this[base + 2] = undefined;
    ***REMOVED***
***REMOVED***;

Promise.prototype._isSettlePromisesQueued = function () ***REMOVED***
    return (this._bitField &
            -1073741824) === -1073741824;
***REMOVED***;

Promise.prototype._setSettlePromisesQueued = function () ***REMOVED***
    this._bitField = this._bitField | -1073741824;
***REMOVED***;

Promise.prototype._unsetSettlePromisesQueued = function () ***REMOVED***
    this._bitField = this._bitField & (~-1073741824);
***REMOVED***;

Promise.prototype._queueSettlePromises = function() ***REMOVED***
    async.settlePromises(this);
    this._setSettlePromisesQueued();
***REMOVED***;

Promise.prototype._fulfillUnchecked = function (value) ***REMOVED***
    if (value === this) ***REMOVED***
        var err = makeSelfResolutionError();
        this._attachExtraTrace(err);
        return this._rejectUnchecked(err, undefined);
    ***REMOVED***
    this._setFulfilled();
    this._settledValue = value;
    this._cleanValues();

    if (this._length() > 0) ***REMOVED***
        this._queueSettlePromises();
    ***REMOVED***
***REMOVED***;

Promise.prototype._rejectUncheckedCheckError = function (reason) ***REMOVED***
    var trace = util.ensureErrorObject(reason);
    this._rejectUnchecked(reason, trace === reason ? undefined : trace);
***REMOVED***;

Promise.prototype._rejectUnchecked = function (reason, trace) ***REMOVED***
    if (reason === this) ***REMOVED***
        var err = makeSelfResolutionError();
        this._attachExtraTrace(err);
        return this._rejectUnchecked(err);
    ***REMOVED***
    this._setRejected();
    this._settledValue = reason;
    this._cleanValues();

    if (this._isFinal()) ***REMOVED***
        async.throwLater(function(e) ***REMOVED***
            if ("stack" in e) ***REMOVED***
                async.invokeFirst(
                    CapturedTrace.unhandledRejection, undefined, e);
            ***REMOVED***
            throw e;
        ***REMOVED***, trace === undefined ? reason : trace);
        return;
    ***REMOVED***

    if (trace !== undefined && trace !== reason) ***REMOVED***
        this._setCarriedStackTrace(trace);
    ***REMOVED***

    if (this._length() > 0) ***REMOVED***
        this._queueSettlePromises();
    ***REMOVED*** else ***REMOVED***
        this._ensurePossibleRejectionHandled();
    ***REMOVED***
***REMOVED***;

Promise.prototype._settlePromises = function () ***REMOVED***
    this._unsetSettlePromisesQueued();
    var len = this._length();
    for (var i = 0; i < len; i++) ***REMOVED***
        this._settlePromiseAt(i);
    ***REMOVED***
***REMOVED***;

util.notEnumerableProp(Promise,
                       "_makeSelfResolutionError",
                       makeSelfResolutionError);

_dereq_("./progress.js")(Promise, PromiseArray);
_dereq_("./method.js")(Promise, INTERNAL, tryConvertToPromise, apiRejection);
_dereq_("./bind.js")(Promise, INTERNAL, tryConvertToPromise);
_dereq_("./finally.js")(Promise, NEXT_FILTER, tryConvertToPromise);
_dereq_("./direct_resolve.js")(Promise);
_dereq_("./synchronous_inspection.js")(Promise);
_dereq_("./join.js")(Promise, PromiseArray, tryConvertToPromise, INTERNAL);
Promise.Promise = Promise;
_dereq_('./map.js')(Promise, PromiseArray, apiRejection, tryConvertToPromise, INTERNAL);
_dereq_('./cancel.js')(Promise);
_dereq_('./using.js')(Promise, apiRejection, tryConvertToPromise, createContext);
_dereq_('./generators.js')(Promise, apiRejection, INTERNAL, tryConvertToPromise);
_dereq_('./nodeify.js')(Promise);
_dereq_('./call_get.js')(Promise);
_dereq_('./props.js')(Promise, PromiseArray, tryConvertToPromise, apiRejection);
_dereq_('./race.js')(Promise, INTERNAL, tryConvertToPromise, apiRejection);
_dereq_('./reduce.js')(Promise, PromiseArray, apiRejection, tryConvertToPromise, INTERNAL);
_dereq_('./settle.js')(Promise, PromiseArray);
_dereq_('./some.js')(Promise, PromiseArray, apiRejection);
_dereq_('./promisify.js')(Promise, INTERNAL);
_dereq_('./any.js')(Promise);
_dereq_('./each.js')(Promise, INTERNAL);
_dereq_('./timers.js')(Promise, INTERNAL);
_dereq_('./filter.js')(Promise, INTERNAL);
                                                         
    util.toFastProperties(Promise);                                          
    util.toFastProperties(Promise.prototype);                                
    function fillTypes(value) ***REMOVED***                                              
        var p = new Promise(INTERNAL);                                       
        p._fulfillmentHandler0 = value;                                      
        p._rejectionHandler0 = value;                                        
        p._progressHandler0 = value;                                         
        p._promise0 = value;                                                 
        p._receiver0 = value;                                                
        p._settledValue = value;                                             
    ***REMOVED***                                                                        
    // Complete slack tracking, opt out of field-type tracking and           
    // stabilize map                                                         
    fillTypes(***REMOVED***a: 1***REMOVED***);                                                       
    fillTypes(***REMOVED***b: 2***REMOVED***);                                                       
    fillTypes(***REMOVED***c: 3***REMOVED***);                                                       
    fillTypes(1);                                                            
    fillTypes(function()***REMOVED******REMOVED***);                                                 
    fillTypes(undefined);                                                    
    fillTypes(false);                                                        
    fillTypes(new Promise(INTERNAL));                                        
    CapturedTrace.setBounds(async.firstLineError, util.lastLineError);       
    return Promise;                                                          

***REMOVED***;

***REMOVED***,***REMOVED***"./any.js":1,"./async.js":2,"./bind.js":3,"./call_get.js":5,"./cancel.js":6,"./captured_trace.js":7,"./catch_filter.js":8,"./context.js":9,"./debuggability.js":10,"./direct_resolve.js":11,"./each.js":12,"./errors.js":13,"./filter.js":15,"./finally.js":16,"./generators.js":17,"./join.js":18,"./map.js":19,"./method.js":20,"./nodeify.js":21,"./progress.js":22,"./promise_array.js":24,"./promise_resolver.js":25,"./promisify.js":26,"./props.js":27,"./race.js":29,"./reduce.js":30,"./settle.js":32,"./some.js":33,"./synchronous_inspection.js":34,"./thenables.js":35,"./timers.js":36,"./using.js":37,"./util.js":38***REMOVED***],24:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, INTERNAL, tryConvertToPromise,
    apiRejection) ***REMOVED***
var util = _dereq_("./util.js");
var isArray = util.isArray;

function toResolutionValue(val) ***REMOVED***
    switch(val) ***REMOVED***
    case -2: return [];
    case -3: return ***REMOVED******REMOVED***;
    ***REMOVED***
***REMOVED***

function PromiseArray(values) ***REMOVED***
    var promise = this._promise = new Promise(INTERNAL);
    var parent;
    if (values instanceof Promise) ***REMOVED***
        parent = values;
        promise._propagateFrom(parent, 1 | 4);
    ***REMOVED***
    this._values = values;
    this._length = 0;
    this._totalResolved = 0;
    this._init(undefined, -2);
***REMOVED***
PromiseArray.prototype.length = function () ***REMOVED***
    return this._length;
***REMOVED***;

PromiseArray.prototype.promise = function () ***REMOVED***
    return this._promise;
***REMOVED***;

PromiseArray.prototype._init = function init(_, resolveValueIfEmpty) ***REMOVED***
    var values = tryConvertToPromise(this._values, this._promise);
    if (values instanceof Promise) ***REMOVED***
        values = values._target();
        this._values = values;
        if (values._isFulfilled()) ***REMOVED***
            values = values._value();
            if (!isArray(values)) ***REMOVED***
                var err = new Promise.TypeError("expecting an array, a promise or a thenable\u000a\u000a    See http://goo.gl/s8MMhc\u000a");
                this.__hardReject__(err);
                return;
            ***REMOVED***
        ***REMOVED*** else if (values._isPending()) ***REMOVED***
            values._then(
                init,
                this._reject,
                undefined,
                this,
                resolveValueIfEmpty
           );
            return;
        ***REMOVED*** else ***REMOVED***
            this._reject(values._reason());
            return;
        ***REMOVED***
    ***REMOVED*** else if (!isArray(values)) ***REMOVED***
        this._promise._reject(apiRejection("expecting an array, a promise or a thenable\u000a\u000a    See http://goo.gl/s8MMhc\u000a")._reason());
        return;
    ***REMOVED***

    if (values.length === 0) ***REMOVED***
        if (resolveValueIfEmpty === -5) ***REMOVED***
            this._resolveEmptyArray();
        ***REMOVED***
        else ***REMOVED***
            this._resolve(toResolutionValue(resolveValueIfEmpty));
        ***REMOVED***
        return;
    ***REMOVED***
    var len = this.getActualLength(values.length);
    this._length = len;
    this._values = this.shouldCopyValues() ? new Array(len) : this._values;
    var promise = this._promise;
    for (var i = 0; i < len; ++i) ***REMOVED***
        var isResolved = this._isResolved();
        var maybePromise = tryConvertToPromise(values[i], promise);
        if (maybePromise instanceof Promise) ***REMOVED***
            maybePromise = maybePromise._target();
            if (isResolved) ***REMOVED***
                maybePromise._ignoreRejections();
            ***REMOVED*** else if (maybePromise._isPending()) ***REMOVED***
                maybePromise._proxyPromiseArray(this, i);
            ***REMOVED*** else if (maybePromise._isFulfilled()) ***REMOVED***
                this._promiseFulfilled(maybePromise._value(), i);
            ***REMOVED*** else ***REMOVED***
                this._promiseRejected(maybePromise._reason(), i);
            ***REMOVED***
        ***REMOVED*** else if (!isResolved) ***REMOVED***
            this._promiseFulfilled(maybePromise, i);
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

PromiseArray.prototype._isResolved = function () ***REMOVED***
    return this._values === null;
***REMOVED***;

PromiseArray.prototype._resolve = function (value) ***REMOVED***
    this._values = null;
    this._promise._fulfill(value);
***REMOVED***;

PromiseArray.prototype.__hardReject__ =
PromiseArray.prototype._reject = function (reason) ***REMOVED***
    this._values = null;
    this._promise._rejectCallback(reason, false, true);
***REMOVED***;

PromiseArray.prototype._promiseProgressed = function (progressValue, index) ***REMOVED***
    this._promise._progress(***REMOVED***
        index: index,
        value: progressValue
    ***REMOVED***);
***REMOVED***;


PromiseArray.prototype._promiseFulfilled = function (value, index) ***REMOVED***
    this._values[index] = value;
    var totalResolved = ++this._totalResolved;
    if (totalResolved >= this._length) ***REMOVED***
        this._resolve(this._values);
    ***REMOVED***
***REMOVED***;

PromiseArray.prototype._promiseRejected = function (reason, index) ***REMOVED***
    this._totalResolved++;
    this._reject(reason);
***REMOVED***;

PromiseArray.prototype.shouldCopyValues = function () ***REMOVED***
    return true;
***REMOVED***;

PromiseArray.prototype.getActualLength = function (len) ***REMOVED***
    return len;
***REMOVED***;

return PromiseArray;
***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],25:[function(_dereq_,module,exports)***REMOVED***
"use strict";
var util = _dereq_("./util.js");
var maybeWrapAsError = util.maybeWrapAsError;
var errors = _dereq_("./errors.js");
var TimeoutError = errors.TimeoutError;
var OperationalError = errors.OperationalError;
var haveGetters = util.haveGetters;
var es5 = _dereq_("./es5.js");

function isUntypedError(obj) ***REMOVED***
    return obj instanceof Error &&
        es5.getPrototypeOf(obj) === Error.prototype;
***REMOVED***

var rErrorKey = /^(?:name|message|stack|cause)$/;
function wrapAsOperationalError(obj) ***REMOVED***
    var ret;
    if (isUntypedError(obj)) ***REMOVED***
        ret = new OperationalError(obj);
        ret.name = obj.name;
        ret.message = obj.message;
        ret.stack = obj.stack;
        var keys = es5.keys(obj);
        for (var i = 0; i < keys.length; ++i) ***REMOVED***
            var key = keys[i];
            if (!rErrorKey.test(key)) ***REMOVED***
                ret[key] = obj[key];
            ***REMOVED***
        ***REMOVED***
        return ret;
    ***REMOVED***
    util.markAsOriginatingFromRejection(obj);
    return obj;
***REMOVED***

function nodebackForPromise(promise) ***REMOVED***
    return function(err, value) ***REMOVED***
        if (promise === null) return;

        if (err) ***REMOVED***
            var wrapped = wrapAsOperationalError(maybeWrapAsError(err));
            promise._attachExtraTrace(wrapped);
            promise._reject(wrapped);
        ***REMOVED*** else if (arguments.length > 2) ***REMOVED***
            var $_len = arguments.length;var args = new Array($_len - 1); for(var $_i = 1; $_i < $_len; ++$_i) ***REMOVED***args[$_i - 1] = arguments[$_i];***REMOVED***
            promise._fulfill(args);
        ***REMOVED*** else ***REMOVED***
            promise._fulfill(value);
        ***REMOVED***

        promise = null;
    ***REMOVED***;
***REMOVED***


var PromiseResolver;
if (!haveGetters) ***REMOVED***
    PromiseResolver = function (promise) ***REMOVED***
        this.promise = promise;
        this.asCallback = nodebackForPromise(promise);
        this.callback = this.asCallback;
    ***REMOVED***;
***REMOVED***
else ***REMOVED***
    PromiseResolver = function (promise) ***REMOVED***
        this.promise = promise;
    ***REMOVED***;
***REMOVED***
if (haveGetters) ***REMOVED***
    var prop = ***REMOVED***
        get: function() ***REMOVED***
            return nodebackForPromise(this.promise);
        ***REMOVED***
    ***REMOVED***;
    es5.defineProperty(PromiseResolver.prototype, "asCallback", prop);
    es5.defineProperty(PromiseResolver.prototype, "callback", prop);
***REMOVED***

PromiseResolver._nodebackForPromise = nodebackForPromise;

PromiseResolver.prototype.toString = function () ***REMOVED***
    return "[object PromiseResolver]";
***REMOVED***;

PromiseResolver.prototype.resolve =
PromiseResolver.prototype.fulfill = function (value) ***REMOVED***
    if (!(this instanceof PromiseResolver)) ***REMOVED***
        throw new TypeError("Illegal invocation, resolver resolve/reject must be called within a resolver context. Consider using the promise constructor instead.\u000a\u000a    See http://goo.gl/sdkXL9\u000a");
    ***REMOVED***
    this.promise._resolveCallback(value);
***REMOVED***;

PromiseResolver.prototype.reject = function (reason) ***REMOVED***
    if (!(this instanceof PromiseResolver)) ***REMOVED***
        throw new TypeError("Illegal invocation, resolver resolve/reject must be called within a resolver context. Consider using the promise constructor instead.\u000a\u000a    See http://goo.gl/sdkXL9\u000a");
    ***REMOVED***
    this.promise._rejectCallback(reason);
***REMOVED***;

PromiseResolver.prototype.progress = function (value) ***REMOVED***
    if (!(this instanceof PromiseResolver)) ***REMOVED***
        throw new TypeError("Illegal invocation, resolver resolve/reject must be called within a resolver context. Consider using the promise constructor instead.\u000a\u000a    See http://goo.gl/sdkXL9\u000a");
    ***REMOVED***
    this.promise._progress(value);
***REMOVED***;

PromiseResolver.prototype.cancel = function (err) ***REMOVED***
    this.promise.cancel(err);
***REMOVED***;

PromiseResolver.prototype.timeout = function () ***REMOVED***
    this.reject(new TimeoutError("timeout"));
***REMOVED***;

PromiseResolver.prototype.isResolved = function () ***REMOVED***
    return this.promise.isResolved();
***REMOVED***;

PromiseResolver.prototype.toJSON = function () ***REMOVED***
    return this.promise.toJSON();
***REMOVED***;

module.exports = PromiseResolver;

***REMOVED***,***REMOVED***"./errors.js":13,"./es5.js":14,"./util.js":38***REMOVED***],26:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, INTERNAL) ***REMOVED***
var THIS = ***REMOVED******REMOVED***;
var util = _dereq_("./util.js");
var nodebackForPromise = _dereq_("./promise_resolver.js")
    ._nodebackForPromise;
var withAppended = util.withAppended;
var maybeWrapAsError = util.maybeWrapAsError;
var canEvaluate = util.canEvaluate;
var TypeError = _dereq_("./errors").TypeError;
var defaultSuffix = "Async";
var defaultPromisified = ***REMOVED***__isPromisified__: true***REMOVED***;
var noCopyProps = [
    "arity",    "length",
    "name",
    "arguments",
    "caller",
    "callee",
    "prototype",
    "__isPromisified__"
];
var noCopyPropsPattern = new RegExp("^(?:" + noCopyProps.join("|") + ")$");

var defaultFilter = function(name) ***REMOVED***
    return util.isIdentifier(name) &&
        name.charAt(0) !== "_" &&
        name !== "constructor";
***REMOVED***;

function propsFilter(key) ***REMOVED***
    return !noCopyPropsPattern.test(key);
***REMOVED***

function isPromisified(fn) ***REMOVED***
    try ***REMOVED***
        return fn.__isPromisified__ === true;
    ***REMOVED***
    catch (e) ***REMOVED***
        return false;
    ***REMOVED***
***REMOVED***

function hasPromisified(obj, key, suffix) ***REMOVED***
    var val = util.getDataPropertyOrDefault(obj, key + suffix,
                                            defaultPromisified);
    return val ? isPromisified(val) : false;
***REMOVED***
function checkValid(ret, suffix, suffixRegexp) ***REMOVED***
    for (var i = 0; i < ret.length; i += 2) ***REMOVED***
        var key = ret[i];
        if (suffixRegexp.test(key)) ***REMOVED***
            var keyWithoutAsyncSuffix = key.replace(suffixRegexp, "");
            for (var j = 0; j < ret.length; j += 2) ***REMOVED***
                if (ret[j] === keyWithoutAsyncSuffix) ***REMOVED***
                    throw new TypeError("Cannot promisify an API that has normal methods with '%s'-suffix\u000a\u000a    See http://goo.gl/iWrZbw\u000a"
                        .replace("%s", suffix));
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***

function promisifiableMethods(obj, suffix, suffixRegexp, filter) ***REMOVED***
    var keys = util.inheritedDataKeys(obj);
    var ret = [];
    for (var i = 0; i < keys.length; ++i) ***REMOVED***
        var key = keys[i];
        var value = obj[key];
        var passesDefaultFilter = filter === defaultFilter
            ? true : defaultFilter(key, value, obj);
        if (typeof value === "function" &&
            !isPromisified(value) &&
            !hasPromisified(obj, key, suffix) &&
            filter(key, value, obj, passesDefaultFilter)) ***REMOVED***
            ret.push(key, value);
        ***REMOVED***
    ***REMOVED***
    checkValid(ret, suffix, suffixRegexp);
    return ret;
***REMOVED***

var escapeIdentRegex = function(str) ***REMOVED***
    return str.replace(/([$])/, "\\$");
***REMOVED***;

var makeNodePromisifiedEval;
if (!true) ***REMOVED***
var switchCaseArgumentOrder = function(likelyArgumentCount) ***REMOVED***
    var ret = [likelyArgumentCount];
    var min = Math.max(0, likelyArgumentCount - 1 - 3);
    for(var i = likelyArgumentCount - 1; i >= min; --i) ***REMOVED***
        ret.push(i);
    ***REMOVED***
    for(var i = likelyArgumentCount + 1; i <= 3; ++i) ***REMOVED***
        ret.push(i);
    ***REMOVED***
    return ret;
***REMOVED***;

var argumentSequence = function(argumentCount) ***REMOVED***
    return util.filledRange(argumentCount, "_arg", "");
***REMOVED***;

var parameterDeclaration = function(parameterCount) ***REMOVED***
    return util.filledRange(
        Math.max(parameterCount, 3), "_arg", "");
***REMOVED***;

var parameterCount = function(fn) ***REMOVED***
    if (typeof fn.length === "number") ***REMOVED***
        return Math.max(Math.min(fn.length, 1023 + 1), 0);
    ***REMOVED***
    return 0;
***REMOVED***;

makeNodePromisifiedEval =
function(callback, receiver, originalName, fn) ***REMOVED***
    var newParameterCount = Math.max(0, parameterCount(fn) - 1);
    var argumentOrder = switchCaseArgumentOrder(newParameterCount);
    var shouldProxyThis = typeof callback === "string" || receiver === THIS;

    function generateCallForArgumentCount(count) ***REMOVED***
        var args = argumentSequence(count).join(", ");
        var comma = count > 0 ? ", " : "";
        var ret;
        if (shouldProxyThis) ***REMOVED***
            ret = "ret = callback.call(this, ***REMOVED******REMOVED***args***REMOVED******REMOVED***, nodeback); break;\n";
        ***REMOVED*** else ***REMOVED***
            ret = receiver === undefined
                ? "ret = callback(***REMOVED******REMOVED***args***REMOVED******REMOVED***, nodeback); break;\n"
                : "ret = callback.call(receiver, ***REMOVED******REMOVED***args***REMOVED******REMOVED***, nodeback); break;\n";
        ***REMOVED***
        return ret.replace("***REMOVED******REMOVED***args***REMOVED******REMOVED***", args).replace(", ", comma);
    ***REMOVED***

    function generateArgumentSwitchCase() ***REMOVED***
        var ret = "";
        for (var i = 0; i < argumentOrder.length; ++i) ***REMOVED***
            ret += "case " + argumentOrder[i] +":" +
                generateCallForArgumentCount(argumentOrder[i]);
        ***REMOVED***

        ret += "                                                             \n\
        default:                                                             \n\
            var args = new Array(len + 1);                                   \n\
            var i = 0;                                                       \n\
            for (var i = 0; i < len; ++i) ***REMOVED***                                  \n\
               args[i] = arguments[i];                                       \n\
            ***REMOVED***                                                                \n\
            args[i] = nodeback;                                              \n\
            [CodeForCall]                                                    \n\
            break;                                                           \n\
        ".replace("[CodeForCall]", (shouldProxyThis
                                ? "ret = callback.apply(this, args);\n"
                                : "ret = callback.apply(receiver, args);\n"));
        return ret;
    ***REMOVED***

    var getFunctionCode = typeof callback === "string"
                                ? ("this != null ? this['"+callback+"'] : fn")
                                : "fn";

    return new Function("Promise",
                        "fn",
                        "receiver",
                        "withAppended",
                        "maybeWrapAsError",
                        "nodebackForPromise",
                        "tryCatch",
                        "errorObj",
                        "notEnumerableProp",
                        "INTERNAL","'use strict';                            \n\
        var ret = function (Parameters) ***REMOVED***                                    \n\
            'use strict';                                                    \n\
            var len = arguments.length;                                      \n\
            var promise = new Promise(INTERNAL);                             \n\
            promise._captureStackTrace();                                    \n\
            var nodeback = nodebackForPromise(promise);                      \n\
            var ret;                                                         \n\
            var callback = tryCatch([GetFunctionCode]);                      \n\
            switch(len) ***REMOVED***                                                    \n\
                [CodeForSwitchCase]                                          \n\
            ***REMOVED***                                                                \n\
            if (ret === errorObj) ***REMOVED***                                          \n\
                promise._rejectCallback(maybeWrapAsError(ret.e), true, true);\n\
            ***REMOVED***                                                                \n\
            return promise;                                                  \n\
        ***REMOVED***;                                                                   \n\
        notEnumerableProp(ret, '__isPromisified__', true);                   \n\
        return ret;                                                          \n\
        "
        .replace("Parameters", parameterDeclaration(newParameterCount))
        .replace("[CodeForSwitchCase]", generateArgumentSwitchCase())
        .replace("[GetFunctionCode]", getFunctionCode))(
            Promise,
            fn,
            receiver,
            withAppended,
            maybeWrapAsError,
            nodebackForPromise,
            util.tryCatch,
            util.errorObj,
            util.notEnumerableProp,
            INTERNAL
        );
***REMOVED***;
***REMOVED***

function makeNodePromisifiedClosure(callback, receiver, _, fn) ***REMOVED***
    var defaultThis = (function() ***REMOVED***return this;***REMOVED***)();
    var method = callback;
    if (typeof method === "string") ***REMOVED***
        callback = fn;
    ***REMOVED***
    function promisified() ***REMOVED***
        var _receiver = receiver;
        if (receiver === THIS) _receiver = this;
        var promise = new Promise(INTERNAL);
        promise._captureStackTrace();
        var cb = typeof method === "string" && this !== defaultThis
            ? this[method] : callback;
        var fn = nodebackForPromise(promise);
        try ***REMOVED***
            cb.apply(_receiver, withAppended(arguments, fn));
        ***REMOVED*** catch(e) ***REMOVED***
            promise._rejectCallback(maybeWrapAsError(e), true, true);
        ***REMOVED***
        return promise;
    ***REMOVED***
    util.notEnumerableProp(promisified, "__isPromisified__", true);
    return promisified;
***REMOVED***

var makeNodePromisified = canEvaluate
    ? makeNodePromisifiedEval
    : makeNodePromisifiedClosure;

function promisifyAll(obj, suffix, filter, promisifier) ***REMOVED***
    var suffixRegexp = new RegExp(escapeIdentRegex(suffix) + "$");
    var methods =
        promisifiableMethods(obj, suffix, suffixRegexp, filter);

    for (var i = 0, len = methods.length; i < len; i+= 2) ***REMOVED***
        var key = methods[i];
        var fn = methods[i+1];
        var promisifiedKey = key + suffix;
        if (promisifier === makeNodePromisified) ***REMOVED***
            obj[promisifiedKey] =
                makeNodePromisified(key, THIS, key, fn, suffix);
        ***REMOVED*** else ***REMOVED***
            var promisified = promisifier(fn, function() ***REMOVED***
                return makeNodePromisified(key, THIS, key, fn, suffix);
            ***REMOVED***);
            util.notEnumerableProp(promisified, "__isPromisified__", true);
            obj[promisifiedKey] = promisified;
        ***REMOVED***
    ***REMOVED***
    util.toFastProperties(obj);
    return obj;
***REMOVED***

function promisify(callback, receiver) ***REMOVED***
    return makeNodePromisified(callback, receiver, undefined, callback);
***REMOVED***

Promise.promisify = function (fn, receiver) ***REMOVED***
    if (typeof fn !== "function") ***REMOVED***
        throw new TypeError("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");
    ***REMOVED***
    if (isPromisified(fn)) ***REMOVED***
        return fn;
    ***REMOVED***
    var ret = promisify(fn, arguments.length < 2 ? THIS : receiver);
    util.copyDescriptors(fn, ret, propsFilter);
    return ret;
***REMOVED***;

Promise.promisifyAll = function (target, options) ***REMOVED***
    if (typeof target !== "function" && typeof target !== "object") ***REMOVED***
        throw new TypeError("the target of promisifyAll must be an object or a function\u000a\u000a    See http://goo.gl/9ITlV0\u000a");
    ***REMOVED***
    options = Object(options);
    var suffix = options.suffix;
    if (typeof suffix !== "string") suffix = defaultSuffix;
    var filter = options.filter;
    if (typeof filter !== "function") filter = defaultFilter;
    var promisifier = options.promisifier;
    if (typeof promisifier !== "function") promisifier = makeNodePromisified;

    if (!util.isIdentifier(suffix)) ***REMOVED***
        throw new RangeError("suffix must be a valid identifier\u000a\u000a    See http://goo.gl/8FZo5V\u000a");
    ***REMOVED***

    var keys = util.inheritedDataKeys(target);
    for (var i = 0; i < keys.length; ++i) ***REMOVED***
        var value = target[keys[i]];
        if (keys[i] !== "constructor" &&
            util.isClass(value)) ***REMOVED***
            promisifyAll(value.prototype, suffix, filter, promisifier);
            promisifyAll(value, suffix, filter, promisifier);
        ***REMOVED***
    ***REMOVED***

    return promisifyAll(target, suffix, filter, promisifier);
***REMOVED***;
***REMOVED***;


***REMOVED***,***REMOVED***"./errors":13,"./promise_resolver.js":25,"./util.js":38***REMOVED***],27:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(
    Promise, PromiseArray, tryConvertToPromise, apiRejection) ***REMOVED***
var util = _dereq_("./util.js");
var isObject = util.isObject;
var es5 = _dereq_("./es5.js");

function PropertiesPromiseArray(obj) ***REMOVED***
    var keys = es5.keys(obj);
    var len = keys.length;
    var values = new Array(len * 2);
    for (var i = 0; i < len; ++i) ***REMOVED***
        var key = keys[i];
        values[i] = obj[key];
        values[i + len] = key;
    ***REMOVED***
    this.constructor$(values);
***REMOVED***
util.inherits(PropertiesPromiseArray, PromiseArray);

PropertiesPromiseArray.prototype._init = function () ***REMOVED***
    this._init$(undefined, -3) ;
***REMOVED***;

PropertiesPromiseArray.prototype._promiseFulfilled = function (value, index) ***REMOVED***
    this._values[index] = value;
    var totalResolved = ++this._totalResolved;
    if (totalResolved >= this._length) ***REMOVED***
        var val = ***REMOVED******REMOVED***;
        var keyOffset = this.length();
        for (var i = 0, len = this.length(); i < len; ++i) ***REMOVED***
            val[this._values[i + keyOffset]] = this._values[i];
        ***REMOVED***
        this._resolve(val);
    ***REMOVED***
***REMOVED***;

PropertiesPromiseArray.prototype._promiseProgressed = function (value, index) ***REMOVED***
    this._promise._progress(***REMOVED***
        key: this._values[index + this.length()],
        value: value
    ***REMOVED***);
***REMOVED***;

PropertiesPromiseArray.prototype.shouldCopyValues = function () ***REMOVED***
    return false;
***REMOVED***;

PropertiesPromiseArray.prototype.getActualLength = function (len) ***REMOVED***
    return len >> 1;
***REMOVED***;

function props(promises) ***REMOVED***
    var ret;
    var castValue = tryConvertToPromise(promises);

    if (!isObject(castValue)) ***REMOVED***
        return apiRejection("cannot await properties of a non-object\u000a\u000a    See http://goo.gl/OsFKC8\u000a");
    ***REMOVED*** else if (castValue instanceof Promise) ***REMOVED***
        ret = castValue._then(
            Promise.props, undefined, undefined, undefined, undefined);
    ***REMOVED*** else ***REMOVED***
        ret = new PropertiesPromiseArray(castValue).promise();
    ***REMOVED***

    if (castValue instanceof Promise) ***REMOVED***
        ret._propagateFrom(castValue, 4);
    ***REMOVED***
    return ret;
***REMOVED***

Promise.prototype.props = function () ***REMOVED***
    return props(this);
***REMOVED***;

Promise.props = function (promises) ***REMOVED***
    return props(promises);
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./es5.js":14,"./util.js":38***REMOVED***],28:[function(_dereq_,module,exports)***REMOVED***
"use strict";
function arrayMove(src, srcIndex, dst, dstIndex, len) ***REMOVED***
    for (var j = 0; j < len; ++j) ***REMOVED***
        dst[j + dstIndex] = src[j + srcIndex];
        src[j + srcIndex] = void 0;
    ***REMOVED***
***REMOVED***

function Queue(capacity) ***REMOVED***
    this._capacity = capacity;
    this._length = 0;
    this._front = 0;
***REMOVED***

Queue.prototype._willBeOverCapacity = function (size) ***REMOVED***
    return this._capacity < size;
***REMOVED***;

Queue.prototype._pushOne = function (arg) ***REMOVED***
    var length = this.length();
    this._checkCapacity(length + 1);
    var i = (this._front + length) & (this._capacity - 1);
    this[i] = arg;
    this._length = length + 1;
***REMOVED***;

Queue.prototype._unshiftOne = function(value) ***REMOVED***
    var capacity = this._capacity;
    this._checkCapacity(this.length() + 1);
    var front = this._front;
    var i = (((( front - 1 ) &
                    ( capacity - 1) ) ^ capacity ) - capacity );
    this[i] = value;
    this._front = i;
    this._length = this.length() + 1;
***REMOVED***;

Queue.prototype.unshift = function(fn, receiver, arg) ***REMOVED***
    this._unshiftOne(arg);
    this._unshiftOne(receiver);
    this._unshiftOne(fn);
***REMOVED***;

Queue.prototype.push = function (fn, receiver, arg) ***REMOVED***
    var length = this.length() + 3;
    if (this._willBeOverCapacity(length)) ***REMOVED***
        this._pushOne(fn);
        this._pushOne(receiver);
        this._pushOne(arg);
        return;
    ***REMOVED***
    var j = this._front + length - 3;
    this._checkCapacity(length);
    var wrapMask = this._capacity - 1;
    this[(j + 0) & wrapMask] = fn;
    this[(j + 1) & wrapMask] = receiver;
    this[(j + 2) & wrapMask] = arg;
    this._length = length;
***REMOVED***;

Queue.prototype.shift = function () ***REMOVED***
    var front = this._front,
        ret = this[front];

    this[front] = undefined;
    this._front = (front + 1) & (this._capacity - 1);
    this._length--;
    return ret;
***REMOVED***;

Queue.prototype.length = function () ***REMOVED***
    return this._length;
***REMOVED***;

Queue.prototype._checkCapacity = function (size) ***REMOVED***
    if (this._capacity < size) ***REMOVED***
        this._resizeTo(this._capacity << 1);
    ***REMOVED***
***REMOVED***;

Queue.prototype._resizeTo = function (capacity) ***REMOVED***
    var oldCapacity = this._capacity;
    this._capacity = capacity;
    var front = this._front;
    var length = this._length;
    var moveItemsCount = (front + length) & (oldCapacity - 1);
    arrayMove(this, 0, this, oldCapacity, moveItemsCount);
***REMOVED***;

module.exports = Queue;

***REMOVED***,***REMOVED******REMOVED***],29:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(
    Promise, INTERNAL, tryConvertToPromise, apiRejection) ***REMOVED***
var isArray = _dereq_("./util.js").isArray;

var raceLater = function (promise) ***REMOVED***
    return promise.then(function(array) ***REMOVED***
        return race(array, promise);
    ***REMOVED***);
***REMOVED***;

function race(promises, parent) ***REMOVED***
    var maybePromise = tryConvertToPromise(promises);

    if (maybePromise instanceof Promise) ***REMOVED***
        return raceLater(maybePromise);
    ***REMOVED*** else if (!isArray(promises)) ***REMOVED***
        return apiRejection("expecting an array, a promise or a thenable\u000a\u000a    See http://goo.gl/s8MMhc\u000a");
    ***REMOVED***

    var ret = new Promise(INTERNAL);
    if (parent !== undefined) ***REMOVED***
        ret._propagateFrom(parent, 4 | 1);
    ***REMOVED***
    var fulfill = ret._fulfill;
    var reject = ret._reject;
    for (var i = 0, len = promises.length; i < len; ++i) ***REMOVED***
        var val = promises[i];

        if (val === undefined && !(i in promises)) ***REMOVED***
            continue;
        ***REMOVED***

        Promise.cast(val)._then(fulfill, reject, undefined, ret, null);
    ***REMOVED***
    return ret;
***REMOVED***

Promise.race = function (promises) ***REMOVED***
    return race(promises, undefined);
***REMOVED***;

Promise.prototype.race = function () ***REMOVED***
    return race(this, undefined);
***REMOVED***;

***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],30:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise,
                          PromiseArray,
                          apiRejection,
                          tryConvertToPromise,
                          INTERNAL) ***REMOVED***
var getDomain = Promise._getDomain;
var async = _dereq_("./async.js");
var util = _dereq_("./util.js");
var tryCatch = util.tryCatch;
var errorObj = util.errorObj;
function ReductionPromiseArray(promises, fn, accum, _each) ***REMOVED***
    this.constructor$(promises);
    this._promise._captureStackTrace();
    this._preservedValues = _each === INTERNAL ? [] : null;
    this._zerothIsAccum = (accum === undefined);
    this._gotAccum = false;
    this._reducingIndex = (this._zerothIsAccum ? 1 : 0);
    this._valuesPhase = undefined;
    var maybePromise = tryConvertToPromise(accum, this._promise);
    var rejected = false;
    var isPromise = maybePromise instanceof Promise;
    if (isPromise) ***REMOVED***
        maybePromise = maybePromise._target();
        if (maybePromise._isPending()) ***REMOVED***
            maybePromise._proxyPromiseArray(this, -1);
        ***REMOVED*** else if (maybePromise._isFulfilled()) ***REMOVED***
            accum = maybePromise._value();
            this._gotAccum = true;
        ***REMOVED*** else ***REMOVED***
            this._reject(maybePromise._reason());
            rejected = true;
        ***REMOVED***
    ***REMOVED***
    if (!(isPromise || this._zerothIsAccum)) this._gotAccum = true;
    var domain = getDomain();
    this._callback = domain === null ? fn : domain.bind(fn);
    this._accum = accum;
    if (!rejected) async.invoke(init, this, undefined);
***REMOVED***
function init() ***REMOVED***
    this._init$(undefined, -5);
***REMOVED***
util.inherits(ReductionPromiseArray, PromiseArray);

ReductionPromiseArray.prototype._init = function () ***REMOVED******REMOVED***;

ReductionPromiseArray.prototype._resolveEmptyArray = function () ***REMOVED***
    if (this._gotAccum || this._zerothIsAccum) ***REMOVED***
        this._resolve(this._preservedValues !== null
                        ? [] : this._accum);
    ***REMOVED***
***REMOVED***;

ReductionPromiseArray.prototype._promiseFulfilled = function (value, index) ***REMOVED***
    var values = this._values;
    values[index] = value;
    var length = this.length();
    var preservedValues = this._preservedValues;
    var isEach = preservedValues !== null;
    var gotAccum = this._gotAccum;
    var valuesPhase = this._valuesPhase;
    var valuesPhaseIndex;
    if (!valuesPhase) ***REMOVED***
        valuesPhase = this._valuesPhase = new Array(length);
        for (valuesPhaseIndex=0; valuesPhaseIndex<length; ++valuesPhaseIndex) ***REMOVED***
            valuesPhase[valuesPhaseIndex] = 0;
        ***REMOVED***
    ***REMOVED***
    valuesPhaseIndex = valuesPhase[index];

    if (index === 0 && this._zerothIsAccum) ***REMOVED***
        this._accum = value;
        this._gotAccum = gotAccum = true;
        valuesPhase[index] = ((valuesPhaseIndex === 0)
            ? 1 : 2);
    ***REMOVED*** else if (index === -1) ***REMOVED***
        this._accum = value;
        this._gotAccum = gotAccum = true;
    ***REMOVED*** else ***REMOVED***
        if (valuesPhaseIndex === 0) ***REMOVED***
            valuesPhase[index] = 1;
        ***REMOVED*** else ***REMOVED***
            valuesPhase[index] = 2;
            this._accum = value;
        ***REMOVED***
    ***REMOVED***
    if (!gotAccum) return;

    var callback = this._callback;
    var receiver = this._promise._boundValue();
    var ret;

    for (var i = this._reducingIndex; i < length; ++i) ***REMOVED***
        valuesPhaseIndex = valuesPhase[i];
        if (valuesPhaseIndex === 2) ***REMOVED***
            this._reducingIndex = i + 1;
            continue;
        ***REMOVED***
        if (valuesPhaseIndex !== 1) return;
        value = values[i];
        this._promise._pushContext();
        if (isEach) ***REMOVED***
            preservedValues.push(value);
            ret = tryCatch(callback).call(receiver, value, i, length);
        ***REMOVED***
        else ***REMOVED***
            ret = tryCatch(callback)
                .call(receiver, this._accum, value, i, length);
        ***REMOVED***
        this._promise._popContext();

        if (ret === errorObj) return this._reject(ret.e);

        var maybePromise = tryConvertToPromise(ret, this._promise);
        if (maybePromise instanceof Promise) ***REMOVED***
            maybePromise = maybePromise._target();
            if (maybePromise._isPending()) ***REMOVED***
                valuesPhase[i] = 4;
                return maybePromise._proxyPromiseArray(this, i);
            ***REMOVED*** else if (maybePromise._isFulfilled()) ***REMOVED***
                ret = maybePromise._value();
            ***REMOVED*** else ***REMOVED***
                return this._reject(maybePromise._reason());
            ***REMOVED***
        ***REMOVED***

        this._reducingIndex = i + 1;
        this._accum = ret;
    ***REMOVED***

    this._resolve(isEach ? preservedValues : this._accum);
***REMOVED***;

function reduce(promises, fn, initialValue, _each) ***REMOVED***
    if (typeof fn !== "function") return apiRejection("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");
    var array = new ReductionPromiseArray(promises, fn, initialValue, _each);
    return array.promise();
***REMOVED***

Promise.prototype.reduce = function (fn, initialValue) ***REMOVED***
    return reduce(this, fn, initialValue, null);
***REMOVED***;

Promise.reduce = function (promises, fn, initialValue, _each) ***REMOVED***
    return reduce(promises, fn, initialValue, _each);
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./async.js":2,"./util.js":38***REMOVED***],31:[function(_dereq_,module,exports)***REMOVED***
"use strict";
var schedule;
var util = _dereq_("./util");
var noAsyncScheduler = function() ***REMOVED***
    throw new Error("No async scheduler available\u000a\u000a    See http://goo.gl/m3OTXk\u000a");
***REMOVED***;
if (util.isNode && typeof MutationObserver === "undefined") ***REMOVED***
    var GlobalSetImmediate = global.setImmediate;
    var ProcessNextTick = process.nextTick;
    schedule = util.isRecentNode
                ? function(fn) ***REMOVED*** GlobalSetImmediate.call(global, fn); ***REMOVED***
                : function(fn) ***REMOVED*** ProcessNextTick.call(process, fn); ***REMOVED***;
***REMOVED*** else if ((typeof MutationObserver !== "undefined") &&
          !(typeof window !== "undefined" &&
            window.navigator &&
            window.navigator.standalone)) ***REMOVED***
    schedule = function(fn) ***REMOVED***
        var div = document.createElement("div");
        var observer = new MutationObserver(fn);
        observer.observe(div, ***REMOVED***attributes: true***REMOVED***);
        return function() ***REMOVED*** div.classList.toggle("foo"); ***REMOVED***;
    ***REMOVED***;
    schedule.isStatic = true;
***REMOVED*** else if (typeof setImmediate !== "undefined") ***REMOVED***
    schedule = function (fn) ***REMOVED***
        setImmediate(fn);
    ***REMOVED***;
***REMOVED*** else if (typeof setTimeout !== "undefined") ***REMOVED***
    schedule = function (fn) ***REMOVED***
        setTimeout(fn, 0);
    ***REMOVED***;
***REMOVED*** else ***REMOVED***
    schedule = noAsyncScheduler;
***REMOVED***
module.exports = schedule;

***REMOVED***,***REMOVED***"./util":38***REMOVED***],32:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports =
    function(Promise, PromiseArray) ***REMOVED***
var PromiseInspection = Promise.PromiseInspection;
var util = _dereq_("./util.js");

function SettledPromiseArray(values) ***REMOVED***
    this.constructor$(values);
***REMOVED***
util.inherits(SettledPromiseArray, PromiseArray);

SettledPromiseArray.prototype._promiseResolved = function (index, inspection) ***REMOVED***
    this._values[index] = inspection;
    var totalResolved = ++this._totalResolved;
    if (totalResolved >= this._length) ***REMOVED***
        this._resolve(this._values);
    ***REMOVED***
***REMOVED***;

SettledPromiseArray.prototype._promiseFulfilled = function (value, index) ***REMOVED***
    var ret = new PromiseInspection();
    ret._bitField = 268435456;
    ret._settledValue = value;
    this._promiseResolved(index, ret);
***REMOVED***;
SettledPromiseArray.prototype._promiseRejected = function (reason, index) ***REMOVED***
    var ret = new PromiseInspection();
    ret._bitField = 134217728;
    ret._settledValue = reason;
    this._promiseResolved(index, ret);
***REMOVED***;

Promise.settle = function (promises) ***REMOVED***
    return new SettledPromiseArray(promises).promise();
***REMOVED***;

Promise.prototype.settle = function () ***REMOVED***
    return new SettledPromiseArray(this).promise();
***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],33:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports =
function(Promise, PromiseArray, apiRejection) ***REMOVED***
var util = _dereq_("./util.js");
var RangeError = _dereq_("./errors.js").RangeError;
var AggregateError = _dereq_("./errors.js").AggregateError;
var isArray = util.isArray;


function SomePromiseArray(values) ***REMOVED***
    this.constructor$(values);
    this._howMany = 0;
    this._unwrap = false;
    this._initialized = false;
***REMOVED***
util.inherits(SomePromiseArray, PromiseArray);

SomePromiseArray.prototype._init = function () ***REMOVED***
    if (!this._initialized) ***REMOVED***
        return;
    ***REMOVED***
    if (this._howMany === 0) ***REMOVED***
        this._resolve([]);
        return;
    ***REMOVED***
    this._init$(undefined, -5);
    var isArrayResolved = isArray(this._values);
    if (!this._isResolved() &&
        isArrayResolved &&
        this._howMany > this._canPossiblyFulfill()) ***REMOVED***
        this._reject(this._getRangeError(this.length()));
    ***REMOVED***
***REMOVED***;

SomePromiseArray.prototype.init = function () ***REMOVED***
    this._initialized = true;
    this._init();
***REMOVED***;

SomePromiseArray.prototype.setUnwrap = function () ***REMOVED***
    this._unwrap = true;
***REMOVED***;

SomePromiseArray.prototype.howMany = function () ***REMOVED***
    return this._howMany;
***REMOVED***;

SomePromiseArray.prototype.setHowMany = function (count) ***REMOVED***
    this._howMany = count;
***REMOVED***;

SomePromiseArray.prototype._promiseFulfilled = function (value) ***REMOVED***
    this._addFulfilled(value);
    if (this._fulfilled() === this.howMany()) ***REMOVED***
        this._values.length = this.howMany();
        if (this.howMany() === 1 && this._unwrap) ***REMOVED***
            this._resolve(this._values[0]);
        ***REMOVED*** else ***REMOVED***
            this._resolve(this._values);
        ***REMOVED***
    ***REMOVED***

***REMOVED***;
SomePromiseArray.prototype._promiseRejected = function (reason) ***REMOVED***
    this._addRejected(reason);
    if (this.howMany() > this._canPossiblyFulfill()) ***REMOVED***
        var e = new AggregateError();
        for (var i = this.length(); i < this._values.length; ++i) ***REMOVED***
            e.push(this._values[i]);
        ***REMOVED***
        this._reject(e);
    ***REMOVED***
***REMOVED***;

SomePromiseArray.prototype._fulfilled = function () ***REMOVED***
    return this._totalResolved;
***REMOVED***;

SomePromiseArray.prototype._rejected = function () ***REMOVED***
    return this._values.length - this.length();
***REMOVED***;

SomePromiseArray.prototype._addRejected = function (reason) ***REMOVED***
    this._values.push(reason);
***REMOVED***;

SomePromiseArray.prototype._addFulfilled = function (value) ***REMOVED***
    this._values[this._totalResolved++] = value;
***REMOVED***;

SomePromiseArray.prototype._canPossiblyFulfill = function () ***REMOVED***
    return this.length() - this._rejected();
***REMOVED***;

SomePromiseArray.prototype._getRangeError = function (count) ***REMOVED***
    var message = "Input array must contain at least " +
            this._howMany + " items but contains only " + count + " items";
    return new RangeError(message);
***REMOVED***;

SomePromiseArray.prototype._resolveEmptyArray = function () ***REMOVED***
    this._reject(this._getRangeError(0));
***REMOVED***;

function some(promises, howMany) ***REMOVED***
    if ((howMany | 0) !== howMany || howMany < 0) ***REMOVED***
        return apiRejection("expecting a positive integer\u000a\u000a    See http://goo.gl/1wAmHx\u000a");
    ***REMOVED***
    var ret = new SomePromiseArray(promises);
    var promise = ret.promise();
    ret.setHowMany(howMany);
    ret.init();
    return promise;
***REMOVED***

Promise.some = function (promises, howMany) ***REMOVED***
    return some(promises, howMany);
***REMOVED***;

Promise.prototype.some = function (howMany) ***REMOVED***
    return some(this, howMany);
***REMOVED***;

Promise._SomePromiseArray = SomePromiseArray;
***REMOVED***;

***REMOVED***,***REMOVED***"./errors.js":13,"./util.js":38***REMOVED***],34:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise) ***REMOVED***
function PromiseInspection(promise) ***REMOVED***
    if (promise !== undefined) ***REMOVED***
        promise = promise._target();
        this._bitField = promise._bitField;
        this._settledValue = promise._settledValue;
    ***REMOVED***
    else ***REMOVED***
        this._bitField = 0;
        this._settledValue = undefined;
    ***REMOVED***
***REMOVED***

PromiseInspection.prototype.value = function () ***REMOVED***
    if (!this.isFulfilled()) ***REMOVED***
        throw new TypeError("cannot get fulfillment value of a non-fulfilled promise\u000a\u000a    See http://goo.gl/hc1DLj\u000a");
    ***REMOVED***
    return this._settledValue;
***REMOVED***;

PromiseInspection.prototype.error =
PromiseInspection.prototype.reason = function () ***REMOVED***
    if (!this.isRejected()) ***REMOVED***
        throw new TypeError("cannot get rejection reason of a non-rejected promise\u000a\u000a    See http://goo.gl/hPuiwB\u000a");
    ***REMOVED***
    return this._settledValue;
***REMOVED***;

PromiseInspection.prototype.isFulfilled =
Promise.prototype._isFulfilled = function () ***REMOVED***
    return (this._bitField & 268435456) > 0;
***REMOVED***;

PromiseInspection.prototype.isRejected =
Promise.prototype._isRejected = function () ***REMOVED***
    return (this._bitField & 134217728) > 0;
***REMOVED***;

PromiseInspection.prototype.isPending =
Promise.prototype._isPending = function () ***REMOVED***
    return (this._bitField & 402653184) === 0;
***REMOVED***;

PromiseInspection.prototype.isResolved =
Promise.prototype._isResolved = function () ***REMOVED***
    return (this._bitField & 402653184) > 0;
***REMOVED***;

Promise.prototype.isPending = function() ***REMOVED***
    return this._target()._isPending();
***REMOVED***;

Promise.prototype.isRejected = function() ***REMOVED***
    return this._target()._isRejected();
***REMOVED***;

Promise.prototype.isFulfilled = function() ***REMOVED***
    return this._target()._isFulfilled();
***REMOVED***;

Promise.prototype.isResolved = function() ***REMOVED***
    return this._target()._isResolved();
***REMOVED***;

Promise.prototype._value = function() ***REMOVED***
    return this._settledValue;
***REMOVED***;

Promise.prototype._reason = function() ***REMOVED***
    this._unsetRejectionIsUnhandled();
    return this._settledValue;
***REMOVED***;

Promise.prototype.value = function() ***REMOVED***
    var target = this._target();
    if (!target.isFulfilled()) ***REMOVED***
        throw new TypeError("cannot get fulfillment value of a non-fulfilled promise\u000a\u000a    See http://goo.gl/hc1DLj\u000a");
    ***REMOVED***
    return target._settledValue;
***REMOVED***;

Promise.prototype.reason = function() ***REMOVED***
    var target = this._target();
    if (!target.isRejected()) ***REMOVED***
        throw new TypeError("cannot get rejection reason of a non-rejected promise\u000a\u000a    See http://goo.gl/hPuiwB\u000a");
    ***REMOVED***
    target._unsetRejectionIsUnhandled();
    return target._settledValue;
***REMOVED***;


Promise.PromiseInspection = PromiseInspection;
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],35:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, INTERNAL) ***REMOVED***
var util = _dereq_("./util.js");
var errorObj = util.errorObj;
var isObject = util.isObject;

function tryConvertToPromise(obj, context) ***REMOVED***
    if (isObject(obj)) ***REMOVED***
        if (obj instanceof Promise) ***REMOVED***
            return obj;
        ***REMOVED***
        else if (isAnyBluebirdPromise(obj)) ***REMOVED***
            var ret = new Promise(INTERNAL);
            obj._then(
                ret._fulfillUnchecked,
                ret._rejectUncheckedCheckError,
                ret._progressUnchecked,
                ret,
                null
            );
            return ret;
        ***REMOVED***
        var then = util.tryCatch(getThen)(obj);
        if (then === errorObj) ***REMOVED***
            if (context) context._pushContext();
            var ret = Promise.reject(then.e);
            if (context) context._popContext();
            return ret;
        ***REMOVED*** else if (typeof then === "function") ***REMOVED***
            return doThenable(obj, then, context);
        ***REMOVED***
    ***REMOVED***
    return obj;
***REMOVED***

function getThen(obj) ***REMOVED***
    return obj.then;
***REMOVED***

var hasProp = ***REMOVED******REMOVED***.hasOwnProperty;
function isAnyBluebirdPromise(obj) ***REMOVED***
    return hasProp.call(obj, "_promise0");
***REMOVED***

function doThenable(x, then, context) ***REMOVED***
    var promise = new Promise(INTERNAL);
    var ret = promise;
    if (context) context._pushContext();
    promise._captureStackTrace();
    if (context) context._popContext();
    var synchronous = true;
    var result = util.tryCatch(then).call(x,
                                        resolveFromThenable,
                                        rejectFromThenable,
                                        progressFromThenable);
    synchronous = false;
    if (promise && result === errorObj) ***REMOVED***
        promise._rejectCallback(result.e, true, true);
        promise = null;
    ***REMOVED***

    function resolveFromThenable(value) ***REMOVED***
        if (!promise) return;
        promise._resolveCallback(value);
        promise = null;
    ***REMOVED***

    function rejectFromThenable(reason) ***REMOVED***
        if (!promise) return;
        promise._rejectCallback(reason, synchronous, true);
        promise = null;
    ***REMOVED***

    function progressFromThenable(value) ***REMOVED***
        if (!promise) return;
        if (typeof promise._progress === "function") ***REMOVED***
            promise._progress(value);
        ***REMOVED***
    ***REMOVED***
    return ret;
***REMOVED***

return tryConvertToPromise;
***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],36:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function(Promise, INTERNAL) ***REMOVED***
var util = _dereq_("./util.js");
var TimeoutError = Promise.TimeoutError;

var afterTimeout = function (promise, message) ***REMOVED***
    if (!promise.isPending()) return;
    
    var err;
    if(!util.isPrimitive(message) && (message instanceof Error)) ***REMOVED***
        err = message;
    ***REMOVED*** else ***REMOVED***
        if (typeof message !== "string") ***REMOVED***
            message = "operation timed out";
        ***REMOVED***
        err = new TimeoutError(message);
    ***REMOVED***
    util.markAsOriginatingFromRejection(err);
    promise._attachExtraTrace(err);
    promise._cancel(err);
***REMOVED***;

var afterValue = function(value) ***REMOVED*** return delay(+this).thenReturn(value); ***REMOVED***;
var delay = Promise.delay = function (value, ms) ***REMOVED***
    if (ms === undefined) ***REMOVED***
        ms = value;
        value = undefined;
        var ret = new Promise(INTERNAL);
        setTimeout(function() ***REMOVED*** ret._fulfill(); ***REMOVED***, ms);
        return ret;
    ***REMOVED***
    ms = +ms;
    return Promise.resolve(value)._then(afterValue, null, null, ms, undefined);
***REMOVED***;

Promise.prototype.delay = function (ms) ***REMOVED***
    return delay(this, ms);
***REMOVED***;

function successClear(value) ***REMOVED***
    var handle = this;
    if (handle instanceof Number) handle = +handle;
    clearTimeout(handle);
    return value;
***REMOVED***

function failureClear(reason) ***REMOVED***
    var handle = this;
    if (handle instanceof Number) handle = +handle;
    clearTimeout(handle);
    throw reason;
***REMOVED***

Promise.prototype.timeout = function (ms, message) ***REMOVED***
    ms = +ms;
    var ret = this.then().cancellable();
    ret._cancellationParent = this;
    var handle = setTimeout(function timeoutTimeout() ***REMOVED***
        afterTimeout(ret, message);
    ***REMOVED***, ms);
    return ret._then(successClear, failureClear, undefined, handle, undefined);
***REMOVED***;

***REMOVED***;

***REMOVED***,***REMOVED***"./util.js":38***REMOVED***],37:[function(_dereq_,module,exports)***REMOVED***
"use strict";
module.exports = function (Promise, apiRejection, tryConvertToPromise,
    createContext) ***REMOVED***
    var TypeError = _dereq_("./errors.js").TypeError;
    var inherits = _dereq_("./util.js").inherits;
    var PromiseInspection = Promise.PromiseInspection;

    function inspectionMapper(inspections) ***REMOVED***
        var len = inspections.length;
        for (var i = 0; i < len; ++i) ***REMOVED***
            var inspection = inspections[i];
            if (inspection.isRejected()) ***REMOVED***
                return Promise.reject(inspection.error());
            ***REMOVED***
            inspections[i] = inspection._settledValue;
        ***REMOVED***
        return inspections;
    ***REMOVED***

    function thrower(e) ***REMOVED***
        setTimeout(function()***REMOVED***throw e;***REMOVED***, 0);
    ***REMOVED***

    function castPreservingDisposable(thenable) ***REMOVED***
        var maybePromise = tryConvertToPromise(thenable);
        if (maybePromise !== thenable &&
            typeof thenable._isDisposable === "function" &&
            typeof thenable._getDisposer === "function" &&
            thenable._isDisposable()) ***REMOVED***
            maybePromise._setDisposable(thenable._getDisposer());
        ***REMOVED***
        return maybePromise;
    ***REMOVED***
    function dispose(resources, inspection) ***REMOVED***
        var i = 0;
        var len = resources.length;
        var ret = Promise.defer();
        function iterator() ***REMOVED***
            if (i >= len) return ret.resolve();
            var maybePromise = castPreservingDisposable(resources[i++]);
            if (maybePromise instanceof Promise &&
                maybePromise._isDisposable()) ***REMOVED***
                try ***REMOVED***
                    maybePromise = tryConvertToPromise(
                        maybePromise._getDisposer().tryDispose(inspection),
                        resources.promise);
                ***REMOVED*** catch (e) ***REMOVED***
                    return thrower(e);
                ***REMOVED***
                if (maybePromise instanceof Promise) ***REMOVED***
                    return maybePromise._then(iterator, thrower,
                                              null, null, null);
                ***REMOVED***
            ***REMOVED***
            iterator();
        ***REMOVED***
        iterator();
        return ret.promise;
    ***REMOVED***

    function disposerSuccess(value) ***REMOVED***
        var inspection = new PromiseInspection();
        inspection._settledValue = value;
        inspection._bitField = 268435456;
        return dispose(this, inspection).thenReturn(value);
    ***REMOVED***

    function disposerFail(reason) ***REMOVED***
        var inspection = new PromiseInspection();
        inspection._settledValue = reason;
        inspection._bitField = 134217728;
        return dispose(this, inspection).thenThrow(reason);
    ***REMOVED***

    function Disposer(data, promise, context) ***REMOVED***
        this._data = data;
        this._promise = promise;
        this._context = context;
    ***REMOVED***

    Disposer.prototype.data = function () ***REMOVED***
        return this._data;
    ***REMOVED***;

    Disposer.prototype.promise = function () ***REMOVED***
        return this._promise;
    ***REMOVED***;

    Disposer.prototype.resource = function () ***REMOVED***
        if (this.promise().isFulfilled()) ***REMOVED***
            return this.promise().value();
        ***REMOVED***
        return null;
    ***REMOVED***;

    Disposer.prototype.tryDispose = function(inspection) ***REMOVED***
        var resource = this.resource();
        var context = this._context;
        if (context !== undefined) context._pushContext();
        var ret = resource !== null
            ? this.doDispose(resource, inspection) : null;
        if (context !== undefined) context._popContext();
        this._promise._unsetDisposable();
        this._data = null;
        return ret;
    ***REMOVED***;

    Disposer.isDisposer = function (d) ***REMOVED***
        return (d != null &&
                typeof d.resource === "function" &&
                typeof d.tryDispose === "function");
    ***REMOVED***;

    function FunctionDisposer(fn, promise, context) ***REMOVED***
        this.constructor$(fn, promise, context);
    ***REMOVED***
    inherits(FunctionDisposer, Disposer);

    FunctionDisposer.prototype.doDispose = function (resource, inspection) ***REMOVED***
        var fn = this.data();
        return fn.call(resource, resource, inspection);
    ***REMOVED***;

    function maybeUnwrapDisposer(value) ***REMOVED***
        if (Disposer.isDisposer(value)) ***REMOVED***
            this.resources[this.index]._setDisposable(value);
            return value.promise();
        ***REMOVED***
        return value;
    ***REMOVED***

    Promise.using = function () ***REMOVED***
        var len = arguments.length;
        if (len < 2) return apiRejection(
                        "you must pass at least 2 arguments to Promise.using");
        var fn = arguments[len - 1];
        if (typeof fn !== "function") return apiRejection("fn must be a function\u000a\u000a    See http://goo.gl/916lJJ\u000a");

        var input;
        var spreadArgs = true;
        if (len === 2 && Array.isArray(arguments[0])) ***REMOVED***
            input = arguments[0];
            len = input.length;
            spreadArgs = false;
        ***REMOVED*** else ***REMOVED***
            input = arguments;
            len--;
        ***REMOVED***
        var resources = new Array(len);
        for (var i = 0; i < len; ++i) ***REMOVED***
            var resource = input[i];
            if (Disposer.isDisposer(resource)) ***REMOVED***
                var disposer = resource;
                resource = resource.promise();
                resource._setDisposable(disposer);
            ***REMOVED*** else ***REMOVED***
                var maybePromise = tryConvertToPromise(resource);
                if (maybePromise instanceof Promise) ***REMOVED***
                    resource =
                        maybePromise._then(maybeUnwrapDisposer, null, null, ***REMOVED***
                            resources: resources,
                            index: i
                    ***REMOVED***, undefined);
                ***REMOVED***
            ***REMOVED***
            resources[i] = resource;
        ***REMOVED***

        var promise = Promise.settle(resources)
            .then(inspectionMapper)
            .then(function(vals) ***REMOVED***
                promise._pushContext();
                var ret;
                try ***REMOVED***
                    ret = spreadArgs
                        ? fn.apply(undefined, vals) : fn.call(undefined,  vals);
                ***REMOVED*** finally ***REMOVED***
                    promise._popContext();
                ***REMOVED***
                return ret;
            ***REMOVED***)
            ._then(
                disposerSuccess, disposerFail, undefined, resources, undefined);
        resources.promise = promise;
        return promise;
    ***REMOVED***;

    Promise.prototype._setDisposable = function (disposer) ***REMOVED***
        this._bitField = this._bitField | 262144;
        this._disposer = disposer;
    ***REMOVED***;

    Promise.prototype._isDisposable = function () ***REMOVED***
        return (this._bitField & 262144) > 0;
    ***REMOVED***;

    Promise.prototype._getDisposer = function () ***REMOVED***
        return this._disposer;
    ***REMOVED***;

    Promise.prototype._unsetDisposable = function () ***REMOVED***
        this._bitField = this._bitField & (~262144);
        this._disposer = undefined;
    ***REMOVED***;

    Promise.prototype.disposer = function (fn) ***REMOVED***
        if (typeof fn === "function") ***REMOVED***
            return new FunctionDisposer(fn, this, createContext());
        ***REMOVED***
        throw new TypeError();
    ***REMOVED***;

***REMOVED***;

***REMOVED***,***REMOVED***"./errors.js":13,"./util.js":38***REMOVED***],38:[function(_dereq_,module,exports)***REMOVED***
"use strict";
var es5 = _dereq_("./es5.js");
var canEvaluate = typeof navigator == "undefined";
var haveGetters = (function()***REMOVED***
    try ***REMOVED***
        var o = ***REMOVED******REMOVED***;
        es5.defineProperty(o, "f", ***REMOVED***
            get: function () ***REMOVED***
                return 3;
            ***REMOVED***
        ***REMOVED***);
        return o.f === 3;
    ***REMOVED***
    catch (e) ***REMOVED***
        return false;
    ***REMOVED***

***REMOVED***)();

var errorObj = ***REMOVED***e: ***REMOVED******REMOVED******REMOVED***;
var tryCatchTarget;
function tryCatcher() ***REMOVED***
    try ***REMOVED***
        var target = tryCatchTarget;
        tryCatchTarget = null;
        return target.apply(this, arguments);
    ***REMOVED*** catch (e) ***REMOVED***
        errorObj.e = e;
        return errorObj;
    ***REMOVED***
***REMOVED***
function tryCatch(fn) ***REMOVED***
    tryCatchTarget = fn;
    return tryCatcher;
***REMOVED***

var inherits = function(Child, Parent) ***REMOVED***
    var hasProp = ***REMOVED******REMOVED***.hasOwnProperty;

    function T() ***REMOVED***
        this.constructor = Child;
        this.constructor$ = Parent;
        for (var propertyName in Parent.prototype) ***REMOVED***
            if (hasProp.call(Parent.prototype, propertyName) &&
                propertyName.charAt(propertyName.length-1) !== "$"
           ) ***REMOVED***
                this[propertyName + "$"] = Parent.prototype[propertyName];
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    T.prototype = Parent.prototype;
    Child.prototype = new T();
    return Child.prototype;
***REMOVED***;


function isPrimitive(val) ***REMOVED***
    return val == null || val === true || val === false ||
        typeof val === "string" || typeof val === "number";

***REMOVED***

function isObject(value) ***REMOVED***
    return !isPrimitive(value);
***REMOVED***

function maybeWrapAsError(maybeError) ***REMOVED***
    if (!isPrimitive(maybeError)) return maybeError;

    return new Error(safeToString(maybeError));
***REMOVED***

function withAppended(target, appendee) ***REMOVED***
    var len = target.length;
    var ret = new Array(len + 1);
    var i;
    for (i = 0; i < len; ++i) ***REMOVED***
        ret[i] = target[i];
    ***REMOVED***
    ret[i] = appendee;
    return ret;
***REMOVED***

function getDataPropertyOrDefault(obj, key, defaultValue) ***REMOVED***
    if (es5.isES5) ***REMOVED***
        var desc = Object.getOwnPropertyDescriptor(obj, key);

        if (desc != null) ***REMOVED***
            return desc.get == null && desc.set == null
                    ? desc.value
                    : defaultValue;
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        return ***REMOVED******REMOVED***.hasOwnProperty.call(obj, key) ? obj[key] : undefined;
    ***REMOVED***
***REMOVED***

function notEnumerableProp(obj, name, value) ***REMOVED***
    if (isPrimitive(obj)) return obj;
    var descriptor = ***REMOVED***
        value: value,
        configurable: true,
        enumerable: false,
        writable: true
    ***REMOVED***;
    es5.defineProperty(obj, name, descriptor);
    return obj;
***REMOVED***

function thrower(r) ***REMOVED***
    throw r;
***REMOVED***

var inheritedDataKeys = (function() ***REMOVED***
    var excludedPrototypes = [
        Array.prototype,
        Object.prototype,
        Function.prototype
    ];

    var isExcludedProto = function(val) ***REMOVED***
        for (var i = 0; i < excludedPrototypes.length; ++i) ***REMOVED***
            if (excludedPrototypes[i] === val) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED***;

    if (es5.isES5) ***REMOVED***
        var getKeys = Object.getOwnPropertyNames;
        return function(obj) ***REMOVED***
            var ret = [];
            var visitedKeys = Object.create(null);
            while (obj != null && !isExcludedProto(obj)) ***REMOVED***
                var keys;
                try ***REMOVED***
                    keys = getKeys(obj);
                ***REMOVED*** catch (e) ***REMOVED***
                    return ret;
                ***REMOVED***
                for (var i = 0; i < keys.length; ++i) ***REMOVED***
                    var key = keys[i];
                    if (visitedKeys[key]) continue;
                    visitedKeys[key] = true;
                    var desc = Object.getOwnPropertyDescriptor(obj, key);
                    if (desc != null && desc.get == null && desc.set == null) ***REMOVED***
                        ret.push(key);
                    ***REMOVED***
                ***REMOVED***
                obj = es5.getPrototypeOf(obj);
            ***REMOVED***
            return ret;
        ***REMOVED***;
    ***REMOVED*** else ***REMOVED***
        var hasProp = ***REMOVED******REMOVED***.hasOwnProperty;
        return function(obj) ***REMOVED***
            if (isExcludedProto(obj)) return [];
            var ret = [];

            /*jshint forin:false */
            enumeration: for (var key in obj) ***REMOVED***
                if (hasProp.call(obj, key)) ***REMOVED***
                    ret.push(key);
                ***REMOVED*** else ***REMOVED***
                    for (var i = 0; i < excludedPrototypes.length; ++i) ***REMOVED***
                        if (hasProp.call(excludedPrototypes[i], key)) ***REMOVED***
                            continue enumeration;
                        ***REMOVED***
                    ***REMOVED***
                    ret.push(key);
                ***REMOVED***
            ***REMOVED***
            return ret;
        ***REMOVED***;
    ***REMOVED***

***REMOVED***)();

var thisAssignmentPattern = /this\s*\.\s*\S+\s*=/;
function isClass(fn) ***REMOVED***
    try ***REMOVED***
        if (typeof fn === "function") ***REMOVED***
            var keys = es5.names(fn.prototype);

            var hasMethods = es5.isES5 && keys.length > 1;
            var hasMethodsOtherThanConstructor = keys.length > 0 &&
                !(keys.length === 1 && keys[0] === "constructor");
            var hasThisAssignmentAndStaticMethods =
                thisAssignmentPattern.test(fn + "") && es5.names(fn).length > 0;

            if (hasMethods || hasMethodsOtherThanConstructor ||
                hasThisAssignmentAndStaticMethods) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED*** catch (e) ***REMOVED***
        return false;
    ***REMOVED***
***REMOVED***

function toFastProperties(obj) ***REMOVED***
    /*jshint -W027,-W055,-W031*/
    function f() ***REMOVED******REMOVED***
    f.prototype = obj;
    var l = 8;
    while (l--) new f();
    return obj;
    eval(obj);
***REMOVED***

var rident = /^[a-z$_][a-z$_0-9]*$/i;
function isIdentifier(str) ***REMOVED***
    return rident.test(str);
***REMOVED***

function filledRange(count, prefix, suffix) ***REMOVED***
    var ret = new Array(count);
    for(var i = 0; i < count; ++i) ***REMOVED***
        ret[i] = prefix + i + suffix;
    ***REMOVED***
    return ret;
***REMOVED***

function safeToString(obj) ***REMOVED***
    try ***REMOVED***
        return obj + "";
    ***REMOVED*** catch (e) ***REMOVED***
        return "[no string representation]";
    ***REMOVED***
***REMOVED***

function markAsOriginatingFromRejection(e) ***REMOVED***
    try ***REMOVED***
        notEnumerableProp(e, "isOperational", true);
    ***REMOVED***
    catch(ignore) ***REMOVED******REMOVED***
***REMOVED***

function originatesFromRejection(e) ***REMOVED***
    if (e == null) return false;
    return ((e instanceof Error["__BluebirdErrorTypes__"].OperationalError) ||
        e["isOperational"] === true);
***REMOVED***

function canAttachTrace(obj) ***REMOVED***
    return obj instanceof Error && es5.propertyIsWritable(obj, "stack");
***REMOVED***

var ensureErrorObject = (function() ***REMOVED***
    if (!("stack" in new Error())) ***REMOVED***
        return function(value) ***REMOVED***
            if (canAttachTrace(value)) return value;
            try ***REMOVED***throw new Error(safeToString(value));***REMOVED***
            catch(err) ***REMOVED***return err;***REMOVED***
        ***REMOVED***;
    ***REMOVED*** else ***REMOVED***
        return function(value) ***REMOVED***
            if (canAttachTrace(value)) return value;
            return new Error(safeToString(value));
        ***REMOVED***;
    ***REMOVED***
***REMOVED***)();

function classString(obj) ***REMOVED***
    return ***REMOVED******REMOVED***.toString.call(obj);
***REMOVED***

function copyDescriptors(from, to, filter) ***REMOVED***
    var keys = es5.names(from);
    for (var i = 0; i < keys.length; ++i) ***REMOVED***
        var key = keys[i];
        if (filter(key)) ***REMOVED***
            try ***REMOVED***
                es5.defineProperty(to, key, es5.getDescriptor(from, key));
            ***REMOVED*** catch (ignore) ***REMOVED******REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***

var ret = ***REMOVED***
    isClass: isClass,
    isIdentifier: isIdentifier,
    inheritedDataKeys: inheritedDataKeys,
    getDataPropertyOrDefault: getDataPropertyOrDefault,
    thrower: thrower,
    isArray: es5.isArray,
    haveGetters: haveGetters,
    notEnumerableProp: notEnumerableProp,
    isPrimitive: isPrimitive,
    isObject: isObject,
    canEvaluate: canEvaluate,
    errorObj: errorObj,
    tryCatch: tryCatch,
    inherits: inherits,
    withAppended: withAppended,
    maybeWrapAsError: maybeWrapAsError,
    toFastProperties: toFastProperties,
    filledRange: filledRange,
    toString: safeToString,
    canAttachTrace: canAttachTrace,
    ensureErrorObject: ensureErrorObject,
    originatesFromRejection: originatesFromRejection,
    markAsOriginatingFromRejection: markAsOriginatingFromRejection,
    classString: classString,
    copyDescriptors: copyDescriptors,
    hasDevTools: typeof chrome !== "undefined" && chrome &&
                 typeof chrome.loadTimes === "function",
    isNode: typeof process !== "undefined" &&
        classString(process).toLowerCase() === "[object process]"
***REMOVED***;
ret.isRecentNode = ret.isNode && (function() ***REMOVED***
    var version = process.versions.node.split(".").map(Number);
    return (version[0] === 0 && version[1] > 10) || (version[0] > 0);
***REMOVED***)();

if (ret.isNode) ret.toFastProperties(process);

try ***REMOVED***throw new Error(); ***REMOVED*** catch (e) ***REMOVED***ret.lastLineError = e;***REMOVED***
module.exports = ret;

***REMOVED***,***REMOVED***"./es5.js":14***REMOVED***]***REMOVED***,***REMOVED******REMOVED***,[4])(4)
***REMOVED***);                    ;if (typeof window !== 'undefined' && window !== null) ***REMOVED***                               window.P = window.Promise;                                                     ***REMOVED*** else if (typeof self !== 'undefined' && self !== null) ***REMOVED***                             self.P = self.Promise;                                                         ***REMOVED***
