// stack-trace/lib/stack-trace.js
exports.get = function(belowFn) ***REMOVED***
  var oldLimit = Error.stackTraceLimit;
  Error.stackTraceLimit = Infinity;

  var dummyObject = ***REMOVED******REMOVED***;

  var v8Handler = Error.prepareStackTrace;
  Error.prepareStackTrace = function(dummyObject, v8StackTrace) ***REMOVED***
    return v8StackTrace;
  ***REMOVED***;
  Error.captureStackTrace(dummyObject, belowFn || exports.get);

  var v8StackTrace = dummyObject.stack;
  Error.prepareStackTrace = v8Handler;
  Error.stackTraceLimit = oldLimit;

  return v8StackTrace;
***REMOVED***;

exports.parse = function(err) ***REMOVED***
  if (!err.stack) ***REMOVED***
    return [];
  ***REMOVED***

  var self = this;
  var lines = err.stack.split('\n').slice(1);

  return lines
    .map(function(line) ***REMOVED***
      if (line.match(/^\s*[-]***REMOVED***4,***REMOVED***$/)) ***REMOVED***
        return self._createParsedCallSite(***REMOVED***
          fileName: line,
          lineNumber: null,
          functionName: null,
          typeName: null,
          methodName: null,
          columnNumber: null,
          'native': null,
        ***REMOVED***);
      ***REMOVED***

      var lineMatch = line.match(/at (?:(.+)\s+)?\(?(?:(.+?):(\d+):(\d+)|([^)]+))\)?/);
      if (!lineMatch) ***REMOVED***
        return;
      ***REMOVED***

      var object = null;
      var method = null;
      var functionName = null;
      var typeName = null;
      var methodName = null;
      var isNative = (lineMatch[5] === 'native');

      if (lineMatch[1]) ***REMOVED***
        var methodMatch = lineMatch[1].match(/([^\.]+)(?:\.(.+))?/);
        object = methodMatch[1];
        method = methodMatch[2];
        functionName = lineMatch[1];
        typeName = 'Object';
      ***REMOVED***

      if (method) ***REMOVED***
        typeName = object;
        methodName = method;
      ***REMOVED***

      if (method === '<anonymous>') ***REMOVED***
        methodName = null;
        functionName = '';
      ***REMOVED***

      var properties = ***REMOVED***
        fileName: lineMatch[2] || null,
        lineNumber: parseInt(lineMatch[3], 10) || null,
        functionName: functionName,
        typeName: typeName,
        methodName: methodName,
        columnNumber: parseInt(lineMatch[4], 10) || null,
        'native': isNative,
      ***REMOVED***;

      return self._createParsedCallSite(properties);
    ***REMOVED***)
    .filter(function(callSite) ***REMOVED***
      return !!callSite;
    ***REMOVED***);
***REMOVED***;

exports._createParsedCallSite = function(properties) ***REMOVED***
  var methods = ***REMOVED******REMOVED***;
  for (var property in properties) ***REMOVED***
    var prefix = 'get';
    if (property === 'native') ***REMOVED***
      prefix = 'is';
    ***REMOVED***
    var method = prefix + property.substr(0, 1).toUpperCase() + property.substr(1);

    (function(property) ***REMOVED***
      methods[method] = function() ***REMOVED***
        return properties[property];
      ***REMOVED***
    ***REMOVED***)(property);
  ***REMOVED***

  var callSite = Object.create(methods);
  for (var property in properties) ***REMOVED***
    callSite[property] = properties[property];
  ***REMOVED***

  return callSite;
***REMOVED***;
