// mongoose/lib/schema/buffer.js
/*!
 * Module dependencies.
 */

var handleBitwiseOperator = require('./operators/bitwise');
var utils = require('../utils');

var MongooseBuffer = require('../types').Buffer;
var SchemaType = require('../schematype');

var Binary = MongooseBuffer.Binary;
var CastError = SchemaType.CastError;
var Document;

/**
 * Buffer SchemaType constructor
 *
 * @param ***REMOVED***String***REMOVED*** key
 * @param ***REMOVED***Object***REMOVED*** options
 * @inherits SchemaType
 * @api public
 */

function SchemaBuffer(key, options) ***REMOVED***
  SchemaType.call(this, key, options, 'Buffer');
***REMOVED***

/**
 * This schema type's name, to defend against minifiers that mangle
 * function names.
 *
 * @api public
 */
SchemaBuffer.schemaName = 'Buffer';

/*!
 * Inherits from SchemaType.
 */
SchemaBuffer.prototype = Object.create(SchemaType.prototype);
SchemaBuffer.prototype.constructor = SchemaBuffer;

/**
 * Check if the given value satisfies a required validator. To satisfy a
 * required validator, a buffer must not be null or undefined and have
 * non-zero length.
 *
 * @param ***REMOVED***Any***REMOVED*** value
 * @param ***REMOVED***Document***REMOVED*** doc
 * @return ***REMOVED***Boolean***REMOVED***
 * @api public
 */

SchemaBuffer.prototype.checkRequired = function(value, doc) ***REMOVED***
  if (SchemaType._isRef(this, value, doc, true)) ***REMOVED***
    return !!value;
  ***REMOVED***
  return !!(value && value.length);
***REMOVED***;

/**
 * Casts contents
 *
 * @param ***REMOVED***Object***REMOVED*** value
 * @param ***REMOVED***Document***REMOVED*** doc document that triggers the casting
 * @param ***REMOVED***Boolean***REMOVED*** init
 * @api private
 */

SchemaBuffer.prototype.cast = function(value, doc, init) ***REMOVED***
  var ret;
  if (SchemaType._isRef(this, value, doc, init)) ***REMOVED***
    // wait! we may need to cast this to a document

    if (value === null || value === undefined) ***REMOVED***
      return value;
    ***REMOVED***

    // lazy load
    Document || (Document = require('./../document'));

    if (value instanceof Document) ***REMOVED***
      value.$__.wasPopulated = true;
      return value;
    ***REMOVED***

    // setting a populated path
    if (Buffer.isBuffer(value)) ***REMOVED***
      return value;
    ***REMOVED*** else if (!utils.isObject(value)) ***REMOVED***
      throw new CastError('buffer', value, this.path);
    ***REMOVED***

    // Handle the case where user directly sets a populated
    // path to a plain object; cast to the Model used in
    // the population query.
    var path = doc.$__fullPath(this.path);
    var owner = doc.ownerDocument ? doc.ownerDocument() : doc;
    var pop = owner.populated(path, true);
    ret = new pop.options.model(value);
    ret.$__.wasPopulated = true;
    return ret;
  ***REMOVED***

  // documents
  if (value && value._id) ***REMOVED***
    value = value._id;
  ***REMOVED***

  if (value && value.isMongooseBuffer) ***REMOVED***
    return value;
  ***REMOVED***

  if (Buffer.isBuffer(value)) ***REMOVED***
    if (!value || !value.isMongooseBuffer) ***REMOVED***
      value = new MongooseBuffer(value, [this.path, doc]);
    ***REMOVED***

    return value;
  ***REMOVED*** else if (value instanceof Binary) ***REMOVED***
    ret = new MongooseBuffer(value.value(true), [this.path, doc]);
    if (typeof value.sub_type !== 'number') ***REMOVED***
      throw new CastError('buffer', value, this.path);
    ***REMOVED***
    ret._subtype = value.sub_type;
    return ret;
  ***REMOVED***

  if (value === null) ***REMOVED***
    return value;
  ***REMOVED***

  var type = typeof value;
  if (type === 'string' || type === 'number' || Array.isArray(value)) ***REMOVED***
    if (type === 'number') ***REMOVED***
      value = [value];
    ***REMOVED***
    ret = new MongooseBuffer(value, [this.path, doc]);
    return ret;
  ***REMOVED***

  throw new CastError('buffer', value, this.path);
***REMOVED***;

/*!
 * ignore
 */
function handleSingle(val) ***REMOVED***
  return this.castForQuery(val);
***REMOVED***

SchemaBuffer.prototype.$conditionalHandlers =
    utils.options(SchemaType.prototype.$conditionalHandlers, ***REMOVED***
      $bitsAllClear: handleBitwiseOperator,
      $bitsAnyClear: handleBitwiseOperator,
      $bitsAllSet: handleBitwiseOperator,
      $bitsAnySet: handleBitwiseOperator,
      $gt: handleSingle,
      $gte: handleSingle,
      $lt: handleSingle,
      $lte: handleSingle
    ***REMOVED***);

/**
 * Casts contents for queries.
 *
 * @param ***REMOVED***String***REMOVED*** $conditional
 * @param ***REMOVED***any***REMOVED*** [value]
 * @api private
 */

SchemaBuffer.prototype.castForQuery = function($conditional, val) ***REMOVED***
  var handler;
  if (arguments.length === 2) ***REMOVED***
    handler = this.$conditionalHandlers[$conditional];
    if (!handler) ***REMOVED***
      throw new Error('Can\'t use ' + $conditional + ' with Buffer.');
    ***REMOVED***
    return handler.call(this, val);
  ***REMOVED***
  val = $conditional;
  return this.cast(val).toObject(***REMOVED*** transform: false, virtuals: false ***REMOVED***);
***REMOVED***;

/*!
 * Module exports.
 */

module.exports = SchemaBuffer;
