// mongoose/lib/types/subdocument.js
var Document = require('../document');
var PromiseProvider = require('../promise_provider');

module.exports = Subdocument;

/**
 * Subdocument constructor.
 *
 * @inherits Document
 * @api private
 */

function Subdocument(value, fields) ***REMOVED***
  this.$isSingleNested = true;
  Document.call(this, value, fields);
***REMOVED***

Subdocument.prototype = Object.create(Document.prototype);

Subdocument.prototype.toBSON = function() ***REMOVED***
  return this.toObject(***REMOVED*** transform: false, virtuals: false ***REMOVED***);
***REMOVED***;

/**
 * Used as a stub for [hooks.js](https://github.com/bnoguchi/hooks-js/tree/31ec571cef0332e21121ee7157e0cf9728572cc3)
 *
 * ####NOTE:
 *
 * _This is a no-op. Does not actually save the doc to the db._
 *
 * @param ***REMOVED***Function***REMOVED*** [fn]
 * @return ***REMOVED***Promise***REMOVED*** resolved Promise
 * @api private
 */

Subdocument.prototype.save = function(fn) ***REMOVED***
  var Promise = PromiseProvider.get();
  return new Promise.ES6(function(resolve) ***REMOVED***
    fn && fn();
    resolve();
  ***REMOVED***);
***REMOVED***;

Subdocument.prototype.$isValid = function(path) ***REMOVED***
  if (this.$parent) ***REMOVED***
    return this.$parent.$isValid([this.$basePath, path].join('.'));
  ***REMOVED***
***REMOVED***;

Subdocument.prototype.markModified = function(path) ***REMOVED***
  Document.prototype.markModified.call(this, path);
  if (this.$parent) ***REMOVED***
    if (this.$parent.isDirectModified(this.$basePath)) ***REMOVED***
      return;
    ***REMOVED***
    this.$parent.markModified([this.$basePath, path].join('.'));
  ***REMOVED***
***REMOVED***;

Subdocument.prototype.$markValid = function(path) ***REMOVED***
  Document.prototype.$markValid.call(this, path);
  if (this.$parent) ***REMOVED***
    this.$parent.$markValid([this.$basePath, path].join('.'));
  ***REMOVED***
***REMOVED***;

Subdocument.prototype.invalidate = function(path, err, val) ***REMOVED***
  Document.prototype.invalidate.call(this, path, err, val);
  if (this.$parent) ***REMOVED***
    this.$parent.invalidate([this.$basePath, path].join('.'), err, val);
  ***REMOVED*** else if (err.kind === 'cast' || err.name === 'CastError') ***REMOVED***
    throw err;
  ***REMOVED***
***REMOVED***;

/**
 * Returns the top level document of this sub-document.
 *
 * @return ***REMOVED***Document***REMOVED***
 */

Subdocument.prototype.ownerDocument = function() ***REMOVED***
  if (this.$__.ownerDocument) ***REMOVED***
    return this.$__.ownerDocument;
  ***REMOVED***

  var parent = this.$parent;
  if (!parent) ***REMOVED***
    return this;
  ***REMOVED***

  while (parent.$parent || parent.__parent) ***REMOVED***
    parent = parent.$parent || parent.__parent;
  ***REMOVED***
  this.$__.ownerDocument = parent;
  return this.$__.ownerDocument;
***REMOVED***;

/**
 * Null-out this subdoc
 *
 * @param ***REMOVED***Object***REMOVED*** [options]
 * @param ***REMOVED***Function***REMOVED*** [callback] optional callback for compatibility with Document.prototype.remove
 */

Subdocument.prototype.remove = function(options, callback) ***REMOVED***
  if (typeof options === 'function') ***REMOVED***
    callback = options;
    options = null;
  ***REMOVED***

  // If removing entire doc, no need to remove subdoc
  if (!options || !options.noop) ***REMOVED***
    this.$parent.set(this.$basePath, null);
    registerRemoveListener(this);
  ***REMOVED***

  if (typeof callback === 'function') ***REMOVED***
    callback(null);
  ***REMOVED***
***REMOVED***;

/*!
 * ignore
 */

Subdocument.prototype.populate = function() ***REMOVED***
  throw new Error('Mongoose does not support calling populate() on nested ' +
    'docs. Instead of `doc.nested.populate("path")`, use ' +
    '`doc.populate("nested.path")`');
***REMOVED***;

/*!
 * Registers remove event listeners for triggering
 * on subdocuments.
 *
 * @param ***REMOVED***EmbeddedDocument***REMOVED*** sub
 * @api private
 */

function registerRemoveListener(sub) ***REMOVED***
  var owner = sub.ownerDocument();

  function emitRemove() ***REMOVED***
    owner.removeListener('save', emitRemove);
    owner.removeListener('remove', emitRemove);
    sub.emit('remove', sub);
    owner = sub = null;
  ***REMOVED***

  owner.on('save', emitRemove);
  owner.on('remove', emitRemove);
***REMOVED***
